{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "31158ef9-d5b8-4d34-b735-97fe829bff12",
   "metadata": {},
   "source": [
    "# Simulation Data Gene Program Inference"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89c7fb21-9b86-46c5-a174-e107a0a1221a",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (sebastian.birk@helmholtz-munich.de)\n",
    "- **Date of Creation:** 12.07.2024\n",
    "- **Date of Last Modification:** 23.08.2024 (Sebastian Birk; <sebastian.birk@helmholtz-munich.de>)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ad6ca50-25d4-45cc-806b-66bf698f81e9",
   "metadata": {},
   "source": [
    "Compare:\n",
    "- NicheCompass niche identification + GP inference\n",
    "- BANKSY niche identification + gene set enrichment analysis (only target genes) with Decoupler as described at https://www.sc-best-practices.org/conditions/gsea_pathway.html#cluster-level-gene-set-enrichment-analysis-with-decoupler\n",
    "- BANKSY niche identification + CCI inference with LIANA as described at https://liana-py.readthedocs.io/en/latest/notebooks/liana_c2c.html with NicheCompass GPs as custom resource as illustrated at https://ccc-protocols.readthedocs.io/en/latest/notebooks/ccc_python/02-Infer-Communication-Scores.html#Alternative-Resources"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ecc11b0-927a-4ee2-97e0-a66c22f7d9f1",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6318931-ea7b-48dc-b557-b1d2cc4e880c",
   "metadata": {},
   "source": [
    "Run this notebook in the nichecompass-reproducibility environment, installable from ```('../../../envs/environment.yaml')```."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a46506a-8379-4055-bdb6-c0f29347b300",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ef43669-df89-4b84-aad9-d62754496575",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58542bc8-95a1-40e8-9fde-4eb9d4de1066",
   "metadata": {},
   "outputs": [],
   "source": [
    "import decoupler\n",
    "import liana as li\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "from sklearn.metrics import f1_score\n",
    "from matplotlib import gridspec\n",
    "\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (create_new_color_dict,\n",
    "                                extract_gp_dict_from_mebocost_ms_interactions,\n",
    "                                extract_gp_dict_from_nichenet_lrt_interactions,\n",
    "                                extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                                filter_and_combine_gp_dict_gps_v2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c16ab975-57c4-45fb-af9d-79d551c78cd1",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b15a2e7-2143-4700-943e-a897bcbfb7f4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "species = \"mouse\"\n",
    "niche_type_key = \"niche_types\"\n",
    "cell_type_key = \"cell_types\"\n",
    "sample_key = \"batch\"\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "differential_gp_test_results_key = \"nichecompass_differential_gp_test_results\"\n",
    "\n",
    "spot_size = 40"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6386925-8e92-4cc2-a889-72a717e24f90",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e1ba8ba-a961-41e8-bddd-520e0dd13c04",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46237877-edd4-4631-bb89-058d67ec91cc",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5feb8575-bbbb-47f6-9829-335c16a9c95e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ga_data_folder_path = \"../../../datasets/ga_data\"\n",
    "gp_data_folder_path = \"../../../datasets/gp_data\"\n",
    "st_data_folder_path = \"../../../datasets/st_data\"\n",
    "st_data_gold_folder_path = st_data_folder_path + \"/gold\"\n",
    "omnipath_lr_network_file_path = f\"{gp_data_folder_path}/omnipath_lr_network.csv\"\n",
    "nichenet_lr_network_file_path = f\"{gp_data_folder_path}/nichenet_lr_network_v2_{species}.csv\"\n",
    "nichenet_ligand_target_matrix_file_path = f\"{gp_data_folder_path}/nichenet_ligand_target_matrix_v2_{species}.csv\"\n",
    "mebocost_enzyme_sensor_interactions_folder_path = f\"{gp_data_folder_path}/metabolite_enzyme_sensor_gps\"\n",
    "gene_orthologs_mapping_file_path = f\"{ga_data_folder_path}/human_mouse_gene_orthologs.csv\"\n",
    "artifacts_folder_path = f\"../../../artifacts\"\n",
    "figure_folder_path = f\"../../../figures\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afa901e6-016a-4d02-8ac7-4a357d813170",
   "metadata": {},
   "source": [
    "### 1.5 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d709c0cd-cb47-446a-891a-0a8b2300d6e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_f1(set1, set2):\n",
    "    # Convert sets to binary vectors\n",
    "    all_elements = list(set1.union(set2))\n",
    "    vec1 = [1 if elem in set1 else 0 for elem in all_elements]\n",
    "    vec2 = [1 if elem in set2 else 0 for elem in all_elements]\n",
    "    \n",
    "    # Compute F1 score\n",
    "    return f1_score(vec1, vec2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4a9f454-6da4-4499-bd98-cdc452ef7468",
   "metadata": {},
   "source": [
    "## 2. GP Inference\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e46adbbe-cbc3-4734-953b-dc5332885794",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 2.1 Load Cell Embeddings from BANKSY\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbfa7066-faf1-4e88-8d07-02d2b50aa1b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get embeddings from BANKSY\n",
    "adata_annotated_banksy = sc.read_h5ad(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_banksy_gpinference.h5ad\")\n",
    "\n",
    "# Get raw gene expression counts\n",
    "adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/sim1_1105genes_10000locs_strongincrements.h5ad\")\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    adata.obsm[f\"banksy_latent_run{run_number}\"] = adata_annotated_banksy.obsm[f\"banksy_latent_run{run_number}\"]\n",
    "del(adata_annotated_banksy)\n",
    "\n",
    "# Preprocess counts\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c55f4c5b-dbc9-43af-8fe2-7db85b367b6a",
   "metadata": {},
   "source": [
    "### 2.2 Get NicheCompass GPs for Use as Prior Knowledge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf82229f-7956-4607-8eda-79ee784bca24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve OmniPath GPs (source: ligand genes; target: receptor genes)\n",
    "omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "    species=species,\n",
    "    load_from_disk=True,\n",
    "    save_to_disk=False,\n",
    "    lr_network_file_path=omnipath_lr_network_file_path,\n",
    "    gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "    plot_gp_gene_count_distributions=False)\n",
    "\n",
    "# Retrieve NicheNet GPs (source: ligand genes; target: receptor genes, target genes)\n",
    "nichenet_gp_dict = extract_gp_dict_from_nichenet_lrt_interactions(\n",
    "    species=species,\n",
    "    version=\"v2\",\n",
    "    load_from_disk=True,\n",
    "    save_to_disk=False,\n",
    "    lr_network_file_path=nichenet_lr_network_file_path,\n",
    "    ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "    gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "    plot_gp_gene_count_distributions=False)\n",
    "\n",
    "# Retrieve MEBOCOST GPs (source: enzyme genes; target: sensor genes)\n",
    "mebocost_gp_dict = extract_gp_dict_from_mebocost_ms_interactions(\n",
    "    dir_path=mebocost_enzyme_sensor_interactions_folder_path,\n",
    "    species=species,\n",
    "    plot_gp_gene_count_distributions=False)\n",
    "\n",
    "# Filter and combine GPs\n",
    "gp_dicts = [omnipath_gp_dict, nichenet_gp_dict, mebocost_gp_dict]\n",
    "combined_gp_dict = filter_and_combine_gp_dict_gps_v2(\n",
    "    gp_dicts,\n",
    "    verbose=True)\n",
    "\n",
    "combined_gp_df = pd.DataFrame(combined_gp_dict.values(), index=combined_gp_dict.keys())\n",
    "\n",
    "# Filter genes to keep only genes in adata\n",
    "adata_genes = adata.var.index.tolist()\n",
    "combined_gp_df[\"sources\"] = combined_gp_df[\"sources\"].apply(lambda x: [gene for gene in x if gene in adata_genes])\n",
    "combined_gp_df[\"targets\"] = combined_gp_df[\"targets\"].apply(lambda x: [gene for gene in x if gene in adata_genes])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6f481c0-4f39-42f6-839b-83f740e05bf7",
   "metadata": {},
   "source": [
    "### 2.3 Get Simulation Ground Truth GPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "303886c5-f384-479e-a610-e4c0d26b36d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get GP ground truth stats\n",
    "increment_mode = \"strong\"\n",
    "sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "sim_gps_stats[\"prior_gps\"] = sim_gps_stats[\"prior_gps\"].apply(\n",
    "    lambda x: [gp for gp in x.split(\",\")])\n",
    "sim_gps_stats = sim_gps_stats[sim_gps_stats[\"increment_param\"] != 1]\n",
    "sim_gps_stats = sim_gps_stats.groupby(\"regionA\")[\"prior_gps\"].sum().reset_index()\n",
    "sim_gps_stats[\"prior_gps\"] = sim_gps_stats[\"prior_gps\"].apply(lambda x: {item for item in x if item not in [None, '']})\n",
    "sim_gps_stats.rename(columns={'regionA': 'niche'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c96845d1-3fd7-47a3-bd53-10f4918140c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get number of upregulated GPs per niche\n",
    "sim_gp_df = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_strongincrements.csv\")\n",
    "sim_gp_df_agg = sim_gp_df[~sim_gp_df['gp'].str.contains(\"Add-On\")][sim_gp_df['increment_param'] != 1].groupby(\"regionA\").agg(set)\n",
    "sim_gp_df_agg[\"niche_n_upregulated_prior_gps\"] = sim_gp_df_agg[\"gp\"].apply(lambda x: len(x))\n",
    "sim_gp_df_agg[\"category\"] = pd.Series(sim_gp_df_agg.index).apply(lambda x: x.replace(\"Region\", \"Niche \")).values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53d14041-b936-4548-acbc-e15ec90b811d",
   "metadata": {},
   "source": [
    "### 2.4 NicheCompass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77d7241b-30b2-4c02-ba02-37455d2da08a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Get differentially expressed genes\n",
    "f1_scores = []\n",
    "n_enriched_gps = []\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    print(f\"Starting run {run_number}...\")\n",
    "    latent_key = f\"nichecompass_latent\"\n",
    "    \n",
    "    load_timestamp = \"20082024_095419_4\" # \"20082024_100032_11\"\n",
    "    \n",
    "    model_label = \"gatv2conv_single_sample_method_benchmarking\"\n",
    "    model_folder_path = f\"{artifacts_folder_path}/sim1_1105genes_10000locs_strongincrements/models/{model_label}/{load_timestamp}/run{run_number}\"\n",
    "\n",
    "    model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                              adata=None,\n",
    "                              adata_file_name=f\"sim1_1105genes_10000locs_strongincrements_{model_label}.h5ad\",\n",
    "                              gp_names_key=gp_names_key) \n",
    "    \n",
    "    sc.pp.neighbors(model.adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    sc.tl.umap(model.adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    print(f\"Computing Leiden clusters...\")\n",
    "    latent_leiden_resolution = 0.3\n",
    "    leiden_resolution_increments = 0.1\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=model.adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = model.adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= 100].index\n",
    "        n_niches = model.adata.obs[model.adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            print(\"Big increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            print(\"Slight increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments/10\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            print(\"Big decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            print(\"Slight decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments/10\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "\n",
    "    cross_tab = pd.crosstab(model.adata.obs[\"pred_niche_types\"], model.adata.obs[\"niche_types\"])\n",
    "    majority_map = cross_tab.idxmax(axis=1)\n",
    "    mapping_dict = {}\n",
    "    for i, niche in enumerate(majority_map):\n",
    "        mapping_dict[f\"{i}\"] = niche \n",
    "    model.adata.obs[\"mapped_pred_niche_types\"] = model.adata.obs[\"pred_niche_types\"].map(mapping_dict)\n",
    "    \n",
    "    latent_cluster_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "        cat_key=\"mapped_pred_niche_types\")\n",
    "    \n",
    "    # Create plot of latent cluster / niche annotations in physical and latent space\n",
    "    groups = None # set this to a specific cluster for easy visualization\n",
    "    save_fig = False\n",
    "    file_path = \"\"\n",
    "    \n",
    "    samples = model.adata.obs[sample_key].unique().tolist()\n",
    "    \n",
    "    fig = plt.figure(figsize=(12, 14))\n",
    "    title = fig.suptitle(t=f\"NicheCompass Niches \" \\\n",
    "                           \"in Latent and Physical Space\",\n",
    "                         y=0.96,\n",
    "                         x=0.55,\n",
    "                         fontsize=20)\n",
    "    spec1 = gridspec.GridSpec(ncols=1,\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1],\n",
    "                              height_ratios=[3, 2])\n",
    "    spec2 = gridspec.GridSpec(ncols=len(samples),\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1] * len(samples),\n",
    "                              height_ratios=[3, 2])\n",
    "    axs = []\n",
    "    axs.append(fig.add_subplot(spec1[0]))\n",
    "    sc.pl.umap(adata=model.adata,\n",
    "               color=[\"mapped_pred_niche_types\"],\n",
    "               groups=groups,\n",
    "               palette=latent_cluster_colors,\n",
    "               title=f\"Niches in Latent Space\",\n",
    "               ax=axs[0],\n",
    "               show=False)\n",
    "    for idx, sample in enumerate(samples):\n",
    "        axs.append(fig.add_subplot(spec2[len(samples) + idx]))\n",
    "        sc.pl.spatial(adata=model.adata[model.adata.obs[sample_key] == sample],\n",
    "                      color=[\"mapped_pred_niche_types\"],\n",
    "                      groups=groups,\n",
    "                      palette=latent_cluster_colors,\n",
    "                      spot_size=spot_size,\n",
    "                      title=f\"Niches in Physical Space \\n\"\n",
    "                            f\"(Sample: {sample})\",\n",
    "                      legend_loc=None,\n",
    "                      ax=axs[idx+1],\n",
    "                      show=False)\n",
    "    \n",
    "    # Create and position shared legend\n",
    "    handles, labels = axs[0].get_legend_handles_labels()\n",
    "    lgd = fig.legend(handles,\n",
    "                     labels,\n",
    "                     loc=\"center left\",\n",
    "                     bbox_to_anchor=(0.98, 0.5))\n",
    "    axs[0].get_legend().remove()\n",
    "    \n",
    "    # Adjust, save and display plot\n",
    "    plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "    if save_fig:\n",
    "        fig.savefig(file_path,\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    \n",
    "    print(f\"Computing enriched GPs...\")\n",
    "    selected_cats = None\n",
    "    comparison_cats = \"rest\"\n",
    "    log_bayes_factor_thresh = 0.\n",
    "    enriched_gps = model.run_differential_gp_tests(\n",
    "        cat_key=\"mapped_pred_niche_types\",\n",
    "        selected_cats=selected_cats,\n",
    "        comparison_cats=comparison_cats,\n",
    "        log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "    \n",
    "    n_enriched_gps.append(len(enriched_gps))\n",
    "\n",
    "    # Get enriched prior and de novo GPs\n",
    "    gp_summary_df = model.get_gp_summary() \n",
    "    enriched_gps = model.adata.uns[differential_gp_test_results_key]\n",
    "    \n",
    "    # Correct GP sign\n",
    "    gp_summary_df[\"gp_sign_source\"] = gp_summary_df[\"gp_source_genes_weights\"].apply(\n",
    "        lambda x: 1 if np.array(x).sum() > 0 else (0 if np.array(x).sum() == 0 else -1))\n",
    "    gp_summary_df[\"gp_sign_target\"] = gp_summary_df[\"gp_target_genes_weights\"].apply(\n",
    "        lambda x: 1 if np.array(x).sum() > 0 else (0 if np.array(x).sum() == 0 else -1))\n",
    "    gp_summary_df[\"gp_sign\"] = gp_summary_df.apply(lambda row: row[\"gp_sign_source\"] if row[\"gp_sign_source\"] != 0 else row[\"gp_sign_target\"], axis=1)\n",
    "    gp_signs = []\n",
    "    for gp in enriched_gps[\"gene_program\"].values.tolist():\n",
    "        gp_signs.append(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_sign\"].values[0])\n",
    "    enriched_gps[\"log_bayes_factor\"] = enriched_gps[\"log_bayes_factor\"] * gp_signs\n",
    "    enriched_gps = enriched_gps[enriched_gps[\"log_bayes_factor\"] > 0]\n",
    "    \n",
    "    # Keep top n upregulated prior GPs for each niche where n matches ground truth for each niche\n",
    "    prior_enriched_gps = enriched_gps[~enriched_gps[\"gene_program\"].str.contains(\"Add-on\")]\n",
    "    prior_enriched_gps = pd.merge(prior_enriched_gps, sim_gp_df_agg, on=\"category\")\n",
    "    prior_enriched_gps = prior_enriched_gps.groupby(\"category\").apply(lambda x: x.head(x['niche_n_upregulated_prior_gps'].iloc[0])).reset_index(drop=True)\n",
    "    enriched_prior_gps_per_niche = prior_enriched_gps.groupby(\n",
    "                \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "\n",
    "    enriched_prior_gps_per_niche[\"niche\"] = enriched_prior_gps_per_niche.index.tolist()\n",
    "    niche_enriched_df = enriched_prior_gps_per_niche[[\"niche\", \"gene_program\"]]\n",
    "    niche_enriched_df[\"enriched_gps\"] = niche_enriched_df[\"gene_program\"].apply(lambda x: set(x))\n",
    "\n",
    "    # Compute F1 scores\n",
    "    merged_df = pd.merge(niche_enriched_df, sim_gps_stats, on='niche', suffixes=('_df1', '_df2'))\n",
    "    merged_df['f1_score'] = merged_df.apply(lambda row: compute_f1(row['enriched_gps'], row['prior_gps']), axis=1)\n",
    "    f1_scores.append(np.mean(merged_df[\"f1_score\"]))\n",
    "    \n",
    "f1_score_df = pd.DataFrame(f1_scores, columns=[\"f1_score\"])\n",
    "f1_score_df[\"run_number\"] = range(1,9)\n",
    "f1_score_df[\"n_enriched_gps\"] = n_enriched_gps\n",
    "f1_score_df.to_csv(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_nichecompass_f1_scores.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "838952b1-34af-4ca5-aa61-82535377bb57",
   "metadata": {},
   "source": [
    "### 2.5 Decoupler GSEA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33cf01e4-31d2-493c-a127-ad4cc0147e98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format GPs for GSEA\n",
    "gp_df_gsea = combined_gp_df.copy()\n",
    "gp_df_gsea[\"geneset\"] = gp_df_gsea.index\n",
    "gp_df_gsea[\"genesymbol\"] = gp_df_gsea[\"targets\"]\n",
    "del(gp_df_gsea[\"targets\"])\n",
    "gp_df_gsea = gp_df_gsea[[\"geneset\", \"genesymbol\"]]\n",
    "gp_df_gsea.reset_index(drop=True, inplace=True)\n",
    "gp_df_gsea = gp_df_gsea.explode(\"genesymbol\").reset_index(drop=True)\n",
    "gp_df_gsea = gp_df_gsea.dropna(subset=['genesymbol'])\n",
    "gp_df_gsea = gp_df_gsea.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8d6ad52-1e8d-4c88-8601-4c8af24ddc43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get differentially expressed genes\n",
    "f1_scores = []\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    print(f\"Starting run {run_number}...\")\n",
    "    latent_key = f\"banksy_latent_run{run_number}\"\n",
    "    sc.pp.neighbors(adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    sc.tl.umap(adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    print(f\"Computing Leiden clusters...\")\n",
    "    latent_leiden_resolution = 0.3\n",
    "    leiden_resolution_increments = 0.1\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= 100].index\n",
    "        n_niches = adata.obs[adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            print(\"Big increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            print(\"Slight increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments/10\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            print(\"Big decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            print(\"Slight decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments/10\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "\n",
    "    cross_tab = pd.crosstab(adata.obs[\"pred_niche_types\"], adata.obs[\"niche_types\"])\n",
    "    majority_map = cross_tab.idxmax(axis=1)\n",
    "    mapping_dict = {}\n",
    "    for i, niche in enumerate(majority_map):\n",
    "        mapping_dict[f\"{i}\"] = niche \n",
    "    adata.obs[\"mapped_pred_niche_types\"] = adata.obs[\"pred_niche_types\"].map(mapping_dict)\n",
    "    \n",
    "    latent_cluster_colors = create_new_color_dict(\n",
    "        adata=adata,\n",
    "        cat_key=\"mapped_pred_niche_types\")\n",
    "    \n",
    "    # Create plot of latent cluster / niche annotations in physical and latent space\n",
    "    groups = None # set this to a specific cluster for easy visualization\n",
    "    save_fig = False\n",
    "    file_path = \"\"\n",
    "    \n",
    "    samples = adata.obs[sample_key].unique().tolist()\n",
    "    \n",
    "    fig = plt.figure(figsize=(12, 14))\n",
    "    title = fig.suptitle(t=f\"BANKSY Niches \" \\\n",
    "                           \"in Latent and Physical Space\",\n",
    "                         y=0.96,\n",
    "                         x=0.55,\n",
    "                         fontsize=20)\n",
    "    spec1 = gridspec.GridSpec(ncols=1,\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1],\n",
    "                              height_ratios=[3, 2])\n",
    "    spec2 = gridspec.GridSpec(ncols=len(samples),\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1] * len(samples),\n",
    "                              height_ratios=[3, 2])\n",
    "    axs = []\n",
    "    axs.append(fig.add_subplot(spec1[0]))\n",
    "    sc.pl.umap(adata=adata,\n",
    "               color=[\"mapped_pred_niche_types\"],\n",
    "               groups=groups,\n",
    "               palette=latent_cluster_colors,\n",
    "               title=f\"Niches in Latent Space\",\n",
    "               ax=axs[0],\n",
    "               show=False)\n",
    "    for idx, sample in enumerate(samples):\n",
    "        axs.append(fig.add_subplot(spec2[len(samples) + idx]))\n",
    "        sc.pl.spatial(adata=adata[adata.obs[sample_key] == sample],\n",
    "                      color=[\"mapped_pred_niche_types\"],\n",
    "                      groups=groups,\n",
    "                      palette=latent_cluster_colors,\n",
    "                      spot_size=spot_size,\n",
    "                      title=f\"Niches in Physical Space \\n\"\n",
    "                            f\"(Sample: {sample})\",\n",
    "                      legend_loc=None,\n",
    "                      ax=axs[idx+1],\n",
    "                      show=False)\n",
    "    \n",
    "    # Create and position shared legend\n",
    "    handles, labels = axs[0].get_legend_handles_labels()\n",
    "    lgd = fig.legend(handles,\n",
    "                     labels,\n",
    "                     loc=\"center left\",\n",
    "                     bbox_to_anchor=(0.98, 0.5))\n",
    "    axs[0].get_legend().remove()\n",
    "    \n",
    "    # Adjust, save and display plot\n",
    "    plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "    if save_fig:\n",
    "        fig.savefig(file_path,\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    \n",
    "    print(f\"Computing GSEA...\")\n",
    "    sc.tl.rank_genes_groups(adata, \"mapped_pred_niche_types\", method=\"t-test\", key_added=\"t-test\")\n",
    "    \n",
    "    # Get enriched GPs per niche\n",
    "    enriched_dict = {}\n",
    "\n",
    "    for niche_type in adata.obs[\"mapped_pred_niche_types\"].unique().tolist():\n",
    "        \n",
    "        # Get number of upregulated GPs per niche\n",
    "        niche_n_upregulated_prior_gps = sim_gp_df_agg[\n",
    "            sim_gp_df_agg[\"category\"] == niche_type][\"niche_n_upregulated_prior_gps\"].values[0]\n",
    "\n",
    "        # Extract differential gene expression scores\n",
    "        t_stats = (\n",
    "            # Get dataframe of DE results for condition vs. rest\n",
    "            sc.get.rank_genes_groups_df(adata, niche_type, key=\"t-test\")\n",
    "            .set_index(\"names\")\n",
    "            # Sort by absolute score\n",
    "            .sort_values(\"scores\", key=np.abs, ascending=False)\n",
    "            # Format for decoupler\n",
    "            [[\"scores\"]]\n",
    "            .rename_axis([niche_type], axis=1)\n",
    "        )\n",
    "\n",
    "        # Get gene set enrichment analysis results\n",
    "        scores, norm, pvals = decoupler.run_gsea(\n",
    "            t_stats.T,\n",
    "            gp_df_gsea,\n",
    "            min_n=0,\n",
    "            source=\"geneset\",\n",
    "            target=\"genesymbol\",\n",
    "        )\n",
    "\n",
    "        gsea_results = (\n",
    "            pd.concat({\"score\": scores.T, \"norm\": norm.T, \"pval\": pvals.T}, axis=1)\n",
    "            .droplevel(level=1, axis=1)\n",
    "            .sort_values(\"pval\")\n",
    "        )\n",
    "\n",
    "        # Keep upregulated pathways with p-value < 0.05\n",
    "        enriched_gps = gsea_results[gsea_results[\"score\"] > 0].index.tolist()\n",
    "        print(gsea_results)\n",
    "        enriched_dict[niche_type] = enriched_gps[:niche_n_upregulated_prior_gps]\n",
    "        \n",
    "    enriched_df = pd.DataFrame(list(enriched_dict.items()), columns=[\"niche\", \"enriched_gps\"])\n",
    "    niche_enriched_df = enriched_df.groupby(\"niche\")[\"enriched_gps\"].sum().reset_index()\n",
    "    niche_enriched_df[\"enriched_gps\"] = niche_enriched_df[\"enriched_gps\"].apply(lambda x: set(x))\n",
    "\n",
    "    # Compute F1 scores\n",
    "    merged_df = pd.merge(niche_enriched_df, sim_gps_stats, on='niche', suffixes=('_df1', '_df2'))\n",
    "    merged_df['f1_score'] = merged_df.apply(lambda row: compute_f1(row['enriched_gps'], row['prior_gps']), axis=1)\n",
    "    f1_scores.append(np.mean(merged_df[\"f1_score\"]))\n",
    "    \n",
    "f1_score_df = pd.DataFrame(f1_scores, columns=[\"f1_score\"])\n",
    "f1_score_df[\"run_number\"] = range(1,9)\n",
    "f1_score_df.to_csv(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_banksy_gsea_f1_scores.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd00ed4b-e11a-480f-96f4-445977d276d9",
   "metadata": {},
   "source": [
    "### 2.6. LIANA CCI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92e48937-b513-474e-bbb9-da49990fcb08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Format GPs for LIANA\n",
    "gp_df_liana = combined_gp_df[[\"sources\", \"targets\"]].copy()\n",
    "gp_df_liana[\"ligand\"] = gp_df_liana[\"sources\"].apply(lambda x: \"_\".join(x))\n",
    "gp_df_liana[\"receptor\"] = gp_df_liana[\"targets\"].apply(lambda x: \"_\".join(x))\n",
    "del(gp_df_liana[\"sources\"])\n",
    "del(gp_df_liana[\"targets\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fc3ec14-3c0d-4714-ae88-fd8d3d180848",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter niches with only one cell type (they will cause a zero division error)\n",
    "adata = adata[~adata.obs[niche_type_key].isin([\"Niche 5\", \"Niche 6\"])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d64b3762-0480-405d-9de4-75cd23b554c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "f1_scores = []\n",
    "\n",
    "for run_number in range(1,9):\n",
    "    print(f\"Starting run {run_number}...\")\n",
    "    latent_key = f\"banksy_latent_run{run_number}\"\n",
    "    sc.pp.neighbors(adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    sc.tl.umap(adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    print(f\"Computing Leiden clusters...\")\n",
    "    latent_leiden_resolution = 0.3\n",
    "    leiden_resolution_increments = 0.1\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= 100].index\n",
    "        n_niches = adata.obs[adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            print(\"Big increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            print(\"Slight increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments/10\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            print(\"Big decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            print(\"Slight decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments/10\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "\n",
    "    cross_tab = pd.crosstab(adata.obs[\"pred_niche_types\"], adata.obs[\"niche_types\"])\n",
    "    majority_map = cross_tab.idxmax(axis=1)\n",
    "    mapping_dict = {}\n",
    "    for i, niche in enumerate(majority_map):\n",
    "        mapping_dict[f\"{i}\"] = niche \n",
    "    adata.obs[\"mapped_pred_niche_types\"] = adata.obs[\"pred_niche_types\"].map(mapping_dict)\n",
    "    \n",
    "    latent_cluster_colors = create_new_color_dict(\n",
    "        adata=adata,\n",
    "        cat_key=\"mapped_pred_niche_types\")\n",
    "    \n",
    "    # Create plot of latent cluster / niche annotations in physical and latent space\n",
    "    groups = None # set this to a specific cluster for easy visualization\n",
    "    save_fig = False\n",
    "    file_path = \"\"\n",
    "    \n",
    "    samples = adata.obs[sample_key].unique().tolist()\n",
    "    \n",
    "    fig = plt.figure(figsize=(12, 14))\n",
    "    title = fig.suptitle(t=f\"BANKSY Niches \" \\\n",
    "                           \"in Latent and Physical Space\",\n",
    "                         y=0.96,\n",
    "                         x=0.55,\n",
    "                         fontsize=20)\n",
    "    spec1 = gridspec.GridSpec(ncols=1,\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1],\n",
    "                              height_ratios=[3, 2])\n",
    "    spec2 = gridspec.GridSpec(ncols=len(samples),\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1] * len(samples),\n",
    "                              height_ratios=[3, 2])\n",
    "    axs = []\n",
    "    axs.append(fig.add_subplot(spec1[0]))\n",
    "    sc.pl.umap(adata=adata,\n",
    "               color=[\"mapped_pred_niche_types\"],\n",
    "               groups=groups,\n",
    "               palette=latent_cluster_colors,\n",
    "               title=f\"Niches in Latent Space\",\n",
    "               ax=axs[0],\n",
    "               show=False)\n",
    "    for idx, sample in enumerate(samples):\n",
    "        axs.append(fig.add_subplot(spec2[len(samples) + idx]))\n",
    "        sc.pl.spatial(adata=adata[adata.obs[sample_key] == sample],\n",
    "                      color=[\"mapped_pred_niche_types\"],\n",
    "                      groups=groups,\n",
    "                      palette=latent_cluster_colors,\n",
    "                      spot_size=spot_size,\n",
    "                      title=f\"Niches in Physical Space \\n\"\n",
    "                            f\"(Sample: {sample})\",\n",
    "                      legend_loc=None,\n",
    "                      ax=axs[idx+1],\n",
    "                      show=False)\n",
    "    \n",
    "    # Create and position shared legend\n",
    "    handles, labels = axs[0].get_legend_handles_labels()\n",
    "    lgd = fig.legend(handles,\n",
    "                     labels,\n",
    "                     loc=\"center left\",\n",
    "                     bbox_to_anchor=(0.98, 0.5))\n",
    "    axs[0].get_legend().remove()\n",
    "    \n",
    "    # Adjust, save and display plot\n",
    "    plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "    if save_fig:\n",
    "        fig.savefig(file_path,\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "    \n",
    "    # Compute CCIs with LIANA\n",
    "    li.mt.rank_aggregate.by_sample(\n",
    "        adata,\n",
    "        groupby=cell_type_key,\n",
    "        resource=gp_df_liana,\n",
    "        sample_key=\"mapped_pred_niche_types\",\n",
    "        use_raw=False,\n",
    "        verbose=True,\n",
    "        n_perms=None,\n",
    "        return_all_lrs=True,\n",
    "        )\n",
    "\n",
    "    # Keep top n results for each niche where n matches ground truth for each niche\n",
    "    liana_gp_results = adata.uns[\"liana_res\"].sort_values(\"magnitude_rank\")\n",
    "    liana_gp_results[\"category\"] = liana_gp_results[\"mapped_pred_niche_types\"]\n",
    "    liana_gp_results[\"niche\"] = liana_gp_results[\"mapped_pred_niche_types\"]\n",
    "    liana_gp_results = pd.merge(liana_gp_results, sim_gp_df_agg, on=\"category\")\n",
    "    liana_gp_results = liana_gp_results.groupby(\"category\").apply(\n",
    "        lambda x: x.head(x['niche_n_upregulated_prior_gps'].iloc[0])).reset_index(drop=True)\n",
    "\n",
    "    # Retrieve GP names based on ligand\n",
    "    liana_gps = []\n",
    "    for ligand in liana_gp_results[\"ligand_complex\"].values.tolist():\n",
    "        liana_gps.append(gp_df_liana[gp_df_liana[\"ligand\"] == ligand].index[0])\n",
    "    liana_gp_results[\"gene_program\"] = liana_gps\n",
    "\n",
    "    # Group GPs by niche\n",
    "    liana_gp_results = liana_gp_results[[\"niche\", \"gene_program\"]]\n",
    "    niche_liana_df = liana_gp_results.groupby(\"niche\").agg(set)\n",
    "\n",
    "    # Add back missing niches\n",
    "    pure_niches_df = pd.DataFrame({\n",
    "        \"niche\": [\"Niche 5\", \"Niche 6\"],\n",
    "        \"gene_program\": [set(), set()]\n",
    "    }).set_index(\"niche\")\n",
    "    niche_liana_df = pd.concat([niche_liana_df, pure_niches_df], axis=0)\n",
    "\n",
    "    # Compute F1 scores\n",
    "    merged_df = pd.merge(niche_liana_df, sim_gps_stats, on='niche', suffixes=('_df1', '_df2'))\n",
    "    merged_df['f1_score'] = merged_df.apply(lambda row: compute_f1(row['gene_program'], row['prior_gps']), axis=1)\n",
    "    f1_scores.append(np.mean(merged_df[\"f1_score\"]))\n",
    "    \n",
    "f1_score_df = pd.DataFrame(f1_scores, columns=[\"f1_score\"])\n",
    "f1_score_df[\"run_number\"] = range(1,9)\n",
    "f1_score_df.to_csv(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_banksy_liana_f1_scores.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad9d59bb-7113-42d9-a1ee-678d6fa4bb05",
   "metadata": {},
   "source": [
    "### 2.7. Visualize Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fc0d765-ce99-4d57-b3a3-bd0a7fa45902",
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_gsea_df = pd.read_csv(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_banksy_gsea_f1_scores.csv\")\n",
    "banksy_liana_df = pd.read_csv(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_banksy_liana_f1_scores.csv\")\n",
    "nichecompass_df = pd.read_csv(f\"{artifacts_folder_path}/single_sample_method_benchmarking/sim1_1105genes_10000locs_strongincrements_nichecompass_f1_scores.csv\")\n",
    "\n",
    "banksy_gsea_df[\"method\"] = \"BANKSY + GSEA\"\n",
    "banksy_liana_df[\"method\"] = \"BANKSY + LIANA\"\n",
    "nichecompass_df[\"method\"] = \"NicheCompass\"\n",
    "summary_df = pd.concat([banksy_gsea_df, banksy_liana_df, nichecompass_df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00bc8714-9a06-4a5c-a526-fd6dd1f9c68f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extended Data Fig. 3f ###\n",
    "plt.figure(figsize=(9, 6))\n",
    "sns.boxplot(x=\"method\", y=\"f1_score\", data=summary_df)\n",
    "plt.xlabel(\"Method\", fontsize=16)\n",
    "plt.ylabel(\"F1 Score\", fontsize=16)\n",
    "plt.title(\"F1 Prior GPs\", fontsize=16)\n",
    "os.makedirs(f\"{figure_folder_path}/single_sample_method_benchmarking\", exist_ok=True)\n",
    "plt.savefig(f\"{figure_folder_path}/single_sample_method_benchmarking/f1_scores_method_comparison.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82a54896-d348-4172-b7e9-be92aad569e9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
