{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# Autotalker Hyperparameter Tuning"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-López Lab\n",
    "- **Date of Creation:** 07.12.2022\n",
    "- **Date of Last Modification:** 10.01.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../../autotalker\") # 'autotalker' repo https://github.com/sebastianbirk/autotalker.git needs to be cloned into same folder as 'autotalker-reproducibility'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "\n",
    "from autotalker.models import Autotalker\n",
    "from autotalker.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                              extract_gp_dict_from_mebocost_es_interactions,\n",
    "                              extract_gp_dict_from_nichenet_ligand_target_mx,\n",
    "                              extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                              filter_and_combine_gp_dict_gps)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Model\n",
    "# Dataset\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo1\"\n",
    "\n",
    "# Anndata keys\n",
    "counts_key = \"counts\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "adj_key = \"spatial_connectivities\"\n",
    "spatial_key = \"spatial\"\n",
    "gp_names_key = \"autotalker_gp_names\"\n",
    "active_gp_names_key = \"autotalker_active_gp_names\"\n",
    "gp_targets_mask_key = \"autotalker_gp_targets\"\n",
    "gp_sources_mask_key = \"autotalker_gp_sources\"\n",
    "latent_key = \"autotalker_latent\"\n",
    "\n",
    "## Others\n",
    "random_seed = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f6cf198-5909-4c1e-9805-5e494a63412d",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3556ca35-3edf-4c42-84b3-2fe58d5fb2f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_folder_path = \"../figures\"\n",
    "model_artifacts_folder_path = \"../model_artifacts\"\n",
    "gp_data_folder_path = \"../datasets/gp_data\" # gene program data\n",
    "srt_data_folder_path = \"../datasets/srt_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "nichenet_ligand_target_mx_file_path = gp_data_folder_path + \"/nichenet_ligand_target_matrix.csv\"\n",
    "omnipath_lr_interactions_file_path = gp_data_folder_path + \"/omnipath_lr_interactions.csv\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(model_artifacts_folder_path, exist_ok=True)\n",
    "os.makedirs(gp_data_folder_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc61ed2d-6ed4-45fa-8e74-b919d74000b5",
   "metadata": {},
   "source": [
    "### 1.5 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cfa64ac5-8032-4aea-b36d-4379ffd38e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_hyperparam_tuning(dataset,\n",
    "                          hyperparam_option_dict,\n",
    "                          n_iters,\n",
    "                          experiment_name,\n",
    "                          gp_mask=\"combined_priors\",\n",
    "                          save_figures=False,\n",
    "                          save_model=False):\n",
    "    # Read dataset\n",
    "    print(\"--- DATASET ---\")\n",
    "    print(f\"Using dataset {dataset}.\")\n",
    "    adata = ad.read_h5ad(f\"{srt_data_gold_folder_path}/{dataset}.h5ad\")\n",
    "    n_nodes = adata.layers[\"counts\"].shape[0]\n",
    "    n_genes = adata.layers[\"counts\"].shape[1]\n",
    "    print(f\"Number of nodes (cells): {n_nodes}.\")\n",
    "    print(f\"Number of node features (genes): {n_genes}.\")\n",
    "        \n",
    "    # Retrieve gene program mask\n",
    "    print(\"\")\n",
    "    print(\"--- GENE PROGRAMS ---\")\n",
    "    print(f\"Using '{gp_mask}' GP mask.\")\n",
    "    \n",
    "    if gp_mask == \"fc\":\n",
    "        n_output = len(adata.var) * 2\n",
    "        adata.varm[gp_targets_mask_key] = np.ones((len(adata.var), len(adata.var)))\n",
    "        adata.varm[gp_sources_mask_key] = np.ones((len(adata.var), len(adata.var)))\n",
    "        adata.uns[gp_names_key] = np.array([f\"FC_GP_{i}\" for i in range(len(adata.var))])\n",
    "    \n",
    "    elif gp_mask == \"combined_priors\":\n",
    "        nichenet_keep_target_ratio = 0.01\n",
    "        omnipath_min_curation_effort = 0\n",
    "        gp_filter_mode = \"subset\"\n",
    "        combine_overlap_gps = True\n",
    "        overlap_thresh_source_genes = 0.9\n",
    "        overlap_thresh_target_genes = 0.9\n",
    "        overlap_thresh_genes = 0.9\n",
    "        min_genes_per_gp = 1\n",
    "        min_source_genes_per_gp = 0\n",
    "        min_target_genes_per_gp = 0\n",
    "        \n",
    "        nichenet_gp_dict = extract_gp_dict_from_nichenet_ligand_target_mx(\n",
    "            keep_target_ratio=nichenet_keep_target_ratio,\n",
    "            load_from_disk=False,\n",
    "            save_to_disk=False,\n",
    "            file_path=nichenet_ligand_target_mx_file_path)\n",
    "        omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "            min_curation_effort=omnipath_min_curation_effort,\n",
    "            load_from_disk=False,\n",
    "            save_to_disk=False,\n",
    "            file_path=omnipath_lr_interactions_file_path)\n",
    "        mebocost_gp_dict = extract_gp_dict_from_mebocost_es_interactions(\n",
    "            dir_path = f\"{gp_data_folder_path}/metabolite_enzyme_sensor_gps/\",\n",
    "            species=\"mouse\",\n",
    "            genes_uppercase=True)\n",
    "        combined_gp_dict = dict(nichenet_gp_dict)\n",
    "        combined_gp_dict.update(omnipath_gp_dict)\n",
    "        combined_gp_dict.update(mebocost_gp_dict)\n",
    "\n",
    "        # Filter and combine gene programs\n",
    "        combined_new_gp_dict = filter_and_combine_gp_dict_gps(\n",
    "            gp_dict=combined_gp_dict,\n",
    "            gp_filter_mode=gp_filter_mode,\n",
    "            combine_overlap_gps=combine_overlap_gps,\n",
    "            overlap_thresh_source_genes=overlap_thresh_source_genes,\n",
    "            overlap_thresh_target_genes=overlap_thresh_target_genes,\n",
    "            overlap_thresh_genes=overlap_thresh_genes,\n",
    "            verbose=True)\n",
    "\n",
    "        print(f\"Number of gene programs before filtering and combining: {len(combined_gp_dict)}.\")\n",
    "        print(f\"Number of gene programs after filtering and combining: {len(combined_new_gp_dict)}.\")\n",
    "        \n",
    "        # Add the gene program dictionary as binary masks to the adata for model training\n",
    "        add_gps_from_gp_dict_to_adata(\n",
    "            gp_dict=combined_new_gp_dict,\n",
    "            adata=adata,\n",
    "            genes_uppercase=True,\n",
    "            gp_targets_mask_key=gp_targets_mask_key,\n",
    "            gp_sources_mask_key=gp_sources_mask_key,\n",
    "            gp_names_key=gp_names_key,\n",
    "            min_genes_per_gp=min_genes_per_gp,\n",
    "            min_source_genes_per_gp=min_source_genes_per_gp,\n",
    "            min_target_genes_per_gp=min_target_genes_per_gp,\n",
    "            max_genes_per_gp=None,\n",
    "            max_source_genes_per_gp=None,\n",
    "            max_target_genes_per_gp=None)\n",
    "        \n",
    "    # Summarize gene programs\n",
    "    print(f\"Number of gene programs with probed genes: {len(adata.uns['autotalker_gp_names'])}.\")\n",
    "    print(f\"Example gene programs: {random.sample(list(adata.uns['autotalker_gp_names']), 5)}.\")\n",
    "    print(f\"Number of gene program target genes: {adata.varm['autotalker_gp_targets'].sum()}.\")\n",
    "    print(f\"Number of gene program source genes: {adata.varm['autotalker_gp_sources'].sum()}.\")\n",
    "\n",
    "    # Loop `n_iters` times through combination of hyperparams \n",
    "    iters = range(n_iters)\n",
    "    hyperparams, hyperparam_values = zip(*hyperparam_option_dict.items())\n",
    "    for hyperparam_comb in itertools.product(*hyperparam_values, iters):\n",
    "        hyperparam_dict = {}\n",
    "        for hyperparam, hyperparam_value in zip(hyperparams, hyperparam_comb):\n",
    "            hyperparam_dict[hyperparam] = hyperparam_value\n",
    "        \n",
    "        experiment = mlflow.set_experiment(experiment_name)\n",
    "            \n",
    "        # Log dataset params\n",
    "        mlflow.log_param(\"dataset\", dataset)\n",
    "        mlflow.log_param(\"n_nodes\", n_nodes)\n",
    "        mlflow.log_param(\"n_genes\", n_genes)\n",
    "    \n",
    "        # Log gp mask params\n",
    "        mlflow.log_param(\"gp_mask\", gp_mask)\n",
    "        mlflow.log_param(\"nichenet_keep_target_ratio\", nichenet_keep_target_ratio)\n",
    "        mlflow.log_param(\"omnipath_min_curation_effort\", omnipath_min_curation_effort)\n",
    "        mlflow.log_param(\"gp_filter_mode\", gp_filter_mode)\n",
    "        mlflow.log_param(\"combine_overlap_gps\", combine_overlap_gps)\n",
    "        mlflow.log_param(\"overlap_thresh_source_genes\", overlap_thresh_source_genes)\n",
    "        mlflow.log_param(\"overlap_thresh_target_genes\", overlap_thresh_target_genes)\n",
    "        mlflow.log_param(\"overlap_thresh_genes\", overlap_thresh_genes)\n",
    "        mlflow.log_param(\"min_genes_per_gp\", min_genes_per_gp)\n",
    "        mlflow.log_param(\"min_source_genes_per_gp\", min_source_genes_per_gp)\n",
    "        mlflow.log_param(\"min_target_genes_per_gp\", min_target_genes_per_gp) \n",
    "        \n",
    "        if save_figures or save_model:\n",
    "            # Get time for timestamping saved artifacts\n",
    "            now = datetime.now()\n",
    "            current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")\n",
    "\n",
    "            run_figure_folder_path = f\"{figure_folder_path}/{dataset}/hyperparam_tuning/{experiment_name}/runs/{current_timestamp}\"\n",
    "            os.makedirs(run_figure_folder_path, exist_ok=True)\n",
    "            run_model_artifacts_folder_path = f\"{model_artifacts_folder_path}/{dataset}/hyperparam_tuning/{experiment_name}/runs/{current_timestamp}\"\n",
    "            os.makedirs(run_model_artifacts_folder_path, exist_ok=True)\n",
    "        \n",
    "        # Copy adata\n",
    "        adata_run = adata.copy()\n",
    "\n",
    "        print(\"\")\n",
    "        print(\"--- SPATIAL CONNECTIVITY ---\")\n",
    "        # Compute spatial neighborhood\n",
    "        sq.gr.spatial_neighbors(adata_run,\n",
    "                                coord_type=\"generic\",\n",
    "                                spatial_key=spatial_key,\n",
    "                                radius=hyperparam_dict[\"radius\"])\n",
    "        avg_edges_per_node = round(adata_run.obsp['spatial_connectivities'].sum(axis=0).mean(), 2)\n",
    "        print(f\"Average number of edges per node: {avg_edges_per_node}.\")\n",
    "        n_edges = int(sp.triu(adata_run.obsp['spatial_connectivities'], k=1).sum())\n",
    "        print(f\"Number of total edges: {n_edges}.\")\n",
    "        mlflow.log_param(\"avg_n_neighbors\", avg_edges_per_node)\n",
    "        mlflow.log_param(\"n_edges\", n_edges)\n",
    "        avg_n_neighbors_rounded = int(round(avg_edges_per_node))\n",
    "\n",
    "        # Initialize model\n",
    "        print(\"\")\n",
    "        if hyperparam_dict[\"n_hidden_encoder_divisor\"] is None:\n",
    "            n_hidden_encoder = len(adata.uns[gp_names_key]) # gp mask size\n",
    "        else:\n",
    "            n_hidden_encoder = int(len(adata.var) / hyperparam_dict[\"n_hidden_encoder_divisor\"])\n",
    "        model = Autotalker(adata_run,\n",
    "                           counts_key=counts_key,\n",
    "                           adj_key=adj_key,\n",
    "                           gp_names_key=gp_names_key,\n",
    "                           active_gp_names_key=active_gp_names_key,\n",
    "                           gp_targets_mask_key=gp_targets_mask_key,\n",
    "                           gp_sources_mask_key=gp_sources_mask_key,\n",
    "                           latent_key=latent_key,\n",
    "                           gene_expr_recon_dist=hyperparam_dict[\"gene_expr_recon_dist\"],\n",
    "                           node_label_method=hyperparam_dict[\"node_label_method\"],\n",
    "                           active_gp_thresh_ratio=hyperparam_dict[\"active_gp_thresh_ratio\"],\n",
    "                           n_hidden_encoder=n_hidden_encoder,\n",
    "                           conv_layer_encoder=hyperparam_dict[\"conv_layer_encoder\"],\n",
    "                           encoder_n_attention_heads=hyperparam_dict[\"encoder_n_attention_heads\"])\n",
    "\n",
    "        # Train model\n",
    "        print(\"\")\n",
    "        model.train(n_epochs=hyperparam_dict[\"n_epochs\"],\n",
    "                    n_epochs_all_gps=hyperparam_dict[\"n_epochs_all_gps\"],\n",
    "                    lr=hyperparam_dict[\"lr\"],\n",
    "                    lambda_edge_recon=hyperparam_dict[\"lambda_edge_recon\"],\n",
    "                    lambda_gene_expr_recon=hyperparam_dict[\"lambda_gene_expr_recon\"],\n",
    "                    lambda_group_lasso=hyperparam_dict[\"lambda_group_lasso\"],\n",
    "                    edge_val_ratio=0.1,\n",
    "                    node_val_ratio=0.1,\n",
    "                    edge_batch_size=hyperparam_dict[\"batch_size\"],\n",
    "                    node_batch_size=hyperparam_dict[\"batch_size\"],\n",
    "                    mlflow_experiment_id=experiment.experiment_id,\n",
    "                    verbose=True,\n",
    "                    seed=random.randint(0, 10000))\n",
    "        \n",
    "        if save_figures:\n",
    "            # Log visualizations\n",
    "            # Use autotalker latent space for UMAP generation\n",
    "            sc.pp.neighbors(model.adata, use_rep=latent_key)\n",
    "            sc.tl.umap(model.adata, min_dist=0.3)\n",
    "\n",
    "            # Create subplot of cell-type annotations in spatial and latent space\n",
    "            fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(10, 29))\n",
    "            title = fig.suptitle(t=\"Cell-type Annotations in Spatial and Latent Space\",\n",
    "                                 x=0.85,\n",
    "                                 y=0.925,\n",
    "                                 fontsize=20)\n",
    "            sc.pl.spatial(adata=model.adata,\n",
    "                          color=[cell_type_key],\n",
    "                          spot_size=0.03,\n",
    "                          title=\"Spatial Space\",\n",
    "                          ax=axs[0],\n",
    "                          show=False)\n",
    "            sc.pl.umap(adata=model.adata,\n",
    "                       color=[cell_type_key],\n",
    "                       title=\"Latent Space\",\n",
    "                       ax=axs[1],\n",
    "                       show=False)\n",
    "            handles, labels = axs[0].get_legend_handles_labels()\n",
    "            lgd = fig.legend(handles, labels, bbox_to_anchor=(1.225, 0.625))\n",
    "            axs[0].get_legend().remove()\n",
    "            axs[1].get_legend().remove()\n",
    "            plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "            fig.savefig(f\"{run_figure_folder_path}/cell_type_annotations_in_spatial_and_latent_space.png\",\n",
    "                        bbox_extra_artists=(lgd, title),\n",
    "                        bbox_inches=\"tight\")\n",
    "            plt.close(fig)\n",
    "\n",
    "            # Create subplot of latent Leiden cluster annotations in latent and spatial space\n",
    "            resolution = 0.3\n",
    "            model.compute_latent_graph_connectivities(adata=model.adata,\n",
    "                                                      n_neighbors=int(round(avg_edges_per_node)),\n",
    "                                                      mode=\"knn\",\n",
    "                                                      seed=random_seed)\n",
    "            sc.tl.leiden(adata=model.adata,\n",
    "                         resolution=resolution,\n",
    "                         random_state=random_seed,\n",
    "                         key_added=f\"leiden_latent_{str(resolution)}\",\n",
    "                         adjacency=model.adata.obsp[\"latent_connectivities\"])\n",
    "            fig, axs = plt.subplots(nrows=2, ncols=1, figsize=(10, 20))\n",
    "            title = fig.suptitle(t=\"Latent Leiden Cluster Annotations in Latent and Spatial Space\",\n",
    "                                 x=0.85,\n",
    "                                 y=0.925,\n",
    "                                 fontsize=20)\n",
    "            sc.pl.umap(adata=model.adata,\n",
    "                       color=[f\"leiden_latent_{str(resolution)}\"],\n",
    "                       title=f\"Latent Space\",\n",
    "                       ax=axs[0],\n",
    "                       show=False)\n",
    "            sc.pl.spatial(adata=model.adata,\n",
    "                          color=[f\"leiden_latent_{str(resolution)}\"],\n",
    "                          spot_size=0.03,\n",
    "                          title=f\"Spatial Space\",\n",
    "                          ax=axs[1],\n",
    "                          show=False)\n",
    "            handles, labels = axs[0].get_legend_handles_labels()\n",
    "            lgd = fig.legend(handles, labels, bbox_to_anchor=(1.025, 0.7))\n",
    "            axs[0].get_legend().remove()\n",
    "            axs[1].get_legend().remove()\n",
    "            plt.subplots_adjust(wspace=0, hspace=0.1)\n",
    "            fig.savefig(f\"{run_figure_folder_path}/latent_leiden_cluster_annotations_in_latent_and_spatial_space.png\",\n",
    "                        bbox_extra_artists=(lgd, title),\n",
    "                        bbox_inches=\"tight\")\n",
    "            plt.close(fig)\n",
    "\n",
    "            mlflow.log_artifacts(run_figure_folder_path,\n",
    "                                 current_timestamp)\n",
    "        \n",
    "        if save_model:            \n",
    "            model.save(dir_path=run_model_artifacts_folder_path,\n",
    "                       overwrite=True,\n",
    "                       save_adata=True,\n",
    "                       adata_file_name=f\"{dataset}.h5ad\")\n",
    "\n",
    "        mlflow.end_run()\n",
    "        print(\"--------------------\")\n",
    "        print(\"\")\n",
    "        print(\"--------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b089dc00-0251-4131-861c-308489b66790",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_hyperparam_benchmarking_metrics(fig_title,\n",
    "                                         df,\n",
    "                                         y_col_name,\n",
    "                                         sort_col=\"val_mse_score\",\n",
    "                                         plot_ratio_active_gps=False,\n",
    "                                         save_fig=False,\n",
    "                                         file_name=\"benchmarking_metrics.png\"):\n",
    "    # Compute evaluation metric ranks for sorting\n",
    "    val_mse_score_ranks = (df.groupby([y_col_name])\n",
    "                           [\"val_mse_score\"].mean().rank(ascending=True)\n",
    "                           .rename(\"val_mse_score_rank\"))\n",
    "    val_auroc_score_ranks = (df.groupby([y_col_name])\n",
    "                                [\"val_auroc_score\"].mean().rank(ascending=False)\n",
    "                                .rename(\"val_auroc_score_rank\"))\n",
    "    df = df.merge(val_mse_score_ranks, on=[y_col_name])\n",
    "    df = df.merge(val_auroc_score_ranks, on=[y_col_name])\n",
    "\n",
    "    if sort_col == \"val_mse_score\":\n",
    "        df.sort_values(by=[\"val_mse_score_rank\", y_col_name],\n",
    "                       inplace=True,\n",
    "                       ascending=False)\n",
    "    elif sort_col == \"val_auroc_score\":\n",
    "        df.sort_values(by=[\"val_auroc_score_rank\", y_col_name],\n",
    "                       inplace=True,\n",
    "                       ascending=False)\n",
    "    elif sort_col == \"total_score\":\n",
    "        df[\"val_total_score_rank\"] = (df[\"val_mse_score_rank\"] + \n",
    "                                      df[\"val_auroc_score_rank\"]).rank(ascending=True)\n",
    "        df.sort_values(by=[\"val_total_score_rank\", y_col_name],\n",
    "                       inplace=True,\n",
    "                       ascending=False)\n",
    "    \n",
    "    if plot_ratio_active_gps:\n",
    "        fig, axes = plt.subplots(3, 1, sharey=True, figsize=(10, 20))\n",
    "        sns.boxplot(data=df, ax=axes[2], x=\"ratio_active_gps\", y=y_col_name)\n",
    "        axes[2].set_title(\"Ratio of Active Gene Programs\")\n",
    "    else:\n",
    "        fig, axes = plt.subplots(2, 1, sharey=True, figsize=(10, 15))\n",
    "    fig.suptitle(fig_title, fontsize=15)\n",
    "    sns.boxplot(data=df, ax=axes[0], x=\"val_auroc_score\", y=y_col_name)\n",
    "    axes[0].set_title(\"Edge Reconstruction Area Under ROC Curve\")\n",
    "    sns.boxplot(data=df, ax=axes[1], x=\"val_mse_score\", y=y_col_name)\n",
    "    axes[1].set_title(\"Gene Expression Reconstruction Mean Squared Error\")\n",
    "    plt.subplots_adjust(left=0.1,\n",
    "                        bottom=0.1,\n",
    "                        right=0.9,\n",
    "                        top=0.94,\n",
    "                        wspace=0.175,\n",
    "                        hspace=0.175)\n",
    "    if save_fig:\n",
    "        # Get time for timestamping saved artefacts\n",
    "        now = datetime.now()\n",
    "        current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")\n",
    "        benchmark_fig_run_dir = f\"{figure_folder_path}/{dataset}/benchmarking/{experiment_name}/results/{current_timestamp}\"\n",
    "        os.makedirs(benchmark_fig_run_dir, exist_ok=True)\n",
    "        plt.savefig(f\"{benchmark_fig_run_dir}/{file_name}\",\n",
    "                    bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7332dc1-413d-4e64-9161-38115e91c450",
   "metadata": {},
   "source": [
    "## 2. Hyperparameter Tuning Experiments"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "864a28bf-fb6f-4442-a45c-0704350dd11a",
   "metadata": {},
   "source": [
    "### 2.1. Loss Weights"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3532074-40e0-4d92-a4f0-3fa32f25db52",
   "metadata": {},
   "source": [
    "- Different combinations of the edge reconstruction loss and gene expression reconstruction loss weighting hyperparameters are tested.\n",
    "- Average number of neighbors is varied. The radius is set in order to get an average number of neighbors of ```2```, ```4```, ```8```, ```16``` and ```32``` with 2 iterations respectively, resulting in a total of 10 runs per loss weight combination."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "94f8b40b-6435-421c-88fc-0d265f3ad3b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_name = \"loss_weights\"\n",
    "\n",
    "# Set hyperparameters that are not varied as part of the experiment\n",
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0]\n",
    "hyperparam_option_dict[\"n_hidden_encoder_divisor\"] = [None]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "\n",
    "hyperparam_option_dict[\"n_epochs\"] = [20]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [10]\n",
    "hyperparam_option_dict[\"lr\"] = [0.001]\n",
    "hyperparam_option_dict[\"batch_size\"] = [64]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fa55161-9831-4c17-9731-cc897a551e1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- DATASET ---\n",
      "Using dataset seqfish_mouse_organogenesis_embryo1.\n",
      "Number of nodes (cells): 19451.\n",
      "Number of node features (genes): 351.\n",
      "\n",
      "--- GENE PROGRAMS ---\n",
      "Using 'combined_priors' GP mask.\n",
      "Downloading NicheNet ligand target potential matrix from the web. This might take a while...\n",
      "Removing GP 'CTF1_ligand_receptor_GP' as it is a subset of GP 'CTF1_ligand_targetgenes_GP'.\n",
      "Removing GP 'LEP_ligand_receptor_GP' as it is a subset of GP 'LEP_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL4_ligand_receptor_GP' as it is a subset of GP 'IL4_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL12A_ligand_receptor_GP' as it is a subset of GP 'IL12A_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL23A_ligand_receptor_GP' as it is a subset of GP 'IL23A_ligand_targetgenes_GP'.\n",
      "Removing GP 'CSF2_ligand_receptor_GP' as it is a subset of GP 'CSF2_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL21_ligand_receptor_GP' as it is a subset of GP 'IL21_ligand_targetgenes_GP'.\n",
      "Removing GP 'TSLP_ligand_receptor_GP' as it is a subset of GP 'TSLP_ligand_targetgenes_GP'.\n",
      "Removing GP 'GH2_ligand_receptor_GP' as it is a subset of GP 'GH2_ligand_targetgenes_GP'.\n",
      "Removing GP 'EGF_ligand_receptor_GP' as it is a subset of GP 'EGF_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL10_ligand_receptor_GP' as it is a subset of GP 'IL10_ligand_targetgenes_GP'.\n",
      "Removing GP 'TNFSF4_ligand_receptor_GP' as it is a subset of GP 'TNFSF4_ligand_targetgenes_GP'.\n",
      "Removing GP 'TNFSF18_ligand_receptor_GP' as it is a subset of GP 'TNFSF18_ligand_targetgenes_GP'.\n",
      "Removing GP 'TGFB1_ligand_receptor_GP' as it is a subset of GP 'TGFB1_ligand_targetgenes_GP'.\n",
      "Removing GP 'BMP2_ligand_receptor_GP' as it is a subset of GP 'BMP2_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL1B_ligand_receptor_GP' as it is a subset of GP 'IL1B_ligand_targetgenes_GP'.\n",
      "Removing GP 'CSH1_ligand_receptor_GP' as it is a subset of GP 'CSH1_ligand_targetgenes_GP'.\n",
      "Removing GP 'EDN1_ligand_receptor_GP' as it is a subset of GP 'EDN1_ligand_targetgenes_GP'.\n",
      "Removing GP 'EDN3_ligand_receptor_GP' as it is a subset of GP 'EDN3_ligand_targetgenes_GP'.\n",
      "Removing GP 'GCG_ligand_receptor_GP' as it is a subset of GP 'GCG_ligand_targetgenes_GP'.\n",
      "Removing GP 'HCRT_ligand_receptor_GP' as it is a subset of GP 'HCRT_ligand_targetgenes_GP'.\n",
      "Removing GP 'IGF1_ligand_receptor_GP' as it is a subset of GP 'IGF1_ligand_targetgenes_GP'.\n",
      "Removing GP 'GRP_ligand_receptor_GP' as it is a subset of GP 'GRP_ligand_targetgenes_GP'.\n",
      "Removing GP 'NTF3_ligand_receptor_GP' as it is a subset of GP 'NTF3_ligand_targetgenes_GP'.\n",
      "Removing GP 'AVP_ligand_receptor_GP' as it is a subset of GP 'AVP_ligand_targetgenes_GP'.\n",
      "Removing GP 'CSHL1_ligand_receptor_GP' as it is a subset of GP 'CSHL1_ligand_targetgenes_GP'.\n",
      "Removing GP 'PROK1_ligand_receptor_GP' as it is a subset of GP 'PROK1_ligand_targetgenes_GP'.\n",
      "Removing GP 'AGT_ligand_receptor_GP' as it is a subset of GP 'AGT_ligand_targetgenes_GP'.\n",
      "Removing GP 'GHRH_ligand_receptor_GP' as it is a subset of GP 'GHRH_ligand_targetgenes_GP'.\n",
      "Removing GP 'DLL3_ligand_receptor_GP' as it is a subset of GP 'DLL3_ligand_targetgenes_GP'.\n",
      "Removing GP 'DLL4_ligand_receptor_GP' as it is a subset of GP 'DLL4_ligand_targetgenes_GP'.\n",
      "Removing GP 'FGF23_ligand_receptor_GP' as it is a subset of GP 'FGF23_ligand_targetgenes_GP'.\n",
      "Removing GP 'GDF6_ligand_receptor_GP' as it is a subset of GP 'GDF6_ligand_targetgenes_GP'.\n",
      "Removing GP 'GDF9_ligand_receptor_GP' as it is a subset of GP 'GDF9_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL12B_ligand_receptor_GP' as it is a subset of GP 'IL12B_ligand_targetgenes_GP'.\n",
      "Removing GP 'JAG1_ligand_receptor_GP' as it is a subset of GP 'JAG1_ligand_targetgenes_GP'.\n",
      "Removing GP 'CRLF1_ligand_receptor_GP' as it is a subset of GP 'CRLF1_ligand_targetgenes_GP'.\n",
      "Removing GP 'VGF_ligand_receptor_GP' as it is a subset of GP 'VGF_ligand_targetgenes_GP'.\n",
      "Removing GP 'IL27_ligand_receptor_GP' as it is a subset of GP 'IL27_ligand_targetgenes_GP'.\n",
      "Removing GP 'CMTM8_ligand_receptor_GP' as it is a subset of GP 'CMTM8_ligand_targetgenes_GP'.\n",
      "Removing GP 'SECTM1_ligand_receptor_GP' as it is a subset of GP 'SECTM1_ligand_targetgenes_GP'.\n",
      "Removing GP 'CMTM2_ligand_receptor_GP' as it is a subset of GP 'CMTM2_ligand_targetgenes_GP'.\n",
      "Removing GP 'BMP8A_ligand_receptor_GP' as it is a subset of GP 'BMP8A_ligand_targetgenes_GP'.\n",
      "Removing GP 'GDF3_ligand_receptor_GP' as it is a subset of GP 'GDF3_ligand_targetgenes_GP'.\n",
      "Removing GP 'EFNB3_ligand_targetgenes_GP' as it is a subset of GP 'EFNB3_ligand_receptor_GP'.\n",
      "Removing GP 'EFNB2_ligand_targetgenes_GP' as it is a subset of GP 'EFNB2_ligand_receptor_GP'.\n",
      "Removing GP 'EFNB1_ligand_targetgenes_GP' as it is a subset of GP 'EFNB1_ligand_receptor_GP'.\n",
      "Removing GP 'CCL23_ligand_targetgenes_GP' as it is a subset of GP 'CCL23_ligand_receptor_GP'.\n",
      "Removing GP 'CCL27_ligand_targetgenes_GP' as it is a subset of GP 'CCL27_ligand_receptor_GP'.\n",
      "Removing GP 'CCL28_ligand_targetgenes_GP' as it is a subset of GP 'CCL28_ligand_receptor_GP'.\n",
      "Removing GP 'CCL16_ligand_targetgenes_GP' as it is a subset of GP 'CCL16_ligand_receptor_GP'.\n",
      "Removing GP 'CXCL5_ligand_targetgenes_GP' as it is a subset of GP 'CXCL5_ligand_receptor_GP'.\n",
      "Removing GP 'PMCH_ligand_targetgenes_GP' as it is a subset of GP 'PMCH_ligand_receptor_GP'.\n",
      "Removing GP 'NMS_ligand_targetgenes_GP' as it is a subset of GP 'NMS_ligand_receptor_GP'.\n",
      "Removing GP 'NMU_ligand_targetgenes_GP' as it is a subset of GP 'NMU_ligand_receptor_GP'.\n",
      "Removing GP 'NPB_ligand_targetgenes_GP' as it is a subset of GP 'NPB_ligand_receptor_GP'.\n",
      "Removing GP 'NPW_ligand_targetgenes_GP' as it is a subset of GP 'NPW_ligand_receptor_GP'.\n",
      "Removing GP 'RLN3_ligand_targetgenes_GP' as it is a subset of GP 'RLN3_ligand_receptor_GP'.\n",
      "Removing GP 'PYY_ligand_targetgenes_GP' as it is a subset of GP 'PYY_ligand_receptor_GP'.\n",
      "Removing GP 'PPY_ligand_targetgenes_GP' as it is a subset of GP 'PPY_ligand_receptor_GP'.\n",
      "Removing GP 'NRG3_ligand_targetgenes_GP' as it is a subset of GP 'NRG3_ligand_receptor_GP'.\n",
      "Removing GP 'NRG4_ligand_targetgenes_GP' as it is a subset of GP 'NRG4_ligand_receptor_GP'.\n",
      "Removing GP 'FGF22_ligand_targetgenes_GP' as it is a subset of GP 'FGF22_ligand_receptor_GP'.\n",
      "Removing GP 'FGF13_ligand_targetgenes_GP' as it is a subset of GP 'FGF13_ligand_receptor_GP'.\n",
      "Removing GP 'FGF11_ligand_targetgenes_GP' as it is a subset of GP 'FGF11_ligand_receptor_GP'.\n",
      "Removing GP 'FGF14_ligand_targetgenes_GP' as it is a subset of GP 'FGF14_ligand_receptor_GP'.\n",
      "Removing GP 'PODXL2_ligand_targetgenes_GP' as it is a subset of GP 'PODXL2_ligand_receptor_GP'.\n",
      "Removing GP 'ADM2_ligand_targetgenes_GP' as it is a subset of GP 'ADM2_ligand_receptor_GP'.\n",
      "Removing GP 'APOB_ligand_targetgenes_GP' as it is a subset of GP 'APOB_ligand_receptor_GP'.\n",
      "Removing GP 'NPS_ligand_targetgenes_GP' as it is a subset of GP 'NPS_ligand_receptor_GP'.\n",
      "Removing GP 'PTH2_ligand_targetgenes_GP' as it is a subset of GP 'PTH2_ligand_receptor_GP'.\n",
      "Removing GP 'INSL3_ligand_targetgenes_GP' as it is a subset of GP 'INSL3_ligand_receptor_GP'.\n",
      "Removing GP 'NECTIN3_ligand_targetgenes_GP' as it is a subset of GP 'NECTIN3_ligand_receptor_GP'.\n",
      "Removing GP 'LAMA1_ligand_targetgenes_GP' as it is a subset of GP 'LAMA1_ligand_receptor_GP'.\n",
      "Removing GP 'WNT9A_ligand_targetgenes_GP' as it is a subset of GP 'WNT9A_ligand_receptor_GP'.\n",
      "Removing GP 'WNT9B_ligand_targetgenes_GP' as it is a subset of GP 'WNT9B_ligand_receptor_GP'.\n",
      "Removing GP 'WNT3_ligand_targetgenes_GP' as it is a subset of GP 'WNT3_ligand_receptor_GP'.\n",
      "Removing GP 'SOST_ligand_targetgenes_GP' as it is a subset of GP 'SOST_ligand_receptor_GP'.\n",
      "Removing GP 'EFNA1_ligand_targetgenes_GP' as it is a subset of GP 'EFNA1_ligand_receptor_GP'.\n",
      "Removing GP 'EFNA2_ligand_targetgenes_GP' as it is a subset of GP 'EFNA2_ligand_receptor_GP'.\n",
      "Removing GP 'EFNA3_ligand_targetgenes_GP' as it is a subset of GP 'EFNA3_ligand_receptor_GP'.\n",
      "Removing GP 'CCL22_ligand_targetgenes_GP' as it is a subset of GP 'CCL22_ligand_receptor_GP'.\n",
      "Removing GP 'NRG2_ligand_targetgenes_GP' as it is a subset of GP 'NRG2_ligand_receptor_GP'.\n",
      "Removing GP 'LAMC2_ligand_targetgenes_GP' as it is a subset of GP 'LAMC2_ligand_receptor_GP'.\n",
      "Removing GP 'ICAM5_ligand_targetgenes_GP' as it is a subset of GP 'ICAM5_ligand_receptor_GP'.\n",
      "Removing GP 'COMP_ligand_targetgenes_GP' as it is a subset of GP 'COMP_ligand_receptor_GP'.\n",
      "Removing GP 'CHAD_ligand_targetgenes_GP' as it is a subset of GP 'CHAD_ligand_receptor_GP'.\n",
      "Removing GP 'THBS4_ligand_targetgenes_GP' as it is a subset of GP 'THBS4_ligand_receptor_GP'.\n",
      "Removing GP 'SEMA7A_ligand_targetgenes_GP' as it is a subset of GP 'SEMA7A_ligand_receptor_GP'.\n",
      "Removing GP 'CCL8_ligand_targetgenes_GP' as it is a subset of GP 'CCL8_ligand_receptor_GP'.\n",
      "Removing GP 'CCL7_ligand_targetgenes_GP' as it is a subset of GP 'CCL7_ligand_receptor_GP'.\n",
      "Removing GP 'XCL1_ligand_targetgenes_GP' as it is a subset of GP 'XCL1_ligand_receptor_GP'.\n",
      "Removing GP 'CCL4L1_ligand_targetgenes_GP' as it is a subset of GP 'CCL4L1_ligand_receptor_GP'.\n",
      "Removing GP 'COPA_ligand_targetgenes_GP' as it is a subset of GP 'COPA_ligand_receptor_GP'.\n",
      "Removing GP 'GUCA2A_ligand_targetgenes_GP' as it is a subset of GP 'GUCA2A_ligand_receptor_GP'.\n",
      "Removing GP 'GUCA2B_ligand_targetgenes_GP' as it is a subset of GP 'GUCA2B_ligand_receptor_GP'.\n",
      "Removing GP 'NTN4_ligand_targetgenes_GP' as it is a subset of GP 'NTN4_ligand_receptor_GP'.\n",
      "Removing GP 'CD99_ligand_targetgenes_GP' as it is a subset of GP 'CD99_ligand_receptor_GP'.\n",
      "Removing GP 'SEMA3C_ligand_targetgenes_GP' as it is a subset of GP 'SEMA3C_ligand_receptor_GP'.\n",
      "Removing GP 'NECTIN2_ligand_targetgenes_GP' as it is a subset of GP 'NECTIN2_ligand_receptor_GP'.\n",
      "Removing GP 'HLA-B_ligand_targetgenes_GP' as it is a subset of GP 'HLA-B_ligand_receptor_GP'.\n",
      "Removing GP 'HLA-C_ligand_targetgenes_GP' as it is a subset of GP 'HLA-C_ligand_receptor_GP'.\n",
      "Removing GP 'CNTN1_ligand_targetgenes_GP' as it is a subset of GP 'CNTN1_ligand_receptor_GP'.\n",
      "Removing GP 'SEMA3F_ligand_targetgenes_GP' as it is a subset of GP 'SEMA3F_ligand_receptor_GP'.\n",
      "Removing GP 'FAM3C_ligand_targetgenes_GP' as it is a subset of GP 'FAM3C_ligand_receptor_GP'.\n",
      "Removing GP 'ULBP2_ligand_targetgenes_GP' as it is a subset of GP 'ULBP2_ligand_receptor_GP'.\n",
      "Removing GP 'GPHA2_ligand_targetgenes_GP' as it is a subset of GP 'GPHA2_ligand_receptor_GP'.\n",
      "Removing GP 'FLRT3_ligand_targetgenes_GP' as it is a subset of GP 'FLRT3_ligand_receptor_GP'.\n",
      "Removing GP 'PRND_ligand_targetgenes_GP' as it is a subset of GP 'PRND_ligand_receptor_GP'.\n",
      "Removing GP 'NDP_ligand_targetgenes_GP' as it is a subset of GP 'NDP_ligand_receptor_GP'.\n",
      "Removing GP 'RSPO2_ligand_targetgenes_GP' as it is a subset of GP 'RSPO2_ligand_receptor_GP'.\n",
      "Removing GP 'RSPO4_ligand_targetgenes_GP' as it is a subset of GP 'RSPO4_ligand_receptor_GP'.\n",
      "Removing GP 'NTN3_ligand_targetgenes_GP' as it is a subset of GP 'NTN3_ligand_receptor_GP'.\n",
      "Removing GP 'DHH_ligand_targetgenes_GP' as it is a subset of GP 'DHH_ligand_receptor_GP'.\n",
      "Removing GP 'DSC1_ligand_targetgenes_GP' as it is a subset of GP 'DSC1_ligand_receptor_GP'.\n",
      "Removing GP 'DSC2_ligand_targetgenes_GP' as it is a subset of GP 'DSC2_ligand_receptor_GP'.\n",
      "Removing GP 'NLGN2_ligand_targetgenes_GP' as it is a subset of GP 'NLGN2_ligand_receptor_GP'.\n",
      "Removing GP 'NXPH1_ligand_targetgenes_GP' as it is a subset of GP 'NXPH1_ligand_receptor_GP'.\n",
      "Removing GP 'NPFF_ligand_targetgenes_GP' as it is a subset of GP 'NPFF_ligand_receptor_GP'.\n",
      "Removing GP 'TAC3_ligand_targetgenes_GP' as it is a subset of GP 'TAC3_ligand_receptor_GP'.\n",
      "Removing GP 'TAC4_ligand_targetgenes_GP' as it is a subset of GP 'TAC4_ligand_receptor_GP'.\n",
      "Removing GP 'NMB_ligand_targetgenes_GP' as it is a subset of GP 'NMB_ligand_receptor_GP'.\n",
      "Removing GP 'GALP_ligand_targetgenes_GP' as it is a subset of GP 'GALP_ligand_receptor_GP'.\n",
      "Removing GP 'MLN_ligand_targetgenes_GP' as it is a subset of GP 'MLN_ligand_receptor_GP'.\n",
      "Removing GP 'IL17B_ligand_targetgenes_GP' as it is a subset of GP 'IL17B_ligand_receptor_GP'.\n",
      "Removing GP 'QRFP_ligand_targetgenes_GP' as it is a subset of GP 'QRFP_ligand_receptor_GP'.\n",
      "Removing GP 'TXLNA_ligand_targetgenes_GP' as it is a subset of GP 'TXLNA_ligand_receptor_GP'.\n",
      "Removing GP 'Deoxyinosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyadenosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Guanosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Inosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Adenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Corticosterone_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Cortisol_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyinosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Deoxyadenosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'D-Galactose_metabolite_enzyme_sensor_GP' as it is a subset of GP 'D-Glucose_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'D-Maltose_metabolite_enzyme_sensor_GP' as it is a subset of GP 'D-Glucose_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Estriol_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Estrone_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Estriol_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Estradiol_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'L-Dopa_metabolite_enzyme_sensor_GP' as it is a subset of GP 'L-Tyrosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyinosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Inosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Deoxyadenosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Inosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Guanosine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Inosine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Stearic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Oleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Oleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Oleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Epinephrine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Norepinephrine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Stearic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Palmitic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Palmitic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Palmitic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Androsterone_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Testosterone_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Liothyronine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Thyroxine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Cytidine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Uridine_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Thymine_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Uracil_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Linoleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Linoleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Linoleic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Stearic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Stearic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Arachidonic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Arachidonic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Arachidonic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Prostaglandin F2a_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Prostaglandin E2_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP '13-cis-Retinoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Retinal_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'alpha-Linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP '13-cis-Retinoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'all-trans-Retinoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Docosahexaenoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Eicosapentaenoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Eicosapentaenoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Docosahexaenoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP '13-cis-Retinoic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP '9-cis-Retinoic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Dihomo-gamma-linolenic acid_metabolite_enzyme_sensor_GP' as it is a subset of GP 'gamma-Linolenic acid_metabolite_enzyme_sensor_GP'.\n",
      "Removing GP 'Calcitriol_metabolite_enzyme_sensor_GP' as it is a subset of GP 'Calcidiol_metabolite_enzyme_sensor_GP'.\n",
      "Number of gene programs before filtering and combining: 1725.\n",
      "Number of gene programs after filtering and combining: 1575.\n",
      "Number of gene programs with probed genes: 489.\n",
      "Example gene programs: ['L-Lactic acid_metabolite_enzyme_sensor_GP', 'DLL3_ligand_targetgenes_GP', 'BMP15_ligand_targetgenes_GP', 'CD40LG_ligand_targetgenes_GP', 'SFRP2_ligand_receptor_GP'].\n",
      "Number of gene program target genes: 7476.\n",
      "Number of gene program source genes: 127.\n",
      "\n",
      "--- SPATIAL CONNECTIVITY ---\n",
      "Average number of edges per node: 2.0.\n",
      "Number of total edges: 19473.\n",
      "\n",
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, gene_expr_recon_dist: nb\n",
      "NODE LABEL METHOD -> one-hop-attention\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "GRAPH ENCODER -> n_input: 351, n_hidden: 489, n_latent: 489, n_addon_latent: 0, conv_layer: gcnconv, n_attention_heads: -, dropout_rate: 0.0\n",
      "DOT PRODUCT GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED GENE EXPRESSION DECODER -> n_input: 489, n_addon_input: 0, n_output: 702\n",
      "\n",
      "--- INITIALIZING TRAINER ---\n",
      "Number of training nodes: 17506\n",
      "Number of validation nodes: 1945\n",
      "Number of training edges: 17526\n",
      "Number of validation edges: 1947\n",
      "\n",
      "--- MODEL TRAINING ---\n",
      "Epoch 1/20 |█-------------------| 5.0% val_auroc_score: 0.5000; val_auprc_score: 0.5000; val_best_acc_score: 0.5000; val_best_f1_score: 0.6667; train_kl_reg_loss: 1.3656; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 585.6582; train_group_lasso_reg_loss: 0.0000; train_global_loss: 587.0237; train_optim_loss: 587.0237; val_kl_reg_loss: 1.3857; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 564.1559; val_group_lasso_reg_loss: 0.0000; val_global_loss: 565.5416; val_optim_loss: 565.5416\n",
      "Epoch 2/20 |██------------------| 10.0% val_auroc_score: 0.5000; val_auprc_score: 0.5000; val_best_acc_score: 0.5000; val_best_f1_score: 0.6667; train_kl_reg_loss: 1.3095; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 556.7712; train_group_lasso_reg_loss: 0.0000; train_global_loss: 558.0807; train_optim_loss: 558.0807; val_kl_reg_loss: 1.4659; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 549.8195; val_group_lasso_reg_loss: 0.0000; val_global_loss: 551.2854; val_optim_loss: 551.2854\n",
      "Epoch 3/20 |███-----------------| 15.0% val_auroc_score: 0.5003; val_auprc_score: 0.5001; val_best_acc_score: 0.5003; val_best_f1_score: 0.6668; train_kl_reg_loss: 1.2301; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 543.8278; train_group_lasso_reg_loss: 0.0000; train_global_loss: 545.0580; train_optim_loss: 545.0580; val_kl_reg_loss: 1.2420; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 538.7464; val_group_lasso_reg_loss: 0.0000; val_global_loss: 539.9884; val_optim_loss: 539.9884\n",
      "Epoch 4/20 |████----------------| 20.0% val_auroc_score: 0.5000; val_auprc_score: 0.5000; val_best_acc_score: 0.5000; val_best_f1_score: 0.6667; train_kl_reg_loss: 1.1980; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 533.7458; train_group_lasso_reg_loss: 0.0000; train_global_loss: 534.9438; train_optim_loss: 534.9438; val_kl_reg_loss: 1.2719; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 528.7338; val_group_lasso_reg_loss: 0.0000; val_global_loss: 530.0057; val_optim_loss: 530.0057\n",
      "Epoch 5/20 |█████---------------| 25.0% val_auroc_score: 0.5008; val_auprc_score: 0.5004; val_best_acc_score: 0.5003; val_best_f1_score: 0.6668; train_kl_reg_loss: 1.1710; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 525.6173; train_group_lasso_reg_loss: 0.0000; train_global_loss: 526.7882; train_optim_loss: 526.7882; val_kl_reg_loss: 1.2265; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 523.4142; val_group_lasso_reg_loss: 0.0000; val_global_loss: 524.6407; val_optim_loss: 524.6407\n",
      "Epoch 6/20 |██████--------------| 30.0% val_auroc_score: 0.5000; val_auprc_score: 0.5000; val_best_acc_score: 0.5000; val_best_f1_score: 0.6667; train_kl_reg_loss: 1.1535; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 518.7971; train_group_lasso_reg_loss: 0.0000; train_global_loss: 519.9506; train_optim_loss: 519.9506; val_kl_reg_loss: 1.2024; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 519.0094; val_group_lasso_reg_loss: 0.0000; val_global_loss: 520.2118; val_optim_loss: 520.2118\n",
      "Epoch 7/20 |███████-------------| 35.0% val_auroc_score: 0.5008; val_auprc_score: 0.5004; val_best_acc_score: 0.5008; val_best_f1_score: 0.6670; train_kl_reg_loss: 1.1437; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 513.2897; train_group_lasso_reg_loss: 0.0000; train_global_loss: 514.4334; train_optim_loss: 514.4334; val_kl_reg_loss: 1.2158; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 513.2237; val_group_lasso_reg_loss: 0.0000; val_global_loss: 514.4396; val_optim_loss: 514.4396\n",
      "Epoch 8/20 |████████------------| 40.0% val_auroc_score: 0.5005; val_auprc_score: 0.5003; val_best_acc_score: 0.5003; val_best_f1_score: 0.6668; train_kl_reg_loss: 1.1424; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 508.7610; train_group_lasso_reg_loss: 0.0000; train_global_loss: 509.9034; train_optim_loss: 509.9034; val_kl_reg_loss: 1.2023; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 509.7882; val_group_lasso_reg_loss: 0.0000; val_global_loss: 510.9905; val_optim_loss: 510.9905\n",
      "Epoch 9/20 |█████████-----------| 45.0% val_auroc_score: 0.5008; val_auprc_score: 0.5004; val_best_acc_score: 0.5008; val_best_f1_score: 0.6670; train_kl_reg_loss: 1.1369; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 504.8627; train_group_lasso_reg_loss: 0.0000; train_global_loss: 505.9996; train_optim_loss: 505.9996; val_kl_reg_loss: 1.2062; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 509.3710; val_group_lasso_reg_loss: 0.0000; val_global_loss: 510.5772; val_optim_loss: 510.5772\n",
      "Epoch 10/20 |██████████----------| 50.0% val_auroc_score: 0.5005; val_auprc_score: 0.5003; val_best_acc_score: 0.5005; val_best_f1_score: 0.6669; train_kl_reg_loss: 1.1383; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 501.8499; train_group_lasso_reg_loss: 0.0000; train_global_loss: 502.9882; train_optim_loss: 502.9882; val_kl_reg_loss: 1.1724; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 506.3991; val_group_lasso_reg_loss: 0.0000; val_global_loss: 507.5715; val_optim_loss: 507.5715\n",
      "Epoch 11/20 |███████████---------| 55.0% val_auroc_score: 0.5010; val_auprc_score: 0.5005; val_best_acc_score: 0.5005; val_best_f1_score: 0.6669; train_kl_reg_loss: 1.1398; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 499.1559; train_group_lasso_reg_loss: 0.0000; train_global_loss: 500.2957; train_optim_loss: 500.2957; val_kl_reg_loss: 1.1971; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 505.0880; val_group_lasso_reg_loss: 0.0000; val_global_loss: 506.2851; val_optim_loss: 506.2851\n",
      "Epoch 12/20 |████████████--------| 60.0% val_auroc_score: 0.5003; val_auprc_score: 0.5001; val_best_acc_score: 0.5003; val_best_f1_score: 0.6668; train_kl_reg_loss: 1.1364; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 497.0682; train_group_lasso_reg_loss: 0.0000; train_global_loss: 498.2046; train_optim_loss: 498.2046; val_kl_reg_loss: 1.2188; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 502.1411; val_group_lasso_reg_loss: 0.0000; val_global_loss: 503.3599; val_optim_loss: 503.3599\n",
      "Epoch 13/20 |█████████████-------| 65.0% val_auroc_score: 0.5015; val_auprc_score: 0.5008; val_best_acc_score: 0.5008; val_best_f1_score: 0.6670; train_kl_reg_loss: 1.1345; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 495.0181; train_group_lasso_reg_loss: 0.0000; train_global_loss: 496.1526; train_optim_loss: 496.1526; val_kl_reg_loss: 1.1692; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 500.9398; val_group_lasso_reg_loss: 0.0000; val_global_loss: 502.1091; val_optim_loss: 502.1091\n",
      "Epoch 14/20 |██████████████------| 70.0% val_auroc_score: 0.5005; val_auprc_score: 0.5003; val_best_acc_score: 0.5003; val_best_f1_score: 0.6668; train_kl_reg_loss: 1.1367; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 493.6307; train_group_lasso_reg_loss: 0.0000; train_global_loss: 494.7674; train_optim_loss: 494.7674; val_kl_reg_loss: 1.2083; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 500.3934; val_group_lasso_reg_loss: 0.0000; val_global_loss: 501.6017; val_optim_loss: 501.6017\n",
      "Epoch 15/20 |███████████████-----| 75.0% val_auroc_score: 0.5008; val_auprc_score: 0.5004; val_best_acc_score: 0.5005; val_best_f1_score: 0.6669; train_kl_reg_loss: 1.1387; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 492.0627; train_group_lasso_reg_loss: 0.0000; train_global_loss: 493.2014; train_optim_loss: 493.2014; val_kl_reg_loss: 1.2022; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 499.6526; val_group_lasso_reg_loss: 0.0000; val_global_loss: 500.8548; val_optim_loss: 500.8548\n",
      "Epoch 16/20 |████████████████----| 80.0% val_auroc_score: 0.5015; val_auprc_score: 0.5008; val_best_acc_score: 0.5008; val_best_f1_score: 0.6670; train_kl_reg_loss: 1.1411; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 490.9448; train_group_lasso_reg_loss: 0.0000; train_global_loss: 492.0859; train_optim_loss: 492.0859; val_kl_reg_loss: 1.2266; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 498.7191; val_group_lasso_reg_loss: 0.0000; val_global_loss: 499.9457; val_optim_loss: 499.9457\n",
      "Epoch 17/20 |█████████████████---| 85.0% val_auroc_score: 0.5008; val_auprc_score: 0.5004; val_best_acc_score: 0.5008; val_best_f1_score: 0.6670; train_kl_reg_loss: 1.1521; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 490.0373; train_group_lasso_reg_loss: 0.0000; train_global_loss: 491.1894; train_optim_loss: 491.1894; val_kl_reg_loss: 1.2057; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 498.2120; val_group_lasso_reg_loss: 0.0000; val_global_loss: 499.4177; val_optim_loss: 499.4177\n",
      "Epoch 18/20 |██████████████████--| 90.0% val_auroc_score: 0.5010; val_auprc_score: 0.5005; val_best_acc_score: 0.5005; val_best_f1_score: 0.6669; train_kl_reg_loss: 1.1399; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 489.1399; train_group_lasso_reg_loss: 0.0000; train_global_loss: 490.2798; train_optim_loss: 490.2798; val_kl_reg_loss: 1.2332; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 496.5231; val_group_lasso_reg_loss: 0.0000; val_global_loss: 497.7563; val_optim_loss: 497.7563\n",
      "Epoch 19/20 |███████████████████-| 95.0% val_auroc_score: 0.5010; val_auprc_score: 0.5005; val_best_acc_score: 0.5008; val_best_f1_score: 0.6670; train_kl_reg_loss: 1.1530; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 488.2171; train_group_lasso_reg_loss: 0.0000; train_global_loss: 489.3701; train_optim_loss: 489.3701; val_kl_reg_loss: 1.2316; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 499.9874; val_group_lasso_reg_loss: 0.0000; val_global_loss: 501.2191; val_optim_loss: 501.2191\n",
      "Epoch 20/20 |████████████████████| 100.0% val_auroc_score: 0.5015; val_auprc_score: 0.5008; val_best_acc_score: 0.5005; val_best_f1_score: 0.6669; train_kl_reg_loss: 1.1473; train_edge_recon_loss: 0.0000; train_gene_expr_recon_loss: 487.5045; train_group_lasso_reg_loss: 0.0000; train_global_loss: 488.6519; train_optim_loss: 488.6519; val_kl_reg_loss: 1.2247; val_edge_recon_loss: 0.0000; val_gene_expr_recon_loss: 500.4554; val_group_lasso_reg_loss: 0.0000; val_global_loss: 501.6801; val_optim_loss: 501.6801\n",
      "\n",
      "Model training finished after 2 min 0 sec.\n",
      "Using best model state, which was in epoch 18.\n",
      "\n",
      "--- MODEL EVALUATION ---\n",
      "Val AUROC score: 0.5026\n",
      "Val AUPRC score: 0.5013\n",
      "Val best accuracy score: 0.5018\n",
      "Val best F1 score: 0.6675\n",
      "Val MSE score: 1.2817\n",
      "--------------------\n",
      "\n",
      "--------------------\n",
      "\n",
      "--- SPATIAL CONNECTIVITY ---\n",
      "Average number of edges per node: 2.0.\n",
      "Number of total edges: 19473.\n",
      "\n",
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, gene_expr_recon_dist: nb\n",
      "NODE LABEL METHOD -> one-hop-attention\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.0\n",
      "GRAPH ENCODER -> n_input: 351, n_hidden: 489, n_latent: 489, n_addon_latent: 0, conv_layer: gcnconv, n_attention_heads: -, dropout_rate: 0.0\n",
      "DOT PRODUCT GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED GENE EXPRESSION DECODER -> n_input: 489, n_addon_input: 0, n_output: 702\n",
      "\n",
      "--- INITIALIZING TRAINER ---\n",
      "Number of training nodes: 17506\n",
      "Number of validation nodes: 1945\n",
      "Number of training edges: 17526\n",
      "Number of validation edges: 1947\n",
      "\n",
      "--- MODEL TRAINING ---\n",
      "Epoch 1/20 |█-------------------| 5.0% val_auroc_score: 0.6960; val_auprc_score: 0.6220; val_best_acc_score: 0.6626; val_best_f1_score: 0.7476; train_kl_reg_loss: 0.3792; train_edge_recon_loss: 4.3133; train_gene_expr_recon_loss: 589.3968; train_group_lasso_reg_loss: 0.0000; train_global_loss: 594.0893; train_optim_loss: 594.0893; val_kl_reg_loss: 0.5608; val_edge_recon_loss: 3.7429; val_gene_expr_recon_loss: 562.2943; val_group_lasso_reg_loss: 0.0000; val_global_loss: 566.5980; val_optim_loss: 566.5980\n",
      "Epoch 2/20 |██------------------| 10.0% val_auroc_score: 0.6877; val_auprc_score: 0.6156; val_best_acc_score: 0.6507; val_best_f1_score: 0.7411; train_kl_reg_loss: 0.6020; train_edge_recon_loss: 3.7111; train_gene_expr_recon_loss: 558.7684; train_group_lasso_reg_loss: 0.0000; train_global_loss: 563.0814; train_optim_loss: 563.0814; val_kl_reg_loss: 0.7177; val_edge_recon_loss: 3.3538; val_gene_expr_recon_loss: 547.6264; val_group_lasso_reg_loss: 0.0000; val_global_loss: 551.6979; val_optim_loss: 551.6979\n",
      "Epoch 3/20 |███-----------------| 15.0% val_auroc_score: 0.6710; val_auprc_score: 0.6032; val_best_acc_score: 0.6225; val_best_f1_score: 0.7260; train_kl_reg_loss: 0.7017; train_edge_recon_loss: 3.1819; train_gene_expr_recon_loss: 545.3592; train_group_lasso_reg_loss: 0.0000; train_global_loss: 549.2428; train_optim_loss: 549.2428; val_kl_reg_loss: 0.7645; val_edge_recon_loss: 3.0532; val_gene_expr_recon_loss: 536.9095; val_group_lasso_reg_loss: 0.0000; val_global_loss: 540.7271; val_optim_loss: 540.7271\n",
      "Epoch 4/20 |████----------------| 20.0% val_auroc_score: 0.6826; val_auprc_score: 0.6117; val_best_acc_score: 0.6305; val_best_f1_score: 0.7302; train_kl_reg_loss: 0.7752; train_edge_recon_loss: 2.8913; train_gene_expr_recon_loss: 534.9390; train_group_lasso_reg_loss: 0.0000; train_global_loss: 538.6055; train_optim_loss: 538.6055; val_kl_reg_loss: 0.8313; val_edge_recon_loss: 2.6347; val_gene_expr_recon_loss: 527.5868; val_group_lasso_reg_loss: 0.0000; val_global_loss: 531.0528; val_optim_loss: 531.0528\n",
      "Epoch 5/20 |█████---------------| 25.0% val_auroc_score: 0.6452; val_auprc_score: 0.5849; val_best_acc_score: 0.5948; val_best_f1_score: 0.7116; train_kl_reg_loss: 0.8244; train_edge_recon_loss: 2.6604; train_gene_expr_recon_loss: 526.7414; train_group_lasso_reg_loss: 0.0000; train_global_loss: 530.2261; train_optim_loss: 530.2261; val_kl_reg_loss: 0.8745; val_edge_recon_loss: 2.8973; val_gene_expr_recon_loss: 520.5911; val_group_lasso_reg_loss: 0.0000; val_global_loss: 524.3629; val_optim_loss: 524.3629\n",
      "Epoch 6/20 |██████--------------| 30.0% val_auroc_score: 0.6870; val_auprc_score: 0.6150; val_best_acc_score: 0.6276; val_best_f1_score: 0.7287; train_kl_reg_loss: 0.8602; train_edge_recon_loss: 2.5056; train_gene_expr_recon_loss: 519.8895; train_group_lasso_reg_loss: 0.0000; train_global_loss: 523.2554; train_optim_loss: 523.2554; val_kl_reg_loss: 0.9428; val_edge_recon_loss: 2.3982; val_gene_expr_recon_loss: 515.8177; val_group_lasso_reg_loss: 0.0000; val_global_loss: 519.1588; val_optim_loss: 519.1588\n",
      "Epoch 7/20 |███████-------------| 35.0% val_auroc_score: 0.6586; val_auprc_score: 0.5942; val_best_acc_score: 0.6009; val_best_f1_score: 0.7148; train_kl_reg_loss: 0.8955; train_edge_recon_loss: 2.3530; train_gene_expr_recon_loss: 514.2433; train_group_lasso_reg_loss: 0.0000; train_global_loss: 517.4918; train_optim_loss: 517.4918; val_kl_reg_loss: 0.9384; val_edge_recon_loss: 2.3302; val_gene_expr_recon_loss: 513.4925; val_group_lasso_reg_loss: 0.0000; val_global_loss: 516.7611; val_optim_loss: 516.7611\n",
      "Epoch 8/20 |████████------------| 40.0% val_auroc_score: 0.6635; val_auprc_score: 0.5978; val_best_acc_score: 0.6138; val_best_f1_score: 0.7213; train_kl_reg_loss: 0.9102; train_edge_recon_loss: 2.2602; train_gene_expr_recon_loss: 509.7862; train_group_lasso_reg_loss: 0.0000; train_global_loss: 512.9566; train_optim_loss: 512.9566; val_kl_reg_loss: 0.9751; val_edge_recon_loss: 2.3358; val_gene_expr_recon_loss: 509.0647; val_group_lasso_reg_loss: 0.0000; val_global_loss: 512.3756; val_optim_loss: 512.3756\n",
      "Epoch 9/20 |█████████-----------| 45.0% val_auroc_score: 0.6821; val_auprc_score: 0.6114; val_best_acc_score: 0.6166; val_best_f1_score: 0.7227; train_kl_reg_loss: 0.9192; train_edge_recon_loss: 2.1515; train_gene_expr_recon_loss: 505.7765; train_group_lasso_reg_loss: 0.0000; train_global_loss: 508.8472; train_optim_loss: 508.8472; val_kl_reg_loss: 0.9837; val_edge_recon_loss: 2.1419; val_gene_expr_recon_loss: 507.4788; val_group_lasso_reg_loss: 0.0000; val_global_loss: 510.6044; val_optim_loss: 510.6044\n",
      "Epoch 10/20 |██████████----------| 50.0% val_auroc_score: 0.6697; val_auprc_score: 0.6022; val_best_acc_score: 0.6091; val_best_f1_score: 0.7190; train_kl_reg_loss: 0.9391; train_edge_recon_loss: 2.1309; train_gene_expr_recon_loss: 502.6472; train_group_lasso_reg_loss: 0.0000; train_global_loss: 505.7172; train_optim_loss: 505.7172; val_kl_reg_loss: 0.9633; val_edge_recon_loss: 2.1523; val_gene_expr_recon_loss: 504.8578; val_group_lasso_reg_loss: 0.0000; val_global_loss: 507.9734; val_optim_loss: 507.9734\n",
      "Epoch 11/20 |███████████---------| 55.0% val_auroc_score: 0.6895; val_auprc_score: 0.6170; val_best_acc_score: 0.6258; val_best_f1_score: 0.7277; train_kl_reg_loss: 0.9448; train_edge_recon_loss: 2.0358; train_gene_expr_recon_loss: 500.0465; train_group_lasso_reg_loss: 0.0000; train_global_loss: 503.0271; train_optim_loss: 503.0271; val_kl_reg_loss: 0.9955; val_edge_recon_loss: 1.8999; val_gene_expr_recon_loss: 503.0362; val_group_lasso_reg_loss: 0.0000; val_global_loss: 505.9316; val_optim_loss: 505.9316\n",
      "Epoch 12/20 |████████████--------| 60.0% val_auroc_score: 0.6733; val_auprc_score: 0.6049; val_best_acc_score: 0.6091; val_best_f1_score: 0.7188; train_kl_reg_loss: 0.9610; train_edge_recon_loss: 1.9630; train_gene_expr_recon_loss: 497.7165; train_group_lasso_reg_loss: 0.0000; train_global_loss: 500.6405; train_optim_loss: 500.6405; val_kl_reg_loss: 0.9666; val_edge_recon_loss: 1.9944; val_gene_expr_recon_loss: 503.3269; val_group_lasso_reg_loss: 0.0000; val_global_loss: 506.2879; val_optim_loss: 506.2879\n",
      "Epoch 13/20 |█████████████-------| 65.0% val_auroc_score: 0.6923; val_auprc_score: 0.6191; val_best_acc_score: 0.6240; val_best_f1_score: 0.7266; train_kl_reg_loss: 0.9661; train_edge_recon_loss: 1.9619; train_gene_expr_recon_loss: 495.8865; train_group_lasso_reg_loss: 0.0000; train_global_loss: 498.8146; train_optim_loss: 498.8146; val_kl_reg_loss: 0.9853; val_edge_recon_loss: 1.8943; val_gene_expr_recon_loss: 502.1268; val_group_lasso_reg_loss: 0.0000; val_global_loss: 505.0064; val_optim_loss: 505.0064\n",
      "Epoch 14/20 |██████████████------| 70.0% val_auroc_score: 0.6887; val_auprc_score: 0.6164; val_best_acc_score: 0.6240; val_best_f1_score: 0.7266; train_kl_reg_loss: 0.9728; train_edge_recon_loss: 1.9263; train_gene_expr_recon_loss: 494.3398; train_group_lasso_reg_loss: 0.0000; train_global_loss: 497.2389; train_optim_loss: 497.2389; val_kl_reg_loss: 0.9987; val_edge_recon_loss: 1.8912; val_gene_expr_recon_loss: 499.6063; val_group_lasso_reg_loss: 0.0000; val_global_loss: 502.4961; val_optim_loss: 502.4961\n",
      "Epoch 15/20 |███████████████-----| 75.0% val_auroc_score: 0.6845; val_auprc_score: 0.6132; val_best_acc_score: 0.6166; val_best_f1_score: 0.7229; train_kl_reg_loss: 0.9766; train_edge_recon_loss: 1.8758; train_gene_expr_recon_loss: 492.7742; train_group_lasso_reg_loss: 0.0000; train_global_loss: 495.6267; train_optim_loss: 495.6267; val_kl_reg_loss: 1.0436; val_edge_recon_loss: 1.8734; val_gene_expr_recon_loss: 498.7989; val_group_lasso_reg_loss: 0.0000; val_global_loss: 501.7159; val_optim_loss: 501.7159\n",
      "Epoch 16/20 |████████████████----| 80.0% val_auroc_score: 0.7030; val_auprc_score: 0.6274; val_best_acc_score: 0.6297; val_best_f1_score: 0.7296; train_kl_reg_loss: 0.9842; train_edge_recon_loss: 1.8220; train_gene_expr_recon_loss: 491.6410; train_group_lasso_reg_loss: 0.0000; train_global_loss: 494.4472; train_optim_loss: 494.4472; val_kl_reg_loss: 1.0535; val_edge_recon_loss: 1.7448; val_gene_expr_recon_loss: 499.1220; val_group_lasso_reg_loss: 0.0000; val_global_loss: 501.9203; val_optim_loss: 501.9203\n",
      "Epoch 17/20 |█████████████████---| 85.0% val_auroc_score: 0.6857; val_auprc_score: 0.6141; val_best_acc_score: 0.6117; val_best_f1_score: 0.7203; train_kl_reg_loss: 0.9871; train_edge_recon_loss: 1.7914; train_gene_expr_recon_loss: 490.6050; train_group_lasso_reg_loss: 0.0000; train_global_loss: 493.3834; train_optim_loss: 493.3834; val_kl_reg_loss: 1.0356; val_edge_recon_loss: 1.8862; val_gene_expr_recon_loss: 500.0390; val_group_lasso_reg_loss: 0.0000; val_global_loss: 502.9608; val_optim_loss: 502.9608\n",
      "Epoch 18/20 |██████████████████--| 90.0% val_auroc_score: 0.7025; val_auprc_score: 0.6270; val_best_acc_score: 0.6279; val_best_f1_score: 0.7288; train_kl_reg_loss: 0.9905; train_edge_recon_loss: 1.7951; train_gene_expr_recon_loss: 489.7070; train_group_lasso_reg_loss: 0.0000; train_global_loss: 492.4926; train_optim_loss: 492.4926; val_kl_reg_loss: 1.0427; val_edge_recon_loss: 1.7954; val_gene_expr_recon_loss: 497.5913; val_group_lasso_reg_loss: 0.0000; val_global_loss: 500.4294; val_optim_loss: 500.4294\n",
      "Epoch 19/20 |███████████████████-| 95.0% val_auroc_score: 0.7032; val_auprc_score: 0.6276; val_best_acc_score: 0.6387; val_best_f1_score: 0.7344; train_kl_reg_loss: 0.9947; train_edge_recon_loss: 1.7679; train_gene_expr_recon_loss: 489.0412; train_group_lasso_reg_loss: 0.0000; train_global_loss: 491.8039; train_optim_loss: 491.8039; val_kl_reg_loss: 1.0345; val_edge_recon_loss: 1.7904; val_gene_expr_recon_loss: 500.1771; val_group_lasso_reg_loss: 0.0000; val_global_loss: 503.0021; val_optim_loss: 503.0021\n",
      "Epoch 20/20 |████████████████████| 100.0% val_auroc_score: 0.7081; val_auprc_score: 0.6315; val_best_acc_score: 0.6333; val_best_f1_score: 0.7316; train_kl_reg_loss: 0.9987; train_edge_recon_loss: 1.7550; train_gene_expr_recon_loss: 488.2639; train_group_lasso_reg_loss: 0.0000; train_global_loss: 491.0175; train_optim_loss: 491.0175; val_kl_reg_loss: 1.0483; val_edge_recon_loss: 1.7397; val_gene_expr_recon_loss: 497.0438; val_group_lasso_reg_loss: 0.0000; val_global_loss: 499.8318; val_optim_loss: 499.8318\n",
      "\n",
      "Model training finished after 2 min 0 sec.\n",
      "Using best model state, which was in epoch 20.\n",
      "\n",
      "--- MODEL EVALUATION ---\n",
      "Val AUROC score: 0.7074\n",
      "Val AUPRC score: 0.6309\n",
      "Val best accuracy score: 0.6384\n",
      "Val best F1 score: 0.7343\n",
      "Val MSE score: 1.2323\n"
     ]
    }
   ],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200] # 0.10241\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [0, 0.1, 0.2, 0.33, 0.5, 1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [1.]\n",
    "\n",
    "model = run_hyperparam_tuning(dataset=dataset,\n",
    "                              hyperparam_option_dict=hyperparam_option_dict,\n",
    "                              n_iters=1,\n",
    "                              experiment_name=experiment_name,\n",
    "                              gp_mask=\"combined_priors\",\n",
    "                              save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2424a1b6-e24f-4c2c-b8cb-3282181bc9f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.10241] # [0.03094, 0.03875, 0.05102, 0.07200] # 0.10241\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0, 0.1, 0.2, 0.33, 0.5]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55e46e48-3231-447d-a7f6-eb480a0c5e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)\n",
    "\n",
    "def get_loss_weights(row):  \n",
    "    return f\"lambda_edge_recon_{row['lambda_edge_recon_']}_+_lambda_gene_expr_recon_{row['lambda_gene_expr_recon_']}\"\n",
    "df[\"loss_weights\"] = df.apply(lambda row: get_loss_weights(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d640db5a-7962-41d7-a64b-dc01e5b80602",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_eval_metrics(fig_title=\"Loss Weights Benchmarking Evaluation Metrics\",\n",
    "                  df=df,\n",
    "                  y_col_name=\"loss_weights\",\n",
    "                  plot_ratio_active_gps=False,\n",
    "                  file_name=\"loss_weights_benchmarking_eval_metrics.png\",\n",
    "                  save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2812dbb-761b-4083-87aa-4e3cd9b9ff4a",
   "metadata": {},
   "source": [
    "### 2.2. Benchmark Node Label Method"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cad5223e-37f1-4a64-a017-bdec0703f835",
   "metadata": {},
   "source": [
    "- Different node label methods are tested.\n",
    "- Average number of neighbors is varied. The radius is set in order to get an average number of neighbors of ```2```, ```4```, ```8```, ```16``` and ```32``` with 2 iterations respectively, resulting in a total of 10 runs per node label method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80a0c0c7-fdaa-4bf9-b81c-e3d2f8ee3f23",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_name = \"benchmark_node_label_method\"\n",
    "\n",
    "# Set hyperparameters that are not varied as part of the experiment\n",
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0.33]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0]\n",
    "hyperparam_option_dict[\"n_hidden_encoder_divisor\"] = [None]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "\n",
    "hyperparam_option_dict[\"n_epochs\"] = [20]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [10]\n",
    "hyperparam_option_dict[\"lr\"] = [0.001]\n",
    "hyperparam_option_dict[\"batch_size\"] = 64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3d37eba-2f74-43c3-b6ff-862c3d0d074d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\", \"one-hop-norm\", \"one-hop-sum\"]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf57b328-1373-42d0-897f-3a79b2bddb55",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d26c093b-4326-4424-bb37-5df881dc3056",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_eval_metrics(fig_title=\"Node Label Method Benchmarking Evaluation Metrics\",\n",
    "                  df=df,\n",
    "                  y_col_name=\"node_label_method_\",\n",
    "                  plot_ratio_active_gps=False,\n",
    "                  file_name=\"node_label_method_benchmarking_eval_metrics.png\",\n",
    "                  save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044a2e48-32d0-4d6d-b112-69be3a613dd0",
   "metadata": {},
   "source": [
    "### 2.3. Benchmark Gene Expression Reconstruction Distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bffa56f0-a86f-495f-ab33-18c75800566c",
   "metadata": {},
   "source": [
    "- Different gene expression reconstruction distributions are tested.\n",
    "- Average number of neighbors is varied. The radius is set in order to get an average number of neighbors of ```2```, ```4```, ```8```, ```16``` and ```32``` with 2 iterations respectively, resulting in a total of 10 runs per gene expression reconstruction distribution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70271076-445b-4c64-9a42-d2ad46e6b57b",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_name = \"benchmark_gene_expr_recon_dist\"\n",
    "\n",
    "# Set hyperparameters that are not varied as part of the experiment\n",
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0.33]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0]\n",
    "hyperparam_option_dict[\"n_hidden_encoder_divisor\"] = [None]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "\n",
    "hyperparam_option_dict[\"n_epochs\"] = [20]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [10]\n",
    "hyperparam_option_dict[\"lr\"] = [0.001]\n",
    "hyperparam_option_dict[\"batch_size\"] = 64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95c8d585-5f37-45df-8fd9-62cc2f39ac1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\", \"zinb\"]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89fe32af-89b6-4f7b-b21e-e9a537240830",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b2bf032-3cca-42dd-ab4a-6df264467b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_eval_metrics(fig_title=\"Gene Expression Reconstruction Distribution Benchmarking Evaluation Metrics\",\n",
    "                  df=df,\n",
    "                  y_col_name=\"gene_expr_recon_dist_\",\n",
    "                  plot_ratio_active_gps=False,\n",
    "                  file_name=\"gene_expr_recon_dist_benchmarking_eval_metrics.png\",\n",
    "                  save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f46fd7da-d441-4930-993c-d2097143f000",
   "metadata": {},
   "source": [
    "### 2.4. Benchmark Encoder Layer Architecture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f63b689e-68f9-4e6c-9b3c-427985059331",
   "metadata": {},
   "source": [
    "- Different encoder layer architectures are tested.\n",
    "- Average number of neighbors is varied. The radius is set in order to get an average number of neighbors of ```2```, ```4```, ```8```, ```16``` and ```32``` with 2 iterations respectively, resulting in a total of 10 runs per encoder layer architecture."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cad3c72-96b2-4af2-b424-13d83095734b",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_name = \"benchmark_encoder_layer_arch\"\n",
    "\n",
    "# Set hyperparameters that are not varied as part of the experiment\n",
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0.2]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\"]\n",
    "hyperparam_option_dict[\"n_hidden_encoder_divisor\"] = [None]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "\n",
    "hyperparam_option_dict[\"n_epochs\"] = [20]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [10]\n",
    "hyperparam_option_dict[\"lr\"] = [0.001]\n",
    "hyperparam_option_dict[\"batch_size\"] = [64]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf17045e-048e-46f3-a545-4178b1a85f69",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "173d439b-4eca-48b3-9dbf-0bf7cb857e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gatv2conv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [1, 2, 4, 8]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85c31ff-8b68-4747-bfbc-15732ba96520",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)\n",
    "\n",
    "def get_encoder_layer_arch(row):  \n",
    "    return f\"conv_layer_encoder_{row['conv_layer_encoder_']}_+_encoder_n_attention_heads_{row['encoder_n_attention_heads_']}\"\n",
    "df[\"encoder_layer_arch\"] = df.apply(lambda row: get_encoder_layer_arch(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca492888-8665-4ba9-ad63-9984a8584f29",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_hyperparam_benchmarking_metrics(fig_title=\"Encoder Layer Architecture Benchmarking Evaluation Metrics\",\n",
    "                                     df=df,\n",
    "                                     y_col_name=\"encoder_layer_arch\",\n",
    "                                     plot_ratio_active_gps=False,\n",
    "                                     file_name=\"encoder_layer_arch_benchmarking_eval_metrics.png\",\n",
    "                                     save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a8c6803-be46-4c5f-b929-897cd0a74d04",
   "metadata": {},
   "source": [
    "### 2.5. Benchmark Encoder Hidden Layer Size"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08602dba-c435-4ba5-aefd-c04f4af92cdd",
   "metadata": {},
   "source": [
    "- Different sizes of the encoder hidden layer are tested.\n",
    "- Average number of neighbors is varied. The radius is set in order to get an average number of neighbors of ```2```, ```4```, ```8```, ```16``` and ```32``` with 2 iterations respectively, resulting in a total of 10 runs per encoder hidden layer size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72c93e2-b701-4d66-9f27-a8d97ff4633d",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_name = \"benchmark_encoder_n_hidden\"\n",
    "\n",
    "# Set hyperparameters that are not varied as part of the experiment\n",
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0.33]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\", \"zinb\"]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0.]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0]\n",
    "\n",
    "hyperparam_option_dict[\"n_epochs\"] = [20]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [10]\n",
    "hyperparam_option_dict[\"lr\"] = [0.001]\n",
    "hyperparam_option_dict[\"batch_size\"] = 64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6db3138e-7a0c-410d-9407-22b5825cd8fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241]\n",
    "hyperparam_option_dict[\"n_hidden_encoder_divisor\"] = [None, 0.5, 1., 2.]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ccd8899-fc3b-42b7-87dc-8866827f9008",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "737d85b6-2a78-4a15-bdb2-bb124d9627ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_eval_metrics(fig_title=\"Encoder Hidden Layer Size Benchmarking Evaluation Metrics\",\n",
    "                  df=df,\n",
    "                  y_col_name=\"encoder_n_hidden_\",\n",
    "                  plot_ratio_active_gps=False,\n",
    "                  file_name=\"encoder_n_hidden_benchmarking_eval_metrics.png\",\n",
    "                  save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e23fc911-a226-4766-805a-f769d6adbc8d",
   "metadata": {},
   "source": [
    "### 2.6. Benchmark Active Gene Programs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d3ed3be-bc7a-4631-98e3-d34530c0cd0e",
   "metadata": {},
   "source": [
    "- Different combinations of the active gene program threshold ratio and the group lasso regularization weighting hyperparameter are tested. Both these hyperparameters contribute to a reduction of active gene programs.\n",
    "- Average number of neighbors is varied. The radius is set in order to get an average number of neighbors of ```2```, ```4```, ```8```, ```16``` and ```32``` with 2 iterations respectively, resulting in a total of 10 runs per active gp hyperparameter combination."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d87d1e38-f428-4ae6-9062-d900d5494f86",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_name = \"benchmark_active_gps\"\n",
    "\n",
    "# Set hyperparameters that are not varied as part of the experiment\n",
    "hyperparam_option_dict = {}\n",
    "hyperparam_option_dict[\"lambda_edge_recon\"] = [1.]\n",
    "hyperparam_option_dict[\"lambda_gene_expr_recon\"] = [0.33]\n",
    "hyperparam_option_dict[\"node_label_method\"] = [\"one-hop-attention\"]\n",
    "hyperparam_option_dict[\"gene_expr_recon_dist\"] = [\"nb\", \"zinb\"]\n",
    "hyperparam_option_dict[\"conv_layer_encoder\"] = [\"gcnconv\"]\n",
    "hyperparam_option_dict[\"encoder_n_attention_heads\"] = [0]\n",
    "hyperparam_option_dict[\"n_hidden_encoder_divisor\"] = [None]\n",
    "\n",
    "hyperparam_option_dict[\"n_epochs\"] = [20]\n",
    "hyperparam_option_dict[\"n_epochs_all_gps\"] = [10]\n",
    "hyperparam_option_dict[\"lr\"] = [0.001]\n",
    "hyperparam_option_dict[\"batch_size\"] = 64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5cecf9c-b7d7-4390-a21c-a6cd47d6b5dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "hyperparam_option_dict[\"radius\"] = [0.03094, 0.03875, 0.05102, 0.07200, 0.10241]\n",
    "hyperparam_option_dict[\"active_gp_thresh_ratio\"] = [0., 0.01, 0.03, 0.1, 0.3]\n",
    "hyperparam_option_dict[\"lambda_group_lasso\"] = [0, 0.01, 0.03, 0.1, 0.3]\n",
    "\n",
    "model = run_hyperparam_benchmarking(dataset=dataset,\n",
    "                                    hyperparam_option_dict=hyperparam_option_dict,\n",
    "                                    n_iters=2,\n",
    "                                    experiment_name=experiment_name,\n",
    "                                    gp_mask=\"combined_priors\",\n",
    "                                    save_model=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57a353f4-36f6-45d4-871e-a79529ff8a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "runs = mlflow.search_runs(experiment_names=[experiment_name],\n",
    "                          output_format=\"list\")\n",
    "\n",
    "data = []\n",
    "for run in runs:\n",
    "    data.append({**run.data.metrics, **run.data.params})\n",
    "df = pd.DataFrame.from_dict(data)\n",
    "\n",
    "# Get active gene program ratio\n",
    "def get_ratio_active_gps(row):\n",
    "    return row[\"n_active_gps\"] / (int(row[\"n_nonaddon_gps_\"]) + int(row[\"n_addon_gps_\"]))\n",
    "df[\"ratio_active_gps\"] = df.apply(lambda row: get_ratio_active_gps(row), axis=1)\n",
    "\n",
    "# Get combination of active gene program threshold and lambda group lasso\n",
    "def get_active_gp_thresh_lambda_group_lasso(row):\n",
    "    return f\"active_gp_thresh_{row['active_gp_thresh_ratio_']}_+_lambda_group_lasso_{row['lambda_group_lasso_']}\"\n",
    "df[\"active_gp_thresh_lambda_group_lasso\"] = df.apply(lambda row: get_active_gp_thresh_lambda_group_lasso(row), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa715ba-3bb0-4821-9b08-c41594145071",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_eval_metrics(fig_title=\"Active GP Benchmarking Evaluation Metrics\",\n",
    "                  df=df,\n",
    "                  y_col_name=\"active_gp_thresh_lambda_group_lasso\",\n",
    "                  plot_ratio_active_gps=True,\n",
    "                  file_name=\"active_gp_benchmarking_eval_metrics.png\",\n",
    "                  save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "469c93e1-47e9-4a5b-883d-7712693d39df",
   "metadata": {},
   "source": [
    "## 3. Method Benchmarking Experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e90e7c8-12c1-476d-969e-3831d4e0fbef",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
