{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# NicheCompass Data Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-López Lab\n",
    "- **Date of Creation:** 22.01.2023\n",
    "- **Date of Last Modification:** 31.08.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6bd6fb-6697-43bd-b3ec-68b061afe511",
   "metadata": {},
   "source": [
    "- In order to run this notebook, a trained model needs to be stored under f\"../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../utils\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import argparse\n",
    "import gc\n",
    "import os\n",
    "import random\n",
    "import shutil\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "import torch\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.pyplot import rc_context\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from pywaffle import Waffle\n",
    "\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                                aggregate_obsp_matrix_per_cell_type,\n",
    "                                create_cell_type_chord_plot_from_df,\n",
    "                                create_new_color_dict,\n",
    "                                generate_enriched_gp_info_plots)\n",
    "\n",
    "from analysis_utils import (add_cell_type_latent_cluster_emphasis,\n",
    "                            add_sub_cell_type,\n",
    "                            compute_cell_type_latent_clusters,\n",
    "                            generate_gp_info_plots,\n",
    "                            plot_physical_latent_for_cell_types,\n",
    "                            plot_cell_type_latent_clusters,\n",
    "                            plot_latent,\n",
    "                            plot_category_in_latent_and_physical_space,\n",
    "                            sankey,\n",
    "                            store_top_gps_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9325589b-a7e4-4c31-97e7-8b7048248896",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "# dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "# dataset = \"vizgen_merfish_human_colon_cancer\"\n",
    "# dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "# dataset = \"vizgen_merfish_human_lung_cancer\"\n",
    "# dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "# dataset = \"starmap_plus_mouse_cns\"\n",
    "# dataset = \"xenium_human_breast_cancer\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ac7b6d-4518-4356-bfdd-e141bf25d504",
   "metadata": {},
   "source": [
    "#### 1.2.1 Generic Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Model\n",
    "# AnnData keys\n",
    "adj_key = \"spatial_connectivities\"\n",
    "spatial_key = \"spatial\"\n",
    "# sub_cell_type_key = \"sub_cell_type\"\n",
    "sub_cell_type_key = \"cell_type_original\"\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "active_gp_names_key = \"nichecompass_active_gp_names\"\n",
    "latent_key = \"nichecompass_latent\"\n",
    "mapping_entity_key = \"mapping_entity\"\n",
    "\n",
    "## Analysis\n",
    "differential_gp_test_results_key = \"nichecompass_differential_gp_test_results\"\n",
    "\n",
    "## Others\n",
    "random_seed = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7be323-bd40-4fa0-a2d5-b8ddf5cc1c51",
   "metadata": {},
   "source": [
    "#### 1.2.2 Dataset-specific Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "01210b2f-017e-43f6-b650-5cc732252f8c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "multimodal = False\n",
    "log_norm_omics_features = False\n",
    "cell_type_groups = []\n",
    "latent_groups = []\n",
    "\n",
    "if dataset == \"seqfish_mouse_organogenesis_imputed\":\n",
    "    load_timestamp = \"09082023_225056_10\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    counts_key = \"log_normalized_counts\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    dataset_str = \"seqFISH Mouse Organogenesis Imputed\"\n",
    "    multimodal = False\n",
    "    sample_key = \"sample\"\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.35\n",
    "    latent_groups = [\"16\", \"15\", \"4\", \"10\", \"5\", \"6\"]\n",
    "    spot_size = 0.05\n",
    "    cell_type_groups = []\n",
    "    multimodal = False\n",
    "elif dataset == \"starmap_plus_mouse_cns\":\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    groups = \"Astrocytes\"\n",
    "    dataset_str = \"STARmap PLUS Mouse Central Nervous System\"\n",
    "    load_timestamp = \"20062023_155306_1\"\n",
    "    #load_timestamp = \"12062023_110439\"\n",
    "    #load_timestamp = \"20062023_164533_42\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.2\n",
    "    multimodal = False\n",
    "    spot_size=0.2\n",
    "    cell_type_groups = [\"Microglia\", \"Pericytes\"]\n",
    "elif dataset == \"visium_human_heart\":\n",
    "    load_timestamp = \"15052023_102158\"\n",
    "    cell_type_key = \"majority_cell_type\"\n",
    "    dataset_str = \"Visium Human Heart\"\n",
    "    sample_key = \"batch\"\n",
    "    samples = [\"batch1\", \"batch2\", \"batch3\"]\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.5\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    latent_cluster_spot_size = 200\n",
    "elif dataset == \"spatial_atac_rna_seq_mouse_brain_batch2\":\n",
    "    load_timestamp = \"09082023_145510_69\"\n",
    "    load_timestamp = \"10082023_091319_10\"\n",
    "    #load_timestamp = \"09082023_130425_21\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    cell_type_key = \"RNA_clusters\"\n",
    "    rna_cluster_key = \"RNA_clusters\"\n",
    "    atac_cluster_key = \"ATAC_clusters\"\n",
    "    dataset_str = \"Spatial ATAC-RNA-Seq Mouse Brain\"\n",
    "    sample_key = \"batch\"\n",
    "    samples = [\"p22\"]\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.23 # 0.3\n",
    "    latent_cluster_spot_size = 30\n",
    "    latent_groups = []\n",
    "    multimodal = True\n",
    "    spot_size = 30\n",
    "    cell_type_groups = []\n",
    "    latent_groups = [\"1\", \"6\", \"7\", \"5\", \"3\"]\n",
    "    log_norm_omics_features = True\n",
    "elif dataset == \"nanostring_cosmx_human_nsclc_filtered\":\n",
    "    load_timestamp = \"03092023_010847_6\"\n",
    "    load_timestamp = \"11072023_130002_1\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    dataset_str = \"Nanostring CosMx Human NSCLC\"\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.03\n",
    "    latent_cluster_spot_size = 60\n",
    "    sample_key = \"batch\"\n",
    "    multimodal = False\n",
    "    spot_size = 50\n",
    "    cell_type_key = \"cell_type\"\n",
    "elif dataset == \"vizgen_merfish_human_colon_cancer\":\n",
    "    load_timestamp = \"27062023_113037_2\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    dataset_str = \"Vizgen MERFISH Human Colon Cancer\"\n",
    "    condition_key = \"sample\"\n",
    "    sample_key = \"batch\"\n",
    "    samples = [\"batch1\", \"batch2\"]\n",
    "    latent_leiden_resolution = 0.5\n",
    "    latent_cluster_spot_size = 30\n",
    "elif dataset == \"vizgen_merfish_human_ovarian_cancer\":\n",
    "    load_timestamp = \"13062023_131002\"\n",
    "    load_timestamp = \"27062023_113037_2\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"Vizgen MERFISH Human Ovarian Cancer\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_type\"\n",
    "elif dataset == \"vizgen_merfish_human_lung_cancer\":\n",
    "    load_timestamp = \"13062023_131002\"\n",
    "    load_timestamp = \"19062023_140621_1\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"Vizgen MERFISH Human Lung Cancer\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_type\"\n",
    "elif dataset == \"xenium_human_breast_cancer\":\n",
    "    load_timestamp = \"28082023_182430_3\" # 12 neighbors\n",
    "    # load_timestamp = \"20062023_120357_1\" # 4 neighbors\n",
    "    model_label = \"reference\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"Xenium Human Breast Cancer\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_states\"\n",
    "elif dataset == \"nanostring_cosmx_human_nsclc\":\n",
    "    load_timestamp = \"03092023_001459_4\"\n",
    "    model_label = \"reference_query_mapping\"\n",
    "    latent_leiden_resolution = 0.4\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"nanoString CosMx Human NSCLC\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_type\"\n",
    "    \n",
    "latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1bf229ca-ab66-4e52-a31e-8655878d290d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "924032d4-24a4-45cf-87e1-0eba197da0c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_folder_path = f\"../artifacts/{dataset}/figures/{model_label}/{load_timestamp}\"\n",
    "model_folder_path = f\"../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"\n",
    "result_folder_path = f\"../artifacts/{dataset}/results/{model_label}/{load_timestamp}\"\n",
    "gp_data_folder_path = \"../datasets/gp_data\" # gene program data\n",
    "srt_data_folder_path = \"../datasets/srt_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(result_folder_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cea1a5c-d075-4e51-90e5-90684bee88a6",
   "metadata": {},
   "source": [
    "## 2. Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d6ede1d-2cdf-4c81-9f56-e4745ae59e6c",
   "metadata": {},
   "source": [
    "### 2.1 Load Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe9e6c2-6ea1-49c7-a625-79e0386ac5bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if model_label == \"sample_integration_method_benchmarking\":\n",
    "    model_folder_path = f\"{model_folder_path}/run{run_number}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04cf6693-e1bf-42f2-8f96-9ea79fb7ebde",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_folder_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dad2ca6-7e34-48bc-89ef-4f994389a0f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if multimodal:\n",
    "    model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                              adata=None,\n",
    "                              adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                              adata_atac=None,\n",
    "                              adata_atac_file_name=f\"{dataset}_{model_label}_atac.h5ad\",\n",
    "                              gp_names_key=gp_names_key)\n",
    "else:\n",
    "    # Load trained model\n",
    "    model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                              adata=None,\n",
    "                              adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                              gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "376e761e-2101-4d1f-958f-5ad483b6f0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f662f73-455e-46c2-bc4b-ec25fb2f1d2b",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "efcc5e38-bc0c-4958-adeb-b7afe3a0886a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.03\n",
      "LOG VARIATIONAL -> True\n",
      "CATEGORICAL COVARIATES EMBEDDINGS INJECTION -> ['gene_expr_decoder']\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 960, n_cat_covariates_embed_input: 0, n_hidden: 960, n_latent: 1494, n_addon_latent: 100, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, use_bn: False\n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 1494, n_addon_gp_input: 100, n_cat_covariates_embed_input: 38, n_output: 960\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 1494, n_addon_gp_input: 100, n_cat_covariates_embed_input: 38, n_output: 960\n"
     ]
    }
   ],
   "source": [
    "# Load trained model\n",
    "model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                          adata=None,\n",
    "                          adata_file_name=f\"{dataset}_{model_label}_postprocessed.h5ad\",\n",
    "                          gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7f625913-a65f-457b-991a-3821acea3591",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = model.adata.obs[sample_key].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fc40859-6a4a-4bfa-8362-d28d4c1c2057",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Check number of active gene programs\n",
    "active_gps = model.get_active_gps()\n",
    "model.adata.uns[active_gp_names_key] = active_gps\n",
    "print(f\"Number of total gene programs: {len(model.adata.uns[gp_names_key])}\")\n",
    "print(f\"Number of active gene programs: {len(active_gps)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d99a385-5769-4c6b-a7dc-62f40028620a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.uns[gp_names_key] = np.array([gp for gp in model.adata.uns[gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp for gp in model.adata.uns[active_gp_names_key] if not \"Add-on \" in gp])\n",
    "active_gps = np.array([gp.replace(\" \", \"_\") for gp in active_gps]) \n",
    "model.adata.uns[gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[gp_names_key]])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[active_gp_names_key]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85cb0edb-1eb6-48b5-b56c-079501cb697d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gp_summary_df = model.get_gp_summary()\n",
    "gp_summary_df[gp_summary_df[\"gp_active\"]][0:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5170a527-58ea-42b0-859b-b582c69d7d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_active_gp_scores_to_obs()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73f67b76-7079-4b6b-a97b-da34f48348ad",
   "metadata": {},
   "source": [
    "### 3.1 Analyze Latent Manifold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7707921-aac1-436d-a2cc-f44052d225aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "if model.adata.obs[mapping_entity_key].nunique() > 1:\n",
    "    # Plot mapping entities in latent and physical space\n",
    "    mapping_entity_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "        cat_key=mapping_entity_key)\n",
    "    \n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/\" \\\n",
    "                \"mapping_entities_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"Mapping Entities\",\n",
    "        cat_key=mapping_entity_key,\n",
    "        groups=None,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=mapping_entity_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b03f95cd-74f6-4e19-bf75-2142850b5801",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot batches in latent and physical space\n",
    "condition_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=condition_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"batches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Batches\",\n",
    "    cat_key=condition_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=condition_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f995026b-cd0b-44b6-ad51-ef27b9a7021c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot cell types in latent and physical space\n",
    "cell_type_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=cell_type_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"cell_types_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Cell Types\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe216d6c-3f23-4356-9e23-3e79f73af61e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot cell types in latent and physical space\n",
    "cell_type_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=sub_cell_type_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"subcell_types_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Cell Types\",\n",
    "    cat_key=sub_cell_type_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "996cf62a-8eb8-4b65-8bd8-d7fb7af460c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot cell types in latent and physical space\n",
    "cell_type_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key='niche')\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"niches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niche\",\n",
    "    cat_key='niche',\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44abc3f7-9d93-408b-be6e-0b1e4e94866d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute latent Leiden clustering\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=latent_leiden_resolution,\n",
    "             key_added=latent_cluster_key,\n",
    "             neighbors_key=latent_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc3a99a-0ce9-4d91-a178-49c91db1f6d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.write(f\"{model_folder_path}/{dataset}_{model_label}_postprocessed.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0204fae9-d88b-4a2a-9d0c-1e0abc579729",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot latent clusters in latent and physical space\n",
    "latent_cluster_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=latent_cluster_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"latent_clusters_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Latent Clusters\",\n",
    "    cat_key=latent_cluster_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=latent_cluster_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aeb14bb-a680-41f2-88d1-6df8ca73950d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for latent_cluster in model.adata.obs[latent_cluster_key].unique().tolist():\n",
    "    # Latent clusters in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_cluster_{latent_cluster}_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"Latent Clusters\",\n",
    "        cat_key=latent_cluster_key,\n",
    "        groups=latent_cluster,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=latent_cluster_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24099fd-afa8-4dea-a3c1-25380cdd1ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# latent_groups = [\"5\", \"6\", \"10\", \"15\"]\n",
    "# latent_groups = [\"4\", \"16\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "928dc642-a1e3-4b55-8bc8-7e4b6c78899f",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_key, color_key = 'niche', cell_type_key\n",
    "df = model.adata.obs.groupby([group_key, color_key]).size().reset_index(name = 'counts')\n",
    "\n",
    "grouped = df.groupby(group_key)['counts'].apply(lambda x: x / x.sum() * 100)\n",
    "grouped = grouped.reset_index()\n",
    "\n",
    "df['proportions'] = grouped['counts']\n",
    "df['waffle_counts'] = (df['proportions'] * 10).astype(int)\n",
    "for group in df[group_key].unique():\n",
    "    temp_df = df[df[group_key] == group]\n",
    "    \n",
    "    data = dict(zip(temp_df[color_key], temp_df['waffle_counts']))\n",
    "    fig = plt.figure(\n",
    "        FigureClass = Waffle, \n",
    "        rows = 8, \n",
    "        values = data, \n",
    "        title = {'label': f'Group {group}', 'loc': 'left'},\n",
    "        labels = [f\"{k} ({v}%)\" for k, v in zip(temp_df[color_key], temp_df['proportions'].round(2))],\n",
    "        #legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': len(data), 'framealpha': 0},\n",
    "        legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': 12, 'framealpha': 0},\n",
    "        figsize = (40, 4)\n",
    "    )\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0680b05d-02c9-4989-9b20-dc3eac1bcb57",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_key, color_key = 'latent_leiden_0.4', 'niche'\n",
    "df = model.adata.obs.groupby([group_key, color_key]).size().reset_index(name = 'counts')\n",
    "\n",
    "grouped = df.groupby(group_key)['counts'].apply(lambda x: x / x.sum() * 100)\n",
    "grouped = grouped.reset_index()\n",
    "\n",
    "df['proportions'] = grouped['counts']\n",
    "df['waffle_counts'] = (df['proportions'] * 10).astype(int)\n",
    "for group in df[group_key].unique():\n",
    "    temp_df = df[df[group_key] == group]\n",
    "    \n",
    "    data = dict(zip(temp_df[color_key], temp_df['waffle_counts']))\n",
    "    fig = plt.figure(\n",
    "        FigureClass = Waffle, \n",
    "        rows = 8, \n",
    "        values = data, \n",
    "        title = {'label': f'Group {group}', 'loc': 'left'},\n",
    "        labels = [f\"{k} ({v}%)\" for k, v in zip(temp_df[color_key], temp_df['proportions'].round(2))],\n",
    "        #legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': len(data), 'framealpha': 0},\n",
    "        legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': 12, 'framealpha': 0},\n",
    "        figsize = (40, 4)\n",
    "    )\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "266cc964-7bde-4664-bafd-b793b584dc7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_key, color_key = 'latent_leiden_0.4', sub_cell_type_key\n",
    "df = model.adata.obs.groupby([group_key, color_key]).size().reset_index(name = 'counts')\n",
    "\n",
    "grouped = df.groupby(group_key)['counts'].apply(lambda x: x / x.sum() * 100)\n",
    "grouped = grouped.reset_index()\n",
    "\n",
    "df['proportions'] = grouped['counts']\n",
    "df['waffle_counts'] = (df['proportions'] * 10).astype(int)\n",
    "for group in df[group_key].unique():\n",
    "    temp_df = df[df[group_key] == group]\n",
    "    \n",
    "    data = dict(zip(temp_df[color_key], temp_df['waffle_counts']))\n",
    "    fig = plt.figure(\n",
    "        FigureClass = Waffle, \n",
    "        rows = 8, \n",
    "        values = data, \n",
    "        title = {'label': f'Group {group}', 'loc': 'left'},\n",
    "        labels = [f\"{k} ({v}%)\" for k, v in zip(temp_df[color_key], temp_df['proportions'].round(2))],\n",
    "        #legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': len(data), 'framealpha': 0},\n",
    "        legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': 12, 'framealpha': 0},\n",
    "        figsize = (40, 4)\n",
    "    )\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5236bdd-b337-449a-a69b-3b5ad59486e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_key, color_key = 'latent_leiden_0.4', cell_type_key\n",
    "df = model.adata.obs.groupby([group_key, color_key]).size().reset_index(name = 'counts')\n",
    "\n",
    "grouped = df.groupby(group_key)['counts'].apply(lambda x: x / x.sum() * 100)\n",
    "grouped = grouped.reset_index()\n",
    "\n",
    "df['proportions'] = grouped['counts']\n",
    "df['waffle_counts'] = (df['proportions'] * 10).astype(int)\n",
    "for group in df[group_key].unique():\n",
    "    temp_df = df[df[group_key] == group]\n",
    "    \n",
    "    data = dict(zip(temp_df[color_key], temp_df['waffle_counts']))\n",
    "    fig = plt.figure(\n",
    "        FigureClass = Waffle, \n",
    "        rows = 8, \n",
    "        values = data, \n",
    "        title = {'label': f'Group {group}', 'loc': 'left'},\n",
    "        labels = [f\"{k} ({v}%)\" for k, v in zip(temp_df[color_key], temp_df['proportions'].round(2))],\n",
    "        #legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': len(data), 'framealpha': 0},\n",
    "        legend = {'loc': 'lower left', 'bbox_to_anchor': (0, -0.4), 'ncol': 12, 'framealpha': 0},\n",
    "        figsize = (40, 4)\n",
    "    )\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a108db-5134-4b51-9975-a3935948b2b9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3.2 Analyze GPs & GP Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "72171e77-6549-4b71-a2df-5460a13fca0a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_ID</th>\n",
       "      <th>patient</th>\n",
       "      <th>batch</th>\n",
       "      <th>fov</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>niche</th>\n",
       "      <th>cell_type_original</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>mapping_entity</th>\n",
       "      <th>FYN_ligand_receptor_GP</th>\n",
       "      <th>...</th>\n",
       "      <th>Add-on_7_GP</th>\n",
       "      <th>Add-on_19_GP</th>\n",
       "      <th>Add-on_31_GP</th>\n",
       "      <th>Add-on_55_GP</th>\n",
       "      <th>Add-on_70_GP</th>\n",
       "      <th>Add-on_82_GP</th>\n",
       "      <th>Add-on_87_GP</th>\n",
       "      <th>Add-on_95_GP</th>\n",
       "      <th>Add-on_98_GP</th>\n",
       "      <th>latent_leiden_0.4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>c_1_1_3</td>\n",
       "      <td>Lung5</td>\n",
       "      <td>lung5_rep1</td>\n",
       "      <td>lung5_rep1_1</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>stroma</td>\n",
       "      <td>endothelial</td>\n",
       "      <td>74.0</td>\n",
       "      <td>reference</td>\n",
       "      <td>0.681425</td>\n",
       "      <td>...</td>\n",
       "      <td>-17.489004</td>\n",
       "      <td>-14.936492</td>\n",
       "      <td>3.799693</td>\n",
       "      <td>-39.409985</td>\n",
       "      <td>2.128621</td>\n",
       "      <td>-12.063638</td>\n",
       "      <td>18.894939</td>\n",
       "      <td>5.632041</td>\n",
       "      <td>-39.309853</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c_1_1_4</td>\n",
       "      <td>Lung5</td>\n",
       "      <td>lung5_rep1</td>\n",
       "      <td>lung5_rep1_1</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>stroma</td>\n",
       "      <td>pDC</td>\n",
       "      <td>60.0</td>\n",
       "      <td>reference</td>\n",
       "      <td>0.273680</td>\n",
       "      <td>...</td>\n",
       "      <td>-13.203317</td>\n",
       "      <td>8.522111</td>\n",
       "      <td>-2.774563</td>\n",
       "      <td>-15.451062</td>\n",
       "      <td>-3.280380</td>\n",
       "      <td>-37.737724</td>\n",
       "      <td>-5.105931</td>\n",
       "      <td>10.302983</td>\n",
       "      <td>-0.375488</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c_1_1_5</td>\n",
       "      <td>Lung5</td>\n",
       "      <td>lung5_rep1</td>\n",
       "      <td>lung5_rep1_1</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>stroma</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>52.0</td>\n",
       "      <td>reference</td>\n",
       "      <td>0.221490</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.162962</td>\n",
       "      <td>11.373133</td>\n",
       "      <td>-5.187509</td>\n",
       "      <td>-19.356821</td>\n",
       "      <td>-3.515104</td>\n",
       "      <td>-35.349342</td>\n",
       "      <td>-1.645619</td>\n",
       "      <td>10.553545</td>\n",
       "      <td>-1.178124</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>c_1_1_7</td>\n",
       "      <td>Lung5</td>\n",
       "      <td>lung5_rep1</td>\n",
       "      <td>lung5_rep1_1</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>plasmablast-enriched stroma</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>419.0</td>\n",
       "      <td>reference</td>\n",
       "      <td>0.427213</td>\n",
       "      <td>...</td>\n",
       "      <td>-22.170458</td>\n",
       "      <td>22.927443</td>\n",
       "      <td>-9.017243</td>\n",
       "      <td>0.945892</td>\n",
       "      <td>-5.096985</td>\n",
       "      <td>22.844490</td>\n",
       "      <td>19.639498</td>\n",
       "      <td>2.279819</td>\n",
       "      <td>-31.737906</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c_1_1_9</td>\n",
       "      <td>Lung5</td>\n",
       "      <td>lung5_rep1</td>\n",
       "      <td>lung5_rep1_1</td>\n",
       "      <td>plasmablast</td>\n",
       "      <td>plasmablast-enriched stroma</td>\n",
       "      <td>plasmablast</td>\n",
       "      <td>249.0</td>\n",
       "      <td>reference</td>\n",
       "      <td>0.568128</td>\n",
       "      <td>...</td>\n",
       "      <td>-26.380348</td>\n",
       "      <td>-14.176794</td>\n",
       "      <td>3.803596</td>\n",
       "      <td>10.833162</td>\n",
       "      <td>-5.772773</td>\n",
       "      <td>22.477608</td>\n",
       "      <td>40.085487</td>\n",
       "      <td>-23.397999</td>\n",
       "      <td>-33.345215</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81230</th>\n",
       "      <td>c_8_20_3998</td>\n",
       "      <td>Lung13</td>\n",
       "      <td>lung13</td>\n",
       "      <td>lung13_20</td>\n",
       "      <td>NK/T cell</td>\n",
       "      <td>macrophages</td>\n",
       "      <td>T CD4 memory</td>\n",
       "      <td>80.0</td>\n",
       "      <td>query</td>\n",
       "      <td>1.114208</td>\n",
       "      <td>...</td>\n",
       "      <td>-88.054962</td>\n",
       "      <td>17.349173</td>\n",
       "      <td>-17.536432</td>\n",
       "      <td>-23.746794</td>\n",
       "      <td>-6.791336</td>\n",
       "      <td>-35.896069</td>\n",
       "      <td>-2.281905</td>\n",
       "      <td>-2.327273</td>\n",
       "      <td>7.340417</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81231</th>\n",
       "      <td>c_8_20_3999</td>\n",
       "      <td>Lung13</td>\n",
       "      <td>lung13</td>\n",
       "      <td>lung13_20</td>\n",
       "      <td>NK/T cell</td>\n",
       "      <td>immune</td>\n",
       "      <td>T CD4 memory</td>\n",
       "      <td>135.0</td>\n",
       "      <td>query</td>\n",
       "      <td>0.136362</td>\n",
       "      <td>...</td>\n",
       "      <td>-51.935818</td>\n",
       "      <td>6.413656</td>\n",
       "      <td>-19.044233</td>\n",
       "      <td>-12.997680</td>\n",
       "      <td>-1.418877</td>\n",
       "      <td>-40.414978</td>\n",
       "      <td>-5.716311</td>\n",
       "      <td>5.811738</td>\n",
       "      <td>8.726611</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81232</th>\n",
       "      <td>c_8_20_4000</td>\n",
       "      <td>Lung13</td>\n",
       "      <td>lung13</td>\n",
       "      <td>lung13_20</td>\n",
       "      <td>NK/T cell</td>\n",
       "      <td>immune</td>\n",
       "      <td>T CD4 memory</td>\n",
       "      <td>74.0</td>\n",
       "      <td>query</td>\n",
       "      <td>0.434632</td>\n",
       "      <td>...</td>\n",
       "      <td>-64.320480</td>\n",
       "      <td>5.044941</td>\n",
       "      <td>-10.139482</td>\n",
       "      <td>-21.754337</td>\n",
       "      <td>-3.457257</td>\n",
       "      <td>-13.774080</td>\n",
       "      <td>-7.491361</td>\n",
       "      <td>-3.326872</td>\n",
       "      <td>10.765855</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81233</th>\n",
       "      <td>c_8_20_4001</td>\n",
       "      <td>Lung13</td>\n",
       "      <td>lung13</td>\n",
       "      <td>lung13_20</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>macrophages</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>102.0</td>\n",
       "      <td>query</td>\n",
       "      <td>0.992373</td>\n",
       "      <td>...</td>\n",
       "      <td>-91.116089</td>\n",
       "      <td>28.286760</td>\n",
       "      <td>-24.224100</td>\n",
       "      <td>-36.481979</td>\n",
       "      <td>-8.841825</td>\n",
       "      <td>0.247141</td>\n",
       "      <td>-2.792174</td>\n",
       "      <td>-1.966371</td>\n",
       "      <td>3.307793</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81235</th>\n",
       "      <td>c_8_20_4004</td>\n",
       "      <td>Lung13</td>\n",
       "      <td>lung13</td>\n",
       "      <td>lung13_20</td>\n",
       "      <td>myeloid</td>\n",
       "      <td>immune</td>\n",
       "      <td>macrophage</td>\n",
       "      <td>69.0</td>\n",
       "      <td>query</td>\n",
       "      <td>0.242673</td>\n",
       "      <td>...</td>\n",
       "      <td>-105.144196</td>\n",
       "      <td>24.266064</td>\n",
       "      <td>-29.696901</td>\n",
       "      <td>-41.929630</td>\n",
       "      <td>-8.734715</td>\n",
       "      <td>-19.529289</td>\n",
       "      <td>4.879042</td>\n",
       "      <td>-9.399879</td>\n",
       "      <td>-15.149623</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>702199 rows × 1018 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_ID patient       batch           fov    cell_type  \\\n",
       "2          c_1_1_3   Lung5  lung5_rep1  lung5_rep1_1  endothelial   \n",
       "3          c_1_1_4   Lung5  lung5_rep1  lung5_rep1_1      myeloid   \n",
       "4          c_1_1_5   Lung5  lung5_rep1  lung5_rep1_1      myeloid   \n",
       "5          c_1_1_7   Lung5  lung5_rep1  lung5_rep1_1      myeloid   \n",
       "7          c_1_1_9   Lung5  lung5_rep1  lung5_rep1_1  plasmablast   \n",
       "...            ...     ...         ...           ...          ...   \n",
       "81230  c_8_20_3998  Lung13      lung13     lung13_20    NK/T cell   \n",
       "81231  c_8_20_3999  Lung13      lung13     lung13_20    NK/T cell   \n",
       "81232  c_8_20_4000  Lung13      lung13     lung13_20    NK/T cell   \n",
       "81233  c_8_20_4001  Lung13      lung13     lung13_20      myeloid   \n",
       "81235  c_8_20_4004  Lung13      lung13     lung13_20      myeloid   \n",
       "\n",
       "                             niche cell_type_original  n_counts  \\\n",
       "2                           stroma        endothelial      74.0   \n",
       "3                           stroma                pDC      60.0   \n",
       "4                           stroma         macrophage      52.0   \n",
       "5      plasmablast-enriched stroma         macrophage     419.0   \n",
       "7      plasmablast-enriched stroma        plasmablast     249.0   \n",
       "...                            ...                ...       ...   \n",
       "81230                  macrophages       T CD4 memory      80.0   \n",
       "81231                       immune       T CD4 memory     135.0   \n",
       "81232                       immune       T CD4 memory      74.0   \n",
       "81233                  macrophages         macrophage     102.0   \n",
       "81235                       immune         macrophage      69.0   \n",
       "\n",
       "      mapping_entity  FYN_ligand_receptor_GP  ...  Add-on_7_GP  Add-on_19_GP  \\\n",
       "2          reference                0.681425  ...   -17.489004    -14.936492   \n",
       "3          reference                0.273680  ...   -13.203317      8.522111   \n",
       "4          reference                0.221490  ...   -16.162962     11.373133   \n",
       "5          reference                0.427213  ...   -22.170458     22.927443   \n",
       "7          reference                0.568128  ...   -26.380348    -14.176794   \n",
       "...              ...                     ...  ...          ...           ...   \n",
       "81230          query                1.114208  ...   -88.054962     17.349173   \n",
       "81231          query                0.136362  ...   -51.935818      6.413656   \n",
       "81232          query                0.434632  ...   -64.320480      5.044941   \n",
       "81233          query                0.992373  ...   -91.116089     28.286760   \n",
       "81235          query                0.242673  ...  -105.144196     24.266064   \n",
       "\n",
       "       Add-on_31_GP  Add-on_55_GP  Add-on_70_GP  Add-on_82_GP  Add-on_87_GP  \\\n",
       "2          3.799693    -39.409985      2.128621    -12.063638     18.894939   \n",
       "3         -2.774563    -15.451062     -3.280380    -37.737724     -5.105931   \n",
       "4         -5.187509    -19.356821     -3.515104    -35.349342     -1.645619   \n",
       "5         -9.017243      0.945892     -5.096985     22.844490     19.639498   \n",
       "7          3.803596     10.833162     -5.772773     22.477608     40.085487   \n",
       "...             ...           ...           ...           ...           ...   \n",
       "81230    -17.536432    -23.746794     -6.791336    -35.896069     -2.281905   \n",
       "81231    -19.044233    -12.997680     -1.418877    -40.414978     -5.716311   \n",
       "81232    -10.139482    -21.754337     -3.457257    -13.774080     -7.491361   \n",
       "81233    -24.224100    -36.481979     -8.841825      0.247141     -2.792174   \n",
       "81235    -29.696901    -41.929630     -8.734715    -19.529289      4.879042   \n",
       "\n",
       "       Add-on_95_GP  Add-on_98_GP  latent_leiden_0.4  \n",
       "2          5.632041    -39.309853                  3  \n",
       "3         10.302983     -0.375488                  0  \n",
       "4         10.553545     -1.178124                  0  \n",
       "5          2.279819    -31.737906                  3  \n",
       "7        -23.397999    -33.345215                  3  \n",
       "...             ...           ...                ...  \n",
       "81230     -2.327273      7.340417                  2  \n",
       "81231      5.811738      8.726611                  2  \n",
       "81232     -3.326872     10.765855                  2  \n",
       "81233     -1.966371      3.307793                  2  \n",
       "81235     -9.399879    -15.149623                  2  \n",
       "\n",
       "[702199 rows x 1018 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9b9a4f9-c6f1-4b84-a8e6-3e65b211faa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get gp scores of all gps (active & inactive)\n",
    "latent, _ = model.get_latent_representation(\n",
    "    adata=model.adata,\n",
    "    counts_key='counts',\n",
    "    adj_key=adj_key,\n",
    "    cat_covariates_keys=[condition_key],\n",
    "    only_active_gps=False,\n",
    "    return_mu_std=True,\n",
    "    node_batch_size=model.node_batch_size_)\n",
    "\n",
    "# Store gp scores, spatial coords and spatial neighbor graph in new adata\n",
    "gp_adata = ad.AnnData(latent)\n",
    "gp_adata.obs_names = model.adata.obs_names\n",
    "gp_adata.var_names = model.adata.uns[gp_names_key]\n",
    "gp_adata.obs[condition_key] = model.adata.obs[condition_key]\n",
    "gp_adata.obsp[adj_key] = model.adata.obsp[adj_key]\n",
    "gp_adata.obsm[spatial_key] = model.adata.obsm[spatial_key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17657dcc-5662-4fef-a0d4-eb2b605b55df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute spatial autocorrelation score Moran I of gene programs\n",
    "# (higher score -> positive spatial autocorrelation)\n",
    "sq.gr.spatial_autocorr(gp_adata,\n",
    "                       mode=\"moran\",\n",
    "                       n_perms=100,\n",
    "                       n_jobs=1)\n",
    "\n",
    "# Add score to summary df\n",
    "gp_summary_df[\"gp_moranI\"] = gp_adata.uns[\"moranI\"].loc[\n",
    "    gp_summary_df[\"gp_name\"].values.tolist(), \"I\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b059064-73c3-4cb5-8c1d-048662918877",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot gp with highest moran I score\n",
    "sq.pl.spatial_scatter(\n",
    "    gp_adata[gp_adata.obs[condition_key] == samples[0]],\n",
    "    shape=None,\n",
    "    color=[gp_summary_df[[\"gp_name\", \"gp_moranI\"]].sort_values(by=\"gp_moranI\", ascending=False).head(1)[\"gp_name\"].values[0]],\n",
    "    size=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ba5ad9a-f7e7-4563-b5d0-4c7c3cde2f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot gp with lowest moran I score\n",
    "sq.pl.spatial_scatter(\n",
    "    gp_adata[gp_adata.obs[condition_key] == samples[0]],\n",
    "    shape=None,\n",
    "    color=[gp_summary_df[[\"gp_name\", \"gp_moranI\"]].sort_values(by=\"gp_moranI\", ascending=False).tail(1)[\"gp_name\"].values[0]],\n",
    "    size=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9e5d9e8-2d0f-4fa9-91cf-5e7acd23d9ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine highly variable gene programs\n",
    "hvgp_threshs = [0.1]\n",
    "\n",
    "hvgp_list = []\n",
    "hvgp_cols = []\n",
    "for hvgp_thresh in hvgp_threshs:\n",
    "\n",
    "    sc.pp.highly_variable_genes(\n",
    "        gp_adata,\n",
    "        n_top_genes=int(len(gp_adata.var) * hvgp_thresh),\n",
    "        flavor=\"seurat\",\n",
    "        # batch_key=condition_key,\n",
    "        subset=False)\n",
    "\n",
    "    hvgp_list.append(gp_adata[:, gp_adata.var[\"highly_variable\"]].var_names.tolist())\n",
    "    \n",
    "for i, hvgp_thresh in enumerate(hvgp_threshs):\n",
    "    gp_summary_df[f\"hvgp_{hvgp_thresh}_thresh\"] = gp_summary_df[\"gp_name\"].isin(hvgp_list[i])\n",
    "    hvgp_cols.append(f\"hvgp_{hvgp_thresh}_thresh\")\n",
    "    \n",
    "gp_summary_df.loc[:, \"gp_category\"] = \"inactive\"\n",
    "gp_summary_df.loc[\n",
    "    gp_summary_df[\"gp_name\"].isin(active_gps), \"gp_category\"] = \"active\"\n",
    "gp_summary_df.loc[\n",
    "    gp_summary_df[\"gp_name\"].isin(hvgp_list[0]), \"gp_category\"] = \"highly_variable\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a11a0189-92cf-4017-9122-0f2cd85b4a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve genes, gene importances, gene ranks, and mean gene expr of target and source genes respectively\n",
    "gp_target_genes_df = gp_summary_df[\n",
    "    [\"gp_name\",\n",
    "     \"gp_active\",\n",
    "     \"gp_category\",\n",
    "     \"gp_moranI\",\n",
    "     \"gp_target_genes\",\n",
    "     \"gp_target_genes_importances\"] +\n",
    "     hvgp_cols].explode(\n",
    "        [\"gp_target_genes\",\n",
    "         \"gp_target_genes_importances\"]).rename(\n",
    "            columns={\"gp_target_genes\": \"target_gene\",\n",
    "                     \"gp_target_genes_importances\": \"target_gene_importance\"}).dropna()\n",
    "gp_target_genes_df[\"target_gene_rank\"] = gp_target_genes_df.groupby(\"gp_name\").cumcount() + 1\n",
    "gp_target_genes_df[\"target_gene_importance\"] = gp_target_genes_df[\"target_gene_importance\"].astype(np.float32)\n",
    "gp_target_genes_df[\"target_gene_mean_log_norm_gene_expr\"] = np.array(\n",
    "    model.adata[:, gp_target_genes_df[\"target_gene\"]].layers[counts_key].mean(0)).flatten()\n",
    "\n",
    "gp_source_genes_df = gp_summary_df[\n",
    "    [\"gp_name\",\n",
    "     \"gp_active\",\n",
    "     \"gp_category\",\n",
    "     \"gp_moranI\",\n",
    "     \"gp_source_genes\",\n",
    "     \"gp_source_genes_importances\"] +\n",
    "     hvgp_cols].explode(\n",
    "        [\"gp_source_genes\",\n",
    "         \"gp_source_genes_importances\"]).rename(\n",
    "            columns={\"gp_source_genes\": \"source_gene\",\n",
    "                     \"gp_source_genes_importances\": \"source_gene_importance\"}).dropna()\n",
    "gp_source_genes_df[\"source_gene_rank\"] = gp_source_genes_df.groupby(\"gp_name\").cumcount() + 1\n",
    "gp_source_genes_df[\"source_gene_importance\"] = gp_source_genes_df[\"source_gene_importance\"].astype(np.float32)\n",
    "gp_source_genes_df[\"source_gene_mean_log_norm_gene_expr\"] = np.array(\n",
    "    model.adata[:, gp_source_genes_df[\"source_gene\"]].layers[counts_key].mean(0)).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11041283-be7d-4b22-aa7e-0270d0c51bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute spatial autocorrelation score Moran I of genes\n",
    "# (higher score -> positive spatial autocorrelation)\n",
    "genes = list(set(list(gp_target_genes_df[\"target_gene\"]) + list(gp_source_genes_df[\"source_gene\"])))\n",
    "\n",
    "sq.gr.spatial_autocorr(model.adata,\n",
    "                       mode=\"moran\",\n",
    "                       genes=genes,\n",
    "                       n_perms=100,\n",
    "                       n_jobs=1)\n",
    "\n",
    "target_genes_moranI_scores = []\n",
    "for gene in gp_target_genes_df[\"target_gene\"]:\n",
    "    target_genes_moranI_scores.append(\n",
    "        model.adata.uns[\"moranI\"].loc[gene, \"I\"])\n",
    "gp_target_genes_df[\"gene_moranI\"] = target_genes_moranI_scores\n",
    "\n",
    "source_genes_moranI_scores = []\n",
    "for gene in gp_source_genes_df[\"source_gene\"]:\n",
    "    source_genes_moranI_scores.append(\n",
    "        model.adata.uns[\"moranI\"].loc[gene, \"I\"])\n",
    "gp_source_genes_df[\"gene_moranI\"] = source_genes_moranI_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a0f2bcc-c42d-4d76-b687-49f2565a5b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create summary plots over all gps, active gps, and inactive gps \n",
    "# with fixed n_top_genes\n",
    "n_top_genes = 3\n",
    "use_only_active_gps = False\n",
    "    \n",
    "gp_target_genes_df_top_genes = gp_target_genes_df[\n",
    "    gp_target_genes_df[\"target_gene_rank\"] <= n_top_genes]\n",
    "gp_source_genes_df_top_genes = gp_source_genes_df[\n",
    "    gp_source_genes_df[\"source_gene_rank\"] <= n_top_genes]\n",
    "\n",
    "all_gp_target_genes_df_top_genes = gp_target_genes_df_top_genes\n",
    "active_gp_target_genes_df_top_genes = gp_target_genes_df_top_genes[\n",
    "    gp_target_genes_df_top_genes[\"gp_active\"]]\n",
    "inactive_gp_target_genes_df_top_genes = gp_target_genes_df_top_genes[\n",
    "    ~gp_target_genes_df_top_genes[\"gp_active\"]]\n",
    "\n",
    "all_gp_source_genes_df_top_genes = gp_source_genes_df_top_genes\n",
    "active_gp_source_genes_df_top_genes = gp_source_genes_df_top_genes[\n",
    "    gp_source_genes_df_top_genes[\"gp_active\"]]\n",
    "inactive_gp_source_genes_df_top_genes = gp_source_genes_df_top_genes[\n",
    "    ~gp_source_genes_df_top_genes[\"gp_active\"]]\n",
    "\n",
    "# Plot distribution of gene importances\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(10, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Gene Importance Distributions\")\n",
    "sns.histplot(all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "             ax=axs[0],\n",
    "             bins=20)\n",
    "axs[0].set(title=\"Target Genes (All GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Target Gene Count\")\n",
    "sns.histplot(all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "             ax=axs[1],\n",
    "             bins=20)\n",
    "axs[1].set(title=\"Source Genes (All GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Source Gene Count\")\n",
    "sns.histplot(active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "             ax=axs[2],\n",
    "             bins=20)\n",
    "axs[2].set(title=\"Target Genes (Active GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Target Gene Count\")\n",
    "sns.histplot(active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "             ax=axs[3],\n",
    "             bins=20)\n",
    "axs[3].set(title=\"Source Genes (Active GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Source Gene Count\")\n",
    "sns.histplot(inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "             ax=axs[4],\n",
    "             bins=20)\n",
    "axs[4].set(title=\"Target Genes (Inactive GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Target Gene Count\")\n",
    "sns.histplot(inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "             ax=axs[5],\n",
    "             bins=20)\n",
    "axs[5].set(title=\"Source Genes (Inactive GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Source Gene Count\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()\n",
    "\n",
    "# Plot relationship between gene importances and log-normalized gene expression\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(20, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between NicheCompass Gene Importances \"\n",
    "             \"and Mean Log-Normalized Gene Expression of Genes\")\n",
    "sns.regplot(x=all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=all_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[0])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "    y=all_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"])\n",
    "all_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[0].set_title(f\"Target Genes (All GPs) - R2: {round(all_gp_target_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[0].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "sns.regplot(x=all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=all_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[1])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "    y=all_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"])\n",
    "all_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[1].set_title(f\"Source Genes (All GPs) - R2: {round(all_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[1].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "sns.regplot(x=active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=active_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[2])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "    y=active_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"])\n",
    "active_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[2].set_title(f\"Target Genes (Active GPs) - R2: {round(active_gp_target_genes_r_squared, 4)};  \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[2].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "sns.regplot(x=active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=active_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[3])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "    y=active_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"])\n",
    "active_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[3].set_title(f\"Source Genes (Active GPs) - R2: {round(active_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[3].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "if len(active_gps) != len(gp_summary_df):\n",
    "    sns.regplot(x=inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "                y=inactive_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"],\n",
    "                ax=axs[4])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "        y=inactive_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"])\n",
    "    inactive_gp_target_genes_r_squared = r_value ** 2\n",
    "    axs[4].set_title(f\"Target Genes (Inactive GPs) - R2: {round(inactive_gp_target_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[4].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "    sns.regplot(x=inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "                y=inactive_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"],\n",
    "                ax=axs[5])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "        y=inactive_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"])\n",
    "    inactive_gp_source_genes_r_squared = r_value ** 2\n",
    "    axs[5].set_title(f\"Source Genes (Inactive GPs) - R2: {round(inactive_gp_source_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[5].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()\n",
    "\n",
    "# Plot relationship between gene importances and Moran's I score\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(20, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between NicheCompass Gene Importances \"\n",
    "             \"and Moran's I Scores of Genes\")\n",
    "sns.regplot(x=all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=all_gp_target_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[0])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_target_genes_df_top_genes.dropna()[\"target_gene_importance\"],\n",
    "    y=all_gp_target_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "all_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[0].set_title(f\"Target Genes (All GPs) - R2: {round(all_gp_target_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[0].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "sns.regplot(x=all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=all_gp_source_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[1])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_source_genes_df_top_genes.dropna()[\"source_gene_importance\"],\n",
    "    y=all_gp_source_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "all_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[1].set_title(f\"Source Genes (All GPs) - R2: {round(all_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[1].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "sns.regplot(x=active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=active_gp_target_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[2])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_target_genes_df_top_genes.dropna()[\"target_gene_importance\"],\n",
    "    y=active_gp_target_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "active_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[2].set_title(f\"Target Genes (Active GPs) - R2: {round(active_gp_target_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[2].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "sns.regplot(x=active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=active_gp_source_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[3])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_source_genes_df_top_genes.dropna()[\"source_gene_importance\"],\n",
    "    y=active_gp_source_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "active_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[3].set_title(f\"Source Genes (Active GPs) - R2: {round(active_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[3].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "if len(active_gps) != len(gp_summary_df):\n",
    "    sns.regplot(x=inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "                y=inactive_gp_target_genes_df_top_genes[\"gene_moranI\"],\n",
    "                ax=axs[4])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_target_genes_df_top_genes.dropna()[\"target_gene_importance\"],\n",
    "        y=inactive_gp_target_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "    inactive_gp_target_genes_r_squared = r_value ** 2\n",
    "    axs[4].set_title(f\"Target Genes (Inactive GPs) - R2: {round(inactive_gp_target_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[4].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Moran's I Score\")\n",
    "    sns.regplot(x=inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "                y=inactive_gp_source_genes_df_top_genes[\"gene_moranI\"],\n",
    "                ax=axs[5])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_source_genes_df_top_genes.dropna()[\"source_gene_importance\"],\n",
    "        y=inactive_gp_source_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "    inactive_gp_source_genes_r_squared = r_value ** 2\n",
    "    axs[5].set_title(f\"Source Genes (Inactive GPs) - R2: {round(inactive_gp_source_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[5].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Moran's I Score\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7afeed9-8b5a-4e20-8734-ac728d51122c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot relationship between gp moran scores and importance weighted gene moran scores\n",
    "gp_target_genes_moran_df = gp_target_genes_df.groupby(\"gp_name\").apply(\n",
    "    lambda x: pd.Series({\"weighted_gene_moranI\": (\n",
    "        x[\"target_gene_importance\"] * x[\"gene_moranI\"]).sum()}))\n",
    "gp_target_genes_moran_df.reset_index(inplace=True)\n",
    "gp_target_genes_moran_df = pd.merge(gp_target_genes_moran_df,\n",
    "                                    gp_summary_df[[\"gp_name\", \"gp_active\", \"gp_moranI\"]],\n",
    "                                    on=\"gp_name\")\n",
    "\n",
    "gp_source_genes_moran_df = gp_source_genes_df.groupby(\"gp_name\").apply(\n",
    "    lambda x: pd.Series({\"weighted_gene_moranI\": (\n",
    "        x[\"source_gene_importance\"] * x[\"gene_moranI\"]).sum()}))\n",
    "gp_source_genes_moran_df.reset_index(inplace=True)\n",
    "gp_source_genes_moran_df = pd.merge(gp_source_genes_moran_df,\n",
    "                                    gp_summary_df[[\"gp_name\", \"gp_active\", \"gp_moranI\"]],\n",
    "                                    on=\"gp_name\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(15, 5))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between GP Moran's I Score and \\n\"\n",
    "             \"Importance Weighted Gene Moran's I Scores\")\n",
    "sns.scatterplot(data=gp_target_genes_moran_df,\n",
    "                x=gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "                y=gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                hue=gp_target_genes_moran_df[\"gp_active\"],\n",
    "                ax=axs[0])\n",
    "axs[0].set(title=\"Target Genes\",\n",
    "           xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.scatterplot(data=gp_source_genes_moran_df,\n",
    "                x=gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "                y=gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                hue=gp_source_genes_moran_df[\"gp_active\"],\n",
    "                ax=axs[1])\n",
    "axs[1].set(title=\"Source Genes\",\n",
    "           xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "plt.subplots_adjust(wspace=0.35, top=0.8)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9aa9915-c406-4fcc-87ee-e91733998a5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot relationship between gp moran scores and importance weighted gene moran scores\n",
    "# across all gps, active gps, and inactive gps\n",
    "\n",
    "all_gp_target_genes_moran_df = gp_target_genes_moran_df\n",
    "active_gp_target_genes_moran_df = gp_target_genes_moran_df[\n",
    "    gp_target_genes_moran_df[\"gp_active\"]]\n",
    "inactive_gp_target_genes_moran_df = gp_target_genes_moran_df[\n",
    "    ~gp_target_genes_moran_df[\"gp_active\"]]\n",
    "\n",
    "all_gp_source_genes_moran_df = gp_source_genes_moran_df\n",
    "active_gp_source_genes_moran_df = gp_source_genes_moran_df[\n",
    "    gp_source_genes_moran_df[\"gp_active\"]]\n",
    "inactive_gp_source_genes_moran_df = gp_source_genes_moran_df[\n",
    "    ~gp_source_genes_moran_df[\"gp_active\"]]\n",
    "\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(20, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between GP Moran's I Score and \\n\"\n",
    "             \"Importance Weighted Gene Moran's I Scores\")\n",
    "sns.regplot(x=all_gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "            y=all_gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[0])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_target_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=all_gp_target_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[0].set_title(f\"Target Genes (All GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[0].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.regplot(x=all_gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "            y=all_gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[1])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_source_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=all_gp_source_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[1].set_title(f\"Source Genes (All GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[1].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.regplot(x=active_gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "            y=active_gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[2])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_target_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=active_gp_target_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[2].set_title(f\"Target Genes (Active GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[2].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.regplot(x=active_gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "            y=active_gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[3])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_source_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=active_gp_source_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[3].set_title(f\"Source Genes (Active GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[3].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "if len(active_gps) != len(gp_summary_df):\n",
    "    sns.regplot(x=inactive_gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "                y=inactive_gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                ax=axs[4])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_target_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "        y=inactive_gp_target_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "    r_squared = r_value ** 2\n",
    "    axs[4].set_title(f\"Target Genes (Inactive GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[4].set(xlabel=\"GP Moran's I Score\",\n",
    "               ylabel=\"Weighted Gene Moran's I Score\")\n",
    "    sns.regplot(x=inactive_gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "                y=inactive_gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                ax=axs[5])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_source_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "        y=inactive_gp_source_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "    r_squared = r_value ** 2\n",
    "    axs[5].set_title(f\"Source Genes (Inactive GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[5].set(xlabel=\"GP Moran's I Score\",\n",
    "               ylabel=\"Weighted Gene Moran's I Score\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3cf43ae-ad12-4852-93c1-8a19a39a0e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_genes_weighted_moran_df = pd.DataFrame(columns=[\"GP Type\", \"GP Category\", \"Weighted Mean Moran's I Score\"])\n",
    "source_genes_weighted_moran_df = pd.DataFrame(columns=[\"GP Type\", \"GP Category\", \"Weighted Mean Moran's I Score\"])\n",
    "\n",
    "for gp_type, gp_type_identifier in zip([\"OmniPath\", \"NicheNet\", \"MEBOCOST\"],\n",
    "                                       [\"receptor\", \"target\", \"metabolite\"]):\n",
    "    gp_type_target_genes_df = gp_target_genes_df[gp_target_genes_df[\"gp_name\"].str.contains(gp_type_identifier)]\n",
    "    gp_type_source_genes_df = gp_source_genes_df[gp_source_genes_df[\"gp_name\"].str.contains(gp_type_identifier)]\n",
    "\n",
    "    target_genes_weighted_moran_sub_df = gp_type_target_genes_df.groupby(\"gp_category\").apply(\n",
    "        lambda x: pd.Series({\"Weighted Mean Moran's I Score\": (\n",
    "            x[\"target_gene_importance\"] * x[\"gene_moranI\"]).sum() / x[\"target_gene_importance\"].sum()}))\n",
    "\n",
    "    #new_row = {\"Weighted Mean Moran's I Score\": gp_type_target_genes_df[\"moranI\"].mean()}\n",
    "    #index_label = \"avg_across_all_gps\"\n",
    "    #new_df = pd.DataFrame(new_row, index=[index_label])\n",
    "    #target_genes_weighted_moran_sub_df = pd.concat([target_genes_weighted_moran_sub_df,\n",
    "    #                                              new_df])\n",
    "    target_genes_weighted_moran_sub_df[\"GP Type\"] = gp_type\n",
    "    target_genes_weighted_moran_sub_df[\"GP Category\"] = target_genes_weighted_moran_sub_df.index\n",
    "    target_genes_weighted_moran_sub_df.reset_index(inplace=True)\n",
    "    \n",
    "    target_genes_weighted_moran_df = pd.concat([target_genes_weighted_moran_df,\n",
    "                                                target_genes_weighted_moran_sub_df])\n",
    "\n",
    "    source_genes_weighted_moran_sub_df = gp_type_source_genes_df.groupby(\"gp_category\").apply(\n",
    "        lambda x: pd.Series({\"Weighted Mean Moran's I Score\": (\n",
    "            x[\"source_gene_importance\"] * x[\"gene_moranI\"]).sum() / x[\"source_gene_importance\"].sum()}))\n",
    "\n",
    "    #new_row = {\"Weighted Mean Moran's I Score\": gp_type_source_genes_df[\"moranI\"].mean()}\n",
    "    #index_label = \"avg_across_all_gps\"\n",
    "    #new_df = pd.DataFrame(new_row, index=[index_label])\n",
    "    #source_genes_weighted_moran_sub_df = pd.concat([source_genes_weighted_moran_sub_df,\n",
    "    #                                              new_df])\n",
    "    source_genes_weighted_moran_sub_df[\"GP Type\"] = gp_type\n",
    "    source_genes_weighted_moran_sub_df[\"GP Category\"] = source_genes_weighted_moran_sub_df.index\n",
    "    source_genes_weighted_moran_sub_df.reset_index(inplace=True)\n",
    "    \n",
    "    source_genes_weighted_moran_df = pd.concat([source_genes_weighted_moran_df,\n",
    "                                                source_genes_weighted_moran_sub_df])\n",
    "    \n",
    "# Plot relationship between gene importances and Moran's I score\n",
    "fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "plt.suptitle(\"Mean Moran's I Score Weighted by NicheCompass Gene Importances \")\n",
    "sns.stripplot(target_genes_weighted_moran_df,\n",
    "              x=\"GP Type\",\n",
    "              y=\"Weighted Mean Moran's I Score\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax1)\n",
    "ax1.set_title(\"Target Genes\",\n",
    "              fontsize=14)\n",
    "ax1.set(xlabel=\"GP Type\",\n",
    "        ylabel=\"Weighted Mean Moran's I Score\")\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles,\n",
    "                 labels,\n",
    "                 loc=\"center left\",\n",
    "                 bbox_to_anchor=(0.98, 0.5))\n",
    "ax1.get_legend().remove()\n",
    "sns.stripplot(source_genes_weighted_moran_df,\n",
    "              x=\"GP Type\",\n",
    "              y=\"Weighted Mean Moran's I Score\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax2)\n",
    "ax2.set_title(\"Source Genes\",\n",
    "              fontsize=14)\n",
    "ax2.set(xlabel=\"GP Type\",\n",
    "        ylabel=\"Weighted Mean Moran's I Score\")\n",
    "ax2.get_legend().remove()\n",
    "plt.subplots_adjust(wspace=0.35, top=0.85)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "199ffcf4-0131-4603-9387-85d89a00ebbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create summary plots with variable n_top_genes\n",
    "target_genes_corr_df = pd.DataFrame(columns=[\"Number of Top Genes\", \"GP Category\", \"Pearson Correlation\"])\n",
    "source_genes_corr_df = pd.DataFrame(columns=[\"Number of Top Genes\", \"GP Category\", \"Pearson Correlation\"])\n",
    "\n",
    "for n_top_genes in [5, 10, 20, 50, 100]:\n",
    "    gp_target_genes_df_top_genes = gp_target_genes_df[gp_target_genes_df[\"target_gene_rank\"] <= n_top_genes]\n",
    "    gp_source_genes_df_top_genes = gp_source_genes_df[gp_source_genes_df[\"source_gene_rank\"] <= n_top_genes]\n",
    "    \n",
    "    for gp_category in gp_target_genes_df_top_genes[\"gp_category\"].unique().tolist():\n",
    "        tmp_cat_df = gp_target_genes_df_top_genes.loc[gp_target_genes_df_top_genes[\"gp_category\"] == gp_category]\n",
    "        pearsonr = tmp_cat_df[\"target_gene_importance\"].corr(tmp_cat_df[\"gene_moranI\"])\n",
    "        temp_row_dict = {\"Number of Top Genes\": n_top_genes,\n",
    "                         \"GP Category\": gp_category,\n",
    "                         \"Pearson Correlation\": pearsonr}\n",
    "        target_genes_corr_df = target_genes_corr_df.append(temp_row_dict, ignore_index=True)\n",
    "        \n",
    "        tmp_cat_df = gp_source_genes_df_top_genes[gp_source_genes_df_top_genes[\"gp_category\"] == gp_category]\n",
    "        pearsonr = tmp_cat_df[\"source_gene_importance\"].corr(tmp_cat_df[\"gene_moranI\"])\n",
    "        temp_row_dict = {\"Number of Top Genes\": n_top_genes,\n",
    "                         \"GP Category\": gp_category,\n",
    "                         \"Pearson Correlation\": pearsonr}\n",
    "        source_genes_corr_df = source_genes_corr_df.append(temp_row_dict, ignore_index=True)\n",
    "        \n",
    "# Plot relationship between gene importances and Moran's I score\n",
    "fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "plt.suptitle(\"Pearson Correlation between NicheCompass Gene Importances \"\n",
    "             \"and Moran's I Scores of Genes\")\n",
    "sns.stripplot(target_genes_corr_df,\n",
    "              x=\"Number of Top Genes\",\n",
    "              y=\"Pearson Correlation\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax1)\n",
    "ax1.set_title(\"Target Genes\",\n",
    "              fontsize=14)\n",
    "ax1.set(xlabel=\"Number of Top Genes\",\n",
    "        ylabel=\"Pearson Correlation\")\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles,\n",
    "                 labels,\n",
    "                 loc=\"center left\",\n",
    "                 bbox_to_anchor=(0.98, 0.5))\n",
    "ax1.get_legend().remove()\n",
    "sns.stripplot(source_genes_corr_df,\n",
    "              x=\"Number of Top Genes\",\n",
    "              y=\"Pearson Correlation\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax2)\n",
    "ax2.set_title(\"Source Genes\",\n",
    "              fontsize=14)\n",
    "ax2.set(xlabel=\"Number of Top Genes\",\n",
    "        ylabel=\"Pearson Correlation\")\n",
    "ax2.get_legend().remove()\n",
    "plt.subplots_adjust(wspace=0.35, top=0.85)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67658a0a-edbb-4e7b-836f-e4e4808246de",
   "metadata": {},
   "source": [
    "### 3.3 Latent Cluster Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ae03333-6864-4c0c-b714-82da734bac39",
   "metadata": {},
   "source": [
    "#### 3.3.1 One Cluster (All) vs Rest of Tissue Differential GP Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d11e6d77-6bcb-49b9-8c59-fb4a55b96116",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cats = None\n",
    "log_bayes_factor_thresh = 2.3 # 2.3 strong threshold; 4.6 decisive threshold (https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "title = f\"NicheCompass Latent Cluster Enriched Gene Programs Log Bayes Factor {log_bayes_factor_thresh}\"\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            f\"latent_clusters_all_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_heatmap.pdf\"\n",
    "\n",
    "# Run differential gp testing\n",
    "enriched_gps = model.run_differential_gp_tests(\n",
    "    cat_key=latent_cluster_key,\n",
    "    selected_cats=selected_cats,\n",
    "    comparison_cats=\"rest\",\n",
    "    log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "# Plot heatmap of enriched gps\n",
    "ax = sc.pl.heatmap(model.adata,\n",
    "                    enriched_gps,\n",
    "                    show_gene_labels=True,\n",
    "                    groupby=latent_cluster_key,\n",
    "                    dendrogram=True,\n",
    "                    swap_axes=True,\n",
    "                    figsize=(model.adata.obs[latent_cluster_key].nunique() * 1.5,\n",
    "                             len(enriched_gps) / 2),\n",
    "                    save=save_fig)\n",
    "if save_fig:\n",
    "    shutil.move(\"figures/heatmap.pdf\", file_path)\n",
    "    os.rmdir(\"figures\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5d5bce1-3f24-4839-8091-191ae425d0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_file = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            f\"latent_clusters_all_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_summary.csv\"\n",
    "\n",
    "gp_summary_cols = [\"gp_name\",\n",
    "                   \"n_source_genes\",\n",
    "                   \"n_non_zero_source_genes\",\n",
    "                   \"n_target_genes\",\n",
    "                   \"n_non_zero_target_genes\",\n",
    "                   \"gp_source_genes\",\n",
    "                   \"gp_target_genes\",\n",
    "                   \"gp_source_genes_weights\",\n",
    "                   \"gp_target_genes_weights\",\n",
    "                   \"gp_source_genes_importances\",\n",
    "                   \"gp_target_genes_importances\"]\n",
    "if multimodal:\n",
    "    gp_summary_cols = gp_summary_cols + [\n",
    "        \"n_source_peaks\",\n",
    "        \"n_target_peaks\",\n",
    "        \"gp_source_peaks\",\n",
    "        \"gp_target_peaks\",\n",
    "        \"gp_source_peaks_weights\",\n",
    "        \"gp_target_peaks_weights\",\n",
    "        \"gp_source_peaks_importances\",\n",
    "        \"gp_target_peaks_importances\"]\n",
    "\n",
    "# Get summary of decisively enriched gene programs\n",
    "enriched_gp_summary_df = gp_summary_df[gp_summary_df[\"gp_name\"].isin(enriched_gps)]\n",
    "cat_dtype = pd.CategoricalDtype(categories=enriched_gps, ordered=True)\n",
    "enriched_gp_summary_df[\"gp_name\"] = enriched_gp_summary_df[\"gp_name\"].astype(cat_dtype)\n",
    "enriched_gp_summary_df = enriched_gp_summary_df.sort_values(by=\"gp_name\")\n",
    "enriched_gp_summary_df = enriched_gp_summary_df[gp_summary_cols]\n",
    "\n",
    "if save_file:\n",
    "    enriched_gp_summary_df.to_csv(f\"{file_path}\")\n",
    "else:\n",
    "    display(enriched_gp_summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7f3e8ec-2117-41f1-937b-52dadaf9f033",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figs = True\n",
    "\n",
    "for i in np.arange(0, len(enriched_gps), 10):\n",
    "    plot_label = f\"res_{latent_leiden_resolution}_\" \\\n",
    "                 f\"latent_clusters_all_vs_rest_log_bayes_factor_\" \\\n",
    "                 f\"{log_bayes_factor_thresh}\"\n",
    "\n",
    "    generate_enriched_gp_info_plots(\n",
    "        plot_label=plot_label,\n",
    "        model=model,\n",
    "        sample_key=sample_key,\n",
    "        differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "        cat_key=latent_cluster_key,\n",
    "        cat_palette=latent_cluster_colors,\n",
    "        n_top_enriched_gp_start_idx=i,\n",
    "        n_top_enriched_gp_end_idx=i+10,\n",
    "        feature_spaces=samples, # [\"latent\"]\n",
    "        n_top_genes_per_gp=5,\n",
    "        n_top_peaks_per_gp=(0 if not multimodal else 5),\n",
    "        scale_omics_ft=False,\n",
    "        save_figs=save_figs,\n",
    "        figure_folder_path=f\"{figure_folder_path}/\",\n",
    "        spot_size=spot_size)\n",
    "    \n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66421db4-1f1d-4522-8afe-3348c9bbc19d",
   "metadata": {},
   "source": [
    "#### 3.2.2 One Cluster (Selected) vs Rest of Tissue Differential GP Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c4c869c-87cf-4344-aee1-92df29b40440",
   "metadata": {},
   "outputs": [],
   "source": [
    "latent_groups = [\"3\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d06e0332-83e3-485c-9d67-c6d69be4a2be",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    title = f\"NicheCompass Latent Cluster Enriched Gene Programs\"\n",
    "    log_bayes_factor_thresh = 2.3 # 4.6 decisive threshold; 2.3 strong threshold (https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_clusters_{'_'.join(latent_groups)}_vs_rest_log_bayes_factor_\" \\\n",
    "                f\"{log_bayes_factor_thresh}_enriched_gps_heatmap.pdf\"\n",
    "\n",
    "    # Run differential gp testing\n",
    "    enriched_gps = model.run_differential_gp_tests(\n",
    "        cat_key=latent_cluster_key,\n",
    "        selected_cats=latent_groups,\n",
    "        comparison_cats=\"rest\",\n",
    "        log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "    # Plot heatmap of enriched gps\n",
    "    ax = sc.pl.heatmap(model.adata,\n",
    "                        enriched_gps,\n",
    "                        show_gene_labels=True,\n",
    "                        groupby=latent_cluster_key,\n",
    "                        dendrogram=True,\n",
    "                        swap_axes=True,\n",
    "                        figsize=(model.adata.obs[latent_cluster_key].nunique() * 1.5,\n",
    "                                 len(enriched_gps) / 2),\n",
    "                        save=save_fig)\n",
    "    if save_fig:\n",
    "        shutil.move(\"figures/heatmap.pdf\", file_path)\n",
    "        os.rmdir(\"figures\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c6ec0f-8dd9-4007-a789-e344db7e6ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    save_file = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_clusters_{'_'.join(latent_groups)}_vs_rest_log_bayes_factor_\" \\\n",
    "                f\"{log_bayes_factor_thresh}_enriched_gps_summary.csv\"\n",
    "\n",
    "    gp_summary_cols = [\"gp_name\",\n",
    "                       \"n_source_genes\",\n",
    "                       \"n_non_zero_source_genes\",\n",
    "                       \"n_target_genes\",\n",
    "                       \"n_non_zero_target_genes\",\n",
    "                       \"gp_source_genes\",\n",
    "                       \"gp_target_genes\",\n",
    "                       \"gp_source_genes_weights\",\n",
    "                       \"gp_target_genes_weights\",\n",
    "                       \"gp_source_genes_importances\",\n",
    "                       \"gp_target_genes_importances\"]\n",
    "    if multimodal:\n",
    "        gp_summary_cols = gp_summary_cols + [\n",
    "            \"n_source_peaks\",\n",
    "            \"n_target_peaks\",\n",
    "            \"gp_source_peaks\",\n",
    "            \"gp_target_peaks\",\n",
    "            \"gp_source_peaks_weights\",\n",
    "            \"gp_target_peaks_weights\",\n",
    "            \"gp_source_peaks_importances\",\n",
    "            \"gp_target_peaks_importances\"]\n",
    "\n",
    "    # Get summary of enriched gene programs\n",
    "    enriched_gp_summary_df = gp_summary_df[gp_summary_df[\"gp_name\"].isin(enriched_gps)]\n",
    "    cat_dtype = pd.CategoricalDtype(categories=enriched_gps, ordered=True)\n",
    "    enriched_gp_summary_df.loc[:, \"gp_name\"] = enriched_gp_summary_df[\"gp_name\"].astype(cat_dtype)\n",
    "    enriched_gp_summary_df = enriched_gp_summary_df.sort_values(by=\"gp_name\")\n",
    "    enriched_gp_summary_df = enriched_gp_summary_df[gp_summary_cols]\n",
    "\n",
    "    if save_file:\n",
    "        enriched_gp_summary_df.to_csv(f\"{file_path}\")\n",
    "    else:\n",
    "        display(enriched_gp_summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba1fd1ba-f399-4557-badc-5719b97226c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    save_figs = True\n",
    "\n",
    "    for i in np.arange(0, len(enriched_gps), 10):\n",
    "        plot_label = f\"res_{latent_leiden_resolution}_\" \\\n",
    "                    f\"latent_clusters_{'_'.join(latent_groups)}_vs_rest_log_bayes_factor_\" \\\n",
    "                    f\"{log_bayes_factor_thresh}\"\n",
    "\n",
    "        generate_enriched_gp_info_plots(\n",
    "            plot_label=plot_label,\n",
    "            model=model,\n",
    "            sample_key=sample_key,\n",
    "            differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "            cat_key=latent_cluster_key,\n",
    "            cat_palette=latent_cluster_colors,\n",
    "            n_top_enriched_gp_start_idx=i,\n",
    "            n_top_enriched_gp_end_idx=i+10,\n",
    "            feature_spaces=[\"p22\"], # [\"latent\"]\n",
    "            n_top_genes_per_gp=5,\n",
    "            n_top_peaks_per_gp=(0 if not multimodal else 5),\n",
    "            scale_omics_ft=False,\n",
    "            save_figs=save_figs,\n",
    "            figure_folder_path=f\"{figure_folder_path}/\",\n",
    "            spot_size=spot_size)\n",
    "\n",
    "        gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64e6685-60bf-4fa6-8c44-0f1dcd548433",
   "metadata": {},
   "source": [
    "### 3.4 Cell Type Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f962bae7-1b93-44eb-bb3d-11f450024c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = \"Gut tube\" # \"Forebrain/Midbrain/Hindbrain\", \"Gut tube\"\n",
    "cell_type_fmt = cell_type.replace('/', '_').lower()\n",
    "cell_type_differential_gp_scores_key = f\"{cell_type}_differential_gp_scores\"\n",
    "cell_type_latent_cluster_differential_gp_scores_key = f\"{cell_type}_latentn_cluster_differential_gp_scores\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0035471a-6afe-4979-99af-c9fbcd5fc4ad",
   "metadata": {},
   "source": [
    "#### 3.4.1 Overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c6933a-7182-4458-a4f5-ff35ebc5e9c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot cell type in physical and latent space\n",
    "plot_physical_latent_for_cell_types(adata=model.adata,\n",
    "                                    cell_types=[cell_type],\n",
    "                                    sample_key=sample_key,\n",
    "                                    cell_type_key=cell_type_key,\n",
    "                                    cell_type_colors=cell_type_colors,\n",
    "                                    figure_folder_path=figure_folder_path,\n",
    "                                    save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044061e0-cde1-4361-b68c-4b45ea32f86a",
   "metadata": {},
   "source": [
    "#### 3.4.2 Cell Type vs Rest of Tissue Differential GP Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a372cba8-98e5-4391-b5df-af3b9c5f0d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cats = [cell_type]\n",
    "log_bayes_factor_thresh = 2.3 # 2.3 strong threshold; 4.6 decisive threshold (https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "title = f\"NicheCompass {cell_type} Enriched Gene Programs Log Bayes Factor {log_bayes_factor_thresh}\"\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/{cell_type_fmt}_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_dotplot.svg\"\n",
    "\n",
    "# Run differential gp testing\n",
    "enriched_gps = model.run_differential_gp_tests(\n",
    "    cat_key=cell_type_key,\n",
    "    selected_cats=selected_cats,\n",
    "    comparison_cats=\"rest\",\n",
    "    log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "sc.pl.heatmap(model.adata,\n",
    "              enriched_gps,\n",
    "              show_gene_labels=True,\n",
    "              groupby=cell_type_key,\n",
    "              dendrogram=True,\n",
    "              swap_axes=True,\n",
    "              figsize=(model.adata.obs[cell_type_key].nunique() * 1.5,\n",
    "                       len(enriched_gps) / 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35709e8a-c998-4685-859f-c3d2dfdbcb5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_file = True\n",
    "file_path = f\"{figure_folder_path}/{cell_type_fmt}_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_summary.csv\"\n",
    "\n",
    "# Get summary of enriched gene programs\n",
    "gp_summary_cols = [\"gp_name\",\n",
    "                   \"n_source_genes\",\n",
    "                   \"n_non_zero_source_genes\",\n",
    "                   \"n_target_genes\",\n",
    "                   \"n_non_zero_target_genes\",\n",
    "                   \"gp_source_genes\",\n",
    "                   \"gp_target_genes\",\n",
    "                   \"gp_source_genes_weights\",\n",
    "                   \"gp_target_genes_weights\",\n",
    "                   \"gp_source_genes_importances\",\n",
    "                   \"gp_target_genes_importances\"]\n",
    "if multimodal:\n",
    "    gp_summary_cols = gp_summary_cols + [\n",
    "        \"n_source_peaks\",\n",
    "        \"n_target_peaks\",\n",
    "        \"gp_source_peaks\",\n",
    "        \"gp_target_peaks\",\n",
    "        \"gp_source_peaks_weights\",\n",
    "        \"gp_target_peaks_weights\",\n",
    "        \"gp_source_peaks_importances\",\n",
    "        \"gp_target_peaks_importances\"]\n",
    "\n",
    "# Get summary of decisively enriched gene programs\n",
    "enriched_gp_summary_df = gp_summary_df[gp_summary_df[\"gp_name\"].isin(enriched_gps)]\n",
    "cat_dtype = pd.CategoricalDtype(categories=enriched_gps, ordered=True)\n",
    "enriched_gp_summary_df[\"gp_name\"] = enriched_gp_summary_df[\"gp_name\"].astype(cat_dtype)\n",
    "enriched_gp_summary_df = enriched_gp_summary_df.sort_values(by=\"gp_name\")\n",
    "enriched_gp_summary_df = enriched_gp_summary_df[gp_summary_cols]\n",
    "\n",
    "if save_file:\n",
    "    enriched_gp_summary_df.to_csv(f\"{file_path}\")\n",
    "else:\n",
    "    display(enriched_gp_summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b64eee9-4dc7-4731-83b3-ced0d2bc313f",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figs = True\n",
    "\n",
    "for i in np.arange(0, len(enriched_gps), 10):\n",
    "    plot_label = f\"{cell_type_fmt}_vs_rest_log_bayes_factor_\" \\\n",
    "                 f\"{log_bayes_factor_thresh}\"\n",
    "\n",
    "    generate_enriched_gp_info_plots(\n",
    "        plot_label=plot_label,\n",
    "        model=model,\n",
    "        sample_key=sample_key,\n",
    "        differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "        cat_key=cell_type_key,\n",
    "        cat_palette=cell_type_colors,\n",
    "        n_top_enriched_gp_start_idx=i,\n",
    "        n_top_enriched_gp_end_idx=i+10,\n",
    "        feature_spaces=samples, # [\"latent\"]\n",
    "        n_top_genes_per_gp=5,\n",
    "        n_top_peaks_per_gp=(0 if not multimodal else 5),\n",
    "        scale_omics_ft=False,\n",
    "        save_figs=save_figs,\n",
    "        figure_folder_path=f\"{figure_folder_path}/\",\n",
    "        spot_size=spot_size)\n",
    "    \n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb08e58-7866-4b2e-baf1-a4264c1a37b8",
   "metadata": {},
   "source": [
    "### 3.5 Save Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e2a7ac-8d25-4d2d-aa75-09a0841dd720",
   "metadata": {},
   "outputs": [],
   "source": [
    "if log_norm_omics_features:\n",
    "    # Log normalize counts for cellxgene server\n",
    "    sc.pp.normalize_total(model.adata, target_sum=1e4)\n",
    "    sc.pp.log1p(model.adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bc19412-c190-4e21-b30b-1bd8cb61b106",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store gp summary in adata\n",
    "gp_summary = model.get_gp_summary()\n",
    "for col in gp_summary.columns:\n",
    "    gp_summary[col] = gp_summary[col].astype(str)\n",
    "model.adata.uns[\"nichecompass_gp_summary\"] = gp_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3910b8-7d79-4f34-99bc-ba85bd0aa668",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.write(f\"{result_folder_path}/{dataset}_analysis.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cda8496-4830-48a9-aac6-f7cdfef78b54",
   "metadata": {},
   "outputs": [],
   "source": [
    "- Forebrain/Midbrain/Hindbrain: Regionally specific developing brain subtypes: separation into Rhombencephalon, Tegmentum, Mesencephalon, Prosencephalon\n",
    "- Mixed Mesenchymal Mesoderm: \"mixed mesenchymal mesoderm, represent \n",
    "a cell state rather than a defined cell type. Mesenchyme represents  a  state  in  which  cells  express  markers  characteristic  of  migratory   cells loosely dispersed within an extracellular matrix56. This strong  overriding transcriptional signature of mesenchyme, irrespective of  location,  makes  it  challenging  to  distinguish  which  cell  types  this   mixed  mesenchymal  mesoderm  population  represents  using  clas- sical scRNA-seq data\"\n",
    "\n",
    "We can identify distinct subpopulations that are spatially defined.\n",
    "Cardiomyocytes are spatially and morphologically distinct from other cell types, endothelium is interspersed and spread across the entire embryo space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ef7a9b4-fd75-4afc-8dbe-886df4118619",
   "metadata": {},
   "outputs": [],
   "source": [
    "f\"{result_folder_path}/{dataset}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be551e91-909b-4e04-9344-8f378e264eee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
