{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# NicheCompass Data Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 22.01.2023\n",
    "- **Date of Last Modification:** 25.08.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6bd6fb-6697-43bd-b3ec-68b061afe511",
   "metadata": {},
   "source": [
    "- In order to run this notebook, a trained model needs to be stored under f\"../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../utils\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import argparse\n",
    "import gc\n",
    "import os\n",
    "import random\n",
    "import shutil\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "import torch\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.pyplot import rc_context\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                                aggregate_obsp_matrix_per_cell_type,\n",
    "                                create_cell_type_chord_plot_from_df,\n",
    "                                create_new_color_dict,\n",
    "                                generate_enriched_gp_info_plots)\n",
    "\n",
    "from analysis_utils import (add_cell_type_latent_cluster_emphasis,\n",
    "                            add_sub_cell_type,\n",
    "                            compute_cell_type_latent_clusters,\n",
    "                            generate_gp_info_plots,\n",
    "                            plot_physical_latent_for_cell_types,\n",
    "                            plot_cell_type_latent_clusters,\n",
    "                            plot_latent,\n",
    "                            plot_category_in_latent_and_physical_space,\n",
    "                            store_top_gps_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9325589b-a7e4-4c31-97e7-8b7048248896",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "# dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "# dataset = \"vizgen_merfish_human_colon_cancer\"\n",
    "# dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "# dataset = \"vizgen_merfish_human_lung_cancer\"\n",
    "# dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "# dataset = \"starmap_plus_mouse_cns\"\n",
    "dataset = \"xenium_human_breast_cancer\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ac7b6d-4518-4356-bfdd-e141bf25d504",
   "metadata": {},
   "source": [
    "#### 1.2.1 Generic Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Model\n",
    "# AnnData keys\n",
    "adj_key = \"spatial_connectivities\"\n",
    "spatial_key = \"spatial\"\n",
    "sub_cell_type_key = \"sub_cell_type\"\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "active_gp_names_key = \"nichecompass_active_gp_names\"\n",
    "latent_key = \"nichecompass_latent\"\n",
    "mapping_entity_key = \"mapping_entity\"\n",
    "\n",
    "## Analysis\n",
    "differential_gp_test_results_key = \"nichecompass_differential_gp_test_results\"\n",
    "\n",
    "## Others\n",
    "random_seed = 0"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7be323-bd40-4fa0-a2d5-b8ddf5cc1c51",
   "metadata": {},
   "source": [
    "#### 1.2.2 Dataset-specific Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01210b2f-017e-43f6-b650-5cc732252f8c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "multimodal = False\n",
    "log_norm_omics_features = False\n",
    "cell_type_groups = []\n",
    "latent_groups = []\n",
    "\n",
    "if dataset == \"seqfish_mouse_organogenesis_imputed\":\n",
    "    load_timestamp = \"09082023_225056_10\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    counts_key = \"log_normalized_counts\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    dataset_str = \"seqFISH Mouse Organogenesis Imputed\"\n",
    "    multimodal = False\n",
    "    sample_key = \"sample\"\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.35\n",
    "    latent_groups = [\"16\", \"15\", \"4\", \"10\", \"5\", \"6\"]\n",
    "    spot_size = 0.05\n",
    "    cell_type_groups = []\n",
    "    multimodal = False\n",
    "elif dataset == \"starmap_plus_mouse_cns\":\n",
    "    cell_type_key = \"Main_molecular_cell_type\"\n",
    "    groups = \"Astrocytes\"\n",
    "    dataset_str = \"STARmap PLUS Mouse Central Nervous System\"\n",
    "    load_timestamp = \"20062023_155306_1\"\n",
    "    #load_timestamp = \"12062023_110439\"\n",
    "    #load_timestamp = \"20062023_164533_42\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.2\n",
    "    multimodal = False\n",
    "    spot_size=0.2\n",
    "    cell_type_groups = [\"Microglia\", \"Pericytes\"]\n",
    "elif dataset == \"visium_human_heart\":\n",
    "    load_timestamp = \"15052023_102158\"\n",
    "    cell_type_key = \"majority_cell_type\"\n",
    "    dataset_str = \"Visium Human Heart\"\n",
    "    sample_key = \"batch\"\n",
    "    samples = [\"batch1\", \"batch2\", \"batch3\"]\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.5\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    latent_cluster_spot_size = 200\n",
    "elif dataset == \"spatial_atac_rna_seq_mouse_brain_batch2\":\n",
    "    load_timestamp = \"09082023_145510_69\"\n",
    "    load_timestamp = \"10082023_091319_10\"\n",
    "    #load_timestamp = \"09082023_130425_21\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    cell_type_key = \"RNA_clusters\"\n",
    "    rna_cluster_key = \"RNA_clusters\"\n",
    "    atac_cluster_key = \"ATAC_clusters\"\n",
    "    dataset_str = \"Spatial ATAC-RNA-Seq Mouse Brain\"\n",
    "    sample_key = \"batch\"\n",
    "    samples = [\"p22\"]\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.23 # 0.3\n",
    "    latent_cluster_spot_size = 30\n",
    "    latent_groups = []\n",
    "    multimodal = True\n",
    "    spot_size = 30\n",
    "    cell_type_groups = []\n",
    "    latent_groups = [\"1\", \"6\", \"7\", \"5\", \"3\"]\n",
    "    log_norm_omics_features = True\n",
    "elif dataset == \"nanostring_cosmx_human_nsclc_filtered\":\n",
    "    load_timestamp = \"12062023_123411\"\n",
    "    load_timestamp = \"12062023_123012\"\n",
    "    load_timestamp = \"11072023_130002_1\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    dataset_str = \"Nanostring CosMx Human NSCLC\"\n",
    "    condition_key = \"batch\"\n",
    "    latent_leiden_resolution = 0.03\n",
    "    latent_cluster_spot_size = 60\n",
    "    sample_key = \"batch\"\n",
    "    multimodal = False\n",
    "    spot_size = 50\n",
    "    cell_type_key = \"cell_type\"\n",
    "elif dataset == \"vizgen_merfish_human_colon_cancer\":\n",
    "    load_timestamp = \"27062023_113037_2\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    dataset_str = \"Vizgen MERFISH Human Colon Cancer\"\n",
    "    condition_key = \"sample\"\n",
    "    sample_key = \"batch\"\n",
    "    samples = [\"batch1\", \"batch2\"]\n",
    "    latent_leiden_resolution = 0.5\n",
    "    latent_cluster_spot_size = 30\n",
    "elif dataset == \"vizgen_merfish_human_ovarian_cancer\":\n",
    "    load_timestamp = \"13062023_131002\"\n",
    "    load_timestamp = \"27062023_113037_2\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"Vizgen MERFISH Human Ovarian Cancer\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_type\"\n",
    "elif dataset == \"vizgen_merfish_human_lung_cancer\":\n",
    "    load_timestamp = \"13062023_131002\"\n",
    "    load_timestamp = \"19062023_140621_1\"\n",
    "    model_label = \"one-hop-norm_reference\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"Vizgen MERFISH Human Lung Cancer\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_type\"\n",
    "elif dataset == \"xenium_human_breast_cancer\":\n",
    "    load_timestamp = \"28082023_182430_3\" # 8 neighbors\n",
    "    # load_timestamp = \"20062023_120357_1\" # 4 neighbors\n",
    "    model_label = \"reference\"\n",
    "    latent_leiden_resolution = 0.2\n",
    "    latent_cluster_spot_size = 0.03\n",
    "    dataset_str = \"Xenium Human Breast Cancer\"\n",
    "    condition_key = \"batch\"\n",
    "    sample_key = \"batch\"\n",
    "    spot_size = 30\n",
    "    cell_type_key = \"cell_states\"\n",
    "    \n",
    "latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf229ca-ab66-4e52-a31e-8655878d290d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "924032d4-24a4-45cf-87e1-0eba197da0c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_folder_path = f\"../artifacts/{dataset}/figures/{model_label}/{load_timestamp}\"\n",
    "model_folder_path = f\"../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"\n",
    "result_folder_path = f\"../artifacts/{dataset}/results/{model_label}/{load_timestamp}\"\n",
    "gp_data_folder_path = \"../datasets/gp_data\" # gene program data\n",
    "srt_data_folder_path = \"../datasets/srt_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(result_folder_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cea1a5c-d075-4e51-90e5-90684bee88a6",
   "metadata": {},
   "source": [
    "## 2. Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d6ede1d-2cdf-4c81-9f56-e4745ae59e6c",
   "metadata": {},
   "source": [
    "### 2.1 Load Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe9e6c2-6ea1-49c7-a625-79e0386ac5bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if model_label == \"sample_integration_method_benchmarking\":\n",
    "    model_folder_path = f\"{model_folder_path}/run{run_number}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dad2ca6-7e34-48bc-89ef-4f994389a0f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "if multimodal:\n",
    "    model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                              adata=None,\n",
    "                              adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                              adata_atac=None,\n",
    "                              adata_atac_file_name=f\"{dataset}_{model_label}_atac.h5ad\",\n",
    "                              gp_names_key=gp_names_key)\n",
    "else:\n",
    "    # Load trained model\n",
    "    model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                              adata=None,\n",
    "                              adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                              gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d99a385-5769-4c6b-a7dc-62f40028620a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.uns[gp_names_key] = np.array([gp for gp in model.adata.uns[gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp for gp in model.adata.uns[active_gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[gp_names_key]])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[active_gp_names_key]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f662f73-455e-46c2-bc4b-ec25fb2f1d2b",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fc40859-6a4a-4bfa-8362-d28d4c1c2057",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Check number of active gene programs\n",
    "active_gps = model.get_active_gps()\n",
    "print(f\"Number of total gene programs: {len(model.adata.uns[gp_names_key])}.\")\n",
    "print(f\"Number of active gene programs: {len(active_gps)}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85cb0edb-1eb6-48b5-b56c-079501cb697d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gp_summary_df = model.get_gp_summary()\n",
    "gp_summary_df[gp_summary_df[\"gp_active\"] == True][0:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5170a527-58ea-42b0-859b-b582c69d7d14",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_active_gp_scores_to_obs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d61c1423-d79d-4e25-bfde-056db4106206",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04a9c809-4877-4e72-8159-fa2024b6baf1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa61dfbd-f31c-494a-bfbe-c14ce7fe02c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1eb19471-c581-4d1a-a581-10c4abd1259f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.heatmap(model.adata,\n",
    "              groupby=\"niche\",)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552770ad-5e55-4d10-a340-036f6c7120f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e55a4e88-b1c3-4a45-aa34-7193546c9000",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ede7740a-37ae-4826-9e81-6d16a6f95244",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed0c1fc8-c621-478e-90d2-6fc82995200d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gps = [\"Col22a1_ligand_receptor_target_gene_GP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4301e6e-e95a-4bf1-922d-ab933803fdf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obs[\"sample\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f9595a6-41d8-442b-83ee-b55cd3737327",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obs[\"Add-on_7_GP\"] = -1 * model.adata.obs[\"Add-on_7_GP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b65eed9b-37ae-4e25-86b0-e15bb2ebeecb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp = \"Add-on_32_GP\"\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "    color=gp,\n",
    "    color_map=\"RdBu\",\n",
    "    spot_size=spot_size,\n",
    "    title=f\"\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc62b16-cc95-496f-b26b-b5ca863a8462",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(\n",
    "    adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "    color=\"Foxf2\",\n",
    "    color_map=\"RdBu_r\",\n",
    "    spot_size=spot_size,\n",
    "    title=f\"\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1026f42d-58da-4ddf-a958-b863c8b4a7f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.stacked_violin(model.adata, var_names=['Wnt5a_ligand_receptor_target_gene_GP',\n",
    "                                             'IHH_ligand_receptor_GP',\n",
    "                                             'Furin_ligand_receptor_target_gene_GP',\n",
    "                                             'Choline_metabolite_enzyme_sensor_GP',\n",
    "                                             'Nodal_ligand_receptor_target_gene_GP',\n",
    "                                             'Lgals3_ligand_receptor_target_gene_GP',\n",
    "                                             'PDGFC_ligand_receptor_GP',\n",
    "                                             'Oxt_ligand_receptor_target_gene_GP'],swap_axes=True, groupby='latent_leiden_0.35',cmap='seismic')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df660162-5cde-4334-a0b6-5857c112328e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=gp,\n",
    "    color_map=\"RdBu_r\",\n",
    "    spot_size=spot_size,\n",
    "    title=f\"\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f0f4b54-7526-4c35-9cfb-11c1f8a0ae1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "value_mapping = {\"5\": \"Forebrain\",\n",
    "                 \"6\": \"Midbrain\",\n",
    "                 \"10\": \"Hindbrain\",\n",
    "                 \"15\": \"Floor Plate (Tegmentum)\"}\n",
    "category_colors = {'Forebrain': latent_cluster_colors[\"5\"],\n",
    "                   'Midbrain': latent_cluster_colors[\"6\"],\n",
    "                   'Hindbrain': latent_cluster_colors[\"10\"],\n",
    "                   'Floor Plate (Tegmentum)': latent_cluster_colors[\"15\"]}\n",
    "\n",
    "def map_values(value):\n",
    "    if value in value_mapping:\n",
    "        return value_mapping[value]\n",
    "    else:\n",
    "        return np.nan\n",
    "\n",
    "#model.adata.obs[\"Calca_ligand_receptor_target_gene_GP\"] = -1 * model.adata.obs[\"Calca_ligand_receptor_target_gene_GP\"]\n",
    "model.adata.obs['Brain Niches'] = model.adata.obs['latent_leiden_0.35'].apply(map_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a72a16-bdc1-4503-800f-5b8e13d946eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Brain\n",
    "ax = sns.scatterplot(x='Col22a1_ligand_receptor_target_gene_GP', # Pixna1_ligand_receptor_target_gene_GP\n",
    "                y='Calca_ligand_receptor_target_gene_GP',\n",
    "                hue='Brain Niches',\n",
    "                data=model.adata.obs,\n",
    "                palette=category_colors)\n",
    "ax.legend().set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e93de7c0-7243-42ad-8076-6e3d98adfaf8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Floor Plate (Tegmentum)\n",
    "ax = sns.scatterplot(x='Shh_ligand_receptor_target_gene_GP', # Pixna1_ligand_receptor_target_gene_GP\n",
    "                y='Calca_ligand_receptor_target_gene_GP',\n",
    "                hue='Brain Niches',\n",
    "                data=model.adata.obs,\n",
    "                palette=category_colors)\n",
    "ax.legend().set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54f79b88-db4d-493d-8819-6c48bf6b4a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Forebrain\n",
    "# Cldn14_ligand_receptor_target_gene_GP\n",
    "ax = sns.scatterplot(x='PCSK1N_ligand_receptor_GP', # \"Spon1_ligand_recetor_target_gene_GP\" PCSK1N_ligand_receptor_GP\n",
    "                y=\"DKK1_ligand_receptor_GP\", # 'Il12b_ligand_receptor_target_gene_GP' ''Il31_ligand_receptor_target_gene_GP''\n",
    "                hue='Brain Niches',\n",
    "                data=model.adata.obs,\n",
    "                palette=category_colors)\n",
    "ax.legend().set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a37eb9d-9641-42b2-bd0f-9edee97da3a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Forebrain\n",
    "ax = sns.scatterplot(x='PCSK1N_ligand_receptor_GP', # \"Spon1_ligand_recetpro_target_gene_GP\"\n",
    "                y=\"DKK1_ligand_receptor_GP\", # 'Il12b_ligand_receptor_target_gene_GP' ''Il31_ligand_receptor_target_gene_GP''\n",
    "                hue='Brain Niches',\n",
    "                data=model.adata.obs,\n",
    "                palette=category_colors)\n",
    "ax.legend().set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49611d28-ebef-4858-9deb-81b90f72184d",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obs[\"Fgf3_ligand_receptor_target_gene_GP\"] = -1 * model.adata.obs[\"Fgf3_ligand_receptor_target_gene_GP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "350982e1-ad8d-4b44-8cd1-b6b99cea9de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hindbrain\n",
    "# Sfrp1_ligand_receptor_target_gene_GP\n",
    "# Gdf10_ligand_receptor_target_gene_GP\n",
    "# Fgf3_ligand_receptor_target_gene_GP\n",
    "# Col22a1_ligand_receptor_target_gene_GP\n",
    "ax = sns.scatterplot(x='Fgf3_ligand_receptor_target_gene_GP', \n",
    "                y='Gdf10_ligand_receptor_target_gene_GP', # 'Col22a1_ligand_receptor_target_gene_GP',\n",
    "                hue='Brain Niches',\n",
    "                data=model.adata.obs,\n",
    "                palette=category_colors)\n",
    "ax.legend().set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04a7ecbd-091b-4c29-ae35-181780de3b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obs[\"Eng_ligand_receptor_target_gene_GP\"] = -1 * model.adata.obs[\"Eng_ligand_receptor_target_gene_GP\"]\n",
    "model.adata.obs[\"Efna2_ligand_receptor_target_gene_GP\"] = -1 * model.adata.obs[\"Efna2_ligand_receptor_target_gene_GP\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d97602b9-392e-477e-b10c-544101c9e39c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Midbrain\n",
    "# Fgf17_ligand_receptor_target_gene_GP\n",
    "# Efna2_ligand_receptor_target_gene_GP\n",
    "ax = sns.scatterplot(x='Fgf17_ligand_receptor_target_gene_GP', # Trh_ligand_receptor_target_gene_GP\n",
    "                y=\"Efna2_ligand_receptor_target_gene_GP\", # 'Eng_ligand_receptor_target_gene_GP', # \n",
    "                hue='Brain Niches',\n",
    "                data=model.adata.obs,\n",
    "                palette=category_colors)\n",
    "ax.legend().set_visible(False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73f67b76-7079-4b6b-a97b-da34f48348ad",
   "metadata": {},
   "source": [
    "### 3.1 Analyze Latent Manifold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f625913-a65f-457b-991a-3821acea3591",
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = model.adata.obs[sample_key].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7707921-aac1-436d-a2cc-f44052d225aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "if model.adata.obs[mapping_entity_key].nunique() > 1:\n",
    "    # Plot mapping entities in latent and physical space\n",
    "    mapping_entity_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "        cat_key=mapping_entity_key)\n",
    "    \n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/\" \\\n",
    "                \"mapping_entities_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"Mapping Entities\",\n",
    "        cat_key=mapping_entity_key,\n",
    "        groups=None,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=mapping_entity_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b03f95cd-74f6-4e19-bf75-2142850b5801",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot batches in latent and physical space\n",
    "condition_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=condition_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"batches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Batches\",\n",
    "    cat_key=condition_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=condition_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f995026b-cd0b-44b6-ad51-ef27b9a7021c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Plot cell types in latent and physical space\n",
    "cell_type_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=cell_type_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"cell_types_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Cell Types\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe216d6c-3f23-4356-9e23-3e79f73af61e",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(cell_type_groups) > 0:\n",
    "    # Plot cell type groups in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/{'_'.join(cell_type_groups).lower()}\" \\\n",
    "                \"_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "        cat_key=cell_type_key,\n",
    "        groups=cell_type_groups,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=cell_type_colors,\n",
    "        size=None,\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c5c87e4b-9a8f-4583-b18d-730d760bda57",
   "metadata": {},
   "source": [
    "del(model)\n",
    "import gc\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab19110c-0795-404f-8685-99fac1c80e30",
   "metadata": {},
   "source": [
    "load_timestamp = \"10082023_024709_36\"\n",
    "load_timestamp = \"09082023_225056_10\"\n",
    "load_timestamp = \"15082023_170146_1\"\n",
    "figure_folder_path = f\"../artifacts/{dataset}/figures/{model_label}/{load_timestamp}\"\n",
    "model_folder_path = f\"../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"\n",
    "result_folder_path = f\"../artifacts/{dataset}/results/{model_label}/{load_timestamp}\"\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(result_folder_path, exist_ok=True)\n",
    "model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                          adata=None,\n",
    "                          adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                          #adata_atac=None,\n",
    "                          #adata_atac_file_name=f\"{dataset}_{model_label}_atac.h5ad\",\n",
    "                          gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "567d1193-01cd-4b6a-a642-528faea844a2",
   "metadata": {},
   "source": [
    "samples = model.adata.obs[sample_key].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44abc3f7-9d93-408b-be6e-0b1e4e94866d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute latent Leiden clustering\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=latent_leiden_resolution,\n",
    "             key_added=latent_cluster_key,\n",
    "             neighbors_key=latent_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0204fae9-d88b-4a2a-9d0c-1e0abc579729",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot latent clusters in latent and physical space\n",
    "latent_cluster_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=latent_cluster_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"latent_clusters_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Latent Clusters\",\n",
    "    cat_key=latent_cluster_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=latent_cluster_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "593e7cd6-2902-427f-995b-79dc12ef14ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "for latent_cluster in model.adata.obs[latent_cluster_key].unique().tolist():\n",
    "    # Latent clusters in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_cluster_{latent_cluster}_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"Latent Clusters\",\n",
    "        cat_key=latent_cluster_key,\n",
    "        groups=latent_cluster,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=latent_cluster_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24099fd-afa8-4dea-a3c1-25380cdd1ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# latent_groups = [\"5\", \"6\", \"10\", \"15\"]\n",
    "# latent_groups = [\"4\", \"16\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62c11630-dd3a-4228-8fa6-7e62e463ccc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    # Plot latent cluster groups in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/{'_'.join(latent_groups).lower()}\" \\\n",
    "                \"_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=f\"{'_'.join(latent_groups)}\",\n",
    "        cat_key=latent_cluster_key,\n",
    "        groups=latent_groups,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=latent_cluster_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f0b6181-059b-4899-83d3-8d354aef43c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "if dataset == \"seqfish_mouse_organogenesis_imputed\":\n",
    "    niche_annotation_dict = {\n",
    "    \"0\": \"Ncam1/Il6st+ Muscles\",\n",
    "    \"1\": \"Vitamin A/9-cis-Retinoic acid+ Splanchnic Mesoderm\",\n",
    "    \"2\": \"IL10/Fgf2+ Angiogenesis\",\n",
    "    \"3\": \"Vip/Fgf8+ Spinal cord\",\n",
    "    \"4\": \"Ihh/Wnt5a+ Ventral Gut\",\n",
    "    \"5\": \"Oleic Acid/L-Threonine+ Midbrain\",\n",
    "    \"6\": \"Fgf17/Reln+ Forebrain\",\n",
    "    \"7\": \"Tgfbi/Vegfc+ Cranial Mesoderm\",\n",
    "    \"8\": \"Guanine/Slitrk1+ Neural Crest (Neural Tube)\",\n",
    "    \"9\": \"Lama5/Pdgfa+ Surface Ectoderm\",\n",
    "    \"10\": \"Gdf10/Fgf3+  Hindbrain\",\n",
    "    \"11\": \"Edn1/Nppa1+  Developing Heart\",\n",
    "    \"12\": \"Epo/Thpo+ Hematopoiesis\",\n",
    "    \"13\": \"Bmp4/Wnt5b+ Lateral Plate Mesoderm\",\n",
    "    \"14\": \"Ccl1/Igf2+ Mixed mesenchymal mesoderm\",\n",
    "    \"15\": \"Calca/Shh+ Floor Plate\",\n",
    "    \"16\": \"Oxt/Pdgfc+ Dorsal Gut\",\n",
    "    \"17\": \"Fgf18/Dll1+ Presomitic Mesoderm\",\n",
    "    \"18\": \"Pgf/Wnt6+ Allontois\",\n",
    "    \"19\": \"Mix\"\n",
    "    }\n",
    "\n",
    "    model.adata.obs[\"niche\"] = model.adata.obs[\"latent_leiden_0.35\"].map(niche_annotation_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "928dc642-a1e3-4b55-8bc8-7e4b6c78899f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.dendrogram(adata=model.adata,\n",
    "                 use_rep=\"nichecompass_latent\",\n",
    "                 linkage_method=\"single\",\n",
    "                 groupby=\"niche\")\n",
    "\n",
    "fig, (ax) = plt.subplots(1, 1, figsize=(3, 8))\n",
    "sc.pl.dendrogram(\n",
    "    adata=model.adata,\n",
    "    groupby=\"niche\",\n",
    "    orientation=\"left\",\n",
    "    ax=ax)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0680b05d-02c9-4989-9b20-dc3eac1bcb57",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04492af7-b0a2-4fcb-b0ff-e84d7aaaa574",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"celltype_mapped_refined\",\n",
    "    groups=\"Mixed mesenchymal mesoderm\",\n",
    "    spot_size=spot_size,\n",
    "    title=f\"\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d7e689b-f853-4554-a973-3ecc84a7ced6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Ccl1_ligand_receptor_target_gene_GP\",\n",
    "    color_map=\"RdBu_r\",\n",
    "    spot_size=spot_size,\n",
    "    title=f\"\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1607ca11-641b-49e3-a428-6f308891cc8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.obs[[\"Ccl1_ligand_receptor_target_gene_GP\", \"niche\"]].groupby(\"niche\").mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dcf1719-7db7-4982-9b1b-0cbd99ebcf1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gps = [\n",
    "    \"Ccl1_ligand_receptor_target_gene_GP\", \"IGF2_ligand_receptor_GP\",\n",
    "    \"BMP4_ligand_receptor_GP\", \"Wnt5b_ligand_receptor_target_gene_GP\",\n",
    "    \"Pgf_ligand_receptor_target_gene_GP\", \"Wnt6_ligand_receptor_target_gene_GP\",\n",
    "    \"Vitamin_A_metabolite_enzyme_sensor_GP\", \"9-cis-Retinoic_acid_metabolite_enzyme_sensor_GP\",\n",
    "    \"Edn1_ligand_receptor_target_gene_GP\", \"NPPA_ligand_receptor_GP\",\n",
    "    \"Oxt_ligand_receptor_target_gene_GP\", \"PDGFC_ligand_receptor_GP\",\n",
    "    \"IHH_ligand_receptor_GP\", \"Wnt5a_ligand_receptor_target_gene_GP\",\n",
    "    \"Lama5_ligand_receptor_target_gene_GP\", \"PDGFA_ligand_receptor_GP\",    \n",
    "    \"Guanine_metabolite_enzyme_sensor_GP\", \"Slitrk1_ligand_receptor_target_gene_GP\",\n",
    "    \"Tgfbi_ligand_receptor_target_gene_GP\", \"Vegfc_ligand_receptor_target_gene_GP\",    \n",
    "    \"Epo_ligand_receptor_target_gene_GP\", \"Thpo_ligand_receptor_target_gene_GP\",    \n",
    "    \"Il10_ligand_receptor_target_gene_GP\", \"Fgf2_ligand_receptor_target_gene_GP\", \n",
    "    \"Fgf17_ligand_receptor_target_gene_GP\", \"Reln_ligand_receptor_target_gene_GP\",    \n",
    "    \"Oleic_acid_metabolite_enzyme_sensor_GP\", \"L-Threonine_metabolite_enzyme_sensor_GP\",    \n",
    "    \"Calca_ligand_receptor_target_gene_GP\", \"Shh_ligand_receptor_target_gene_GP\",    \n",
    "    \"Gdf10_ligand_receptor_target_gene_GP\", \"Fgf3_ligand_receptor_target_gene_GP\",    \n",
    "    \"Vip_ligand_receptor_target_gene_GP\", \"Fgf8_ligand_receptor_target_gene_GP\", \n",
    "    \"Ncam1_ligand_receptor_target_gene_GP\", \"IL6ST_ligand_receptor_GP\",\n",
    "    \"Fgf18_ligand_receptor_target_gene_GP\", \"Dll1_ligand_receptor_target_gene_GP\",\n",
    "]\n",
    "\n",
    "reverse_sign_gps = [\n",
    "    \"IL6ST_ligand_receptor_GP\",\n",
    "    \"Vitamin_A_metabolite_enzyme_sensor_GP\",\n",
    "    \"9-cis-Retinoic_acid_metabolite_enzyme_sensor_GP\",\n",
    "    \"Il10_ligand_receptor_target_gene_GP\",\n",
    "    \"Fgf8_ligand_receptor_target_gene_GP\",\n",
    "    \"Oleic_acid_metabolite_enzyme_sensor_GP\",\n",
    "    \"L-Threonine_metabolite_enzyme_sensor_GP\",\n",
    "    \"Reln_ligand_receptor_target_gene_GP\",\n",
    "    \"Slitrk1_ligand_receptor_target_gene_GP\",\n",
    "    \"Fgf3_ligand_receptor_target_gene_GP\",\n",
    "    \"NPPA_ligand_receptor_GP\",\n",
    "    \"Thpo_ligand_receptor_target_gene_GP\",\n",
    "    \"Wnt5b_ligand_receptor_target_gene_GP\",\n",
    "    \"Calca_ligand_receptor_target_gene_GP\",\n",
    "    \"PDGFC_ligand_receptor_GP\",\n",
    "    \"Dll1_ligand_receptor_target_gene_GP\",\n",
    "    \"Wnt6_ligand_receptor_target_gene_GP\",\n",
    "]\n",
    "\n",
    "for gp in reverse_sign_gps:\n",
    "    model.adata.obs[gp] = -1 * model.adata.obs[gp]\n",
    "\n",
    "df = model.adata.obs[[\"niche\"] + gps].groupby(\"niche\").mean()\n",
    "df = df.reindex(model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"][::-1])\n",
    "\n",
    "scaler = MinMaxScaler()\n",
    "normalized_columns = scaler.fit_transform(df)\n",
    "normalized_df = pd.DataFrame(normalized_columns, columns=df.columns)\n",
    "normalized_df.index = df.index\n",
    "normalized_df.columns = [col.split(\"_lig\")[0].split(\"_met\")[0] for col in normalized_df.columns]\n",
    "\n",
    "plt.figure(figsize=(20, 6))  # Set the figure size\n",
    "sns.heatmap(normalized_df,\n",
    "            cmap='magma',\n",
    "            annot=False,\n",
    "            linewidths=0)\n",
    "plt.xticks(rotation=90,\n",
    "           ha='center')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "330124c9-17d3-4e1b-b0dc-6b307464858c",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10, 5))\n",
    "sankey(\n",
    "    x=adata_query_label_transfer.obs[ground_truth_key],\n",
    "    y=adata_query_label_transfer.obs[transfer_label_key],\n",
    "    title=âOriginal label vs. predicted annotationâ,\n",
    "    title_left=âOriginal labelâ,\n",
    "    title_right=âPredicted annotationâ,\n",
    "    ax=ax,\n",
    "    fontsize=â5\",  # âxx-smallâ,\n",
    "    left_order=adata_query_label_transfer.obs[ground_truth_key].unique().tolist(),\n",
    "    colorside=âleftâ,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7907109-c8c8-423e-9a0d-d2dad08d2302",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cb42305-d20a-4910-8d18-8cc49930cd7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "if dataset == \"spatial_atac_rna_seq_mouse_brain_batch2\":\n",
    "    rna_cluster_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "       cat_key=rna_cluster_key)\n",
    "    atac_cluster_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "       cat_key=atac_cluster_key)\n",
    "    \n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/\" \\\n",
    "                \"rna_cluster_annotations_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"RNA Clusters\",\n",
    "        cat_key=rna_cluster_key,\n",
    "        groups=None,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=rna_cluster_colors,\n",
    "        size=None,\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)\n",
    "\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/\" \\\n",
    "                \"atac_cluster_annotations_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"ATAC Clusters\",\n",
    "        cat_key=atac_cluster_key,\n",
    "        groups=None,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=atac_cluster_colors,\n",
    "        size=None,\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73a108db-5134-4b51-9975-a3935948b2b9",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 3.2 Analyze GPs & GP Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9b9a4f9-c6f1-4b84-a8e6-3e65b211faa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get gp scores of all gps (active & inactive)\n",
    "latent, _ = model.get_latent_representation(\n",
    "    adata=model.adata,\n",
    "    counts_key=counts_key,\n",
    "    adj_key=adj_key,\n",
    "    condition_key=condition_key,\n",
    "    only_active_gps=False,\n",
    "    return_mu_std=True,\n",
    "    node_batch_size=model.node_batch_size_)\n",
    "\n",
    "# Store gp scores, spatial coords and spatial neighbor graph in new adata\n",
    "gp_adata = ad.AnnData(latent)\n",
    "gp_adata.obs_names = model.adata.obs_names\n",
    "gp_adata.var_names = model.adata.uns[gp_names_key]\n",
    "gp_adata.obs[condition_key] = model.adata.obs[condition_key]\n",
    "gp_adata.obsp[adj_key] = model.adata.obsp[adj_key]\n",
    "gp_adata.obsm[spatial_key] = model.adata.obsm[spatial_key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17657dcc-5662-4fef-a0d4-eb2b605b55df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute spatial autocorrelation score Moran I of gene programs\n",
    "# (higher score -> positive spatial autocorrelation)\n",
    "sq.gr.spatial_autocorr(gp_adata,\n",
    "                       mode=\"moran\",\n",
    "                       n_perms=100,\n",
    "                       n_jobs=1)\n",
    "\n",
    "# Add score to summary df\n",
    "gp_summary_df[\"gp_moranI\"] = gp_adata.uns[\"moranI\"].loc[\n",
    "    gp_summary_df[\"gp_name\"].values.tolist(), \"I\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b059064-73c3-4cb5-8c1d-048662918877",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot gp with highest moran I score\n",
    "sq.pl.spatial_scatter(\n",
    "    gp_adata[gp_adata.obs[condition_key] == samples[0]],\n",
    "    shape=None,\n",
    "    color=[gp_summary_df[[\"gp_name\", \"gp_moranI\"]].sort_values(by=\"gp_moranI\", ascending=False).head(1)[\"gp_name\"].values[0]],\n",
    "    size=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ba5ad9a-f7e7-4563-b5d0-4c7c3cde2f74",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot gp with lowest moran I score\n",
    "sq.pl.spatial_scatter(\n",
    "    gp_adata[gp_adata.obs[condition_key] == samples[0]],\n",
    "    shape=None,\n",
    "    color=[gp_summary_df[[\"gp_name\", \"gp_moranI\"]].sort_values(by=\"gp_moranI\", ascending=False).tail(1)[\"gp_name\"].values[0]],\n",
    "    size=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9e5d9e8-2d0f-4fa9-91cf-5e7acd23d9ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine highly variable gene programs\n",
    "hvgp_threshs = [0.1]\n",
    "\n",
    "hvgp_list = []\n",
    "hvgp_cols = []\n",
    "for hvgp_thresh in hvgp_threshs:\n",
    "\n",
    "    sc.pp.highly_variable_genes(\n",
    "        gp_adata,\n",
    "        n_top_genes=int(len(gp_adata.var) * hvgp_thresh),\n",
    "        flavor=\"seurat\",\n",
    "        # batch_key=condition_key,\n",
    "        subset=False)\n",
    "\n",
    "    hvgp_list.append(gp_adata[:, gp_adata.var[\"highly_variable\"]].var_names.tolist())\n",
    "    \n",
    "for i, hvgp_thresh in enumerate(hvgp_threshs):\n",
    "    gp_summary_df[f\"hvgp_{hvgp_thresh}_thresh\"] = gp_summary_df[\"gp_name\"].isin(hvgp_list[i])\n",
    "    hvgp_cols.append(f\"hvgp_{hvgp_thresh}_thresh\")\n",
    "    \n",
    "gp_summary_df.loc[:, \"gp_category\"] = \"inactive\"\n",
    "gp_summary_df.loc[\n",
    "    gp_summary_df[\"gp_name\"].isin(active_gps), \"gp_category\"] = \"active\"\n",
    "gp_summary_df.loc[\n",
    "    gp_summary_df[\"gp_name\"].isin(hvgp_list[0]), \"gp_category\"] = \"highly_variable\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a11a0189-92cf-4017-9122-0f2cd85b4a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve genes, gene importances, gene ranks, and mean gene expr of target and source genes respectively\n",
    "gp_target_genes_df = gp_summary_df[\n",
    "    [\"gp_name\",\n",
    "     \"gp_active\",\n",
    "     \"gp_category\",\n",
    "     \"gp_moranI\",\n",
    "     \"gp_target_genes\",\n",
    "     \"gp_target_genes_importances\"] +\n",
    "     hvgp_cols].explode(\n",
    "        [\"gp_target_genes\",\n",
    "         \"gp_target_genes_importances\"]).rename(\n",
    "            columns={\"gp_target_genes\": \"target_gene\",\n",
    "                     \"gp_target_genes_importances\": \"target_gene_importance\"}).dropna()\n",
    "gp_target_genes_df[\"target_gene_rank\"] = gp_target_genes_df.groupby(\"gp_name\").cumcount() + 1\n",
    "gp_target_genes_df[\"target_gene_importance\"] = gp_target_genes_df[\"target_gene_importance\"].astype(np.float32)\n",
    "gp_target_genes_df[\"target_gene_mean_log_norm_gene_expr\"] = np.array(\n",
    "    model.adata[:, gp_target_genes_df[\"target_gene\"]].layers[counts_key].mean(0)).flatten()\n",
    "\n",
    "gp_source_genes_df = gp_summary_df[\n",
    "    [\"gp_name\",\n",
    "     \"gp_active\",\n",
    "     \"gp_category\",\n",
    "     \"gp_moranI\",\n",
    "     \"gp_source_genes\",\n",
    "     \"gp_source_genes_importances\"] +\n",
    "     hvgp_cols].explode(\n",
    "        [\"gp_source_genes\",\n",
    "         \"gp_source_genes_importances\"]).rename(\n",
    "            columns={\"gp_source_genes\": \"source_gene\",\n",
    "                     \"gp_source_genes_importances\": \"source_gene_importance\"}).dropna()\n",
    "gp_source_genes_df[\"source_gene_rank\"] = gp_source_genes_df.groupby(\"gp_name\").cumcount() + 1\n",
    "gp_source_genes_df[\"source_gene_importance\"] = gp_source_genes_df[\"source_gene_importance\"].astype(np.float32)\n",
    "gp_source_genes_df[\"source_gene_mean_log_norm_gene_expr\"] = np.array(\n",
    "    model.adata[:, gp_source_genes_df[\"source_gene\"]].layers[counts_key].mean(0)).flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11041283-be7d-4b22-aa7e-0270d0c51bca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute spatial autocorrelation score Moran I of genes\n",
    "# (higher score -> positive spatial autocorrelation)\n",
    "genes = list(set(list(gp_target_genes_df[\"target_gene\"]) + list(gp_source_genes_df[\"source_gene\"])))\n",
    "\n",
    "sq.gr.spatial_autocorr(model.adata,\n",
    "                       mode=\"moran\",\n",
    "                       genes=genes,\n",
    "                       n_perms=100,\n",
    "                       n_jobs=1)\n",
    "\n",
    "target_genes_moranI_scores = []\n",
    "for gene in gp_target_genes_df[\"target_gene\"]:\n",
    "    target_genes_moranI_scores.append(\n",
    "        model.adata.uns[\"moranI\"].loc[gene, \"I\"])\n",
    "gp_target_genes_df[\"gene_moranI\"] = target_genes_moranI_scores\n",
    "\n",
    "source_genes_moranI_scores = []\n",
    "for gene in gp_source_genes_df[\"source_gene\"]:\n",
    "    source_genes_moranI_scores.append(\n",
    "        model.adata.uns[\"moranI\"].loc[gene, \"I\"])\n",
    "gp_source_genes_df[\"gene_moranI\"] = source_genes_moranI_scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a0f2bcc-c42d-4d76-b687-49f2565a5b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create summary plots over all gps, active gps, and inactive gps \n",
    "# with fixed n_top_genes\n",
    "n_top_genes = 3\n",
    "use_only_active_gps = False\n",
    "    \n",
    "gp_target_genes_df_top_genes = gp_target_genes_df[\n",
    "    gp_target_genes_df[\"target_gene_rank\"] <= n_top_genes]\n",
    "gp_source_genes_df_top_genes = gp_source_genes_df[\n",
    "    gp_source_genes_df[\"source_gene_rank\"] <= n_top_genes]\n",
    "\n",
    "all_gp_target_genes_df_top_genes = gp_target_genes_df_top_genes\n",
    "active_gp_target_genes_df_top_genes = gp_target_genes_df_top_genes[\n",
    "    gp_target_genes_df_top_genes[\"gp_active\"]]\n",
    "inactive_gp_target_genes_df_top_genes = gp_target_genes_df_top_genes[\n",
    "    ~gp_target_genes_df_top_genes[\"gp_active\"]]\n",
    "\n",
    "all_gp_source_genes_df_top_genes = gp_source_genes_df_top_genes\n",
    "active_gp_source_genes_df_top_genes = gp_source_genes_df_top_genes[\n",
    "    gp_source_genes_df_top_genes[\"gp_active\"]]\n",
    "inactive_gp_source_genes_df_top_genes = gp_source_genes_df_top_genes[\n",
    "    ~gp_source_genes_df_top_genes[\"gp_active\"]]\n",
    "\n",
    "# Plot distribution of gene importances\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(10, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Gene Importance Distributions\")\n",
    "sns.histplot(all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "             ax=axs[0],\n",
    "             bins=20)\n",
    "axs[0].set(title=\"Target Genes (All GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Target Gene Count\")\n",
    "sns.histplot(all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "             ax=axs[1],\n",
    "             bins=20)\n",
    "axs[1].set(title=\"Source Genes (All GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Source Gene Count\")\n",
    "sns.histplot(active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "             ax=axs[2],\n",
    "             bins=20)\n",
    "axs[2].set(title=\"Target Genes (Active GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Target Gene Count\")\n",
    "sns.histplot(active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "             ax=axs[3],\n",
    "             bins=20)\n",
    "axs[3].set(title=\"Source Genes (Active GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Source Gene Count\")\n",
    "sns.histplot(inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "             ax=axs[4],\n",
    "             bins=20)\n",
    "axs[4].set(title=\"Target Genes (Inactive GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Target Gene Count\")\n",
    "sns.histplot(inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "             ax=axs[5],\n",
    "             bins=20)\n",
    "axs[5].set(title=\"Source Genes (Inactive GPs)\",\n",
    "           xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Source Gene Count\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()\n",
    "\n",
    "# Plot relationship between gene importances and log-normalized gene expression\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(20, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between NicheCompass Gene Importances \"\n",
    "             \"and Mean Log-Normalized Gene Expression of Genes\")\n",
    "sns.regplot(x=all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=all_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[0])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "    y=all_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"])\n",
    "all_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[0].set_title(f\"Target Genes (All GPs) - R2: {round(all_gp_target_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[0].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "sns.regplot(x=all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=all_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[1])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "    y=all_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"])\n",
    "all_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[1].set_title(f\"Source Genes (All GPs) - R2: {round(all_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[1].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "sns.regplot(x=active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=active_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[2])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "    y=active_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"])\n",
    "active_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[2].set_title(f\"Target Genes (Active GPs) - R2: {round(active_gp_target_genes_r_squared, 4)};  \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[2].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "sns.regplot(x=active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=active_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"],\n",
    "            ax=axs[3])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "    y=active_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"])\n",
    "active_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[3].set_title(f\"Source Genes (Active GPs) - R2: {round(active_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[3].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "if len(active_gps) != len(gp_summary_df):\n",
    "    sns.regplot(x=inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "                y=inactive_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"],\n",
    "                ax=axs[4])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "        y=inactive_gp_target_genes_df_top_genes[\"target_gene_mean_log_norm_gene_expr\"])\n",
    "    inactive_gp_target_genes_r_squared = r_value ** 2\n",
    "    axs[4].set_title(f\"Target Genes (Inactive GPs) - R2: {round(inactive_gp_target_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[4].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "    sns.regplot(x=inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "                y=inactive_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"],\n",
    "                ax=axs[5])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "        y=inactive_gp_source_genes_df_top_genes[\"source_gene_mean_log_norm_gene_expr\"])\n",
    "    inactive_gp_source_genes_r_squared = r_value ** 2\n",
    "    axs[5].set_title(f\"Source Genes (Inactive GPs) - R2: {round(inactive_gp_source_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[5].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Mean Log-Normalized Gene Expr\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()\n",
    "\n",
    "# Plot relationship between gene importances and Moran's I score\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(20, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between NicheCompass Gene Importances \"\n",
    "             \"and Moran's I Scores of Genes\")\n",
    "sns.regplot(x=all_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=all_gp_target_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[0])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_target_genes_df_top_genes.dropna()[\"target_gene_importance\"],\n",
    "    y=all_gp_target_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "all_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[0].set_title(f\"Target Genes (All GPs) - R2: {round(all_gp_target_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[0].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "sns.regplot(x=all_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=all_gp_source_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[1])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_source_genes_df_top_genes.dropna()[\"source_gene_importance\"],\n",
    "    y=all_gp_source_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "all_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[1].set_title(f\"Source Genes (All GPs) - R2: {round(all_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[1].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "sns.regplot(x=active_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "            y=active_gp_target_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[2])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_target_genes_df_top_genes.dropna()[\"target_gene_importance\"],\n",
    "    y=active_gp_target_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "active_gp_target_genes_r_squared = r_value ** 2\n",
    "axs[2].set_title(f\"Target Genes (Active GPs) - R2: {round(active_gp_target_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[2].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "sns.regplot(x=active_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "            y=active_gp_source_genes_df_top_genes[\"gene_moranI\"],\n",
    "            ax=axs[3])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_source_genes_df_top_genes.dropna()[\"source_gene_importance\"],\n",
    "    y=active_gp_source_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "active_gp_source_genes_r_squared = r_value ** 2\n",
    "axs[3].set_title(f\"Source Genes (Active GPs) - R2: {round(active_gp_source_genes_r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[3].set(xlabel=\"Gene Importance\",\n",
    "           ylabel=\"Moran's I Score\")\n",
    "if len(active_gps) != len(gp_summary_df):\n",
    "    sns.regplot(x=inactive_gp_target_genes_df_top_genes[\"target_gene_importance\"],\n",
    "                y=inactive_gp_target_genes_df_top_genes[\"gene_moranI\"],\n",
    "                ax=axs[4])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_target_genes_df_top_genes.dropna()[\"target_gene_importance\"],\n",
    "        y=inactive_gp_target_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "    inactive_gp_target_genes_r_squared = r_value ** 2\n",
    "    axs[4].set_title(f\"Target Genes (Inactive GPs) - R2: {round(inactive_gp_target_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[4].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Moran's I Score\")\n",
    "    sns.regplot(x=inactive_gp_source_genes_df_top_genes[\"source_gene_importance\"],\n",
    "                y=inactive_gp_source_genes_df_top_genes[\"gene_moranI\"],\n",
    "                ax=axs[5])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_source_genes_df_top_genes.dropna()[\"source_gene_importance\"],\n",
    "        y=inactive_gp_source_genes_df_top_genes.dropna()[\"gene_moranI\"])\n",
    "    inactive_gp_source_genes_r_squared = r_value ** 2\n",
    "    axs[5].set_title(f\"Source Genes (Inactive GPs) - R2: {round(inactive_gp_source_genes_r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[5].set(xlabel=\"Gene Importance\",\n",
    "               ylabel=\"Moran's I Score\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7afeed9-8b5a-4e20-8734-ac728d51122c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot relationship between gp moran scores and importance weighted gene moran scores\n",
    "gp_target_genes_moran_df = gp_target_genes_df.groupby(\"gp_name\").apply(\n",
    "    lambda x: pd.Series({\"weighted_gene_moranI\": (\n",
    "        x[\"target_gene_importance\"] * x[\"gene_moranI\"]).sum()}))\n",
    "gp_target_genes_moran_df.reset_index(inplace=True)\n",
    "gp_target_genes_moran_df = pd.merge(gp_target_genes_moran_df,\n",
    "                                    gp_summary_df[[\"gp_name\", \"gp_active\", \"gp_moranI\"]],\n",
    "                                    on=\"gp_name\")\n",
    "\n",
    "gp_source_genes_moran_df = gp_source_genes_df.groupby(\"gp_name\").apply(\n",
    "    lambda x: pd.Series({\"weighted_gene_moranI\": (\n",
    "        x[\"source_gene_importance\"] * x[\"gene_moranI\"]).sum()}))\n",
    "gp_source_genes_moran_df.reset_index(inplace=True)\n",
    "gp_source_genes_moran_df = pd.merge(gp_source_genes_moran_df,\n",
    "                                    gp_summary_df[[\"gp_name\", \"gp_active\", \"gp_moranI\"]],\n",
    "                                    on=\"gp_name\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(15, 5))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between GP Moran's I Score and \\n\"\n",
    "             \"Importance Weighted Gene Moran's I Scores\")\n",
    "sns.scatterplot(data=gp_target_genes_moran_df,\n",
    "                x=gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "                y=gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                hue=gp_target_genes_moran_df[\"gp_active\"],\n",
    "                ax=axs[0])\n",
    "axs[0].set(title=\"Target Genes\",\n",
    "           xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.scatterplot(data=gp_source_genes_moran_df,\n",
    "                x=gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "                y=gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                hue=gp_source_genes_moran_df[\"gp_active\"],\n",
    "                ax=axs[1])\n",
    "axs[1].set(title=\"Source Genes\",\n",
    "           xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "plt.subplots_adjust(wspace=0.35, top=0.8)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9aa9915-c406-4fcc-87ee-e91733998a5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot relationship between gp moran scores and importance weighted gene moran scores\n",
    "# across all gps, active gps, and inactive gps\n",
    "\n",
    "all_gp_target_genes_moran_df = gp_target_genes_moran_df\n",
    "active_gp_target_genes_moran_df = gp_target_genes_moran_df[\n",
    "    gp_target_genes_moran_df[\"gp_active\"]]\n",
    "inactive_gp_target_genes_moran_df = gp_target_genes_moran_df[\n",
    "    ~gp_target_genes_moran_df[\"gp_active\"]]\n",
    "\n",
    "all_gp_source_genes_moran_df = gp_source_genes_moran_df\n",
    "active_gp_source_genes_moran_df = gp_source_genes_moran_df[\n",
    "    gp_source_genes_moran_df[\"gp_active\"]]\n",
    "inactive_gp_source_genes_moran_df = gp_source_genes_moran_df[\n",
    "    ~gp_source_genes_moran_df[\"gp_active\"]]\n",
    "\n",
    "fig, axs = plt.subplots(nrows=3, ncols=2, figsize=(20, 15))\n",
    "axs = axs.flatten()\n",
    "plt.suptitle(\"Relationship between GP Moran's I Score and \\n\"\n",
    "             \"Importance Weighted Gene Moran's I Scores\")\n",
    "sns.regplot(x=all_gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "            y=all_gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[0])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_target_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=all_gp_target_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[0].set_title(f\"Target Genes (All GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[0].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.regplot(x=all_gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "            y=all_gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[1])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=all_gp_source_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=all_gp_source_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[1].set_title(f\"Source Genes (All GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[1].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.regplot(x=active_gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "            y=active_gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[2])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_target_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=active_gp_target_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[2].set_title(f\"Target Genes (Active GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[2].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "sns.regplot(x=active_gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "            y=active_gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "            ax=axs[3])\n",
    "_, _, r_value, p_value, _ = stats.linregress(\n",
    "    x=active_gp_source_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "    y=active_gp_source_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "r_squared = r_value ** 2\n",
    "axs[3].set_title(f\"Source Genes (Active GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                 f\"P: {round(p_value, 4)}\",\n",
    "                 fontsize=14)\n",
    "axs[3].set(xlabel=\"GP Moran's I Score\",\n",
    "           ylabel=\"Weighted Gene Moran's I Score\")\n",
    "if len(active_gps) != len(gp_summary_df):\n",
    "    sns.regplot(x=inactive_gp_target_genes_moran_df[\"gp_moranI\"],\n",
    "                y=inactive_gp_target_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                ax=axs[4])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_target_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "        y=inactive_gp_target_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "    r_squared = r_value ** 2\n",
    "    axs[4].set_title(f\"Target Genes (Inactive GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[4].set(xlabel=\"GP Moran's I Score\",\n",
    "               ylabel=\"Weighted Gene Moran's I Score\")\n",
    "    sns.regplot(x=inactive_gp_source_genes_moran_df[\"gp_moranI\"],\n",
    "                y=inactive_gp_source_genes_moran_df[\"weighted_gene_moranI\"],\n",
    "                ax=axs[5])\n",
    "    _, _, r_value, p_value, _ = stats.linregress(\n",
    "        x=inactive_gp_source_genes_moran_df.dropna()[\"gp_moranI\"],\n",
    "        y=inactive_gp_source_genes_moran_df.dropna()[\"weighted_gene_moranI\"])\n",
    "    r_squared = r_value ** 2\n",
    "    axs[5].set_title(f\"Source Genes (Inactive GPs) - R2: {round(r_squared, 4)}; \"\n",
    "                     f\"P: {round(p_value, 4)}\",\n",
    "                     fontsize=14)\n",
    "    axs[5].set(xlabel=\"GP Moran's I Score\",\n",
    "               ylabel=\"Weighted Gene Moran's I Score\")\n",
    "plt.subplots_adjust(wspace=0.35, hspace=0.3)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3cf43ae-ad12-4852-93c1-8a19a39a0e14",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_genes_weighted_moran_df = pd.DataFrame(columns=[\"GP Type\", \"GP Category\", \"Weighted Mean Moran's I Score\"])\n",
    "source_genes_weighted_moran_df = pd.DataFrame(columns=[\"GP Type\", \"GP Category\", \"Weighted Mean Moran's I Score\"])\n",
    "\n",
    "for gp_type, gp_type_identifier in zip([\"OmniPath\", \"NicheNet\", \"MEBOCOST\"],\n",
    "                                       [\"receptor\", \"target\", \"metabolite\"]):\n",
    "    gp_type_target_genes_df = gp_target_genes_df[gp_target_genes_df[\"gp_name\"].str.contains(gp_type_identifier)]\n",
    "    gp_type_source_genes_df = gp_source_genes_df[gp_source_genes_df[\"gp_name\"].str.contains(gp_type_identifier)]\n",
    "\n",
    "    target_genes_weighted_moran_sub_df = gp_type_target_genes_df.groupby(\"gp_category\").apply(\n",
    "        lambda x: pd.Series({\"Weighted Mean Moran's I Score\": (\n",
    "            x[\"target_gene_importance\"] * x[\"gene_moranI\"]).sum() / x[\"target_gene_importance\"].sum()}))\n",
    "\n",
    "    #new_row = {\"Weighted Mean Moran's I Score\": gp_type_target_genes_df[\"moranI\"].mean()}\n",
    "    #index_label = \"avg_across_all_gps\"\n",
    "    #new_df = pd.DataFrame(new_row, index=[index_label])\n",
    "    #target_genes_weighted_moran_sub_df = pd.concat([target_genes_weighted_moran_sub_df,\n",
    "    #                                              new_df])\n",
    "    target_genes_weighted_moran_sub_df[\"GP Type\"] = gp_type\n",
    "    target_genes_weighted_moran_sub_df[\"GP Category\"] = target_genes_weighted_moran_sub_df.index\n",
    "    target_genes_weighted_moran_sub_df.reset_index(inplace=True)\n",
    "    \n",
    "    target_genes_weighted_moran_df = pd.concat([target_genes_weighted_moran_df,\n",
    "                                                target_genes_weighted_moran_sub_df])\n",
    "\n",
    "    source_genes_weighted_moran_sub_df = gp_type_source_genes_df.groupby(\"gp_category\").apply(\n",
    "        lambda x: pd.Series({\"Weighted Mean Moran's I Score\": (\n",
    "            x[\"source_gene_importance\"] * x[\"gene_moranI\"]).sum() / x[\"source_gene_importance\"].sum()}))\n",
    "\n",
    "    #new_row = {\"Weighted Mean Moran's I Score\": gp_type_source_genes_df[\"moranI\"].mean()}\n",
    "    #index_label = \"avg_across_all_gps\"\n",
    "    #new_df = pd.DataFrame(new_row, index=[index_label])\n",
    "    #source_genes_weighted_moran_sub_df = pd.concat([source_genes_weighted_moran_sub_df,\n",
    "    #                                              new_df])\n",
    "    source_genes_weighted_moran_sub_df[\"GP Type\"] = gp_type\n",
    "    source_genes_weighted_moran_sub_df[\"GP Category\"] = source_genes_weighted_moran_sub_df.index\n",
    "    source_genes_weighted_moran_sub_df.reset_index(inplace=True)\n",
    "    \n",
    "    source_genes_weighted_moran_df = pd.concat([source_genes_weighted_moran_df,\n",
    "                                                source_genes_weighted_moran_sub_df])\n",
    "    \n",
    "# Plot relationship between gene importances and Moran's I score\n",
    "fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "plt.suptitle(\"Mean Moran's I Score Weighted by NicheCompass Gene Importances \")\n",
    "sns.stripplot(target_genes_weighted_moran_df,\n",
    "              x=\"GP Type\",\n",
    "              y=\"Weighted Mean Moran's I Score\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax1)\n",
    "ax1.set_title(\"Target Genes\",\n",
    "              fontsize=14)\n",
    "ax1.set(xlabel=\"GP Type\",\n",
    "        ylabel=\"Weighted Mean Moran's I Score\")\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles,\n",
    "                 labels,\n",
    "                 loc=\"center left\",\n",
    "                 bbox_to_anchor=(0.98, 0.5))\n",
    "ax1.get_legend().remove()\n",
    "sns.stripplot(source_genes_weighted_moran_df,\n",
    "              x=\"GP Type\",\n",
    "              y=\"Weighted Mean Moran's I Score\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax2)\n",
    "ax2.set_title(\"Source Genes\",\n",
    "              fontsize=14)\n",
    "ax2.set(xlabel=\"GP Type\",\n",
    "        ylabel=\"Weighted Mean Moran's I Score\")\n",
    "ax2.get_legend().remove()\n",
    "plt.subplots_adjust(wspace=0.35, top=0.85)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "199ffcf4-0131-4603-9387-85d89a00ebbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create summary plots with variable n_top_genes\n",
    "target_genes_corr_df = pd.DataFrame(columns=[\"Number of Top Genes\", \"GP Category\", \"Pearson Correlation\"])\n",
    "source_genes_corr_df = pd.DataFrame(columns=[\"Number of Top Genes\", \"GP Category\", \"Pearson Correlation\"])\n",
    "\n",
    "for n_top_genes in [5, 10, 20, 50, 100]:\n",
    "    gp_target_genes_df_top_genes = gp_target_genes_df[gp_target_genes_df[\"target_gene_rank\"] <= n_top_genes]\n",
    "    gp_source_genes_df_top_genes = gp_source_genes_df[gp_source_genes_df[\"source_gene_rank\"] <= n_top_genes]\n",
    "    \n",
    "    for gp_category in gp_target_genes_df_top_genes[\"gp_category\"].unique().tolist():\n",
    "        tmp_cat_df = gp_target_genes_df_top_genes.loc[gp_target_genes_df_top_genes[\"gp_category\"] == gp_category]\n",
    "        pearsonr = tmp_cat_df[\"target_gene_importance\"].corr(tmp_cat_df[\"gene_moranI\"])\n",
    "        temp_row_dict = {\"Number of Top Genes\": n_top_genes,\n",
    "                         \"GP Category\": gp_category,\n",
    "                         \"Pearson Correlation\": pearsonr}\n",
    "        target_genes_corr_df = target_genes_corr_df.append(temp_row_dict, ignore_index=True)\n",
    "        \n",
    "        tmp_cat_df = gp_source_genes_df_top_genes[gp_source_genes_df_top_genes[\"gp_category\"] == gp_category]\n",
    "        pearsonr = tmp_cat_df[\"source_gene_importance\"].corr(tmp_cat_df[\"gene_moranI\"])\n",
    "        temp_row_dict = {\"Number of Top Genes\": n_top_genes,\n",
    "                         \"GP Category\": gp_category,\n",
    "                         \"Pearson Correlation\": pearsonr}\n",
    "        source_genes_corr_df = source_genes_corr_df.append(temp_row_dict, ignore_index=True)\n",
    "        \n",
    "# Plot relationship between gene importances and Moran's I score\n",
    "fig, (ax1, ax2) = plt.subplots(ncols=2, figsize=(10, 5))\n",
    "plt.suptitle(\"Pearson Correlation between NicheCompass Gene Importances \"\n",
    "             \"and Moran's I Scores of Genes\")\n",
    "sns.stripplot(target_genes_corr_df,\n",
    "              x=\"Number of Top Genes\",\n",
    "              y=\"Pearson Correlation\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax1)\n",
    "ax1.set_title(\"Target Genes\",\n",
    "              fontsize=14)\n",
    "ax1.set(xlabel=\"Number of Top Genes\",\n",
    "        ylabel=\"Pearson Correlation\")\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "lgd = fig.legend(handles,\n",
    "                 labels,\n",
    "                 loc=\"center left\",\n",
    "                 bbox_to_anchor=(0.98, 0.5))\n",
    "ax1.get_legend().remove()\n",
    "sns.stripplot(source_genes_corr_df,\n",
    "              x=\"Number of Top Genes\",\n",
    "              y=\"Pearson Correlation\",\n",
    "              hue=\"GP Category\",\n",
    "              ax=ax2)\n",
    "ax2.set_title(\"Source Genes\",\n",
    "              fontsize=14)\n",
    "ax2.set(xlabel=\"Number of Top Genes\",\n",
    "        ylabel=\"Pearson Correlation\")\n",
    "ax2.get_legend().remove()\n",
    "plt.subplots_adjust(wspace=0.35, top=0.85)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67658a0a-edbb-4e7b-836f-e4e4808246de",
   "metadata": {},
   "source": [
    "### 3.3 Latent Cluster Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ae03333-6864-4c0c-b714-82da734bac39",
   "metadata": {},
   "source": [
    "#### 3.3.1 One Cluster (All) vs Rest of Tissue Differential GP Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fd12bee-cf19-4abe-b407-3dc9ebf927e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cats = None\n",
    "log_bayes_factor_thresh = 2.3 # 2.3 strong threshold; 4.6 decisive threshold (https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "title = f\"NicheCompass Latent Cluster Enriched Gene Programs Log Bayes Factor {log_bayes_factor_thresh}\"\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            f\"latent_clusters_all_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_heatmap.pdf\"\n",
    "\n",
    "# Run differential gp testing\n",
    "enriched_gps = model.run_differential_gp_tests(\n",
    "    cat_key=latent_cluster_key,\n",
    "    selected_cats=selected_cats,\n",
    "    comparison_cats=\"rest\",\n",
    "    log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "# Plot heatmap of enriched gps\n",
    "ax = sc.pl.heatmap(model.adata,\n",
    "                    enriched_gps,\n",
    "                    show_gene_labels=True,\n",
    "                    groupby=latent_cluster_key,\n",
    "                    dendrogram=True,\n",
    "                    swap_axes=True,\n",
    "                    figsize=(model.adata.obs[latent_cluster_key].nunique() * 1.5,\n",
    "                             len(enriched_gps) / 2),\n",
    "                    save=save_fig)\n",
    "if save_fig:\n",
    "    shutil.move(\"figures/heatmap.pdf\", file_path)\n",
    "    os.rmdir(\"figures\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5d5bce1-3f24-4839-8091-191ae425d0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_file = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            f\"latent_clusters_all_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_summary.csv\"\n",
    "\n",
    "gp_summary_cols = [\"gp_name\",\n",
    "                   \"n_source_genes\",\n",
    "                   \"n_non_zero_source_genes\",\n",
    "                   \"n_target_genes\",\n",
    "                   \"n_non_zero_target_genes\",\n",
    "                   \"gp_source_genes\",\n",
    "                   \"gp_target_genes\",\n",
    "                   \"gp_source_genes_weights\",\n",
    "                   \"gp_target_genes_weights\",\n",
    "                   \"gp_source_genes_importances\",\n",
    "                   \"gp_target_genes_importances\"]\n",
    "if multimodal:\n",
    "    gp_summary_cols = gp_summary_cols + [\n",
    "        \"n_source_peaks\",\n",
    "        \"n_target_peaks\",\n",
    "        \"gp_source_peaks\",\n",
    "        \"gp_target_peaks\",\n",
    "        \"gp_source_peaks_weights\",\n",
    "        \"gp_target_peaks_weights\",\n",
    "        \"gp_source_peaks_importances\",\n",
    "        \"gp_target_peaks_importances\"]\n",
    "\n",
    "# Get summary of decisively enriched gene programs\n",
    "enriched_gp_summary_df = gp_summary_df[gp_summary_df[\"gp_name\"].isin(enriched_gps)]\n",
    "cat_dtype = pd.CategoricalDtype(categories=enriched_gps, ordered=True)\n",
    "enriched_gp_summary_df[\"gp_name\"] = enriched_gp_summary_df[\"gp_name\"].astype(cat_dtype)\n",
    "enriched_gp_summary_df = enriched_gp_summary_df.sort_values(by=\"gp_name\")\n",
    "enriched_gp_summary_df = enriched_gp_summary_df[gp_summary_cols]\n",
    "\n",
    "if save_file:\n",
    "    enriched_gp_summary_df.to_csv(f\"{file_path}\")\n",
    "else:\n",
    "    display(enriched_gp_summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3348ee54-3b18-4891-bcce-2a043d3f7d37",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figs = True\n",
    "\n",
    "for i in np.arange(40, 50, 10):\n",
    "    plot_label = f\"res_{latent_leiden_resolution}_\" \\\n",
    "                 f\"latent_clusters_all_vs_rest_log_bayes_factor_\" \\\n",
    "                 f\"{log_bayes_factor_thresh}\"\n",
    "\n",
    "    generate_enriched_gp_info_plots(\n",
    "        plot_label=plot_label,\n",
    "        model=model,\n",
    "        sample_key=sample_key,\n",
    "        differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "        cat_key=latent_cluster_key,\n",
    "        cat_palette=latent_cluster_colors,\n",
    "        n_top_enriched_gp_start_idx=i,\n",
    "        n_top_enriched_gp_end_idx=i+10,\n",
    "        feature_spaces=samples, # [\"latent\"]\n",
    "        n_top_genes_per_gp=5,\n",
    "        n_top_peaks_per_gp=(0 if not multimodal else 5),\n",
    "        scale_omics_ft=False,\n",
    "        save_figs=save_figs,\n",
    "        figure_folder_path=f\"{figure_folder_path}/\",\n",
    "        spot_size=spot_size)\n",
    "    \n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66421db4-1f1d-4522-8afe-3348c9bbc19d",
   "metadata": {},
   "source": [
    "#### 3.2.2 One Cluster (Selected) vs Rest of Tissue Differential GP Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c4c869c-87cf-4344-aee1-92df29b40440",
   "metadata": {},
   "outputs": [],
   "source": [
    "latent_groups = [\"0\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d06e0332-83e3-485c-9d67-c6d69be4a2be",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    title = f\"NicheCompass Latent Cluster Enriched Gene Programs\"\n",
    "    log_bayes_factor_thresh = 2.3 # 4.6 decisive threshold; 2.3 strong threshold (https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_clusters_{'_'.join(latent_groups)}_vs_rest_log_bayes_factor_\" \\\n",
    "                f\"{log_bayes_factor_thresh}_enriched_gps_heatmap.pdf\"\n",
    "\n",
    "    # Run differential gp testing\n",
    "    enriched_gps = model.run_differential_gp_tests(\n",
    "        cat_key=latent_cluster_key,\n",
    "        selected_cats=latent_groups,\n",
    "        comparison_cats=\"rest\",\n",
    "        log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "    # Plot heatmap of enriched gps\n",
    "    ax = sc.pl.heatmap(model.adata,\n",
    "                        enriched_gps,\n",
    "                        show_gene_labels=True,\n",
    "                        groupby=latent_cluster_key,\n",
    "                        dendrogram=True,\n",
    "                        swap_axes=True,\n",
    "                        figsize=(model.adata.obs[latent_cluster_key].nunique() * 1.5,\n",
    "                                 len(enriched_gps) / 2),\n",
    "                        save=save_fig)\n",
    "    if save_fig:\n",
    "        shutil.move(\"figures/heatmap.pdf\", file_path)\n",
    "        os.rmdir(\"figures\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c6ec0f-8dd9-4007-a789-e344db7e6ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    save_file = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_clusters_{'_'.join(latent_groups)}_vs_rest_log_bayes_factor_\" \\\n",
    "                f\"{log_bayes_factor_thresh}_enriched_gps_summary.csv\"\n",
    "\n",
    "    gp_summary_cols = [\"gp_name\",\n",
    "                       \"n_source_genes\",\n",
    "                       \"n_non_zero_source_genes\",\n",
    "                       \"n_target_genes\",\n",
    "                       \"n_non_zero_target_genes\",\n",
    "                       \"gp_source_genes\",\n",
    "                       \"gp_target_genes\",\n",
    "                       \"gp_source_genes_weights\",\n",
    "                       \"gp_target_genes_weights\",\n",
    "                       \"gp_source_genes_importances\",\n",
    "                       \"gp_target_genes_importances\"]\n",
    "    if multimodal:\n",
    "        gp_summary_cols = gp_summary_cols + [\n",
    "            \"n_source_peaks\",\n",
    "            \"n_target_peaks\",\n",
    "            \"gp_source_peaks\",\n",
    "            \"gp_target_peaks\",\n",
    "            \"gp_source_peaks_weights\",\n",
    "            \"gp_target_peaks_weights\",\n",
    "            \"gp_source_peaks_importances\",\n",
    "            \"gp_target_peaks_importances\"]\n",
    "\n",
    "    # Get summary of enriched gene programs\n",
    "    enriched_gp_summary_df = gp_summary_df[gp_summary_df[\"gp_name\"].isin(enriched_gps)]\n",
    "    cat_dtype = pd.CategoricalDtype(categories=enriched_gps, ordered=True)\n",
    "    enriched_gp_summary_df.loc[:, \"gp_name\"] = enriched_gp_summary_df[\"gp_name\"].astype(cat_dtype)\n",
    "    enriched_gp_summary_df = enriched_gp_summary_df.sort_values(by=\"gp_name\")\n",
    "    enriched_gp_summary_df = enriched_gp_summary_df[gp_summary_cols]\n",
    "\n",
    "    if save_file:\n",
    "        enriched_gp_summary_df.to_csv(f\"{file_path}\")\n",
    "    else:\n",
    "        display(enriched_gp_summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba1fd1ba-f399-4557-badc-5719b97226c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(latent_groups) > 0:\n",
    "    save_figs = True\n",
    "\n",
    "    for i in np.arange(0, len(enriched_gps), 10):\n",
    "        plot_label = f\"res_{latent_leiden_resolution}_\" \\\n",
    "                    f\"latent_clusters_{'_'.join(latent_groups)}_vs_rest_log_bayes_factor_\" \\\n",
    "                    f\"{log_bayes_factor_thresh}\"\n",
    "\n",
    "        generate_enriched_gp_info_plots(\n",
    "            plot_label=plot_label,\n",
    "            model=model,\n",
    "            sample_key=sample_key,\n",
    "            differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "            cat_key=latent_cluster_key,\n",
    "            cat_palette=latent_cluster_colors,\n",
    "            n_top_enriched_gp_start_idx=i,\n",
    "            n_top_enriched_gp_end_idx=i+10,\n",
    "            feature_spaces=[\"p22\"], # [\"latent\"]\n",
    "            n_top_genes_per_gp=5,\n",
    "            n_top_peaks_per_gp=(0 if not multimodal else 5),\n",
    "            scale_omics_ft=False,\n",
    "            save_figs=save_figs,\n",
    "            figure_folder_path=f\"{figure_folder_path}/\",\n",
    "            spot_size=spot_size)\n",
    "\n",
    "        gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c64e6685-60bf-4fa6-8c44-0f1dcd548433",
   "metadata": {},
   "source": [
    "### 3.4 Cell Type Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f962bae7-1b93-44eb-bb3d-11f450024c4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = \"Gut tube\" # \"Forebrain/Midbrain/Hindbrain\", \"Gut tube\"\n",
    "cell_type_fmt = cell_type.replace('/', '_').lower()\n",
    "cell_type_differential_gp_scores_key = f\"{cell_type}_differential_gp_scores\"\n",
    "cell_type_latent_cluster_differential_gp_scores_key = f\"{cell_type}_latentn_cluster_differential_gp_scores\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0035471a-6afe-4979-99af-c9fbcd5fc4ad",
   "metadata": {},
   "source": [
    "#### 3.4.1 Overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c6933a-7182-4458-a4f5-ff35ebc5e9c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot cell type in physical and latent space\n",
    "plot_physical_latent_for_cell_types(adata=model.adata,\n",
    "                                    cell_types=[cell_type],\n",
    "                                    sample_key=sample_key,\n",
    "                                    cell_type_key=cell_type_key,\n",
    "                                    cell_type_colors=cell_type_colors,\n",
    "                                    figure_folder_path=figure_folder_path,\n",
    "                                    save_fig=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044061e0-cde1-4361-b68c-4b45ea32f86a",
   "metadata": {},
   "source": [
    "#### 3.4.2 Cell Type vs Rest of Tissue Differential GP Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a372cba8-98e5-4391-b5df-af3b9c5f0d4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_cats = [cell_type]\n",
    "log_bayes_factor_thresh = 2.3 # 2.3 strong threshold; 4.6 decisive threshold (https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "title = f\"NicheCompass {cell_type} Enriched Gene Programs Log Bayes Factor {log_bayes_factor_thresh}\"\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/{cell_type_fmt}_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_dotplot.svg\"\n",
    "\n",
    "# Run differential gp testing\n",
    "enriched_gps = model.run_differential_gp_tests(\n",
    "    cat_key=cell_type_key,\n",
    "    selected_cats=selected_cats,\n",
    "    comparison_cats=\"rest\",\n",
    "    log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "sc.pl.heatmap(model.adata,\n",
    "              enriched_gps,\n",
    "              show_gene_labels=True,\n",
    "              groupby=cell_type_key,\n",
    "              dendrogram=True,\n",
    "              swap_axes=True,\n",
    "              figsize=(model.adata.obs[cell_type_key].nunique() * 1.5,\n",
    "                       len(enriched_gps) / 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35709e8a-c998-4685-859f-c3d2dfdbcb5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_file = True\n",
    "file_path = f\"{figure_folder_path}/{cell_type_fmt}_vs_rest_log_bayes_factor_\" \\\n",
    "            f\"{log_bayes_factor_thresh}_enriched_gps_summary.csv\"\n",
    "\n",
    "# Get summary of enriched gene programs\n",
    "gp_summary_cols = [\"gp_name\",\n",
    "                   \"n_source_genes\",\n",
    "                   \"n_non_zero_source_genes\",\n",
    "                   \"n_target_genes\",\n",
    "                   \"n_non_zero_target_genes\",\n",
    "                   \"gp_source_genes\",\n",
    "                   \"gp_target_genes\",\n",
    "                   \"gp_source_genes_weights\",\n",
    "                   \"gp_target_genes_weights\",\n",
    "                   \"gp_source_genes_importances\",\n",
    "                   \"gp_target_genes_importances\"]\n",
    "if multimodal:\n",
    "    gp_summary_cols = gp_summary_cols + [\n",
    "        \"n_source_peaks\",\n",
    "        \"n_target_peaks\",\n",
    "        \"gp_source_peaks\",\n",
    "        \"gp_target_peaks\",\n",
    "        \"gp_source_peaks_weights\",\n",
    "        \"gp_target_peaks_weights\",\n",
    "        \"gp_source_peaks_importances\",\n",
    "        \"gp_target_peaks_importances\"]\n",
    "\n",
    "# Get summary of decisively enriched gene programs\n",
    "enriched_gp_summary_df = gp_summary_df[gp_summary_df[\"gp_name\"].isin(enriched_gps)]\n",
    "cat_dtype = pd.CategoricalDtype(categories=enriched_gps, ordered=True)\n",
    "enriched_gp_summary_df[\"gp_name\"] = enriched_gp_summary_df[\"gp_name\"].astype(cat_dtype)\n",
    "enriched_gp_summary_df = enriched_gp_summary_df.sort_values(by=\"gp_name\")\n",
    "enriched_gp_summary_df = enriched_gp_summary_df[gp_summary_cols]\n",
    "\n",
    "if save_file:\n",
    "    enriched_gp_summary_df.to_csv(f\"{file_path}\")\n",
    "else:\n",
    "    display(enriched_gp_summary_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b64eee9-4dc7-4731-83b3-ced0d2bc313f",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figs = True\n",
    "\n",
    "for i in np.arange(0, len(enriched_gps), 10):\n",
    "    plot_label = f\"{cell_type_fmt}_vs_rest_log_bayes_factor_\" \\\n",
    "                 f\"{log_bayes_factor_thresh}\"\n",
    "\n",
    "    generate_enriched_gp_info_plots(\n",
    "        plot_label=plot_label,\n",
    "        model=model,\n",
    "        sample_key=sample_key,\n",
    "        differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "        cat_key=cell_type_key,\n",
    "        cat_palette=cell_type_colors,\n",
    "        n_top_enriched_gp_start_idx=i,\n",
    "        n_top_enriched_gp_end_idx=i+10,\n",
    "        feature_spaces=samples, # [\"latent\"]\n",
    "        n_top_genes_per_gp=5,\n",
    "        n_top_peaks_per_gp=(0 if not multimodal else 5),\n",
    "        scale_omics_ft=False,\n",
    "        save_figs=save_figs,\n",
    "        figure_folder_path=f\"{figure_folder_path}/\",\n",
    "        spot_size=spot_size)\n",
    "    \n",
    "    gc.collect()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb08e58-7866-4b2e-baf1-a4264c1a37b8",
   "metadata": {},
   "source": [
    "### 3.5 Save Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e2a7ac-8d25-4d2d-aa75-09a0841dd720",
   "metadata": {},
   "outputs": [],
   "source": [
    "if log_norm_omics_features:\n",
    "    # Log normalize counts for cellxgene server\n",
    "    sc.pp.normalize_total(model.adata, target_sum=1e4)\n",
    "    sc.pp.log1p(model.adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bc19412-c190-4e21-b30b-1bd8cb61b106",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store gp summary in adata\n",
    "gp_summary = model.get_gp_summary()\n",
    "for col in gp_summary.columns:\n",
    "    gp_summary[col] = gp_summary[col].astype(str)\n",
    "model.adata.uns[\"nichecompass_gp_summary\"] = gp_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df3910b8-7d79-4f34-99bc-ba85bd0aa668",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.write(f\"{result_folder_path}/{dataset}_analysis.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cda8496-4830-48a9-aac6-f7cdfef78b54",
   "metadata": {},
   "outputs": [],
   "source": [
    "- Forebrain/Midbrain/Hindbrain: Regionally specific developing brain subtypes: separation into Rhombencephalon, Tegmentum, Mesencephalon, Prosencephalon\n",
    "- Mixed Mesenchymal Mesoderm: \"mixed mesenchymal mesoderm, represent \n",
    "a cell state rather than a defined cell type. Mesenchyme represents  a  state  in  which  cells  express  markers  characteristic  of  migratory   cells loosely dispersed within an extracellular matrix56. This strong  overriding transcriptional signature of mesenchyme, irrespective of  location,  makes  it  challenging  to  distinguish  which  cell  types  this   mixed  mesenchymal  mesoderm  population  represents  using  clas- sical scRNA-seq data\"\n",
    "\n",
    "We can identify distinct subpopulations that are spatially defined.\n",
    "Cardiomyocytes are spatially and morphologically distinct from other cell types, endothelium is interspersed and spread across the entire embryo space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68607d03-1863-414a-a015-7aa52d8783ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
