{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bdc8086b-2b80-43b4-b88a-aa0c87c820a6",
   "metadata": {},
   "source": [
    "# Slurm Job Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be88d0ca-4c09-44e3-83fc-fb6bcf4563fd",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>).\n",
    "- **Affiliation:** Helmholtz Munich, Institute of Computational Biology (ICB), Talavera-LÃ³pez Lab\n",
    "- **Date of Creation:** 20.03.2023\n",
    "- **Date of Last Modification:** 21.08.2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeeeb1fe-eaa1-4cdd-aa21-ed78e0b7be81",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e58b0a3-c3fe-43d5-a420-14111aa545f6",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48d297b0-e0dc-42b9-a349-e6ac2edba0e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab82f4c-edd8-47dd-be56-559cef739439",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0491ee0e-3587-4243-856f-e7f71e6dced6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda_env_name = \"nichecompass\"\n",
    "reproducibility_conda_env_name = \"nichecompass-reproducibility\"\n",
    "reproducibility_conda_env_name = \"nichecompass-test\"\n",
    "conda_env_name = \"nichecompass-test\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "004ab52a-9cc3-4e40-b0e6-f9fed7f4a549",
   "metadata": {},
   "source": [
    "### 1.3 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b2d6cf0-b360-428f-9548-c7942378de68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_python_script(\n",
    "        job_name_prefix,\n",
    "        job_id,\n",
    "        job_folder_path,\n",
    "        conda_env_name,\n",
    "        script_folder_path,\n",
    "        script_name,\n",
    "        script_args,\n",
    "        t=\"48:00:00\",\n",
    "        p=\"gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"gpu\",\n",
    "        nice=10000):\n",
    "    job_name = f\"{job_name_prefix}_{job_id}\"\n",
    "    # Account for fact that submit node has different home path than compute node\n",
    "    job_file_path = f\"{job_folder_path.replace('/aih', '')}/job_{job_name}.cmd\"\n",
    "    out_file_path = f\"{job_folder_path}/logs/out_{job_name}.txt\"\n",
    "    err_file_path = f\"{job_folder_path}/logs/err_{job_name}.txt\"\n",
    "    \n",
    "    os.makedirs(job_folder_path + \"/logs\", exist_ok=True)\n",
    "    \n",
    "    with open(job_file_path, \"w\") as handle:\n",
    "        handle.writelines(\"#!/bin/bash\\n\")\n",
    "        handle.writelines(f\"#SBATCH -J {job_name}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -o {out_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -e {err_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -t {t}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -p {p}\\n\")\n",
    "        handle.writelines(\"#SBATCH -c 6\\n\")\n",
    "        if \"gpu\" in p:\n",
    "            handle.writelines(f\"#SBATCH --gres={gres}\\n\")\n",
    "            handle.writelines(f\"#SBATCH --qos={qos}\\n\")       \n",
    "        handle.writelines(\"#SBATCH --mem=128G\\n\")\n",
    "        handle.writelines(f\"#SBATCH --nice={nice}\\n\")\n",
    "        handle.writelines(\"source $HOME/.bashrc\\n\")\n",
    "        handle.writelines(f\"conda activate {conda_env_name}\\n\")\n",
    "        handle.writelines(\"cd /\\n\")\n",
    "        handle.writelines(f\"cd {script_folder_path}\\n\")\n",
    "        handle.writelines(f\"python ../{script_name}\")\n",
    "        handle.writelines(f\"{script_args}\")\n",
    "        handle.writelines(\"\\n\")\n",
    "        \n",
    "    os.system(f\"sbatch {job_file_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4624e92-33eb-457c-8767-9006e44d850e",
   "metadata": {},
   "source": [
    "## 2. NicheCompass Ablation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc99f810-89a4-4070-ac09-d3f6da04c95d",
   "metadata": {},
   "source": [
    "### 2.1 Loss Weights & GP Mask"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4617a3-879b-4591-95d9-76576cc0031f",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Loss weights: Gene Expression Reconstruction, Edge Reconstruction, KL Divergence\n",
    "- Gene program Mask: FC GP Mask, NicheCompass Default GP Mask (Not Filtered, Weakly Filtered, Strongly Filtered)\n",
    "\n",
    "Additional covariate:\n",
    "- Spatial Neighborhood Graph: Number of Neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e95996-f6e0-459c-a400-5a2ae3193f47",
   "metadata": {},
   "source": [
    "#### 2.1.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84cfe23e-ec00-4c8d-af02-2aa210997565",
   "metadata": {},
   "source": [
    "##### 2.1.1.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0621ca71-70a3-423e-9f79-91e706f5d4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f136c4-a8a6-4907-962d-3bc8a2282b23",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc62e692-6c03-42e6-9fd8-c6bdde35d409",
   "metadata": {},
   "source": [
    "##### 2.1.1.2 NicheCompass GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48a49459-8035-4d3e-82bb-9e594f12533b",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 53\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a7c6489-5007-46b9-9d8f-650e3125e292",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 89\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"xenium_human_breast_cancer\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f3fbdd9-c303-4781-91e6-9485dabf8e33",
   "metadata": {},
   "source": [
    "#### 2.1.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e888c2-82b8-4d9f-b5e7-ac0d371bb0da",
   "metadata": {},
   "source": [
    "##### 2.1.2.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "333a8e02-74cc-474e-9baa-49b02d348617",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "038d37a5-46a9-4c7f-9d19-fe33a0398c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f29be6dc-ee82-43bf-9a07-c8b4cacab09b",
   "metadata": {},
   "source": [
    "##### 2.1.2.2 NicheCompass GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790911e2-ce3f-45cc-8e3d-3789cf1661e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 53\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4e9286b-b22c-441f-b3a4-25b9050eae73",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 89\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"starmap_plus_mouse_cns\"\n",
    "            reference_batches = \"batch1\"\n",
    "            species = \"mouse\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 1024\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb42f90-41a3-43e6-8747-18d91b527a92",
   "metadata": {},
   "source": [
    "#### 2.1.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5558252c-7c8a-4755-997e-2553e7af9d5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 100006\n",
    "for lambda_edge_recon in [50000000]:\n",
    "    for lambda_gene_expr_recon in [3000]:\n",
    "        for n_neighbors in [16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "            n_sampled_neighbors = -1\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 79\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e989cecb-0197-490e-8d11-5348bbdeffbc",
   "metadata": {},
   "source": [
    "##### 2.1.3.1 Fully Connected GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a9d1e2b-a939-4b6f-bdee-87f767743618",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ce2d525-2c68-4816-bbea-e6ca62309774",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          f\" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ff6fd10-e19a-4d55-ab17-ae11a82665a3",
   "metadata": {},
   "source": [
    "##### 2.1.3.2 NicheCompass GP Mask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4125af6c-9aed-40a2-a5ad-ac00ccae7e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 53\n",
    "for lambda_edge_recon in [0, 50000, 500000]:\n",
    "    for lambda_gene_expr_recon in [0, 30, 300]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9dc3f3b-b7c1-485c-8291-e0e2ad4c25ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# missing\n",
    "job_id = 89\n",
    "for lambda_edge_recon in [5000000, 50000000]:\n",
    "    for lambda_gene_expr_recon in [3000, 30000]:\n",
    "        for n_neighbors in [4, 8, 12, 16]:\n",
    "            task = \"ablation\"\n",
    "            ablation_task = \"loss_weights\"\n",
    "            dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "            reference_batches = \"batch2\"\n",
    "            species = \"human\"\n",
    "            node_label_method = \"one-hop-norm\"\n",
    "            edge_batch_size = 512\n",
    "\n",
    "            job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "            script_name = \"train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          \" --active_gp_thresh_ratio 0.\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          \" --conv_layer_encoder gcnconv\" \\\n",
    "                          \" --n_epochs 100\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked 0.\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors -1\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "            submit_python_script(\n",
    "                    job_name_prefix=job_name_prefix,\n",
    "                    job_id=job_id,\n",
    "                    job_folder_path=job_folder_path,\n",
    "                    conda_env_name=conda_env_name,\n",
    "                    script_folder_path=script_folder_path,\n",
    "                    script_name=script_name,\n",
    "                    script_args=script_args,\n",
    "                    nice=10000)\n",
    "\n",
    "            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ab4f058-e6e7-45d5-bce1-9432c1279590",
   "metadata": {},
   "source": [
    "#### 2.1.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2190df7-8f1e-4e06-979d-c2bb016394be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(60, 65, 5):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4c49be3-1031-427f-bb45-3245c60e98b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(5, 70, 5):\n",
    "    task = \"one-hop-norm_reference\"\n",
    "    dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"celltype_mapped_refined\"\n",
    "    batch_keys = \"batch\"\n",
    "    file_name = f\"mlflow_summary_loss_weights_ablation_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_loss_weights_ablation_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/ablation/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/ablation\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33a9fd4a-d69f-400d-8b1c-8c43e53f256f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(5, 45, 5):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"xenium_human_breast_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_states\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dd8b859-b990-4a88-a0de-2fb01190d204",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 1\n",
    "for i in np.arange(45, 65, 5):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198ebdc9-e759-438b-a624-51e4d49ac5aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(5, 45, 5):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b72aa3ff-636d-49a7-b406-47b4a3d14745",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(5, 45, 5):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db88fc23-3075-491e-bf2c-5fbdc4482a56",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(20, 240, 20):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "    ablation_task = \"loss_weights\"\n",
    "    cell_type_keys = \"cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ecdcbd2-4616-4175-b09f-3d853d51984d",
   "metadata": {},
   "source": [
    "### 2.2 Neighbor Sampling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016fccd5-e4f7-4262-ad93-e2e27e848248",
   "metadata": {},
   "source": [
    "Best parameters from previous ablations:\n",
    "- lambda_edge_recon = 50,000,000\n",
    "- lambda_gene_expr_recon = 30,000\n",
    "- nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "Ablating:\n",
    "- Neighbor sampling: no sampling, 2 neighbors sampling, 4 neighbors sampling, 8 neighbors sampling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbecce69-187e-4bdc-8b6f-2a2c348ca4f8",
   "metadata": {},
   "source": [
    "#### 2.2.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22828726-54f6-4d3d-9f91-3a34d05b1944",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"xenium_human_breast_cancer\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 128\n",
    "        lambda_edge_recon = 500000\n",
    "        lambda_gene_expr_recon = 300\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc9e436-8c7e-4bfe-b31f-19e974b3c083",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 15\n",
    "for n_neighbors in [20]:\n",
    "    for n_sampled_neighbors in [8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"xenium_human_breast_cancer\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 256\n",
    "        lambda_edge_recon = 500000\n",
    "        lambda_gene_expr_recon = 300\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8b28b8e-943e-4636-80f4-2300aa11e01a",
   "metadata": {},
   "source": [
    "#### 2.2.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12566fe1-76c8-4034-bde2-b5440c7407cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"starmap_plus_mouse_cns\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"mouse\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 1024\n",
    "        lambda_edge_recon = 500000\n",
    "        lambda_gene_expr_recon = 300\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b3617d6-b2ab-4fa9-9c07-eb76538a88b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 13\n",
    "for n_neighbors in [20, 24, 28, 32]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"starmap_plus_mouse_cns\"\n",
    "        reference_batches = \"batch1\"\n",
    "        species = \"mouse\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 512\n",
    "        lambda_edge_recon = 500000\n",
    "        lambda_gene_expr_recon = 300\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bd17731-1ce8-45a4-b02d-a0a92c56b895",
   "metadata": {},
   "source": [
    "#### 2.2.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e4e4b9b-20ef-42b8-b7e4-2279b4de9a78",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "        reference_batches = \"batch2\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 512\n",
    "        lambda_edge_recon = 500000\n",
    "        lambda_gene_expr_recon = 300\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd07fca5-4076-4f65-9d9c-79109b886bf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_neighbors in [20, 24, 28, 32]:\n",
    "    for n_sampled_neighbors in [2, 4, 8]:\n",
    "        task = \"ablation\"\n",
    "        ablation_task = \"neighbor_sampling\"\n",
    "        dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "        reference_batches = \"batch2\"\n",
    "        species = \"human\"\n",
    "        node_label_method = \"one-hop-norm\"\n",
    "        edge_batch_size = 256\n",
    "        lambda_edge_recon = \n",
    "        lambda_gene_expr_recon =\n",
    "        nichenet_keep_target_genes_ratio = 1.0\n",
    "\n",
    "        job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "        script_name = \"train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      f\" --nichenet_keep_target_genes_ratio {nichenet_keep_target_genes_ratio}\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      \" --active_gp_thresh_ratio 0.\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      \" --conv_layer_encoder gcnconv\" \\\n",
    "                      \" --n_epochs 100\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked 0.\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23fd2150-6666-445f-bada-f63bdc2ab9e6",
   "metadata": {},
   "source": [
    "#### 2.2.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8546a7c8-2eb9-49c9-b2b9-a856426ee7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(2, 22, 2):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"xenium_human_breast_cancer\"\n",
    "    ablation_task = \"neighbor_sampling\"\n",
    "    cell_type_keys = \"cell_states\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0a67899-5f7a-4d0f-a2ae-762d1a1ef113",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(2, 14, 2):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"neighbor_sampling\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1695f80-6b0c-48ee-8868-58ad387d4ce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(2, 14, 2):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "    ablation_task = \"neighbor_sampling\"\n",
    "    cell_type_keys = \"cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8b9a684-f40e-4cb3-a54e-936e6a7ed3bf",
   "metadata": {},
   "source": [
    "### 2.3 Encoder Architecture"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5501bd0a-337b-439b-9592-1abd7dd08c6e",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Encoder Number of Message Passing Layers: 1, 2\n",
    "- Encoder Type of Message Passing Layers: GCNConv vs GATv2Conv (4 attention heads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "544da563-d23d-4cfc-b288-e7e13f394f25",
   "metadata": {},
   "source": [
    "#### 2.3.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36b5743-f2a0-4b98-b68d-ab1525b96c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1, 2]:\n",
    "                for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"xenium_human_breast_cancer\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f05acdc8-844d-4bd9-a531-6e9efc12ef13",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 33\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1, 2]:\n",
    "                for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"xenium_human_breast_cancer\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b8c1653-0820-43a3-afd6-0655a6fd5394",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 8\n",
    "for n_layers_encoder in [1]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1]:\n",
    "                for n_neighbors in [16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"xenium_human_breast_cancer\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=0)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7804d861-fc83-4e82-890d-3e61c32b16c5",
   "metadata": {},
   "source": [
    "#### 2.3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7331e624-4c83-41a3-9a43-3dbeea67bb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1, 2]:\n",
    "                for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"starmap_plus_mouse_cns\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"mouse\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 256\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d82d10a2-e787-4c2e-be63-34509c2952a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 33\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1, 2]:\n",
    "                for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"starmap_plus_mouse_cns\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"mouse\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 256\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edeba52c-128d-4f1b-9391-d1173bde0466",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 28\n",
    "for n_layers_encoder in [2]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1]:\n",
    "                for n_neighbors in [16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"starmap_plus_mouse_cns\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"mouse\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe6c40e-d47c-40f2-ab49-194d977e0cd7",
   "metadata": {},
   "source": [
    "#### 2.3.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6c24a4-cecc-4ba0-81e2-8dcf7da09995",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1, 2]:\n",
    "                for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                    reference_batches = \"batch2\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec03d0e7-3800-4b18-9c50-8e3c945258f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 33\n",
    "for n_layers_encoder in [1, 2]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1, 2]:\n",
    "                for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                    reference_batches = \"batch2\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c7c1e9-39f7-4ccd-ae09-64f6d5268890",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 11\n",
    "for n_layers_encoder in [1]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1]:\n",
    "                for n_neighbors in [12]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                    reference_batches = \"batch2\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 256\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 50\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eb0c2c8-61c7-4843-99df-8473c13817ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 27\n",
    "for n_layers_encoder in [2]:\n",
    "    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_fc_layers_encoder in [1]:\n",
    "                for n_neighbors in [12]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                    reference_batches = \"batch2\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 256\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio 0.\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 50\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked 0.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d002ed6f-dc0c-4454-a130-86af5d46b439",
   "metadata": {},
   "source": [
    "#### 2.3.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aac34f4a-c335-4f22-b001-e9a608bdcc93",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(4, 36, 4):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"xenium_human_breast_cancer\"\n",
    "    ablation_task = \"encoder_architecture\"\n",
    "    cell_type_keys = \"cell_states\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d903ed41-10cc-4ffc-acdf-e03b062f149f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for i in np.arange(4, 36, 4):\n",
    "    task = \"ablation\"\n",
    "    dataset = \"starmap_plus_mouse_cns\"\n",
    "    ablation_task = \"encoder_architecture\"\n",
    "    cell_type_keys = \"Main_molecular_cell_type\"\n",
    "    batch_keys = \"None\"\n",
    "    file_name = f\"mlflow_summary_{ablation_task}_{task}_{dataset}_{i}.csv\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_{ablation_task}_{task}_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_metrics.py\"\n",
    "    script_args = f\" --task {ablation_task}_{task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --datasets {dataset}\" \\\n",
    "                  f\" --cell_type_keys {cell_type_keys}\" \\\n",
    "                  f\" --batch_keys {batch_keys}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)\n",
    "    \n",
    "    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8734b56d-9239-4960-9689-f80ff56bf927",
   "metadata": {},
   "source": [
    "### 2.3 Gene Program & Gene Selection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7757c17-695f-4a02-9f39-645513a96ec4",
   "metadata": {},
   "source": [
    "Ablating:\n",
    "- Gene Program Selection: Active GP Threshold Ratio, Group Lasso Regularization\n",
    "- Gene Selection: L1 Regularization\n",
    "- Spatial Neighborhood Graph: Number of Neighbors"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "985a6ac0-5138-497a-817d-4aca2cf73df9",
   "metadata": {},
   "source": [
    "#### 2.3.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "895f4c4e-2941-41aa-85af-7f6b1b856eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.03, 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0., 10.]:\n",
    "        for lambda_l1_masked in [0., 10.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"feature_selection\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 128\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "                n_sampled_neighbors = 4\n",
    "                n_fc_layers_encoder = 1\n",
    "                n_layers_encoder = 1\n",
    "                n_hidden_encoder = \"None\"\n",
    "                conv_layer_encoder = \"gcnconv\"\n",
    "                \n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba1b231d-c2ba-4628-975a-9c8793568012",
   "metadata": {},
   "source": [
    "#### 2.3.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c74da10f-1b0e-482f-9028-f1172312a902",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.03, 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0., 10.]:\n",
    "        for lambda_l1_masked in [0., 10.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"starmap_plus_mouse_cns\"\n",
    "                    reference_batches = \"batch1\"\n",
    "                    species = \"mouse\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 256\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "                    n_fc_layers_encoder = 1\n",
    "                    n_layers_encoder = 1\n",
    "                    n_hidden_encoder = \"None\"\n",
    "                    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                                  f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d51e95-d594-4fa6-bdcb-c44001904a73",
   "metadata": {},
   "source": [
    "#### 2.3.3 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f96de33-56ea-48db-bd1f-6437da817d90",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.03, 0.05, 0.1]:\n",
    "    for lambda_group_lasso in [0., 10.]:\n",
    "        for lambda_l1_masked in [0., 10.]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                    task = \"ablation\"\n",
    "                    ablation_task = \"encoder_architecture\"\n",
    "                    dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "                    reference_batches = \"batch2\"\n",
    "                    species = \"human\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    edge_batch_size = 128\n",
    "                    lambda_edge_recon = 500000.\n",
    "                    lambda_gene_expr_recon = 300.\n",
    "                    n_sampled_neighbors = 4\n",
    "                    n_fc_layers_encoder = 1\n",
    "                    n_layers_encoder = 1\n",
    "                    n_hidden_encoder = \"None\"\n",
    "                    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --no-filter_genes\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  f\" --model_label {ablation_task}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  f\" --n_fc_layers_encoder {n_fc_layers_encoder}\" \\\n",
    "                                  f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                                  f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                                  f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29ad5d87-e074-423f-b9c5-43a0405ed227",
   "metadata": {},
   "source": [
    "### 2.5 Categorical Covariate Embedding & Contrastive Integration"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e8207b-2b89-4fb1-89b0-c32cc207a8f3",
   "metadata": {},
   "source": [
    "#### 2.5.1 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9ebf56-a5a1-4b64-b341-1108d269bbd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for cat_covariates_embeds_injection in [\"encoder\", \"gene_expr_decoder\"]:\n",
    "for lambda_cat_covariates_contrastive in [0., 1, 2]:\n",
    "    for contrastive_logits_pos_ratio in [\"gcnconv\", \"gatv2conv\"]:\n",
    "        for n_hidden_encoder in [None]:\n",
    "            for n_neighbors in [4, 8, 12, 16]:\n",
    "                task = \"ablation\"\n",
    "                ablation_task = \"encoder_architecture\"\n",
    "                dataset = \"xenium_human_breast_cancer\"\n",
    "                reference_batches = \"batch1\"\n",
    "                species = \"human\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                edge_batch_size = 512\n",
    "                lambda_edge_recon = 500000.\n",
    "                lambda_gene_expr_recon = 300.\n",
    "                active_gp_thresh_ratio = 0.05\n",
    "                lambda_group_lasso = 0.\n",
    "                n_layers_encoder = 1\n",
    "                n_hidden_encoder = None\n",
    "                conv_layer_encoder = \"gcnconv\"\n",
    "                \n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --no-filter_genes\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              f\" --model_label {ablation_task}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              f\" --n_layers_encoder {n_layers_encoder}\" \\\n",
    "                              f\" --n_hidden_encoder {n_hidden_encoder}\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              f\" --n_epochs_no_cat_covariates_contrastive {n_epochs_no_cat_covariates_contrastive}\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                              f\" --lambda_group_lasso {lambda_group_lasso}\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a80b724d-8200-44b4-8dba-02269e4358e0",
   "metadata": {},
   "source": [
    "## 3. NicheCompass Single Sample Method Benchmarking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb67595-8ede-4823-a1fe-a313dee704f5",
   "metadata": {},
   "source": [
    "### 3.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52c28e8-6fc8-492f-9c31-d3ab9b815980",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe70711-9d03-47d5-892a-5cdf77714389",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6ac715-da37-447e-a72d-d67c838bd477",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ecebdb-2b38-4021-8116-b21061bef3b1",
   "metadata": {},
   "source": [
    "#### 3.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3995e5-3c35-427e-838c-4831370d44c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2d9c53-f844-48ec-b541-37071ac86e99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048 2048 2048\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19574d56-9c79-4874-9b4d-ee39248f0028",
   "metadata": {},
   "source": [
    "### 3.2 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe560799-9d98-4740-af7a-b972c7596d18",
   "metadata": {},
   "source": [
    "#### 3.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85bd3ffa-40d7-4414-b9fc-0a9e577f6a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory\n",
    "n_svg = 5000\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_gcnconv_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               \" --n_hvg 0\" \\\n",
    "               f\" --n_svg {n_svg}\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label gcnconv_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1548b5-4dca-4762-b2b1-3b70211ed57f",
   "metadata": {},
   "source": [
    "#### 3.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5242ada-198f-46a6-b545-048ab451c42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory\n",
    "    n_svg = 3000\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_one-hop-norm_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   \" --n_hvg 0\" \\\n",
    "                   f\" --n_svg {n_svg}\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label one-hop-norm_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 0\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4111eee6-88e7-4180-bc5b-a05cdde41061",
   "metadata": {},
   "source": [
    "### 3.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcce8ca8-124d-4814-a224-900bd546165e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7842f6f0-664b-4405-ad30-add0af93ec7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_batch5\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"32768 32768 32768 32768 32768 32768 32768 32768\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff1805d-42a6-4a8d-9277-2f98bfe4ffec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_batch5\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"512 512 512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None\" \\\n",
    "               \" --seeds 7 6 5 4\" \\\n",
    "               \" --run_index 8 7 6 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24ab2f3-1d57-4b83-aa65-e6948aca20c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_batch5\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"512 512 512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None\" \\\n",
    "               \" --seeds 3 2 1 0\" \\\n",
    "               \" --run_index 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d54d8a-bea3-40a9-acfa-4e78d2cffffe",
   "metadata": {},
   "source": [
    "#### 3.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7b792d-e60d-4a14-a971-7aca558dca0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"32768 32768 32768 32768 32768 32768 32768 32768\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "365a815d-5acd-4006-8f01-727b55e2b86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "for subsample_pct in [5, 10]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"32768 32768 32768 32768 32768 32768 32768 32768\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc04dd24-f620-46af-b075-be2ee4b3df7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f35e99-5f89-48ee-9894-73b7fb3b9fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "# GATv2 encoder\n",
    "for subsample_pct in [25]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                    f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c88c8e4-3f52-4ac4-8e9b-8ed601511180",
   "metadata": {},
   "source": [
    "### 3.4 Vizgen MERFISH Mouse Liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91da577-1daa-484b-809f-4c7c080aa129",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.4.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4f57ce-ed16-4994-a6dd-6dafe87a719f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"4096 4096 4096 4096 4096\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3\" \\\n",
    "               \" --run_index 8 7 6 5 4\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dc27972-079e-41d8-9aa1-e773ad39b34c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"4096 4096 4096\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None\" \\\n",
    "               \" --seeds 2 1 0\" \\\n",
    "               \" --run_index 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c89ecd3-de28-4a93-a243-5765d1d1d489",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a8ad41-8fa2-4b3b-8a87-29a3f740fc75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0039d426-bda8-4891-8120-d02fdd663477",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d15848a7-9057-44e8-be6f-c2ff35a26241",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65df455a-dc5d-48d1-9390-2f1cbd15c9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 3\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a60f728-fa4c-445e-94bf-0ded90bf5bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 6 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 4\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d826697c-54bd-44ad-a33a-2fe56af946db",
   "metadata": {},
   "source": [
    "#### 3.4.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa871f9-7357-4d38-a931-09a67728e0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"4096 4096 4096 4096 4096 4096 4096 4096\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27024f2f-dc7f-4d20-89fa-93543f8b85e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 7\" \\\n",
    "                   \" --run_index 8\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e3822a-3400-4d71-aa3b-24733b818b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 6\" \\\n",
    "                   \" --run_index 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e867ca2-1440-4a59-9a81-fc96cdb1ac72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 5\" \\\n",
    "                   \" --run_index 6\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d13bf7a-d45b-43a3-acb9-6a1d722bfcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 4\" \\\n",
    "                   \" --run_index 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f08ff180-3913-4da9-a2be-81ba62448596",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 3\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32bd9a75-b996-43f6-931d-1c6fda5e8a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [25]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2dea5a0-d09d-4c5e-aeca-9f543192a49c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [25]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 1 0\" \\\n",
    "                   \" --run_index 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d56926b9-cd38-4fb7-b16d-9f99f75f6843",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c39bdcc-6c2f-45ae-bc75-305db782e1e2",
   "metadata": {},
   "source": [
    "### 3.5 Slide-seqV2 Mouse Hippocampus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "278ca85b-f667-460d-b54f-98101eb221d4",
   "metadata": {},
   "source": [
    "#### 3.5.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f0bccd-24db-4252-933a-fffb5e17c618",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ece3508-371d-4930-9287-822da20f2236",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed715aa-9f15-445b-a37a-4ea469ca91b2",
   "metadata": {},
   "source": [
    "#### 3.5.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c94ac68-2133-41d2-913f-94438f7e26f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07a9315-cb35-4343-970e-2ff857b746ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fea47ff-9bfe-4850-9efb-5bd29fec260e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735ce16a-90a6-4e2d-b784-6fbcd3c17476",
   "metadata": {},
   "source": [
    "### 3.6 DeepLinc Models"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85aad8b2-b7ce-407f-856f-75331a35b730",
   "metadata": {},
   "source": [
    "### 3.7 Metrics Computation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1124e093-1499-421f-bf87-a4a4d68e186f",
   "metadata": {},
   "source": [
    "#### 3.6.1 NicheCompass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "455409d5-1191-4e62-b2a0-4569ddbc2cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e9a41a1-6517-4e95-8881-a06e012690bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4854fc25-77e5-4596-a9af-bf611cd05648",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd6ae43-2a23-47ce-bf8a-847ebb131d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb00efa-876f-490d-8ea1-2108a8af9623",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6e0860-1529-458c-94e0-285e5246d9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [#\"vizgen_merfish_mouse_liver\",\n",
    "                #\"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63e8020-49b7-422d-94f5-0947cf38f008",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_gcnconv_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "621a8e6b-966a-4467-8b7d-acea178463e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_gcnconv_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e42e7fd-aa6f-4856-858e-b66f93db1d5e",
   "metadata": {},
   "source": [
    "#### 3.6.2 GraphST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d39cdd-732b-45c3-ba71-239c23f18ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36cf5b6f-69df-446e-afe8-83c21290e4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12061d47-4585-4d81-9592-c543587c92f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41babdbf-f415-4bbb-8072-32c77fceb801",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\"]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd53e480-ebd7-481c-b50a-96b07b3b393c",
   "metadata": {},
   "source": [
    "#### 3.6.3 DeepLinc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16579147-7bb6-40ee-9301-2d66554e35b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_deeplinc.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"deeplinc_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key deeplinc_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ddfdb24-d761-4b18-96d9-b49106f5e073",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_deeplinc.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"deeplinc_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key deeplinc_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f0d25c-47a4-43d1-963f-97bf86974c74",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_deeplinc.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"deeplinc_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key deeplinc_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3070a0c0-979b-4df1-8290-eb4b22ede5dc",
   "metadata": {},
   "source": [
    "#### 3.6.4 SageNet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abeca64b-eb8f-4772-b78d-85675e8ccc2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_sagenet.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"sagenet_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key sagenet_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef0cf21c-9cf1-4bae-92f5-168a5b021b27",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_sagenet.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"sagenet_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key sagenet_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d30e24a-c8c6-4740-b80b-a1e2afcb6c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver_subsample_1pct\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_sagenet.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"sagenet_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key sagenet_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a8a76db-4b4a-48d7-a0d7-5dcea32c6910",
   "metadata": {},
   "source": [
    "#### 3.6.5 scVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b70c8d-1b55-493a-bf4b-b8481f2db9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21c715be-3352-4d76-b94f-4ee2b451cd76",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42df767-8650-428a-b74f-160a58090fca",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11227ee6-b65c-4f3a-aeda-e69819766070",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac03756-453e-41a6-b823-4a7da0211942",
   "metadata": {},
   "source": [
    "#### 3.6.6 expiMap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36802aa9-aa15-4455-939f-2701b0ac64ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d082e4f-db2e-4584-9976-b1dea885418d",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68169441-1003-49dd-b9fe-9b42a357e18a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a8c262f-eea4-434f-b3b2-5abe01e94234",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "adata = sc.read_h5ad(\"../artifacts/single_sample_method_benchmarking/vizgen_merfish_mouse_liver_subsample_25pct_expimap.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56f5c637-5dd8-491f-bddd-02901cd65fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d61960c-95e6-4396-a5df-8ccd83ff73c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]: \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ece5ce1-b630-4d02-ae72-4e8b483e5a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622f6c1d-65a1-4318-b03c-c12444fd0f4c",
   "metadata": {},
   "source": [
    "## 4. NicheCompass Sample Integration Method Benchmarking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd147edf-4ccd-402d-aca9-ce19cbd1b34f",
   "metadata": {},
   "source": [
    "### 4.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b865230b-ab67-4474-a29c-dcb1d2294340",
   "metadata": {},
   "source": [
    "#### 4.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f764998-8726-419f-bafe-d627f75a1257",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072\"\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = True\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f649c1e6-5384-45ce-90dd-a7413806b323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256 256\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = True\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None\" \\\n",
    "               \" --seeds 7 6 5 4\" \\\n",
    "               \" --run_index 8 7 6 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f0c2165-4fe0-43dd-9532-42d21da62de8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256 256\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = True\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None\" \\\n",
    "               \" --seeds 3 2 1 0\" \\\n",
    "               \" --run_index 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac84fe1-0698-4c04-868f-2dd6e7d73356",
   "metadata": {},
   "source": [
    "#### 4.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f15cc6-4e6d-4473-a126-dccb02b889a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048\"\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = True\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95eb73ce-64bf-4097-92fd-e5431f66652f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    node_label_method = \"one-hop-norm\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384\"\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = True\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method {node_label_method}\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b334564-7198-4344-8e8b-b20897765f3c",
   "metadata": {},
   "source": [
    "### 4.2 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfecebf-0b36-47f9-a3a4-b59938733399",
   "metadata": {},
   "source": [
    "#### 4.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679dc44d-14de-493b-8407-6bcf0debf032",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 5000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label one-hop-norm_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 100\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a05506-ee44-4d31-a43a-63a82c40ad1f",
   "metadata": {},
   "source": [
    "#### 4.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92069eca-64a0-4ab1-be3d-a5b0059145be",
   "metadata": {},
   "outputs": [],
   "source": [
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 5000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label one-hop-norm_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ab268b-d86a-4843-9623-6f687c4f0a11",
   "metadata": {},
   "source": [
    "### 4.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfda56b3-cb23-4b88-95f4-b3d8e8651ec2",
   "metadata": {},
   "source": [
    "#### 4.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c27c72a-b425-4514-a4d5-875309a40987",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8870fde7-98ad-410d-b431-b51d9501b50c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7 \" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da355e82-fdac-4d61-a665-844dcb2a59c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6 \" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eca55161-e696-478f-9cc5-ae6174459bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5 \" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c75ab927-fa76-4d59-b309-be7038fe1fef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192 8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a38cd78-7ee2-4121-89d2-92aaefc96504",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192 8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5db4fe3-e259-4370-90fd-f80f34ef0d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb25914-1791-4677-b8c1-460d637bb993",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "045bc3c8-10d1-4e11-9c5d-f2845269a9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b010d3-5626-453f-a82a-5d84e82613c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "933371f3-44a0-49e6-b40c-0b0433853405",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512 512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eae3159-6692-449e-8e7e-5f74cdd1b31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512 512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c112831f-87a4-41f8-9478-02f71d678a3f",
   "metadata": {},
   "source": [
    "#### 4.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dd3fe18-fd00-4fcd-9dfe-2700ff332153",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 7 6\" \\\n",
    "                   \" --run_index 8 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ab15aad-2467-40b3-a145-06168ebe2ab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 5 4\" \\\n",
    "                   \" --run_index 6 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8121ada5-3134-4b63-8898-5a3dfc6670ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73704f6-331a-4f60-987d-6a4c8121d428",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "296d296e-a8aa-46ba-a5f9-d62393aedbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 7 6\" \\\n",
    "                   \" --run_index 8 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e18f26c9-a86c-4f09-85c5-4575cf4a9856",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 5 4\" \\\n",
    "                   \" --run_index 6 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77ddc526-afc4-4153-85f6-f167ae190129",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a77d8f04-2cf4-4bb2-ae1f-5840338bc237",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8995fe1f-d843-444b-b288-01936b2e4544",
   "metadata": {},
   "source": [
    "### 4.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c4a4cbe-402e-4b61-a82e-6c7b71ec7c13",
   "metadata": {},
   "source": [
    "#### 4.4.1 NicheCompass"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ce8451-0b91-4bbd-9da6-7aadb4f2cd93",
   "metadata": {},
   "source": [
    "#### 4.4.2 GraphST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe2c202-e672-4742-a743-d3182d239662",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw bgc blisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b3e5e85-0b12-4b56-8c1d-0ed0ef332d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:     \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw bgc blisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4376fbd-d4b4-443a-ac33-047fc729c487",
   "metadata": {},
   "source": [
    "#### 4.4.3 scVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71925802-31c2-41a6-8f35-766bc471330e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw bgc blisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2963e3bf-efe6-47b3-82af-53b9d15f3f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw bgc blisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=reproducibility_conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828f673a-9f55-4b43-9ff2-ca286d8ed45a",
   "metadata": {},
   "source": [
    "## 5. NicheCompass Reference Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56180669-13e7-4873-8887-5804360c2b75",
   "metadata": {},
   "source": [
    "### 5.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6bed3a-6613-4101-81e5-9c19bb217b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0]):\n",
    "            for n_addon_gp in [10]:\n",
    "                for n_svg in [5000]:\n",
    "                    for conv_layer_encoder in [\"gcnconv\"]:\n",
    "                        for n_neighbors in [16]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 2\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97971162-badb-456b-8fc0-de1acfae91c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "job_id = 9998\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0]):\n",
    "            for n_addon_gp in [10]:\n",
    "                for n_svg in [3000]:\n",
    "                    for conv_layer_encoder in [\"gcnconv\"]:\n",
    "                        for n_neighbors in [16]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch3 batch4\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"None\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "                            job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aee0b50-4cfa-49f3-84b4-134a62b061b4",
   "metadata": {},
   "source": [
    "### 5.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11bdc0b5-5a82-4786-b53f-2e13cbe4f42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 2048\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"20\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 35\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd09c068-30ba-4237-a53d-c841dcbba9d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 2048\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fde3a4b-fd6f-43ba-9f37-436885f328df",
   "metadata": {},
   "source": [
    "### 5.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463fcdec-a4d3-4d97-8f6c-6d4a9251158b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 8 # 4, 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84963ec-99d0-47f9-be16-09b2d8d79a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 8 # 4, 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 20\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8d8a8b2-e5d2-4b7c-8422-c6ae5f923e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "job_id = 3\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\"\n",
    "n_neighbors = 8 # 4, 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ea84c27-e0ec-4924-9148-17ec1af9c156",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_filtered\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\"\n",
    "n_neighbors = 8 # 4, 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov patient\"\n",
    "cat_covariates_no_edges = \"True False True\"\n",
    "cat_covariates_embeds_nums = \"3 20 5\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7209ec-5334-4239-a142-d44ae5971bd9",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "671d5295-2e42-4365-811d-dcbaab2b0b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 2048\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"patient batch\"\n",
    "cat_covariates_no_edges = \"True True\"\n",
    "cat_covariates_embeds_nums = \"2 4\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4694727-8563-417f-8832-dc125b4af932",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_human_ovarian_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024 # 2048\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"4\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c96dfb0-6b04-4bed-ae41-5af7591090a8",
   "metadata": {},
   "source": [
    "### 5.4 Vizgen MERFISH Human Lung Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16844a7c-93da-4f56-bb1f-1cd17701d9b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096 # 2048\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.125 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7380bdd-79ba-48a7-aeaa-ba20b8978edd",
   "metadata": {},
   "source": [
    "### 5.5 Spatial ATAC-RNA-Seq Mouse Embryo & Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "916e9324-c9de-464a-b7ea-b1ed81980b92",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000, 5000000], [3000, 5000]):\n",
    "    for lambda_chrom_access_recon in [1000, 3000, 5000]:\n",
    "        for active_gp_thresh_ratio in [0.03, 0.05, 0.1]:\n",
    "            for lambda_l1_masked, lambda_l1_addon in zip([0, 50], [0, 50]):\n",
    "                for n_svg in [3000]:\n",
    "                    for n_svp in [0, 3000, 15000]:\n",
    "                        task = \"reference\"\n",
    "                        dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                        reference_batches = \"None\"\n",
    "                        n_neighbors = 8\n",
    "                        species = \"mouse\"\n",
    "                        node_label_method = \"one-hop-sum\"\n",
    "                        conv_layer_encoder = \"gatv2conv\"\n",
    "                        edge_batch_size = \"256\"\n",
    "                        n_sampled_neighbors = 4\n",
    "                        lambda_cat_covariates_contrastive = 0.\n",
    "                        contrastive_logits_pos_ratio = 0.\n",
    "                        contrastive_logits_neg_ratio = 0.\n",
    "                        cat_covariates_embeds_injection = \"None\"\n",
    "                        cat_covariates_keys = \"batch\"\n",
    "                        cat_covariates_no_edges = \"True\"\n",
    "                        cat_covariates_embeds_nums = 0\n",
    "                        n_hvg = 0\n",
    "                        n_addon_gp = 20\n",
    "\n",
    "                        job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                        job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                        script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                        script_name = \"train_nichecompass_reference_model.py\"\n",
    "                        script_args = f\" --dataset {dataset}\" \\\n",
    "                                      f\" --reference_batches {reference_batches}\" \\\n",
    "                                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                      \" --filter_genes\" \\\n",
    "                                      f\" --n_hvg {n_hvg}\" \\\n",
    "                                      f\" --n_svg {n_svg}\" \\\n",
    "                                      f\" --n_svp {n_svp}\" \\\n",
    "                                      \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                      \" --include_mebocost_gps\" \\\n",
    "                                      \" --include_collectri_gps\" \\\n",
    "                                      \" --include_brain_marker_gps\" \\\n",
    "                                      f\" --species {species}\" \\\n",
    "                                      \" --gp_filter_mode subset\" \\\n",
    "                                      \" --combine_overlap_gps\" \\\n",
    "                                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                                      \" --counts_key counts\" \\\n",
    "                                      \" --condition_key batch\" \\\n",
    "                                      f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                      f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                      \" --spatial_key spatial\" \\\n",
    "                                      \" --adj_key spatial_connectivities\" \\\n",
    "                                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                      \" --include_atac_modality\" \\\n",
    "                                      \" --filter_peaks\" \\\n",
    "                                      \" --min_cell_peak_thresh_ratio 0.01\" \\\n",
    "                                      f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                      \" --latent_key nichecompass_latent\" \\\n",
    "                                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                      \" --gene_expr_recon_dist nb\" \\\n",
    "                                      f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                      f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                      \" --log_variational\" \\\n",
    "                                      f\" --node_label_method {node_label_method}\" \\\n",
    "                                      \" --n_layers_encoder 1\" \\\n",
    "                                      \" --n_hidden_encoder None\" \\\n",
    "                                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                      \" --n_epochs 100\" \\\n",
    "                                      \" --n_epochs_all_gps 25\" \\\n",
    "                                      \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                      \" --lr 0.001\" \\\n",
    "                                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                      f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                                      f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                      f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                      f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                      \" --lambda_group_lasso 0.\" \\\n",
    "                                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                      f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                      \" --node_batch_size None\" \\\n",
    "                                      f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                        submit_python_script(\n",
    "                                job_name_prefix=job_name_prefix,\n",
    "                                job_id=job_id,\n",
    "                                job_folder_path=job_folder_path,\n",
    "                                conda_env_name=conda_env_name,\n",
    "                                script_folder_path=script_folder_path,\n",
    "                                script_name=script_name,\n",
    "                                script_args=script_args,\n",
    "                                nice=10000)\n",
    "\n",
    "                        job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9be90fc-602a-4d35-aa32-2cc7832503b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon, lambda_gene_expr_recon, lambda_chrom_access_recon in zip([5000000, 5000000], [3000, 3000], [1000, 3000]):\n",
    "    for active_gp_thresh_ratio in [0.01, 0.03, 0.05]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0, 20], [0, 20]):\n",
    "            for n_svg in [3000, 5000]:\n",
    "                for n_svp in [3000, 15000]:\n",
    "                    task = \"reference\"\n",
    "                    dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                    reference_batches = \"None\"\n",
    "                    n_neighbors = 8\n",
    "                    species = \"mouse\"\n",
    "                    node_label_method = \"one-hop-norm\"\n",
    "                    conv_layer_encoder = \"gatv2conv\"\n",
    "                    edge_batch_size = \"256\"\n",
    "                    n_sampled_neighbors = -1\n",
    "                    lambda_cat_covariates_contrastive = 0.\n",
    "                    contrastive_logits_pos_ratio = 0.\n",
    "                    contrastive_logits_neg_ratio = 0.\n",
    "                    cat_covariates_embeds_injection = \"None\"\n",
    "                    cat_covariates_keys = \"batch\"\n",
    "                    cat_covariates_no_edges = \"True\"\n",
    "                    cat_covariates_embeds_nums = 0\n",
    "                    n_hvg = 0\n",
    "                    n_addon_gp = 100\n",
    "\n",
    "                    job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                    script_name = \"train_nichecompass_reference_model.py\"\n",
    "                    script_args = f\" --dataset {dataset}\" \\\n",
    "                                  f\" --reference_batches {reference_batches}\" \\\n",
    "                                  f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                  \" --filter_genes\" \\\n",
    "                                  f\" --n_hvg {n_hvg}\" \\\n",
    "                                  f\" --n_svg {n_svg}\" \\\n",
    "                                  f\" --n_svp {n_svp}\" \\\n",
    "                                  \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                  \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                  \" --include_mebocost_gps\" \\\n",
    "                                  \" --include_collectri_gps\" \\\n",
    "                                  \" --include_brain_marker_gps\" \\\n",
    "                                  f\" --species {species}\" \\\n",
    "                                  \" --gp_filter_mode subset\" \\\n",
    "                                  \" --combine_overlap_gps\" \\\n",
    "                                  \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                  \" --overlap_thresh_genes 0.9\" \\\n",
    "                                  \" --counts_key counts\" \\\n",
    "                                  \" --condition_key batch\" \\\n",
    "                                  f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                  f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                  \" --spatial_key spatial\" \\\n",
    "                                  \" --adj_key spatial_connectivities\" \\\n",
    "                                  \" --mapping_entity_key mapping_entity\" \\\n",
    "                                  \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                  \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                  \" --include_atac_modality\" \\\n",
    "                                  \" --filter_peaks\" \\\n",
    "                                  \" --min_cell_peak_thresh_ratio 0.01\" \\\n",
    "                                  f\" --model_label {node_label_method}_{task}\" \\\n",
    "                                  \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                  \" --latent_key nichecompass_latent\" \\\n",
    "                                  f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                  f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                  \" --gene_expr_recon_dist nb\" \\\n",
    "                                  f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                  f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                  \" --log_variational\" \\\n",
    "                                  f\" --node_label_method {node_label_method}\" \\\n",
    "                                  \" --n_layers_encoder 1\" \\\n",
    "                                  \" --n_hidden_encoder None\" \\\n",
    "                                  f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                  \" --n_epochs 100\" \\\n",
    "                                  \" --n_epochs_all_gps 25\" \\\n",
    "                                  \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                  \" --lr 0.001\" \\\n",
    "                                  f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                  f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                  f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                                  f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                  f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                  f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                  \" --lambda_group_lasso 0.\" \\\n",
    "                                  f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                  f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                                  f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                  \" --node_batch_size None\" \\\n",
    "                                  f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                  f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                    submit_python_script(\n",
    "                            job_name_prefix=job_name_prefix,\n",
    "                            job_id=job_id,\n",
    "                            job_folder_path=job_folder_path,\n",
    "                            conda_env_name=conda_env_name,\n",
    "                            script_folder_path=script_folder_path,\n",
    "                            script_name=script_name,\n",
    "                            script_args=script_args,\n",
    "                            nice=10000)\n",
    "\n",
    "                    job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15d70ac5-c32f-4fe7-806c-1f29e680c458",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 19\n",
    "for lambda_edge_recon, lambda_gene_expr_recon, lambda_chrom_access_recon in zip([5000000, 10000000, 50000000], [3000, 6000, 30000], [1000, 2000, 10000]):\n",
    "    for active_gp_thresh_ratio in [0.01, 0.03, 0.05]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0, 20], [0, 20]):\n",
    "            for n_addon_gp in [100]:\n",
    "                task = \"reference\"\n",
    "                dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                reference_batches = \"None\"\n",
    "                n_neighbors = 8\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                conv_layer_encoder = \"gatv2conv\"\n",
    "                edge_batch_size = \"256\"\n",
    "                n_sampled_neighbors = 4\n",
    "                lambda_cat_covariates_contrastive = 0.\n",
    "                contrastive_logits_pos_ratio = 0.\n",
    "                contrastive_logits_neg_ratio = 0.\n",
    "                cat_covariates_embeds_injection = \"None\"\n",
    "                cat_covariates_keys = \"batch\"\n",
    "                cat_covariates_no_edges = \"True\"\n",
    "                cat_covariates_embeds_nums = 0\n",
    "                n_hvg = 3000\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --filter_genes\" \\\n",
    "                              f\" --n_hvg {n_hvg}\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              \" --include_collectri_gps\" \\\n",
    "                              \" --include_brain_marker_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --condition_key batch\" \\\n",
    "                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              \" --include_atac_modality\" \\\n",
    "                              \" --filter_peaks\" \\\n",
    "                              \" --min_cell_peak_thresh_ratio 0.01\" \\\n",
    "                              f\" --model_label {node_label_method}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                              \" --lambda_group_lasso 0.\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a44effd0-6542-4d16-bdd0-df3b89aa927f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 37\n",
    "for lambda_edge_recon, lambda_gene_expr_recon, lambda_chrom_access_recon in zip([5000000, 10000000], [3000, 6000], [1000, 2000]):\n",
    "    for active_gp_thresh_ratio in [0.01, 0.03, 0.05]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0, 20], [0, 20]):\n",
    "            for n_addon_gp in [100]:\n",
    "                task = \"reference\"\n",
    "                dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                reference_batches = \"None\"\n",
    "                n_neighbors = 8\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                conv_layer_encoder = \"gcnconv\"\n",
    "                edge_batch_size = \"4096\"\n",
    "                n_sampled_neighbors = -1\n",
    "                lambda_cat_covariates_contrastive = 0.\n",
    "                contrastive_logits_pos_ratio = 0.\n",
    "                contrastive_logits_neg_ratio = 0.\n",
    "                cat_covariates_embeds_injection = \"None\"\n",
    "                cat_covariates_keys = \"batch\"\n",
    "                cat_covariates_no_edges = \"True\"\n",
    "                cat_covariates_embeds_nums = 0\n",
    "                n_hvg = 3000\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --filter_genes\" \\\n",
    "                              f\" --n_hvg {n_hvg}\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              \" --include_collectri_gps\" \\\n",
    "                              \" --include_brain_marker_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --condition_key batch\" \\\n",
    "                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              \" --include_atac_modality\" \\\n",
    "                              \" --filter_peaks\" \\\n",
    "                              \" --min_cell_peak_thresh_ratio 0.01\" \\\n",
    "                              f\" --model_label {node_label_method}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                              \" --lambda_group_lasso 0.\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c8cba5a-c094-441e-8d33-14f09bd8b38d",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 55\n",
    "for lambda_edge_recon, lambda_gene_expr_recon, lambda_chrom_access_recon in zip([5000000, 10000000, 50000000], [3000, 6000, 30000], [1000, 2000, 10000]):\n",
    "    for active_gp_thresh_ratio in [0.01, 0.03, 0.05]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0, 20], [0, 20]):\n",
    "            for n_addon_gp in [100]:\n",
    "                task = \"reference\"\n",
    "                dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "                reference_batches = \"None\"\n",
    "                n_neighbors = 8\n",
    "                species = \"mouse\"\n",
    "                node_label_method = \"one-hop-norm\"\n",
    "                conv_layer_encoder = \"gatv2conv\"\n",
    "                edge_batch_size = \"256\"\n",
    "                n_sampled_neighbors = 4\n",
    "                lambda_cat_covariates_contrastive = 0.\n",
    "                contrastive_logits_pos_ratio = 0.\n",
    "                contrastive_logits_neg_ratio = 0.\n",
    "                cat_covariates_embeds_injection = \"None\"\n",
    "                cat_covariates_keys = \"batch\"\n",
    "                cat_covariates_no_edges = \"True\"\n",
    "                cat_covariates_embeds_nums = 0\n",
    "                n_hvg = 3000\n",
    "\n",
    "                job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "                script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                script_name = \"train_nichecompass_reference_model.py\"\n",
    "                script_args = f\" --dataset {dataset}\" \\\n",
    "                              f\" --reference_batches {reference_batches}\" \\\n",
    "                              f\" --n_neighbors {n_neighbors}\" \\\n",
    "                              \" --filter_genes\" \\\n",
    "                              f\" --n_hvg {n_hvg}\" \\\n",
    "                              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                              \" --include_mebocost_gps\" \\\n",
    "                              \" --include_collectri_gps\" \\\n",
    "                              \" --include_brain_marker_gps\" \\\n",
    "                              f\" --species {species}\" \\\n",
    "                              \" --gp_filter_mode subset\" \\\n",
    "                              \" --combine_overlap_gps\" \\\n",
    "                              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                              \" --overlap_thresh_genes 0.9\" \\\n",
    "                              \" --counts_key counts\" \\\n",
    "                              \" --condition_key batch\" \\\n",
    "                              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                              \" --spatial_key spatial\" \\\n",
    "                              \" --adj_key spatial_connectivities\" \\\n",
    "                              \" --mapping_entity_key mapping_entity\" \\\n",
    "                              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                              \" --include_atac_modality\" \\\n",
    "                              \" --filter_peaks\" \\\n",
    "                              \" --min_cell_peak_thresh_ratio 0.01\" \\\n",
    "                              f\" --model_label {node_label_method}_{task}\" \\\n",
    "                              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                              \" --latent_key nichecompass_latent\" \\\n",
    "                              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                              \" --gene_expr_recon_dist nb\" \\\n",
    "                              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                              \" --log_variational\" \\\n",
    "                              f\" --node_label_method {node_label_method}\" \\\n",
    "                              \" --n_layers_encoder 1\" \\\n",
    "                              \" --n_hidden_encoder None\" \\\n",
    "                              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                              \" --n_epochs 100\" \\\n",
    "                              \" --n_epochs_all_gps 25\" \\\n",
    "                              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                              \" --lr 0.001\" \\\n",
    "                              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                              f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "                              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                              \" --lambda_group_lasso 0.\" \\\n",
    "                              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                              f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "                              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                              \" --node_batch_size None\" \\\n",
    "                              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                submit_python_script(\n",
    "                        job_name_prefix=job_name_prefix,\n",
    "                        job_id=job_id,\n",
    "                        job_folder_path=job_folder_path,\n",
    "                        conda_env_name=conda_env_name,\n",
    "                        script_folder_path=script_folder_path,\n",
    "                        script_name=script_name,\n",
    "                        script_args=script_args,\n",
    "                        nice=10000)\n",
    "\n",
    "                job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01341dcb-64ad-45bd-8ef6-939e77d5b0f8",
   "metadata": {},
   "source": [
    "### 5.6 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c5b999-11e0-42e5-b886-8fb3a47d82f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"xenium_human_breast_cancer\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 4 # 4, 8, 12\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-attention\"\n",
    "edge_batch_size = 4096 # 4096 (4, 8), 1024 (12)\n",
    "cat_covariates_embeds_injection = \"encoder gene_expr_decoder\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0. # 0., 5., 10.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001945c5-6d59-4bdb-9efe-d8bbb33edc33",
   "metadata": {},
   "source": [
    "## 6. NicheCompass Reference Query Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bf6b4d-34b5-4324-8deb-095bdfa7d0e2",
   "metadata": {},
   "source": [
    "### 6.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a6a826d-f6ad-4ee5-85b2-66a3670a574e",
   "metadata": {},
   "source": [
    "#### 6.1.1 Reference Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdc61c79-6652-49fb-8bef-71ba0dda3b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_neighbors = 12\n",
    "n_hvg = 3000\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 4096\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_reference_only\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.05\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7db14469-4ee2-4be2-b00d-535bc0439620",
   "metadata": {},
   "source": [
    "#### 6.1.2 Query Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17761bc0-3178-42dd-9b0e-3935be8d2a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"01072023_165203_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 4096\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label {node_label_method}_{task}_query_only\" \\\n",
    "              f\" --reference_query_model_label {node_label_method}_{task}_query_mapping\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5391f85b-21e3-4723-bcef-6f7b0d948b4c",
   "metadata": {},
   "source": [
    "### 6.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "210672eb-c6ef-4cf5-9953-c2e4fcdec335",
   "metadata": {},
   "source": [
    "#### 6.2.1 Reference Model Training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90d450f9-673b-42ba-b11d-03fcd295693d",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 4\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 1024\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 250000. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0.0625 # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 0.1\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "149598af-5167-4466-b1e2-89de2f7375e2",
   "metadata": {},
   "source": [
    "#### 6.2.2 Query Mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a73f7984-551c-49f4-90f6-47eecf5092d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "query_batches = \"batch3\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"06072023_114143_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 1024\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label {node_label_method}_{task}_reference_only\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label {node_label_method}_{task}_query_only\" \\\n",
    "              f\" --reference_query_model_label {node_label_method}_{task}_query_mapping\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da211ed3-d9d9-4e55-a008-a33c0ca4f34a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  },
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
