{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "992dc6ee-ffc4-4d15-a364-486eaa5f6edc",
   "metadata": {},
   "source": [
    "# seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>)\n",
    "- **Date of Creation:** 22.01.2023\n",
    "- **Date of Last Modification:** 10.01.2025 (Sebastian Birk; <sebastian.birk@helmholtz-munich.de>)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6bd6fb-6697-43bd-b3ec-68b061afe511",
   "metadata": {},
   "source": [
    "- In order to run this notebook, a trained model needs to be stored under ```../../artifacts/{dataset}/models/{model_label}/{load_timestamp}```.\n",
    "    - dataset: ```seqfish_mouse_organogenesis_imputed```\n",
    "    - model_label: ```reference```\n",
    "    - load_timestamp: ```09082023_225056_10```\n",
    "- Run this notebook in the nichecompass-reproducibility environment, installable from ```('../../envs/environment.yaml')```."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../../utils\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import argparse\n",
    "import gc\n",
    "import os\n",
    "import random\n",
    "import shutil\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "import torch\n",
    "from matplotlib import font_manager, gridspec\n",
    "from matplotlib_venn import venn2, venn3\n",
    "from matplotlib.pyplot import rc_context\n",
    "from pywaffle import Waffle\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from sklearn.metrics import silhouette_score\n",
    "\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                                aggregate_obsp_matrix_per_cell_type,\n",
    "                                create_cell_type_chord_plot_from_df,\n",
    "                                create_new_color_dict,\n",
    "                                compute_communication_gp_network,\n",
    "                                visualize_communication_gp_network,\n",
    "                                generate_enriched_gp_info_plots)\n",
    "\n",
    "from analysis_utils import (add_cell_type_latent_cluster_emphasis,\n",
    "                            add_sub_cell_type,\n",
    "                            compute_cell_type_latent_clusters,\n",
    "                            generate_gp_info_plots,\n",
    "                            plot_physical_latent_for_cell_types,\n",
    "                            plot_cell_type_latent_clusters,\n",
    "                            plot_latent,\n",
    "                            plot_category_in_latent_and_physical_space,\n",
    "                            sankey,\n",
    "                            store_top_gps_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb8b324-ef18-4ab0-9ffb-8332d652458c",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9325589b-a7e4-4c31-97e7-8b7048248896",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"seqfish_mouse_organogenesis_imputed\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ac7b6d-4518-4356-bfdd-e141bf25d504",
   "metadata": {},
   "source": [
    "#### 1.2.1 Generic Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Model\n",
    "# AnnData keys\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "active_gp_names_key = \"nichecompass_active_gp_names\"\n",
    "latent_key = \"nichecompass_latent\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7be323-bd40-4fa0-a2d5-b8ddf5cc1c51",
   "metadata": {},
   "source": [
    "#### 1.2.2 Dataset-specific Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01210b2f-017e-43f6-b650-5cc732252f8c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "load_timestamp = \"09082023_225056_10\"\n",
    "model_label = \"reference\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "sample_key = \"sample\"\n",
    "condition_key = \"batch\"\n",
    "latent_leiden_resolution = 0.35\n",
    "spot_size = 0.05\n",
    "samples = [\"embryo2\"]\n",
    "    \n",
    "latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf229ca-ab66-4e52-a31e-8655878d290d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "924032d4-24a4-45cf-87e1-0eba197da0c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59250145-d819-4339-97cc-a039dcf4a8d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "plt.rcParams['font.size'] = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_folder_path = f\"../../artifacts/{dataset}/figures/{model_label}/{load_timestamp}\"\n",
    "model_folder_path = f\"../../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"\n",
    "result_folder_path = f\"../../artifacts/{dataset}/results/{model_label}/{load_timestamp}\"\n",
    "srt_data_folder_path = \"../../datasets/st_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(result_folder_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cea1a5c-d075-4e51-90e5-90684bee88a6",
   "metadata": {},
   "source": [
    "## 2. Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dad2ca6-7e34-48bc-89ef-4f994389a0f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load trained model\n",
    "model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                          adata=None,\n",
    "                          adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                          gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d99a385-5769-4c6b-a7dc-62f40028620a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.uns[gp_names_key] = np.array([gp for gp in model.adata.uns[gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp for gp in model.adata.uns[active_gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[gp_names_key]])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[active_gp_names_key]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fc40859-6a4a-4bfa-8362-d28d4c1c2057",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Check number of active gene programs\n",
    "active_gps = model.get_active_gps()\n",
    "print(f\"Number of total gene programs: {len(model.adata.uns[gp_names_key])}.\")\n",
    "print(f\"Number of active gene programs: {len(active_gps)}.\")\n",
    "\n",
    "gp_summary_df = model.get_gp_summary()\n",
    "gp_summary_df[gp_summary_df[\"gp_active\"] == True][20:]\n",
    "\n",
    "samples = model.adata.obs[sample_key].unique().tolist()\n",
    "model.add_active_gp_scores_to_obs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5486e215-7a06-424c-ae13-a209fc2c0173",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute Leiden clustering\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=latent_leiden_resolution,\n",
    "             key_added=latent_cluster_key,\n",
    "             neighbors_key=latent_key)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f662f73-455e-46c2-bc4b-ec25fb2f1d2b",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73f67b76-7079-4b6b-a97b-da34f48348ad",
   "metadata": {},
   "source": [
    "### 3.1 Create Figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44abc3f7-9d93-408b-be6e-0b1e4e94866d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2a ###\n",
    "niche_annotation_dict = {\n",
    "    \"0\": \"Saa1/Il6st+ Muscles\",\n",
    "    \"1\": \"Bmp8a/9-cis-Ret. acid+ Splanch. Mesoderm\",\n",
    "    \"2\": \"IL10/Fgf2+ Vasculature (Angiogenesis)\",\n",
    "    \"3\": \"Nptn/Fgf8+ Spinal Cord\",\n",
    "    \"4\": \"Ihh/Spint1+ Ventral Gut\",\n",
    "    \"5\": \"Pcsk1n/Dkk1+ Forebrain\",\n",
    "    \"6\": \"Fgf17/Efna2+ Midbrain\",\n",
    "    \"7\": \"Tgfbi/Fgf6+ Cranial Mesoderm\",\n",
    "    \"8\": \"Guanine/Slitrk1+ Neural Crest (Neural Tube)\",\n",
    "    \"9\": \"Sfrp4/Pdgfa+ Surface Ectoderm\",\n",
    "    \"10\": \"Gdf10/Fgf3+ Hindbrain\",\n",
    "    \"11\": \"Edn1/Nppa1+ Developing Heart\",\n",
    "    \"12\": \"Epo/Thpo+ Vasculature (Hematopoiesis)\",\n",
    "    \"13\": \"Bmp4/Vegfb+ Lateral Plate Mesoderm\",\n",
    "    \"14\": \"Cdh13/Col1a1+ Mixed Mesench. Mesoderm\",\n",
    "    \"15\": \"Calca/Shh+ Floor Plate\",\n",
    "    \"16\": \"Cthrc1/Pdgfc+ Dorsal Gut\",\n",
    "    \"17\": \"Fgf18/Dll1+ Presomitic Mesoderm\",\n",
    "    \"18\": \"Pgf/Cd44+ Allantois\",\n",
    "    \"19\": \"Mix\"\n",
    "}\n",
    "\n",
    "latent_cluster_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    color_palette=\"default\",\n",
    "    cat_key=latent_cluster_key)\n",
    "\n",
    "model.adata.obs[\"niche\"] = model.adata.obs[f\"latent_leiden_{latent_leiden_resolution}\"].map(niche_annotation_dict)\n",
    "niche_colors = {niche: latent_cluster_colors[cluster] for cluster, niche in niche_annotation_dict.items()}\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"niches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niches\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b03f95cd-74f6-4e19-bf75-2142850b5801",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Supplementary Fig. 2a ###\n",
    "condition_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    color_palette=\"batch\",\n",
    "    cat_key=condition_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"batches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Batches\",\n",
    "    cat_key=condition_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=condition_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f995026b-cd0b-44b6-ad51-ef27b9a7021c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Fig. 2b ###\n",
    "cell_type_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    color_palette=(\"cell_type_20\" if dataset == \"xenium_human_breast_cancer\" else \"cell_type_30\"),\n",
    "    cat_key=cell_type_key)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            \"cell_types_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Cell Types\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85b23bd0-29db-4df0-98fd-2bd085976d17",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2d ###\n",
    "sc.tl.dendrogram(adata=model.adata,\n",
    "                 use_rep=\"nichecompass_latent\",\n",
    "                 linkage_method=\"single\",\n",
    "                 groupby=\"niche\")\n",
    "\n",
    "fig, (ax) = plt.subplots(1, 1, figsize=(0.6, 2))\n",
    "sc.pl.dendrogram(\n",
    "    adata=model.adata,\n",
    "    groupby=\"niche\",\n",
    "    orientation=\"left\",\n",
    "    ax=ax,\n",
    "    save=\"_nichecompass_latent.svg\")\n",
    "plt.show()\n",
    "if os.path.exists(f\"{figure_folder_path}/dendrogram_nichecompass_latent.svg\"):\n",
    "    os.remove(f\"{figure_folder_path}/dendrogram_nichecompass_latent.svg\")\n",
    "shutil.move(f\"./figures/dendrogram_nichecompass_latent.svg\", figure_folder_path)\n",
    "shutil.rmtree(\"./figures/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96c0f4b8-cfc8-4cc0-8cf2-f3849895598d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2c ###\n",
    "tmp = pd.crosstab(model.adata.obs[\"niche\"],model.adata.obs['batch'], normalize='index')\n",
    "tmp = tmp.reindex(model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"][::])\n",
    "tmp.index = [gp_name.split(\"+ \")[1].replace(\"Mesenchymal\", \"Mesench.\") if \"+\" in gp_name else gp_name for gp_name in tmp.index.tolist()]\n",
    "ax = tmp.plot.barh(color=condition_colors, stacked=True, figsize=(2, 6)).legend(loc='center left', bbox_to_anchor=(1.0, 0.5))\n",
    "plt.xlabel(\"Data Source Proportions\")\n",
    "plt.xticks([0.1, 0.5, 1.0])\n",
    "plt.savefig(f\"{figure_folder_path}/niche_batch_proportions.svg\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93192065-a984-4e4a-a792-ca045c0c2ba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2f ###\n",
    "tmp = pd.crosstab(model.adata.obs[\"niche\"], model.adata.obs[cell_type_key], normalize='index')\n",
    "tmp = tmp.reindex(model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"][::])\n",
    "ax = tmp.plot.barh(color=cell_type_colors, stacked=True, figsize=(4, 6)).legend(loc='center left', bbox_to_anchor=(1.0, 0.5))\n",
    "plt.xlabel(\"Cell Type Proportions\")\n",
    "plt.xticks([0.1, 0.5, 1.0])\n",
    "plt.savefig(f\"{figure_folder_path}/niche_cell_type_proportions.svg\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9836defc-d203-4dea-9fd8-6a096f3fdcd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 2b ###\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "sankey(\n",
    "    x=model.adata.obs[cell_type_key],\n",
    "    y=model.adata.obs[\"niche\"],\n",
    "    title=\"Cell Type to Niche Mapping\",\n",
    "    title_left=\"Cell Type Label\",\n",
    "    title_right=\"Niche Label\",\n",
    "    ax=ax,\n",
    "    fontsize=\"5\",  # \"xx-small\",\n",
    "    #left_order=[\"Mixed mesenchymal mesoderm\", \"Lateral plate mesoderm\", \"Intermediate mesoderm\", \"Allantois\", \"Splanchnic mesoderm\",\n",
    "    #            \"ExE endoderm\", \"Cardiomyocytes\", \"Definitive endoderm\", \"Gut tube\", \"Surface ectoderm\", \"Neural crest\", \"Cranial mesoderm\",\n",
    "    #            \"Erythroid\", \"Blood progenitors\", \"Endothelium\", \"Haematoendothelial progenitors\", \"Forebrain/Midbrain/Hindbrain\",\n",
    "    #            \"Spinal cord\", \"NMP\", \"Dermomyotome\", \"Sclerotome\", \"Anterior somitic tissues\", \"Presomitic mesoderm\"],\n",
    "    colors=cell_type_colors,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.savefig(f\"{figure_folder_path}/sankey_cell_type_to_niche.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "129e4dbc-dfe0-40db-8e40-c71b4c4f787b",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2e ###\n",
    "gp_resign = True\n",
    "\n",
    "# Adjust signs of GPs\n",
    "gps = [\n",
    "        \"Cdh13_ligand_receptor_target_gene_GP\", \"Col1a1_ligand_receptor_target_gene_GP\",        \n",
    "        \"BMP4_ligand_receptor_GP\", \"Vegfb_ligand_receptor_target_gene_GP\",  \n",
    "        \"Pgf_ligand_receptor_target_gene_GP\", \"Cd44_ligand_receptor_target_gene_GP\",\n",
    "        \"Bmp8a_ligand_receptor_target_gene_GP\", \"9-cis-Retinoic_acid_metabolite_enzyme_sensor_GP\",\n",
    "        \"Edn1_ligand_receptor_target_gene_GP\", \"NPPA_ligand_receptor_GP\",\n",
    "        \"Cthrc1_ligand_receptor_target_gene_GP\", \"PDGFC_ligand_receptor_GP\",\n",
    "        \"IHH_ligand_receptor_GP\", \"Spint1_ligand_receptor_target_gene_GP\",\n",
    "        \"SFRP4_ligand_receptor_GP\", \"PDGFA_ligand_receptor_GP\",\n",
    "        \"Guanine_metabolite_enzyme_sensor_GP\", \"Slitrk1_ligand_receptor_target_gene_GP\",\n",
    "        \"Tgfbi_ligand_receptor_target_gene_GP\", \"FGF6_ligand_receptor_GP\",\n",
    "        \"Epo_ligand_receptor_target_gene_GP\", \"Thpo_ligand_receptor_target_gene_GP\",\n",
    "        \"Il10_ligand_receptor_target_gene_GP\", \"Fgf2_ligand_receptor_target_gene_GP\",\n",
    "        \"Fgf17_ligand_receptor_target_gene_GP\", \"Efna2_ligand_receptor_target_gene_GP\",\n",
    "        \"PCSK1N_ligand_receptor_GP\", \"DKK1_ligand_receptor_GP\",\n",
    "        \"Calca_ligand_receptor_target_gene_GP\", \"Shh_ligand_receptor_target_gene_GP\",\n",
    "        \"Gdf10_ligand_receptor_target_gene_GP\", \"Fgf3_ligand_receptor_target_gene_GP\",\n",
    "        \"NPTN_ligand_receptor_GP\", \"Fgf8_ligand_receptor_target_gene_GP\",\n",
    "        \"SAA1_ligand_receptor_GP\", \"IL6ST_ligand_receptor_GP\",\n",
    "        \"Fgf18_ligand_receptor_target_gene_GP\", \"Dll1_ligand_receptor_target_gene_GP\",    \n",
    "    ]\n",
    "    \n",
    "resign_gps = [\n",
    "    \"IL6ST_ligand_receptor_GP\",\n",
    "    \"9-cis-Retinoic_acid_metabolite_enzyme_sensor_GP\",\n",
    "    \"Il10_ligand_receptor_target_gene_GP\",\n",
    "    \"Fgf8_ligand_receptor_target_gene_GP\",\n",
    "    \"Oleic_acid_metabolite_enzyme_sensor_GP\",\n",
    "    \"L-Threonine_metabolite_enzyme_sensor_GP\",\n",
    "    \"Efna2_ligand_receptor_target_gene_GP\",\n",
    "    \"Slitrk1_ligand_receptor_target_gene_GP\",\n",
    "    \"Fgf3_ligand_receptor_target_gene_GP\",\n",
    "    \"NPPA_ligand_receptor_GP\",\n",
    "    \"Thpo_ligand_receptor_target_gene_GP\",\n",
    "    \"Calca_ligand_receptor_target_gene_GP\",\n",
    "    \"PDGFC_ligand_receptor_GP\",\n",
    "    \"Dll1_ligand_receptor_target_gene_GP\",\n",
    "    \"Cthrc1_ligand_receptor_target_gene_GP\",\n",
    "    \"Bmp8a_ligand_receptor_target_gene_GP\",\n",
    "    \"Spint1_ligand_receptor_target_gene_GP\",\n",
    "    \"FGF6_ligand_receptor_GP\",\n",
    "    \"SFRP4_ligand_receptor_GP\",\n",
    "    \"VCAM1_ligand_receptor_GP\",\n",
    "]\n",
    "\n",
    "if gp_resign:\n",
    "    for gp in resign_gps:\n",
    "        model.adata.obs[gp] = -1 * model.adata.obs[gp]\n",
    "        \n",
    "df = model.adata.obs[[\"niche\"] + gps].groupby(\"niche\").mean()\n",
    "df = df.reindex(model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"][::-1])\n",
    "\n",
    "scaler = MinMaxScaler()\n",
    "normalized_columns = scaler.fit_transform(df)\n",
    "normalized_df = pd.DataFrame(normalized_columns, columns=df.columns)\n",
    "normalized_df.index = df.index\n",
    "normalized_df.columns = [col.split(\"_lig\")[0].split(\"_met\")[0].replace(\"_\", \" \").replace(\"9-cis-Retinoic\", \"9-cis-Ret.\").capitalize() + \" GP\" for col in normalized_df.columns]\n",
    "\n",
    "plt.figure(figsize=(16, 6))  # Set the figure size\n",
    "ax = sns.heatmap(normalized_df,\n",
    "            cmap='viridis',\n",
    "            annot=False,\n",
    "            linewidths=0)\n",
    "plt.xticks(rotation=45,\n",
    "           ha=\"right\"\n",
    "          )\n",
    "plt.savefig(f\"{figure_folder_path}/enriched_gps_heatmap.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "plt.figure(figsize=(6, 3))  # Set the figure size\n",
    "ax = sns.heatmap(normalized_df,\n",
    "            cmap='viridis',\n",
    "            annot=False,\n",
    "            linewidths=0,\n",
    "            cbar_kws={\"orientation\": \"horizontal\", \"location\": \"bottom\", \"shrink\": 0.4})\n",
    "plt.xticks([])\n",
    "plt.savefig(f\"{figure_folder_path}/heatmap_colorbar.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "495e6eae-6852-44d2-990a-40fca455ac6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2h ###\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"forebrain_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niches\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=[\"Pcsk1n/Dkk1+ Forebrain\"],\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"midbrain_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niches\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=[\"Fgf17/Efna2+ Midbrain\"],\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"hindbrain_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niches\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=[\"Gdf10/Fgf3+ Hindbrain\"],\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"floorplate_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niches\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=[\"Calca/Shh+ Floor Plate\"],\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32312fc-ee27-400d-9a3d-9d0715ccb5bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2h ###\n",
    "color_map = \"RdGy_r\"\n",
    "adata_top = model.adata[model.adata.obsm[\"spatial\"][:,1] < 0]\n",
    "adata_top = adata_top[adata_top.obs[\"sample\"] == \"embryo2\"]\n",
    "\n",
    "for gp in gps:\n",
    "    scaler = MinMaxScaler()\n",
    "    # Fit and transform the column to normalize its values\n",
    "    adata_top.obs[f'{gp}_normalized'] = scaler.fit_transform(adata_top.obs[[gp]])\n",
    "\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "\n",
    "# Forebrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"PCSK1N_ligand_receptor_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pcsk1n GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"DKK1_ligand_receptor_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Dkk1 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "\n",
    "# Midbrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgf17_ligand_receptor_target_gene_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgf17 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Efna2_ligand_receptor_target_gene_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Efna2 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "\n",
    "# Floor plate\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Calca_ligand_receptor_target_gene_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Calca GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Shh_ligand_receptor_target_gene_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Shh GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "\n",
    "# Hindbrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Gdf10_ligand_receptor_target_gene_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gdf10 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgf3_ligand_receptor_target_gene_GP_normalized\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgf3 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 1],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/brain_enriched_gps_top.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d14c0f8-5792-4cd4-b30d-313f3f36eb1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2g ###\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=5,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "\n",
    "# Forebrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Pcsk1n\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pcsk1n\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Dkk1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Dkk1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "\n",
    "# Midbrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgf17\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgf17\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgfr2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgfr2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Efna2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Efna2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "\n",
    "# Floor plate\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Calca\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Calca\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Shh\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Shh\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "\n",
    "# Hindbrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Gdf10\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gdf10\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgf3\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgf3\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[4, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgfr1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgfr1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[4, 1],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/brain_enriched_gps_genes.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5241ea3-73b8-49e3-a5e7-4c0269dd2cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2jn ###\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                            ncols=2,\n",
    "                            figsize=(15, 15))\n",
    "\n",
    "# Ligand\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Shh\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Shh\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "\n",
    "# Receptor\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Ptch1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ptch1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "\n",
    "# Top 5 target genes\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Nkx2-9\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nkx2-9\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Foxf2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Foxf2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Slit2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Slit2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Mnx1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Mnx1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Foxd1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Foxd1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/shh_ligand_receptor_target_gene_gp.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48080610-50e0-4f3c-9214-bb8e3964f61c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2g ###\n",
    "adata_gut = model.adata[model.adata.obsm[\"spatial\"][:,1] < 2.8]\n",
    "adata_gut = adata_gut[adata_gut.obsm[\"spatial\"][:,1] > -1.6]\n",
    "adata_gut = adata_gut[adata_gut.obsm[\"spatial\"][:,0] < 1.3]\n",
    "adata_gut = adata_gut[adata_gut.obsm[\"spatial\"][:,0] > -1.2]\n",
    "adata_gut = adata_gut[adata_gut.obs[\"sample\"] == \"embryo2\"]\n",
    "\n",
    "for gp in gps:\n",
    "    scaler = MinMaxScaler()\n",
    "    # Fit and transform the column to normalize its values\n",
    "    adata_gut.obs[f'{gp}_normalized'] = scaler.fit_transform(adata_gut.obs[[gp]])\n",
    "\n",
    "color_map = \"RdGy_r\"\n",
    "fig, axs = plt.subplots(nrows=2,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "\n",
    "# Ventral Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"IHH_ligand_receptor_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ihh GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Spint1_ligand_receptor_target_gene_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Spint1 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "\n",
    "# Dorsal Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Cthrc1_ligand_receptor_target_gene_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cthrc1 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"PDGFC_ligand_receptor_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pdgfc GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/gut_enriched_gps.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "471df3dc-75ee-4b1f-a1d0-1f43818ac719",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2g ###\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "\n",
    "# Ventral Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Ihh\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ihh\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Smo\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Smo\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Spint1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Spint1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"St14\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"St14\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "\n",
    "# Dorsal Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Cthrc1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cthrc1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Fzd3\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fzd3\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Pdgfc\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pdgfc\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/gut_enriched_gps_genes.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eb8e808-caad-4caf-9045-1cf13b0da915",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2i ###\n",
    "adata_gut.obs[\"Vtn_ligand_receptor_target_gene_GP\"] = -1 *  adata_gut.obs[\"Vtn_ligand_receptor_target_gene_GP\"] \n",
    "\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 15))\n",
    "\n",
    "# Ligand\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Vtn\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Vtn\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "\n",
    "# Receptor\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Itga2b\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Itga2b\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Itga5\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Itga5\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)    \n",
    "\n",
    "# Target Genes\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Pxdn\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pxdn\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Mecom\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Mecom\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Kdr\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Kdr\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Crem\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Crem\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "\n",
    "# GP\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "adata=adata_gut,\n",
    "color=\"Vtn_ligand_receptor_target_gene_GP\",\n",
    "color_map=color_map,\n",
    "spot_size=spot_size,\n",
    "title=f\"Vtn GP\",\n",
    "legend_loc=None,\n",
    "colorbar_loc=\"bottom\",\n",
    "ax=axs[3, 1],\n",
    "show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/vtn_ligand_receptor_target_gene_gp.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "520df671-8d1e-4385-b269-fd2b8b6538a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extended Data Fig. 1 ###\n",
    "model.model.features_idx_dict_[\"target_reconstructed_rna_idx\"] = np.sort(np.concatenate((model.model.features_idx_dict_[\"target_masked_rna_idx\"], model.model.features_idx_dict_[\"target_unmasked_rna_idx\"])))\n",
    "model.model.features_idx_dict_[\"source_reconstructed_rna_idx\"] = np.sort(np.concatenate((model.model.features_idx_dict_[\"source_masked_rna_idx\"], model.model.features_idx_dict_[\"source_unmasked_rna_idx\"])))\n",
    "\n",
    "output = model.get_omics_decoder_outputs(\n",
    "                adata=model.adata,\n",
    "                only_active_gps=True,\n",
    "                node_batch_size=2048)\n",
    "\n",
    "for gene in [\"Spint1\", \"Ihh\", \"Cthrc1\", \"Pdgfc\"]:\n",
    "    model.adata.obs[f\"{gene}_rec\"] = output[\"source_rna_nb_means\"][:, model.adata.var_names.tolist().index(gene)]\n",
    "\n",
    "for gene in [\"St14\", \"Smo\", \"Fzd3\"]:\n",
    "    model.adata.obs[f\"{gene}_rec\"] = output[\"target_rna_nb_means\"][:, model.adata.var_names.tolist().index(gene)]\n",
    "    \n",
    "adata_gut = model.adata[model.adata.obsm[\"spatial\"][:,1] < 2.8]\n",
    "adata_gut = adata_gut[adata_gut.obsm[\"spatial\"][:,1] > -1.6]\n",
    "adata_gut = adata_gut[adata_gut.obsm[\"spatial\"][:,0] < 1.3]\n",
    "adata_gut = adata_gut[adata_gut.obsm[\"spatial\"][:,0] > -1.2]\n",
    "adata_gut = adata_gut[adata_gut.obs[\"sample\"] == \"embryo2\"]\n",
    "\n",
    "# Reconstructed gene expression\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "# Ventral Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Spint1_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Spint1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"St14_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"St14\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Ihh_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ihh\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Smo_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Smo\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "\n",
    "# Dorsal Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Cthrc1_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cthrc1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Fzd3_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fzd3\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)    \n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Pdgfc_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pdgfc\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/gut_enriched_gps_reconstructed_genes.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a004f74a-ffb0-4c6b-8fbb-a280a9dd6565",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extended Data Fig. 1 ###\n",
    "for gene in [\"Pcsk1n\", \"Dkk1\", \"Fgf17\", \"Efna2\", \"Calca\", \"Shh\", \"Gdf10\", \"Fgf3\"]:\n",
    "    model.adata.obs[f\"{gene}_rec\"] = output[\"source_rna_nb_means\"][:, model.adata.var_names.tolist().index(gene)]\n",
    "\n",
    "for gene in [\"Fgfr2\", \"Ptch1\", \"Fgfr1\"]:\n",
    "    model.adata.obs[f\"{gene}_rec\"] = output[\"target_rna_nb_means\"][:, model.adata.var_names.tolist().index(gene)]\n",
    "\n",
    "adata_top = model.adata[model.adata.obsm[\"spatial\"][:,1] < 0]\n",
    "adata_top = adata_top[adata_top.obs[\"sample\"] == \"embryo2\"]\n",
    "\n",
    "# Reconstructed gene expression\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "# Forebrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Pcsk1n_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Pcsk1n\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Dkk1_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Dkk1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "\n",
    "# Midbrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgf17_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgf17\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgfr2_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgfr2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Efna2_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Efna2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "\n",
    "# Floor plate\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Calca_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Calca\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Shh_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Shh\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Ptch1_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ptch1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 1],\n",
    "    show=False)\n",
    "\n",
    "# Hindbrain\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Gdf10_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gdf10\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[4, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgf3_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgf3\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[4, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Fgfr1_rec\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fgfr1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[5, 0],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/brain_enriched_gps_reconstructed_genes.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16308f93-8345-4098-97d4-7b35264e962e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 4 ###\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "# Floor Plate markers\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Foxa1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Foxa1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Foxa2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Foxa2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_top,\n",
    "    color=\"Ntn1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ntn1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/floor_plate_markers.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b9d8264-cc17-4e34-9ddf-62599b6148cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2i ###\n",
    "niche_groups = [\"IL10/Fgf2+ Vasculature (Angiogenesis)\"]\n",
    "\n",
    "if len(niche_groups) > 0:\n",
    "    # Plot latent cluster groups in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/angiogenesis_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=f\"{'_'.join(niche_groups)}\",\n",
    "        cat_key=\"niche\",\n",
    "        groups=niche_groups,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=niche_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)\n",
    "\n",
    "niche_groups = [\"Bmp8a/9-cis-Ret. acid+ Splanch. Mesoderm\"]\n",
    "\n",
    "if len(niche_groups) > 0:\n",
    "    # Plot latent cluster groups in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/splanchnic_mesoderm_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=f\"{'_'.join(niche_groups)}\",\n",
    "        cat_key=\"niche\",\n",
    "        groups=niche_groups,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=niche_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "744406e6-24a1-4de5-b817-40a2c4f63b03",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2i,j: Niche cell-cell communication ###\n",
    "model.adata.obs[\"niche_abr\"] = model.adata.obs[\"niche\"].apply(lambda x: x.split(\"GP+ \")[1].replace(\" \", \"\\n\") if \"GP+ \" in x else x)\n",
    "niche_colors = {niche: latent_cluster_colors[cluster] for cluster, niche in niche_annotation_dict.items()}\n",
    "niche_abreviated_colors = {}\n",
    "\n",
    "for key, value in niche_colors.items():\n",
    "    if \"GP+ \" in key:\n",
    "        niche_abreviated_colors[key.split(\"GP+ \")[1].replace(\" \", \"\\n\")] = value\n",
    "    else:\n",
    "        niche_abreviated_colors[key] = value\n",
    "\n",
    "\n",
    "network_df = compute_communication_gp_network(\n",
    "    gp_list=[\"Shh_ligand_receptor_target_gene_GP\", \"Vtn_ligand_receptor_target_gene_GP\"],\n",
    "    model=model,\n",
    "    group_key=\"niche_abr\", # celltype\n",
    "    n_neighbors=90)\n",
    "\n",
    "shh_network_df = network_df[network_df[\"edge_type\"] == \"Shh_ligand_receptor_target_gene_GP\"]\n",
    "vtn_network_df = network_df[network_df[\"edge_type\"] == \"Vtn_ligand_receptor_target_gene_GP\"]\n",
    "\n",
    "visualize_communication_gp_network(\n",
    "    adata=model.adata,\n",
    "    network_df=shh_network_df,\n",
    "    figsize=(12, 7),\n",
    "    cat_colors=niche_abreviated_colors,\n",
    "    edge_type_colors=[\"#1f77b4\"], \n",
    "    cat_key=\"niche_abr\",\n",
    "    save=True,\n",
    "    save_path=f\"{figure_folder_path}/gp_network_shh.svg\",\n",
    "    )\n",
    "\n",
    "visualize_communication_gp_network(\n",
    "    adata=model.adata,\n",
    "    network_df=vtn_network_df,\n",
    "    figsize=(8, 7),\n",
    "    cat_colors=niche_abreviated_colors,\n",
    "    edge_type_colors=[\"#ff7f0e\"],\n",
    "    cat_key=\"niche_abr\",\n",
    "    save=True,\n",
    "    save_path=f\"{figure_folder_path}/gp_network_vtn.svg\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ee5037-443f-462f-bd68-8ff4fd1e44c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 3 ###\n",
    "def return_niche_cell_types(row, niche):\n",
    "    if row[\"niche\"] == niche:\n",
    "        return row[cell_type_key]\n",
    "    else:\n",
    "        return \"-1\"\n",
    "\n",
    "niche = \"Edn1/Nppa1+ Developing Heart\"\n",
    "model.adata.obs[f\"{niche}_{cell_type_key}\"] = model.adata.obs.apply(\n",
    "    lambda row: return_niche_cell_types(row, niche), axis=1)\n",
    "\n",
    "spot_size = 0.15\n",
    "\n",
    "cell_type_groups = [\"Cardiomyocytes\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"{cell_type_groups[0].lower()}_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "    cat_key=f\"{niche}_{cell_type_key}\",\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(1440000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"{niche.split('+ ')[1].lower()}_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=niche,\n",
    "    cat_key=\"niche\",\n",
    "    groups=[niche],\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(1440000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00d1bd0a-89be-4110-95d8-75e55a8c20f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 3 ###\n",
    "def return_niche_cell_types(row, niche):\n",
    "    if row[\"niche\"] == niche:\n",
    "        return row[cell_type_key]\n",
    "    else:\n",
    "        return \"-1\"\n",
    "\n",
    "niche = \"IL10/Fgf2+ Vasculature (Angiogenesis)\"\n",
    "model.adata.obs[f\"{niche}_{cell_type_key}\"] = model.adata.obs.apply(\n",
    "    lambda row: return_niche_cell_types(row, niche), axis=1)\n",
    "\n",
    "spot_size = 0.15\n",
    "\n",
    "cell_type_groups = [\"Endothelium\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"{cell_type_groups[0].lower()}_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "    cat_key=f\"{niche}_{cell_type_key}\",\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "spot_size = 0.15\n",
    "\n",
    "cell_type_groups = [\"Haematoendothelial progenitors\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"{cell_type_groups[0].lower()}_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "    cat_key=f\"{niche}_{cell_type_key}\",\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"{niche.split('+ ')[1].lower()}_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=niche,\n",
    "    cat_key=\"niche\",\n",
    "    groups=[niche],\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f7e371e-21b1-4962-88bf-e228392fbff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 3 ###\n",
    "def return_niche_cell_types(row, niches):\n",
    "    if row[\"niche\"] in niches:\n",
    "        return row[cell_type_key]\n",
    "    else:\n",
    "        return \"-1\"\n",
    "\n",
    "niches = [\"Fgf17/Efna2+ Midbrain\",\n",
    "          \"Pcsk1n/Dkk1+ Forebrain\",\n",
    "          \"Calca/Shh+ Floor Plate\",\n",
    "          \"Gdf10/Fgf3+ Hindbrain\"]\n",
    "model.adata.obs[f\"brain_niches_{cell_type_key}\"] = model.adata.obs.apply(\n",
    "    lambda row: return_niche_cell_types(row, niches), axis=1)\n",
    "\n",
    "spot_size = 0.15\n",
    "\n",
    "cell_type_groups = [\"Forebrain/Midbrain/Hindbrain\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"brain_cell_types_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "    cat_key=f\"brain_niches_{cell_type_key}\",\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"brain_niches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=niches,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e1ba759-0798-4bee-9e4a-07a47a3441b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 3 ###\n",
    "def return_niche_cell_types(row, niches):\n",
    "    if row[\"niche\"] in niches:\n",
    "        return row[cell_type_key]\n",
    "    else:\n",
    "        return \"-1\"\n",
    "\n",
    "niches = [\"Ihh/Spint1+ Ventral Gut\",\n",
    "          \"Cthrc1/Pdgfc+ Dorsal Gut\"]\n",
    "model.adata.obs[f\"gut_niches_{cell_type_key}\"] = model.adata.obs.apply(\n",
    "    lambda row: return_niche_cell_types(row, niches), axis=1)\n",
    "\n",
    "spot_size = 0.15\n",
    "\n",
    "cell_type_groups = [\"Gut tube\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"gut_tube_cell_types_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "    cat_key=f\"gut_niches_{cell_type_key}\",\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"gut_tube_niches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=niches,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61970f49-7349-4d3c-bda7-a37a1ee4f0bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 5 ###\n",
    "spot_size = 0.15\n",
    "\n",
    "cell_type_groups = [\"Allantois\", \"Presomitic mesoderm\", \"Lateral plate mesoderm\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/{'_'.join(cell_type_groups).lower()}\" \\\n",
    "            \"_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=f\"{'_'.join(cell_type_groups)}\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "niches = [\"Bmp4/Vegfb+ Lateral Plate Mesoderm\",\n",
    "          \"Pgf/Cd44+ Allantois\",\n",
    "          \"Fgf18/Dll1+ Presomitic Mesoderm\"]\n",
    "\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"missing_niches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=niches,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9f27b58-9cf6-4264-afa1-e63f28ea3168",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 4b ###\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "\n",
    "# Ventral Gut\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Nog\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nog\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=adata_gut,\n",
    "    color=\"Cthrc1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cthrc1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/notochord_markers.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2f84a6-af5e-440d-8499-1aa69d34bffc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 7a,c ###\n",
    "color_map = \"RdPu\"\n",
    "fig, axs = plt.subplots(nrows=4,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "# Vtn GP\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "    color=\"Vtn_ligand_receptor_target_gene_GP_source_score\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Vtn GP source score\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "    color=\"Vtn_ligand_receptor_target_gene_GP_target_score\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Vtn GP target score\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "# Shh GP\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "    color=\"Shh_ligand_receptor_target_gene_GP_source_score\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Shh GP source score\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata[model.adata.obs[\"sample\"] == \"embryo2\"],\n",
    "    color=\"Shh_ligand_receptor_target_gene_GP_target_score\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Shh GP target score\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/cci_gp_source_target_scores.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a623f548-e5c2-4d7d-8efd-237cea820c30",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 7b,d ###\n",
    "network_df = compute_communication_gp_network(\n",
    "    gp_list=[\"Shh_ligand_receptor_target_gene_GP\", \"Vtn_ligand_receptor_target_gene_GP\"],\n",
    "    model=model,\n",
    "    group_key=cell_type_key, # celltype\n",
    "    n_neighbors=90)\n",
    "\n",
    "shh_network_df = network_df[network_df[\"edge_type\"] == \"Shh_ligand_receptor_target_gene_GP\"]\n",
    "vtn_network_df = network_df[network_df[\"edge_type\"] == \"Vtn_ligand_receptor_target_gene_GP\"]\n",
    "\n",
    "visualize_communication_gp_network(\n",
    "    adata=model.adata,\n",
    "    network_df=shh_network_df,\n",
    "    figsize=(10, 8),\n",
    "    cat_colors=cell_type_colors,\n",
    "    edge_type_colors=[\"#1f77b4\"], \n",
    "    cat_key=cell_type_key,\n",
    "    save=True,\n",
    "    save_path=f\"{figure_folder_path}/gp_network_shh_cell_types.svg\",\n",
    "    )\n",
    "\n",
    "visualize_communication_gp_network(\n",
    "    adata=model.adata,\n",
    "    network_df=vtn_network_df,\n",
    "    figsize=(12, 9),\n",
    "    cat_colors=cell_type_colors,\n",
    "    edge_type_colors=[\"#ff7f0e\"],\n",
    "    cat_key=cell_type_key,\n",
    "    save=True,\n",
    "    save_path=f\"{figure_folder_path}/gp_network_vtn_cell_types.svg\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc0f9c9-9a57-4387-9f20-ea9d2776e383",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 7b,d ###\n",
    "cell_type_groups = [\"Definitive endoderm\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"definitive_endoderm_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Definitive endoderm\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "cell_type_groups = [\"Forebrain/Midbrain/Hindbrain\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"forebrain_midbrain_hindbrain_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Forebrain/Midbrain/Hindbrain\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "cell_type_groups = [\"Gut tube\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"gut_tube_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Gut tube\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)\n",
    "\n",
    "cell_type_groups = [\"Endothelium\"]\n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/\" \\\n",
    "            f\"endothelium_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Endothelium\",\n",
    "    cat_key=cell_type_key,\n",
    "    groups=cell_type_groups,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=cell_type_colors,\n",
    "    size=(2160000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aeb14bb-a680-41f2-88d1-6df8ca73950d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Fig. 2g,h ###\n",
    "for latent_cluster in model.adata.obs[latent_cluster_key].unique().tolist():\n",
    "    # Latent clusters in latent and physical space\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "                f\"latent_cluster_{latent_cluster}_latent_physical_space.svg\"\n",
    "\n",
    "    plot_category_in_latent_and_physical_space(\n",
    "        adata=model.adata,\n",
    "        plot_label=\"Latent Clusters\",\n",
    "        cat_key=latent_cluster_key,\n",
    "        groups=latent_cluster,\n",
    "        sample_key=sample_key,\n",
    "        samples=samples,\n",
    "        cat_colors=latent_cluster_colors,\n",
    "        size=(720000 / len(model.adata)),\n",
    "        spot_size=spot_size,\n",
    "        save_fig=save_fig,\n",
    "        file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2015d708-728b-4311-97b2-49db64988394",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find best Leiden resolution\n",
    "sc.pp.normalize_total(model.adata, target_sum=1e4)\n",
    "sc.pp.log1p(model.adata)\n",
    "# model.adata.raw = model.adata  # freeze the state in `.raw`\n",
    "\n",
    "sc.pp.highly_variable_genes(model.adata,\n",
    "                            subset=False,\n",
    "                            n_top_genes=2000)\n",
    "\n",
    "sc.pp.pca(model.adata)\n",
    "\n",
    "sc.pp.neighbors(model.adata,\n",
    "                use_rep='X_pca')\n",
    "\n",
    "sc.tl.umap(model.adata)\n",
    "\n",
    "# Find resolution with highest silhouette\n",
    "resolutions = [0.1,\n",
    "               0.15,\n",
    "               0.2,\n",
    "               0.25,\n",
    "               0.3,\n",
    "               0.35,\n",
    "               0.4,\n",
    "               0.45,\n",
    "               0.5,\n",
    "               0.55,\n",
    "               0.6,\n",
    "               0.65,\n",
    "               0.7,\n",
    "               0.75,\n",
    "               0.8,\n",
    "               0.85,\n",
    "               0.9,\n",
    "               0.95,\n",
    "               1,\n",
    "               1.05,\n",
    "               1.1,\n",
    "               1.15,\n",
    "               1.2,\n",
    "               1.25,\n",
    "               1.3,\n",
    "               1.35,\n",
    "               1.4,\n",
    "               1.45,\n",
    "               1.5,\n",
    "               1.55,\n",
    "               1.6,\n",
    "               1.65,\n",
    "               1.7,\n",
    "               1.75,\n",
    "               1.8,\n",
    "               1.85,\n",
    "               1.9,\n",
    "               1.95,\n",
    "               2]\n",
    "\n",
    "silhouette_scores = []\n",
    "\n",
    "for res in resolutions:\n",
    "    sc.tl.leiden(model.adata, resolution=res, key_added=f'leiden_{res}')\n",
    "    \n",
    "    labels = model.adata.obs[f'leiden_{res}']\n",
    "    silhouette_avg = silhouette_score(model.adata.obsm['X_pca'], labels, metric='euclidean')\n",
    "    silhouette_scores.append(silhouette_avg)\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.bar(range(len(resolutions)), silhouette_scores, color='skyblue', edgecolor='black')\n",
    "plt.xticks(range(len(resolutions)), [str(r) for r in resolutions],fontsize=5)\n",
    "plt.xlabel('Resolution')\n",
    "plt.ylabel('Silhouette Score')\n",
    "plt.title('Silhouette Scores for Different Leiden Resolutions')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "369c9319-7948-4ae6-ad0d-70babac7c18d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get top differentially expressed genes\n",
    "# Perform clustering with best Leiden resolution\n",
    "latent_leiden_resolution = 0.55\n",
    "sc.tl.leiden(model.adata,\n",
    "             resolution=latent_leiden_resolution,\n",
    "             key_added=f'leiden_{latent_leiden_resolution}')\n",
    "model.adata.obs[f'leiden_{latent_leiden_resolution}'] = model.adata.obs[\n",
    "    f'leiden_{latent_leiden_resolution}'].astype(str).apply(lambda x: f'c_{x}')\n",
    "\n",
    "# Rank genes for each cluster against all other clusters\n",
    "sc.tl.rank_genes_groups(model.adata,\n",
    "                        f'leiden_{latent_leiden_resolution}',\n",
    "                        method='wilcoxon')\n",
    "\n",
    "# Get the top 10/20 most significantly upregulated genes for each cluster\n",
    "top_10_degs = set()\n",
    "top_20_degs = set()\n",
    "for cluster in model.adata.obs[f'leiden_{latent_leiden_resolution}'].cat.categories:\n",
    "    cluster_genes = model.adata.uns['rank_genes_groups']['names'][cluster]\n",
    "    cluster_pvals_adj = model.adata.uns['rank_genes_groups']['pvals_adj'][cluster]\n",
    "    cluster_logfoldchanges = model.adata.uns['rank_genes_groups']['logfoldchanges'][cluster]\n",
    "    cluster_df = pd.DataFrame({\n",
    "        'gene': cluster_genes,\n",
    "        'pval_adj': cluster_pvals_adj,\n",
    "        'logfoldchange': cluster_logfoldchanges\n",
    "    })\n",
    "\n",
    "    # Filter for significant up-regulated genes\n",
    "    significant_genes = cluster_df[(cluster_df['pval_adj'] < 0.05)]\n",
    "\n",
    "    # Select the top 10/20 genes based on p-value\n",
    "    top_10_degs_cluster = significant_genes.head(10)['gene']\n",
    "    top_20_degs_cluster = significant_genes.head(20)['gene']\n",
    "    top_10_degs.update(top_10_degs_cluster)\n",
    "    top_20_degs.update(top_20_degs_cluster)\n",
    "\n",
    "top_10_degs = list(top_10_degs)\n",
    "top_20_degs = list(top_20_degs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "657de625-f5bb-47da-8ed1-d34bd389f110",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get top enriched GP genes\n",
    "# Perform differential testing\n",
    "selected_cats = None\n",
    "comparison_cats = \"rest\"\n",
    "title = f\"NicheCompass Strongly Enriched Niche GPs\"\n",
    "log_bayes_factor_thresh = 2.3\n",
    "enriched_gps = model.run_differential_gp_tests(\n",
    "    cat_key='niche',\n",
    "    selected_cats=selected_cats,\n",
    "    comparison_cats=comparison_cats,\n",
    "    log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "df = model.adata.uns['nichecompass_differential_gp_test_results']\n",
    "gp_list_without_denovo_gps = df[~df['gene_program'].str.startswith('Add-')]\n",
    "target_gps = list(np.unique(gp_list_without_denovo_gps['gene_program']))\n",
    "\n",
    "filtered_df = gp_summary_df[gp_summary_df['gp_name'].isin(target_gps)]\n",
    "\n",
    "# Function to filter genes based on importance\n",
    "def filter_genes_by_importance(genes, importances, threshold=0.05):\n",
    "    return [gene for gene, importance in zip(genes, importances) if importance > threshold]\n",
    "\n",
    "# Apply filtering function to source and target genes\n",
    "filtered_df['filtered_source_genes'] = filtered_df.apply(lambda row: filter_genes_by_importance(row['gp_source_genes'], row['gp_source_genes_importances']), axis=1)\n",
    "filtered_df['filtered_target_genes'] = filtered_df.apply(lambda row: filter_genes_by_importance(row['gp_target_genes'], row['gp_target_genes_importances']), axis=1)\n",
    "\n",
    "# Combine the filtered source and target genes into a single list\n",
    "filtered_df['total_genes'] = filtered_df.apply(lambda row: row['filtered_source_genes'] + row['filtered_target_genes'], axis=1)\n",
    "\n",
    "# Aggregate all genes from the filtered GPs\n",
    "top_enriched = set(gene for genes in filtered_df['total_genes'] for gene in genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93eb94a9-fc9b-49fc-9707-8ce724f49a34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get characterizing GP genes\n",
    "target_gps = [\n",
    "    'Cdh13_ligand_receptor_target_gene_GP', 'Col1a1_ligand_receptor_target_gene_GP',\n",
    "    'Bmp4_ligand_receptor_target_gene_GP', 'Vegfb_ligand_receptor_target_gene_GP',\n",
    "    'Pgf_ligand_receptor_target_gene_GP', 'Cd44_ligand_receptor_target_gene_GP',\n",
    "    'Bmp8a_ligand_receptor_target_gene_GP', '9-cis-Retinoic_acid_metabolite_enzyme_sensor_GP',\n",
    "    'Edn1_ligand_receptor_target_gene_GP', 'NPPA_ligand_receptor_GP',\n",
    "    'Cthrc1_ligand_receptor_target_gene_GP', 'Pdgfc_ligand_receptor_target_gene_GP',\n",
    "    'Ihh_ligand_receptor_target_gene_GP', 'Spint1_ligand_receptor_target_gene_GP',\n",
    "    'SFRP4_ligand_receptor_GP', 'PDGFA_ligand_receptor_GP',\n",
    "    'Guanine_metabolite_enzyme_sensor_GP', 'Slitrk1_ligand_receptor_target_gene_GP',\n",
    "    'Tgfbi_ligand_receptor_target_gene_GP', 'FGF6_ligand_receptor_GP',\n",
    "    'Epo_ligand_receptor_target_gene_GP', 'Thpo_ligand_receptor_target_gene_GP',\n",
    "    'Il10_ligand_receptor_target_gene_GP', 'Fgf2_ligand_receptor_target_gene_GP',\n",
    "    'Fgf17_ligand_receptor_target_gene_GP', 'Efna2_ligand_receptor_target_gene_GP',\n",
    "    'PCSK1N_ligand_receptor_GP', 'DKK1_ligand_receptor_GP',\n",
    "    'Calca_ligand_receptor_target_gene_GP', 'Shh_ligand_receptor_target_gene_GP',\n",
    "    'Gdf10_ligand_receptor_target_gene_GP', 'Fgf3_ligand_receptor_target_gene_GP',\n",
    "    'NPTN_ligand_receptor_GP', 'Fgf8_ligand_receptor_target_gene_GP',\n",
    "    'SAA1_ligand_receptor_GP', 'IL6ST_ligand_receptor_GP',\n",
    "    'Fgf18_ligand_receptor_target_gene_GP', 'Dll1_ligand_receptor_target_gene_GP'\n",
    "]\n",
    "\n",
    "# Filter the dataframe for the specified GPs\n",
    "filtered_df = gp_summary_df[gp_summary_df['gp_name'].isin(target_gps)]\n",
    "\n",
    "# Function to filter genes based on importance\n",
    "def filter_genes_by_importance(genes, importances, threshold=0.05):\n",
    "    return [gene for gene, importance in zip(genes, importances) if importance > threshold]\n",
    "\n",
    "# Apply filtering function to source and target genes\n",
    "filtered_df['filtered_source_genes'] = filtered_df.apply(lambda row: filter_genes_by_importance(row['gp_source_genes'], row['gp_source_genes_importances']), axis=1)\n",
    "filtered_df['filtered_target_genes'] = filtered_df.apply(lambda row: filter_genes_by_importance(row['gp_target_genes'], row['gp_target_genes_importances']), axis=1)\n",
    "\n",
    "# Combine the filtered source and target genes into a single list\n",
    "filtered_df['total_genes'] = filtered_df.apply(lambda row: row['filtered_source_genes'] + row['filtered_target_genes'], axis=1)\n",
    "\n",
    "# Aggregate all genes from the filtered GPs\n",
    "top_characterizing = set(gene for genes in filtered_df['total_genes'] for gene in genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dacb25bd-b2f1-4016-93c2-dc59bc493c91",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 35 ###\n",
    "np.random.seed(42)\n",
    "\n",
    "overlap_top_enriched_top_10_degs = top_enriched.intersection(top_10_degs)\n",
    "\n",
    "# Compute random baseline (overlap between differentially expressed genes and random genes)\n",
    "all_genes = list(model.adata.var_names)\n",
    "\n",
    "# Number of random selections and number of genes to select each time\n",
    "n_selections = 10\n",
    "n_genes_to_select = len(top_10_degs)\n",
    "\n",
    "overlap_random_top_enriched = []\n",
    "overlap_random_top_10_degs = []\n",
    "overlap_random_both = []\n",
    "\n",
    "# Perform the random selections and calculate overlaps\n",
    "for _ in range(n_selections):\n",
    "    random_genes = set(np.random.choice(all_genes, n_genes_to_select, replace=False))\n",
    "    overlap_top_enriched = len(random_genes.intersection(top_enriched))\n",
    "    overlap_top_10_degs = len(random_genes.intersection(top_10_degs))\n",
    "    overlap_both = len(random_genes.intersection(overlap_top_enriched_top_10_degs))\n",
    "    overlap_random_top_enriched.append(overlap_top_enriched)\n",
    "    overlap_random_top_10_degs.append(overlap_top_10_degs)\n",
    "    overlap_random_both.append(overlap_both)\n",
    "\n",
    "mean_overlap_random_top_enriched = np.mean(overlap_random_top_enriched)\n",
    "mean_overlap_random_top_10_degs = np.mean(overlap_random_top_10_degs)\n",
    "mean_overlap_random_both = np.mean(overlap_random_both)\n",
    "\n",
    "mean_enriched_only = len(top_enriched) - len(overlap_top_enriched_top_10_degs) - mean_overlap_random_top_enriched\n",
    "mean_top_10_degs_only = len(top_10_degs) - len(overlap_top_enriched_top_10_degs) - mean_overlap_random_top_10_degs\n",
    "mean_random_only = len(random_genes) - mean_overlap_random_top_enriched - mean_overlap_random_top_10_degs\n",
    "\n",
    "print(f\"Mean enriched only: {mean_enriched_only}\")\n",
    "print(f\"Mean top 10 DEGs only: {mean_top_10_degs_only}\")\n",
    "print(f\"Mean random only: {mean_random_only}\")\n",
    "print(f\"Mean overlap random & enriched: {mean_overlap_random_top_enriched}\")\n",
    "print(f\"Mean overlap random & top 10 DEGs: {mean_overlap_random_top_10_degs}\")\n",
    "print(f\"Overlap enriched & top 10 DEGs: {len(overlap_top_enriched_top_10_degs)}\")\n",
    "print(f\"Mean overlap all: {mean_overlap_random_both}\")\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn=venn3(subsets=(mean_enriched_only,\n",
    "                    mean_top_10_degs_only,\n",
    "                    len(overlap_top_enriched_top_10_degs),\n",
    "                    round(mean_random_only, 2),\n",
    "                    mean_overlap_random_top_enriched,\n",
    "                    mean_overlap_random_top_10_degs,\n",
    "                    mean_overlap_random_both),\n",
    "           set_labels=('Enriched GP genes',\n",
    "                       'Top 10 DEGs in non-spatial clustering',\n",
    "                       'Randomly selected baseline genes'))\n",
    "\n",
    "font_properties = font_manager.FontProperties(size=12)\n",
    "for label in venn.set_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "for label in venn.subset_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "\n",
    "plt.savefig(f'{figure_folder_path}/overlap_enriched_10_degs.svg', format='svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b53a8a68-36b8-4ab5-be33-c267d3a3c332",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 35 ###\n",
    "np.random.seed(42)\n",
    "\n",
    "overlap_top_enriched_top_20_degs = top_enriched.intersection(top_20_degs)\n",
    "\n",
    "# Compute random baseline (overlap between differentially expressed genes and random genes)\n",
    "all_genes = list(model.adata.var_names)\n",
    "\n",
    "# Number of random selections and number of genes to select each time\n",
    "n_selections = 10\n",
    "n_genes_to_select = len(top_20_degs)\n",
    "\n",
    "overlap_random_top_enriched = []\n",
    "overlap_random_top_20_degs = []\n",
    "overlap_random_both = []\n",
    "\n",
    "# Perform the random selections and calculate overlaps\n",
    "for _ in range(n_selections):\n",
    "    random_genes = set(np.random.choice(all_genes, n_genes_to_select, replace=False))\n",
    "    overlap_top_enriched = len(random_genes.intersection(top_enriched))\n",
    "    overlap_top_20_degs = len(random_genes.intersection(top_20_degs))\n",
    "    overlap_both = len(random_genes.intersection(overlap_top_enriched_top_20_degs))\n",
    "    overlap_random_top_enriched.append(overlap_top_enriched)\n",
    "    overlap_random_top_20_degs.append(overlap_top_20_degs)\n",
    "    overlap_random_both.append(overlap_both)\n",
    "\n",
    "mean_overlap_random_top_enriched = np.mean(overlap_random_top_enriched)\n",
    "mean_overlap_random_top_20_degs = np.mean(overlap_random_top_20_degs)\n",
    "mean_overlap_random_both = np.mean(overlap_random_both)\n",
    "\n",
    "mean_enriched_only = len(top_enriched) - len(overlap_top_enriched_top_20_degs) - mean_overlap_random_top_enriched\n",
    "mean_top_20_degs_only = len(top_20_degs) - len(overlap_top_enriched_top_20_degs) - mean_overlap_random_top_20_degs\n",
    "mean_random_only = len(random_genes) - mean_overlap_random_top_enriched - mean_overlap_random_top_20_degs\n",
    "\n",
    "print(f\"Mean enriched only: {mean_enriched_only}\")\n",
    "print(f\"Mean top 20 DEGs only: {mean_top_20_degs_only}\")\n",
    "print(f\"Mean random only: {mean_random_only}\")\n",
    "print(f\"Mean overlap random & enriched: {mean_overlap_random_top_enriched}\")\n",
    "print(f\"Mean overlap random & top 20 DEGs: {mean_overlap_random_top_20_degs}\")\n",
    "print(f\"Overlap enriched & top 20 DEGs: {len(overlap_top_enriched_top_20_degs)}\")\n",
    "print(f\"Mean overlap all: {mean_overlap_random_both}\")\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn=venn3(subsets=(mean_enriched_only,\n",
    "                    mean_top_20_degs_only,\n",
    "                    len(overlap_top_enriched_top_20_degs),\n",
    "                    round(mean_random_only, 2),\n",
    "                    mean_overlap_random_top_enriched,\n",
    "                    mean_overlap_random_top_20_degs,\n",
    "                    mean_overlap_random_both),\n",
    "           set_labels=('Enriched GP genes',\n",
    "                       'Top 20 DEGs in non-spatial clustering',\n",
    "                       'Randomly selected baseline genes'))\n",
    "\n",
    "font_properties = font_manager.FontProperties(size=12)\n",
    "for label in venn.set_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "for label in venn.subset_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "\n",
    "plt.savefig(f'{figure_folder_path}/overlap_enriched_20_degs.svg', format='svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e75ec849-d04a-4c03-b381-7443e8b576d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 35 ###\n",
    "np.random.seed(42)\n",
    "\n",
    "overlap_top_characterizing_top_10_degs = top_characterizing.intersection(top_10_degs)\n",
    "\n",
    "# Compute random baseline (overlap between differentially expressed genes and random genes)\n",
    "all_genes = list(model.adata.var_names)\n",
    "\n",
    "# Number of random selections and number of genes to select each time\n",
    "n_selections = 10\n",
    "n_genes_to_select = len(top_10_degs)\n",
    "\n",
    "overlap_random_top_characterizing = []\n",
    "overlap_random_top_10_degs = []\n",
    "overlap_random_both = []\n",
    "\n",
    "# Perform the random selections and calculate overlaps\n",
    "for _ in range(n_selections):\n",
    "    random_genes = set(np.random.choice(all_genes, n_genes_to_select, replace=False))\n",
    "    overlap_top_characterizing = len(random_genes.intersection(top_characterizing))\n",
    "    overlap_top_10_degs = len(random_genes.intersection(top_10_degs))\n",
    "    overlap_both = len(random_genes.intersection(overlap_top_characterizing_top_10_degs))\n",
    "    overlap_random_top_characterizing.append(overlap_top_characterizing)\n",
    "    overlap_random_top_10_degs.append(overlap_top_10_degs)\n",
    "    overlap_random_both.append(overlap_both)\n",
    "\n",
    "mean_overlap_random_top_characterizing = np.mean(overlap_random_top_characterizing)\n",
    "mean_overlap_random_top_10_degs = np.mean(overlap_random_top_10_degs)\n",
    "mean_overlap_random_both = np.mean(overlap_random_both)\n",
    "\n",
    "mean_characterizing_only = len(top_characterizing) - len(overlap_top_characterizing_top_10_degs) - mean_overlap_random_top_characterizing\n",
    "mean_top_10_degs_only = len(top_10_degs) - len(overlap_top_characterizing_top_10_degs) - mean_overlap_random_top_10_degs\n",
    "mean_random_only = len(random_genes) - mean_overlap_random_top_characterizing - mean_overlap_random_top_10_degs\n",
    "\n",
    "print(f\"Mean characterizing only: {mean_characterizing_only}\")\n",
    "print(f\"Mean top 10 DEGs only: {mean_top_10_degs_only}\")\n",
    "print(f\"Mean random only: {mean_random_only}\")\n",
    "print(f\"Mean overlap random & characterizing: {mean_overlap_random_top_characterizing}\")\n",
    "print(f\"Mean overlap random & top 10 DEGs: {mean_overlap_random_top_10_degs}\")\n",
    "print(f\"Overlap characterizing & top 10 DEGs: {len(overlap_top_characterizing_top_10_degs)}\")\n",
    "print(f\"Mean overlap all: {mean_overlap_random_both}\")\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn=venn3(subsets=(mean_characterizing_only,\n",
    "                    mean_top_10_degs_only,\n",
    "                    len(overlap_top_characterizing_top_10_degs),\n",
    "                    round(mean_random_only, 2),\n",
    "                    mean_overlap_random_top_characterizing,\n",
    "                    mean_overlap_random_top_10_degs,\n",
    "                    mean_overlap_random_both),\n",
    "           set_labels=('characterizing GP genes',\n",
    "                       'Top 10 DEGs in non-spatial clustering',\n",
    "                       'Randomly selected baseline genes'))\n",
    "\n",
    "font_properties = font_manager.FontProperties(size=12)\n",
    "for label in venn.set_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "for label in venn.subset_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "\n",
    "plt.savefig(f'{figure_folder_path}/overlap_characterizing_10_degs.svg', format='svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd67b5e9-9975-4a67-993d-60a507ff19bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 35 ###\n",
    "np.random.seed(42)\n",
    "\n",
    "overlap_top_characterizing_top_20_degs = top_characterizing.intersection(top_20_degs)\n",
    "\n",
    "# Compute random baseline (overlap between differentially expressed genes and random genes)\n",
    "all_genes = list(model.adata.var_names)\n",
    "\n",
    "# Number of random selections and number of genes to select each time\n",
    "n_selections = 10\n",
    "n_genes_to_select = len(top_20_degs)\n",
    "\n",
    "overlap_random_top_characterizing = []\n",
    "overlap_random_top_20_degs = []\n",
    "overlap_random_both = []\n",
    "\n",
    "# Perform the random selections and calculate overlaps\n",
    "for _ in range(n_selections):\n",
    "    random_genes = set(np.random.choice(all_genes, n_genes_to_select, replace=False))\n",
    "    overlap_top_characterizing = len(random_genes.intersection(top_characterizing))\n",
    "    overlap_top_20_degs = len(random_genes.intersection(top_20_degs))\n",
    "    overlap_both = len(random_genes.intersection(overlap_top_characterizing_top_20_degs))\n",
    "    overlap_random_top_characterizing.append(overlap_top_characterizing)\n",
    "    overlap_random_top_20_degs.append(overlap_top_20_degs)\n",
    "    overlap_random_both.append(overlap_both)\n",
    "\n",
    "mean_overlap_random_top_characterizing = np.mean(overlap_random_top_characterizing)\n",
    "mean_overlap_random_top_20_degs = np.mean(overlap_random_top_20_degs)\n",
    "mean_overlap_random_both = np.mean(overlap_random_both)\n",
    "\n",
    "mean_characterizing_only = len(top_characterizing) - len(overlap_top_characterizing_top_20_degs) - mean_overlap_random_top_characterizing\n",
    "mean_top_20_degs_only = len(top_20_degs) - len(overlap_top_characterizing_top_20_degs) - mean_overlap_random_top_20_degs\n",
    "mean_random_only = len(random_genes) - mean_overlap_random_top_characterizing - mean_overlap_random_top_20_degs\n",
    "\n",
    "print(f\"Mean characterizing only: {mean_characterizing_only}\")\n",
    "print(f\"Mean top 20 DEGs only: {mean_top_20_degs_only}\")\n",
    "print(f\"Mean random only: {mean_random_only}\")\n",
    "print(f\"Mean overlap random & characterizing: {mean_overlap_random_top_characterizing}\")\n",
    "print(f\"Mean overlap random & top 20 DEGs: {mean_overlap_random_top_20_degs}\")\n",
    "print(f\"Overlap characterizing & top 20 DEGs: {len(overlap_top_characterizing_top_20_degs)}\")\n",
    "print(f\"Mean overlap all: {mean_overlap_random_both}\")\n",
    "\n",
    "# Create the Venn diagram\n",
    "venn=venn3(subsets=(mean_characterizing_only,\n",
    "                    mean_top_20_degs_only,\n",
    "                    len(overlap_top_characterizing_top_20_degs),\n",
    "                    round(mean_random_only, 2),\n",
    "                    mean_overlap_random_top_characterizing,\n",
    "                    mean_overlap_random_top_20_degs,\n",
    "                    mean_overlap_random_both),\n",
    "           set_labels=('characterizing GP genes',\n",
    "                       'Top 20 DEGs in non-spatial clustering',\n",
    "                       'Randomly selected baseline genes'))\n",
    "\n",
    "font_properties = font_manager.FontProperties(size=12)\n",
    "for label in venn.set_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "for label in venn.subset_labels:\n",
    "    label.set_fontproperties(font_properties)\n",
    "\n",
    "plt.savefig(f'{figure_folder_path}/overlap_characterizing_20_degs.svg', format='svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb08e58-7866-4b2e-baf1-a4264c1a37b8",
   "metadata": {},
   "source": [
    "### 3.2 Save Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64e2a7ac-8d25-4d2d-aa75-09a0841dd720",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log normalize counts for cellxgene server\n",
    "sc.pp.normalize_total(model.adata, target_sum=1e4)\n",
    "sc.pp.log1p(model.adata)\n",
    "\n",
    "# Store gp summary in adata\n",
    "gp_summary = model.get_gp_summary()\n",
    "for col in gp_summary.columns:\n",
    "    gp_summary[col] = gp_summary[col].astype(str)\n",
    "model.adata.uns[\"nichecompass_gp_summary\"] = gp_summary\n",
    "\n",
    "model.adata.write(f\"{result_folder_path}/{dataset}_analysis.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e93ecbde-e779-4522-83ce-b14bee6169bf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
