{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "992dc6ee-ffc4-4d15-a364-486eaa5f6edc",
   "metadata": {},
   "source": [
    "# Spatial ATAC-RNA Seq Mouse Brain"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>)\n",
    "- **Date of Creation:** 22.01.2023\n",
    "- **Date of Last Modification:** 22.12.2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6bd6fb-6697-43bd-b3ec-68b061afe511",
   "metadata": {},
   "source": [
    "- In order to run this notebook, a trained model needs to be stored under ```../artifacts/{dataset}/models/{model_label}/{load_timestamp}```.\n",
    "- Load timestamp: ```13092023_174230```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5d3297-0290-47c0-846f-705e74cdb3de",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dbd4925-7505-4d01-bf91-95b54bb8b533",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../../utils\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import argparse\n",
    "import gc\n",
    "import os\n",
    "import random\n",
    "import shutil\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import anndata as ad\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import scipy.stats as stats\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "import torch\n",
    "from matplotlib import gridspec\n",
    "from matplotlib.pyplot import rc_context\n",
    "from pywaffle import Waffle\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                                aggregate_obsp_matrix_per_cell_type,\n",
    "                                create_cell_type_chord_plot_from_df,\n",
    "                                create_new_color_dict,\n",
    "                                compute_communication_gp_network,\n",
    "                                generate_enriched_gp_info_plots,\n",
    "                                visualize_communication_gp_network)\n",
    "\n",
    "from analysis_utils import (add_cell_type_latent_cluster_emphasis,\n",
    "                            add_sub_cell_type,\n",
    "                            compute_cell_type_latent_clusters,\n",
    "                            generate_gp_info_plots,\n",
    "                            plot_category_in_latent_and_physical_space,\n",
    "                            plot_cell_type_latent_clusters,\n",
    "                            plot_latent,\n",
    "                            plot_physical_latent_for_cell_types,\n",
    "                            sankey,\n",
    "                            store_top_gps_summary)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eeb8b324-ef18-4ab0-9ffb-8332d652458c",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9325589b-a7e4-4c31-97e7-8b7048248896",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ac7b6d-4518-4356-bfdd-e141bf25d504",
   "metadata": {},
   "source": [
    "#### 1.2.1 Generic Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## Model\n",
    "# AnnData keys\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "active_gp_names_key = \"nichecompass_active_gp_names\"\n",
    "latent_key = \"nichecompass_latent\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7be323-bd40-4fa0-a2d5-b8ddf5cc1c51",
   "metadata": {},
   "source": [
    "#### 1.2.2 Dataset-specific Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01210b2f-017e-43f6-b650-5cc732252f8c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "load_timestamp = \"13092023_174230\"\n",
    "model_label = \"reference\"\n",
    "cell_type_key = \"celltype\"\n",
    "sample_key = \"batch\"\n",
    "samples = [\"p22\"]\n",
    "latent_leiden_resolution = 0.3\n",
    "spot_size = 30\n",
    "    \n",
    "latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf229ca-ab66-4e52-a31e-8655878d290d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))\n",
    "# sc.set_figure_params(dpi=300)\n",
    "sns.set_style(\"whitegrid\", {'axes.grid' : False})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "924032d4-24a4-45cf-87e1-0eba197da0c2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Ignore future warnings and user warnings\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=UserWarning)\n",
    "warnings.simplefilter(action=\"ignore\", category=RuntimeWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "445898ac-b061-48c4-b611-904b6ef15787",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "plt.rcParams['font.size'] = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define paths\n",
    "figure_folder_path = f\"../../artifacts/{dataset}/figures/{model_label}/{load_timestamp}\"\n",
    "model_folder_path = f\"../../artifacts/{dataset}/models/{model_label}/{load_timestamp}\"\n",
    "result_folder_path = f\"../../artifacts/{dataset}/results/{model_label}/{load_timestamp}\"\n",
    "srt_data_folder_path = \"../../datasets/st_data\" # spatially resolved transcriptomics data\n",
    "srt_data_gold_folder_path = f\"{srt_data_folder_path}/gold\"\n",
    "\n",
    "# Create required directories\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(result_folder_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cea1a5c-d075-4e51-90e5-90684bee88a6",
   "metadata": {},
   "source": [
    "## 2. Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dad2ca6-7e34-48bc-89ef-4f994389a0f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "model = NicheCompass.load(dir_path=model_folder_path,\n",
    "                          adata=None,\n",
    "                          adata_file_name=f\"{dataset}_{model_label}.h5ad\",\n",
    "                          adata_atac=None,\n",
    "                          adata_atac_file_name=f\"{dataset}_{model_label}_atac.h5ad\",\n",
    "                          gp_names_key=gp_names_key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d99a385-5769-4c6b-a7dc-62f40028620a",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.adata.uns[gp_names_key] = np.array([gp for gp in model.adata.uns[gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp for gp in model.adata.uns[active_gp_names_key] if not \"Add-on \" in gp])\n",
    "model.adata.uns[gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[gp_names_key]])\n",
    "model.adata.uns[active_gp_names_key] = np.array([gp.replace(\" \", \"_\") for gp in model.adata.uns[active_gp_names_key]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dac39074-398d-4fc7-a0f7-86416a8e4e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataset-specific metadata\n",
    "cell_type_df = pd.read_csv(f\"{srt_data_gold_folder_path}/spatial_atac_rna_seq_mouse_brain_batch2_cell_type_annotations.csv\", index_col=0)\n",
    "cell_type_df.rename({\"predicted.celltype\": \"celltype\"}, axis=1, inplace=True)\n",
    "cell_type_df.drop(\"ATAC_clusters\", axis=1, inplace=True)\n",
    "model.adata.obs = model.adata.obs.merge(cell_type_df, left_index=True, right_index=True, how=\"left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ee19f45-486f-4b55-9a9e-ac55e8ef1c0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check number of active gene programs\n",
    "active_gps = model.get_active_gps()\n",
    "print(f\"Number of total gene programs: {len(model.adata.uns[gp_names_key])}.\")\n",
    "print(f\"Number of active gene programs: {len(active_gps)}.\")\n",
    "\n",
    "gp_summary_df = model.get_gp_summary()\n",
    "gp_summary_df[gp_summary_df[\"gp_active\"] == True][20:]\n",
    "\n",
    "samples = model.adata.obs[sample_key].unique().tolist()\n",
    "model.add_active_gp_scores_to_obs()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f662f73-455e-46c2-bc4b-ec25fb2f1d2b",
   "metadata": {},
   "source": [
    "## 3. Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec2e6f68-8f30-4ec8-85fc-742e5fc185c5",
   "metadata": {},
   "source": [
    "### 3.1 Create Figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acea0b4e-d809-49cf-8c29-f6a7cb214f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 24a: Sparse marker gene expression ###\n",
    "adata_all_genes = sc.read_h5ad(f\"{srt_data_gold_folder_path}/spatial_atac_rna_seq_mouse_brain_batch2.h5ad\")\n",
    "fig, axs = plt.subplots(nrows=2,\n",
    "                        ncols=2,\n",
    "                        figsize=(15, 15))\n",
    "\n",
    "# Lmo4 TF & Target Gene Peaks\n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=adata_all_genes,\n",
    "    color=\"Drd3\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Drd3\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)\n",
    "\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/drd3_expression.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "del(adata_all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44abc3f7-9d93-408b-be6e-0b1e4e94866d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 24b: Niches ###\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=latent_leiden_resolution,\n",
    "             key_added=latent_cluster_key,\n",
    "             neighbors_key=latent_key)\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=0.05,\n",
    "             key_added=latent_cluster_key,\n",
    "             restrict_to=(latent_cluster_key, [\"7\"]),\n",
    "             neighbors_key=latent_key)\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=0.05,\n",
    "             key_added=latent_cluster_key,\n",
    "             restrict_to=(latent_cluster_key, [\"8\"]),\n",
    "             neighbors_key=latent_key)\n",
    "sc.tl.leiden(adata=model.adata,\n",
    "             resolution=0.05,\n",
    "             key_added=latent_cluster_key,\n",
    "             restrict_to=(latent_cluster_key, [\"6\"]),\n",
    "             neighbors_key=latent_key)\n",
    "\n",
    "niche_annotation_dict = {\n",
    "    \"0\":\"Sox5/Taf1+ L5-L6\",\n",
    "    \"1\":\"Foxa1/Id3+ L6a-L6b\",\n",
    "    \"2\":\"Meis2/Ebf1+ STR_ACB_core\",\n",
    "    \"3\":\"Hlf/Neurod2+ L1_L4_vasc_olg\",\n",
    "    \"4\":\"Addon16/Irf8+ STR_caudoputamen_lateral\",\n",
    "    \"5\":\"Nr1i3/Lmx1a+ L1-L4\",\n",
    "    \"6,0\":\"Pax2/L-Glutamine+ L5-L6\",\n",
    "    \"6,1\":\"Zbtb7a/Myf5+ L5\",\n",
    "    \"7,0\":\"Gfi1b/Ep300+ STR_caudoputamen_medial\",\n",
    "    \"7,1\":\"Neurog3/Nkx2-2+ Anterior_commisure\",\n",
    "    \"8,0\":\"Pitx3/Cadm1+ STR_ACB_shell\",\n",
    "    \"8,1\":\"Usf2/Lmo4+ islm_Major_island_of_Calleja\",\n",
    "    \"9\":\"Nkx6-2/Nfya+ Corpus_callosum\",\n",
    "    \"10\":\"Plxna2/Hoxb1+ L1-L3\",\n",
    "    \"11\":\"Smad3/Col1a2+ Meninges\",\n",
    "    \"12\":\"Klf10/Crx+ Lateral_septum_complex\",\n",
    "    \"13\":\"Lmo2/Nkx6-1+ Lateral_septal_nucleus\",\n",
    "    \"14\":\"L5-L6\",\n",
    "    \"15\":\"Dach1/Pitx3+ STR_medial\",\n",
    "    \"16\":\"Bcl11a/Brca1+ Lateral_ventricle\"}\n",
    "\n",
    "model.adata.obs[\"niche\"] = model.adata.obs[latent_cluster_key].map(niche_annotation_dict)\n",
    "\n",
    "latent_cluster_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    cat_key=latent_cluster_key)\n",
    "latent_cluster_colors[\"7,0\"] = \"#7d87b9\"\n",
    "latent_cluster_colors[\"4\"] = \"#4a6fe3\"\n",
    "latent_cluster_colors[\"2\"] = \"#023fa5\"\n",
    "latent_cluster_colors[\"1\"] = \"#DAB6C4\"\n",
    "latent_cluster_colors[\"15\"] = \"#9EB9F3\"\n",
    "latent_cluster_colors[\"7,1\"] = \"#9D88A2\"\n",
    "latent_cluster_colors[\"13\"] = \"#FF9CDA\"\n",
    "latent_cluster_colors[\"8,1\"] = \"#CE5A57\"\n",
    "niche_colors = {niche: latent_cluster_colors[cluster] for cluster, niche in niche_annotation_dict.items()}\n",
    "\n",
    "model.adata.obs[\"niche_abr\"] = model.adata.obs[\"niche\"].apply(lambda x: x.split(\"GP+ \")[1].replace(\" \", \"\\n\") if \"GP+ \" in x else x)\n",
    "\n",
    "niche_abreviated_colors = {}\n",
    "for key, value in niche_colors.items():\n",
    "    if \"GP+ \" in key:\n",
    "        niche_abreviated_colors[key.split(\"GP+ \")[1].replace(\" \", \"\\n\")] = value\n",
    "    else:\n",
    "        niche_abreviated_colors[key] = value\n",
    "        \n",
    "save_fig = True\n",
    "file_path = f\"{figure_folder_path}/res_{latent_leiden_resolution}_\" \\\n",
    "            \"niches_latent_physical_space.svg\"\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "    adata=model.adata,\n",
    "    plot_label=\"Niches\",\n",
    "    cat_key=\"niche\",\n",
    "    groups=None,\n",
    "    sample_key=sample_key,\n",
    "    samples=samples,\n",
    "    cat_colors=niche_colors,\n",
    "    size=(720000 / len(model.adata)),\n",
    "    spot_size=spot_size,\n",
    "    save_fig=save_fig,\n",
    "    file_path=file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dae2e20-70fd-4116-9d2f-1eab50e651a6",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sc' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m### Supplementary Fig. 24c: Tissue niche hierarchy ###\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43msc\u001b[49m\u001b[38;5;241m.\u001b[39mtl\u001b[38;5;241m.\u001b[39mdendrogram(adata\u001b[38;5;241m=\u001b[39mmodel\u001b[38;5;241m.\u001b[39madata,\n\u001b[1;32m      3\u001b[0m                  use_rep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnichecompass_latent\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      4\u001b[0m                  linkage_method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mward\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m      5\u001b[0m                  groupby\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mniche\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      7\u001b[0m fig, (ax) \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39msubplots(\u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m1\u001b[39m, figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m0.6\u001b[39m, \u001b[38;5;241m2\u001b[39m))\n\u001b[1;32m      8\u001b[0m sc\u001b[38;5;241m.\u001b[39mpl\u001b[38;5;241m.\u001b[39mdendrogram(\n\u001b[1;32m      9\u001b[0m     adata\u001b[38;5;241m=\u001b[39mmodel\u001b[38;5;241m.\u001b[39madata,\n\u001b[1;32m     10\u001b[0m     groupby\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mniche\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     11\u001b[0m     orientation\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mright\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m     12\u001b[0m     ax\u001b[38;5;241m=\u001b[39max,\n\u001b[1;32m     13\u001b[0m     save\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_nichecompass_latent.svg\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sc' is not defined"
     ]
    }
   ],
   "source": [
    "### Supplementary Fig. 24c: Tissue niche hierarchy ###\n",
    "sc.tl.dendrogram(adata=model.adata,\n",
    "                 use_rep=\"nichecompass_latent\",\n",
    "                 linkage_method=\"ward\",\n",
    "                 groupby=\"niche\")\n",
    "\n",
    "fig, (ax) = plt.subplots(1, 1, figsize=(0.6, 2))\n",
    "sc.pl.dendrogram(\n",
    "    adata=model.adata,\n",
    "    groupby=\"niche\",\n",
    "    orientation=\"left\",\n",
    "    ax=ax,\n",
    "    save=\"_nichecompass_latent.svg\")\n",
    "plt.show()\n",
    "if os.path.exists(f\"{figure_folder_path}/dendrogram_nichecompass_latent.svg\"):\n",
    "    os.remove(f\"{figure_folder_path}/dendrogram_nichecompass_latent.svg\")\n",
    "shutil.move(f\"./figures/dendrogram_nichecompass_latent.svg\", figure_folder_path)\n",
    "shutil.rmtree(\"./figures/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eebbc5eb-86fd-49d9-913b-a5e98f798059",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 24d: Niche-enriched gene programs ###\n",
    "gp_resign = True # True\n",
    "\n",
    "gps = [ \n",
    "    \"Foxa1_TF_target_genes_GP\", \"Id3_TF_target_genes_GP\",\n",
    "    \"Plxna2_ligand_receptor_target_gene_GP\", \"Hoxb1_TF_target_genes_GP\",\n",
    "    \"Sox5_TF_target_genes_GP\", \"Taf1_TF_target_genes_GP\", # Xanthine_metabolite_enzyme_sensor_GP    \n",
    "    \"Zbtb7a_TF_target_genes_GP\", \"Myf5_TF_target_genes_GP\",\n",
    "    \"Pax2_TF_target_genes_GP\", \"L-Glutamine_metabolite_enzyme_sensor_GP\",\n",
    "    \"Nr1i3_TF_target_genes_GP\", \"Lmx1a_TF_target_genes_GP\",\n",
    "    \"Hlf_TF_target_genes_GP\", \"Neurod2_TF_target_genes_GP\",\n",
    "    \"Add-on_16_GP\", \"Irf8_TF_target_genes_GP\",\n",
    "    \"Meis2_TF_target_genes_GP\", \"Ebf1_TF_target_genes_GP\",\n",
    "    \"Gfi1b_TF_target_genes_GP\", \"Ep300_TF_target_genes_GP\",\n",
    "    \"Nkx6-2_TF_target_genes_GP\", \"Nfya_TF_target_genes_GP\",\n",
    "    \"Neurog3_TF_target_genes_GP\", \"Nkx2-2_TF_target_genes_GP\",\n",
    "    \"Lmo2_TF_target_genes_GP\", \"Nkx6-1_TF_target_genes_GP\",\n",
    "    \"Klf10_TF_target_genes_GP\", \"Crx_TF_target_genes_GP\",\n",
    "    \"Smad3_TF_target_genes_GP\", \"Col1a2_ligand_receptor_target_gene_GP\",\n",
    "    \"Bcl11a_TF_target_genes_GP\", \"Brca1_TF_target_genes_GP\", # Draxin_ligand_receptor_target_gene_GP\n",
    "    \"Usf2_TF_target_genes_GP\", \"Lmo4_TF_target_genes_GP\",\n",
    "    \"Pitx3_TF_target_genes_GP\", \"Cadm1_ligand_receptor_target_gene_GP\",\n",
    "    \"Dach1_TF_target_genes_GP\", \"Pitx3_TF_target_genes_GP\",\n",
    "]\n",
    "\n",
    "resign_gps = [\n",
    "    \"Taf1_TF_target_genes_GP\",\n",
    "    \"Zbtb7a_TF_target_genes_GP\",\n",
    "    \"Meis2_TF_target_genes_GP\",\n",
    "    \"Ebf1_TF_target_genes_GP\",\n",
    "    \"Neurod2_TF_target_genes_GP\",\n",
    "    \"Nr1i3_TF_target_genes_GP\",\n",
    "    \"Hlf_TF_target_genes_GP\",\n",
    "    \"L-Glutamine_metabolite_enzyme_sensor_GP\",\n",
    "    \"Gfi1b_TF_target_genes_GP\",\n",
    "    \"Ep300_TF_target_genes_GP\",\n",
    "    \"Pitx3_TF_target_genes_GP\",\n",
    "    \"Cadm1_ligand_receptor_target_gene_GP\",\n",
    "    \"Usf2_TF_target_genes_GP\",\n",
    "    \"Lmo4_TF_target_genes_GP\",\n",
    "    \"Rps19_ligand_receptor_target_gene_GP\",\n",
    "    \"Nfya_TF_target_genes_GP\",\n",
    "    \"Nkx6-2_TF_target_genes_GP\",\n",
    "    \"Lmo2_TF_target_genes_GP\",\n",
    "    \"Pitx3_TF_target_genes_GP\",\n",
    "    \"Brca1_TF_target_genes_GP\",\n",
    "    \"Bcl11a_TF_target_genes_GP\",\n",
    "    \"Nkx6-1_TF_target_genes_GP\",\n",
    "    \"Klf10_TF_target_genes_GP\",\n",
    "    \"Dach1_TF_target_genes_GP\",\n",
    "    \"Pitx3_TF_target_genes_GP\"]\n",
    "\n",
    "if gp_resign:\n",
    "    for gp in resign_gps:\n",
    "        model.adata.obs[gp] = -1 * model.adata.obs[gp]\n",
    "\n",
    "df = model.adata.obs[[\"niche\"] + gps].groupby(\"niche\").mean()\n",
    "df = df.reindex(model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"][::-1])\n",
    "\n",
    "scaler = MinMaxScaler()\n",
    "normalized_columns = scaler.fit_transform(df)\n",
    "normalized_df = pd.DataFrame(normalized_columns, columns=df.columns)\n",
    "normalized_df.index = df.index\n",
    "normalized_df.columns = [col.split(\"_lig\")[0].split(\"_met\")[0].split(\"_TF\")[0].replace(\"Add-on_16_GP\", \"Addon16\").replace(\"_acid\", \" acid\") + \" GP\" for col in normalized_df.columns]\n",
    "\n",
    "plt.figure(figsize=(16, 6))  # Set the figure size\n",
    "ax = sns.heatmap(normalized_df,\n",
    "            cmap='viridis',\n",
    "            annot=False,\n",
    "            linewidths=0)\n",
    "plt.xticks(rotation=45,\n",
    "           ha=\"right\"\n",
    "          )\n",
    "plt.savefig(f\"{figure_folder_path}/enriched_gps_heatmap.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "plt.figure(figsize=(6, 3))  # Set the figure size\n",
    "ax = sns.heatmap(normalized_df,\n",
    "            cmap='viridis',\n",
    "            annot=False,\n",
    "            linewidths=0,\n",
    "            cbar_kws={\"orientation\": \"horizontal\", \"location\": \"bottom\", \"shrink\": 0.4})\n",
    "plt.xticks([])\n",
    "plt.savefig(f\"{figure_folder_path}/heatmap_colorbar.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a92cd6e-5eca-430d-8688-dd248adc5e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 24e: Niche composition ###\n",
    "cell_type_colors = create_new_color_dict(\n",
    "    adata=model.adata,\n",
    "    color_palette=(\"cell_type_30\"),\n",
    "    cat_key=cell_type_key)\n",
    "cell_type_colors[\"CA3GL\"] = \"#1f77b4\"\n",
    "cell_type_colors[\"D1MSN\"] = \"#ff7f0e\"\n",
    "cell_type_colors[\"LSXGA\"] = \"#279e68\"\n",
    "cell_type_colors[\"MXD\"] = \"#d62728\"\n",
    "cell_type_colors[\"STRGA\"] = \"#aa40fc\"\n",
    "\n",
    "tmp = pd.crosstab(model.adata.obs[\"niche\"], model.adata.obs[cell_type_key], normalize='index')\n",
    "tmp = tmp.reindex(model.adata.uns[\"dendrogram_niche\"][\"categories_ordered\"][::])\n",
    "ax = tmp.plot.barh(color=cell_type_colors, stacked=True, figsize=(4, 6)).legend(loc='center left', bbox_to_anchor=(1.0, 0.5))\n",
    "plt.xlabel(\"Cell Type Proportions\")\n",
    "plt.savefig(f\"{figure_folder_path}/niche_cell_type_proportions.svg\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c23db600-9faa-442c-973d-f59b7bbabff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 24f: Niche composition ###\n",
    "fig, ax = plt.subplots(figsize=(17, 21))\n",
    "sankey(\n",
    "    x=model.adata.obs[cell_type_key],\n",
    "    y=model.adata.obs[\"niche_abr\"],\n",
    "    title=\"Cell Type to Niche Mapping\",\n",
    "    title_left=\"Cell Type\",\n",
    "    title_right=\"Niche\",\n",
    "    ax=ax,\n",
    "    fontsize=\"16\",  # \"xx-small\",\n",
    "    #left_order=model.adata.obs[cell_type_key].unique().tolist(),\n",
    "    colors=cell_type_colors,\n",
    "    alpha=0.5)\n",
    "plt.tight_layout()\n",
    "plt.savefig(f\"{figure_folder_path}/sankey_cell_type_to_niche.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8165efe9-97c8-402b-85c8-f7e8e0824e9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 25a: Major Island of Calleja niche: enriched prior GPs ###\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Usf2_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Usf2 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Prkar2b\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Prkar2b\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Apc\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Apc\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Pou2f1\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"Pou2f1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr18:34247310-34248119\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr18:34247310-34248119\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr1:165934747-165935645\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr1:165934747-165935645\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)    \n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/usf2_gp_genes.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cebpz_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cebpz GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cebpz\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cebpz\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Col11a1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Col11a1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Apc\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Apc\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr3:114030022-114030924\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr3:114030022-114030924\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr18:34247310-34248119\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr18:34247310-34248119\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr11:53300121-53301011\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr11:53300121-53301011\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/cebpz_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                            figsize=(25, 25))\n",
    "model.adata.obs[\"Rela_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Rela_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Rela_TF_target_genes_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Rela GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Bclaf1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Bclaf1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Ing4\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ing4\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Kmt2a\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Kmt2a\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr10:20311949-20312805\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr10:20311949-20312805\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/rela_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Rarb_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Rarb_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Rarb_TF_target_genes_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Rarb GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Isl1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Isl1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Hap1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Hap1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr13:116304790-116305726\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr13:116304790-116305726\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr13:116309366-116310250\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr13:116309366-116310250\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/rarb_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Rreb1_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Rreb1_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Rreb1_TF_target_genes_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Rreb1 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Rreb1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Rreb1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/rreb1_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Gata1_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Gata1_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gata1_TF_target_genes_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gata1 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Tac1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Tac1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Runx1t1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Runx1t1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Furin\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Furin\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr12:104689761-104690616\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr12:104689761-104690616\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/Gata1_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Lmo4_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Lmo4\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Lmo4\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Lmo4\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr11:70029079-70029942\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr11:70029079-70029942\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr11:70021137-70021886\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"chr11:70021137-70021886\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/lmo4_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5954e7fa-e315-46a2-8ee0-66d3e630065c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 26b: Major Island Of Calleja: enriched de novo GP ###\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Add-on_89_GP_sign_corrected\"] = -1 * model.adata.obs[\"Add-on_89_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Add-on_89_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Add-on_89_GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr13:9583665-9584550\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr13:9583665-9584550\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr9:104796453-104797222\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr9:104796453-104797222\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr1:170358844-170359717\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr1:170358844-170359717\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr5:65770160-65771077\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr5:65770160-65771077\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr2:156658586-156659426\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr2:156658586-156659426\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr13:102943643-102944509\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr13:102943643-102944509\",\n",
    "    use_raw=False,\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/add-on89_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbfb917b-8a74-41de-89b7-deaac5077b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 26a: Corpus Callosum niche: enriched prior GPs ###\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                            figsize=(25, 25))\n",
    "#model.adata.obs[\"Nfya_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Nfya_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Nfya_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nfya GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cldn11\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cldn11\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Phgdh\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Phgdh\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Hmgcs1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Hmgcs1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Fdps\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Fdps\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Ldlr\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ldlr\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/nfya_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Klf6_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Klf6_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Klf6_TF_target_genes_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Klf6 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cers2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cers2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gstp2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gstp2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gstp1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gstp1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Tuba1a\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Tuba1a\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/klf6_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Tp73_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Tp73 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Apod\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Apod\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cryab\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cryab\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gstp2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gstp2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gstp1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gstp1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Ybx1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ybx1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/tp73_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Rps19_ligand_receptor_target_gene_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Rps19 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Rps19\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Rps19\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cd82\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cd82\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Scd2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Scd2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Ndrg1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ndrg1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Enpp2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Enpp2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/rps19_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Nkx6-2_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nkx6-2 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Plp1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Plp1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Mbp\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Mbp\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/nkx6-2_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Nkx2-2_TF_target_genes_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nkx2-2 GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Ermn\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Ermn\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Sirt2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Sirt2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Arx\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Arx\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/nkx2-2_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Olig2_TF_target_genes_GP_sign_corrected\"] = -1 * model.adata.obs[\"Olig2_TF_target_genes_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Olig2_TF_target_genes_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Olig2  GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Olig1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Olig1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Sox10\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Sox10\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Olig2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Olig2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/olig2_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Cd82_ligand_receptor_target_gene_GP_sign_corrected\"] = -1 * model.adata.obs[\"Cd82_ligand_receptor_target_gene_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cd82_ligand_receptor_target_gene_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cd82  GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Cd82\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Cd82\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gsn\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gsn\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gfra1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gfra1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Tcf7l2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Tcf7l2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Zfp36l1\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Zfp36l1\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/cd82_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Nfasc_ligand_receptor_target_gene_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nfasc  GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Nfasc\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Nfasc\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Mgll\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Mgll\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Enpp2\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Enpp2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/nfasc_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")\n",
    "\n",
    "fig, axs = plt.subplots(nrows=6,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "model.adata.obs[\"Glycine_metabolite_enzyme_sensor_GP_sign_corrected\"] = -1 * model.adata.obs[\"Glycine_metabolite_enzyme_sensor_GP\"]\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Glycine_metabolite_enzyme_sensor_GP_sign_corrected\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Glycine  GP\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Lap3\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Lap3\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Gatm\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Gatm\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    use_raw=False,\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/glycine_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62cf487c-e38e-4c59-bc78-c47a1f4e51db",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 26b: Corpus Callosum niche: enriched de novo gps ###\n",
    "fig, axs = plt.subplots(nrows=5,\n",
    "                        ncols=2,\n",
    "                        figsize=(25, 25))\n",
    "color_map = \"RdGy_r\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Add-on_89_GP\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Add-on 89\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 0],\n",
    "    show=False)    \n",
    "color_map = \"RdPu\"\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Mobp\",\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Mobp\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[0, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Tspan2\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"Tspan2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Pllp\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"Pllp\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[1, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Trf\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"Trf\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata,\n",
    "    color=\"Pdlim2\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"Pdlim2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[2, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr1:24612403-24613300\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=f\"chr1:24612403-24613300\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr1:172349675-172350564\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"chr1:172349675-172350564\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[3, 1],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr4:48629122-48629953\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"chr4:48629122-48629953\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[4, 0],\n",
    "    show=False)\n",
    "sc.pl.spatial(\n",
    "    adata=model.adata_atac,\n",
    "    color=\"chr14:54935868-54936644\",\n",
    "    use_raw=False,\n",
    "    color_map=color_map,\n",
    "    spot_size=spot_size,\n",
    "    title=\"chr14:54935868-54936644\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    ax=axs[4, 1],\n",
    "    show=False)\n",
    "plt.show()\n",
    "fig.savefig(f\"{figure_folder_path}/addon89_gp_genes_peaks.svg\",\n",
    "            bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7eb08e58-7866-4b2e-baf1-a4264c1a37b8",
   "metadata": {},
   "source": [
    "### 3.2 Save Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f117beb1-7834-4cd1-b9eb-cf2a14f9ca5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log normalize counts for cellxgene server\n",
    "sc.pp.normalize_total(model.adata, target_sum=1e4)\n",
    "sc.pp.log1p(model.adata)\n",
    "    \n",
    "# Store gp summary in adata\n",
    "gp_summary = model.get_gp_summary()\n",
    "for col in gp_summary.columns:\n",
    "    gp_summary[col] = gp_summary[col].astype(str)\n",
    "model.adata.uns[\"nichecompass_gp_summary\"] = gp_summary\n",
    "\n",
    "model.adata.write(f\"{result_folder_path}/{dataset}_analysis.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe28ef4-cf38-4d28-a9ef-26118caa2892",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "toc-autonumbering": false,
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
