{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bdc8086b-2b80-43b4-b88a-aa0c87c820a6",
   "metadata": {},
   "source": [
    "# Slurm Job Submission"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be88d0ca-4c09-44e3-83fc-fb6bcf4563fd",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>)\n",
    "- **Date of Creation:** 20.03.2023\n",
    "- **Date of Last Modification:** 09.10.2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aeeeb1fe-eaa1-4cdd-aa21-ed78e0b7be81",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e58b0a3-c3fe-43d5-a420-14111aa545f6",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48d297b0-e0dc-42b9-a349-e6ac2edba0e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aab82f4c-edd8-47dd-be56-559cef739439",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0491ee0e-3587-4243-856f-e7f71e6dced6",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda_env_name = \"nichecompass-reproducibility\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "004ab52a-9cc3-4e40-b0e6-f9fed7f4a549",
   "metadata": {},
   "source": [
    "### 1.3 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b2d6cf0-b360-428f-9548-c7942378de68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def submit_python_script(\n",
    "        job_name_prefix,\n",
    "        job_id,\n",
    "        job_folder_path,\n",
    "        conda_env_name,\n",
    "        script_folder_path,\n",
    "        script_name,\n",
    "        script_args,\n",
    "        t=\"48:00:00\",\n",
    "        p=\"gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"gpu_normal\",\n",
    "        mem=\"156G\",\n",
    "        nice=10000):\n",
    "    job_name = f\"{job_name_prefix}_{job_id}\"\n",
    "    # account for fact that submit node has different home path than compute node\n",
    "    job_file_path = f\"{job_folder_path.replace('/aih', '')}/job_{job_name}.cmd\"\n",
    "    out_file_path = f\"{job_folder_path}/logs/out_{job_name}.txt\"\n",
    "    err_file_path = f\"{job_folder_path}/logs/err_{job_name}.txt\"\n",
    "    \n",
    "    os.makedirs(job_folder_path + \"/logs\", exist_ok=True)\n",
    "    \n",
    "    with open(job_file_path, \"w\") as handle:\n",
    "        handle.writelines(\"#!/bin/bash\\n\")\n",
    "        handle.writelines(f\"#SBATCH -J {job_name}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -o {out_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -e {err_file_path}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -t {t}\\n\")\n",
    "        handle.writelines(f\"#SBATCH -p {p}\\n\")\n",
    "        handle.writelines(\"#SBATCH -c 6\\n\")\n",
    "        #handle.writelines(\"#SBATCH --exclude=supergpu02,supergpu03\\n\")\n",
    "        if \"gpu\" in p:\n",
    "            handle.writelines(f\"#SBATCH --gres={gres}\\n\")\n",
    "        handle.writelines(f\"#SBATCH --qos={qos}\\n\")       \n",
    "        handle.writelines(f\"#SBATCH --mem={mem}\\n\")\n",
    "        handle.writelines(f\"#SBATCH --nice={nice}\\n\")\n",
    "        handle.writelines(\"source $HOME/.bashrc\\n\")\n",
    "        handle.writelines(f\"conda activate {conda_env_name}\\n\")\n",
    "        handle.writelines(\"cd /\\n\")\n",
    "        handle.writelines(f\"cd {script_folder_path}\\n\")\n",
    "        handle.writelines(f\"python ../{script_name}\")\n",
    "        handle.writelines(f\"{script_args}\")\n",
    "        handle.writelines(\"\\n\")\n",
    "        \n",
    "    os.system(f\"sbatch {job_file_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4624e92-33eb-457c-8767-9006e44d850e",
   "metadata": {},
   "source": [
    "## 2. NicheCompass Ablation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc99f810-89a4-4070-ac09-d3f6da04c95d",
   "metadata": {},
   "source": [
    "### 2.1 Loss Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1454dcbe-7988-4118-a212-e2c8ca65c429",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"loss\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "n_neighbors = 6\n",
    "active_gp_thresh_ratio = 0.01\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_id = 1\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([500000., 0., 500000., 50000., 500000.], [300., 300., 0., 300., 30]):\n",
    "    for seed in range(8):\n",
    "        script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                      f\" --n_epochs {n_epochs}\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                      f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                      f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors 4\" \\\n",
    "                      f\" --seed {seed}\" \\\n",
    "                      f\" --use_new_gp_mask\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87fc9414-f4e8-4be3-b30b-5bd438367abb",
   "metadata": {},
   "source": [
    "### 2.2 Loss Ablation Extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "383ae45c-b8db-40c5-86c3-7fde50c0b317",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"loss_extended\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "n_neighbors = 6\n",
    "active_gp_thresh_ratio = 0.01\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_edge_recon = 500000.\n",
    "lambda_gene_expr_recon = 300.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_id = 1\n",
    "for lambda_l1_masked, lambda_l1_addon in zip([0., 0., 0., 0., 3., 30., 300.], [0., 3., 30., 300., 3., 30., 300.]):\n",
    "    for seed in range(8):\n",
    "        script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                      f\" --n_epochs {n_epochs}\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                      f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                      f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors 4\" \\\n",
    "                      f\" --seed {seed}\" \\\n",
    "                      f\" --use_new_gp_mask\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d60206d8-ce51-4096-9b20-f7b5a52697b6",
   "metadata": {},
   "source": [
    "### 2.3 Encoder Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1db10ee7-239f-45d8-9f3a-05c8b335840c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"encoder\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "n_neighbors = 6\n",
    "active_gp_thresh_ratio = 0.01\n",
    "node_label_method = \"one-hop-norm\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_edge_recon = 500000.\n",
    "lambda_gene_expr_recon = 300.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_id = 1\n",
    "for conv_layer_encoder in [\"gatv2conv\", \"gcnconv\"]:\n",
    "    for seed in range(8):\n",
    "        script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                      f\" --n_epochs {n_epochs}\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                      f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                      f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors 4\" \\\n",
    "                      f\" --seed {seed}\" \\\n",
    "                      f\" --use_new_gp_mask\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e889786-ea04-483c-a64e-471c9d0eb658",
   "metadata": {},
   "source": [
    "### 2.4 Neighbor Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7677322-620c-4853-9511-c1b66abaece2",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"neighbor\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "active_gp_thresh_ratio = 0.01\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_edge_recon = 500000.\n",
    "lambda_gene_expr_recon = 300.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_id = 1\n",
    "for n_neighbors in [4, 8, 12, 16, 20]:\n",
    "    for seed in range(8):\n",
    "        script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                      f\" --n_epochs {n_epochs}\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                      f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                      f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors 4\" \\\n",
    "                      f\" --seed {seed}\" \\\n",
    "                      f\" --use_new_gp_mask\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a593ca2-49b2-4133-bad9-6a4f66719436",
   "metadata": {},
   "source": [
    "### 2.5 De Novo GP Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a0cb745-1533-422a-b3da-dcf10ece7466",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"denovogp\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "n_neighbors = 6\n",
    "active_gp_thresh_ratio = 0.01\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_edge_recon = 500000.\n",
    "lambda_gene_expr_recon = 300.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_id = 1\n",
    "for n_addon_gp in [0, 10, 30, 100, 500]:\n",
    "    for seed in range(8):\n",
    "        script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                      f\" --n_epochs {n_epochs}\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                      f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                      f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors 4\" \\\n",
    "                      f\" --seed {seed}\" \\\n",
    "                      f\" --use_new_gp_mask\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a8956f3-058b-4e3c-b253-b1b61aa8d41d",
   "metadata": {},
   "source": [
    "### 2.6 GP Selection Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71a07623-3d5c-4595-b2a2-b9018ce1e4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"gpselection\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "n_neighbors = 6\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_edge_recon = 500000.\n",
    "lambda_gene_expr_recon = 300.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_id = 1\n",
    "for active_gp_thresh_ratio in [0., 0.01, 0.03, 0.1, 0.3, 0.5, 1]:\n",
    "    for seed in range(8):\n",
    "        script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "        script_args = f\" --dataset {dataset}\" \\\n",
    "                      f\" --reference_batches {reference_batches}\" \\\n",
    "                      f\" --n_neighbors {n_neighbors}\" \\\n",
    "                      \" --no-filter_genes\" \\\n",
    "                      \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                      \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                      \" --include_mebocost_gps\" \\\n",
    "                      f\" --species {species}\" \\\n",
    "                      \" --gp_filter_mode subset\" \\\n",
    "                      \" --combine_overlap_gps\" \\\n",
    "                      \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                      \" --overlap_thresh_genes 0.9\" \\\n",
    "                      \" --counts_key counts\" \\\n",
    "                      \" --spatial_key spatial\" \\\n",
    "                      \" --adj_key spatial_connectivities\" \\\n",
    "                      \" --mapping_entity_key mapping_entity\" \\\n",
    "                      \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                      \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                      \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                      f\" --model_label {ablation_task}_{task}\" \\\n",
    "                      \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                      \" --latent_key nichecompass_latent\" \\\n",
    "                      f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                      f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                      \" --gene_expr_recon_dist nb\" \\\n",
    "                      \" --log_variational\" \\\n",
    "                      f\" --node_label_method {node_label_method}\" \\\n",
    "                      \" --n_layers_encoder 1\" \\\n",
    "                      \" --n_hidden_encoder None\" \\\n",
    "                      f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                      f\" --n_epochs {n_epochs}\" \\\n",
    "                      \" --n_epochs_all_gps 25\" \\\n",
    "                      \" --lr 0.001\" \\\n",
    "                      f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                      f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                      \" --lambda_group_lasso 0.\" \\\n",
    "                      f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                      f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                      f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                      f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                      f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                      \" --node_batch_size None\" \\\n",
    "                      f\" --n_sampled_neighbors 4\" \\\n",
    "                      f\" --seed {seed}\" \\\n",
    "                      f\" --use_new_gp_mask\" \\\n",
    "                      f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69e3c2dd-114d-43a5-bb6a-957a05ffee12",
   "metadata": {},
   "source": [
    "### 2.7 No Prior GP Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec9abd72-e945-4f86-9837-2522639f96d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"ablation\"\n",
    "ablation_task = \"nopriorgp\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "reference_batches = \"batch1\"\n",
    "species = \"mouse\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{ablation_task}_{task}\"\n",
    "job_folder_path = f\"./{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk_new/workspace/projects/nichecompass-reproducibility/analysis/{task}\"\n",
    "\n",
    "n_neighbors = 6\n",
    "active_gp_thresh_ratio = 0.01\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_epochs = 400\n",
    "\n",
    "lambda_edge_recon = 500000.\n",
    "lambda_gene_expr_recon = 300.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "l1_targets_categories = \"target_gene\"\n",
    "l1_sources_categories = None\n",
    "edge_batch_size = 256\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_id = 1\n",
    "for add_fc_gps_instead_of_gp_dict_gps in [True, False]:\n",
    "    for seed in range(8):\n",
    "        if add_fc_gps_instead_of_gp_dict_gps:\n",
    "            script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --add_fc_gps_instead_of_gp_dict_gps\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                          f\" --n_epochs {n_epochs}\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                          f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                          f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors 4\" \\\n",
    "                          f\" --seed {seed}\" \\\n",
    "                          f\" --use_new_gp_mask\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "        else:\n",
    "            script_name = \"./data_analysis/train_nichecompass_reference_model.py\"\n",
    "            script_args = f\" --dataset {dataset}\" \\\n",
    "                          f\" --reference_batches {reference_batches}\" \\\n",
    "                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                          \" --no-filter_genes\" \\\n",
    "                          \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                          \" --include_mebocost_gps\" \\\n",
    "                          f\" --species {species}\" \\\n",
    "                          \" --gp_filter_mode subset\" \\\n",
    "                          \" --combine_overlap_gps\" \\\n",
    "                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                          \" --counts_key counts\" \\\n",
    "                          \" --spatial_key spatial\" \\\n",
    "                          \" --adj_key spatial_connectivities\" \\\n",
    "                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                          f\" --model_label {ablation_task}_{task}\" \\\n",
    "                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                          \" --latent_key nichecompass_latent\" \\\n",
    "                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                          \" --gene_expr_recon_dist nb\" \\\n",
    "                          \" --log_variational\" \\\n",
    "                          f\" --node_label_method {node_label_method}\" \\\n",
    "                          \" --n_layers_encoder 1\" \\\n",
    "                          \" --n_hidden_encoder None\" \\\n",
    "                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                          f\" --n_epochs {n_epochs}\" \\\n",
    "                          \" --n_epochs_all_gps 25\" \\\n",
    "                          \" --lr 0.001\" \\\n",
    "                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                          \" --lambda_group_lasso 0.\" \\\n",
    "                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                          f\" --l1_targets_categories {l1_targets_categories}\" \\\n",
    "                          f\" --l1_sources_categories {l1_sources_categories}\" \\\n",
    "                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                          \" --node_batch_size None\" \\\n",
    "                          f\" --n_sampled_neighbors 4\" \\\n",
    "                          f\" --seed {seed}\" \\\n",
    "                          f\" --use_new_gp_mask\" \\\n",
    "                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "        submit_python_script(\n",
    "                job_name_prefix=job_name_prefix,\n",
    "                job_id=job_id,\n",
    "                job_folder_path=job_folder_path,\n",
    "                conda_env_name=conda_env_name,\n",
    "                script_folder_path=script_folder_path,\n",
    "                script_name=script_name,\n",
    "                script_args=script_args,\n",
    "                nice=10000)\n",
    "\n",
    "        job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a80b724d-8200-44b4-8dba-02269e4358e0",
   "metadata": {},
   "source": [
    "## 3. Single Sample Method Benchmarking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb67595-8ede-4823-a1fe-a313dee704f5",
   "metadata": {},
   "source": [
    "### 3.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52c28e8-6fc8-492f-9c31-d3ab9b815980",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fe70711-9d03-47d5-892a-5cdf77714389",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c6ac715-da37-447e-a72d-d67c838bd477",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24ecebdb-2b38-4021-8116-b21061bef3b1",
   "metadata": {},
   "source": [
    "#### 3.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c3995e5-3c35-427e-838c-4831370d44c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c2d9c53-f844-48ec-b541-37071ac86e99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048 2048 2048\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19574d56-9c79-4874-9b4d-ee39248f0028",
   "metadata": {},
   "source": [
    "### 3.2 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2ab9803-1c84-42ac-a8b1-4ff9356a3493",
   "metadata": {},
   "source": [
    "This is not used for the manuscript."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe560799-9d98-4740-af7a-b972c7596d18",
   "metadata": {},
   "source": [
    "#### 3.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85bd3ffa-40d7-4414-b9fc-0a9e577f6a4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_embryo2\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory\n",
    "n_svg = 5000\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_gcnconv_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               \" --n_hvg 0\" \\\n",
    "               f\" --n_svg {n_svg}\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label gcnconv_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b1548b5-4dca-4762-b2b1-3b70211ed57f",
   "metadata": {},
   "source": [
    "#### 3.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5242ada-198f-46a6-b545-048ab451c42e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct_embryo2\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory\n",
    "    n_svg = 3000\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_one-hop-norm_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   \" --n_hvg 0\" \\\n",
    "                   f\" --n_svg {n_svg}\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label one-hop-norm_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 0\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   \" --conv_layer_encoder gcnconv\" \\\n",
    "                   \" --n_epochs 100\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4111eee6-88e7-4180-bc5b-a05cdde41061",
   "metadata": {},
   "source": [
    "### 3.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcce8ca8-124d-4814-a224-900bd546165e",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7842f6f0-664b-4405-ad30-add0af93ec7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_batch5\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"32768 32768 32768 32768 32768 32768 32768 32768\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff1805d-42a6-4a8d-9277-2f98bfe4ffec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_batch5\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"512 512 512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None\" \\\n",
    "               \" --seeds 7 6 5 4\" \\\n",
    "               \" --run_index 8 7 6 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a24ab2f3-1d57-4b83-aa65-e6948aca20c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc_batch5\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"512 512 512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None\" \\\n",
    "               \" --seeds 3 2 1 0\" \\\n",
    "               \" --run_index 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59d54d8a-bea3-40a9-acfa-4e78d2cffffe",
   "metadata": {},
   "source": [
    "#### 3.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7b792d-e60d-4a14-a971-7aca558dca0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"32768 32768 32768 32768 32768 32768 32768 32768\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "365a815d-5acd-4006-8f01-727b55e2b86b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "for subsample_pct in [5, 10]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"32768 32768 32768 32768 32768 32768 32768 32768\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc04dd24-f620-46af-b075-be2ee4b3df7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f35e99-5f89-48ee-9894-73b7fb3b9fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "# GATv2 encoder\n",
    "for subsample_pct in [25]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct_batch5\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                    f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                  \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c88c8e4-3f52-4ac4-8e9b-8ed601511180",
   "metadata": {},
   "source": [
    "### 3.4 Vizgen MERFISH Mouse Liver"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a91da577-1daa-484b-809f-4c7c080aa129",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### 3.4.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4f57ce-ed16-4994-a6dd-6dafe87a719f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"4096 4096 4096 4096 4096\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3\" \\\n",
    "               \" --run_index 8 7 6 5 4\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dc27972-079e-41d8-9aa1-e773ad39b34c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"4096 4096 4096\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None\" \\\n",
    "               \" --seeds 2 1 0\" \\\n",
    "               \" --run_index 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c89ecd3-de28-4a93-a243-5765d1d1d489",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1a8ad41-8fa2-4b3b-8a87-29a3f740fc75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0039d426-bda8-4891-8120-d02fdd663477",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d15848a7-9057-44e8-be6f-c2ff35a26241",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "# 50 epochs due to 2 day time limit\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 50\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65df455a-dc5d-48d1-9390-2f1cbd15c9fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 3\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a60f728-fa4c-445e-94bf-0ded90bf5bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 6 due to 2 day limit)\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"vizgen_merfish_mouse_liver\"\n",
    "cell_type_key = \"Cell_Type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"512 512\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 4\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d826697c-54bd-44ad-a33a-2fe56af946db",
   "metadata": {},
   "source": [
    "#### 3.4.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faa871f9-7357-4d38-a931-09a67728e0f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"4096 4096 4096 4096 4096 4096 4096 4096\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27024f2f-dc7f-4d20-89fa-93543f8b85e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 7\" \\\n",
    "                   \" --run_index 8\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4e3822a-3400-4d71-aa3b-24733b818b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 6\" \\\n",
    "                   \" --run_index 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e867ca2-1440-4a59-9a81-fc96cdb1ac72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 5\" \\\n",
    "                   \" --run_index 6\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d13bf7a-d45b-43a3-acb9-6a1d722bfcf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 4\" \\\n",
    "                   \" --run_index 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f08ff180-3913-4da9-a2be-81ba62448596",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 3\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32bd9a75-b996-43f6-931d-1c6fda5e8a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [25]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2dea5a0-d09d-4c5e-aeca-9f543192a49c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [25]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 1 0\" \\\n",
    "                   \" --run_index 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d56926b9-cd38-4fb7-b16d-9f99f75f6843",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"vizgen_merfish_mouse_liver_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"Cell_Type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c39bdcc-6c2f-45ae-bc75-305db782e1e2",
   "metadata": {},
   "source": [
    "### 3.5 Slide-seqV2 Mouse Hippocampus"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "278ca85b-f667-460d-b54f-98101eb221d4",
   "metadata": {},
   "source": [
    "#### 3.5.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1f0bccd-24db-4252-933a-fffb5e17c618",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384\" # out of memory\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ece3508-371d-4930-9287-822da20f2236",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"slideseqv2_mouse_hippocampus\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed715aa-9f15-445b-a37a-4ea469ca91b2",
   "metadata": {},
   "source": [
    "#### 3.5.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c94ac68-2133-41d2-913f-94438f7e26f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384\" # out of memory\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07a9315-cb35-4343-970e-2ff857b746ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fea47ff-9bfe-4850-9efb-5bd29fec260e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "for subsample_pct in [50]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    dataset = f\"slideseqv2_mouse_hippocampus_subsample_{subsample_pct}pct\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   \" --reference_batches None\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --mapping_entity_key mapping_entity\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   \" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "735ce16a-90a6-4e2d-b784-6fbcd3c17476",
   "metadata": {},
   "source": [
    "### 3.6 Simulation Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "720639fc-a61b-4d06-b581-9546c0517dde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "n_sim = \"1\"\n",
    "increment_mode = \"strong\" # \"weak\", strong\"\n",
    "n_genes = 1105\n",
    "n_locs = 10000\n",
    "dataset = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\"\n",
    "cell_type_key = \"cell_types\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"16384 16384 16384 16384 16384 16384 16384 16384\" # full dataset\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 100\" \\\n",
    "               \" --active_gp_thresh_ratio 0.03\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 30.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e1a812b-45ef-46ab-a99f-6f7b50108204",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "n_sim = \"1\"\n",
    "increment_mode = \"strong\" # \"weak\", strong\"\n",
    "n_genes = 1105\n",
    "n_locs = 10000\n",
    "dataset = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\"\n",
    "cell_type_key = \"cell_types\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 40\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 100\" \\\n",
    "               \" --active_gp_thresh_ratio 0.03\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 30.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        qos=\"gpu_long\",\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17b1464b-a809-402e-9cef-12178b11829d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "n_sim = \"1\"\n",
    "increment_mode = \"strong\" # \"weak\", strong\"\n",
    "n_genes = 1105\n",
    "n_locs = 10000\n",
    "dataset = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\"\n",
    "cell_type_key = \"cell_types\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"2048 2048 2048 2048 2048 2048 2048 2048\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 41\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 6 6 6 6 6 6 6 6\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 100\" \\\n",
    "               \" --active_gp_thresh_ratio 0.03\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 30.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        qos=\"gpu_long\",\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c28d134-9429-4127-b56b-349708ba3398",
   "metadata": {},
   "source": [
    "### 3.7 SDMBench Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14dbe99d-bf49-4f28-bb38-9c6d47c0f11c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"starmap_mouse_mpfc\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a982c771-3058-4e99-bdde-ebf1351d66c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"starmap_mouse_mpfc\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c924ed4c-726b-424a-9b46-e090831371c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"stereoseq_mouse_embryo\"\n",
    "cell_type_key = \"leiden\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"131072 131072 131072 131072 131072 131072 131072 131072\" # full dataset\n",
    "n_svg = 3000\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               \" --n_hvg 0\" \\\n",
    "               f\" --n_svg {n_svg}\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label gcnconv_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               \" --conv_layer_encoder gcnconv\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f22450e0-92fd-4a64-9bdd-e9dfa8f22c0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "dataset = \"stereoseq_mouse_embryo\"\n",
    "cell_type_key = \"leiden\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"1024 1024 1024 1024 1024 1024 1024 1024\"\n",
    "n_svg = 3000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               \" --reference_batches None\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --mapping_entity_key mapping_entity\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               \" --n_hvg 0\" \\\n",
    "               f\" --n_svg {n_svg}\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 500000.\" \\\n",
    "               \" --lambda_gene_expr_recon 300.\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               \" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               \" --use_new_gp_mask\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85aad8b2-b7ce-407f-856f-75331a35b730",
   "metadata": {},
   "source": [
    "### 3.8 Metrics Computation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1124e093-1499-421f-bf87-a4a4d68e186f",
   "metadata": {},
   "source": [
    "#### 3.8.1 NicheCompass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "455409d5-1191-4e62-b2a0-4569ddbc2cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e9a41a1-6517-4e95-8881-a06e012690bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4854fc25-77e5-4596-a9af-bf611cd05648",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd6ae43-2a23-47ce-bf8a-847ebb131d76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb00efa-876f-490d-8ea1-2108a8af9623",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e6e0860-1529-458c-94e0-285e5246d9c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63e8020-49b7-422d-94f5-0947cf38f008",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_gcnconv_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "621a8e6b-966a-4467-8b7d-acea178463e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"\n",
    "]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_gcnconv_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ab5ba3a-465c-468c-9fa9-cd7856afe33f",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "encoder = \"gcnconv\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}_{encoder}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b28453a0-2a04-4a3b-b714-031320b45ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "encoder = \"gatv2conv\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}_{encoder}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "089b890f-d2cc-4353-97de-4213da1cad55",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "encoder = \"gcnconv\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"starmap_mouse_mpfc\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}_{encoder}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87fe1b51-a176-489f-8410-be7429854cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "encoder = \"gatv2conv\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"starmap_mouse_mpfc\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}_{encoder}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ed9df30-d58c-464e-90a1-f42dec9794d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "encoder = \"gcnconv\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"stereoseq_mouse_embryo\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}_{encoder}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3178b4b-0058-47e7-b7e2-c650ddc63b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "encoder = \"gatv2conv\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"stereoseq_mouse_embryo\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}_{encoder}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3bf998b-50da-4d28-a073-b4bde8abf7be",
   "metadata": {},
   "source": [
    "#### 3.8.2 STACI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c643a28f-6d5c-4112-94e1-72b430fae5e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"\n",
    "]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"staci_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            t=\"48:00:00\",\n",
    "            p=\"gpu_p\",\n",
    "            qos=\"gpu_normal\",\n",
    "            mem=\"160G\",\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad391052-6fca-498a-ade0-9b0912980b8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"\n",
    "               ]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"staci_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            t=\"48:00:00\",\n",
    "            p=\"gpu_p\",\n",
    "            qos=\"gpu_normal\",\n",
    "            mem=\"160G\",\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fbd07da-9081-427c-bccd-8aaf0490a6e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"\n",
    "]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"staci_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            t=\"48:00:00\",\n",
    "            p=\"gpu_p\",\n",
    "            qos=\"gpu_normal\",\n",
    "            mem=\"160G\",\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e59f526-6401-4741-ac75-2ca36cfc8895",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"staci_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            t=\"48:00:00\",\n",
    "            p=\"gpu_p\",\n",
    "            qos=\"gpu_normal\",\n",
    "            mem=\"160G\",\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ffcf61e-2cfd-4047-b23b-a811aed77116",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"staci\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a830391b-5617-4902-8d6a-6a7a3fba34b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"staci\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"starmap_mouse_mpfc\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e42e7fd-aa6f-4856-858e-b66f93db1d5e",
   "metadata": {},
   "source": [
    "#### 3.8.3 GraphST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0d39cdd-732b-45c3-ba71-239c23f18ab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36cf5b6f-69df-446e-afe8-83c21290e4f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12061d47-4585-4d81-9592-c543587c92f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41babdbf-f415-4bbb-8072-32c77fceb801",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\"]:\n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c96cf85-5df4-4872-b641-db28682867ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"graphst\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e6cb872-2454-4a22-85b9-ef85d7d0ffba",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"graphst\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"starmap_mouse_mpfc\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a1515f0-00f7-43c9-866b-7ac0d1951e56",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"graphst\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"stereoseq_mouse_embryo\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd53e480-ebd7-481c-b50a-96b07b3b393c",
   "metadata": {},
   "source": [
    "#### 3.8.4 DeepLinc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16579147-7bb6-40ee-9301-2d66554e35b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_deeplinc.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"deeplinc_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key deeplinc_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ddfdb24-d761-4b18-96d9-b49106f5e073",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_deeplinc.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"deeplinc_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key deeplinc_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50f0d25c-47a4-43d1-963f-97bf86974c74",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_deeplinc.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"deeplinc_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key deeplinc_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6978d45e-c4ed-497f-bceb-ed1cb78c787f",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"deeplinc\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43cc8ff7-a026-4195-8283-dcda4110a919",
   "metadata": {},
   "source": [
    "#### 3.8.5 CellCharter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7af517d5-d115-4e79-bb60-43c233e3b5b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"\n",
    "               ]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cffece7-3dcb-4cd4-96d8-96e8ba7ab422",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"\n",
    "               ]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8b2f494-500c-4d88-bb59-5f680a67a032",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"\n",
    "               ]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83ed0800-3a49-4abf-b0e3-cc9a3cef586e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"\n",
    "               ]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcf3d52b-c048-42c2-8605-e466132e6c40",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"cellcharter\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26d9eaae-1262-4b81-adc7-96f04506054c",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"cellcharter\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"starmap_mouse_mpfc\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1334da31-b06f-47ba-a4f1-d1af1f4b1d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"cellcharter\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"stereoseq_mouse_embryo\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e79cdba4-db9c-4373-ae2f-b8dbad1e4044",
   "metadata": {},
   "source": [
    "#### 3.8.6 BANKSY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea7da2b-69f6-431b-ba1a-f8323de3f19e",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61cfce68-dc9c-45a7-aafb-bec7a0a1ceaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3393461-ba2c-41f3-ac49-d4dbc58f678b",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "079cf6b8-8ba6-4cc0-8ebb-18bd7a4564d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\b\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cff33d30-a828-4ec2-8cac-a7184523f0d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93f76b2c-2e14-4fcf-9eba-725dbd07cd1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"starmap_mouse_mpfc\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42754a6d-d26a-4d3c-81df-d938ad625d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "niche_type_key = \"niche_type\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"stereoseq_mouse_embryo\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/single_sample_method_benchmarking/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --niche_type_key {niche_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  \" --include_sdmbench\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a8a76db-4b4a-48d7-a0d7-5dcea32c6910",
   "metadata": {},
   "source": [
    "#### 3.8.7 scVI (Not Included)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b70c8d-1b55-493a-bf4b-b8481f2db9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                #\"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                #\"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                #\"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                #\"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                #\"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"\n",
    "               ]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21c715be-3352-4d76-b94f-4ee2b451cd76",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d42df767-8650-428a-b74f-160a58090fca",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11227ee6-b65c-4f3a-aeda-e69819766070",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"]:    \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac03756-453e-41a6-b823-4a7da0211942",
   "metadata": {},
   "source": [
    "#### 3.8.8 expiMap (Not Included)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36802aa9-aa15-4455-939f-2701b0ac64ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct_embryo2\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct_embryo2\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d082e4f-db2e-4584-9976-b1dea885418d",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct_batch5\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct_batch5\"]:   \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d61960c-95e6-4396-a5df-8ccd83ff73c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"vizgen_merfish_mouse_liver\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_50pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_25pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_10pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_5pct\",\n",
    "                \"vizgen_merfish_mouse_liver_subsample_1pct\"]: \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ece5ce1-b630-4d02-ae72-4e8b483e5a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"slideseqv2_mouse_hippocampus\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_50pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_25pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_10pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_5pct\",\n",
    "                \"slideseqv2_mouse_hippocampus_subsample_1pct\"\n",
    "               ]:  \n",
    "    task = \"single_sample_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"None\"\n",
    "    file_name = f\"{dataset}_expimap.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "    job_name_prefix = f\"expimap_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key expimap_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622f6c1d-65a1-4318-b03c-c12444fd0f4c",
   "metadata": {},
   "source": [
    "## 4. NicheCompass Sample Integration Method Benchmarking"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd147edf-4ccd-402d-aca9-ce19cbd1b34f",
   "metadata": {},
   "source": [
    "### 4.1 seqFISH Mouse Organogenesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b865230b-ab67-4474-a29c-dcb1d2294340",
   "metadata": {},
   "source": [
    "#### 4.1.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f764998-8726-419f-bafe-d627f75a1257",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"65536 65536 65536 65536 65536 65536 65536 65536\"\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = True\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f649c1e6-5384-45ce-90dd-a7413806b323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256 256 256 256 256 256 256 256\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = True\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac84fe1-0698-4c04-868f-2dd6e7d73356",
   "metadata": {},
   "source": [
    "#### 4.1.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63f15cc6-4e6d-4473-a126-dccb02b889a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"65536 65536 65536 65536 65536 65536 65536 65536\"\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = True\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95eb73ce-64bf-4097-92fd-e5431f66652f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\"\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = True\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d949f3a-99b1-4e68-8eb7-62cd37a67604",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"512 512 512 512 512 512 512 512\"\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = True\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b334564-7198-4344-8e8b-b20897765f3c",
   "metadata": {},
   "source": [
    "### 4.2 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4d74ae3-ea69-42e8-824f-5c014103f9ed",
   "metadata": {},
   "source": [
    "This is not used for the manuscript."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dfecebf-0b36-47f9-a3a4-b59938733399",
   "metadata": {},
   "source": [
    "#### 4.2.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51df66d0-94b4-4e12-9e50-f91fc5535f0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"visium_mouse_brain\"\n",
    "reference_batches = \"batch1 batch2\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 5000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c01381b-77f3-4b67-8098-08efe359f842",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d4a594f-c90a-4fd2-b360-e42465f1ed5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "679dc44d-14de-493b-8407-6bcf0debf032",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None None None None None None None\" \\\n",
    "               \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "               \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14b3c6cc-195f-4295-bcac-5900ff9004a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b75b91a4-0c56-4fce-8148-7ac27f64faa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c00536d3-5956-46e1-93d5-5259485409f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf62bc85-0ae9-40d0-9743-20066708c323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00bb057e-1010-44a0-9c16-0b0a8962735a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256 256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23e672a8-2aec-4150-8e93-0adf9fb78936",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "species = \"mouse\"\n",
    "edge_batch_size_str = \"256 256\" # out of memory \n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --filter_genes\" \\\n",
    "               f\" --n_svg 3000\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               \" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61a05506-ee44-4d31-a43a-63a82c40ad1f",
   "metadata": {},
   "source": [
    "#### 4.2.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ba78064-ce5d-4f56-b025-d2762607ad43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [50, 25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 3000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab6eb152-be84-4685-952b-67c592d68ea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tmp\n",
    "for subsample_pct in [5]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 3000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "842b8886-ede3-4fe2-8a87-7374484970f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 7 6\" \\\n",
    "                   \" --run_index 8 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 3000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e709b55-80ef-4071-b11e-7ad8a83c421b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 5 4\" \\\n",
    "                   \" --run_index 6 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 3000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f803ece-528a-4b3a-9cdb-205ae4bb164e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256 256 256\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 3000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3979c1c4-cf88-4fa7-8e0f-afe759c8a7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"seqfish_mouse_organogenesis_imputed_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    cell_type_key = \"celltype_mapped_refined\"\n",
    "    species = \"mouse\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\" # out of memory \n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --filter_genes\" \\\n",
    "                   f\" --n_svg 3000\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   \" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ab268b-d86a-4843-9623-6f687c4f0a11",
   "metadata": {},
   "source": [
    "### 4.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfda56b3-cb23-4b88-95f4-b3d8e8651ec2",
   "metadata": {},
   "source": [
    "#### 4.3.1 Spatial Transcriptomics Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f44c05d0-6f28-46c5-99d7-c182917f2850",
   "metadata": {},
   "source": [
    "##### 4.3.1.1 GCN W/O FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c27c72a-b425-4514-a4d5-875309a40987",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8870fde7-98ad-410d-b431-b51d9501b50c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7 \" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da355e82-fdac-4d61-a665-844dcb2a59c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6 \" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eca55161-e696-478f-9cc5-ae6174459bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5 \" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c75ab927-fa76-4d59-b309-be7038fe1fef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192 8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a38cd78-7ee2-4121-89d2-92aaefc96504",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192 8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a8200c4-1f70-4313-8561-f658135d2bf1",
   "metadata": {},
   "source": [
    "##### 4.3.1.2 GCN with FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3bceaec-cd5f-43e1-ab60-b15d1d12303c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac0da1d-9480-44fa-a49f-b6e3e260c1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33b10567-5c96-48f0-90b5-bcaa81f122a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 3\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "017d9d1e-c5f0-4564-81e2-a0fedf017655",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 4\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8bbd153-65ba-49da-b197-0b8a091ad253",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192 8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 5\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e5def3a-7d35-41ef-9921-cbc784f53cb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "edge_batch_size_str = \"8192 8192\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 6\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00b8467f-3fe4-4247-a875-b6afff87ffbd",
   "metadata": {},
   "source": [
    "##### 4.3.1.3 GATv2 W/O FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5db4fe3-e259-4370-90fd-f80f34ef0d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb25914-1791-4677-b8c1-460d637bb993",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "045bc3c8-10d1-4e11-9c5d-f2845269a9b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7b010d3-5626-453f-a82a-5d84e82613c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "933371f3-44a0-49e6-b40c-0b0433853405",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512 512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eae3159-6692-449e-8e7e-5f74cdd1b31f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"512 512\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc681836-c890-434d-92e9-285ec77ebbb8",
   "metadata": {},
   "source": [
    "##### 4.3.1.4 GATv2 with FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5652004-1e58-41e8-91ef-c004b4449d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 1\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 7\" \\\n",
    "               \" --run_index 8\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fae3537-ad20-411f-83b5-09e8b5850902",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 2\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 16\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 6\" \\\n",
    "               \" --run_index 7\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10059c4b-4dd0-466b-b68c-eff07fb7e70e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 3\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 5\" \\\n",
    "               \" --run_index 6\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a98ce408-bb68-4778-b421-899a05084367",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 4\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 12\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None\" \\\n",
    "               \" --seeds 4\" \\\n",
    "               \" --run_index 5\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65513179-16cc-4467-820b-d0eea88e7f42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256 256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 5\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 8 8\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 3 2\" \\\n",
    "               \" --run_index 4 3\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acdfa499-10fb-416c-8ed1-bc2162acb538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 6 due to 2 day limit)\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "cell_type_key = \"cell_type\"\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size_str = \"256 256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov\"\n",
    "cat_covariates_no_edges = \"True False\"\n",
    "cat_covariates_embeds_nums = \"3 30\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "job_id = 6\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "script_args =  \" --adata_new_name None\" \\\n",
    "               \" --n_neighbors_list 4 4\" \\\n",
    "               f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "               \" --node_batch_size_list None None\" \\\n",
    "               \" --seeds 1 0\" \\\n",
    "               \" --run_index 2 1\" \\\n",
    "               f\" --cell_type_key {cell_type_key}\" \\\n",
    "               \" --no-filter_genes\" \\\n",
    "               \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "               \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "               \" --include_mebocost_gps\" \\\n",
    "               f\" --species {species}\" \\\n",
    "               \" --gp_filter_mode subset\" \\\n",
    "               \" --combine_overlap_gps\" \\\n",
    "               \" --overlap_thresh_source_genes 0.9\" \\\n",
    "               \" --overlap_thresh_target_genes 0.9\" \\\n",
    "               \" --overlap_thresh_genes 0.9\" \\\n",
    "               f\" --dataset {dataset}\" \\\n",
    "               f\" --reference_batches {reference_batches}\" \\\n",
    "               \" --counts_key counts\" \\\n",
    "               f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "               f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "               \" --spatial_key spatial\" \\\n",
    "               \" --adj_key spatial_connectivities\" \\\n",
    "               \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "               \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "               \" --gp_names_key nichecompass_gp_names\" \\\n",
    "               f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "               \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "               \" --latent_key nichecompass_latent\" \\\n",
    "               \" --n_addon_gp 10\" \\\n",
    "               \" --active_gp_thresh_ratio 0.\" \\\n",
    "               \" --gene_expr_recon_dist nb\" \\\n",
    "               f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "               f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "               \" --log_variational\" \\\n",
    "               f\" --node_label_method one-hop-norm\" \\\n",
    "               \" --n_layers_encoder 1\" \\\n",
    "               \" --n_hidden_encoder None\" \\\n",
    "               f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "               \" --n_epochs 400\" \\\n",
    "               \" --n_epochs_all_gps 25\" \\\n",
    "               \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "               \" --lr 0.001\" \\\n",
    "               \" --lambda_edge_recon 5000000.\" \\\n",
    "               \" --lambda_gene_expr_recon 3000.\" \\\n",
    "               f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "               f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "               f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "               \" --lambda_group_lasso 0.\" \\\n",
    "               f\" --lambda_l1_masked 0.\" \\\n",
    "               \" --lambda_l1_addon 0.\" \\\n",
    "               f\" --n_sampled_neighbors 4\" \\\n",
    "               f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c112831f-87a4-41f8-9478-02f71d678a3f",
   "metadata": {},
   "source": [
    "#### 4.3.2 Spatial Transcriptomics Data Subsamples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afc9a425-299e-488f-8682-64692896640a",
   "metadata": {},
   "source": [
    "##### 4.3.2.1 GCN W/O FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dd3fe18-fd00-4fcd-9dfe-2700ff332153",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 7\" \\\n",
    "                   \" --run_index 8\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffe16725-7ea6-4102-856a-02b0db964b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 6\" \\\n",
    "                   \" --run_index 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ab15aad-2467-40b3-a145-06168ebe2ab7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None\" \\\n",
    "                   \" --seeds 5 4\" \\\n",
    "                   \" --run_index 6 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8121ada5-3134-4b63-8898-5a3dfc6670ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 4 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e210d211-f81c-4098-af7f-17adee682eb4",
   "metadata": {},
   "source": [
    "##### 4.3.2.2 GCN with FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3307ad2-85da-4463-b7ef-6e629e3e444b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 7\" \\\n",
    "                   \" --run_index 8\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3d323e0-74ec-4a0b-aa55-94e4d7527c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 2\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 6\" \\\n",
    "                   \" --run_index 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47c82166-d254-42d3-9c4f-2eb9b105549b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 3\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 5\" \\\n",
    "                   \" --run_index 6\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f28a9fc-33de-4f6b-a95d-8a3563bd848f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 4 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 4\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 4\" \\\n",
    "                   \" --run_index 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c0de440-62df-4e6b-8049-0b962e8cb17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder (run 5 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 5\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73704f6-331a-4f60-987d-6a4c8121d428",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "for subsample_pct in [25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"8192 8192 8192 8192 8192 8192 8192 8192\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gcnconv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27762183-3e28-45cd-aa83-34d04bf9aa15",
   "metadata": {},
   "source": [
    "##### 4.3.2.3 GATv2 W/O FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "296d296e-a8aa-46ba-a5f9-d62393aedbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 7\" \\\n",
    "                   \" --run_index 8\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beedf017-a04a-4348-a207-0aab213e5c4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 6\" \\\n",
    "                   \" --run_index 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e18f26c9-a86c-4f09-85c5-4575cf4a9856",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 5\" \\\n",
    "                   \" --run_index 6\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e3e902a-17d0-4182-93e4-807fcb2a8a5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 4\" \\\n",
    "                   \" --run_index 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77ddc526-afc4-4153-85f6-f167ae190129",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"512 512 512 512\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a77d8f04-2cf4-4bb2-ae1f-5840338bc237",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch\"\n",
    "    cat_covariates_no_edges = \"True\"\n",
    "    cat_covariates_embeds_nums = \"3\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3ef9fb0-77ef-47d7-b1a5-676019ac97fe",
   "metadata": {},
   "source": [
    "##### 4.3.2.1 GATv2 with FOV Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c721c0-b3a0-41e5-a761-31961afb134b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 1 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 7\" \\\n",
    "                   \" --run_index 8\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfceb76f-b10f-4ffa-8071-722184615f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 2 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 2\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 6\" \\\n",
    "                   \" --run_index 7\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60361a0f-9859-4e63-9f4e-ad25d3819bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 3 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 3\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 5\" \\\n",
    "                   \" --run_index 6\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e86776-fb83-4631-b191-c903bda90f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 4 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 4\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 12\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None\" \\\n",
    "                   \" --seeds 4\" \\\n",
    "                   \" --run_index 5\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4c6893c-cb82-4608-b981-76574acef11b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder (run 5 due to 2 day limit)\n",
    "for subsample_pct in [50]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256 256 256 256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 5\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None\" \\\n",
    "                   \" --seeds 3 2 1 0\" \\\n",
    "                   \" --run_index 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35561bbe-5e42-43b2-8d75-70d54b172945",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "for subsample_pct in [25, 10, 5, 1]:\n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    dataset = f\"nanostring_cosmx_human_nsclc_subsample_{subsample_pct}pct\"\n",
    "    reference_batches = \"batch1 batch2 batch3\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    species = \"human\"\n",
    "    edge_batch_size_str = \"256 256 256 256 256 256 256 256\"\n",
    "    cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "    cat_covariates_keys = \"batch fov\"\n",
    "    cat_covariates_no_edges = \"True False\"\n",
    "    cat_covariates_embeds_nums = \"3 30\"\n",
    "    lambda_cat_covariates_contrastive = 0.\n",
    "    contrastive_logits_pos_ratio = 0.\n",
    "    contrastive_logits_neg_ratio = 0.\n",
    "    conv_layer_encoder = \"gatv2conv\"\n",
    "\n",
    "    job_name_prefix = f\"{dataset}_nichecompass_{conv_layer_encoder}_{task}\"\n",
    "    job_id = 1\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"train_nichecompass_benchmarking_models.py\"\n",
    "    script_args =  \" --adata_new_name None\" \\\n",
    "                   \" --n_neighbors_list 16 16 12 12 8 8 4 4\" \\\n",
    "                   f\" --edge_batch_size_list {edge_batch_size_str}\" \\\n",
    "                   \" --node_batch_size_list None None None None None None None None\" \\\n",
    "                   \" --seeds 7 6 5 4 3 2 1 0\" \\\n",
    "                   \" --run_index 8 7 6 5 4 3 2 1\" \\\n",
    "                   f\" --cell_type_key {cell_type_key}\" \\\n",
    "                   \" --no-filter_genes\" \\\n",
    "                   \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                   \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                   \" --include_mebocost_gps\" \\\n",
    "                   f\" --species {species}\" \\\n",
    "                   \" --gp_filter_mode subset\" \\\n",
    "                   \" --combine_overlap_gps\" \\\n",
    "                   \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                   \" --overlap_thresh_genes 0.9\" \\\n",
    "                   f\" --dataset {dataset}\" \\\n",
    "                   f\" --reference_batches {reference_batches}\" \\\n",
    "                   \" --counts_key counts\" \\\n",
    "                   f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                   f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                   \" --spatial_key spatial\" \\\n",
    "                   \" --adj_key spatial_connectivities\" \\\n",
    "                   \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                   \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                   \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                   f\" --model_label {conv_layer_encoder}_{task}\" \\\n",
    "                   \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                   \" --latent_key nichecompass_latent\" \\\n",
    "                   \" --n_addon_gp 10\" \\\n",
    "                   \" --active_gp_thresh_ratio 0.\" \\\n",
    "                   \" --gene_expr_recon_dist nb\" \\\n",
    "                   f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                   f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                   \" --log_variational\" \\\n",
    "                   f\" --node_label_method one-hop-norm\" \\\n",
    "                   \" --n_layers_encoder 1\" \\\n",
    "                   \" --n_hidden_encoder None\" \\\n",
    "                   f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                   \" --n_epochs 400\" \\\n",
    "                   \" --n_epochs_all_gps 25\" \\\n",
    "                   \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                   \" --lr 0.001\" \\\n",
    "                   \" --lambda_edge_recon 5000000.\" \\\n",
    "                   \" --lambda_gene_expr_recon 3000.\" \\\n",
    "                   f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                   f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                   f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                   \" --lambda_group_lasso 0.\" \\\n",
    "                   f\" --lambda_l1_masked 0.\" \\\n",
    "                   \" --lambda_l1_addon 0.\" \\\n",
    "                   f\" --n_sampled_neighbors 4\" \\\n",
    "                   f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args,\n",
    "            nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8995fe1f-d843-444b-b288-01936b2e4544",
   "metadata": {},
   "source": [
    "### 4.4 Metrics Computation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c4a4cbe-402e-4b61-a82e-6c7b71ec7c13",
   "metadata": {},
   "source": [
    "#### 4.4.1 NicheCompass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75d9705a-5225-46c0-a091-efedab7873df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d77fd59-9f0a-4d9a-ab98-57952616b018",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 2\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f9df2b0-bd58-4f1c-8cbf-b689a8199263",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5772627b-6d78-4c16-8693-26c61e975e47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder\n",
    "job_id = 2\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "692e322e-8fef-4961-b123-a8625c502560",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder W/O FoV embedding\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e2b7a4f-468e-4d28-b207-21be9da0310c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder with FoV embedding\n",
    "job_id = 2\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv_fov.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00da0400-7129-47c5-a557-ddaa30dd2a55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GCN encoder with contrastive loss\n",
    "job_id = 3\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_nichecompass_gcnconv_cont.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af62db45-2169-41f8-ac57-c777a3451c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder W/O FoV embedding\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba901b31-39fb-4162-b51e-a9b477fa58ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder with FoV embedding\n",
    "job_id = 2\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv_fov.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8e3cc45-8c53-4a1f-a13a-01562379c87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# GATv2 encoder with contrastive loss\n",
    "job_id = 3\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_nichecompass_gatv2conv_cont.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"nichecompass_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key nichecompass_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b52f9b3-c547-41db-8a89-4b588a99367f",
   "metadata": {},
   "source": [
    "#### 4.4.2 STACI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea174183-27fc-4cf1-b054-ed24b9b17a75",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:    \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"staci_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6cdfaed-007c-4952-9bf6-a14f0e930293",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"staci_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "628985a6-8dd6-4791-a542-d04fb61eecab",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                #\"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    run_number = \"8\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_staci_run{run_number}.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"staci_sample_integration_method_benchmarking_{dataset}_metrics_computation_run{run_number}\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\" \\\n",
    "                  f\" --run_number {run_number}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1288fd7-42be-40bf-8b1b-dda83075530c",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:     \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_staci.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"staci_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key staci_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10ce8451-0b91-4bbd-9da6-7aadb4f2cd93",
   "metadata": {},
   "source": [
    "#### 4.4.3 GraphST"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe2c202-e672-4742-a743-d3182d239662",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b33cf69-f1cb-4b2a-8768-c15093bb6a78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PASTE alignment\n",
    "job_id = 2\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_graphst_paste.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "774b3bcb-aae1-4c7a-81da-96ebd4e34d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee57e738-455b-432a-9c25-64ac0e2a2933",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PASTE alignment\n",
    "job_id = 2\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_graphst_paste.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b3e5e85-0b12-4b56-8c1d-0ed0ef332d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"]:     \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_graphst.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eebe7a1e-c7c8-4c99-9273-bbe9b30a1d11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PASTE alignment\n",
    "job_id = 2\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:     \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_graphst_paste.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"graphst_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key graphst_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec8c6fb-4cfb-4cfb-8bb8-01b7622f26fd",
   "metadata": {},
   "source": [
    "#### 4.4.4 CellCharter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32fc2a1e-b2f3-45bb-b75b-fad5036855ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5132b045-d342-426b-80aa-f9d45de3cce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5606fbf0-b572-4fcb-b25d-a4f723ca041d",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"\n",
    "               ]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_cellcharter.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"cellcharter_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key cellcharter_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bda88c5-4973-4754-b11d-02c27409ca83",
   "metadata": {},
   "source": [
    "#### 4.4.5 BANKSY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1af2923-fbac-4362-a58b-04454b9f0417",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "batch_key = \"batch\"\n",
    "batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:   \n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d07f3442-cef5-482c-8997-037dc5aa5d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "               ]:  \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c57fcbef-ca2a-4296-9b63-46bd028e6c47",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"banksy\"\n",
    "task = \"sample_integration_method_benchmarking\"\n",
    "cell_type_key = \"cell_type\"\n",
    "batch_key = \"batch\"\n",
    "batches = \"batch1 batch2 batch3\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./benchmarking/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4376fbd-d4b4-443a-ac33-047fc729c487",
   "metadata": {},
   "source": [
    "#### 4.4.6 scVI (Not Included)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71925802-31c2-41a6-8f35-766bc471330e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bb66751-7a5d-4d62-8083-bdac567f8a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis_imputed\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2963e3bf-efe6-47b3-82af-53b9d15f3f15",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for dataset in [\"nanostring_cosmx_human_nsclc\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_50pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_25pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_10pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_5pct\",\n",
    "                \"nanostring_cosmx_human_nsclc_subsample_1pct\"]:   \n",
    "    task = \"sample_integration_method_benchmarking\"\n",
    "    cell_type_key = \"cell_type\"\n",
    "    batch_key = \"batch\"\n",
    "    batches = \"batch1 batch2 batch3\"\n",
    "    file_name = f\"{dataset}_scvi.h5ad\"\n",
    "    metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi basw blisi kbet pcr\"\n",
    "\n",
    "    job_name_prefix = f\"scvi_sample_integration_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --batches {batches}\" \\\n",
    "                  f\" --latent_key scvi_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "828f673a-9f55-4b43-9ff2-ca286d8ed45a",
   "metadata": {},
   "source": [
    "## 5. NicheCompass Reference Model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56180669-13e7-4873-8887-5804360c2b75",
   "metadata": {},
   "source": [
    "### 5.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e6bed3a-6613-4101-81e5-9c19bb217b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0]):\n",
    "            for n_addon_gp in [10]:\n",
    "                for n_svg in [3000]:\n",
    "                    for conv_layer_encoder in [\"gcnconv\"]:\n",
    "                        for n_neighbors in [4]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"./reference/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c481d469-ffec-4669-be35-97736e5f2a79",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 2\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.01]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0]):\n",
    "            for n_addon_gp in [100]:\n",
    "                for n_svg in [5000]:\n",
    "                    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "                        for n_neighbors in [8]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch1 batch2 batch5 batch6\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e73e87a-c22f-4ad6-9403-5b25018ffbb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 3\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.01]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0.]):\n",
    "            for n_addon_gp in [100]:\n",
    "                for n_svg in [5000]:\n",
    "                    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "                        for n_neighbors in [8]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f384c59-faac-437b-b5e6-cd800b191b61",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 4\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.01]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0.]):\n",
    "            for n_addon_gp in [100]:\n",
    "                for n_svg in [5000]:\n",
    "                    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "                        for n_neighbors in [4]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f43575-1a26-453b-936e-29329f723bec",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 5\n",
    "for lambda_edge_recon, lambda_gene_expr_recon in zip([5000000], [3000]):\n",
    "    for active_gp_thresh_ratio in [0.01]:\n",
    "        for lambda_l1_masked, lambda_l1_addon in zip([0], [0.]):\n",
    "            for n_addon_gp in [100]:\n",
    "                for n_svg in [5000]:\n",
    "                    for conv_layer_encoder in [\"gatv2conv\"]:\n",
    "                        for n_neighbors in [12]:\n",
    "                            task = \"reference\"\n",
    "                            dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "                            reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "                            n_sampled_neighbors = 4\n",
    "                            n_hvg = 0\n",
    "                            species = \"mouse\"\n",
    "                            node_label_method = \"one-hop-norm\"\n",
    "                            edge_batch_size = \"256\" # \"512\"\n",
    "                            cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "                            cat_covariates_keys = \"batch\"\n",
    "                            cat_covariates_no_edges = \"True\"\n",
    "                            cat_covariates_embeds_nums = \"3\"\n",
    "                            lambda_cat_covariates_contrastive = 0.\n",
    "                            contrastive_logits_pos_ratio = 0.\n",
    "                            contrastive_logits_neg_ratio = 0.\n",
    "\n",
    "                            job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "                            job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "                            script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "                            script_name = \"train_nichecompass_reference_model.py\"\n",
    "                            script_args = f\" --dataset {dataset}\" \\\n",
    "                                          f\" --reference_batches {reference_batches}\" \\\n",
    "                                          f\" --n_neighbors {n_neighbors}\" \\\n",
    "                                          \" --filter_genes\" \\\n",
    "                                          f\" --n_hvg {n_hvg}\" \\\n",
    "                                          f\" --n_svg {n_svg}\" \\\n",
    "                                          \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "                                          \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "                                          \" --include_mebocost_gps\" \\\n",
    "                                          f\" --species {species}\" \\\n",
    "                                          \" --gp_filter_mode subset\" \\\n",
    "                                          \" --combine_overlap_gps\" \\\n",
    "                                          \" --overlap_thresh_source_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_target_genes 0.9\" \\\n",
    "                                          \" --overlap_thresh_genes 0.9\" \\\n",
    "                                          \" --counts_key counts\" \\\n",
    "                                          \" --condition_key batch\" \\\n",
    "                                          f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "                                          f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "                                          \" --spatial_key spatial\" \\\n",
    "                                          \" --adj_key spatial_connectivities\" \\\n",
    "                                          \" --mapping_entity_key mapping_entity\" \\\n",
    "                                          \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "                                          \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "                                          \" --gp_names_key nichecompass_gp_names\" \\\n",
    "                                          f\" --model_label {task}\" \\\n",
    "                                          \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "                                          \" --latent_key nichecompass_latent\" \\\n",
    "                                          f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "                                          f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "                                          \" --gene_expr_recon_dist nb\" \\\n",
    "                                          f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "                                          f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "                                          \" --log_variational\" \\\n",
    "                                          f\" --node_label_method {node_label_method}\" \\\n",
    "                                          \" --n_layers_encoder 1\" \\\n",
    "                                          \" --n_hidden_encoder None\" \\\n",
    "                                          f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "                                          \" --n_epochs 100\" \\\n",
    "                                          \" --n_epochs_all_gps 25\" \\\n",
    "                                          \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "                                          \" --lr 0.001\" \\\n",
    "                                          f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "                                          f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "                                          f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "                                          f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "                                          f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "                                          \" --lambda_group_lasso 0.\" \\\n",
    "                                          f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "                                          f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "                                          f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "                                          \" --node_batch_size None\" \\\n",
    "                                          f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "                                          f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "                            submit_python_script(\n",
    "                                    job_name_prefix=job_name_prefix,\n",
    "                                    job_id=job_id,\n",
    "                                    job_folder_path=job_folder_path,\n",
    "                                    conda_env_name=conda_env_name,\n",
    "                                    script_folder_path=script_folder_path,\n",
    "                                    script_name=script_name,\n",
    "                                    script_args=script_args,\n",
    "                                    nice=10000)\n",
    "\n",
    "                            job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3e4152d-8ca1-47d4-b8f3-4f3001836639",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 6 # different random seed #1\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 100\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 42\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "994556b5-9c96-4719-9aaa-c77dd96bfa82",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 6 # different random seed #2\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 100\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 43 # 43\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80a30f22-49f2-4be7-9dc7-24d974987dd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 7\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 100\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a9b74cc-90de-4171-b0a0-8899dc3882f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 8\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 100\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c801c8-a395-4a63-8273-6e623f159d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 8 # Different prior GP sets (Omnipath)\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_omnipath_gps\" \\\n",
    "              \" --no-include_nichenet_gps\" \\\n",
    "              \" --no-include_mebocost_gps\" \\\n",
    "              \" --no-include_collectri_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf50363a-9937-49b1-b7d1-d992ce09b780",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 9 # Different prior GP sets (NicheNet)\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --no-include_omnipath_gps\" \\\n",
    "              \" --include_nichenet_gps\" \\\n",
    "              \" --no-include_mebocost_gps\" \\\n",
    "              \" --no-include_collectri_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06b7f382-875d-48b4-92e8-c820342f2863",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 10 # Different prior GP sets (MEBOCOST)\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --no-include_omnipath_gps\" \\\n",
    "              \" --no-include_nichenet_gps\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              \" --no-include_collectri_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f642e07c-9201-4b48-8f1d-7ea920acf1d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 11 # Different prior GP sets (CollecTRI)\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --no-include_omnipath_gps\" \\\n",
    "              \" --no-include_nichenet_gps\" \\\n",
    "              \" --no-include_mebocost_gps\" \\\n",
    "              \" --include_collectri_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "594e803c-fa53-4e25-8361-763a3a4f8635",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 12 # Subsample Leave out Embryo 3\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_50pct\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e117a066-13df-4ff3-a65a-b1e10303c1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 13 # Subsample Leave Out Embryo 3\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_25pct\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a105703-3025-4694-b8a2-449a6ef79079",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 14 # Subsample Leave Out Embryo 3\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_10pct\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4797160-6717-425c-8ec8-6d54b4903364",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 15 # Subsample Leave Out Embryo 3\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_5pct\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b51112-29cc-4f5f-b489-719ae62a06da",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 16 # Subsample Leave Out Embryo 3\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 0\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "n_neighbors = 8\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7fa865f-8110-4c2e-a982-dc4b90fa9cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 17 # radius-based knn graph\n",
    "lambda_edge_recon = 5000000\n",
    "lambda_gene_expr_recon = 3000\n",
    "active_gp_thresh_ratio = 0.01\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "n_addon_gp = 100\n",
    "n_svg = 5000\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "graph_type = \"radius\"\n",
    "radius = 0.08\n",
    "task = \"reference\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6\"\n",
    "n_sampled_neighbors = 4\n",
    "n_hvg = 0\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = \"256\"\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "seed = 0\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --graph_type {graph_type}\" \\\n",
    "              f\" --radius {radius}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --seed {seed}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aee0b50-4cfa-49f3-84b4-134a62b061b4",
   "metadata": {},
   "source": [
    "### 5.2 STARmap PLUS Mouse Central Nervous System"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7829046a-eb97-408c-8c8e-e3086fdeb8f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(\"../datasets/st_data/gold/starmap_plus_mouse_alzheimers.h5ad\")\n",
    "adata2 = sc.read_h5ad(\"../datasets/st_data/gold/starmap_plus_mouse_cns.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "368f12e9-a4bc-4591-abf4-e7360acdd9d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = set([gene.upper() for gene in adata.var.index.tolist()])\n",
    "genes2 = set(adata2.var.index.tolist())\n",
    "genes_intersect = genes.intersection(genes2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54ec8b19-b343-449c-a4cb-303e98953e3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx in range(1,21):\n",
    "    adata_test = sc.read_h5ad(f\"../datasets/st_data/gold/starmap_plus_mouse_cns_batch{idx}.h5ad\")\n",
    "    adata_test.write(f\"../datasets/st_data/gold/starmap_plus_mouse_batch{idx}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da28522d-5c31-470e-a6e1-4b01217361bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx in range(1,9):\n",
    "    adata_test = sc.read_h5ad(f\"../datasets/st_data/gold/starmap_plus_mouse_alzheimers_batch{idx}.h5ad\")\n",
    "    adata_test.write(f\"../datasets/st_data/gold/starmap_plus_mouse_batch{20+idx}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05bf11ae-694e-4fba-86fb-ed31cdc3cd5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_test = sc.read_h5ad(f\"../datasets/st_data/gold/starmap_plus_mouse_batch25.h5ad\")\n",
    "adata_test = adata_test[:, [gene for gene in adata_test.var.index.tolist() if gene.upper() in genes_intersect]]\n",
    "selected_genes = adata_test.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02473ba-42b7-4a76-a486-5fec310869bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx in range(1,29):\n",
    "    adata_test = sc.read_h5ad(f\"../datasets/st_data/gold/starmap_plus_mouse_batch{idx}.h5ad\")\n",
    "    adata_test = adata_test[:, [gene for gene in adata_test.var.index.tolist() if gene.upper() in genes_intersect]]\n",
    "    adata_test.obs[\"batch\"] = f\"batch{idx}\"\n",
    "    adata_test.var.index = selected_genes\n",
    "    adata_test.write(f\"../datasets/st_data/gold/starmap_plus_mouse_batch{idx}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5532080a-b064-4d97-83fa-70b466d9371d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "# batch, dataset, experimental_batch (?)\n",
    "adata = sc.read_h5ad(\"../datasets/st_data/gold/starmap_plus_mouse_alzheimers.h5ad\")\n",
    "adata.obs.rename(columns={'batch': 'experimental_batch'}, inplace=True)\n",
    "adata.obs[\"dataset\"] = \"alzheimers\"\n",
    "adata.layers[\"counts\"] = adata.X.copy()\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "for idx, donor in enumerate(adata.obs[\"donor_id\"].unique().tolist()):\n",
    "    adata_donor = adata[adata.obs[\"donor_id\"] == donor]\n",
    "    adata_donor.obs[\"batch\"] = f\"batch{idx+1}\"\n",
    "    adata_donor.write(f\"../datasets/st_data/gold/starmap_plus_mouse_alzheimers_batch{idx+1}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef7e6b23-2510-4fa5-ba14-9885d8f3aea0",
   "metadata": {},
   "outputs": [],
   "source": [
    "0.0625 / 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f628a19-b96d-4f04-abbb-7b7418b6529f",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse\"\n",
    "job_id = 26\n",
    "#reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "#                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "#                    \" batch19 batch20 batch21 batch22 batch23 batch24 batch25 batch26 batch27 batch28\"\n",
    "reference_batches = \"batch10 batch15 batch21 batch22 batch23\"\n",
    "n_neighbors = 8\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "edge_batch_size = 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"dataset\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"50\"\n",
    "lambda_cat_covariates_contrastive = 300.\n",
    "contrastive_logits_pos_ratio = 0.015625\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 100.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.01\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method one-hop-norm\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 400\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37e166f-455e-4092-89f6-2cde8139eecc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11bdc0b5-5a82-4786-b53f-2e13cbe4f42c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" \\\n",
    "                    \" batch9 batch10 batch11 batch12 batch13 batch14 batch15 batch16 batch17 batch18\" \\\n",
    "                    \" batch19 batch20\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 2048\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"20\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 35\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd09c068-30ba-4237-a53d-c841dcbba9d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"starmap_plus_mouse_cns\"\n",
    "job_id = 2\n",
    "reference_batches = \"batch1 batch2 batch3\"\n",
    "n_neighbors = 8 # 8, 12\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 2048\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"3\"\n",
    "lambda_cat_covariates_contrastive = 0. # 0., 10000, 100000, 500000\n",
    "contrastive_logits_pos_ratio = 0. # 0., 0.125, 0.0625, 0.03125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fde3a4b-fd6f-43ba-9f37-436885f328df",
   "metadata": {},
   "source": [
    "### 5.3 nanoString CosMx SMI Human Non-Small-Cell Lung Cancer (NSCLC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463fcdec-a4d3-4d97-8f6c-6d4a9251158b",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 5\n",
    "reference_batches = \"batch1 batch2 batch3 batch4 batch5 batch6 batch7 batch8\" # leave out: batch3, batch8\n",
    "n_neighbors = 9\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "node_label_method = \"one-hop-norm\"\n",
    "edge_batch_size = 512 # 4096, 2048, 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\" # \"encoder gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch fov patient\"\n",
    "cat_covariates_no_edges = \"True False True\"\n",
    "cat_covariates_embeds_nums = \"16 60 10\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0. # 0.015625, 0.0078125\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 300.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --n_addon_gp 100\" \\\n",
    "              \" --active_gp_thresh_ratio 0.01\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 400\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 50000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 30000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              \" --lambda_l1_addon 300.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7380bdd-79ba-48a7-aeaa-ba20b8978edd",
   "metadata": {},
   "source": [
    "### 5.5 Spatial ATAC-RNA-Seq Mouse Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "916e9324-c9de-464a-b7ea-b1ed81980b92",
   "metadata": {},
   "outputs": [],
   "source": [
    "job_id = 1\n",
    "\n",
    "task = \"reference\"\n",
    "dataset = \"spatial_atac_rna_seq_mouse_brain_batch2\"\n",
    "reference_batches = \"None\"\n",
    "species = \"mouse\"\n",
    "lambda_edge_recon = 5000000.\n",
    "lambda_gene_expr_recon = 3000.\n",
    "lambda_chrom_access_recon = 1000.\n",
    "active_gp_thresh_ratio = 0.03\n",
    "lambda_l1_masked = 300.\n",
    "lambda_l1_addon = 300.\n",
    "n_svg = 3000\n",
    "n_svp = 15000\n",
    "n_neighbors = 8\n",
    "nichenet_max_n_target_genes_per_gp = 250\n",
    "node_label_method = \"one-hop-norm\"\n",
    "conv_layer_encoder = \"gatv2conv\"\n",
    "edge_batch_size = \"256\"\n",
    "n_sampled_neighbors = 4\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "cat_covariates_embeds_injection = \"None\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = 0\n",
    "n_hvg = 0\n",
    "n_addon_gp = 100\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --filter_genes\" \\\n",
    "              f\" --n_hvg {n_hvg}\" \\\n",
    "              f\" --n_svg {n_svg}\" \\\n",
    "              f\" --n_svp {n_svp}\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              f\" --nichenet_max_n_target_genes_per_gp {nichenet_max_n_target_genes_per_gp}\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              \" --include_collectri_gps\" \\\n",
    "              \" --include_brain_marker_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              \" --condition_key batch\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              \" --include_atac_modality\" \\\n",
    "              \" --filter_peaks\" \\\n",
    "              \" --min_cell_peak_thresh_ratio 0.005\" \\\n",
    "              \" --min_cell_gene_thresh_ratio 0.005\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              f\" --n_addon_gp {n_addon_gp}\" \\\n",
    "              f\" --active_gp_thresh_ratio {active_gp_thresh_ratio}\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method {node_label_method}\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              f\" --conv_layer_encoder {conv_layer_encoder}\" \\\n",
    "              \" --n_epochs 400\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              f\" --lambda_edge_recon {lambda_edge_recon}\" \\\n",
    "              f\" --lambda_gene_expr_recon {lambda_gene_expr_recon}\" \\\n",
    "              f\" --lambda_chrom_access_recon {lambda_chrom_access_recon}\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --lambda_l1_addon {lambda_l1_addon}.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)\n",
    "\n",
    "job_id += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01341dcb-64ad-45bd-8ef6-939e77d5b0f8",
   "metadata": {},
   "source": [
    "### 5.6 Xenium Human Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c5b999-11e0-42e5-b886-8fb3a47d82f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"xenium_human_breast_cancer\"\n",
    "job_id = 1\n",
    "reference_batches = \"batch1 batch2\"\n",
    "n_neighbors = 8\n",
    "n_sampled_neighbors = 4\n",
    "species = \"human\"\n",
    "edge_batch_size = 512\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"batch\"\n",
    "cat_covariates_no_edges = \"True\"\n",
    "cat_covariates_embeds_nums = \"2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 100.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.01\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method one-hop-norm\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gatv2conv\" \\\n",
    "              \" --n_epochs 400\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23c40603-6658-4e3e-926c-9b6402b58fba",
   "metadata": {},
   "source": [
    "### 5.8 MERFISH Mouse Brain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0caa9674-1d77-4981-a258-79a5ea1b95b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference\"\n",
    "dataset = \"merfish_mouse_brain\"\n",
    "job_id = 1\n",
    "reference_batches = \" \".join([f\"batch{i}\" for i in range(1, 240)])\n",
    "n_neighbors = 4\n",
    "n_sampled_neighbors = 4\n",
    "species = \"mouse\"\n",
    "edge_batch_size = 1024 # 1024 #4096 # 1024\n",
    "cat_covariates_embeds_injection = \"gene_expr_decoder\"\n",
    "cat_covariates_keys = \"donor_id batch\"\n",
    "cat_covariates_no_edges = \"True True\"\n",
    "cat_covariates_embeds_nums = \"4 239\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "lambda_l1_addon = 0.\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"train_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --reference_batches {reference_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --no-filter_genes\" \\\n",
    "              \" --nichenet_keep_target_genes_ratio 1.0\" \\\n",
    "              \" --nichenet_max_n_target_genes_per_gp 250\" \\\n",
    "              \" --include_mebocost_gps\" \\\n",
    "              f\" --species {species}\" \\\n",
    "              \" --gp_filter_mode subset\" \\\n",
    "              \" --combine_overlap_gps\" \\\n",
    "              \" --overlap_thresh_source_genes 0.9\" \\\n",
    "              \" --overlap_thresh_target_genes 0.9\" \\\n",
    "              \" --overlap_thresh_genes 0.9\" \\\n",
    "              \" --counts_key counts\" \\\n",
    "              f\" --cat_covariates_keys {cat_covariates_keys}\" \\\n",
    "              f\" --cat_covariates_no_edges {cat_covariates_no_edges}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --adj_key spatial_connectivities\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_targets_mask_key nichecompass_gp_targets\" \\\n",
    "              \" --gp_sources_mask_key nichecompass_gp_sources\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --model_label {task}\" \\\n",
    "              \" --active_gp_names_key nichecompass_active_gp_names\" \\\n",
    "              \" --latent_key nichecompass_latent\" \\\n",
    "              \" --active_gp_thresh_ratio 0.01\" \\\n",
    "              \" --gene_expr_recon_dist nb\" \\\n",
    "              f\" --cat_covariates_embeds_injection {cat_covariates_embeds_injection}\" \\\n",
    "              f\" --cat_covariates_embeds_nums {cat_covariates_embeds_nums}\" \\\n",
    "              \" --log_variational\" \\\n",
    "              f\" --node_label_method one-hop-norm\" \\\n",
    "              \" --n_layers_encoder 1\" \\\n",
    "              \" --n_hidden_encoder None\" \\\n",
    "              \" --conv_layer_encoder gcnconv\" \\\n",
    "              \" --n_epochs 8\" \\\n",
    "              \" --n_epochs_all_gps 2\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              \" --edge_val_ratio 0.01\" \\\n",
    "              \" --node_val_ratio 0.01\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "              f\" --no-mlflow_tracking\" \\\n",
    "              f\" --no-compute_knn_graph\" \\\n",
    "              f\" --timestamp_suffix _{job_id}\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "997a31b2-7af9-4c73-96b0-a18d62ec6509",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"merfish_mouse_brain\"\n",
    "task = \"postprocessing\"\n",
    "model_label = \"reference\"\n",
    "load_timestamp = \"24022024_170907_82\" # \"21022024_194703_55\" # \"19022024_170806_46\" # \"21022024_194703_56\"\n",
    "job_id = 7\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"postprocess_nichecompass_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --model_label {model_label}\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              \" --compute_latent\" \\\n",
    "              \" --compute_pca\" \\\n",
    "              \" --compute_knn_graph\" \\\n",
    "              \" --compute_umap\" \\\n",
    "              \" --compute_leiden\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        p=\"cpu_p\",\n",
    "        gres=\"cpu:1\",\n",
    "        qos=\"cpu_normal\",\n",
    "        mem=\"300G\",\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c040bc3-bc03-4b1d-aa0c-1e3fc83c1f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset = \"merfish_mouse_brain\"\n",
    "task = \"postprocessing\"\n",
    "model_label = \"reference\"\n",
    "load_timestamp = \"21022024_194703_55\" # \"21022024_194703_55\" # \"19022024_170806_46\" # \"21022024_194703_56\"\n",
    "job_id = 9\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"postprocess_nichecompass_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --model_label {model_label}\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              \" --compute_umap\" \\\n",
    "              \" --no-compute_leiden\"\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        p=\"cpu_p\",\n",
    "        gres=\"cpu:1\",\n",
    "        qos=\"cpu_normal\",\n",
    "        mem=\"300G\",\n",
    "        nice=10000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001945c5-6d59-4bdb-9efe-d8bbb33edc33",
   "metadata": {},
   "source": [
    "## 6. NicheCompass Query Mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f54d6013-72d5-4b11-8a85-e5d52c3e7ab2",
   "metadata": {},
   "source": [
    "### 6.1 nanoString CosMx Human NSCLC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caf1ee21-4e25-449a-823e-d5ad9ef07c61",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"nanostring_cosmx_human_nsclc\"\n",
    "job_id = 105\n",
    "query_batches = \"batch3 batch8\"\n",
    "n_neighbors = 4\n",
    "n_sampled_neighbors = 4\n",
    "load_timestamp = \"25092023_232129_105\"  #\n",
    "lambda_cat_covariates_contrastive = 1000000.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 0.\n",
    "edge_batch_size = 512\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}_query_mapping\"\n",
    "job_folder_path = f\"../scripts/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/scripts/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label reference_query_mapping\" \\\n",
    "              \" --n_epochs 400\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "              f\" --n_sampled_neighbors {n_sampled_neighbors}\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bf6b4d-34b5-4324-8deb-095bdfa7d0e2",
   "metadata": {},
   "source": [
    "### 6.1 seqFISH Mouse Organogenesis Imputed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17761bc0-3178-42dd-9b0e-3935be8d2a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 12\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"01072023_165203_1\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "lambda_l1_masked = 5.\n",
    "edge_batch_size = 4096\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{task}\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 500000.\" \\\n",
    "              \" --lambda_gene_expr_recon 300.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked {lambda_l1_masked}\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args,\n",
    "        t=\"12:00:00\",\n",
    "        p=\"interactive_gpu_p\",\n",
    "        gres=\"gpu:1\",\n",
    "        qos=\"interactive_gpu\",\n",
    "        nice=9999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8bc239f-54d5-4911-a2e9-ddcd63f95bd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "query_batches = \"batch3 batch4\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"01082024_191835_2\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0d88209-2847-4f03-8e2b-867b0fb02faa",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 1\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"01082024_191836_3\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de92ab81-64f8-41d4-a05c-704b20e95116",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 10\n",
    "query_batches = \"batch1 batch2 batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"07082024_124745_8\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "582b51ad-1f25-4d66-8fe6-982870a7a9b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 11\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"07082024_124745_7\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --dataset {dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f33f9b2-22a7-4f82-8c69-dddbec4e610e",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_50pct\"\n",
    "query_dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 12\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"08082024_095147_12\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --reference_dataset {dataset}\" \\\n",
    "              f\" --query_dataset {query_dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "587de2e7-4430-4e7b-b2f8-77550af92379",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_25pct\"\n",
    "query_dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 13\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"08082024_095147_13\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --reference_dataset {dataset}\" \\\n",
    "              f\" --query_dataset {query_dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba223cff-ae0a-485b-92e7-e9b9b19f762c",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_10pct\"\n",
    "query_dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 14\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"08082024_140042_14\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --reference_dataset {dataset}\" \\\n",
    "              f\" --query_dataset {query_dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e68b03e-acbf-4aa9-8f2f-5b24e1edcf9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_5pct\"\n",
    "query_dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 15\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"08082024_140042_15\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --reference_dataset {dataset}\" \\\n",
    "              f\" --query_dataset {query_dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42af9975-07c3-4caa-ab10-68d476354312",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = \"reference_query\"\n",
    "dataset = \"seqfish_mouse_organogenesis_imputed_subsample_1pct\"\n",
    "query_dataset = \"seqfish_mouse_organogenesis_imputed\"\n",
    "job_id = 15\n",
    "query_batches = \"batch5 batch6\"\n",
    "n_neighbors = 8\n",
    "node_label_method = \"one-hop-norm\"\n",
    "load_timestamp = \"08082024_140042_16\"\n",
    "lambda_cat_covariates_contrastive = 0.\n",
    "contrastive_logits_pos_ratio = 0.\n",
    "contrastive_logits_neg_ratio = 0.\n",
    "edge_batch_size = 256\n",
    "\n",
    "job_name_prefix = f\"{dataset}_nichecompass_{node_label_method}_{task}_query_mapping\"\n",
    "job_folder_path = f\"./data_analysis/{task}/slurm_jobs\"\n",
    "script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/data_analysis/{task}\"\n",
    "script_name = \"map_query_on_nichecompass_reference_model.py\"\n",
    "script_args = f\" --reference_dataset {dataset}\" \\\n",
    "              f\" --query_dataset {query_dataset}\" \\\n",
    "              f\" --query_batches {query_batches}\" \\\n",
    "              f\" --n_neighbors {n_neighbors}\" \\\n",
    "              \" --spatial_key spatial\" \\\n",
    "              \" --mapping_entity_key mapping_entity\" \\\n",
    "              \" --gp_names_key nichecompass_gp_names\" \\\n",
    "              f\" --reference_model_label reference\" \\\n",
    "              f\" --load_timestamp {load_timestamp}\" \\\n",
    "              f\" --query_model_label query\" \\\n",
    "              f\" --reference_query_model_label {task}\" \\\n",
    "              \" --n_epochs 100\" \\\n",
    "              \" --n_epochs_all_gps 25\" \\\n",
    "              \" --n_epochs_no_cat_covariates_contrastive 0\" \\\n",
    "              \" --lr 0.001\" \\\n",
    "              \" --lambda_edge_recon 5000000.\" \\\n",
    "              \" --lambda_gene_expr_recon 3000.\" \\\n",
    "              f\" --lambda_cat_covariates_contrastive {lambda_cat_covariates_contrastive}\" \\\n",
    "              f\" --contrastive_logits_pos_ratio {contrastive_logits_pos_ratio}\" \\\n",
    "              f\" --contrastive_logits_neg_ratio {contrastive_logits_neg_ratio}\" \\\n",
    "              \" --lambda_group_lasso 0.\" \\\n",
    "              f\" --lambda_l1_masked 0.\" \\\n",
    "              f\" --edge_batch_size {edge_batch_size}\" \\\n",
    "              \" --node_batch_size None\" \\\n",
    "\n",
    "submit_python_script(\n",
    "        job_name_prefix=job_name_prefix,\n",
    "        job_id=job_id,\n",
    "        job_folder_path=job_folder_path,\n",
    "        conda_env_name=conda_env_name,\n",
    "        script_folder_path=script_folder_path,\n",
    "        script_name=script_name,\n",
    "        script_args=script_args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c96f5a04-ca93-42ed-a536-bc14e18ed630",
   "metadata": {},
   "outputs": [],
   "source": [
    "method = \"nichecompass\"\n",
    "task = \"single_sample_method_benchmarking\"\n",
    "cell_type_key = \"celltype_mapped_refined\"\n",
    "batch_key = \"None\"\n",
    "metrics = \"gcs mlami cas clisis nasw cnmi cari casw clisi\"\n",
    "\n",
    "job_id = 1\n",
    "for dataset in [\"seqfish_mouse_organogenesis\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_50pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_25pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_10pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_5pct\",\n",
    "                \"seqfish_mouse_organogenesis_subsample_1pct\"\n",
    "               ]:\n",
    "    file_name = f\"{dataset}_{method}.h5ad\"\n",
    "\n",
    "    job_name_prefix = f\"{method}_single_sample_method_benchmarking_{dataset}_metrics_computation\"\n",
    "    job_folder_path = f\"./data_analysis/reference_query/slurm_jobs\"\n",
    "    script_folder_path = f\"/home/aih/sebastian.birk/workspace/projects/nichecompass-reproducibility/analysis/benchmarking/{task}\"\n",
    "    script_name = \"compute_benchmarking_metrics.py\"\n",
    "    script_args = f\" --dataset {dataset}\" \\\n",
    "                  f\" --task {task}\" \\\n",
    "                  f\" --file_name {file_name}\" \\\n",
    "                  f\" --cell_type_key {cell_type_key}\" \\\n",
    "                  f\" --batch_key {batch_key}\" \\\n",
    "                  f\" --latent_key {method}_latent\" \\\n",
    "                  f\" --metrics {metrics}\"\n",
    "\n",
    "    submit_python_script(\n",
    "            job_name_prefix=job_name_prefix,\n",
    "            job_id=job_id,\n",
    "            job_folder_path=job_folder_path,\n",
    "            conda_env_name=conda_env_name,\n",
    "            script_folder_path=script_folder_path,\n",
    "            script_name=script_name,\n",
    "            script_args=script_args)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "toc-showcode": false,
  "toc-showmarkdowntxt": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
