{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# Simulation Model Ablation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>)\n",
    "- **Date of Creation:** 26.06.2024\n",
    "- **Date of Last Modification:** 28.07.2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94c7551a-013a-444f-8e2c-7d94290d6388",
   "metadata": {},
   "source": [
    "Run this notebook in the nichecompass-reproducibility environment, installable from ```('../../../envs/environment.yaml')```.\n",
    "\n",
    "Before running this notebook:\n",
    "- Run the data simulation notebook at ```('../data_simulation/data_simulation.ipynb')```\n",
    "- Clone SDMBench from https://github.com/zhaofangyuan98/SDMBench/tree/main/SDMBench into ```('../benchmarking')``` (some slight modifications to the SDMBench source code were necessary to remove technical bugs)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7c756075-c6ff-4c0b-8bd4-20885ca56b9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "35b00887-6603-4c0a-a834-f427d7610b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../benchmarking/SDMBench/SDMBench\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Failed to download from `https://omnipathdb.org/`.\n",
      "WARNING:root:Traceback (most recent call last):\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 715, in urlopen\n",
      "    httplib_response = self._make_request(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 404, in _make_request\n",
      "    self._validate_conn(conn)\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 1060, in _validate_conn\n",
      "    conn.connect()\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/connection.py\", line 419, in connect\n",
      "    self.sock = ssl_wrap_socket(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 449, in ssl_wrap_socket\n",
      "    ssl_sock = _ssl_wrap_socket_impl(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/util/ssl_.py\", line 493, in _ssl_wrap_socket_impl\n",
      "    return ssl_context.wrap_socket(sock, server_hostname=server_hostname)\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/ssl.py\", line 501, in wrap_socket\n",
      "    return self.sslsocket_class._create(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/ssl.py\", line 1074, in _create\n",
      "    self.do_handshake()\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/ssl.py\", line 1343, in do_handshake\n",
      "    self._sslobj.do_handshake()\n",
      "ssl.SSLZeroReturnError: TLS/SSL connection has been closed (EOF) (_ssl.c:1133)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/requests/adapters.py\", line 667, in send\n",
      "    resp = conn.urlopen(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/connectionpool.py\", line 801, in urlopen\n",
      "    retries = retries.increment(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/urllib3/util/retry.py\", line 594, in increment\n",
      "    raise MaxRetryError(_pool, url, error or ResponseError(cause))\n",
      "urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='omnipathdb.org', port=443): Max retries exceeded with url: /queries/enzsub?format=json (Caused by SSLError(SSLZeroReturnError(6, 'TLS/SSL connection has been closed (EOF) (_ssl.c:1133)')))\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/omnipath/_core/downloader/_downloader.py\", line 143, in maybe_download\n",
      "    res = self._download(req)\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/omnipath/_core/downloader/_downloader.py\", line 178, in _download\n",
      "    with self._session.send(\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/requests/sessions.py\", line 703, in send\n",
      "    r = adapter.send(request, **kwargs)\n",
      "  File \"/home/aih/sebastian.birk/miniconda3/envs/nichecompass-reproducibility/lib/python3.9/site-packages/requests/adapters.py\", line 698, in send\n",
      "    raise SSLError(e, request=request)\n",
      "requests.exceptions.SSLError: HTTPSConnectionPool(host='omnipathdb.org', port=443): Max retries exceeded with url: /queries/enzsub?format=json (Caused by SSLError(SSLZeroReturnError(6, 'TLS/SSL connection has been closed (EOF) (_ssl.c:1133)')))\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "from matplotlib import gridspec\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                                compute_communication_gp_network,\n",
    "                                visualize_communication_gp_network,\n",
    "                                create_new_color_dict,\n",
    "                                extract_gp_dict_from_mebocost_ms_interactions,\n",
    "                                extract_gp_dict_from_nichenet_lrt_interactions,\n",
    "                                extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                                filter_and_combine_gp_dict_gps_v2,\n",
    "                                generate_enriched_gp_info_plots)\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "from SDMBench import sdmbench"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {},
   "source": [
    "### 1.2 Define Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Dataset (Fixed) ###\n",
    "species = \"mouse\"\n",
    "spatial_key = \"spatial\"\n",
    "niche_type_key = \"niche_types\"\n",
    "sample_key = \"batch\"\n",
    "\n",
    "### Model ###\n",
    "# AnnData Keys\n",
    "counts_key = \"counts\"\n",
    "adj_key = \"spatial_connectivities\"\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "active_gp_names_key = \"nichecompass_active_gp_names\"\n",
    "gp_targets_mask_key = \"nichecompass_gp_targets\"\n",
    "gp_targets_categories_mask_key = \"nichecompass_gp_targets_categories\"\n",
    "gp_sources_mask_key = \"nichecompass_gp_sources\"\n",
    "gp_sources_categories_mask_key = \"nichecompass_gp_sources_categories\"\n",
    "latent_key = \"nichecompass_latent\"\n",
    "\n",
    "# Trainer\n",
    "n_epochs = 400\n",
    "n_epochs_all_gps = 25\n",
    "lr = 0.001\n",
    "edge_batch_size = 1024\n",
    "use_cuda_if_available = True\n",
    "\n",
    "### Evaluation ###\n",
    "spot_size = 40\n",
    "differential_gp_test_results_key = \"nichecompass_differential_gp_test_results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "efefff3d-4666-4ac4-a0cb-43159dff5823",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Dataset (Variable) ###\n",
    "n_sim = 1\n",
    "n_genes = 1105\n",
    "n_locs = 10000\n",
    "increment_mode = \"strong\"\n",
    "dataset_name = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "05caf2ee-887a-4429-aed6-9a89a22e7062",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\")\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ab22fbaf-aeed-464b-8a20-b640b9e055fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get time of notebook execution for timestamping saved artifacts\n",
    "now = datetime.now()\n",
    "current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths\n",
    "ga_data_folder_path = \"../../datasets/ga_data\"\n",
    "gp_data_folder_path = \"../../datasets/gp_data\"\n",
    "st_data_folder_path = \"../../datasets/st_data\"\n",
    "st_data_gold_folder_path = st_data_folder_path + \"/gold\"\n",
    "omnipath_lr_network_file_path = f\"{gp_data_folder_path}/omnipath_lr_network.csv\"\n",
    "nichenet_lr_network_file_path = f\"{gp_data_folder_path}/nichenet_lr_network_v2_{species}.csv\"\n",
    "nichenet_ligand_target_matrix_file_path = f\"{gp_data_folder_path}/nichenet_ligand_target_matrix_v2_{species}.csv\"\n",
    "mebocost_enzyme_sensor_interactions_folder_path = f\"{gp_data_folder_path}/metabolite_enzyme_sensor_gps\"\n",
    "gene_orthologs_mapping_file_path = f\"{ga_data_folder_path}/human_mouse_gene_orthologs.csv\"\n",
    "artifacts_folder_path = f\"../../artifacts\"\n",
    "model_folder_path = f\"{artifacts_folder_path}/ablation/{current_timestamp}/model\"\n",
    "figure_folder_path = f\"{artifacts_folder_path}/ablation/{current_timestamp}/figures\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756c2e46-6815-4f82-addf-5bfbed664175",
   "metadata": {},
   "source": [
    "### 1.5 Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "69c9f8f1-d15a-4cc9-b54c-35b14b0b43c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(model_folder_path, exist_ok=True)\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(f\"{artifacts_folder_path}/ablation\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83a4e030-3928-4da7-b321-1c74d9a5239c",
   "metadata": {},
   "source": [
    "### 1.6 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b98333b3-5d8c-463f-9b52-3760755fd223",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_gps(species,\n",
    "                omnipath_lr_network_file_path,\n",
    "                nichenet_lr_network_file_path,\n",
    "                nichenet_ligand_target_matrix_file_path,\n",
    "                gene_orthologs_mapping_file_path):\n",
    "    # Retrieve OmniPath GPs (source: ligand genes; target: receptor genes)\n",
    "    omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "        species=species,\n",
    "        load_from_disk=True,\n",
    "        save_to_disk=False,\n",
    "        lr_network_file_path=omnipath_lr_network_file_path,\n",
    "        gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "        plot_gp_gene_count_distributions=False)\n",
    "    \n",
    "    # Retrieve NicheNet GPs (source: ligand genes; target: receptor genes, target genes)\n",
    "    nichenet_gp_dict = extract_gp_dict_from_nichenet_lrt_interactions(\n",
    "        species=species,\n",
    "        version=\"v2\",\n",
    "        load_from_disk=True,\n",
    "        save_to_disk=False,\n",
    "        lr_network_file_path=nichenet_lr_network_file_path,\n",
    "        ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "        gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "        plot_gp_gene_count_distributions=False)\n",
    "    \n",
    "    # Retrieve MEBOCOST GPs (source: enzyme genes; target: sensor genes)\n",
    "    mebocost_gp_dict = extract_gp_dict_from_mebocost_ms_interactions(\n",
    "        dir_path=mebocost_enzyme_sensor_interactions_folder_path,\n",
    "        species=species,\n",
    "        plot_gp_gene_count_distributions=False)\n",
    "    \n",
    "    # Filter and combine GPs\n",
    "    gp_dicts = [omnipath_gp_dict, nichenet_gp_dict, mebocost_gp_dict]\n",
    "    combined_gp_dict = filter_and_combine_gp_dict_gps_v2(\n",
    "        gp_dicts,\n",
    "        verbose=True)\n",
    "    \n",
    "    print(f\"Number of gene programs after filtering and combining: \"\n",
    "          f\"{len(combined_gp_dict)}.\")\n",
    "\n",
    "    return combined_gp_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f2fef25c-709f-4cc5-b234-98f7c0c4f3e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data(combined_gp_dict,\n",
    "              run_dict):\n",
    "    # Load adata\n",
    "    adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/{dataset_name}.h5ad\")\n",
    "    \n",
    "    # Compute spatial neighborhood\n",
    "    sq.gr.spatial_neighbors(adata,\n",
    "                            coord_type=\"generic\",\n",
    "                            spatial_key=spatial_key,\n",
    "                            n_neighs=run_dict[\"n_neighbors\"][-1])\n",
    "    \n",
    "    # Make adjacency matrix symmetric\n",
    "    adata.obsp[adj_key] = (\n",
    "        adata.obsp[adj_key].maximum(\n",
    "            adata.obsp[adj_key].T))\n",
    "    \n",
    "    #combined_gp_dict = combined_gp_dict.copy()\n",
    "    #for gp in combined_gp_dict.keys():\n",
    "    #    if gp not in simulation_gps:\n",
    "    #        del(combined_gp_dict[gp])\n",
    "    \n",
    "    # Add the GP dictionary as binary masks to the adata\n",
    "    add_gps_from_gp_dict_to_adata(\n",
    "        gp_dict=combined_gp_dict,\n",
    "        adata=adata,\n",
    "        gp_targets_mask_key=gp_targets_mask_key,\n",
    "        gp_targets_categories_mask_key=gp_targets_categories_mask_key,\n",
    "        gp_sources_mask_key=gp_sources_mask_key,\n",
    "        gp_sources_categories_mask_key=gp_sources_categories_mask_key,\n",
    "        gp_names_key=gp_names_key,\n",
    "        min_genes_per_gp=2,\n",
    "        min_source_genes_per_gp=1,\n",
    "        min_target_genes_per_gp=1,\n",
    "        max_genes_per_gp=None,\n",
    "        max_source_genes_per_gp=None,\n",
    "        max_target_genes_per_gp=None)\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9d08320a-547a-4743-bd25-a11c41f16714",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(adata,\n",
    "                run_dict,\n",
    "                mlflow_experiment_id):\n",
    "    # Initialize model\n",
    "    model = NicheCompass(adata,\n",
    "                         counts_key=counts_key,\n",
    "                         adj_key=adj_key,\n",
    "                         gp_names_key=gp_names_key,\n",
    "                         active_gp_names_key=active_gp_names_key,\n",
    "                         gp_targets_mask_key=gp_targets_mask_key,\n",
    "                         gp_targets_categories_mask_key=gp_targets_categories_mask_key,\n",
    "                         gp_sources_mask_key=gp_sources_mask_key,\n",
    "                         gp_sources_categories_mask_key=gp_sources_categories_mask_key,\n",
    "                         latent_key=latent_key,\n",
    "                         n_addon_gp=run_dict[\"n_addon_gp\"][-1],\n",
    "                         conv_layer_encoder=run_dict[\"conv_layer_encoder\"][-1],\n",
    "                         active_gp_thresh_ratio=run_dict[\"active_gp_thresh_ratio\"][-1],\n",
    "                         seed=run_dict[\"model_seed\"][-1])\n",
    "    \n",
    "    # Train model\n",
    "    model.train(n_epochs=n_epochs,\n",
    "                n_epochs_all_gps=n_epochs_all_gps,\n",
    "                lr=lr,\n",
    "                lambda_edge_recon=run_dict[\"lambda_edge_recon\"][-1],\n",
    "                lambda_gene_expr_recon=run_dict[\"lambda_gene_expr_recon\"][-1],\n",
    "                lambda_l1_masked=run_dict[\"lambda_l1_masked\"][-1],\n",
    "                edge_batch_size=edge_batch_size,\n",
    "                mlflow_experiment_id=mlflow_experiment_id,\n",
    "                n_sampled_neighbors=run_dict[\"n_neighbors\"][-1],\n",
    "                use_cuda_if_available=use_cuda_if_available,\n",
    "                seed=run_dict[\"model_seed\"][-1],\n",
    "                verbose=True)\n",
    "    \n",
    "    # Compute latent neighbor graph\n",
    "    sc.pp.neighbors(model.adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    \n",
    "    # Compute UMAP embedding\n",
    "    sc.tl.umap(model.adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Save trained model\n",
    "    os.makedirs(model_folder_path + f\"/run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\", exist_ok=True)\n",
    "    model.save(dir_path=model_folder_path + f\"/run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\",\n",
    "               overwrite=True,\n",
    "               save_adata=True,\n",
    "               adata_file_name=\"adata.h5ad\")\n",
    "\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ed936ca4-3060-4b13-b665-ace0cde89b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_niches(model,\n",
    "                    run_dict):\n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    latent_leiden_resolution = 0.3\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=model.adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = model.adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= 100].index\n",
    "        n_niches = model.adata.obs[model.adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            latent_leiden_resolution += 0.1\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            latent_leiden_resolution += 0.01\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            latent_leiden_resolution -= 0.1\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            latent_leiden_resolution -= 0.01\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "    \n",
    "    latent_cluster_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "        cat_key=\"pred_niche_types\")\n",
    "    \n",
    "    # Create plot of latent cluster / niche annotations in physical and latent space\n",
    "    groups = None # set this to a specific cluster for easy visualization\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/\" \\\n",
    "                f\"{ablation}_{dataset_name}_run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\" \\\n",
    "                \"niches_latent_physical_space.svg\"\n",
    "    \n",
    "    model.adata.obs[sample_key] = \"batch1\"\n",
    "    samples = model.adata.obs[sample_key].unique().tolist()\n",
    "    \n",
    "    fig = plt.figure(figsize=(12, 14))\n",
    "    title = fig.suptitle(t=f\"NicheCompass Niches \" \\\n",
    "                           \"in Latent and Physical Space\",\n",
    "                         y=0.96,\n",
    "                         x=0.55,\n",
    "                         fontsize=20)\n",
    "    spec1 = gridspec.GridSpec(ncols=1,\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1],\n",
    "                              height_ratios=[3, 2])\n",
    "    spec2 = gridspec.GridSpec(ncols=len(samples),\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1] * len(samples),\n",
    "                              height_ratios=[3, 2])\n",
    "    axs = []\n",
    "    axs.append(fig.add_subplot(spec1[0]))\n",
    "    sc.pl.umap(adata=model.adata,\n",
    "               color=[\"pred_niche_types\"],\n",
    "               groups=groups,\n",
    "               palette=latent_cluster_colors,\n",
    "               title=f\"Niches in Latent Space\",\n",
    "               ax=axs[0],\n",
    "               show=False)\n",
    "    for idx, sample in enumerate(samples):\n",
    "        axs.append(fig.add_subplot(spec2[len(samples) + idx]))\n",
    "        sc.pl.spatial(adata=model.adata[model.adata.obs[sample_key] == sample],\n",
    "                      color=[\"pred_niche_types\"],\n",
    "                      groups=groups,\n",
    "                      palette=latent_cluster_colors,\n",
    "                      spot_size=spot_size,\n",
    "                      title=f\"Niches in Physical Space \\n\"\n",
    "                            f\"(Sample: {sample})\",\n",
    "                      legend_loc=None,\n",
    "                      ax=axs[idx+1],\n",
    "                      show=False)\n",
    "    \n",
    "    # Create and position shared legend\n",
    "    handles, labels = axs[0].get_legend_handles_labels()\n",
    "    lgd = fig.legend(handles,\n",
    "                     labels,\n",
    "                     loc=\"center left\",\n",
    "                     bbox_to_anchor=(0.98, 0.5))\n",
    "    axs[0].get_legend().remove()\n",
    "    \n",
    "    # Adjust, save and display plot\n",
    "    plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "    if save_fig:\n",
    "        fig.savefig(file_path,\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "37e37c34-0591-406e-98dd-0a94559b00ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enriched_gps(model):\n",
    "    # Run differential GP testing: each niche vs rest\n",
    "    selected_cats = None\n",
    "    comparison_cats = \"rest\"\n",
    "    enriched_gps = model.run_differential_gp_tests(\n",
    "        cat_key=\"niche_types\",\n",
    "        selected_cats=selected_cats,\n",
    "        comparison_cats=comparison_cats,\n",
    "        log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "    # Get enriched prior and de novo GPs\n",
    "    gp_summary_df = model.get_gp_summary() \n",
    "    enriched_gps = model.adata.uns[differential_gp_test_results_key]\n",
    "    prior_enriched_gps = enriched_gps[~enriched_gps[\"gene_program\"].str.contains(\"Add-on\")]\n",
    "    denovo_enriched_gp = enriched_gps[enriched_gps[\"gene_program\"].str.contains(\"Add-on\")]\n",
    "    \n",
    "    # Get enriched prior gene programs for each niche according to differential gene program test\n",
    "    enriched_prior_gps_per_niche = prior_enriched_gps.groupby(\n",
    "        \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "    \n",
    "    # Get enriched de-novo gene programs for each niche according to differential gene program test\n",
    "    enriched_denovo_gps_per_niche = denovo_enriched_gp.groupby(\n",
    "        \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "\n",
    "    return enriched_prior_gps_per_niche, enriched_denovo_gps_per_niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "57d43ed3-d261-4e25-ab38-83433403c3fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define eval helper functions\n",
    "def calculate_intersection(row, col1, col2):\n",
    "    return len(row[col1].intersection(row[col2]))\n",
    "\n",
    "def calculate_surplus(row, col1, col2):\n",
    "    return row[col1] - row[col2]\n",
    "\n",
    "def jaccard_index_elementwise(sets1, sets2):\n",
    "    def jaccard_index(set1, set2):\n",
    "        intersection = len(set1 & set2)\n",
    "        union = len(set1 | set2)\n",
    "        if union == 0:\n",
    "            return 0.0\n",
    "        return intersection / union\n",
    "    \n",
    "    return np.array([jaccard_index(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "def precision_score_elementwise(sets1, sets2):\n",
    "    def precision_score(set1, set2):\n",
    "        tp = len(set1 & set2)  # true positives\n",
    "        fp = len(set1 - set2)  # false positives\n",
    "        \n",
    "        if tp == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        precision = tp / (tp + fp) if (tp + fp) > 0 else 0\n",
    "        \n",
    "        return precision\n",
    "    \n",
    "    return np.array([precision_score(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "def recall_score_elementwise(sets1, sets2):\n",
    "    def recall_score(set1, set2):\n",
    "        tp = len(set1 & set2)  # true positives\n",
    "        fn = len(set2 - set1)  # false negatives\n",
    "        \n",
    "        if tp == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        recall = tp / (tp + fn) if (tp + fn) > 0 else 0\n",
    "        \n",
    "        return recall\n",
    "\n",
    "    return np.array([recall_score(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "def f1_score_elementwise(sets1, sets2):\n",
    "    def f1_score(set1, set2):\n",
    "        tp = len(set1 & set2)  # true positives\n",
    "        fp = len(set1 - set2)  # false positives\n",
    "        fn = len(set2 - set1)  # false negatives\n",
    "        \n",
    "        if tp == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        precision = tp / (tp + fp) if (tp + fp) > 0 else 0\n",
    "        recall = tp / (tp + fn) if (tp + fn) > 0 else 0\n",
    "        \n",
    "        if precision + recall == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        return 2 * (precision * recall) / (precision + recall)\n",
    "    \n",
    "    return np.array([f1_score(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "\n",
    "def compute_metrics(model,\n",
    "                    enriched_prior_gps_per_niche,\n",
    "                    enriched_denovo_gps_per_niche,\n",
    "                    run_dict,\n",
    "                    debug=False):\n",
    "    #######################################################################\n",
    "    ### Prepare enriched and incremented GPs for GP recovery evaluation ###\n",
    "    #######################################################################\n",
    "    print(\"Prepare enriched and incremented GPs for GP recovery evaluation...\")\n",
    "\n",
    "    gp_summary_df = model.get_gp_summary()\n",
    "    \n",
    "    gp_eval_dict = {}\n",
    "    for niche_type in sim_gps_stats[\"regionA\"].unique().tolist():\n",
    "        if niche_type in enriched_prior_gps_per_niche.index:\n",
    "            gp_eval_dict[niche_type] = {\"enriched_prior_gps\": set(enriched_prior_gps_per_niche.loc[niche_type][\"gene_program\"])}\n",
    "            source_genes_top3 = []\n",
    "            source_genes_top5 = []\n",
    "            source_genes_top10 = []\n",
    "            target_genes_top3 = []\n",
    "            target_genes_top5 = []\n",
    "            target_genes_top10 = []\n",
    "            for gp, log_bayes_factor in zip(enriched_prior_gps_per_niche.loc[niche_type][\"gene_program\"], enriched_prior_gps_per_niche.loc[niche_type][\"log_bayes_factor\"]):\n",
    "                # Determine gp sign factor\n",
    "                gp_sign_factor = np.sign(log_bayes_factor)\n",
    "                \n",
    "                gp_source_genes_weights = np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_source_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes\"].values[0]\n",
    "                top_gp_source_genes = np.array(gp_source_genes)[np.argsort(gp_source_genes_weights)[::-1][:10]].tolist()\n",
    "                source_genes_top3.extend(top_gp_source_genes[:3])\n",
    "                source_genes_top5.extend(top_gp_source_genes[:5])\n",
    "                source_genes_top10.extend(top_gp_source_genes[:10])\n",
    "            \n",
    "                gp_target_genes_weights = np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_target_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].values[0]\n",
    "                top_gp_target_genes = np.array(gp_target_genes)[np.argsort(gp_target_genes_weights)[::-1][:10]].tolist()\n",
    "                target_genes_top3.extend(top_gp_target_genes[:3])\n",
    "                target_genes_top5.extend(top_gp_target_genes[:5])\n",
    "                target_genes_top10.extend(top_gp_target_genes[:10])\n",
    "            \n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top3\"] = set(source_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top5\"] = set(source_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top10\"] = set(source_genes_top10)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top3\"] = set(target_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top5\"] = set(target_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top10\"] = set(target_genes_top10)\n",
    "        else:\n",
    "            gp_eval_dict[niche_type] = {\"enriched_prior_gps\": set()}\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top10\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top10\"] = set()\n",
    "            \n",
    "        if niche_type in enriched_denovo_gps_per_niche.index:\n",
    "            source_genes_top3 = []\n",
    "            source_genes_top5 = []\n",
    "            source_genes_top10 = []\n",
    "            target_genes_top3 = []\n",
    "            target_genes_top5 = []\n",
    "            target_genes_top10 = []\n",
    "            for gp, log_bayes_factor in zip(enriched_denovo_gps_per_niche.loc[niche_type][\"gene_program\"], enriched_denovo_gps_per_niche.loc[niche_type][\"log_bayes_factor\"]):\n",
    "                # Determine gp sign factor\n",
    "                gp_sign_factor = np.sign(log_bayes_factor)\n",
    "                \n",
    "                gp_source_genes_weights = np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_source_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes\"].values[0]\n",
    "                top_gp_source_genes = np.array(gp_source_genes)[np.argsort(gp_source_genes_weights)[::-1][:10]].tolist()\n",
    "                source_genes_top3.extend(top_gp_source_genes[:3])\n",
    "                source_genes_top5.extend(top_gp_source_genes[:5])\n",
    "                source_genes_top10.extend(top_gp_source_genes[:10])\n",
    "            \n",
    "                gp_target_genes_weights =  np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_target_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].values[0]\n",
    "                top_gp_target_genes = np.array(gp_target_genes)[np.argsort(gp_target_genes_weights)[::-1][:10]].tolist()\n",
    "                target_genes_top3.extend(top_gp_target_genes[:3])\n",
    "                target_genes_top5.extend(top_gp_target_genes[:5])\n",
    "                target_genes_top10.extend(top_gp_target_genes[:10])\n",
    "            \n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top3\"] = set(source_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top5\"] = set(source_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top10\"] = set(source_genes_top10)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top3\"] = set(target_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top5\"] = set(target_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top10\"] = set(target_genes_top10)\n",
    "        else:\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top10\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top10\"] = set()\n",
    "            \n",
    "        for increment_param in increment_params:\n",
    "            if len(sim_gps_stats[\n",
    "            (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gps\"].values[0]) == 0:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gps\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_source_genes\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_target_genes\"] = set()\n",
    "            else:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gps\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gps\"].apply(\n",
    "                    lambda x: [gp for gp in x.split(\",\")]).values[0])\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_source_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gp_sources\"].values[0].split(\",\"))\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_target_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gp_targets\"].values[0].split(\",\"))\n",
    "    \n",
    "            if len(sim_gps_stats[\n",
    "            (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gps\"].values[0]) == 0:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gps\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_source_genes\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_target_genes\"] = set()\n",
    "            else:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gps\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gps\"].apply(\n",
    "                    lambda x: [gp for gp in x.split(\",\")]).values[0])\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_source_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gp_sources\"].values[0].split(\",\"))\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_target_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gp_targets\"].values[0].split(\",\"))\n",
    "    \n",
    "    gp_eval_df = pd.DataFrame(gp_eval_dict.values(), index=gp_eval_dict.keys())\n",
    "    \n",
    "    # Aggregate incremented elements to compute summary metrics\n",
    "    for element_type in [\"prior_gps\",\n",
    "                         \"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        gp_eval_df[f\"incremented_{element_type}\"] = gp_eval_df.apply(lambda row:\n",
    "                                                                     row[f\"increment_param_8_{element_type}\"] |\n",
    "                                                                     row[f\"increment_param_5_{element_type}\"] |\n",
    "                                                                     row[f\"increment_param_3_{element_type}\"], axis=1)\n",
    "\n",
    "    \n",
    "    ########################################\n",
    "    ### Compute metrics for enriched GPs ###\n",
    "    ########################################\n",
    "    print(\"Compute GP recovery metrics for enriched GPs...\")\n",
    "    \n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        run_dict[f\"f1_enriched_{element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "        run_dict[f\"precision_enriched_{element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "        run_dict[f\"recall_enriched_{element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "        run_dict[f\"jaccard_index_enriched_{element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "\n",
    " \n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            run_dict[f\"f1_enriched_{top_n_element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            run_dict[f\"precision_enriched_{top_n_element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            run_dict[f\"recall_enriched_{top_n_element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "\n",
    "    \n",
    "    ####################\n",
    "    ### Get GP stats ###\n",
    "    ####################\n",
    "    # Get all genes, prior gps, and prior and de novo gp source and target genes\n",
    "    all_genes = set(model.adata.var_names)\n",
    "    \n",
    "    n_prior_gps = model.adata.varm[gp_sources_mask_key].shape[1]\n",
    "    prior_gps = set(model.adata.uns[gp_names_key][:n_prior_gps])\n",
    "    \n",
    "    prior_gp_source_gene_indices = np.where(np.any(model.adata.varm[gp_sources_mask_key] == 1, axis=1))[0]\n",
    "    prior_gp_source_genes = set(model.adata.var_names[prior_gp_source_gene_indices])\n",
    "    \n",
    "    prior_gp_target_gene_indices = np.where(np.any(model.adata.varm[gp_targets_mask_key] == 1, axis=1))[0]\n",
    "    prior_gp_target_genes = set(model.adata.var_names[prior_gp_target_gene_indices])\n",
    "    \n",
    "    denovo_gp_source_genes = all_genes - prior_gp_source_genes\n",
    "    denovo_gp_target_genes = all_genes - prior_gp_target_genes\n",
    "    \n",
    "    \n",
    "    ##################################################\n",
    "    ### Compute metrics for random GPs as baseline ###\n",
    "    ##################################################\n",
    "    print(\"Compute GP recovery metrics for random GPs...\")\n",
    "    random_gp_dict = {}\n",
    "    \n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        random_gp_dict[f\"f1_random_{element_type}\"] = []\n",
    "        random_gp_dict[f\"precision_random_{element_type}\"] = []\n",
    "        random_gp_dict[f\"recall_random_{element_type}\"] = []\n",
    "        random_gp_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            random_gp_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "            random_gp_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "            random_gp_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "            random_gp_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "    \n",
    "    for seed in range(n_seeds):\n",
    "        random.seed(seed)\n",
    "        # Sample random gp from prior knowledge gps so that number matches enriched gps\n",
    "        gp_eval_df[f\"random_prior_gps_seed{seed}\"] = gp_eval_df[\"enriched_prior_gps\"].apply(\n",
    "            lambda x: set(random.sample([gp for gp in prior_gps], len(x))))\n",
    "        \n",
    "        # Sample random genes from prior knowledge gps so that number matches enriched genes from prior gps\n",
    "        for n in [3, 5, 10]:\n",
    "            gp_eval_df[f\"random_prior_gp_target_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_prior_gp_target_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in prior_gp_target_genes], len(x))))\n",
    "            gp_eval_df[f\"random_prior_gp_source_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_prior_gp_source_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in prior_gp_source_genes], len(x))))\n",
    "        \n",
    "            # Sample random genes from denovo gps so that number matches enriched genes from denovo gps\n",
    "            gp_eval_df[f\"random_denovo_gp_target_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_denovo_gp_target_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in denovo_gp_target_genes], len(x))))\n",
    "            gp_eval_df[f\"random_denovo_gp_source_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_denovo_gp_source_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in denovo_gp_source_genes], len(x))))\n",
    "\n",
    "        # Aggregate metrics over niches\n",
    "        for element_type in [\"prior_gps\"]:\n",
    "            random_gp_dict[f\"f1_random_{element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            random_gp_dict[f\"precision_random_{element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            random_gp_dict[f\"recall_random_{element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            random_gp_dict[f\"jaccard_index_random_{element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "\n",
    "        for element_type in [\"prior_gp_source_genes\",\n",
    "                             \"prior_gp_target_genes\",\n",
    "                             \"denovo_gp_source_genes\",\n",
    "                             \"denovo_gp_target_genes\"]:\n",
    "            for n in [3, 5, 10]:\n",
    "                top_n_element_type = element_type + f\"_top{n}\"\n",
    "                random_gp_dict[f\"f1_random_{top_n_element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "                random_gp_dict[f\"precision_random_{top_n_element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "                random_gp_dict[f\"recall_random_{top_n_element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "                random_gp_dict[f\"jaccard_index_random_{top_n_element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "    \n",
    "    # Aggregate metrics over random seeds\n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        run_dict[f\"f1_random_{element_type}\"].append(np.mean(random_gp_dict[f\"f1_random_{element_type}\"]))\n",
    "        run_dict[f\"precision_random_{element_type}\"].append(np.mean(random_gp_dict[f\"precision_random_{element_type}\"]))\n",
    "        run_dict[f\"recall_random_{element_type}\"].append(np.mean(random_gp_dict[f\"recall_random_{element_type}\"]))\n",
    "        run_dict[f\"jaccard_index_random_{element_type}\"].append(np.mean(random_gp_dict[f\"jaccard_index_random_{element_type}\"]))\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            run_dict[f\"f1_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"f1_random_{top_n_element_type}\"]))\n",
    "            run_dict[f\"precision_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"precision_random_{top_n_element_type}\"]))\n",
    "            run_dict[f\"recall_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"recall_random_{top_n_element_type}\"]))\n",
    "            run_dict[f\"jaccard_index_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"jaccard_index_random_{top_n_element_type}\"]))\n",
    "\n",
    "\n",
    "    ####################################\n",
    "    ### Compute GP recovery z-scores ###\n",
    "    ####################################\n",
    "    \"\"\"\n",
    "    print(\"Compute GP recovery z-scores...\")\n",
    "    # Compute overlaps between enriched elements and incremented elements\n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        for increment_param in increment_params:\n",
    "            gp_eval_df[f\"enriched_{element_type}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                row, col1=f\"enriched_{element_type}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            for increment_param in increment_params:\n",
    "                gp_eval_df[f\"enriched_{top_n_element_type}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                    row, col1=f\"enriched_{top_n_element_type}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "    # Compute overlaps between random elements and incremented elements\n",
    "    for seed in range(n_seeds):\n",
    "        for element_type in [\"prior_gps\"]:\n",
    "            for increment_param in increment_params:\n",
    "                gp_eval_df[f\"random_{element_type}_seed{seed}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                    row, col1=f\"random_{element_type}_seed{seed}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "        for element_type in [\"prior_gp_source_genes\",\n",
    "                             \"prior_gp_target_genes\",\n",
    "                             \"denovo_gp_source_genes\",\n",
    "                             \"denovo_gp_target_genes\"]:\n",
    "            for n in [3, 5, 10]:\n",
    "                top_n_element_type = element_type + f\"_top{n}\"\n",
    "                for increment_param in increment_params:\n",
    "                    gp_eval_df[f\"random_{top_n_element_type}_seed{seed}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                        row, col1=f\"random_{top_n_element_type}_seed{seed}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "    # Compute z-scores\n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        for increment_param in increment_params:\n",
    "            agg_cols = [col for col in gp_eval_df.columns if f\"random_{element_type}_seed\" in col and f\"increment_param_{increment_param}_overlap\" in col]\n",
    "\n",
    "            # Compute means and stds over random seeds\n",
    "            means = gp_eval_df[agg_cols].mean(axis=1)\n",
    "            stds = gp_eval_df[agg_cols].std(axis=1)\n",
    "\n",
    "            # Compute means over regions\n",
    "            run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"].append(\n",
    "                np.mean((gp_eval_df[f\"enriched_{element_type}_increment_param_{increment_param}_overlap\"] - means)/stds))\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            for increment_param in increment_params:\n",
    "                agg_cols = [col for col in gp_eval_df.columns if f\"random_{top_n_element_type}_seed\" in col and f\"increment_param_{increment_param}_overlap\" in col]\n",
    "    \n",
    "                # Compute means and stds over random seeds\n",
    "                means = gp_eval_df[agg_cols].mean(axis=1)\n",
    "                stds = gp_eval_df[agg_cols].std(axis=1)\n",
    "    \n",
    "                # Compute means over regions\n",
    "                run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"].append(\n",
    "                    np.mean((gp_eval_df[f\"enriched_{top_n_element_type}_increment_param_{increment_param}_overlap\"] - means)/stds))\n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    ################################################\n",
    "    ### Compute metrics for niche identification ###\n",
    "    ################################################\n",
    "    print(\"Compute niche identification metrics...\")\n",
    "    run_dict[\"ari\"].append(sdmbench.compute_ARI(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"nmi\"].append(sdmbench.compute_NMI(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"chaos\"].append(sdmbench.compute_CHAOS(\n",
    "        model.adata,\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"pas\"].append(sdmbench.compute_PAS(\n",
    "        model.adata,\n",
    "        \"pred_niche_types\",\n",
    "        spatial_key=\"spatial\"))\n",
    "    run_dict[\"asw\"].append(sdmbench.compute_ASW(\n",
    "        model.adata,\n",
    "        \"pred_niche_types\",\n",
    "        spatial_key=\"spatial\"))\n",
    "    run_dict[\"hom\"].append(sdmbench.compute_HOM(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"com\"].append(sdmbench.compute_COM(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    #moranI, gearyC = sdmbench.marker_score(\n",
    "    #    model.adata,\n",
    "    #    \"niche_types\")\n",
    "    #run_dict[\"moranI\"].append(moranI)\n",
    "    #run_dict[\"gearyC\"].append(gearyC)\n",
    "\n",
    "    if debug:\n",
    "        return gp_eval_df, run_dict\n",
    "    else:\n",
    "        return run_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ab9e4f51-1512-4fcd-8206-33d866433968",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_simulation(combined_gp_dict,\n",
    "                   run_dict,\n",
    "                   mlflow_experiment_id):\n",
    "    # Load data\n",
    "    adata = load_data(combined_gp_dict,\n",
    "                      run_dict)\n",
    "\n",
    "    # Train NicheCompass\n",
    "    model = train_model(adata,\n",
    "                        run_dict,\n",
    "                        mlflow_experiment_id)\n",
    "\n",
    "    # Identify niches\n",
    "    model = identify_niches(model,\n",
    "                            run_dict)\n",
    "\n",
    "    # Get enriched prior and de-novo GPs\n",
    "    enriched_prior_gps_per_niche, enriched_denovo_gps_per_niche = get_enriched_gps(model)\n",
    "\n",
    "    # Compute metrics\n",
    "    run_dict = compute_metrics(model,\n",
    "                               enriched_prior_gps_per_niche,\n",
    "                               enriched_denovo_gps_per_niche,\n",
    "                               run_dict)\n",
    "\n",
    "    return run_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "649c17ae-48c6-443c-b018-c222ce4d813c",
   "metadata": {},
   "source": [
    "## 2. Test Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92de7919-6829-4298-b69f-4d74243da1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_gp_dict = prepare_gps(species=species,\n",
    "                               omnipath_lr_network_file_path=omnipath_lr_network_file_path,\n",
    "                               nichenet_lr_network_file_path=nichenet_lr_network_file_path,\n",
    "                               nichenet_ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "                               gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c65c9bc7-b4eb-4912-9b6a-62b42aa1d012",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"n_neighbors\"] = [6]\n",
    "\n",
    "adata = load_data(combined_gp_dict,\n",
    "          run_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "023f6e5c-71d0-4a74-a3ab-b7f4c80bee1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"model_seed\"] = [0]\n",
    "run_dict[\"run_number\"] = [42]\n",
    "run_dict[\"n_neighbors\"] = [6]\n",
    "run_dict[\"n_addon_gp\"] = [100]\n",
    "run_dict[\"conv_layer_encoder\"] = [\"gatv2conv\"]\n",
    "run_dict[\"active_gp_thresh_ratio\"] = [0.01]\n",
    "run_dict[\"lambda_edge_recon\"] = [500000.]\n",
    "run_dict[\"lambda_gene_expr_recon\"] = [300.]\n",
    "run_dict[\"lambda_l1_masked\"] = [0.]\n",
    "run_dict[\"lamba_l1_addon\"] = [0.]\n",
    "\n",
    "mlflow_experiment_id = None\n",
    "\n",
    "model = train_model(adata,\n",
    "                    run_dict,\n",
    "                    mlflow_experiment_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c392d21f-a33b-4a79-a9be-d1241bc36a43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"model_seed\"] = [0]\n",
    "run_dict[\"run_number\"] = [42]\n",
    "\n",
    "ablation = \"test\"\n",
    "\n",
    "model = identify_niches(model,\n",
    "                        run_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a856643-040f-4512-9434-bc9dc7244769",
   "metadata": {},
   "outputs": [],
   "source": [
    "log_bayes_factor_thresh = 2.3\n",
    "\n",
    "enriched_prior_gps_per_niche, enriched_denovo_gps_per_niche = get_enriched_gps(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f334e20-bf5b-4db7-bbf0-c62321f41e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if n_genes == 1105:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "elif n_genes == 12086:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_stats_{increment_mode}increments.csv\"\n",
    "\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "\n",
    "sim_dict = {}\n",
    "for element in [\"prior_gps\",\n",
    "                \"denovo_gps\",\n",
    "                \"prior_gp_sources\",\n",
    "                \"prior_gp_targets\",\n",
    "                \"denovo_gp_sources\",\n",
    "                \"denovo_gp_targets\"]:\n",
    "    sim_dict[element] = set()\n",
    "\n",
    "for i, row in sim_gps_stats.iterrows():\n",
    "    for element in [\"prior_gps\",\n",
    "                    \"denovo_gps\",\n",
    "                    \"prior_gp_sources\",\n",
    "                    \"prior_gp_targets\",\n",
    "                    \"denovo_gp_sources\",\n",
    "                    \"denovo_gp_targets\"]:\n",
    "        if len(row[element]) > 0:\n",
    "            sim_dict[element].update(row[element].split(\",\"))\n",
    "\n",
    "simulation_gps = sim_dict['prior_gps'] | sim_dict['denovo_gps']\n",
    "simulation_denovo_genes = sim_dict['denovo_gp_sources'] | sim_dict['denovo_gp_targets']\n",
    "simulation_prior_genes = sim_dict['prior_gp_sources'] | sim_dict['prior_gp_targets']\n",
    "\n",
    "\n",
    "print(f\"\"\"Number of GPs : {(\n",
    "    len(simulation_gps))}\"\"\")\n",
    "\n",
    "print(f\"\"\"Number of genes : {(\n",
    "    len(simulation_denovo_genes) +\n",
    "    len(simulation_prior_genes))}\"\"\")\n",
    "\n",
    "print(f\"\"\"Number of denovo genes : {len(simulation_denovo_genes)}\"\"\")\n",
    "\n",
    "print(f\"\"\"Number of prior genes : {len(simulation_prior_genes)}\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb03684a-6e84-4904-96c4-6d5c09027912",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "\n",
    "increment_params = [3, 5, 8]\n",
    "\n",
    "for element_type in [\n",
    "    \"prior_gps\",\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "top_n_genes = 3\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 10\n",
    "\n",
    "eval_df, run_dict = compute_metrics(model,\n",
    "                                    enriched_prior_gps_per_niche,\n",
    "                                    enriched_denovo_gps_per_niche,\n",
    "                                    run_dict,\n",
    "                                    debug=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2da62c5-5e5c-4a40-9e9b-e488ad6332dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "enriched_prior_gps_per_niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdb381ae-a558-4ca5-8bee-fd6a49b01640",
   "metadata": {},
   "outputs": [],
   "source": [
    "enriched_denovo_gps_per_niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eec12d64-af9c-4f49-bb12-8f12b13db4d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_summary_df = model.get_gp_summary()\n",
    "model.add_active_gp_scores_to_obs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c25e222-67e1-44d4-9dc9-905a2d647e63",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp = \"Add-on_59_GP\"\n",
    "gp_summary_df[gp_summary_df[\"gp_name\"] == gp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe31d635-457d-4aee-bc8d-8e49937a30ce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "idx = 0\n",
    "gene = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].tolist()[0][idx]\n",
    "\n",
    "sc.pl.umap(adata, color=[gp], layer=\"counts\")\n",
    "\n",
    "gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].values.tolist()[0].index(gene)\n",
    "sc.pl.umap(adata, color=[gene], layer=\"counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d3cdd67-f375-4190-85e4-717b0fb3ef90",
   "metadata": {},
   "outputs": [],
   "source": [
    "niche_type = \"Niche 6\"\n",
    "increment_param = 8\n",
    "col = \"denovo_gp_targets\"\n",
    "genes = sim_gps_stats[(sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][col].tolist()[0].split(\",\")\n",
    "\n",
    "gene in genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c32978-764a-4fca-a3af-6eeec94914d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def f1_score(set1, set2):\n",
    "    tp = len(set1 & set2)  # true positives\n",
    "    fp = len(set1 - set2)  # false positives\n",
    "    fn = len(set2 - set1)  # false negatives\n",
    "\n",
    "    print(tp)\n",
    "    print(fp)\n",
    "    print(fn)\n",
    "    \n",
    "    if tp == 0:\n",
    "        return 0.0\n",
    "    \n",
    "    precision = tp / (tp + fp) if (tp + fp) > 0 else 0\n",
    "    recall = tp / (tp + fn) if (tp + fn) > 0 else 0\n",
    "    \n",
    "    if precision + recall == 0:\n",
    "        return 0.0\n",
    "    \n",
    "    return 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "niche = \"Niche 7\" # \"Niche 6\"\n",
    "set1_col = \"random_denovo_gp_target_genes_top3_seed0\" # enriched_denovo_gp_target_genes_top3, random_denovo_gp_target_genes_top3_seed0\n",
    "set2_col = \"incremented_denovo_gp_target_genes\"\n",
    "\n",
    "print(eval_df.loc[niche][set1_col])\n",
    "print(eval_df.loc[niche][set2_col])\n",
    "\n",
    "print(f1_score(eval_df.loc[niche][set1_col], eval_df.loc[niche][set2_col]))\n",
    "\n",
    "f1_scores = []\n",
    "for niche in eval_df.index:\n",
    "    f1_scores.append(f1_score(eval_df.loc[niche][set1_col], eval_df.loc[niche][set2_col]))\n",
    "\n",
    "print(np.mean(f1_scores))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0edeae-4452-4730-b49d-d16c498c069a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_df = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_strongincrements.csv\")\n",
    "gp_df[(gp_df[\"regionA\"] == \"Region8\") & (gp_df[\"increment_param\"] == 8)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b875e4-a6dc-4879-855a-4fd868ac9818",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene = \"Slc37a2\"\n",
    "\n",
    "sc.pl.umap(adata, color=[gene], layer=\"counts\")\n",
    "sq.pl.spatial_scatter(adata, color=gene, shape=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f53d723-4fd7-49dc-844b-58cf8b9585a2",
   "metadata": {},
   "source": [
    "## 3. Perform Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "66340427-aed0-47d4-877c-058494f5b004",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combining ACE2_ligand_receptor_GP and Ace2_ligand_receptor_target_gene_GP.\n",
      "Combining ACE2_ligand_receptor_GP and L-Phenylalanine_metabolite_enzyme_sensor_GP.\n",
      "Combining ADCYAP1_ligand_receptor_GP and Adcyap1_ligand_receptor_target_gene_GP.\n",
      "Combining ADGRE5_ligand_receptor_GP and Adgre5_ligand_receptor_target_gene_GP.\n",
      "Combining ADIPOQ_ligand_receptor_GP and Adipoq_ligand_receptor_target_gene_GP.\n",
      "Combining ADM_ligand_receptor_GP and Adm_ligand_receptor_target_gene_GP.\n",
      "Combining ADM2_ligand_receptor_GP and Adm2_ligand_receptor_target_gene_GP.\n",
      "Combining AGRP_ligand_receptor_GP and Agrp_ligand_receptor_target_gene_GP.\n",
      "Combining AGT_ligand_receptor_GP and Agt_ligand_receptor_target_gene_GP.\n",
      "Combining ALCAM_ligand_receptor_GP and Alcam_ligand_receptor_target_gene_GP.\n",
      "Combining ALKAL1_ligand_receptor_GP and Alkal1_ligand_receptor_target_gene_GP.\n",
      "Combining ALKAL2_ligand_receptor_GP and Alkal2_ligand_receptor_target_gene_GP.\n",
      "Combining AMBN_ligand_receptor_GP and Ambn_ligand_receptor_target_gene_GP.\n",
      "Combining AMELX_ligand_receptor_GP and Amelx_ligand_receptor_target_gene_GP.\n",
      "Combining AMH_ligand_receptor_GP and Amh_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPT1_ligand_receptor_GP and Angpt1_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPT2_ligand_receptor_GP and Angpt2_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPT4_ligand_receptor_GP and Angpt4_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL1_ligand_receptor_GP and Angptl1_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL2_ligand_receptor_GP and Angptl2_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL3_ligand_receptor_GP and Angptl3_ligand_receptor_target_gene_GP.\n",
      "Combining ANGPTL4_ligand_receptor_GP and Angptl4_ligand_receptor_target_gene_GP.\n",
      "Combining ANXA1_ligand_receptor_GP and Anxa1_ligand_receptor_target_gene_GP.\n",
      "Combining ANXA2_ligand_receptor_GP and Anxa2_ligand_receptor_target_gene_GP.\n",
      "Combining APLN_ligand_receptor_GP and Apln_ligand_receptor_target_gene_GP.\n",
      "Combining APOB_ligand_receptor_GP and Apob_ligand_receptor_target_gene_GP.\n",
      "Combining APOD_ligand_receptor_GP and Apod_ligand_receptor_target_gene_GP.\n",
      "Combining APOE_ligand_receptor_GP and Apoe_ligand_receptor_target_gene_GP.\n",
      "Combining AREG_ligand_receptor_GP and Areg_ligand_receptor_target_gene_GP.\n",
      "Combining ARTN_ligand_receptor_GP and Artn_ligand_receptor_target_gene_GP.\n",
      "Combining AVP_ligand_receptor_GP and Avp_ligand_receptor_target_gene_GP.\n",
      "Combining AZGP1_ligand_receptor_GP and Azgp1_ligand_receptor_target_gene_GP.\n",
      "Combining BDNF_ligand_receptor_GP and Bdnf_ligand_receptor_target_gene_GP.\n",
      "Combining BMP10_ligand_receptor_GP and Bmp10_ligand_receptor_target_gene_GP.\n",
      "Combining BMP15_ligand_receptor_GP and Bmp15_ligand_receptor_target_gene_GP.\n",
      "Combining BMP2_ligand_receptor_GP and Bmp2_ligand_receptor_target_gene_GP.\n",
      "Combining BMP3_ligand_receptor_GP and Bmp3_ligand_receptor_target_gene_GP.\n",
      "Combining BMP4_ligand_receptor_GP and Bmp4_ligand_receptor_target_gene_GP.\n",
      "Combining BMP5_ligand_receptor_GP and Bmp5_ligand_receptor_target_gene_GP.\n",
      "Combining BMP6_ligand_receptor_GP and Bmp6_ligand_receptor_target_gene_GP.\n",
      "Combining BMP7_ligand_receptor_GP and Bmp7_ligand_receptor_target_gene_GP.\n",
      "Combining BMP8A_ligand_receptor_GP and Bmp8a_ligand_receptor_target_gene_GP.\n",
      "Combining BST2_ligand_receptor_GP and Bst2_ligand_receptor_target_gene_GP.\n",
      "Combining BTC_ligand_receptor_GP and Btc_ligand_receptor_target_gene_GP.\n",
      "Combining C1QTNF1_ligand_receptor_GP and C1qtnf1_ligand_receptor_target_gene_GP.\n",
      "Combining C1QTNF5_ligand_receptor_GP and C1qtnf5_ligand_receptor_target_gene_GP.\n",
      "Combining C4B_ligand_receptor_GP and C4b_ligand_receptor_target_gene_GP.\n",
      "Combining CALCA_ligand_receptor_GP and Calca_ligand_receptor_target_gene_GP.\n",
      "Combining CALCB_ligand_receptor_GP and Calcb_ligand_receptor_target_gene_GP.\n",
      "Combining CALR_ligand_receptor_GP and Calr_ligand_receptor_target_gene_GP.\n",
      "Combining CCK_ligand_receptor_GP and Cck_ligand_receptor_target_gene_GP.\n",
      "Combining CCL1_ligand_receptor_GP and Ccl1_ligand_receptor_target_gene_GP.\n",
      "Combining CCL11_ligand_receptor_GP and Ccl11_ligand_receptor_target_gene_GP.\n",
      "Combining CCL17_ligand_receptor_GP and Ccl17_ligand_receptor_target_gene_GP.\n",
      "Combining CCL18_ligand_receptor_GP and CCL3_ligand_receptor_GP.\n",
      "Combining CCL18_ligand_receptor_GP and Ccl3_ligand_receptor_target_gene_GP.\n",
      "Combining CCL19_ligand_receptor_GP and Ccl19_ligand_receptor_target_gene_GP.\n",
      "Combining CCL2_ligand_receptor_GP and Ccl2_ligand_receptor_target_gene_GP.\n",
      "Combining CCL20_ligand_receptor_GP and Ccl20_ligand_receptor_target_gene_GP.\n",
      "Combining CCL22_ligand_receptor_GP and Ccl22_ligand_receptor_target_gene_GP.\n",
      "Combining CCL24_ligand_receptor_GP and Ccl24_ligand_receptor_target_gene_GP.\n",
      "Combining CCL25_ligand_receptor_GP and Ccl25_ligand_receptor_target_gene_GP.\n",
      "Combining CCL26_ligand_receptor_GP and Ccl26_ligand_receptor_target_gene_GP.\n",
      "Combining CCL28_ligand_receptor_GP and Ccl28_ligand_receptor_target_gene_GP.\n",
      "Combining CCL3_ligand_receptor_GP and CCL18_ligand_receptor_GP.\n",
      "Combining CCL3_ligand_receptor_GP and Ccl3_ligand_receptor_target_gene_GP.\n",
      "Combining CCL4_ligand_receptor_GP and CCL4L2_ligand_receptor_GP.\n",
      "Combining CCL4_ligand_receptor_GP and Ccl4_ligand_receptor_target_gene_GP.\n",
      "Combining CCL4L2_ligand_receptor_GP and CCL4_ligand_receptor_GP.\n",
      "Combining CCL4L2_ligand_receptor_GP and Ccl4_ligand_receptor_target_gene_GP.\n",
      "Combining CCL5_ligand_receptor_GP and Ccl5_ligand_receptor_target_gene_GP.\n",
      "Combining CCL7_ligand_receptor_GP and Ccl7_ligand_receptor_target_gene_GP.\n",
      "Combining CCL8_ligand_receptor_GP and Ccl8_ligand_receptor_target_gene_GP.\n",
      "Combining CCN1_ligand_receptor_GP and Ccn1_ligand_receptor_target_gene_GP.\n",
      "Combining CCN2_ligand_receptor_GP and Ccn2_ligand_receptor_target_gene_GP.\n",
      "Combining CCN3_ligand_receptor_GP and Ccn3_ligand_receptor_target_gene_GP.\n",
      "Combining CD160_ligand_receptor_GP and Cd160_ligand_receptor_target_gene_GP.\n",
      "Combining CD200_ligand_receptor_GP and Cd200_ligand_receptor_target_gene_GP.\n",
      "Combining CD274_ligand_receptor_GP and Cd274_ligand_receptor_target_gene_GP.\n",
      "Combining CD34_ligand_receptor_GP and Cd34_ligand_receptor_target_gene_GP.\n",
      "Combining CD40LG_ligand_receptor_GP and Cd40lg_ligand_receptor_target_gene_GP.\n",
      "Combining CD48_ligand_receptor_GP and Cd48_ligand_receptor_target_gene_GP.\n",
      "Combining CD52_ligand_receptor_GP and Cd52_ligand_receptor_target_gene_GP.\n",
      "Combining CD5L_ligand_receptor_GP and Cd5l_ligand_receptor_target_gene_GP.\n",
      "Combining CD70_ligand_receptor_GP and Cd70_ligand_receptor_target_gene_GP.\n",
      "Combining CD72_ligand_receptor_GP and Cd72_ligand_receptor_target_gene_GP.\n",
      "Combining CD80_ligand_receptor_GP and Cd80_ligand_receptor_target_gene_GP.\n",
      "Combining CD99_ligand_receptor_GP and Cd99_ligand_receptor_target_gene_GP.\n",
      "Combining CEACAM5_ligand_receptor_GP and Ceacam5_ligand_receptor_target_gene_GP.\n",
      "Combining CGA_ligand_receptor_GP and Cga_ligand_receptor_target_gene_GP.\n",
      "Combining CHAD_ligand_receptor_GP and Chad_ligand_receptor_target_gene_GP.\n",
      "Combining CKLF_ligand_receptor_GP and Cklf_ligand_receptor_target_gene_GP.\n",
      "Combining CLCF1_ligand_receptor_GP and Clcf1_ligand_receptor_target_gene_GP.\n",
      "Combining CLDN1_ligand_receptor_GP and Cldn1_ligand_receptor_target_gene_GP.\n",
      "Combining CLDN16_ligand_receptor_GP and Cldn16_ligand_receptor_target_gene_GP.\n",
      "Combining CLEC4G_ligand_receptor_GP and Clec4g_ligand_receptor_target_gene_GP.\n",
      "Combining CLU_ligand_receptor_GP and Clu_ligand_receptor_target_gene_GP.\n",
      "Combining CMTM8_ligand_receptor_GP and Cmtm8_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN1_ligand_receptor_GP and Cntn1_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN2_ligand_receptor_GP and Cntn2_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN3_ligand_receptor_GP and Cntn3_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN4_ligand_receptor_GP and Cntn4_ligand_receptor_target_gene_GP.\n",
      "Combining CNTN6_ligand_receptor_GP and Cntn6_ligand_receptor_target_gene_GP.\n",
      "Combining COL10A1_ligand_receptor_GP and Col10a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL11A2_ligand_receptor_GP and Col11a2_ligand_receptor_target_gene_GP.\n",
      "Combining COL12A1_ligand_receptor_GP and Col12a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL13A1_ligand_receptor_GP and Col13a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL15A1_ligand_receptor_GP and Col15a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL16A1_ligand_receptor_GP and Col16a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL17A1_ligand_receptor_GP and Col17a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL19A1_ligand_receptor_GP and Col19a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL20A1_ligand_receptor_GP and Col20a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL22A1_ligand_receptor_GP and Col22a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL24A1_ligand_receptor_GP and Col24a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL26A1_ligand_receptor_GP and Col26a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL27A1_ligand_receptor_GP and Col27a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL28A1_ligand_receptor_GP and Col28a1_ligand_receptor_target_gene_GP.\n",
      "Combining COL6A3_ligand_receptor_GP and Col6a3_ligand_receptor_target_gene_GP.\n",
      "Combining COL6A5_ligand_receptor_GP and Col6a5_ligand_receptor_target_gene_GP.\n",
      "Combining COL6A6_ligand_receptor_GP and Col6a6_ligand_receptor_target_gene_GP.\n",
      "Combining COL8A2_ligand_receptor_GP and Col8a2_ligand_receptor_target_gene_GP.\n",
      "Combining COMP_ligand_receptor_GP and Comp_ligand_receptor_target_gene_GP.\n",
      "Combining CORT_ligand_receptor_GP and Cort_ligand_receptor_target_gene_GP.\n",
      "Combining CRH_ligand_receptor_GP and Crh_ligand_receptor_target_gene_GP.\n",
      "Combining CRLF1_ligand_receptor_GP and Crlf1_ligand_receptor_target_gene_GP.\n",
      "Combining CSF1_ligand_receptor_GP and Csf1_ligand_receptor_target_gene_GP.\n",
      "Combining CSF2_ligand_receptor_GP and Csf2_ligand_receptor_target_gene_GP.\n",
      "Combining CSF3_ligand_receptor_GP and Csf3_ligand_receptor_target_gene_GP.\n",
      "Combining CSH1_ligand_receptor_GP and GH1_ligand_receptor_GP.\n",
      "Combining CSH1_ligand_receptor_GP and GH2_ligand_receptor_GP.\n",
      "Combining CSH1_ligand_receptor_GP and Gh_ligand_receptor_target_gene_GP.\n",
      "Combining CTF1_ligand_receptor_GP and Ctf1_ligand_receptor_target_gene_GP.\n",
      "Combining CTSG_ligand_receptor_GP and Ctsg_ligand_receptor_target_gene_GP.\n",
      "Combining CX3CL1_ligand_receptor_GP and Cx3cl1_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL1_ligand_receptor_GP and Cxcl1_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL10_ligand_receptor_GP and Cxcl10_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL11_ligand_receptor_GP and Cxcl11_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL12_ligand_receptor_GP and Cxcl12_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL13_ligand_receptor_GP and Cxcl13_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL14_ligand_receptor_GP and Cxcl14_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL17_ligand_receptor_GP and Cxcl17_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL2_ligand_receptor_GP and Cxcl2_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL5_ligand_receptor_GP and CXCL6_ligand_receptor_GP.\n",
      "Combining CXCL5_ligand_receptor_GP and Cxcl5_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL6_ligand_receptor_GP and CXCL5_ligand_receptor_GP.\n",
      "Combining CXCL6_ligand_receptor_GP and Cxcl5_ligand_receptor_target_gene_GP.\n",
      "Combining CXCL9_ligand_receptor_GP and Cxcl9_ligand_receptor_target_gene_GP.\n",
      "Combining DCN_ligand_receptor_GP and Dcn_ligand_receptor_target_gene_GP.\n",
      "Combining DEFB103A_ligand_receptor_GP and Defb14_ligand_receptor_target_gene_GP.\n",
      "Combining DHH_ligand_receptor_GP and Dhh_ligand_receptor_target_gene_GP.\n",
      "Combining DKK1_ligand_receptor_GP and Dkk1_ligand_receptor_target_gene_GP.\n",
      "Combining DKK2_ligand_receptor_GP and Dkk2_ligand_receptor_target_gene_GP.\n",
      "Combining DKK4_ligand_receptor_GP and Dkk4_ligand_receptor_target_gene_GP.\n",
      "Combining DLK1_ligand_receptor_GP and Dlk1_ligand_receptor_target_gene_GP.\n",
      "Combining DLK2_ligand_receptor_GP and Dlk2_ligand_receptor_target_gene_GP.\n",
      "Combining DLL1_ligand_receptor_GP and Dll1_ligand_receptor_target_gene_GP.\n",
      "Combining DLL3_ligand_receptor_GP and Dll3_ligand_receptor_target_gene_GP.\n",
      "Combining DLL4_ligand_receptor_GP and Dll4_ligand_receptor_target_gene_GP.\n",
      "Combining DRAXIN_ligand_receptor_GP and Draxin_ligand_receptor_target_gene_GP.\n",
      "Combining DSC1_ligand_receptor_GP and Dsc1_ligand_receptor_target_gene_GP.\n",
      "Combining DSC2_ligand_receptor_GP and Dsc2_ligand_receptor_target_gene_GP.\n",
      "Combining DSC3_ligand_receptor_GP and Dsc3_ligand_receptor_target_gene_GP.\n",
      "Combining EBI3_ligand_receptor_GP and Ebi3_ligand_receptor_target_gene_GP.\n",
      "Combining EDA_ligand_receptor_GP and Eda_ligand_receptor_target_gene_GP.\n",
      "Combining EDIL3_ligand_receptor_GP and Edil3_ligand_receptor_target_gene_GP.\n",
      "Combining EDN1_ligand_receptor_GP and Edn1_ligand_receptor_target_gene_GP.\n",
      "Combining EDN2_ligand_receptor_GP and Edn2_ligand_receptor_target_gene_GP.\n",
      "Combining EDN3_ligand_receptor_GP and Edn3_ligand_receptor_target_gene_GP.\n",
      "Combining EFEMP1_ligand_receptor_GP and Efemp1_ligand_receptor_target_gene_GP.\n",
      "Combining EFEMP2_ligand_receptor_GP and Efemp2_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA1_ligand_receptor_GP and Efna1_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA2_ligand_receptor_GP and Efna2_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA3_ligand_receptor_GP and Efna3_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA4_ligand_receptor_GP and Efna4_ligand_receptor_target_gene_GP.\n",
      "Combining EFNA5_ligand_receptor_GP and Efna5_ligand_receptor_target_gene_GP.\n",
      "Combining EFNB1_ligand_receptor_GP and Efnb1_ligand_receptor_target_gene_GP.\n",
      "Combining EFNB2_ligand_receptor_GP and Efnb2_ligand_receptor_target_gene_GP.\n",
      "Combining EFNB3_ligand_receptor_GP and Efnb3_ligand_receptor_target_gene_GP.\n",
      "Combining EGF_ligand_receptor_GP and Egf_ligand_receptor_target_gene_GP.\n",
      "Combining ENHO_ligand_receptor_GP and Enho_ligand_receptor_target_gene_GP.\n",
      "Combining EPGN_ligand_receptor_GP and Epgn_ligand_receptor_target_gene_GP.\n",
      "Combining EPO_ligand_receptor_GP and Epo_ligand_receptor_target_gene_GP.\n",
      "Combining EREG_ligand_receptor_GP and Ereg_ligand_receptor_target_gene_GP.\n",
      "Combining FASLG_ligand_receptor_GP and Fasl_ligand_receptor_target_gene_GP.\n",
      "Combining FBLN1_ligand_receptor_GP and Fbln1_ligand_receptor_target_gene_GP.\n",
      "Combining FBLN2_ligand_receptor_GP and Fbln2_ligand_receptor_target_gene_GP.\n",
      "Combining FGF1_ligand_receptor_GP and Fgf1_ligand_receptor_target_gene_GP.\n",
      "Combining FGF10_ligand_receptor_GP and Fgf10_ligand_receptor_target_gene_GP.\n",
      "Combining FGF11_ligand_receptor_GP and Fgf11_ligand_receptor_target_gene_GP.\n",
      "Combining FGF12_ligand_receptor_GP and Fgf12_ligand_receptor_target_gene_GP.\n",
      "Combining FGF13_ligand_receptor_GP and Fgf13_ligand_receptor_target_gene_GP.\n",
      "Combining FGF14_ligand_receptor_GP and Fgf14_ligand_receptor_target_gene_GP.\n",
      "Combining FGF16_ligand_receptor_GP and Fgf16_ligand_receptor_target_gene_GP.\n",
      "Combining FGF17_ligand_receptor_GP and Fgf17_ligand_receptor_target_gene_GP.\n",
      "Combining FGF18_ligand_receptor_GP and Fgf18_ligand_receptor_target_gene_GP.\n",
      "Combining FGF19_ligand_receptor_GP and Fgf15_ligand_receptor_target_gene_GP.\n",
      "Combining FGF2_ligand_receptor_GP and Fgf2_ligand_receptor_target_gene_GP.\n",
      "Combining FGF20_ligand_receptor_GP and Fgf20_ligand_receptor_target_gene_GP.\n",
      "Combining FGF21_ligand_receptor_GP and Fgf21_ligand_receptor_target_gene_GP.\n",
      "Combining FGF22_ligand_receptor_GP and Fgf22_ligand_receptor_target_gene_GP.\n",
      "Combining FGF23_ligand_receptor_GP and Fgf23_ligand_receptor_target_gene_GP.\n",
      "Combining FGF3_ligand_receptor_GP and Fgf3_ligand_receptor_target_gene_GP.\n",
      "Combining FGF4_ligand_receptor_GP and Fgf4_ligand_receptor_target_gene_GP.\n",
      "Combining FGF5_ligand_receptor_GP and Fgf5_ligand_receptor_target_gene_GP.\n",
      "Combining FGF6_ligand_receptor_GP and Fgf6_ligand_receptor_target_gene_GP.\n",
      "Combining FGF7_ligand_receptor_GP and Fgf7_ligand_receptor_target_gene_GP.\n",
      "Combining FGF8_ligand_receptor_GP and Fgf8_ligand_receptor_target_gene_GP.\n",
      "Combining FGF9_ligand_receptor_GP and Fgf9_ligand_receptor_target_gene_GP.\n",
      "Combining FGL1_ligand_receptor_GP and Fgl1_ligand_receptor_target_gene_GP.\n",
      "Combining FLRT3_ligand_receptor_GP and Flrt3_ligand_receptor_target_gene_GP.\n",
      "Combining FLT3LG_ligand_receptor_GP and Flt3l_ligand_receptor_target_gene_GP.\n",
      "Combining FSHB_ligand_receptor_GP and Fshb_ligand_receptor_target_gene_GP.\n",
      "Combining GAL_ligand_receptor_GP and Gal_ligand_receptor_target_gene_GP.\n",
      "Combining GALP_ligand_receptor_GP and Galp_ligand_receptor_target_gene_GP.\n",
      "Combining GAS6_ligand_receptor_GP and Gas6_ligand_receptor_target_gene_GP.\n",
      "Combining GAST_ligand_receptor_GP and Gast_ligand_receptor_target_gene_GP.\n",
      "Combining GC_ligand_receptor_GP and Gc_ligand_receptor_target_gene_GP.\n",
      "Combining GCG_ligand_receptor_GP and Gcg_ligand_receptor_target_gene_GP.\n",
      "Combining GDF10_ligand_receptor_GP and Gdf10_ligand_receptor_target_gene_GP.\n",
      "Combining GDF11_ligand_receptor_GP and Gdf11_ligand_receptor_target_gene_GP.\n",
      "Combining GDF15_ligand_receptor_GP and Gdf15_ligand_receptor_target_gene_GP.\n",
      "Combining GDF2_ligand_receptor_GP and Gdf2_ligand_receptor_target_gene_GP.\n",
      "Combining GDF3_ligand_receptor_GP and Gdf3_ligand_receptor_target_gene_GP.\n",
      "Combining GDF5_ligand_receptor_GP and Gdf5_ligand_receptor_target_gene_GP.\n",
      "Combining GDF6_ligand_receptor_GP and Gdf6_ligand_receptor_target_gene_GP.\n",
      "Combining GDF7_ligand_receptor_GP and Gdf7_ligand_receptor_target_gene_GP.\n",
      "Combining GDF9_ligand_receptor_GP and Gdf9_ligand_receptor_target_gene_GP.\n",
      "Combining GDNF_ligand_receptor_GP and Gdnf_ligand_receptor_target_gene_GP.\n",
      "Combining GH1_ligand_receptor_GP and CSH1_ligand_receptor_GP.\n",
      "Combining GH1_ligand_receptor_GP and GH2_ligand_receptor_GP.\n",
      "Combining GH1_ligand_receptor_GP and Gh_ligand_receptor_target_gene_GP.\n",
      "Combining GH2_ligand_receptor_GP and CSH1_ligand_receptor_GP.\n",
      "Combining GH2_ligand_receptor_GP and GH1_ligand_receptor_GP.\n",
      "Combining GH2_ligand_receptor_GP and Gh_ligand_receptor_target_gene_GP.\n",
      "Combining GHRH_ligand_receptor_GP and Ghrh_ligand_receptor_target_gene_GP.\n",
      "Combining GHRL_ligand_receptor_GP and Ghrl_ligand_receptor_target_gene_GP.\n",
      "Combining GIP_ligand_receptor_GP and Gip_ligand_receptor_target_gene_GP.\n",
      "Combining GNRH1_ligand_receptor_GP and Gnrh1_ligand_receptor_target_gene_GP.\n",
      "Combining GPHA2_ligand_receptor_GP and Gpha2_ligand_receptor_target_gene_GP.\n",
      "Combining GPR15LG_ligand_receptor_GP and 2610528A11Rik_ligand_receptor_target_gene_GP.\n",
      "Combining GREM1_ligand_receptor_GP and Grem1_ligand_receptor_target_gene_GP.\n",
      "Combining GRN_ligand_receptor_GP and Grn_ligand_receptor_target_gene_GP.\n",
      "Combining GRP_ligand_receptor_GP and Grp_ligand_receptor_target_gene_GP.\n",
      "Combining GUCA2A_ligand_receptor_GP and Guca2a_ligand_receptor_target_gene_GP.\n",
      "Combining GUCA2B_ligand_receptor_GP and Guca2b_ligand_receptor_target_gene_GP.\n",
      "Combining HBEGF_ligand_receptor_GP and Hbegf_ligand_receptor_target_gene_GP.\n",
      "Combining HCRT_ligand_receptor_GP and Hcrt_ligand_receptor_target_gene_GP.\n",
      "Combining HGF_ligand_receptor_GP and Hgf_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DMA_ligand_receptor_GP and H2-DMa_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DQB1_ligand_receptor_GP and HLA-DQB2_ligand_receptor_GP.\n",
      "Combining HLA-DQB1_ligand_receptor_GP and H2-Ab1_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DQB2_ligand_receptor_GP and HLA-DQB1_ligand_receptor_GP.\n",
      "Combining HLA-DQB2_ligand_receptor_GP and H2-Ab1_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DRA_ligand_receptor_GP and H2-Ea_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DRB1_ligand_receptor_GP and HLA-DRB5_ligand_receptor_GP.\n",
      "Combining HLA-DRB1_ligand_receptor_GP and H2-Eb2_ligand_receptor_target_gene_GP.\n",
      "Combining HLA-DRB5_ligand_receptor_GP and HLA-DRB1_ligand_receptor_GP.\n",
      "Combining HLA-DRB5_ligand_receptor_GP and H2-Eb2_ligand_receptor_target_gene_GP.\n",
      "Combining HP_ligand_receptor_GP and Hp_ligand_receptor_target_gene_GP.\n",
      "Combining IAPP_ligand_receptor_GP and Iapp_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM1_ligand_receptor_GP and Icam1_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM2_ligand_receptor_GP and Icam2_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM4_ligand_receptor_GP and Icam4_ligand_receptor_target_gene_GP.\n",
      "Combining ICAM5_ligand_receptor_GP and Icam5_ligand_receptor_target_gene_GP.\n",
      "Combining ICOSLG_ligand_receptor_GP and Icosl_ligand_receptor_target_gene_GP.\n",
      "Combining IFNB1_ligand_receptor_GP and Ifnb1_ligand_receptor_target_gene_GP.\n",
      "Combining IFNE_ligand_receptor_GP and Ifne_ligand_receptor_target_gene_GP.\n",
      "Combining IFNG_ligand_receptor_GP and Ifng_ligand_receptor_target_gene_GP.\n",
      "Combining IFNK_ligand_receptor_GP and Ifnk_ligand_receptor_target_gene_GP.\n",
      "Combining IGF1_ligand_receptor_GP and Igf1_ligand_receptor_target_gene_GP.\n",
      "Combining IGF2_ligand_receptor_GP and Igf2_ligand_receptor_target_gene_GP.\n",
      "Combining IGFL2_ligand_receptor_GP and IGFL3_ligand_receptor_GP.\n",
      "Combining IGFL2_ligand_receptor_GP and Igfl3_ligand_receptor_target_gene_GP.\n",
      "Combining IGFL3_ligand_receptor_GP and IGFL2_ligand_receptor_GP.\n",
      "Combining IGFL3_ligand_receptor_GP and Igfl3_ligand_receptor_target_gene_GP.\n",
      "Combining IHH_ligand_receptor_GP and Ihh_ligand_receptor_target_gene_GP.\n",
      "Combining IL10_ligand_receptor_GP and Il10_ligand_receptor_target_gene_GP.\n",
      "Combining IL11_ligand_receptor_GP and Il11_ligand_receptor_target_gene_GP.\n",
      "Combining IL12A_ligand_receptor_GP and Il12a_ligand_receptor_target_gene_GP.\n",
      "Combining IL12B_ligand_receptor_GP and Il12b_ligand_receptor_target_gene_GP.\n",
      "Combining IL13_ligand_receptor_GP and Il13_ligand_receptor_target_gene_GP.\n",
      "Combining IL15_ligand_receptor_GP and Il15_ligand_receptor_target_gene_GP.\n",
      "Combining IL16_ligand_receptor_GP and Il16_ligand_receptor_target_gene_GP.\n",
      "Combining IL17A_ligand_receptor_GP and Il17a_ligand_receptor_target_gene_GP.\n",
      "Combining IL17B_ligand_receptor_GP and Il17b_ligand_receptor_target_gene_GP.\n",
      "Combining IL17C_ligand_receptor_GP and Il17c_ligand_receptor_target_gene_GP.\n",
      "Combining IL17F_ligand_receptor_GP and Il17f_ligand_receptor_target_gene_GP.\n",
      "Combining IL18_ligand_receptor_GP and Il18_ligand_receptor_target_gene_GP.\n",
      "Combining IL19_ligand_receptor_GP and Il19_ligand_receptor_target_gene_GP.\n",
      "Combining IL1A_ligand_receptor_GP and Il1a_ligand_receptor_target_gene_GP.\n",
      "Combining IL1B_ligand_receptor_GP and Il1b_ligand_receptor_target_gene_GP.\n",
      "Combining IL1F10_ligand_receptor_GP and Il1f10_ligand_receptor_target_gene_GP.\n",
      "Combining IL1RN_ligand_receptor_GP and Il1rn_ligand_receptor_target_gene_GP.\n",
      "Combining IL2_ligand_receptor_GP and Il2_ligand_receptor_target_gene_GP.\n",
      "Combining IL20_ligand_receptor_GP and Il20_ligand_receptor_target_gene_GP.\n",
      "Combining IL21_ligand_receptor_GP and Il21_ligand_receptor_target_gene_GP.\n",
      "Combining IL23A_ligand_receptor_GP and Il23a_ligand_receptor_target_gene_GP.\n",
      "Combining IL24_ligand_receptor_GP and Il24_ligand_receptor_target_gene_GP.\n",
      "Combining IL25_ligand_receptor_GP and Il25_ligand_receptor_target_gene_GP.\n",
      "Combining IL27_ligand_receptor_GP and Il27_ligand_receptor_target_gene_GP.\n",
      "Combining IL3_ligand_receptor_GP and Il3_ligand_receptor_target_gene_GP.\n",
      "Combining IL31_ligand_receptor_GP and Il31_ligand_receptor_target_gene_GP.\n",
      "Combining IL33_ligand_receptor_GP and Il33_ligand_receptor_target_gene_GP.\n",
      "Combining IL34_ligand_receptor_GP and Il34_ligand_receptor_target_gene_GP.\n",
      "Combining IL36A_ligand_receptor_GP and Il36a_ligand_receptor_target_gene_GP.\n",
      "Combining IL36B_ligand_receptor_GP and Il36b_ligand_receptor_target_gene_GP.\n",
      "Combining IL36G_ligand_receptor_GP and Il36g_ligand_receptor_target_gene_GP.\n",
      "Combining IL36RN_ligand_receptor_GP and Il36rn_ligand_receptor_target_gene_GP.\n",
      "Combining IL4_ligand_receptor_GP and Il4_ligand_receptor_target_gene_GP.\n",
      "Combining IL5_ligand_receptor_GP and Il5_ligand_receptor_target_gene_GP.\n",
      "Combining IL6_ligand_receptor_GP and Il6_ligand_receptor_target_gene_GP.\n",
      "Combining IL7_ligand_receptor_GP and Il7_ligand_receptor_target_gene_GP.\n",
      "Combining IL9_ligand_receptor_GP and Il9_ligand_receptor_target_gene_GP.\n",
      "Combining INHA_ligand_receptor_GP and Inha_ligand_receptor_target_gene_GP.\n",
      "Combining INHBA_ligand_receptor_GP and Inhba_ligand_receptor_target_gene_GP.\n",
      "Combining INHBB_ligand_receptor_GP and Inhbb_ligand_receptor_target_gene_GP.\n",
      "Combining INHBC_ligand_receptor_GP and Inhbc_ligand_receptor_target_gene_GP.\n",
      "Combining INHBE_ligand_receptor_GP and Inhbe_ligand_receptor_target_gene_GP.\n",
      "Combining INSL3_ligand_receptor_GP and Insl3_ligand_receptor_target_gene_GP.\n",
      "Combining INSL5_ligand_receptor_GP and Insl5_ligand_receptor_target_gene_GP.\n",
      "Combining IZUMO1_ligand_receptor_GP and Izumo1_ligand_receptor_target_gene_GP.\n",
      "Combining JAG1_ligand_receptor_GP and Jag1_ligand_receptor_target_gene_GP.\n",
      "Combining JAG2_ligand_receptor_GP and Jag2_ligand_receptor_target_gene_GP.\n",
      "Combining JAM2_ligand_receptor_GP and Jam2_ligand_receptor_target_gene_GP.\n",
      "Combining JAM3_ligand_receptor_GP and Jam3_ligand_receptor_target_gene_GP.\n",
      "Combining JAML_ligand_receptor_GP and Jaml_ligand_receptor_target_gene_GP.\n",
      "Combining KITLG_ligand_receptor_GP and Kitl_ligand_receptor_target_gene_GP.\n",
      "Combining KL_ligand_receptor_GP and Kl_ligand_receptor_target_gene_GP.\n",
      "Combining L1CAM_ligand_receptor_GP and L1cam_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA1_ligand_receptor_GP and Lama1_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA2_ligand_receptor_GP and Lama2_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA3_ligand_receptor_GP and Lama3_ligand_receptor_target_gene_GP.\n",
      "Combining LAMA5_ligand_receptor_GP and Lama5_ligand_receptor_target_gene_GP.\n",
      "Combining LAMB1_ligand_receptor_GP and Lamb1_ligand_receptor_target_gene_GP.\n",
      "Combining LAMB2_ligand_receptor_GP and Lamb2_ligand_receptor_target_gene_GP.\n",
      "Combining LAMB3_ligand_receptor_GP and Lamb3_ligand_receptor_target_gene_GP.\n",
      "Combining LAMC1_ligand_receptor_GP and Lamc1_ligand_receptor_target_gene_GP.\n",
      "Combining LAMC2_ligand_receptor_GP and Lamc2_ligand_receptor_target_gene_GP.\n",
      "Combining LEAP2_ligand_receptor_GP and Leap2_ligand_receptor_target_gene_GP.\n",
      "Combining LEP_ligand_receptor_GP and Lep_ligand_receptor_target_gene_GP.\n",
      "Combining LFNG_ligand_receptor_GP and Lfng_ligand_receptor_target_gene_GP.\n",
      "Combining LGALS1_ligand_receptor_GP and Lgals1_ligand_receptor_target_gene_GP.\n",
      "Combining LGALS3_ligand_receptor_GP and Lgals3_ligand_receptor_target_gene_GP.\n",
      "Combining LGALS3BP_ligand_receptor_GP and Lgals3bp_ligand_receptor_target_gene_GP.\n",
      "Combining LGI3_ligand_receptor_GP and Lgi3_ligand_receptor_target_gene_GP.\n",
      "Combining LGI4_ligand_receptor_GP and Lgi4_ligand_receptor_target_gene_GP.\n",
      "Combining LHB_ligand_receptor_GP and Lhb_ligand_receptor_target_gene_GP.\n",
      "Combining LIF_ligand_receptor_GP and Lif_ligand_receptor_target_gene_GP.\n",
      "Combining LRIG1_ligand_receptor_GP and Lrig1_ligand_receptor_target_gene_GP.\n",
      "Combining LRRC4B_ligand_receptor_GP and Lrrc4b_ligand_receptor_target_gene_GP.\n",
      "Combining LTA_ligand_receptor_GP and Lta_ligand_receptor_target_gene_GP.\n",
      "Combining LTB_ligand_receptor_GP and Ltb_ligand_receptor_target_gene_GP.\n",
      "Combining LTBP3_ligand_receptor_GP and Ltbp3_ligand_receptor_target_gene_GP.\n",
      "Combining MDK_ligand_receptor_GP and Mdk_ligand_receptor_target_gene_GP.\n",
      "Combining MFAP2_ligand_receptor_GP and Mfap2_ligand_receptor_target_gene_GP.\n",
      "Combining MFGE8_ligand_receptor_GP and Mfge8_ligand_receptor_target_gene_GP.\n",
      "Combining MIA_ligand_receptor_GP and Mia_ligand_receptor_target_gene_GP.\n",
      "Combining MIF_ligand_receptor_GP and Mif_ligand_receptor_target_gene_GP.\n",
      "Combining MST1_ligand_receptor_GP and Mst1_ligand_receptor_target_gene_GP.\n",
      "Combining MSTN_ligand_receptor_GP and Mstn_ligand_receptor_target_gene_GP.\n",
      "Combining MUC1_ligand_receptor_GP and Muc1_ligand_receptor_target_gene_GP.\n",
      "Combining MYOC_ligand_receptor_GP and Myoc_ligand_receptor_target_gene_GP.\n",
      "Combining NDP_ligand_receptor_GP and Ndp_ligand_receptor_target_gene_GP.\n",
      "Combining NECTIN2_ligand_receptor_GP and Nectin2_ligand_receptor_target_gene_GP.\n",
      "Combining NECTIN3_ligand_receptor_GP and Nectin3_ligand_receptor_target_gene_GP.\n",
      "Combining NECTIN4_ligand_receptor_GP and Nectin4_ligand_receptor_target_gene_GP.\n",
      "Combining NELL2_ligand_receptor_GP and Nell2_ligand_receptor_target_gene_GP.\n",
      "Combining NGF_ligand_receptor_GP and Ngf_ligand_receptor_target_gene_GP.\n",
      "Combining NID1_ligand_receptor_GP and Nid1_ligand_receptor_target_gene_GP.\n",
      "Combining NLGN1_ligand_receptor_GP and Nlgn1_ligand_receptor_target_gene_GP.\n",
      "Combining NLGN2_ligand_receptor_GP and Nlgn2_ligand_receptor_target_gene_GP.\n",
      "Combining NLGN3_ligand_receptor_GP and Nlgn3_ligand_receptor_target_gene_GP.\n",
      "Combining NMB_ligand_receptor_GP and Nmb_ligand_receptor_target_gene_GP.\n",
      "Combining NMS_ligand_receptor_GP and Nms_ligand_receptor_target_gene_GP.\n",
      "Combining NMU_ligand_receptor_GP and Nmu_ligand_receptor_target_gene_GP.\n",
      "Combining NODAL_ligand_receptor_GP and Nodal_ligand_receptor_target_gene_GP.\n",
      "Combining NOG_ligand_receptor_GP and Nog_ligand_receptor_target_gene_GP.\n",
      "Combining NPB_ligand_receptor_GP and Npb_ligand_receptor_target_gene_GP.\n",
      "Combining NPFF_ligand_receptor_GP and Npff_ligand_receptor_target_gene_GP.\n",
      "Combining NPNT_ligand_receptor_GP and Npnt_ligand_receptor_target_gene_GP.\n",
      "Combining NPPA_ligand_receptor_GP and Nppa_ligand_receptor_target_gene_GP.\n",
      "Combining NPPB_ligand_receptor_GP and Nppb_ligand_receptor_target_gene_GP.\n",
      "Combining NPPC_ligand_receptor_GP and Nppc_ligand_receptor_target_gene_GP.\n",
      "Combining NPS_ligand_receptor_GP and Nps_ligand_receptor_target_gene_GP.\n",
      "Combining NPTX2_ligand_receptor_GP and Nptx2_ligand_receptor_target_gene_GP.\n",
      "Combining NPVF_ligand_receptor_GP and Npvf_ligand_receptor_target_gene_GP.\n",
      "Combining NPW_ligand_receptor_GP and Npw_ligand_receptor_target_gene_GP.\n",
      "Combining NPY_ligand_receptor_GP and Npy_ligand_receptor_target_gene_GP.\n",
      "Combining NRG1_ligand_receptor_GP and Nrg1_ligand_receptor_target_gene_GP.\n",
      "Combining NRG2_ligand_receptor_GP and Nrg2_ligand_receptor_target_gene_GP.\n",
      "Combining NRG3_ligand_receptor_GP and Nrg3_ligand_receptor_target_gene_GP.\n",
      "Combining NRG4_ligand_receptor_GP and Nrg4_ligand_receptor_target_gene_GP.\n",
      "Combining NRP1_ligand_receptor_GP and Nrp1_ligand_receptor_target_gene_GP.\n",
      "Combining NRTN_ligand_receptor_GP and Nrtn_ligand_receptor_target_gene_GP.\n",
      "Combining NTF3_ligand_receptor_GP and Ntf3_ligand_receptor_target_gene_GP.\n",
      "Combining NTF4_ligand_receptor_GP and Ntf5_ligand_receptor_target_gene_GP.\n",
      "Combining NTN1_ligand_receptor_GP and Ntn1_ligand_receptor_target_gene_GP.\n",
      "Combining NTN3_ligand_receptor_GP and Ntn3_ligand_receptor_target_gene_GP.\n",
      "Combining NTN4_ligand_receptor_GP and Ntn4_ligand_receptor_target_gene_GP.\n",
      "Combining NTNG2_ligand_receptor_GP and Ntng2_ligand_receptor_target_gene_GP.\n",
      "Combining NTS_ligand_receptor_GP and Nts_ligand_receptor_target_gene_GP.\n",
      "Combining NXPH1_ligand_receptor_GP and Nxph1_ligand_receptor_target_gene_GP.\n",
      "Combining NXPH2_ligand_receptor_GP and Nxph2_ligand_receptor_target_gene_GP.\n",
      "Combining NXPH3_ligand_receptor_GP and Nxph3_ligand_receptor_target_gene_GP.\n",
      "Combining OCLN_ligand_receptor_GP and Ocln_ligand_receptor_target_gene_GP.\n",
      "Combining OLFM2_ligand_receptor_GP and Olfm2_ligand_receptor_target_gene_GP.\n",
      "Combining OSM_ligand_receptor_GP and Osm_ligand_receptor_target_gene_GP.\n",
      "Combining OSTN_ligand_receptor_GP and Ostn_ligand_receptor_target_gene_GP.\n",
      "Combining OXT_ligand_receptor_GP and Oxt_ligand_receptor_target_gene_GP.\n",
      "Combining PAM_ligand_receptor_GP and Pam_ligand_receptor_target_gene_GP.\n",
      "Combining PCSK9_ligand_receptor_GP and Pcsk9_ligand_receptor_target_gene_GP.\n",
      "Combining PDCD1LG2_ligand_receptor_GP and Pdcd1lg2_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFA_ligand_receptor_GP and Pdgfa_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFB_ligand_receptor_GP and Pdgfb_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFC_ligand_receptor_GP and Pdgfc_ligand_receptor_target_gene_GP.\n",
      "Combining PDGFD_ligand_receptor_GP and Pdgfd_ligand_receptor_target_gene_GP.\n",
      "Combining PDYN_ligand_receptor_GP and Pdyn_ligand_receptor_target_gene_GP.\n",
      "Combining PECAM1_ligand_receptor_GP and Pecam1_ligand_receptor_target_gene_GP.\n",
      "Combining PENK_ligand_receptor_GP and Penk_ligand_receptor_target_gene_GP.\n",
      "Combining PF4_ligand_receptor_GP and Pf4_ligand_receptor_target_gene_GP.\n",
      "Combining PGF_ligand_receptor_GP and Pgf_ligand_receptor_target_gene_GP.\n",
      "Combining PGLYRP1_ligand_receptor_GP and Pglyrp1_ligand_receptor_target_gene_GP.\n",
      "Combining PIP_ligand_receptor_GP and Pip_ligand_receptor_target_gene_GP.\n",
      "Combining PLG_ligand_receptor_GP and Plg_ligand_receptor_target_gene_GP.\n",
      "Combining PMCH_ligand_receptor_GP and Pmch_ligand_receptor_target_gene_GP.\n",
      "Combining PNOC_ligand_receptor_GP and Pnoc_ligand_receptor_target_gene_GP.\n",
      "Combining PODXL2_ligand_receptor_GP and Podxl2_ligand_receptor_target_gene_GP.\n",
      "Combining POMC_ligand_receptor_GP and Pomc_ligand_receptor_target_gene_GP.\n",
      "Combining POSTN_ligand_receptor_GP and Postn_ligand_receptor_target_gene_GP.\n",
      "Combining PPBP_ligand_receptor_GP and Ppbp_ligand_receptor_target_gene_GP.\n",
      "Combining PPY_ligand_receptor_GP and Ppy_ligand_receptor_target_gene_GP.\n",
      "Combining PRLH_ligand_receptor_GP and Prlh_ligand_receptor_target_gene_GP.\n",
      "Combining PRND_ligand_receptor_GP and Prnd_ligand_receptor_target_gene_GP.\n",
      "Combining PROK1_ligand_receptor_GP and Prok1_ligand_receptor_target_gene_GP.\n",
      "Combining PROK2_ligand_receptor_GP and Prok2_ligand_receptor_target_gene_GP.\n",
      "Combining PSPN_ligand_receptor_GP and Pspn_ligand_receptor_target_gene_GP.\n",
      "Combining PTH_ligand_receptor_GP and Pth_ligand_receptor_target_gene_GP.\n",
      "Combining PTH2_ligand_receptor_GP and Pth2_ligand_receptor_target_gene_GP.\n",
      "Combining PTHLH_ligand_receptor_GP and Pthlh_ligand_receptor_target_gene_GP.\n",
      "Combining PTN_ligand_receptor_GP and Ptn_ligand_receptor_target_gene_GP.\n",
      "Combining PYY_ligand_receptor_GP and Pyy_ligand_receptor_target_gene_GP.\n",
      "Combining QRFP_ligand_receptor_GP and Qrfp_ligand_receptor_target_gene_GP.\n",
      "Combining RAET1E_ligand_receptor_GP and Raet1e_ligand_receptor_target_gene_GP.\n",
      "Combining RARRES2_ligand_receptor_GP and Rarres2_ligand_receptor_target_gene_GP.\n",
      "Combining RBP3_ligand_receptor_GP and Rbp3_ligand_receptor_target_gene_GP.\n",
      "Combining RELN_ligand_receptor_GP and Reln_ligand_receptor_target_gene_GP.\n",
      "Combining RETN_ligand_receptor_GP and Retn_ligand_receptor_target_gene_GP.\n",
      "Combining RGMA_ligand_receptor_GP and Rgma_ligand_receptor_target_gene_GP.\n",
      "Combining RLN1_ligand_receptor_GP and RLN2_ligand_receptor_GP.\n",
      "Combining RLN1_ligand_receptor_GP and Rln1_ligand_receptor_target_gene_GP.\n",
      "Combining RLN2_ligand_receptor_GP and RLN1_ligand_receptor_GP.\n",
      "Combining RLN2_ligand_receptor_GP and Rln1_ligand_receptor_target_gene_GP.\n",
      "Combining RLN3_ligand_receptor_GP and Rln3_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO1_ligand_receptor_GP and Rspo1_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO2_ligand_receptor_GP and Rspo2_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO3_ligand_receptor_GP and Rspo3_ligand_receptor_target_gene_GP.\n",
      "Combining RSPO4_ligand_receptor_GP and Rspo4_ligand_receptor_target_gene_GP.\n",
      "Combining S100A10_ligand_receptor_GP and S100a10_ligand_receptor_target_gene_GP.\n",
      "Combining S100A4_ligand_receptor_GP and S100a4_ligand_receptor_target_gene_GP.\n",
      "Combining S100A8_ligand_receptor_GP and S100a8_ligand_receptor_target_gene_GP.\n",
      "Combining S100B_ligand_receptor_GP and S100b_ligand_receptor_target_gene_GP.\n",
      "Combining SCGB1A1_ligand_receptor_GP and Scgb1a1_ligand_receptor_target_gene_GP.\n",
      "Combining SCGB3A1_ligand_receptor_GP and Scgb3a1_ligand_receptor_target_gene_GP.\n",
      "Combining SCGB3A2_ligand_receptor_GP and Scgb3a2_ligand_receptor_target_gene_GP.\n",
      "Combining SCT_ligand_receptor_GP and Sct_ligand_receptor_target_gene_GP.\n",
      "Combining SDC2_ligand_receptor_GP and Sdc2_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3A_ligand_receptor_GP and Sema3a_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3B_ligand_receptor_GP and Sema3b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3C_ligand_receptor_GP and Sema3c_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3D_ligand_receptor_GP and Sema3d_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3E_ligand_receptor_GP and Sema3e_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3F_ligand_receptor_GP and Sema3f_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA3G_ligand_receptor_GP and Sema3g_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA4B_ligand_receptor_GP and Sema4b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA5A_ligand_receptor_GP and Sema5a_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA5B_ligand_receptor_GP and Sema5b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA6B_ligand_receptor_GP and Sema6b_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA6D_ligand_receptor_GP and Sema6d_ligand_receptor_target_gene_GP.\n",
      "Combining SEMA7A_ligand_receptor_GP and Sema7a_ligand_receptor_target_gene_GP.\n",
      "Combining SFRP2_ligand_receptor_GP and Sfrp2_ligand_receptor_target_gene_GP.\n",
      "Combining SFTPA1_ligand_receptor_GP and Sftpa1_ligand_receptor_target_gene_GP.\n",
      "Combining SFTPD_ligand_receptor_GP and Sftpd_ligand_receptor_target_gene_GP.\n",
      "Combining SHH_ligand_receptor_GP and Shh_ligand_receptor_target_gene_GP.\n",
      "Combining SIGLEC9_ligand_receptor_GP and Siglece_ligand_receptor_target_gene_GP.\n",
      "Combining SLIT1_ligand_receptor_GP and Slit1_ligand_receptor_target_gene_GP.\n",
      "Combining SLIT2_ligand_receptor_GP and Slit2_ligand_receptor_target_gene_GP.\n",
      "Combining SLIT3_ligand_receptor_GP and Slit3_ligand_receptor_target_gene_GP.\n",
      "Combining SLITRK1_ligand_receptor_GP and Slitrk1_ligand_receptor_target_gene_GP.\n",
      "Combining SLURP1_ligand_receptor_GP and Slurp1_ligand_receptor_target_gene_GP.\n",
      "Combining SORL1_ligand_receptor_GP and Sorl1_ligand_receptor_target_gene_GP.\n",
      "Combining SOST_ligand_receptor_GP and Sost_ligand_receptor_target_gene_GP.\n",
      "Combining SPON1_ligand_receptor_GP and Spon1_ligand_receptor_target_gene_GP.\n",
      "Combining SPON2_ligand_receptor_GP and Spon2_ligand_receptor_target_gene_GP.\n",
      "Combining SPP1_ligand_receptor_GP and Spp1_ligand_receptor_target_gene_GP.\n",
      "Combining SPX_ligand_receptor_GP and Spx_ligand_receptor_target_gene_GP.\n",
      "Combining SST_ligand_receptor_GP and Sst_ligand_receptor_target_gene_GP.\n",
      "Combining TAC1_ligand_receptor_GP and Tac1_ligand_receptor_target_gene_GP.\n",
      "Combining TAC3_ligand_receptor_GP and Tac2_ligand_receptor_target_gene_GP.\n",
      "Combining TAC4_ligand_receptor_GP and Tac4_ligand_receptor_target_gene_GP.\n",
      "Combining TF_ligand_receptor_GP and Trf_ligand_receptor_target_gene_GP.\n",
      "Combining TFF1_ligand_receptor_GP and Tff1_ligand_receptor_target_gene_GP.\n",
      "Combining TFF2_ligand_receptor_GP and Tff2_ligand_receptor_target_gene_GP.\n",
      "Combining TG_ligand_receptor_GP and Tg_ligand_receptor_target_gene_GP.\n",
      "Combining TGFA_ligand_receptor_GP and Tgfa_ligand_receptor_target_gene_GP.\n",
      "Combining TGFB1_ligand_receptor_GP and Tgfb1_ligand_receptor_target_gene_GP.\n",
      "Combining TGFB2_ligand_receptor_GP and Tgfb2_ligand_receptor_target_gene_GP.\n",
      "Combining TGFB3_ligand_receptor_GP and Tgfb3_ligand_receptor_target_gene_GP.\n",
      "Combining THBS1_ligand_receptor_GP and Thbs1_ligand_receptor_target_gene_GP.\n",
      "Combining THBS2_ligand_receptor_GP and Thbs2_ligand_receptor_target_gene_GP.\n",
      "Combining THPO_ligand_receptor_GP and Thpo_ligand_receptor_target_gene_GP.\n",
      "Combining TNC_ligand_receptor_GP and Tnc_ligand_receptor_target_gene_GP.\n",
      "Combining TNF_ligand_receptor_GP and Tnf_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF10_ligand_receptor_GP and Tnfsf10_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF11_ligand_receptor_GP and Tnfsf11_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF12_ligand_receptor_GP and Tnfsf12_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF13_ligand_receptor_GP and Tnfsf13_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF13B_ligand_receptor_GP and Tnfsf13b_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF14_ligand_receptor_GP and Tnfsf14_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF15_ligand_receptor_GP and Tnfsf15_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF18_ligand_receptor_GP and Tnfsf18_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF4_ligand_receptor_GP and Tnfsf4_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF8_ligand_receptor_GP and Tnfsf8_ligand_receptor_target_gene_GP.\n",
      "Combining TNFSF9_ligand_receptor_GP and Tnfsf9_ligand_receptor_target_gene_GP.\n",
      "Combining TRH_ligand_receptor_GP and Trh_ligand_receptor_target_gene_GP.\n",
      "Combining TSHB_ligand_receptor_GP and Tshb_ligand_receptor_target_gene_GP.\n",
      "Combining TSLP_ligand_receptor_GP and Tslp_ligand_receptor_target_gene_GP.\n",
      "Combining UCN_ligand_receptor_GP and Ucn_ligand_receptor_target_gene_GP.\n",
      "Combining UCN2_ligand_receptor_GP and Ucn2_ligand_receptor_target_gene_GP.\n",
      "Combining UCN3_ligand_receptor_GP and Ucn3_ligand_receptor_target_gene_GP.\n",
      "Combining UTS2_ligand_receptor_GP and Uts2_ligand_receptor_target_gene_GP.\n",
      "Combining UTS2B_ligand_receptor_GP and Uts2b_ligand_receptor_target_gene_GP.\n",
      "Combining VCAM1_ligand_receptor_GP and Vcam1_ligand_receptor_target_gene_GP.\n",
      "Combining VCAN_ligand_receptor_GP and Vcan_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFA_ligand_receptor_GP and Vegfa_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFB_ligand_receptor_GP and Vegfb_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFC_ligand_receptor_GP and Vegfc_ligand_receptor_target_gene_GP.\n",
      "Combining VEGFD_ligand_receptor_GP and Vegfd_ligand_receptor_target_gene_GP.\n",
      "Combining VIP_ligand_receptor_GP and Vip_ligand_receptor_target_gene_GP.\n",
      "Combining VTN_ligand_receptor_GP and Vtn_ligand_receptor_target_gene_GP.\n",
      "Combining VWF_ligand_receptor_GP and Vwf_ligand_receptor_target_gene_GP.\n",
      "Combining WNT1_ligand_receptor_GP and Wnt1_ligand_receptor_target_gene_GP.\n",
      "Combining WNT10B_ligand_receptor_GP and Wnt10b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT11_ligand_receptor_GP and Wnt11_ligand_receptor_target_gene_GP.\n",
      "Combining WNT2_ligand_receptor_GP and Wnt2_ligand_receptor_target_gene_GP.\n",
      "Combining WNT2B_ligand_receptor_GP and Wnt2b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT3_ligand_receptor_GP and Wnt3_ligand_receptor_target_gene_GP.\n",
      "Combining WNT3A_ligand_receptor_GP and Wnt3a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT4_ligand_receptor_GP and Wnt4_ligand_receptor_target_gene_GP.\n",
      "Combining WNT5A_ligand_receptor_GP and Wnt5a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT5B_ligand_receptor_GP and Wnt5b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT6_ligand_receptor_GP and Wnt6_ligand_receptor_target_gene_GP.\n",
      "Combining WNT7A_ligand_receptor_GP and Wnt7a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT7B_ligand_receptor_GP and Wnt7b_ligand_receptor_target_gene_GP.\n",
      "Combining WNT8A_ligand_receptor_GP and Wnt8a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT9A_ligand_receptor_GP and Wnt9a_ligand_receptor_target_gene_GP.\n",
      "Combining WNT9B_ligand_receptor_GP and Wnt9b_ligand_receptor_target_gene_GP.\n",
      "Combining XCL1_ligand_receptor_GP and XCL2_ligand_receptor_GP.\n",
      "Combining XCL1_ligand_receptor_GP and Xcl1_ligand_receptor_target_gene_GP.\n",
      "Combining XCL2_ligand_receptor_GP and XCL1_ligand_receptor_GP.\n",
      "Combining XCL2_ligand_receptor_GP and Xcl1_ligand_receptor_target_gene_GP.\n",
      "Combining ZP3_ligand_receptor_GP and Zp3_ligand_receptor_target_gene_GP.\n",
      "Combining 2610528A11Rik_ligand_receptor_target_gene_GP and GPR15LG_ligand_receptor_GP.\n",
      "Combining Ace2_ligand_receptor_target_gene_GP and ACE2_ligand_receptor_GP.\n",
      "Combining Ace2_ligand_receptor_target_gene_GP and L-Phenylalanine_metabolite_enzyme_sensor_GP.\n",
      "Combining Adcyap1_ligand_receptor_target_gene_GP and ADCYAP1_ligand_receptor_GP.\n",
      "Combining Adgre5_ligand_receptor_target_gene_GP and ADGRE5_ligand_receptor_GP.\n",
      "Combining Adipoq_ligand_receptor_target_gene_GP and ADIPOQ_ligand_receptor_GP.\n",
      "Combining Adm_ligand_receptor_target_gene_GP and ADM_ligand_receptor_GP.\n",
      "Combining Adm2_ligand_receptor_target_gene_GP and ADM2_ligand_receptor_GP.\n",
      "Combining Agrp_ligand_receptor_target_gene_GP and AGRP_ligand_receptor_GP.\n",
      "Combining Agt_ligand_receptor_target_gene_GP and AGT_ligand_receptor_GP.\n",
      "Combining Alcam_ligand_receptor_target_gene_GP and ALCAM_ligand_receptor_GP.\n",
      "Combining Alkal1_ligand_receptor_target_gene_GP and ALKAL1_ligand_receptor_GP.\n",
      "Combining Alkal2_ligand_receptor_target_gene_GP and ALKAL2_ligand_receptor_GP.\n",
      "Combining Ambn_ligand_receptor_target_gene_GP and AMBN_ligand_receptor_GP.\n",
      "Combining Amelx_ligand_receptor_target_gene_GP and AMELX_ligand_receptor_GP.\n",
      "Combining Amh_ligand_receptor_target_gene_GP and AMH_ligand_receptor_GP.\n",
      "Combining Angpt1_ligand_receptor_target_gene_GP and ANGPT1_ligand_receptor_GP.\n",
      "Combining Angpt2_ligand_receptor_target_gene_GP and ANGPT2_ligand_receptor_GP.\n",
      "Combining Angpt4_ligand_receptor_target_gene_GP and ANGPT4_ligand_receptor_GP.\n",
      "Combining Angptl1_ligand_receptor_target_gene_GP and ANGPTL1_ligand_receptor_GP.\n",
      "Combining Angptl2_ligand_receptor_target_gene_GP and ANGPTL2_ligand_receptor_GP.\n",
      "Combining Angptl3_ligand_receptor_target_gene_GP and ANGPTL3_ligand_receptor_GP.\n",
      "Combining Angptl4_ligand_receptor_target_gene_GP and ANGPTL4_ligand_receptor_GP.\n",
      "Combining Anxa1_ligand_receptor_target_gene_GP and ANXA1_ligand_receptor_GP.\n",
      "Combining Anxa2_ligand_receptor_target_gene_GP and ANXA2_ligand_receptor_GP.\n",
      "Combining Apln_ligand_receptor_target_gene_GP and APLN_ligand_receptor_GP.\n",
      "Combining Apob_ligand_receptor_target_gene_GP and APOB_ligand_receptor_GP.\n",
      "Combining Apod_ligand_receptor_target_gene_GP and APOD_ligand_receptor_GP.\n",
      "Combining Apoe_ligand_receptor_target_gene_GP and APOE_ligand_receptor_GP.\n",
      "Combining Areg_ligand_receptor_target_gene_GP and AREG_ligand_receptor_GP.\n",
      "Combining Artn_ligand_receptor_target_gene_GP and ARTN_ligand_receptor_GP.\n",
      "Combining Avp_ligand_receptor_target_gene_GP and AVP_ligand_receptor_GP.\n",
      "Combining Azgp1_ligand_receptor_target_gene_GP and AZGP1_ligand_receptor_GP.\n",
      "Combining Bdnf_ligand_receptor_target_gene_GP and BDNF_ligand_receptor_GP.\n",
      "Combining Bmp10_ligand_receptor_target_gene_GP and BMP10_ligand_receptor_GP.\n",
      "Combining Bmp15_ligand_receptor_target_gene_GP and BMP15_ligand_receptor_GP.\n",
      "Combining Bmp2_ligand_receptor_target_gene_GP and BMP2_ligand_receptor_GP.\n",
      "Combining Bmp3_ligand_receptor_target_gene_GP and BMP3_ligand_receptor_GP.\n",
      "Combining Bmp4_ligand_receptor_target_gene_GP and BMP4_ligand_receptor_GP.\n",
      "Combining Bmp5_ligand_receptor_target_gene_GP and BMP5_ligand_receptor_GP.\n",
      "Combining Bmp6_ligand_receptor_target_gene_GP and BMP6_ligand_receptor_GP.\n",
      "Combining Bmp7_ligand_receptor_target_gene_GP and BMP7_ligand_receptor_GP.\n",
      "Combining Bmp8a_ligand_receptor_target_gene_GP and BMP8A_ligand_receptor_GP.\n",
      "Combining Bst2_ligand_receptor_target_gene_GP and BST2_ligand_receptor_GP.\n",
      "Combining Btc_ligand_receptor_target_gene_GP and BTC_ligand_receptor_GP.\n",
      "Combining C1qtnf1_ligand_receptor_target_gene_GP and C1QTNF1_ligand_receptor_GP.\n",
      "Combining C1qtnf5_ligand_receptor_target_gene_GP and C1QTNF5_ligand_receptor_GP.\n",
      "Combining C4b_ligand_receptor_target_gene_GP and C4B_ligand_receptor_GP.\n",
      "Combining Calca_ligand_receptor_target_gene_GP and CALCA_ligand_receptor_GP.\n",
      "Combining Calcb_ligand_receptor_target_gene_GP and CALCB_ligand_receptor_GP.\n",
      "Combining Calr_ligand_receptor_target_gene_GP and CALR_ligand_receptor_GP.\n",
      "Combining Cck_ligand_receptor_target_gene_GP and CCK_ligand_receptor_GP.\n",
      "Combining Ccl1_ligand_receptor_target_gene_GP and CCL1_ligand_receptor_GP.\n",
      "Combining Ccl11_ligand_receptor_target_gene_GP and CCL11_ligand_receptor_GP.\n",
      "Combining Ccl17_ligand_receptor_target_gene_GP and CCL17_ligand_receptor_GP.\n",
      "Combining Ccl19_ligand_receptor_target_gene_GP and CCL19_ligand_receptor_GP.\n",
      "Combining Ccl2_ligand_receptor_target_gene_GP and CCL2_ligand_receptor_GP.\n",
      "Combining Ccl20_ligand_receptor_target_gene_GP and CCL20_ligand_receptor_GP.\n",
      "Combining Ccl22_ligand_receptor_target_gene_GP and CCL22_ligand_receptor_GP.\n",
      "Combining Ccl24_ligand_receptor_target_gene_GP and CCL24_ligand_receptor_GP.\n",
      "Combining Ccl25_ligand_receptor_target_gene_GP and CCL25_ligand_receptor_GP.\n",
      "Combining Ccl26_ligand_receptor_target_gene_GP and CCL26_ligand_receptor_GP.\n",
      "Combining Ccl28_ligand_receptor_target_gene_GP and CCL28_ligand_receptor_GP.\n",
      "Combining Ccl3_ligand_receptor_target_gene_GP and CCL18_ligand_receptor_GP.\n",
      "Combining Ccl3_ligand_receptor_target_gene_GP and CCL3_ligand_receptor_GP.\n",
      "Combining Ccl4_ligand_receptor_target_gene_GP and CCL4_ligand_receptor_GP.\n",
      "Combining Ccl4_ligand_receptor_target_gene_GP and CCL4L2_ligand_receptor_GP.\n",
      "Combining Ccl5_ligand_receptor_target_gene_GP and CCL5_ligand_receptor_GP.\n",
      "Combining Ccl7_ligand_receptor_target_gene_GP and CCL7_ligand_receptor_GP.\n",
      "Combining Ccl8_ligand_receptor_target_gene_GP and CCL8_ligand_receptor_GP.\n",
      "Combining Ccn1_ligand_receptor_target_gene_GP and CCN1_ligand_receptor_GP.\n",
      "Combining Ccn2_ligand_receptor_target_gene_GP and CCN2_ligand_receptor_GP.\n",
      "Combining Ccn3_ligand_receptor_target_gene_GP and CCN3_ligand_receptor_GP.\n",
      "Combining Cd160_ligand_receptor_target_gene_GP and CD160_ligand_receptor_GP.\n",
      "Combining Cd200_ligand_receptor_target_gene_GP and CD200_ligand_receptor_GP.\n",
      "Combining Cd274_ligand_receptor_target_gene_GP and CD274_ligand_receptor_GP.\n",
      "Combining Cd34_ligand_receptor_target_gene_GP and CD34_ligand_receptor_GP.\n",
      "Combining Cd40lg_ligand_receptor_target_gene_GP and CD40LG_ligand_receptor_GP.\n",
      "Combining Cd48_ligand_receptor_target_gene_GP and CD48_ligand_receptor_GP.\n",
      "Combining Cd52_ligand_receptor_target_gene_GP and CD52_ligand_receptor_GP.\n",
      "Combining Cd5l_ligand_receptor_target_gene_GP and CD5L_ligand_receptor_GP.\n",
      "Combining Cd70_ligand_receptor_target_gene_GP and CD70_ligand_receptor_GP.\n",
      "Combining Cd72_ligand_receptor_target_gene_GP and CD72_ligand_receptor_GP.\n",
      "Combining Cd80_ligand_receptor_target_gene_GP and CD80_ligand_receptor_GP.\n",
      "Combining Cd99_ligand_receptor_target_gene_GP and CD99_ligand_receptor_GP.\n",
      "Combining Ceacam5_ligand_receptor_target_gene_GP and CEACAM5_ligand_receptor_GP.\n",
      "Combining Cga_ligand_receptor_target_gene_GP and CGA_ligand_receptor_GP.\n",
      "Combining Chad_ligand_receptor_target_gene_GP and CHAD_ligand_receptor_GP.\n",
      "Combining Cklf_ligand_receptor_target_gene_GP and CKLF_ligand_receptor_GP.\n",
      "Combining Clcf1_ligand_receptor_target_gene_GP and CLCF1_ligand_receptor_GP.\n",
      "Combining Cldn1_ligand_receptor_target_gene_GP and CLDN1_ligand_receptor_GP.\n",
      "Combining Cldn16_ligand_receptor_target_gene_GP and CLDN16_ligand_receptor_GP.\n",
      "Combining Clec4g_ligand_receptor_target_gene_GP and CLEC4G_ligand_receptor_GP.\n",
      "Combining Clu_ligand_receptor_target_gene_GP and CLU_ligand_receptor_GP.\n",
      "Combining Cmtm8_ligand_receptor_target_gene_GP and CMTM8_ligand_receptor_GP.\n",
      "Combining Cntn1_ligand_receptor_target_gene_GP and CNTN1_ligand_receptor_GP.\n",
      "Combining Cntn2_ligand_receptor_target_gene_GP and CNTN2_ligand_receptor_GP.\n",
      "Combining Cntn3_ligand_receptor_target_gene_GP and CNTN3_ligand_receptor_GP.\n",
      "Combining Cntn4_ligand_receptor_target_gene_GP and CNTN4_ligand_receptor_GP.\n",
      "Combining Cntn6_ligand_receptor_target_gene_GP and CNTN6_ligand_receptor_GP.\n",
      "Combining Col10a1_ligand_receptor_target_gene_GP and COL10A1_ligand_receptor_GP.\n",
      "Combining Col11a2_ligand_receptor_target_gene_GP and COL11A2_ligand_receptor_GP.\n",
      "Combining Col12a1_ligand_receptor_target_gene_GP and COL12A1_ligand_receptor_GP.\n",
      "Combining Col13a1_ligand_receptor_target_gene_GP and COL13A1_ligand_receptor_GP.\n",
      "Combining Col15a1_ligand_receptor_target_gene_GP and COL15A1_ligand_receptor_GP.\n",
      "Combining Col16a1_ligand_receptor_target_gene_GP and COL16A1_ligand_receptor_GP.\n",
      "Combining Col17a1_ligand_receptor_target_gene_GP and COL17A1_ligand_receptor_GP.\n",
      "Combining Col19a1_ligand_receptor_target_gene_GP and COL19A1_ligand_receptor_GP.\n",
      "Combining Col20a1_ligand_receptor_target_gene_GP and COL20A1_ligand_receptor_GP.\n",
      "Combining Col22a1_ligand_receptor_target_gene_GP and COL22A1_ligand_receptor_GP.\n",
      "Combining Col24a1_ligand_receptor_target_gene_GP and COL24A1_ligand_receptor_GP.\n",
      "Combining Col26a1_ligand_receptor_target_gene_GP and COL26A1_ligand_receptor_GP.\n",
      "Combining Col27a1_ligand_receptor_target_gene_GP and COL27A1_ligand_receptor_GP.\n",
      "Combining Col28a1_ligand_receptor_target_gene_GP and COL28A1_ligand_receptor_GP.\n",
      "Combining Col6a3_ligand_receptor_target_gene_GP and COL6A3_ligand_receptor_GP.\n",
      "Combining Col6a5_ligand_receptor_target_gene_GP and COL6A5_ligand_receptor_GP.\n",
      "Combining Col6a6_ligand_receptor_target_gene_GP and COL6A6_ligand_receptor_GP.\n",
      "Combining Col8a2_ligand_receptor_target_gene_GP and COL8A2_ligand_receptor_GP.\n",
      "Combining Comp_ligand_receptor_target_gene_GP and COMP_ligand_receptor_GP.\n",
      "Combining Cort_ligand_receptor_target_gene_GP and CORT_ligand_receptor_GP.\n",
      "Combining Crh_ligand_receptor_target_gene_GP and CRH_ligand_receptor_GP.\n",
      "Combining Crlf1_ligand_receptor_target_gene_GP and CRLF1_ligand_receptor_GP.\n",
      "Combining Csf1_ligand_receptor_target_gene_GP and CSF1_ligand_receptor_GP.\n",
      "Combining Csf2_ligand_receptor_target_gene_GP and CSF2_ligand_receptor_GP.\n",
      "Combining Csf3_ligand_receptor_target_gene_GP and CSF3_ligand_receptor_GP.\n",
      "Combining Ctf1_ligand_receptor_target_gene_GP and CTF1_ligand_receptor_GP.\n",
      "Combining Ctsg_ligand_receptor_target_gene_GP and CTSG_ligand_receptor_GP.\n",
      "Combining Cx3cl1_ligand_receptor_target_gene_GP and CX3CL1_ligand_receptor_GP.\n",
      "Combining Cxcl1_ligand_receptor_target_gene_GP and CXCL1_ligand_receptor_GP.\n",
      "Combining Cxcl10_ligand_receptor_target_gene_GP and CXCL10_ligand_receptor_GP.\n",
      "Combining Cxcl11_ligand_receptor_target_gene_GP and CXCL11_ligand_receptor_GP.\n",
      "Combining Cxcl12_ligand_receptor_target_gene_GP and CXCL12_ligand_receptor_GP.\n",
      "Combining Cxcl13_ligand_receptor_target_gene_GP and CXCL13_ligand_receptor_GP.\n",
      "Combining Cxcl14_ligand_receptor_target_gene_GP and CXCL14_ligand_receptor_GP.\n",
      "Combining Cxcl17_ligand_receptor_target_gene_GP and CXCL17_ligand_receptor_GP.\n",
      "Combining Cxcl2_ligand_receptor_target_gene_GP and CXCL2_ligand_receptor_GP.\n",
      "Combining Cxcl5_ligand_receptor_target_gene_GP and CXCL5_ligand_receptor_GP.\n",
      "Combining Cxcl5_ligand_receptor_target_gene_GP and CXCL6_ligand_receptor_GP.\n",
      "Combining Cxcl9_ligand_receptor_target_gene_GP and CXCL9_ligand_receptor_GP.\n",
      "Combining Dbh_ligand_receptor_target_gene_GP and Norepinephrine_metabolite_enzyme_sensor_GP.\n",
      "Combining Dcn_ligand_receptor_target_gene_GP and DCN_ligand_receptor_GP.\n",
      "Combining Defb14_ligand_receptor_target_gene_GP and DEFB103A_ligand_receptor_GP.\n",
      "Combining Dhh_ligand_receptor_target_gene_GP and DHH_ligand_receptor_GP.\n",
      "Combining Dkk1_ligand_receptor_target_gene_GP and DKK1_ligand_receptor_GP.\n",
      "Combining Dkk2_ligand_receptor_target_gene_GP and DKK2_ligand_receptor_GP.\n",
      "Combining Dkk4_ligand_receptor_target_gene_GP and DKK4_ligand_receptor_GP.\n",
      "Combining Dlk1_ligand_receptor_target_gene_GP and DLK1_ligand_receptor_GP.\n",
      "Combining Dlk2_ligand_receptor_target_gene_GP and DLK2_ligand_receptor_GP.\n",
      "Combining Dll1_ligand_receptor_target_gene_GP and DLL1_ligand_receptor_GP.\n",
      "Combining Dll3_ligand_receptor_target_gene_GP and DLL3_ligand_receptor_GP.\n",
      "Combining Dll4_ligand_receptor_target_gene_GP and DLL4_ligand_receptor_GP.\n",
      "Combining Draxin_ligand_receptor_target_gene_GP and DRAXIN_ligand_receptor_GP.\n",
      "Combining Dsc1_ligand_receptor_target_gene_GP and DSC1_ligand_receptor_GP.\n",
      "Combining Dsc2_ligand_receptor_target_gene_GP and DSC2_ligand_receptor_GP.\n",
      "Combining Dsc3_ligand_receptor_target_gene_GP and DSC3_ligand_receptor_GP.\n",
      "Combining Ebi3_ligand_receptor_target_gene_GP and EBI3_ligand_receptor_GP.\n",
      "Combining Eda_ligand_receptor_target_gene_GP and EDA_ligand_receptor_GP.\n",
      "Combining Edil3_ligand_receptor_target_gene_GP and EDIL3_ligand_receptor_GP.\n",
      "Combining Edn1_ligand_receptor_target_gene_GP and EDN1_ligand_receptor_GP.\n",
      "Combining Edn2_ligand_receptor_target_gene_GP and EDN2_ligand_receptor_GP.\n",
      "Combining Edn3_ligand_receptor_target_gene_GP and EDN3_ligand_receptor_GP.\n",
      "Combining Efemp1_ligand_receptor_target_gene_GP and EFEMP1_ligand_receptor_GP.\n",
      "Combining Efemp2_ligand_receptor_target_gene_GP and EFEMP2_ligand_receptor_GP.\n",
      "Combining Efna1_ligand_receptor_target_gene_GP and EFNA1_ligand_receptor_GP.\n",
      "Combining Efna2_ligand_receptor_target_gene_GP and EFNA2_ligand_receptor_GP.\n",
      "Combining Efna3_ligand_receptor_target_gene_GP and EFNA3_ligand_receptor_GP.\n",
      "Combining Efna4_ligand_receptor_target_gene_GP and EFNA4_ligand_receptor_GP.\n",
      "Combining Efna5_ligand_receptor_target_gene_GP and EFNA5_ligand_receptor_GP.\n",
      "Combining Efnb1_ligand_receptor_target_gene_GP and EFNB1_ligand_receptor_GP.\n",
      "Combining Efnb2_ligand_receptor_target_gene_GP and EFNB2_ligand_receptor_GP.\n",
      "Combining Efnb3_ligand_receptor_target_gene_GP and EFNB3_ligand_receptor_GP.\n",
      "Combining Egf_ligand_receptor_target_gene_GP and EGF_ligand_receptor_GP.\n",
      "Combining Enho_ligand_receptor_target_gene_GP and ENHO_ligand_receptor_GP.\n",
      "Combining Epgn_ligand_receptor_target_gene_GP and EPGN_ligand_receptor_GP.\n",
      "Combining Epo_ligand_receptor_target_gene_GP and EPO_ligand_receptor_GP.\n",
      "Combining Ereg_ligand_receptor_target_gene_GP and EREG_ligand_receptor_GP.\n",
      "Combining Fasl_ligand_receptor_target_gene_GP and FASLG_ligand_receptor_GP.\n",
      "Combining Fbln1_ligand_receptor_target_gene_GP and FBLN1_ligand_receptor_GP.\n",
      "Combining Fbln2_ligand_receptor_target_gene_GP and FBLN2_ligand_receptor_GP.\n",
      "Combining Fgf1_ligand_receptor_target_gene_GP and FGF1_ligand_receptor_GP.\n",
      "Combining Fgf10_ligand_receptor_target_gene_GP and FGF10_ligand_receptor_GP.\n",
      "Combining Fgf11_ligand_receptor_target_gene_GP and FGF11_ligand_receptor_GP.\n",
      "Combining Fgf12_ligand_receptor_target_gene_GP and FGF12_ligand_receptor_GP.\n",
      "Combining Fgf13_ligand_receptor_target_gene_GP and FGF13_ligand_receptor_GP.\n",
      "Combining Fgf14_ligand_receptor_target_gene_GP and FGF14_ligand_receptor_GP.\n",
      "Combining Fgf15_ligand_receptor_target_gene_GP and FGF19_ligand_receptor_GP.\n",
      "Combining Fgf16_ligand_receptor_target_gene_GP and FGF16_ligand_receptor_GP.\n",
      "Combining Fgf17_ligand_receptor_target_gene_GP and FGF17_ligand_receptor_GP.\n",
      "Combining Fgf18_ligand_receptor_target_gene_GP and FGF18_ligand_receptor_GP.\n",
      "Combining Fgf2_ligand_receptor_target_gene_GP and FGF2_ligand_receptor_GP.\n",
      "Combining Fgf20_ligand_receptor_target_gene_GP and FGF20_ligand_receptor_GP.\n",
      "Combining Fgf21_ligand_receptor_target_gene_GP and FGF21_ligand_receptor_GP.\n",
      "Combining Fgf22_ligand_receptor_target_gene_GP and FGF22_ligand_receptor_GP.\n",
      "Combining Fgf23_ligand_receptor_target_gene_GP and FGF23_ligand_receptor_GP.\n",
      "Combining Fgf3_ligand_receptor_target_gene_GP and FGF3_ligand_receptor_GP.\n",
      "Combining Fgf4_ligand_receptor_target_gene_GP and FGF4_ligand_receptor_GP.\n",
      "Combining Fgf5_ligand_receptor_target_gene_GP and FGF5_ligand_receptor_GP.\n",
      "Combining Fgf6_ligand_receptor_target_gene_GP and FGF6_ligand_receptor_GP.\n",
      "Combining Fgf7_ligand_receptor_target_gene_GP and FGF7_ligand_receptor_GP.\n",
      "Combining Fgf8_ligand_receptor_target_gene_GP and FGF8_ligand_receptor_GP.\n",
      "Combining Fgf9_ligand_receptor_target_gene_GP and FGF9_ligand_receptor_GP.\n",
      "Combining Fgl1_ligand_receptor_target_gene_GP and FGL1_ligand_receptor_GP.\n",
      "Combining Flrt3_ligand_receptor_target_gene_GP and FLRT3_ligand_receptor_GP.\n",
      "Combining Flt3l_ligand_receptor_target_gene_GP and FLT3LG_ligand_receptor_GP.\n",
      "Combining Fshb_ligand_receptor_target_gene_GP and FSHB_ligand_receptor_GP.\n",
      "Combining Gal_ligand_receptor_target_gene_GP and GAL_ligand_receptor_GP.\n",
      "Combining Galp_ligand_receptor_target_gene_GP and GALP_ligand_receptor_GP.\n",
      "Combining Gas6_ligand_receptor_target_gene_GP and GAS6_ligand_receptor_GP.\n",
      "Combining Gast_ligand_receptor_target_gene_GP and GAST_ligand_receptor_GP.\n",
      "Combining Gc_ligand_receptor_target_gene_GP and GC_ligand_receptor_GP.\n",
      "Combining Gcg_ligand_receptor_target_gene_GP and GCG_ligand_receptor_GP.\n",
      "Combining Gdf10_ligand_receptor_target_gene_GP and GDF10_ligand_receptor_GP.\n",
      "Combining Gdf11_ligand_receptor_target_gene_GP and GDF11_ligand_receptor_GP.\n",
      "Combining Gdf15_ligand_receptor_target_gene_GP and GDF15_ligand_receptor_GP.\n",
      "Combining Gdf2_ligand_receptor_target_gene_GP and GDF2_ligand_receptor_GP.\n",
      "Combining Gdf3_ligand_receptor_target_gene_GP and GDF3_ligand_receptor_GP.\n",
      "Combining Gdf5_ligand_receptor_target_gene_GP and GDF5_ligand_receptor_GP.\n",
      "Combining Gdf6_ligand_receptor_target_gene_GP and GDF6_ligand_receptor_GP.\n",
      "Combining Gdf7_ligand_receptor_target_gene_GP and GDF7_ligand_receptor_GP.\n",
      "Combining Gdf9_ligand_receptor_target_gene_GP and GDF9_ligand_receptor_GP.\n",
      "Combining Gdnf_ligand_receptor_target_gene_GP and GDNF_ligand_receptor_GP.\n",
      "Combining Gh_ligand_receptor_target_gene_GP and CSH1_ligand_receptor_GP.\n",
      "Combining Gh_ligand_receptor_target_gene_GP and GH1_ligand_receptor_GP.\n",
      "Combining Gh_ligand_receptor_target_gene_GP and GH2_ligand_receptor_GP.\n",
      "Combining Ghrh_ligand_receptor_target_gene_GP and GHRH_ligand_receptor_GP.\n",
      "Combining Ghrl_ligand_receptor_target_gene_GP and GHRL_ligand_receptor_GP.\n",
      "Combining Gip_ligand_receptor_target_gene_GP and GIP_ligand_receptor_GP.\n",
      "Combining Gnrh1_ligand_receptor_target_gene_GP and GNRH1_ligand_receptor_GP.\n",
      "Combining Gpha2_ligand_receptor_target_gene_GP and GPHA2_ligand_receptor_GP.\n",
      "Combining Grem1_ligand_receptor_target_gene_GP and GREM1_ligand_receptor_GP.\n",
      "Combining Grn_ligand_receptor_target_gene_GP and GRN_ligand_receptor_GP.\n",
      "Combining Grp_ligand_receptor_target_gene_GP and GRP_ligand_receptor_GP.\n",
      "Combining Guca2a_ligand_receptor_target_gene_GP and GUCA2A_ligand_receptor_GP.\n",
      "Combining Guca2b_ligand_receptor_target_gene_GP and GUCA2B_ligand_receptor_GP.\n",
      "Combining H2-Ab1_ligand_receptor_target_gene_GP and HLA-DQB1_ligand_receptor_GP.\n",
      "Combining H2-Ab1_ligand_receptor_target_gene_GP and HLA-DQB2_ligand_receptor_GP.\n",
      "Combining H2-DMa_ligand_receptor_target_gene_GP and HLA-DMA_ligand_receptor_GP.\n",
      "Combining H2-Ea_ligand_receptor_target_gene_GP and HLA-DRA_ligand_receptor_GP.\n",
      "Combining H2-Eb2_ligand_receptor_target_gene_GP and HLA-DRB1_ligand_receptor_GP.\n",
      "Combining H2-Eb2_ligand_receptor_target_gene_GP and HLA-DRB5_ligand_receptor_GP.\n",
      "Combining Hbegf_ligand_receptor_target_gene_GP and HBEGF_ligand_receptor_GP.\n",
      "Combining Hcrt_ligand_receptor_target_gene_GP and HCRT_ligand_receptor_GP.\n",
      "Combining Hgf_ligand_receptor_target_gene_GP and HGF_ligand_receptor_GP.\n",
      "Combining Hp_ligand_receptor_target_gene_GP and HP_ligand_receptor_GP.\n",
      "Combining Iapp_ligand_receptor_target_gene_GP and IAPP_ligand_receptor_GP.\n",
      "Combining Icam1_ligand_receptor_target_gene_GP and ICAM1_ligand_receptor_GP.\n",
      "Combining Icam2_ligand_receptor_target_gene_GP and ICAM2_ligand_receptor_GP.\n",
      "Combining Icam4_ligand_receptor_target_gene_GP and ICAM4_ligand_receptor_GP.\n",
      "Combining Icam5_ligand_receptor_target_gene_GP and ICAM5_ligand_receptor_GP.\n",
      "Combining Icosl_ligand_receptor_target_gene_GP and ICOSLG_ligand_receptor_GP.\n",
      "Combining Ifnb1_ligand_receptor_target_gene_GP and IFNB1_ligand_receptor_GP.\n",
      "Combining Ifne_ligand_receptor_target_gene_GP and IFNE_ligand_receptor_GP.\n",
      "Combining Ifng_ligand_receptor_target_gene_GP and IFNG_ligand_receptor_GP.\n",
      "Combining Ifnk_ligand_receptor_target_gene_GP and IFNK_ligand_receptor_GP.\n",
      "Combining Igf1_ligand_receptor_target_gene_GP and IGF1_ligand_receptor_GP.\n",
      "Combining Igf2_ligand_receptor_target_gene_GP and IGF2_ligand_receptor_GP.\n",
      "Combining Igfl3_ligand_receptor_target_gene_GP and IGFL2_ligand_receptor_GP.\n",
      "Combining Igfl3_ligand_receptor_target_gene_GP and IGFL3_ligand_receptor_GP.\n",
      "Combining Ihh_ligand_receptor_target_gene_GP and IHH_ligand_receptor_GP.\n",
      "Combining Il10_ligand_receptor_target_gene_GP and IL10_ligand_receptor_GP.\n",
      "Combining Il11_ligand_receptor_target_gene_GP and IL11_ligand_receptor_GP.\n",
      "Combining Il12a_ligand_receptor_target_gene_GP and IL12A_ligand_receptor_GP.\n",
      "Combining Il12b_ligand_receptor_target_gene_GP and IL12B_ligand_receptor_GP.\n",
      "Combining Il13_ligand_receptor_target_gene_GP and IL13_ligand_receptor_GP.\n",
      "Combining Il15_ligand_receptor_target_gene_GP and IL15_ligand_receptor_GP.\n",
      "Combining Il16_ligand_receptor_target_gene_GP and IL16_ligand_receptor_GP.\n",
      "Combining Il17a_ligand_receptor_target_gene_GP and IL17A_ligand_receptor_GP.\n",
      "Combining Il17b_ligand_receptor_target_gene_GP and IL17B_ligand_receptor_GP.\n",
      "Combining Il17c_ligand_receptor_target_gene_GP and IL17C_ligand_receptor_GP.\n",
      "Combining Il17f_ligand_receptor_target_gene_GP and IL17F_ligand_receptor_GP.\n",
      "Combining Il18_ligand_receptor_target_gene_GP and IL18_ligand_receptor_GP.\n",
      "Combining Il19_ligand_receptor_target_gene_GP and IL19_ligand_receptor_GP.\n",
      "Combining Il1a_ligand_receptor_target_gene_GP and IL1A_ligand_receptor_GP.\n",
      "Combining Il1b_ligand_receptor_target_gene_GP and IL1B_ligand_receptor_GP.\n",
      "Combining Il1f10_ligand_receptor_target_gene_GP and IL1F10_ligand_receptor_GP.\n",
      "Combining Il1rn_ligand_receptor_target_gene_GP and IL1RN_ligand_receptor_GP.\n",
      "Combining Il2_ligand_receptor_target_gene_GP and IL2_ligand_receptor_GP.\n",
      "Combining Il20_ligand_receptor_target_gene_GP and IL20_ligand_receptor_GP.\n",
      "Combining Il21_ligand_receptor_target_gene_GP and IL21_ligand_receptor_GP.\n",
      "Combining Il23a_ligand_receptor_target_gene_GP and IL23A_ligand_receptor_GP.\n",
      "Combining Il24_ligand_receptor_target_gene_GP and IL24_ligand_receptor_GP.\n",
      "Combining Il25_ligand_receptor_target_gene_GP and IL25_ligand_receptor_GP.\n",
      "Combining Il27_ligand_receptor_target_gene_GP and IL27_ligand_receptor_GP.\n",
      "Combining Il3_ligand_receptor_target_gene_GP and IL3_ligand_receptor_GP.\n",
      "Combining Il31_ligand_receptor_target_gene_GP and IL31_ligand_receptor_GP.\n",
      "Combining Il33_ligand_receptor_target_gene_GP and IL33_ligand_receptor_GP.\n",
      "Combining Il34_ligand_receptor_target_gene_GP and IL34_ligand_receptor_GP.\n",
      "Combining Il36a_ligand_receptor_target_gene_GP and IL36A_ligand_receptor_GP.\n",
      "Combining Il36b_ligand_receptor_target_gene_GP and IL36B_ligand_receptor_GP.\n",
      "Combining Il36g_ligand_receptor_target_gene_GP and IL36G_ligand_receptor_GP.\n",
      "Combining Il36rn_ligand_receptor_target_gene_GP and IL36RN_ligand_receptor_GP.\n",
      "Combining Il4_ligand_receptor_target_gene_GP and IL4_ligand_receptor_GP.\n",
      "Combining Il5_ligand_receptor_target_gene_GP and IL5_ligand_receptor_GP.\n",
      "Combining Il6_ligand_receptor_target_gene_GP and IL6_ligand_receptor_GP.\n",
      "Combining Il7_ligand_receptor_target_gene_GP and IL7_ligand_receptor_GP.\n",
      "Combining Il9_ligand_receptor_target_gene_GP and IL9_ligand_receptor_GP.\n",
      "Combining Inha_ligand_receptor_target_gene_GP and INHA_ligand_receptor_GP.\n",
      "Combining Inhba_ligand_receptor_target_gene_GP and INHBA_ligand_receptor_GP.\n",
      "Combining Inhbb_ligand_receptor_target_gene_GP and INHBB_ligand_receptor_GP.\n",
      "Combining Inhbc_ligand_receptor_target_gene_GP and INHBC_ligand_receptor_GP.\n",
      "Combining Inhbe_ligand_receptor_target_gene_GP and INHBE_ligand_receptor_GP.\n",
      "Combining Insl3_ligand_receptor_target_gene_GP and INSL3_ligand_receptor_GP.\n",
      "Combining Insl5_ligand_receptor_target_gene_GP and INSL5_ligand_receptor_GP.\n",
      "Combining Izumo1_ligand_receptor_target_gene_GP and IZUMO1_ligand_receptor_GP.\n",
      "Combining Jag1_ligand_receptor_target_gene_GP and JAG1_ligand_receptor_GP.\n",
      "Combining Jag2_ligand_receptor_target_gene_GP and JAG2_ligand_receptor_GP.\n",
      "Combining Jam2_ligand_receptor_target_gene_GP and JAM2_ligand_receptor_GP.\n",
      "Combining Jam3_ligand_receptor_target_gene_GP and JAM3_ligand_receptor_GP.\n",
      "Combining Jaml_ligand_receptor_target_gene_GP and JAML_ligand_receptor_GP.\n",
      "Combining Kitl_ligand_receptor_target_gene_GP and KITLG_ligand_receptor_GP.\n",
      "Combining Kl_ligand_receptor_target_gene_GP and KL_ligand_receptor_GP.\n",
      "Combining L1cam_ligand_receptor_target_gene_GP and L1CAM_ligand_receptor_GP.\n",
      "Combining Lama1_ligand_receptor_target_gene_GP and LAMA1_ligand_receptor_GP.\n",
      "Combining Lama2_ligand_receptor_target_gene_GP and LAMA2_ligand_receptor_GP.\n",
      "Combining Lama3_ligand_receptor_target_gene_GP and LAMA3_ligand_receptor_GP.\n",
      "Combining Lama5_ligand_receptor_target_gene_GP and LAMA5_ligand_receptor_GP.\n",
      "Combining Lamb1_ligand_receptor_target_gene_GP and LAMB1_ligand_receptor_GP.\n",
      "Combining Lamb2_ligand_receptor_target_gene_GP and LAMB2_ligand_receptor_GP.\n",
      "Combining Lamb3_ligand_receptor_target_gene_GP and LAMB3_ligand_receptor_GP.\n",
      "Combining Lamc1_ligand_receptor_target_gene_GP and LAMC1_ligand_receptor_GP.\n",
      "Combining Lamc2_ligand_receptor_target_gene_GP and LAMC2_ligand_receptor_GP.\n",
      "Combining Leap2_ligand_receptor_target_gene_GP and LEAP2_ligand_receptor_GP.\n",
      "Combining Lep_ligand_receptor_target_gene_GP and LEP_ligand_receptor_GP.\n",
      "Combining Lfng_ligand_receptor_target_gene_GP and LFNG_ligand_receptor_GP.\n",
      "Combining Lgals1_ligand_receptor_target_gene_GP and LGALS1_ligand_receptor_GP.\n",
      "Combining Lgals3_ligand_receptor_target_gene_GP and LGALS3_ligand_receptor_GP.\n",
      "Combining Lgals3bp_ligand_receptor_target_gene_GP and LGALS3BP_ligand_receptor_GP.\n",
      "Combining Lgi3_ligand_receptor_target_gene_GP and LGI3_ligand_receptor_GP.\n",
      "Combining Lgi4_ligand_receptor_target_gene_GP and LGI4_ligand_receptor_GP.\n",
      "Combining Lhb_ligand_receptor_target_gene_GP and LHB_ligand_receptor_GP.\n",
      "Combining Lif_ligand_receptor_target_gene_GP and LIF_ligand_receptor_GP.\n",
      "Combining Lrig1_ligand_receptor_target_gene_GP and LRIG1_ligand_receptor_GP.\n",
      "Combining Lrrc4b_ligand_receptor_target_gene_GP and LRRC4B_ligand_receptor_GP.\n",
      "Combining Lta_ligand_receptor_target_gene_GP and LTA_ligand_receptor_GP.\n",
      "Combining Ltb_ligand_receptor_target_gene_GP and LTB_ligand_receptor_GP.\n",
      "Combining Ltbp3_ligand_receptor_target_gene_GP and LTBP3_ligand_receptor_GP.\n",
      "Combining Mdk_ligand_receptor_target_gene_GP and MDK_ligand_receptor_GP.\n",
      "Combining Mfap2_ligand_receptor_target_gene_GP and MFAP2_ligand_receptor_GP.\n",
      "Combining Mfge8_ligand_receptor_target_gene_GP and MFGE8_ligand_receptor_GP.\n",
      "Combining Mia_ligand_receptor_target_gene_GP and MIA_ligand_receptor_GP.\n",
      "Combining Mif_ligand_receptor_target_gene_GP and MIF_ligand_receptor_GP.\n",
      "Combining Mst1_ligand_receptor_target_gene_GP and MST1_ligand_receptor_GP.\n",
      "Combining Mstn_ligand_receptor_target_gene_GP and MSTN_ligand_receptor_GP.\n",
      "Combining Muc1_ligand_receptor_target_gene_GP and MUC1_ligand_receptor_GP.\n",
      "Combining Myoc_ligand_receptor_target_gene_GP and MYOC_ligand_receptor_GP.\n",
      "Combining Ndp_ligand_receptor_target_gene_GP and NDP_ligand_receptor_GP.\n",
      "Combining Nectin2_ligand_receptor_target_gene_GP and NECTIN2_ligand_receptor_GP.\n",
      "Combining Nectin3_ligand_receptor_target_gene_GP and NECTIN3_ligand_receptor_GP.\n",
      "Combining Nectin4_ligand_receptor_target_gene_GP and NECTIN4_ligand_receptor_GP.\n",
      "Combining Nell2_ligand_receptor_target_gene_GP and NELL2_ligand_receptor_GP.\n",
      "Combining Ngf_ligand_receptor_target_gene_GP and NGF_ligand_receptor_GP.\n",
      "Combining Nid1_ligand_receptor_target_gene_GP and NID1_ligand_receptor_GP.\n",
      "Combining Nlgn1_ligand_receptor_target_gene_GP and NLGN1_ligand_receptor_GP.\n",
      "Combining Nlgn2_ligand_receptor_target_gene_GP and NLGN2_ligand_receptor_GP.\n",
      "Combining Nlgn3_ligand_receptor_target_gene_GP and NLGN3_ligand_receptor_GP.\n",
      "Combining Nmb_ligand_receptor_target_gene_GP and NMB_ligand_receptor_GP.\n",
      "Combining Nms_ligand_receptor_target_gene_GP and NMS_ligand_receptor_GP.\n",
      "Combining Nmu_ligand_receptor_target_gene_GP and NMU_ligand_receptor_GP.\n",
      "Combining Nodal_ligand_receptor_target_gene_GP and NODAL_ligand_receptor_GP.\n",
      "Combining Nog_ligand_receptor_target_gene_GP and NOG_ligand_receptor_GP.\n",
      "Combining Npb_ligand_receptor_target_gene_GP and NPB_ligand_receptor_GP.\n",
      "Combining Npff_ligand_receptor_target_gene_GP and NPFF_ligand_receptor_GP.\n",
      "Combining Npnt_ligand_receptor_target_gene_GP and NPNT_ligand_receptor_GP.\n",
      "Combining Nppa_ligand_receptor_target_gene_GP and NPPA_ligand_receptor_GP.\n",
      "Combining Nppb_ligand_receptor_target_gene_GP and NPPB_ligand_receptor_GP.\n",
      "Combining Nppc_ligand_receptor_target_gene_GP and NPPC_ligand_receptor_GP.\n",
      "Combining Nps_ligand_receptor_target_gene_GP and NPS_ligand_receptor_GP.\n",
      "Combining Nptx2_ligand_receptor_target_gene_GP and NPTX2_ligand_receptor_GP.\n",
      "Combining Npvf_ligand_receptor_target_gene_GP and NPVF_ligand_receptor_GP.\n",
      "Combining Npw_ligand_receptor_target_gene_GP and NPW_ligand_receptor_GP.\n",
      "Combining Npy_ligand_receptor_target_gene_GP and NPY_ligand_receptor_GP.\n",
      "Combining Nrg1_ligand_receptor_target_gene_GP and NRG1_ligand_receptor_GP.\n",
      "Combining Nrg2_ligand_receptor_target_gene_GP and NRG2_ligand_receptor_GP.\n",
      "Combining Nrg3_ligand_receptor_target_gene_GP and NRG3_ligand_receptor_GP.\n",
      "Combining Nrg4_ligand_receptor_target_gene_GP and NRG4_ligand_receptor_GP.\n",
      "Combining Nrp1_ligand_receptor_target_gene_GP and NRP1_ligand_receptor_GP.\n",
      "Combining Nrtn_ligand_receptor_target_gene_GP and NRTN_ligand_receptor_GP.\n",
      "Combining Ntf3_ligand_receptor_target_gene_GP and NTF3_ligand_receptor_GP.\n",
      "Combining Ntf5_ligand_receptor_target_gene_GP and NTF4_ligand_receptor_GP.\n",
      "Combining Ntn1_ligand_receptor_target_gene_GP and NTN1_ligand_receptor_GP.\n",
      "Combining Ntn3_ligand_receptor_target_gene_GP and NTN3_ligand_receptor_GP.\n",
      "Combining Ntn4_ligand_receptor_target_gene_GP and NTN4_ligand_receptor_GP.\n",
      "Combining Ntng2_ligand_receptor_target_gene_GP and NTNG2_ligand_receptor_GP.\n",
      "Combining Nts_ligand_receptor_target_gene_GP and NTS_ligand_receptor_GP.\n",
      "Combining Nxph1_ligand_receptor_target_gene_GP and NXPH1_ligand_receptor_GP.\n",
      "Combining Nxph2_ligand_receptor_target_gene_GP and NXPH2_ligand_receptor_GP.\n",
      "Combining Nxph3_ligand_receptor_target_gene_GP and NXPH3_ligand_receptor_GP.\n",
      "Combining Ocln_ligand_receptor_target_gene_GP and OCLN_ligand_receptor_GP.\n",
      "Combining Olfm2_ligand_receptor_target_gene_GP and OLFM2_ligand_receptor_GP.\n",
      "Combining Osm_ligand_receptor_target_gene_GP and OSM_ligand_receptor_GP.\n",
      "Combining Ostn_ligand_receptor_target_gene_GP and OSTN_ligand_receptor_GP.\n",
      "Combining Oxt_ligand_receptor_target_gene_GP and OXT_ligand_receptor_GP.\n",
      "Combining Pam_ligand_receptor_target_gene_GP and PAM_ligand_receptor_GP.\n",
      "Combining Pcsk9_ligand_receptor_target_gene_GP and PCSK9_ligand_receptor_GP.\n",
      "Combining Pdcd1lg2_ligand_receptor_target_gene_GP and PDCD1LG2_ligand_receptor_GP.\n",
      "Combining Pdgfa_ligand_receptor_target_gene_GP and PDGFA_ligand_receptor_GP.\n",
      "Combining Pdgfb_ligand_receptor_target_gene_GP and PDGFB_ligand_receptor_GP.\n",
      "Combining Pdgfc_ligand_receptor_target_gene_GP and PDGFC_ligand_receptor_GP.\n",
      "Combining Pdgfd_ligand_receptor_target_gene_GP and PDGFD_ligand_receptor_GP.\n",
      "Combining Pdyn_ligand_receptor_target_gene_GP and PDYN_ligand_receptor_GP.\n",
      "Combining Pecam1_ligand_receptor_target_gene_GP and PECAM1_ligand_receptor_GP.\n",
      "Combining Penk_ligand_receptor_target_gene_GP and PENK_ligand_receptor_GP.\n",
      "Combining Pf4_ligand_receptor_target_gene_GP and PF4_ligand_receptor_GP.\n",
      "Combining Pgf_ligand_receptor_target_gene_GP and PGF_ligand_receptor_GP.\n",
      "Combining Pglyrp1_ligand_receptor_target_gene_GP and PGLYRP1_ligand_receptor_GP.\n",
      "Combining Pip_ligand_receptor_target_gene_GP and PIP_ligand_receptor_GP.\n",
      "Combining Plg_ligand_receptor_target_gene_GP and PLG_ligand_receptor_GP.\n",
      "Combining Pmch_ligand_receptor_target_gene_GP and PMCH_ligand_receptor_GP.\n",
      "Combining Pnoc_ligand_receptor_target_gene_GP and PNOC_ligand_receptor_GP.\n",
      "Combining Podxl2_ligand_receptor_target_gene_GP and PODXL2_ligand_receptor_GP.\n",
      "Combining Pomc_ligand_receptor_target_gene_GP and POMC_ligand_receptor_GP.\n",
      "Combining Postn_ligand_receptor_target_gene_GP and POSTN_ligand_receptor_GP.\n",
      "Combining Ppbp_ligand_receptor_target_gene_GP and PPBP_ligand_receptor_GP.\n",
      "Combining Ppy_ligand_receptor_target_gene_GP and PPY_ligand_receptor_GP.\n",
      "Combining Prlh_ligand_receptor_target_gene_GP and PRLH_ligand_receptor_GP.\n",
      "Combining Prnd_ligand_receptor_target_gene_GP and PRND_ligand_receptor_GP.\n",
      "Combining Prok1_ligand_receptor_target_gene_GP and PROK1_ligand_receptor_GP.\n",
      "Combining Prok2_ligand_receptor_target_gene_GP and PROK2_ligand_receptor_GP.\n",
      "Combining Pspn_ligand_receptor_target_gene_GP and PSPN_ligand_receptor_GP.\n",
      "Combining Pth_ligand_receptor_target_gene_GP and PTH_ligand_receptor_GP.\n",
      "Combining Pth2_ligand_receptor_target_gene_GP and PTH2_ligand_receptor_GP.\n",
      "Combining Pthlh_ligand_receptor_target_gene_GP and PTHLH_ligand_receptor_GP.\n",
      "Combining Ptn_ligand_receptor_target_gene_GP and PTN_ligand_receptor_GP.\n",
      "Combining Pyy_ligand_receptor_target_gene_GP and PYY_ligand_receptor_GP.\n",
      "Combining Qrfp_ligand_receptor_target_gene_GP and QRFP_ligand_receptor_GP.\n",
      "Combining Raet1e_ligand_receptor_target_gene_GP and RAET1E_ligand_receptor_GP.\n",
      "Combining Rarres2_ligand_receptor_target_gene_GP and RARRES2_ligand_receptor_GP.\n",
      "Combining Rbp3_ligand_receptor_target_gene_GP and RBP3_ligand_receptor_GP.\n",
      "Combining Reln_ligand_receptor_target_gene_GP and RELN_ligand_receptor_GP.\n",
      "Combining Retn_ligand_receptor_target_gene_GP and RETN_ligand_receptor_GP.\n",
      "Combining Rgma_ligand_receptor_target_gene_GP and RGMA_ligand_receptor_GP.\n",
      "Combining Rln1_ligand_receptor_target_gene_GP and RLN1_ligand_receptor_GP.\n",
      "Combining Rln1_ligand_receptor_target_gene_GP and RLN2_ligand_receptor_GP.\n",
      "Combining Rln3_ligand_receptor_target_gene_GP and RLN3_ligand_receptor_GP.\n",
      "Combining Rspo1_ligand_receptor_target_gene_GP and RSPO1_ligand_receptor_GP.\n",
      "Combining Rspo2_ligand_receptor_target_gene_GP and RSPO2_ligand_receptor_GP.\n",
      "Combining Rspo3_ligand_receptor_target_gene_GP and RSPO3_ligand_receptor_GP.\n",
      "Combining Rspo4_ligand_receptor_target_gene_GP and RSPO4_ligand_receptor_GP.\n",
      "Combining S100a10_ligand_receptor_target_gene_GP and S100A10_ligand_receptor_GP.\n",
      "Combining S100a4_ligand_receptor_target_gene_GP and S100A4_ligand_receptor_GP.\n",
      "Combining S100a8_ligand_receptor_target_gene_GP and S100A8_ligand_receptor_GP.\n",
      "Combining S100b_ligand_receptor_target_gene_GP and S100B_ligand_receptor_GP.\n",
      "Combining Scgb1a1_ligand_receptor_target_gene_GP and SCGB1A1_ligand_receptor_GP.\n",
      "Combining Scgb3a1_ligand_receptor_target_gene_GP and SCGB3A1_ligand_receptor_GP.\n",
      "Combining Scgb3a2_ligand_receptor_target_gene_GP and SCGB3A2_ligand_receptor_GP.\n",
      "Combining Sct_ligand_receptor_target_gene_GP and SCT_ligand_receptor_GP.\n",
      "Combining Sdc2_ligand_receptor_target_gene_GP and SDC2_ligand_receptor_GP.\n",
      "Combining Sema3a_ligand_receptor_target_gene_GP and SEMA3A_ligand_receptor_GP.\n",
      "Combining Sema3b_ligand_receptor_target_gene_GP and SEMA3B_ligand_receptor_GP.\n",
      "Combining Sema3c_ligand_receptor_target_gene_GP and SEMA3C_ligand_receptor_GP.\n",
      "Combining Sema3d_ligand_receptor_target_gene_GP and SEMA3D_ligand_receptor_GP.\n",
      "Combining Sema3e_ligand_receptor_target_gene_GP and SEMA3E_ligand_receptor_GP.\n",
      "Combining Sema3f_ligand_receptor_target_gene_GP and SEMA3F_ligand_receptor_GP.\n",
      "Combining Sema3g_ligand_receptor_target_gene_GP and SEMA3G_ligand_receptor_GP.\n",
      "Combining Sema4b_ligand_receptor_target_gene_GP and SEMA4B_ligand_receptor_GP.\n",
      "Combining Sema5a_ligand_receptor_target_gene_GP and SEMA5A_ligand_receptor_GP.\n",
      "Combining Sema5b_ligand_receptor_target_gene_GP and SEMA5B_ligand_receptor_GP.\n",
      "Combining Sema6b_ligand_receptor_target_gene_GP and SEMA6B_ligand_receptor_GP.\n",
      "Combining Sema6d_ligand_receptor_target_gene_GP and SEMA6D_ligand_receptor_GP.\n",
      "Combining Sema7a_ligand_receptor_target_gene_GP and SEMA7A_ligand_receptor_GP.\n",
      "Combining Sfrp2_ligand_receptor_target_gene_GP and SFRP2_ligand_receptor_GP.\n",
      "Combining Sftpa1_ligand_receptor_target_gene_GP and SFTPA1_ligand_receptor_GP.\n",
      "Combining Sftpd_ligand_receptor_target_gene_GP and SFTPD_ligand_receptor_GP.\n",
      "Combining Shh_ligand_receptor_target_gene_GP and SHH_ligand_receptor_GP.\n",
      "Combining Siglece_ligand_receptor_target_gene_GP and SIGLEC9_ligand_receptor_GP.\n",
      "Combining Slit1_ligand_receptor_target_gene_GP and SLIT1_ligand_receptor_GP.\n",
      "Combining Slit2_ligand_receptor_target_gene_GP and SLIT2_ligand_receptor_GP.\n",
      "Combining Slit3_ligand_receptor_target_gene_GP and SLIT3_ligand_receptor_GP.\n",
      "Combining Slitrk1_ligand_receptor_target_gene_GP and SLITRK1_ligand_receptor_GP.\n",
      "Combining Slurp1_ligand_receptor_target_gene_GP and SLURP1_ligand_receptor_GP.\n",
      "Combining Sorl1_ligand_receptor_target_gene_GP and SORL1_ligand_receptor_GP.\n",
      "Combining Sost_ligand_receptor_target_gene_GP and SOST_ligand_receptor_GP.\n",
      "Combining Spon1_ligand_receptor_target_gene_GP and SPON1_ligand_receptor_GP.\n",
      "Combining Spon2_ligand_receptor_target_gene_GP and SPON2_ligand_receptor_GP.\n",
      "Combining Spp1_ligand_receptor_target_gene_GP and SPP1_ligand_receptor_GP.\n",
      "Combining Spx_ligand_receptor_target_gene_GP and SPX_ligand_receptor_GP.\n",
      "Combining Sst_ligand_receptor_target_gene_GP and SST_ligand_receptor_GP.\n",
      "Combining Tac1_ligand_receptor_target_gene_GP and TAC1_ligand_receptor_GP.\n",
      "Combining Tac2_ligand_receptor_target_gene_GP and TAC3_ligand_receptor_GP.\n",
      "Combining Tac4_ligand_receptor_target_gene_GP and TAC4_ligand_receptor_GP.\n",
      "Combining Tff1_ligand_receptor_target_gene_GP and TFF1_ligand_receptor_GP.\n",
      "Combining Tff2_ligand_receptor_target_gene_GP and TFF2_ligand_receptor_GP.\n",
      "Combining Tg_ligand_receptor_target_gene_GP and TG_ligand_receptor_GP.\n",
      "Combining Tgfa_ligand_receptor_target_gene_GP and TGFA_ligand_receptor_GP.\n",
      "Combining Tgfb1_ligand_receptor_target_gene_GP and TGFB1_ligand_receptor_GP.\n",
      "Combining Tgfb2_ligand_receptor_target_gene_GP and TGFB2_ligand_receptor_GP.\n",
      "Combining Tgfb3_ligand_receptor_target_gene_GP and TGFB3_ligand_receptor_GP.\n",
      "Combining Thbs1_ligand_receptor_target_gene_GP and THBS1_ligand_receptor_GP.\n",
      "Combining Thbs2_ligand_receptor_target_gene_GP and THBS2_ligand_receptor_GP.\n",
      "Combining Thpo_ligand_receptor_target_gene_GP and THPO_ligand_receptor_GP.\n",
      "Combining Tnc_ligand_receptor_target_gene_GP and TNC_ligand_receptor_GP.\n",
      "Combining Tnf_ligand_receptor_target_gene_GP and TNF_ligand_receptor_GP.\n",
      "Combining Tnfsf10_ligand_receptor_target_gene_GP and TNFSF10_ligand_receptor_GP.\n",
      "Combining Tnfsf11_ligand_receptor_target_gene_GP and TNFSF11_ligand_receptor_GP.\n",
      "Combining Tnfsf12_ligand_receptor_target_gene_GP and TNFSF12_ligand_receptor_GP.\n",
      "Combining Tnfsf13_ligand_receptor_target_gene_GP and TNFSF13_ligand_receptor_GP.\n",
      "Combining Tnfsf13b_ligand_receptor_target_gene_GP and TNFSF13B_ligand_receptor_GP.\n",
      "Combining Tnfsf14_ligand_receptor_target_gene_GP and TNFSF14_ligand_receptor_GP.\n",
      "Combining Tnfsf15_ligand_receptor_target_gene_GP and TNFSF15_ligand_receptor_GP.\n",
      "Combining Tnfsf18_ligand_receptor_target_gene_GP and TNFSF18_ligand_receptor_GP.\n",
      "Combining Tnfsf4_ligand_receptor_target_gene_GP and TNFSF4_ligand_receptor_GP.\n",
      "Combining Tnfsf8_ligand_receptor_target_gene_GP and TNFSF8_ligand_receptor_GP.\n",
      "Combining Tnfsf9_ligand_receptor_target_gene_GP and TNFSF9_ligand_receptor_GP.\n",
      "Combining Tpo_ligand_receptor_target_gene_GP and Liothyronine_metabolite_enzyme_sensor_GP.\n",
      "Combining Trf_ligand_receptor_target_gene_GP and TF_ligand_receptor_GP.\n",
      "Combining Trh_ligand_receptor_target_gene_GP and TRH_ligand_receptor_GP.\n",
      "Combining Tshb_ligand_receptor_target_gene_GP and TSHB_ligand_receptor_GP.\n",
      "Combining Tslp_ligand_receptor_target_gene_GP and TSLP_ligand_receptor_GP.\n",
      "Combining Ucn_ligand_receptor_target_gene_GP and UCN_ligand_receptor_GP.\n",
      "Combining Ucn2_ligand_receptor_target_gene_GP and UCN2_ligand_receptor_GP.\n",
      "Combining Ucn3_ligand_receptor_target_gene_GP and UCN3_ligand_receptor_GP.\n",
      "Combining Uts2_ligand_receptor_target_gene_GP and UTS2_ligand_receptor_GP.\n",
      "Combining Uts2b_ligand_receptor_target_gene_GP and UTS2B_ligand_receptor_GP.\n",
      "Combining Vcam1_ligand_receptor_target_gene_GP and VCAM1_ligand_receptor_GP.\n",
      "Combining Vcan_ligand_receptor_target_gene_GP and VCAN_ligand_receptor_GP.\n",
      "Combining Vegfa_ligand_receptor_target_gene_GP and VEGFA_ligand_receptor_GP.\n",
      "Combining Vegfb_ligand_receptor_target_gene_GP and VEGFB_ligand_receptor_GP.\n",
      "Combining Vegfc_ligand_receptor_target_gene_GP and VEGFC_ligand_receptor_GP.\n",
      "Combining Vegfd_ligand_receptor_target_gene_GP and VEGFD_ligand_receptor_GP.\n",
      "Combining Vip_ligand_receptor_target_gene_GP and VIP_ligand_receptor_GP.\n",
      "Combining Vtn_ligand_receptor_target_gene_GP and VTN_ligand_receptor_GP.\n",
      "Combining Vwf_ligand_receptor_target_gene_GP and VWF_ligand_receptor_GP.\n",
      "Combining Wnt1_ligand_receptor_target_gene_GP and WNT1_ligand_receptor_GP.\n",
      "Combining Wnt10b_ligand_receptor_target_gene_GP and WNT10B_ligand_receptor_GP.\n",
      "Combining Wnt11_ligand_receptor_target_gene_GP and WNT11_ligand_receptor_GP.\n",
      "Combining Wnt2_ligand_receptor_target_gene_GP and WNT2_ligand_receptor_GP.\n",
      "Combining Wnt2b_ligand_receptor_target_gene_GP and WNT2B_ligand_receptor_GP.\n",
      "Combining Wnt3_ligand_receptor_target_gene_GP and WNT3_ligand_receptor_GP.\n",
      "Combining Wnt3a_ligand_receptor_target_gene_GP and WNT3A_ligand_receptor_GP.\n",
      "Combining Wnt4_ligand_receptor_target_gene_GP and WNT4_ligand_receptor_GP.\n",
      "Combining Wnt5a_ligand_receptor_target_gene_GP and WNT5A_ligand_receptor_GP.\n",
      "Combining Wnt5b_ligand_receptor_target_gene_GP and WNT5B_ligand_receptor_GP.\n",
      "Combining Wnt6_ligand_receptor_target_gene_GP and WNT6_ligand_receptor_GP.\n",
      "Combining Wnt7a_ligand_receptor_target_gene_GP and WNT7A_ligand_receptor_GP.\n",
      "Combining Wnt7b_ligand_receptor_target_gene_GP and WNT7B_ligand_receptor_GP.\n",
      "Combining Wnt8a_ligand_receptor_target_gene_GP and WNT8A_ligand_receptor_GP.\n",
      "Combining Wnt9a_ligand_receptor_target_gene_GP and WNT9A_ligand_receptor_GP.\n",
      "Combining Wnt9b_ligand_receptor_target_gene_GP and WNT9B_ligand_receptor_GP.\n",
      "Combining Xcl1_ligand_receptor_target_gene_GP and XCL1_ligand_receptor_GP.\n",
      "Combining Xcl1_ligand_receptor_target_gene_GP and XCL2_ligand_receptor_GP.\n",
      "Combining Zp3_ligand_receptor_target_gene_GP and ZP3_ligand_receptor_GP.\n",
      "Combining L-Phenylalanine_metabolite_enzyme_sensor_GP and ACE2_ligand_receptor_GP.\n",
      "Combining L-Phenylalanine_metabolite_enzyme_sensor_GP and Ace2_ligand_receptor_target_gene_GP.\n",
      "Combining Norepinephrine_metabolite_enzyme_sensor_GP and Dbh_ligand_receptor_target_gene_GP.\n",
      "Combining Liothyronine_metabolite_enzyme_sensor_GP and Tpo_ligand_receptor_target_gene_GP.\n",
      "Combining Vitamin D3_metabolite_enzyme_sensor_GP and Calcidiol_metabolite_enzyme_sensor_GP.\n",
      "Combining all-trans-Retinoic acid_metabolite_enzyme_sensor_GP and 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining all-trans-Retinoic acid_metabolite_enzyme_sensor_GP and 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP and all-trans-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP and 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining Calcidiol_metabolite_enzyme_sensor_GP and Vitamin D3_metabolite_enzyme_sensor_GP.\n",
      "Combining 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP and all-trans-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining 13-cis-Retinoic acid_metabolite_enzyme_sensor_GP and 9-cis-Retinoic acid_metabolite_enzyme_sensor_GP.\n",
      "Combining ACE2_combined_GP and Ace2_combined_GP.\n",
      "Combining CCL18_combined_GP and CCL3_combined_GP.\n",
      "Combining CCL3_combined_GP and CCL18_combined_GP.\n",
      "Combining CCL4_combined_GP and CCL4L2_combined_GP.\n",
      "Combining CCL4L2_combined_GP and CCL4_combined_GP.\n",
      "Combining CSH1_combined_GP and GH1_combined_GP.\n",
      "Combining CSH1_combined_GP and GH2_combined_GP.\n",
      "Combining CXCL5_combined_GP and CXCL6_combined_GP.\n",
      "Combining CXCL6_combined_GP and CXCL5_combined_GP.\n",
      "Combining GH1_combined_GP and CSH1_combined_GP.\n",
      "Combining GH1_combined_GP and GH2_combined_GP.\n",
      "Combining GH2_combined_GP and CSH1_combined_GP.\n",
      "Combining GH2_combined_GP and GH1_combined_GP.\n",
      "Combining HLA-DQB1_combined_GP and HLA-DQB2_combined_GP.\n",
      "Combining HLA-DQB2_combined_GP and HLA-DQB1_combined_GP.\n",
      "Combining HLA-DRB1_combined_GP and HLA-DRB5_combined_GP.\n",
      "Combining HLA-DRB5_combined_GP and HLA-DRB1_combined_GP.\n",
      "Combining IGFL2_combined_GP and IGFL3_combined_GP.\n",
      "Combining IGFL3_combined_GP and IGFL2_combined_GP.\n",
      "Combining RLN1_combined_GP and RLN2_combined_GP.\n",
      "Combining RLN2_combined_GP and RLN1_combined_GP.\n",
      "Combining XCL1_combined_GP and XCL2_combined_GP.\n",
      "Combining XCL2_combined_GP and XCL1_combined_GP.\n",
      "Combining Ace2_combined_GP and ACE2_combined_GP.\n",
      "Combining all-trans-Retinoic acid_combined_GP and 9-cis-Retinoic acid_combined_GP.\n",
      "Combining 9-cis-Retinoic acid_combined_GP and all-trans-Retinoic acid_combined_GP.\n",
      "Number of gene programs after filtering and combining: 1582.\n",
      "Number of simulation GPs: 314.\n"
     ]
    }
   ],
   "source": [
    "###############\n",
    "### Get GPs ###\n",
    "###############\n",
    "combined_gp_dict = prepare_gps(species=species,\n",
    "                               omnipath_lr_network_file_path=omnipath_lr_network_file_path,\n",
    "                               nichenet_lr_network_file_path=nichenet_lr_network_file_path,\n",
    "                               nichenet_ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "                               gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path)\n",
    "\n",
    "####################\n",
    "### Get GP stats ###\n",
    "####################\n",
    "if n_genes == 1105:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "elif n_genes == 12086:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_stats_{increment_mode}increments.csv\"\n",
    "\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "\n",
    "simulation_gps = set()\n",
    "for gp in sim_gps_stats[\"prior_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "for gp in sim_gps_stats[\"denovo_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "\n",
    "print(f\"Number of simulation GPs: {len(simulation_gps)}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d621701-56aa-4b22-90c6-20e5d1166774",
   "metadata": {},
   "source": [
    "### 2.1 Loss Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10f79e70-1e93-4a73-ab34-868537ec8928",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ### Define model configuration ###\n",
    "#################################\n",
    "ablation = \"loss_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000., 0., 500000., 50000., 500000.]\n",
    "config[\"lambda_gene_expr_recon\"] = [300., 300., 0., 300., 30]\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lamba_l1_addon\"] = [0.] * 5\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb03f219-aef0-4bfd-9b8b-c69e00eadd92",
   "metadata": {},
   "source": [
    "### 2.2 Loss Ablation Extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c429f22c-195c-406a-968e-e753043a2123",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- INITIALIZING NEW NETWORK MODULE: VARIATIONAL GENE PROGRAM GRAPH AUTOENCODER ---\n",
      "LOSS -> include_edge_recon_loss: True, include_gene_expr_recon_loss: True, rna_recon_loss: nb\n",
      "NODE LABEL METHOD -> one-hop-norm\n",
      "ACTIVE GP THRESHOLD RATIO -> 0.01\n",
      "LOG VARIATIONAL -> True\n",
      "ONE HOP GCN NORM RNA NODE LABEL AGGREGATOR\n",
      "ENCODER -> n_input: 1105, n_cat_covariates_embed_input: 0, n_hidden: 440, n_latent: 297, n_addon_latent: 143, n_fc_layers: 1, n_layers: 1, conv_layer: gatv2conv, n_attention_heads: 4, dropout_rate: 0.0, \n",
      "COSINE SIM GRAPH DECODER -> dropout_rate: 0.0\n",
      "MASKED TARGET RNA DECODER -> n_prior_gp_input: 297, n_addon_gp_input: 143, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "MASKED SOURCE RNA DECODER -> n_prior_gp_input: 297, n_addon_gp_input: 143, n_cat_covariates_embed_input: 0, n_output: 1105\n",
      "\n",
      "--- INITIALIZING TRAINER ---\n",
      "Number of training nodes: 9000\n",
      "Number of validation nodes: 1000\n",
      "Number of training edges: 31752\n",
      "Number of validation edges: 3527\n",
      "Edge batch size: 1024\n",
      "Node batch size: None\n",
      "\n",
      "--- MODEL TRAINING ---\n",
      "Epoch 1/400 |--------------------| 0.2% val_auroc_score: 0.8568; val_auprc_score: 0.8449; val_best_acc_score: 0.7681; val_best_f1_score: 0.7801; train_kl_reg_loss: 883.0897; train_edge_recon_loss: 337372.0820; train_gene_expr_recon_loss: 600532.6797; train_masked_gp_l1_reg_loss: 141086.3354; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 1079874.1836; train_optim_loss: 1079874.1836; val_kl_reg_loss: 1497.7918; val_edge_recon_loss: 311331.7188; val_gene_expr_recon_loss: 591571.5312; val_masked_gp_l1_reg_loss: 55023.7734; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 959424.8125; val_optim_loss: 959424.8125\n",
      "Epoch 2/400 |--------------------| 0.5% val_auroc_score: 0.8729; val_auprc_score: 0.8579; val_best_acc_score: 0.7944; val_best_f1_score: 0.8020; train_kl_reg_loss: 2621.6389; train_edge_recon_loss: 288104.2969; train_gene_expr_recon_loss: 590732.5176; train_masked_gp_l1_reg_loss: 20582.2310; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 902040.6895; train_optim_loss: 902040.6895; val_kl_reg_loss: 3286.3976; val_edge_recon_loss: 294889.4609; val_gene_expr_recon_loss: 587789.2344; val_masked_gp_l1_reg_loss: 5367.4365; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 891332.5312; val_optim_loss: 891332.5312\n",
      "Epoch 3/400 |--------------------| 0.8% val_auroc_score: 0.9165; val_auprc_score: 0.8947; val_best_acc_score: 0.8572; val_best_f1_score: 0.8598; train_kl_reg_loss: 2828.2640; train_edge_recon_loss: 275928.9609; train_gene_expr_recon_loss: 587518.6582; train_masked_gp_l1_reg_loss: 2546.1722; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 868822.0625; train_optim_loss: 868822.0625; val_kl_reg_loss: 2698.9651; val_edge_recon_loss: 275142.3672; val_gene_expr_recon_loss: 584682.0156; val_masked_gp_l1_reg_loss: 1361.9041; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 863885.2031; val_optim_loss: 863885.2031\n",
      "Epoch 4/400 |--------------------| 1.0% val_auroc_score: 0.9212; val_auprc_score: 0.8953; val_best_acc_score: 0.8734; val_best_f1_score: 0.8773; train_kl_reg_loss: 2751.0119; train_edge_recon_loss: 269977.2793; train_gene_expr_recon_loss: 584308.5039; train_masked_gp_l1_reg_loss: 1197.9285; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 858234.7148; train_optim_loss: 858234.7148; val_kl_reg_loss: 2742.8220; val_edge_recon_loss: 272714.4531; val_gene_expr_recon_loss: 582761.7812; val_masked_gp_l1_reg_loss: 1111.3234; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 859330.3594; val_optim_loss: 859330.3594\n",
      "Epoch 5/400 |--------------------| 1.2% val_auroc_score: 0.9355; val_auprc_score: 0.9102; val_best_acc_score: 0.8869; val_best_f1_score: 0.8921; train_kl_reg_loss: 2667.7917; train_edge_recon_loss: 269562.9951; train_gene_expr_recon_loss: 582195.8438; train_masked_gp_l1_reg_loss: 1111.4356; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 855538.0645; train_optim_loss: 855538.0645; val_kl_reg_loss: 2622.6681; val_edge_recon_loss: 268940.8594; val_gene_expr_recon_loss: 580141.0625; val_masked_gp_l1_reg_loss: 1107.0551; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 852811.6562; val_optim_loss: 852811.6562\n",
      "Epoch 6/400 |--------------------| 1.5% val_auroc_score: 0.9353; val_auprc_score: 0.9103; val_best_acc_score: 0.8894; val_best_f1_score: 0.8950; train_kl_reg_loss: 2554.6057; train_edge_recon_loss: 266709.6030; train_gene_expr_recon_loss: 580009.0410; train_masked_gp_l1_reg_loss: 1090.3969; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 850363.6504; train_optim_loss: 850363.6504; val_kl_reg_loss: 2543.4395; val_edge_recon_loss: 268327.7734; val_gene_expr_recon_loss: 578090.5469; val_masked_gp_l1_reg_loss: 1088.3844; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 850050.1406; val_optim_loss: 850050.1406\n",
      "Epoch 7/400 |--------------------| 1.8% val_auroc_score: 0.9464; val_auprc_score: 0.9226; val_best_acc_score: 0.9025; val_best_f1_score: 0.9075; train_kl_reg_loss: 2548.9422; train_edge_recon_loss: 264489.4561; train_gene_expr_recon_loss: 577808.0781; train_masked_gp_l1_reg_loss: 1073.0363; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 845919.5117; train_optim_loss: 845919.5117; val_kl_reg_loss: 2526.7053; val_edge_recon_loss: 265516.2188; val_gene_expr_recon_loss: 575870.9219; val_masked_gp_l1_reg_loss: 1068.7260; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 844982.5781; val_optim_loss: 844982.5781\n",
      "Epoch 8/400 |--------------------| 2.0% val_auroc_score: 0.9375; val_auprc_score: 0.9141; val_best_acc_score: 0.9001; val_best_f1_score: 0.9052; train_kl_reg_loss: 2494.6183; train_edge_recon_loss: 262086.4365; train_gene_expr_recon_loss: 576043.8379; train_masked_gp_l1_reg_loss: 1062.3645; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 841687.2598; train_optim_loss: 841687.2598; val_kl_reg_loss: 2504.7740; val_edge_recon_loss: 270500.4062; val_gene_expr_recon_loss: 573919.2500; val_masked_gp_l1_reg_loss: 1052.1552; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 847976.5625; val_optim_loss: 847976.5625\n",
      "Epoch 9/400 |--------------------| 2.2% val_auroc_score: 0.9432; val_auprc_score: 0.9218; val_best_acc_score: 0.9029; val_best_f1_score: 0.9084; train_kl_reg_loss: 2432.5543; train_edge_recon_loss: 262633.9834; train_gene_expr_recon_loss: 574170.6602; train_masked_gp_l1_reg_loss: 1055.7064; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 840292.9121; train_optim_loss: 840292.9121; val_kl_reg_loss: 2425.6481; val_edge_recon_loss: 266806.5234; val_gene_expr_recon_loss: 572023.3281; val_masked_gp_l1_reg_loss: 1044.8151; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 842300.2969; val_optim_loss: 842300.2969\n",
      "Epoch 10/400 |--------------------| 2.5% val_auroc_score: 0.9457; val_auprc_score: 0.9213; val_best_acc_score: 0.9080; val_best_f1_score: 0.9124; train_kl_reg_loss: 2407.7751; train_edge_recon_loss: 264142.7388; train_gene_expr_recon_loss: 572561.2793; train_masked_gp_l1_reg_loss: 1051.2949; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 840163.0879; train_optim_loss: 840163.0879; val_kl_reg_loss: 2439.5801; val_edge_recon_loss: 263844.3750; val_gene_expr_recon_loss: 570945.7656; val_masked_gp_l1_reg_loss: 1051.7688; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 838281.4844; val_optim_loss: 838281.4844\n",
      "Epoch 11/400 |--------------------| 2.8% val_auroc_score: 0.9453; val_auprc_score: 0.9247; val_best_acc_score: 0.9045; val_best_f1_score: 0.9096; train_kl_reg_loss: 2352.8961; train_edge_recon_loss: 262398.8447; train_gene_expr_recon_loss: 571292.7207; train_masked_gp_l1_reg_loss: 1052.1151; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 837096.5742; train_optim_loss: 837096.5742; val_kl_reg_loss: 2405.8726; val_edge_recon_loss: 263895.6172; val_gene_expr_recon_loss: 569087.9531; val_masked_gp_l1_reg_loss: 1057.2078; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 836446.6406; val_optim_loss: 836446.6406\n",
      "Epoch 12/400 |--------------------| 3.0% val_auroc_score: 0.9463; val_auprc_score: 0.9205; val_best_acc_score: 0.9144; val_best_f1_score: 0.9192; train_kl_reg_loss: 2372.7724; train_edge_recon_loss: 261365.8008; train_gene_expr_recon_loss: 569545.5820; train_masked_gp_l1_reg_loss: 1050.3095; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 834334.4668; train_optim_loss: 834334.4668; val_kl_reg_loss: 2357.0897; val_edge_recon_loss: 265663.2188; val_gene_expr_recon_loss: 568175.9844; val_masked_gp_l1_reg_loss: 1052.2902; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 837248.5938; val_optim_loss: 837248.5938\n",
      "Epoch 13/400 |--------------------| 3.2% val_auroc_score: 0.9517; val_auprc_score: 0.9283; val_best_acc_score: 0.9174; val_best_f1_score: 0.9209; train_kl_reg_loss: 2366.7489; train_edge_recon_loss: 260355.6699; train_gene_expr_recon_loss: 567777.1289; train_masked_gp_l1_reg_loss: 1047.0946; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 831546.6426; train_optim_loss: 831546.6426; val_kl_reg_loss: 2457.6057; val_edge_recon_loss: 261525.3906; val_gene_expr_recon_loss: 566623.1562; val_masked_gp_l1_reg_loss: 1044.1880; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 831650.3438; val_optim_loss: 831650.3438\n",
      "Epoch 14/400 |--------------------| 3.5% val_auroc_score: 0.9492; val_auprc_score: 0.9254; val_best_acc_score: 0.9149; val_best_f1_score: 0.9191; train_kl_reg_loss: 2407.9777; train_edge_recon_loss: 260404.8667; train_gene_expr_recon_loss: 566501.7344; train_masked_gp_l1_reg_loss: 1050.0998; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 830364.6777; train_optim_loss: 830364.6777; val_kl_reg_loss: 2377.7974; val_edge_recon_loss: 263257.0156; val_gene_expr_recon_loss: 565007.3594; val_masked_gp_l1_reg_loss: 1059.8125; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 831702.0000; val_optim_loss: 831702.0000\n",
      "Epoch 15/400 |--------------------| 3.8% val_auroc_score: 0.9504; val_auprc_score: 0.9263; val_best_acc_score: 0.9215; val_best_f1_score: 0.9258; train_kl_reg_loss: 2368.2420; train_edge_recon_loss: 261018.3877; train_gene_expr_recon_loss: 565399.9434; train_masked_gp_l1_reg_loss: 1052.6074; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 829839.1758; train_optim_loss: 829839.1758; val_kl_reg_loss: 2351.6630; val_edge_recon_loss: 262775.5664; val_gene_expr_recon_loss: 563701.8281; val_masked_gp_l1_reg_loss: 1056.4982; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 829885.5625; val_optim_loss: 829885.5625\n",
      "Epoch 16/400 |--------------------| 4.0% val_auroc_score: 0.9532; val_auprc_score: 0.9304; val_best_acc_score: 0.9223; val_best_f1_score: 0.9263; train_kl_reg_loss: 2372.2900; train_edge_recon_loss: 260775.9458; train_gene_expr_recon_loss: 563881.7324; train_masked_gp_l1_reg_loss: 1057.9905; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 828087.9473; train_optim_loss: 828087.9473; val_kl_reg_loss: 2350.2505; val_edge_recon_loss: 261306.1484; val_gene_expr_recon_loss: 562579.3750; val_masked_gp_l1_reg_loss: 1054.6464; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 827290.3906; val_optim_loss: 827290.3906\n",
      "Epoch 17/400 |--------------------| 4.2% val_auroc_score: 0.9487; val_auprc_score: 0.9261; val_best_acc_score: 0.9141; val_best_f1_score: 0.9189; train_kl_reg_loss: 2371.1792; train_edge_recon_loss: 261073.7842; train_gene_expr_recon_loss: 562623.3672; train_masked_gp_l1_reg_loss: 1060.4599; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 827128.7891; train_optim_loss: 827128.7891; val_kl_reg_loss: 2422.9689; val_edge_recon_loss: 263614.5625; val_gene_expr_recon_loss: 561438.5938; val_masked_gp_l1_reg_loss: 1056.5266; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 828532.6250; val_optim_loss: 828532.6250\n",
      "Epoch 18/400 |--------------------| 4.5% val_auroc_score: 0.9539; val_auprc_score: 0.9336; val_best_acc_score: 0.9239; val_best_f1_score: 0.9281; train_kl_reg_loss: 2361.0732; train_edge_recon_loss: 260186.4204; train_gene_expr_recon_loss: 561472.8848; train_masked_gp_l1_reg_loss: 1060.1326; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 825080.5059; train_optim_loss: 825080.5059; val_kl_reg_loss: 2393.9082; val_edge_recon_loss: 261054.3906; val_gene_expr_recon_loss: 559635.4375; val_masked_gp_l1_reg_loss: 1062.7124; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 824146.4219; val_optim_loss: 824146.4219\n",
      "Epoch 19/400 |--------------------| 4.8% val_auroc_score: 0.9519; val_auprc_score: 0.9314; val_best_acc_score: 0.9202; val_best_f1_score: 0.9244; train_kl_reg_loss: 2358.5789; train_edge_recon_loss: 260136.5947; train_gene_expr_recon_loss: 560102.6562; train_masked_gp_l1_reg_loss: 1063.0912; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 823660.9219; train_optim_loss: 823660.9219; val_kl_reg_loss: 2353.0903; val_edge_recon_loss: 261184.8359; val_gene_expr_recon_loss: 558612.6250; val_masked_gp_l1_reg_loss: 1066.0863; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 823216.6094; val_optim_loss: 823216.6094\n",
      "Epoch 20/400 |-------------------| 5.0% val_auroc_score: 0.9464; val_auprc_score: 0.9197; val_best_acc_score: 0.9186; val_best_f1_score: 0.9233; train_kl_reg_loss: 2361.7765; train_edge_recon_loss: 260203.6714; train_gene_expr_recon_loss: 558910.0508; train_masked_gp_l1_reg_loss: 1063.1120; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 822538.6016; train_optim_loss: 822538.6016; val_kl_reg_loss: 2342.5420; val_edge_recon_loss: 262071.4961; val_gene_expr_recon_loss: 558247.2500; val_masked_gp_l1_reg_loss: 1063.5778; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 823724.8594; val_optim_loss: 823724.8594\n",
      "Epoch 21/400 |-------------------| 5.2% val_auroc_score: 0.9557; val_auprc_score: 0.9383; val_best_acc_score: 0.9250; val_best_f1_score: 0.9288; train_kl_reg_loss: 2347.5960; train_edge_recon_loss: 260044.4927; train_gene_expr_recon_loss: 557895.1855; train_masked_gp_l1_reg_loss: 1064.0764; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 821351.3535; train_optim_loss: 821351.3535; val_kl_reg_loss: 2378.8864; val_edge_recon_loss: 259616.3672; val_gene_expr_recon_loss: 556926.2812; val_masked_gp_l1_reg_loss: 1065.3577; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 819986.9062; val_optim_loss: 819986.9062\n",
      "Epoch 22/400 |-------------------| 5.5% val_auroc_score: 0.9533; val_auprc_score: 0.9327; val_best_acc_score: 0.9213; val_best_f1_score: 0.9259; train_kl_reg_loss: 2332.0674; train_edge_recon_loss: 259638.9624; train_gene_expr_recon_loss: 556652.3750; train_masked_gp_l1_reg_loss: 1066.1994; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 819689.6172; train_optim_loss: 819689.6172; val_kl_reg_loss: 2438.3687; val_edge_recon_loss: 262579.3047; val_gene_expr_recon_loss: 555087.2656; val_masked_gp_l1_reg_loss: 1065.1792; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 821170.1406; val_optim_loss: 821170.1406\n",
      "Epoch 23/400 |-------------------| 5.8% val_auroc_score: 0.9537; val_auprc_score: 0.9323; val_best_acc_score: 0.9246; val_best_f1_score: 0.9291; train_kl_reg_loss: 2319.6038; train_edge_recon_loss: 258892.6797; train_gene_expr_recon_loss: 555548.1680; train_masked_gp_l1_reg_loss: 1067.0930; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 817827.5352; train_optim_loss: 817827.5352; val_kl_reg_loss: 2416.1537; val_edge_recon_loss: 260385.4180; val_gene_expr_recon_loss: 554893.2031; val_masked_gp_l1_reg_loss: 1063.0497; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 818757.8438; val_optim_loss: 818757.8438\n",
      "Epoch 24/400 |-------------------| 6.0% val_auroc_score: 0.9528; val_auprc_score: 0.9311; val_best_acc_score: 0.9236; val_best_f1_score: 0.9283; train_kl_reg_loss: 2363.6858; train_edge_recon_loss: 260758.6792; train_gene_expr_recon_loss: 554748.1289; train_masked_gp_l1_reg_loss: 1059.9747; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 818930.4805; train_optim_loss: 818930.4805; val_kl_reg_loss: 2356.5015; val_edge_recon_loss: 260369.8555; val_gene_expr_recon_loss: 553518.7812; val_masked_gp_l1_reg_loss: 1056.6356; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 817301.7500; val_optim_loss: 817301.7500\n",
      "Epoch 25/400 |-------------------| 6.2% val_auroc_score: 0.9562; val_auprc_score: 0.9387; val_best_acc_score: 0.9261; val_best_f1_score: 0.9300; train_kl_reg_loss: 2355.9675; train_edge_recon_loss: 260058.2300; train_gene_expr_recon_loss: 553271.8633; train_masked_gp_l1_reg_loss: 1064.4313; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 816750.4805; train_optim_loss: 816750.4805; val_kl_reg_loss: 2374.0652; val_edge_recon_loss: 258843.1172; val_gene_expr_recon_loss: 552422.4844; val_masked_gp_l1_reg_loss: 1062.5800; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 814702.2344; val_optim_loss: 814702.2344\n",
      "Epoch 26/400 |-------------------| 6.5% val_auroc_score: 0.9509; val_auprc_score: 0.9251; val_best_acc_score: 0.9217; val_best_f1_score: 0.9265; train_kl_reg_loss: 2266.9915; train_edge_recon_loss: 259561.6587; train_gene_expr_recon_loss: 516264.7168; train_masked_gp_l1_reg_loss: 1035.9214; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 779129.2852; train_optim_loss: 779129.2852; val_kl_reg_loss: 2109.6649; val_edge_recon_loss: 262474.5820; val_gene_expr_recon_loss: 514705.8750; val_masked_gp_l1_reg_loss: 1013.1900; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 780303.3125; val_optim_loss: 780303.3125\n",
      "Epoch 27/400 |-------------------| 6.8% val_auroc_score: 0.9524; val_auprc_score: 0.9332; val_best_acc_score: 0.9200; val_best_f1_score: 0.9249; train_kl_reg_loss: 2081.0554; train_edge_recon_loss: 258749.1338; train_gene_expr_recon_loss: 515195.1592; train_masked_gp_l1_reg_loss: 1007.9733; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 777033.3242; train_optim_loss: 777033.3242; val_kl_reg_loss: 1987.1713; val_edge_recon_loss: 261521.2539; val_gene_expr_recon_loss: 514097.7969; val_masked_gp_l1_reg_loss: 1003.0186; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 778609.2500; val_optim_loss: 778609.2500\n",
      "Epoch 28/400 |-------------------| 7.0% val_auroc_score: 0.9500; val_auprc_score: 0.9274; val_best_acc_score: 0.9244; val_best_f1_score: 0.9287; train_kl_reg_loss: 1986.6558; train_edge_recon_loss: 258750.0601; train_gene_expr_recon_loss: 513851.0645; train_masked_gp_l1_reg_loss: 1005.2730; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 775593.0508; train_optim_loss: 775593.0508; val_kl_reg_loss: 1868.5413; val_edge_recon_loss: 260784.1016; val_gene_expr_recon_loss: 512087.8281; val_masked_gp_l1_reg_loss: 1008.9714; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 775749.4688; val_optim_loss: 775749.4688\n",
      "Epoch 29/400 |-------------------| 7.2% val_auroc_score: 0.9558; val_auprc_score: 0.9382; val_best_acc_score: 0.9273; val_best_f1_score: 0.9313; train_kl_reg_loss: 1934.6313; train_edge_recon_loss: 257289.5542; train_gene_expr_recon_loss: 511582.5137; train_masked_gp_l1_reg_loss: 1002.6456; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 771809.3457; train_optim_loss: 771809.3457; val_kl_reg_loss: 1888.8553; val_edge_recon_loss: 259652.8711; val_gene_expr_recon_loss: 510213.6328; val_masked_gp_l1_reg_loss: 1006.9166; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 772762.2969; val_optim_loss: 772762.2969\n",
      "Epoch 30/400 |-------------------| 7.5% val_auroc_score: 0.9527; val_auprc_score: 0.9319; val_best_acc_score: 0.9264; val_best_f1_score: 0.9308; train_kl_reg_loss: 1919.8916; train_edge_recon_loss: 259648.6851; train_gene_expr_recon_loss: 510452.9033; train_masked_gp_l1_reg_loss: 1006.9692; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 773028.4531; train_optim_loss: 773028.4531; val_kl_reg_loss: 1800.3836; val_edge_recon_loss: 260517.8555; val_gene_expr_recon_loss: 509009.0938; val_masked_gp_l1_reg_loss: 1012.2908; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 772339.6406; val_optim_loss: 772339.6406\n",
      "Epoch 31/400 |-------------------| 7.8% val_auroc_score: 0.9548; val_auprc_score: 0.9332; val_best_acc_score: 0.9277; val_best_f1_score: 0.9315; train_kl_reg_loss: 1873.4367; train_edge_recon_loss: 260172.1060; train_gene_expr_recon_loss: 509456.5820; train_masked_gp_l1_reg_loss: 1009.8993; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 772512.0352; train_optim_loss: 772512.0352; val_kl_reg_loss: 1800.5013; val_edge_recon_loss: 259310.3984; val_gene_expr_recon_loss: 508330.5625; val_masked_gp_l1_reg_loss: 1002.0140; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 770443.4531; val_optim_loss: 770443.4531\n",
      "Epoch 32/400 |-------------------| 8.0% val_auroc_score: 0.9561; val_auprc_score: 0.9335; val_best_acc_score: 0.9261; val_best_f1_score: 0.9304; train_kl_reg_loss: 1850.5826; train_edge_recon_loss: 258476.7417; train_gene_expr_recon_loss: 508677.2041; train_masked_gp_l1_reg_loss: 1007.7157; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 770012.2402; train_optim_loss: 770012.2402; val_kl_reg_loss: 1826.2032; val_edge_recon_loss: 259883.8867; val_gene_expr_recon_loss: 507411.9297; val_masked_gp_l1_reg_loss: 1007.3016; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 770129.3438; val_optim_loss: 770129.3438\n",
      "Epoch 33/400 |-------------------| 8.2% val_auroc_score: 0.9494; val_auprc_score: 0.9238; val_best_acc_score: 0.9209; val_best_f1_score: 0.9258; train_kl_reg_loss: 1865.5602; train_edge_recon_loss: 258867.2388; train_gene_expr_recon_loss: 508040.7686; train_masked_gp_l1_reg_loss: 1009.8733; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 769783.4375; train_optim_loss: 769783.4375; val_kl_reg_loss: 1768.4384; val_edge_recon_loss: 261574.3242; val_gene_expr_recon_loss: 507093.2188; val_masked_gp_l1_reg_loss: 1003.7825; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 771439.7812; val_optim_loss: 771439.7812\n",
      "Epoch 34/400 |-------------------| 8.5% val_auroc_score: 0.9517; val_auprc_score: 0.9279; val_best_acc_score: 0.9239; val_best_f1_score: 0.9284; train_kl_reg_loss: 1853.9779; train_edge_recon_loss: 258983.3945; train_gene_expr_recon_loss: 507481.4346; train_masked_gp_l1_reg_loss: 1008.3659; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 769327.1719; train_optim_loss: 769327.1719; val_kl_reg_loss: 1795.2569; val_edge_recon_loss: 260288.4336; val_gene_expr_recon_loss: 506972.2266; val_masked_gp_l1_reg_loss: 1013.7602; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 770069.6562; val_optim_loss: 770069.6562\n",
      "Epoch 35/400 |-------------------| 8.8% val_auroc_score: 0.9542; val_auprc_score: 0.9322; val_best_acc_score: 0.9276; val_best_f1_score: 0.9319; train_kl_reg_loss: 1838.1810; train_edge_recon_loss: 258863.5259; train_gene_expr_recon_loss: 506287.8662; train_masked_gp_l1_reg_loss: 1009.2216; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 767998.7949; train_optim_loss: 767998.7949; val_kl_reg_loss: 1744.7379; val_edge_recon_loss: 258951.3438; val_gene_expr_recon_loss: 505204.3516; val_masked_gp_l1_reg_loss: 1010.5911; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 766911.0000; val_optim_loss: 766911.0000\n",
      "Epoch 36/400 |-------------------| 9.0% val_auroc_score: 0.9573; val_auprc_score: 0.9378; val_best_acc_score: 0.9263; val_best_f1_score: 0.9306; train_kl_reg_loss: 1820.7479; train_edge_recon_loss: 259047.3789; train_gene_expr_recon_loss: 505803.6709; train_masked_gp_l1_reg_loss: 1007.0599; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 767678.8535; train_optim_loss: 767678.8535; val_kl_reg_loss: 1777.9794; val_edge_recon_loss: 258202.6250; val_gene_expr_recon_loss: 504973.6641; val_masked_gp_l1_reg_loss: 1001.7505; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 765956.0312; val_optim_loss: 765956.0312\n",
      "Epoch 37/400 |-------------------| 9.2% val_auroc_score: 0.9583; val_auprc_score: 0.9395; val_best_acc_score: 0.9278; val_best_f1_score: 0.9318; train_kl_reg_loss: 1849.9161; train_edge_recon_loss: 257929.4736; train_gene_expr_recon_loss: 505151.0010; train_masked_gp_l1_reg_loss: 1007.4364; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 765937.8340; train_optim_loss: 765937.8340; val_kl_reg_loss: 1804.3960; val_edge_recon_loss: 258823.6328; val_gene_expr_recon_loss: 504014.0000; val_masked_gp_l1_reg_loss: 1015.4738; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 765657.5156; val_optim_loss: 765657.5156\n",
      "Epoch 38/400 |-------------------| 9.5% val_auroc_score: 0.9559; val_auprc_score: 0.9382; val_best_acc_score: 0.9243; val_best_f1_score: 0.9287; train_kl_reg_loss: 1830.9491; train_edge_recon_loss: 257085.6006; train_gene_expr_recon_loss: 504358.8594; train_masked_gp_l1_reg_loss: 1012.2199; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 764287.6270; train_optim_loss: 764287.6270; val_kl_reg_loss: 1809.7386; val_edge_recon_loss: 260233.1875; val_gene_expr_recon_loss: 502930.5078; val_masked_gp_l1_reg_loss: 1018.6754; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 765992.1094; val_optim_loss: 765992.1094\n",
      "Epoch 39/400 |-------------------| 9.8% val_auroc_score: 0.9532; val_auprc_score: 0.9306; val_best_acc_score: 0.9283; val_best_f1_score: 0.9327; train_kl_reg_loss: 1840.2605; train_edge_recon_loss: 258681.2075; train_gene_expr_recon_loss: 503503.5957; train_masked_gp_l1_reg_loss: 1016.9231; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 765041.9844; train_optim_loss: 765041.9844; val_kl_reg_loss: 1768.2192; val_edge_recon_loss: 259122.4297; val_gene_expr_recon_loss: 502580.1719; val_masked_gp_l1_reg_loss: 1012.6366; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 764483.4531; val_optim_loss: 764483.4531\n",
      "Epoch 40/400 |------------------| 10.0% val_auroc_score: 0.9549; val_auprc_score: 0.9314; val_best_acc_score: 0.9277; val_best_f1_score: 0.9317; train_kl_reg_loss: 1825.4255; train_edge_recon_loss: 258858.7778; train_gene_expr_recon_loss: 502389.6816; train_masked_gp_l1_reg_loss: 1017.4709; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 764091.3535; train_optim_loss: 764091.3535; val_kl_reg_loss: 1797.1089; val_edge_recon_loss: 258432.2070; val_gene_expr_recon_loss: 502166.1797; val_masked_gp_l1_reg_loss: 1007.5406; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 763403.0781; val_optim_loss: 763403.0781\n",
      "Epoch 41/400 |------------------| 10.2% val_auroc_score: 0.9558; val_auprc_score: 0.9310; val_best_acc_score: 0.9259; val_best_f1_score: 0.9302; train_kl_reg_loss: 1824.7262; train_edge_recon_loss: 259272.3374; train_gene_expr_recon_loss: 501968.1123; train_masked_gp_l1_reg_loss: 1017.5919; train_group_lasso_reg_loss: 0.0000; train_addon_gp_l1_reg_loss: 0.0000; train_global_loss: 764082.7656; train_optim_loss: 764082.7656; val_kl_reg_loss: 1721.7434; val_edge_recon_loss: 258803.3516; val_gene_expr_recon_loss: 501699.2734; val_masked_gp_l1_reg_loss: 1022.4436; val_group_lasso_reg_loss: 0.0000; val_addon_gp_l1_reg_loss: 0.0000; val_global_loss: 763246.7969; val_optim_loss: 763246.7969\n"
     ]
    }
   ],
   "source": [
    "### ### Define model configuration ###\n",
    "#################################\n",
    "ablation = \"loss_ablation_extended\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [300., 3000., 3000., 300., 30000.]\n",
    "config[\"lamba_l1_addon\"] = [300., 3000., 300., 3000., 30000.]\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b2fd48-14f1-4f21-86b1-6e240d18ff51",
   "metadata": {},
   "source": [
    "### 2.3 Encoder Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa707ec5-06a3-47c4-b821-bc6dd00e1bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ### Define model configuration ###\n",
    "#################################\n",
    "ablation = \"encoder_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_neighbors\"] = [6] * 2\n",
    "config[\"conv_layer_encoder\"] = [\"gcnconv\", \"gatv2conv\"]\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 2\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 2\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 2\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 2\n",
    "config[\"lambda_l1_masked\"] = [0.] * 2\n",
    "config[\"lamba_l1_addon\"] = [0.] * 2\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d594652-f345-41c3-84a4-7c7384e020db",
   "metadata": {},
   "source": [
    "### 2.4 Neighbor Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca1a6f5a-cb97-4efd-819b-c92a71a421a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "### ### Define model configuration ###\n",
    "#################################\n",
    "ablation = \"neighbor_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_neighbors\"] = [4, 8, 12, 16, 20]\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lamba_l1_addon\"] = [0.] * 5\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaf904cf-3d60-4150-93cb-d24d28b88080",
   "metadata": {},
   "source": [
    "## 4. Visualize Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9de86c4-d578-4437-9d0a-6bf0e3aeff1f",
   "metadata": {},
   "source": [
    "### 4.1 Loss Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15aad70d-9112-477f-9826-939f9c89e87b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"loss_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")\n",
    "#run_df2 = pd.read_csv(f\"{artifacts_folder_path}/ablation/loss_ablation_results_4.csv\")\n",
    "#run_df = pd.concat([run_df, run_df2])\n",
    "#run_df = run_df.drop_duplicates()\n",
    "#run_df.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b2cbb8f-c756-4754-84a1-3f25ec1ea8d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"Loss\"\n",
    "title = \"Edge & Gene Expression Reconstruction Ablation\"\n",
    "\n",
    "mapping_dict = {\"500000.0_300.0\": \"Balanced Edge & Gene Expr Recon\",\n",
    "                \"0.0_300.0\": \"Only Gene Expr Recon\",\n",
    "                \"500000.0_0.0\": \"Only Edge Recon\",\n",
    "                \"50000.0_300.0\": \"Weak Edge Recon\",\n",
    "                \"500000.0_30.0\": \"Weak Gene Expr Recon\"\n",
    "               }\n",
    "\n",
    "col1 = \"lambda_edge_recon\"\n",
    "col2 = \"lambda_gene_expr_recon\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get((str(row[col1]) + \"_\" + str(row[col2])), \"NA\")\n",
    "\n",
    "run_df[\"Loss\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top3\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top3\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5db79495-097e-4c75-b11e-7a70fc3b60af",
   "metadata": {},
   "source": [
    "### 4.1.2 Loss Ablation Extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "423208de-ff47-461e-9860-492bfd54f936",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"loss_ablation_extended\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c01d7ba9-9d99-403b-b89c-b8a597c50cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48216c46-bb0c-45bd-a14a-a1a083b4e417",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"Loss\"\n",
    "title = \"Gene Expr Regularization Ablation\"\n",
    "\n",
    "mapping_dict = {\"50.0_50.0\": \"Light Prior & Light De-novo Reg\",\n",
    "                \"0.0_50.0\": \"Light De-novo Reg\",\n",
    "                \"0.0_300.0\": \"Strong De-novo Reg\",\n",
    "                \"50.0_300.0\": \"Light Prior & Strong De-novo Reg\",\n",
    "                \"30.0_30.0\": \"Very Light Prior & De-novo Reg\"\n",
    "               }\n",
    "\n",
    "col1 = \"lambda_l1_masked\"\n",
    "col2 = \"lamba_l1_addon\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get((str(row[col1]) + \"_\" + str(row[col2])), \"NA\")\n",
    "\n",
    "run_df[\"Loss\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top3\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top3\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a1ba922-d4d5-4e00-acd9-81e0e5eb67cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_df.columns.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee9ad5e4-b900-465e-958a-70bc2b9f5a84",
   "metadata": {},
   "source": [
    "## 3. Miscellaneous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a01117bf-2c29-421a-b974-63d4aab14603",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"loss_ablation\"\n",
    "\n",
    "df_list = []\n",
    "for file_path in [\n",
    "    \"./loss_ablation_results_save.csv\",\n",
    "    \"./loss_ablation_results_save2.csv\",\n",
    "    \"./loss_ablation_results_save3.csv\", \n",
    "    \"./loss_ablation_results_save4.csv\",\n",
    "]:\n",
    "    df_list.append(pd.read_csv(file_path))\n",
    "\n",
    "df = pd.concat(df_list, ignore_index=True)\n",
    "df.to_csv(f\"{artifacts_folder_path}/data_simulation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "855db9e0-6944-4ae0-a4ee-fbe05ef4d791",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_active_gp_scores_to_obs()\n",
    "\n",
    "gp = \"Spag11b_ligand_receptor_target_gene_GP\"\n",
    "gp_summary_df[gp_summary_df[\"gp_name\"] == gp]\n",
    "# Visualize niche types in tissue\n",
    "sq.pl.spatial_scatter(model.adata, color=gp, shape=None) \n",
    "# Visualize niche types in tissue\n",
    "sq.pl.spatial_scatter(model.adata, color=\"Spag11b\", shape=None) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4ea27c-ecb2-4414-88f3-158566db9d45",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_df = pd.DataFrame([[nmi,hom,com,chaos,pas,asw,moranI,gearyC]],\n",
    "               index = [\"NicheCompass\"],#your method name\n",
    "               columns=[['Accuracy','Accuracy','Accuracy','Continuity','Continuity','Continuity','Marker score','Marker score'],\n",
    "                        ['NMI','HOM','COM','CHAOS','PAS','ASW','Moran\\'I','Geary\\'s C']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8f504bd-62e0-4e2f-9ed0-b0914b73af7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_eval_df[\"incremented_prior_gps\"][\"Region5\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fcdf7f5-0990-45a1-a110-7c0f12d91cd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_eval_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dae2e02-c014-47b9-a3b8-93b5af7a7378",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_eval_df[\"incremented_denovo_gps\"][\"Region5\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "662ac80d-e71d-4bb7-b8fc-e368c76482f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_label = f\"log_bayes_factor_{log_bayes_factor_thresh}_all_vs_rest\"\n",
    "save_figs = True\n",
    "\n",
    "generate_enriched_gp_info_plots(\n",
    "    plot_label=plot_label,\n",
    "    model=model,\n",
    "    sample_key=sample_key,\n",
    "    differential_gp_test_results_key=differential_gp_test_results_key,\n",
    "    cat_key=latent_cluster_key,\n",
    "    cat_palette=latent_cluster_colors,\n",
    "    n_top_enriched_gp_start_idx=0,\n",
    "    n_top_enriched_gp_end_idx=10,\n",
    "    feature_spaces=samples, # [\"latent\"]\n",
    "    n_top_genes_per_gp=3,\n",
    "    save_figs=save_figs,\n",
    "    figure_folder_path=f\"{figure_folder_path}/\",\n",
    "    spot_size=spot_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53de48ce-e6c7-4410-aa47-bebb15dde367",
   "metadata": {},
   "outputs": [],
   "source": [
    "network_df = compute_communication_gp_network(\n",
    "    gp_list=[gp],\n",
    "    model=model,\n",
    "    group_key=latent_cluster_key,\n",
    "    n_neighbors=n_neighbors)\n",
    "\n",
    "visualize_communication_gp_network(\n",
    "    adata=model.adata,\n",
    "    network_df=network_df,\n",
    "    figsize=(3, 7),\n",
    "    cat_colors=latent_cluster_colors,\n",
    "    edge_type_colors=[\"#1f77b4\"], \n",
    "    cat_key=latent_cluster_key,\n",
    "    save=True,\n",
    "    save_path=f\"{figure_folder_path}/gp_network_{gp_name}.svg\",\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
