{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1b1daabf-c8e3-43d4-aa32-9789abcc8619",
   "metadata": {},
   "source": [
    "# Simulation Model Ablation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c325d921-2077-42b3-90a8-6f6c7dd928c5",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>)\n",
    "- **Date of Creation:** 26.06.2024\n",
    "- **Date of Last Modification:** 18.08.2024"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d464ebf-354f-4f61-8040-c3dcbcc893b9",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94c7551a-013a-444f-8e2c-7d94290d6388",
   "metadata": {},
   "source": [
    "Run this notebook in the nichecompass-reproducibility environment, installable from ```('../../../envs/environment.yaml')```.\n",
    "\n",
    "Before running this notebook:\n",
    "- Run the data simulation notebook at ```('../data_simulation/data_simulation.ipynb')```\n",
    "- Clone SDMBench from https://github.com/zhaofangyuan98/SDMBench.git into ```('../benchmarking')``` (some slight modifications to the SDMBench source code were necessary to remove technical bugs).\n",
    "    - Move _compute_CHAOS function into compute_CHAOS\n",
    "    - Move _compute_PAS function into compute_PAS\n",
    "    - Move fx_kNN function into compute_PAS\n",
    "    - Move fx_1NN function into compute_CHAOS"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5631277-d8d2-4194-a376-280f4f149b7d",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c756075-c6ff-4c0b-8bd4-20885ca56b9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35b00887-6603-4c0a-a834-f427d7610b9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../benchmarking/SDMBench/SDMBench\")\n",
    "sys.path.append(\"../../utils\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4aa3c6a-5a44-4cf1-84a3-5e514457cf3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import random\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import mlflow\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import squidpy as sq\n",
    "from matplotlib import gridspec\n",
    "from nichecompass.models import NicheCompass\n",
    "from nichecompass.utils import (add_gps_from_gp_dict_to_adata,\n",
    "                                compute_communication_gp_network,\n",
    "                                visualize_communication_gp_network,\n",
    "                                create_new_color_dict,\n",
    "                                extract_gp_dict_from_mebocost_ms_interactions,\n",
    "                                extract_gp_dict_from_nichenet_lrt_interactions,\n",
    "                                extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                                filter_and_combine_gp_dict_gps_v2,\n",
    "                                generate_enriched_gp_info_plots)\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "from SDMBench import sdmbench\n",
    "from benchmarking_utils import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f91f345-c297-47cb-8650-c6d9c60a8b7e",
   "metadata": {},
   "source": [
    "### 1.2 Define Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d4e6d9-239f-47c5-a30b-579b16d4c27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Dataset (Fixed) ###\n",
    "species = \"mouse\"\n",
    "spatial_key = \"spatial\"\n",
    "niche_type_key = \"niche_types\"\n",
    "sample_key = \"batch\"\n",
    "\n",
    "### Model ###\n",
    "# AnnData Keys\n",
    "counts_key = \"counts\"\n",
    "adj_key = \"spatial_connectivities\"\n",
    "gp_names_key = \"nichecompass_gp_names\"\n",
    "active_gp_names_key = \"nichecompass_active_gp_names\"\n",
    "gp_targets_mask_key = \"nichecompass_gp_targets\"\n",
    "gp_targets_categories_mask_key = \"nichecompass_gp_targets_categories\"\n",
    "gp_sources_mask_key = \"nichecompass_gp_sources\"\n",
    "gp_sources_categories_mask_key = \"nichecompass_gp_sources_categories\"\n",
    "latent_key = \"nichecompass_latent\"\n",
    "source_genes_idx_key: str=\"nichecompass_source_genes_idx\"\n",
    "target_genes_idx_key: str=\"nichecompass_target_genes_idx\"\n",
    "genes_idx_key: str=\"nichecompass_genes_idx\"\n",
    "\n",
    "# Trainer\n",
    "n_epochs_all_gps = 25\n",
    "lr = 0.001\n",
    "edge_batch_size = 32 #1024\n",
    "use_cuda_if_available = True\n",
    "\n",
    "### Evaluation ###\n",
    "spot_size = 40\n",
    "differential_gp_test_results_key = \"nichecompass_differential_gp_test_results\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efefff3d-4666-4ac4-a0cb-43159dff5823",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Dataset (Variable) ###\n",
    "n_sim = 1\n",
    "n_genes = 1105\n",
    "n_locs = 10000\n",
    "increment_mode = \"strong\"\n",
    "dataset_name = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6c3b7fa-fa59-43dd-81d2-7e116650915f",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05caf2ee-887a-4429-aed6-9a89a22e7062",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\")\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab22fbaf-aeed-464b-8a20-b640b9e055fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get time of notebook execution for timestamping saved artifacts\n",
    "now = datetime.now()\n",
    "current_timestamp = now.strftime(\"%d%m%Y_%H%M%S\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b886225-6b6d-433c-b117-47b9a8f8091f",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36675337-808d-40a9-93fc-d6cfa9473f66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define paths\n",
    "ga_data_folder_path = \"../../datasets/ga_data\"\n",
    "gp_data_folder_path = \"../../datasets/gp_data\"\n",
    "st_data_folder_path = \"../../datasets/st_data\"\n",
    "st_data_gold_folder_path = st_data_folder_path + \"/gold\"\n",
    "omnipath_lr_network_file_path = f\"{gp_data_folder_path}/omnipath_lr_network.csv\"\n",
    "nichenet_lr_network_file_path = f\"{gp_data_folder_path}/nichenet_lr_network_v2_{species}.csv\"\n",
    "nichenet_ligand_target_matrix_file_path = f\"{gp_data_folder_path}/nichenet_ligand_target_matrix_v2_{species}.csv\"\n",
    "mebocost_enzyme_sensor_interactions_folder_path = f\"{gp_data_folder_path}/metabolite_enzyme_sensor_gps\"\n",
    "gene_orthologs_mapping_file_path = f\"{ga_data_folder_path}/human_mouse_gene_orthologs.csv\"\n",
    "artifacts_folder_path = f\"../../artifacts\"\n",
    "model_folder_path = f\"{artifacts_folder_path}/ablation/{current_timestamp}/model\"\n",
    "figure_folder_path = f\"{artifacts_folder_path}/ablation/{current_timestamp}/figures\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756c2e46-6815-4f82-addf-5bfbed664175",
   "metadata": {},
   "source": [
    "### 1.5 Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c9f8f1-d15a-4cc9-b54c-35b14b0b43c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(model_folder_path, exist_ok=True)\n",
    "os.makedirs(figure_folder_path, exist_ok=True)\n",
    "os.makedirs(f\"{artifacts_folder_path}/ablation\", exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83a4e030-3928-4da7-b321-1c74d9a5239c",
   "metadata": {},
   "source": [
    "### 1.6 Define Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b98333b3-5d8c-463f-9b52-3760755fd223",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_gps(species,\n",
    "                omnipath_lr_network_file_path,\n",
    "                nichenet_lr_network_file_path,\n",
    "                nichenet_ligand_target_matrix_file_path,\n",
    "                gene_orthologs_mapping_file_path):\n",
    "    # Retrieve OmniPath GPs (source: ligand genes; target: receptor genes)\n",
    "    omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "        species=species,\n",
    "        load_from_disk=True,\n",
    "        save_to_disk=False,\n",
    "        lr_network_file_path=omnipath_lr_network_file_path,\n",
    "        gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "        plot_gp_gene_count_distributions=False)\n",
    "    \n",
    "    # Retrieve NicheNet GPs (source: ligand genes; target: receptor genes, target genes)\n",
    "    nichenet_gp_dict = extract_gp_dict_from_nichenet_lrt_interactions(\n",
    "        species=species,\n",
    "        version=\"v2\",\n",
    "        load_from_disk=True,\n",
    "        save_to_disk=False,\n",
    "        lr_network_file_path=nichenet_lr_network_file_path,\n",
    "        ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "        gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "        plot_gp_gene_count_distributions=False)\n",
    "    \n",
    "    # Retrieve MEBOCOST GPs (source: enzyme genes; target: sensor genes)\n",
    "    mebocost_gp_dict = extract_gp_dict_from_mebocost_ms_interactions(\n",
    "        dir_path=mebocost_enzyme_sensor_interactions_folder_path,\n",
    "        species=species,\n",
    "        plot_gp_gene_count_distributions=False)\n",
    "    \n",
    "    # Filter and combine GPs\n",
    "    gp_dicts = [omnipath_gp_dict, nichenet_gp_dict, mebocost_gp_dict]\n",
    "    combined_gp_dict = filter_and_combine_gp_dict_gps_v2(\n",
    "        gp_dicts,\n",
    "        verbose=True)\n",
    "    \n",
    "    print(f\"Number of gene programs after filtering and combining: \"\n",
    "          f\"{len(combined_gp_dict)}.\")\n",
    "\n",
    "    return combined_gp_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2fef25c-709f-4cc5-b234-98f7c0c4f3e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_data(combined_gp_dict,\n",
    "              run_dict,\n",
    "              use_prior_gp_mask=True,\n",
    "              dataset_name=dataset_name):\n",
    "    # Load adata\n",
    "    adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/{dataset_name}.h5ad\")\n",
    "    \n",
    "    # Compute spatial neighborhood\n",
    "    sq.gr.spatial_neighbors(adata,\n",
    "                            coord_type=\"generic\",\n",
    "                            spatial_key=spatial_key,\n",
    "                            n_neighs=run_dict[\"n_neighbors\"][-1])\n",
    "    \n",
    "    # Make adjacency matrix symmetric\n",
    "    adata.obsp[adj_key] = (\n",
    "        adata.obsp[adj_key].maximum(\n",
    "            adata.obsp[adj_key].T))\n",
    "    \n",
    "    #combined_gp_dict = combined_gp_dict.copy()\n",
    "    #for gp in combined_gp_dict.keys():\n",
    "    #    if gp not in simulation_gps:\n",
    "    #        del(combined_gp_dict[gp])\n",
    "    \n",
    "    if use_prior_gp_mask:\n",
    "        # Add the GP dictionary as binary masks to the adata\n",
    "        add_gps_from_gp_dict_to_adata(\n",
    "            gp_dict=combined_gp_dict,\n",
    "            adata=adata,\n",
    "            gp_targets_mask_key=gp_targets_mask_key,\n",
    "            gp_targets_categories_mask_key=gp_targets_categories_mask_key,\n",
    "            gp_sources_mask_key=gp_sources_mask_key,\n",
    "            gp_sources_categories_mask_key=gp_sources_categories_mask_key,\n",
    "            gp_names_key=gp_names_key,\n",
    "            min_genes_per_gp=2,\n",
    "            min_source_genes_per_gp=1,\n",
    "            min_target_genes_per_gp=1,\n",
    "            max_genes_per_gp=None,\n",
    "            max_source_genes_per_gp=None,\n",
    "            max_target_genes_per_gp=None)\n",
    "    else:\n",
    "        adata.varm[gp_targets_mask_key] = np.ones((adata.n_vars, run_dict[\"n_latent\"][-1]))\n",
    "        adata.varm[gp_sources_mask_key] = np.ones((adata.n_vars, run_dict[\"n_latent\"][-1]))\n",
    "        adata.uns[gp_names_key] = np.array([f\"FC_GP{i+1}\" for i in range(run_dict[\"n_latent\"][-1])])\n",
    "        adata.uns[source_genes_idx_key] = np.nonzero(\n",
    "            adata.varm[gp_sources_mask_key].sum(axis=1))[0]\n",
    "        adata.uns[target_genes_idx_key] = np.nonzero(\n",
    "            adata.varm[gp_targets_mask_key].sum(axis=1))[0]\n",
    "        adata.uns[genes_idx_key] = np.concatenate(\n",
    "            (adata.uns[target_genes_idx_key],\n",
    "             adata.uns[source_genes_idx_key] + adata.n_vars), axis=0)\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d08320a-547a-4743-bd25-a11c41f16714",
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_model(adata,\n",
    "                run_dict,\n",
    "                mlflow_experiment_id):\n",
    "    # Initialize model\n",
    "    model = NicheCompass(adata,\n",
    "                         counts_key=counts_key,\n",
    "                         adj_key=adj_key,\n",
    "                         gp_names_key=gp_names_key,\n",
    "                         active_gp_names_key=active_gp_names_key,\n",
    "                         gp_targets_mask_key=gp_targets_mask_key,\n",
    "                         gp_targets_categories_mask_key=gp_targets_categories_mask_key,\n",
    "                         gp_sources_mask_key=gp_sources_mask_key,\n",
    "                         gp_sources_categories_mask_key=gp_sources_categories_mask_key,\n",
    "                         latent_key=latent_key,\n",
    "                         n_addon_gp=run_dict[\"n_addon_gp\"][-1],\n",
    "                         conv_layer_encoder=run_dict[\"conv_layer_encoder\"][-1],\n",
    "                         active_gp_thresh_ratio=run_dict[\"active_gp_thresh_ratio\"][-1],\n",
    "                         seed=run_dict[\"model_seed\"][-1])\n",
    "    \n",
    "    # Train model\n",
    "    model.train(n_epochs=run_dict[\"n_epochs\"][-1],\n",
    "                n_epochs_all_gps=n_epochs_all_gps,\n",
    "                lr=lr,\n",
    "                lambda_edge_recon=run_dict[\"lambda_edge_recon\"][-1],\n",
    "                lambda_gene_expr_recon=run_dict[\"lambda_gene_expr_recon\"][-1],\n",
    "                lambda_l1_masked=run_dict[\"lambda_l1_masked\"][-1],\n",
    "                lambda_l1_addon=run_dict[\"lambda_l1_addon\"][-1],\n",
    "                l1_targets_categories=run_dict[\"l1_targets_categories\"][-1],\n",
    "                l1_sources_categories=run_dict[\"l1_sources_categories\"][-1],\n",
    "                edge_batch_size=edge_batch_size,\n",
    "                mlflow_experiment_id=mlflow_experiment_id,\n",
    "                n_sampled_neighbors=run_dict[\"n_neighbors\"][-1],\n",
    "                use_cuda_if_available=use_cuda_if_available,\n",
    "                seed=run_dict[\"model_seed\"][-1],\n",
    "                verbose=True)\n",
    "    \n",
    "    # Compute latent neighbor graph\n",
    "    sc.pp.neighbors(model.adata,\n",
    "                    use_rep=latent_key,\n",
    "                    key_added=latent_key)\n",
    "    \n",
    "    # Compute UMAP embedding\n",
    "    sc.tl.umap(model.adata,\n",
    "               neighbors_key=latent_key)\n",
    "    \n",
    "    # Save trained model\n",
    "    os.makedirs(model_folder_path + f\"/run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\", exist_ok=True)\n",
    "    model.save(dir_path=model_folder_path + f\"/run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\",\n",
    "               overwrite=True,\n",
    "               save_adata=True,\n",
    "               adata_file_name=\"adata.h5ad\")\n",
    "\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed936ca4-3060-4b13-b665-ace0cde89b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_niches(model,\n",
    "                    run_dict,\n",
    "                    figure_folder_path=figure_folder_path,\n",
    "                    latent_leiden_resolution=0.3,\n",
    "                    leiden_resolution_increments=0.1,\n",
    "                    valid_niche_thresh=100):\n",
    "    # Compute Leiden clustering of latent space until 8 niches are obtained (to match ground truth number)\n",
    "    latent_cluster_key = f\"latent_leiden_{str(latent_leiden_resolution)}\"\n",
    "    counter = 0\n",
    "    while True:\n",
    "        sc.tl.leiden(adata=model.adata,\n",
    "                     resolution=latent_leiden_resolution,\n",
    "                     key_added=\"pred_niche_types\",\n",
    "                     neighbors_key=latent_key)\n",
    "        \n",
    "        niche_counts = model.adata.obs[\"pred_niche_types\"].value_counts()\n",
    "        valid_niches = niche_counts[niche_counts >= valid_niche_thresh].index\n",
    "        n_niches = model.adata.obs[model.adata.obs[\"pred_niche_types\"].isin(valid_niches)][\"pred_niche_types\"].nunique()\n",
    "        print(f\"Current number of niches: {n_niches}\")\n",
    "        print(f\"Cluster counter: {counter}\")\n",
    "        if n_niches == 8:\n",
    "            break\n",
    "        elif n_niches < 7 and counter < 30:\n",
    "            print(\"Big increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments\n",
    "        elif n_niches < 8 and counter < 60:\n",
    "            print(\"Slight increase of clustering resolution...\")\n",
    "            latent_leiden_resolution += leiden_resolution_increments/10\n",
    "        elif n_niches > 9 and counter < 30:\n",
    "            print(\"Big decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments\n",
    "        elif n_niches > 8 and counter < 60:\n",
    "            print(\"Slight decrease of clustering resolution...\")\n",
    "            latent_leiden_resolution -= leiden_resolution_increments/10\n",
    "        elif counter > 60:\n",
    "            break\n",
    "        counter += 1\n",
    "    \n",
    "    latent_cluster_colors = create_new_color_dict(\n",
    "        adata=model.adata,\n",
    "        cat_key=\"pred_niche_types\")\n",
    "    \n",
    "    # Create plot of latent cluster / niche annotations in physical and latent space\n",
    "    groups = None # set this to a specific cluster for easy visualization\n",
    "    save_fig = True\n",
    "    file_path = f\"{figure_folder_path}/\" \\\n",
    "                f\"{ablation}_{dataset_name}_run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\" \\\n",
    "                \"niches_latent_physical_space.svg\"\n",
    "    \n",
    "    model.adata.obs[sample_key] = \"batch1\"\n",
    "    samples = model.adata.obs[sample_key].unique().tolist()\n",
    "    \n",
    "    fig = plt.figure(figsize=(12, 14))\n",
    "    title = fig.suptitle(t=f\"NicheCompass Niches \" \\\n",
    "                           \"in Latent and Physical Space\",\n",
    "                         y=0.96,\n",
    "                         x=0.55,\n",
    "                         fontsize=20)\n",
    "    spec1 = gridspec.GridSpec(ncols=1,\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1],\n",
    "                              height_ratios=[3, 2])\n",
    "    spec2 = gridspec.GridSpec(ncols=len(samples),\n",
    "                              nrows=2,\n",
    "                              width_ratios=[1] * len(samples),\n",
    "                              height_ratios=[3, 2])\n",
    "    axs = []\n",
    "    axs.append(fig.add_subplot(spec1[0]))\n",
    "    sc.pl.umap(adata=model.adata,\n",
    "               color=[\"pred_niche_types\"],\n",
    "               groups=groups,\n",
    "               palette=latent_cluster_colors,\n",
    "               title=f\"Niches in Latent Space\",\n",
    "               ax=axs[0],\n",
    "               show=False)\n",
    "    for idx, sample in enumerate(samples):\n",
    "        axs.append(fig.add_subplot(spec2[len(samples) + idx]))\n",
    "        sc.pl.spatial(adata=model.adata[model.adata.obs[sample_key] == sample],\n",
    "                      color=[\"pred_niche_types\"],\n",
    "                      groups=groups,\n",
    "                      palette=latent_cluster_colors,\n",
    "                      spot_size=spot_size,\n",
    "                      title=f\"Niches in Physical Space \\n\"\n",
    "                            f\"(Sample: {sample})\",\n",
    "                      legend_loc=None,\n",
    "                      ax=axs[idx+1],\n",
    "                      show=False)\n",
    "    \n",
    "    # Create and position shared legend\n",
    "    handles, labels = axs[0].get_legend_handles_labels()\n",
    "    lgd = fig.legend(handles,\n",
    "                     labels,\n",
    "                     loc=\"center left\",\n",
    "                     bbox_to_anchor=(0.98, 0.5))\n",
    "    axs[0].get_legend().remove()\n",
    "    \n",
    "    # Adjust, save and display plot\n",
    "    plt.subplots_adjust(wspace=0.2, hspace=0.25)\n",
    "    if save_fig:\n",
    "        fig.savefig(file_path,\n",
    "                    bbox_extra_artists=(lgd, title),\n",
    "                    bbox_inches=\"tight\")\n",
    "    plt.show()\n",
    "\n",
    "    return model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37e37c34-0591-406e-98dd-0a94559b00ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_enriched_gps(model,\n",
    "                     use_prior_gp_mask=True,\n",
    "                     use_ground_truth_gp_numbers=True):\n",
    "    if use_ground_truth_gp_numbers:\n",
    "        # Get number of upregulated GPs per niche\n",
    "        sim_gp_df = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_strongincrements.csv\")\n",
    "        sim_gp_prior_df_agg = sim_gp_df[~sim_gp_df['gp'].str.contains(\"Add-On\")][sim_gp_df['increment_param'] != 1].groupby(\"regionA\").agg(set)\n",
    "        sim_gp_prior_df_agg[\"niche_n_upregulated_prior_gps\"] = sim_gp_prior_df_agg[\"gp\"].apply(lambda x: len(x))\n",
    "        sim_gp_prior_df_agg[\"category\"] = pd.Series(sim_gp_prior_df_agg.index).apply(lambda x: x.replace(\"Region\", \"Niche \")).values\n",
    "    \n",
    "    # Run differential GP testing: each niche vs rest\n",
    "    selected_cats = None\n",
    "    comparison_cats = \"rest\"\n",
    "    enriched_gps = model.run_differential_gp_tests(\n",
    "        cat_key=\"niche_types\",\n",
    "        selected_cats=selected_cats,\n",
    "        comparison_cats=comparison_cats,\n",
    "        log_bayes_factor_thresh=log_bayes_factor_thresh)\n",
    "\n",
    "    # Get enriched prior and de novo GPs with corrected GP sign\n",
    "    gp_summary_df = model.get_gp_summary() \n",
    "    enriched_gps = model.adata.uns[differential_gp_test_results_key]\n",
    "    gp_summary_df[\"gp_sign_source\"] = gp_summary_df[\"gp_source_genes_weights\"].apply(\n",
    "        lambda x: 1 if np.array(x).sum() > 0 else (0 if np.array(x).sum() == 0 else -1))\n",
    "    gp_summary_df[\"gp_sign_target\"] = gp_summary_df[\"gp_target_genes_weights\"].apply(\n",
    "        lambda x: 1 if np.array(x).sum() > 0 else (0 if np.array(x).sum() == 0 else -1))\n",
    "    gp_summary_df[\"gp_sign\"] = gp_summary_df.apply(lambda row: row[\"gp_sign_source\"] if row[\"gp_sign_source\"] != 0 else row[\"gp_sign_target\"], axis=1)\n",
    "    gp_signs = []\n",
    "    for gp in enriched_gps[\"gene_program\"].values.tolist():\n",
    "        gp_signs.append(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_sign\"].values[0])\n",
    "    enriched_gps[\"log_bayes_factor\"] = enriched_gps[\"log_bayes_factor\"] * gp_signs\n",
    "    enriched_gps = enriched_gps[enriched_gps[\"log_bayes_factor\"] > 0]\n",
    "    \n",
    "    prior_enriched_gps = enriched_gps[~enriched_gps[\"gene_program\"].str.contains(\"Add-on\")]\n",
    "    if use_ground_truth_gp_numbers:\n",
    "        prior_enriched_gps = pd.merge(prior_enriched_gps, sim_gp_prior_df_agg, on=\"category\")\n",
    "        prior_enriched_gps = prior_enriched_gps.groupby(\"category\").apply(lambda x: x.head(x['niche_n_upregulated_prior_gps'].iloc[0])).reset_index(drop=True)\n",
    "       \n",
    "    denovo_enriched_gps = enriched_gps[enriched_gps[\"gene_program\"].str.contains(\"Add-on\")]\n",
    "    denovo_enriched_gps = denovo_enriched_gps[denovo_enriched_gps[\"log_bayes_factor\"] > 2.3]\n",
    "    \n",
    "    if use_prior_gp_mask:\n",
    "        # Get enriched prior gene programs for each niche according to differential gene program test\n",
    "        enriched_prior_gps_per_niche = prior_enriched_gps.groupby(\n",
    "            \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "\n",
    "        # Get enriched de-novo gene programs for each niche according to differential gene program test\n",
    "        enriched_denovo_gps_per_niche = denovo_enriched_gps.groupby(\n",
    "            \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "        \n",
    "        return enriched_prior_gps_per_niche, enriched_denovo_gps_per_niche\n",
    "    else:\n",
    "        enriched_gps_per_niche = prior_enriched_gps.groupby(\n",
    "            \"category\")[[\"gene_program\", \"log_bayes_factor\"]].agg(list)\n",
    "        \n",
    "        return enriched_gps_per_niche, enriched_gps_per_niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d43ed3-d261-4e25-ab38-83433403c3fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define eval helper functions\n",
    "def calculate_intersection(row, col1, col2):\n",
    "    return len(row[col1].intersection(row[col2]))\n",
    "\n",
    "def calculate_surplus(row, col1, col2):\n",
    "    return row[col1] - row[col2]\n",
    "\n",
    "def jaccard_index_elementwise(sets1, sets2):\n",
    "    def jaccard_index(set1, set2):\n",
    "        intersection = len(set1 & set2)\n",
    "        union = len(set1 | set2)\n",
    "        if union == 0:\n",
    "            return 0.0\n",
    "        return intersection / union\n",
    "    \n",
    "    return np.array([jaccard_index(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "def precision_score_elementwise(sets1, sets2):\n",
    "    def precision_score(set1, set2):\n",
    "        tp = len(set1 & set2)  # true positives\n",
    "        fp = len(set1 - set2)  # false positives\n",
    "        \n",
    "        if tp == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        precision = tp / (tp + fp) if (tp + fp) > 0 else 0\n",
    "        \n",
    "        return precision\n",
    "    \n",
    "    return np.array([precision_score(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "def recall_score_elementwise(sets1, sets2):\n",
    "    def recall_score(set1, set2):\n",
    "        tp = len(set1 & set2)  # true positives\n",
    "        fn = len(set2 - set1)  # false negatives\n",
    "        \n",
    "        if tp == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        recall = tp / (tp + fn) if (tp + fn) > 0 else 0\n",
    "        \n",
    "        return recall\n",
    "\n",
    "    return np.array([recall_score(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "def f1_score_elementwise(sets1, sets2):\n",
    "    def f1_score(set1, set2):\n",
    "        tp = len(set1 & set2)  # true positives\n",
    "        fp = len(set1 - set2)  # false positives\n",
    "        fn = len(set2 - set1)  # false negatives\n",
    "        \n",
    "        if tp == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        precision = tp / (tp + fp) if (tp + fp) > 0 else 0\n",
    "        recall = tp / (tp + fn) if (tp + fn) > 0 else 0\n",
    "        \n",
    "        if precision + recall == 0:\n",
    "            return 0.0\n",
    "        \n",
    "        return 2 * (precision * recall) / (precision + recall)\n",
    "    \n",
    "    return np.array([f1_score(s1, s2) for s1, s2 in zip(sets1, sets2)])\n",
    "\n",
    "\n",
    "def compute_metrics(model,\n",
    "                    enriched_prior_gps_per_niche,\n",
    "                    enriched_denovo_gps_per_niche,\n",
    "                    run_dict,\n",
    "                    use_prior_gp_mask=True,\n",
    "                    debug=False):\n",
    "    #######################################################################\n",
    "    ### Prepare enriched and incremented GPs for GP recovery evaluation ###\n",
    "    #######################################################################\n",
    "    print(\"Prepare enriched and incremented GPs for GP recovery evaluation...\")\n",
    "\n",
    "    gp_summary_df = model.get_gp_summary()\n",
    "    \n",
    "    gp_eval_dict = {}\n",
    "    for niche_type in sim_gps_stats[\"regionA\"].unique().tolist():\n",
    "        if niche_type in enriched_prior_gps_per_niche.index:\n",
    "            gp_eval_dict[niche_type] = {\"enriched_prior_gps\": set(enriched_prior_gps_per_niche.loc[niche_type][\"gene_program\"])}\n",
    "            source_genes_top3 = []\n",
    "            source_genes_top5 = []\n",
    "            source_genes_top10 = []\n",
    "            target_genes_top3 = []\n",
    "            target_genes_top5 = []\n",
    "            target_genes_top10 = []\n",
    "            for gp, log_bayes_factor in zip(enriched_prior_gps_per_niche.loc[niche_type][\"gene_program\"], enriched_prior_gps_per_niche.loc[niche_type][\"log_bayes_factor\"]):\n",
    "                # Determine gp sign factor\n",
    "                gp_sign_factor = np.sign(log_bayes_factor)\n",
    "                \n",
    "                gp_source_genes_weights = np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_source_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes\"].values[0]\n",
    "                top_gp_source_genes = np.array(gp_source_genes)[np.argsort(gp_source_genes_weights)[::-1][:10]].tolist()\n",
    "                source_genes_top3.extend(top_gp_source_genes[:3])\n",
    "                source_genes_top5.extend(top_gp_source_genes[:5])\n",
    "                source_genes_top10.extend(top_gp_source_genes[:10])\n",
    "            \n",
    "                gp_target_genes_weights = np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_target_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].values[0]\n",
    "                top_gp_target_genes = np.array(gp_target_genes)[np.argsort(gp_target_genes_weights)[::-1][:10]].tolist()\n",
    "                target_genes_top3.extend(top_gp_target_genes[:3])\n",
    "                target_genes_top5.extend(top_gp_target_genes[:5])\n",
    "                target_genes_top10.extend(top_gp_target_genes[:10])\n",
    "            \n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top3\"] = set(source_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top5\"] = set(source_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top10\"] = set(source_genes_top10)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top3\"] = set(target_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top5\"] = set(target_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top10\"] = set(target_genes_top10)\n",
    "        else:\n",
    "            gp_eval_dict[niche_type] = {\"enriched_prior_gps\": set()}\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_source_genes_top10\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_prior_gp_target_genes_top10\"] = set()\n",
    "            \n",
    "        if niche_type in enriched_denovo_gps_per_niche.index:\n",
    "            source_genes_top3 = []\n",
    "            source_genes_top5 = []\n",
    "            source_genes_top10 = []\n",
    "            target_genes_top3 = []\n",
    "            target_genes_top5 = []\n",
    "            target_genes_top10 = []\n",
    "            for gp, log_bayes_factor in zip(enriched_denovo_gps_per_niche.loc[niche_type][\"gene_program\"], enriched_denovo_gps_per_niche.loc[niche_type][\"log_bayes_factor\"]):\n",
    "                # Determine gp sign factor\n",
    "                gp_sign_factor = np.sign(log_bayes_factor)\n",
    "                \n",
    "                gp_source_genes_weights = np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_source_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_source_genes\"].values[0]\n",
    "                top_gp_source_genes = np.array(gp_source_genes)[np.argsort(gp_source_genes_weights)[::-1][:10]].tolist()\n",
    "                source_genes_top3.extend(top_gp_source_genes[:3])\n",
    "                source_genes_top5.extend(top_gp_source_genes[:5])\n",
    "                source_genes_top10.extend(top_gp_source_genes[:10])\n",
    "            \n",
    "                gp_target_genes_weights =  np.array(gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes_weights\"].values[0]) * gp_sign_factor\n",
    "                gp_target_genes = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].values[0]\n",
    "                top_gp_target_genes = np.array(gp_target_genes)[np.argsort(gp_target_genes_weights)[::-1][:10]].tolist()\n",
    "                target_genes_top3.extend(top_gp_target_genes[:3])\n",
    "                target_genes_top5.extend(top_gp_target_genes[:5])\n",
    "                target_genes_top10.extend(top_gp_target_genes[:10])\n",
    "            \n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top3\"] = set(source_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top5\"] = set(source_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top10\"] = set(source_genes_top10)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top3\"] = set(target_genes_top3)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top5\"] = set(target_genes_top5)\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top10\"] = set(target_genes_top10)\n",
    "        else:\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_source_genes_top10\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top3\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top5\"] = set()\n",
    "            gp_eval_dict[niche_type][\"enriched_denovo_gp_target_genes_top10\"] = set()\n",
    "            \n",
    "        for increment_param in increment_params:\n",
    "            if len(sim_gps_stats[\n",
    "            (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gps\"].values[0]) == 0:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gps\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_source_genes\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_target_genes\"] = set()\n",
    "            else:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gps\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gps\"].apply(\n",
    "                    lambda x: [gp for gp in x.split(\",\")]).values[0])\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_source_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gp_sources\"].values[0].split(\",\"))\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_prior_gp_target_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"prior_gp_targets\"].values[0].split(\",\"))\n",
    "    \n",
    "            if len(sim_gps_stats[\n",
    "            (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gps\"].values[0]) == 0:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gps\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_source_genes\"] = set()\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_target_genes\"] = set()\n",
    "            else:\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gps\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gps\"].apply(\n",
    "                    lambda x: [gp for gp in x.split(\",\")]).values[0])\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_source_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gp_sources\"].values[0].split(\",\"))\n",
    "                gp_eval_dict[niche_type][f\"increment_param_{increment_param}_denovo_gp_target_genes\"] = set(sim_gps_stats[\n",
    "                (sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][\"denovo_gp_targets\"].values[0].split(\",\"))\n",
    "    \n",
    "    gp_eval_df = pd.DataFrame(gp_eval_dict.values(), index=gp_eval_dict.keys())\n",
    "    \n",
    "    # Aggregate incremented elements to compute summary metrics\n",
    "    for element_type in [\"prior_gps\",\n",
    "                         \"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        gp_eval_df[f\"incremented_{element_type}\"] = gp_eval_df.apply(lambda row:\n",
    "                                                                     row[f\"increment_param_8_{element_type}\"] |\n",
    "                                                                     row[f\"increment_param_5_{element_type}\"] |\n",
    "                                                                     row[f\"increment_param_3_{element_type}\"], axis=1)\n",
    "\n",
    "    \n",
    "    ########################################\n",
    "    ### Compute metrics for enriched GPs ###\n",
    "    ########################################\n",
    "    print(\"Compute GP recovery metrics for enriched GPs...\")\n",
    "    \n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        run_dict[f\"f1_enriched_{element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "        run_dict[f\"precision_enriched_{element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "        run_dict[f\"recall_enriched_{element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "        run_dict[f\"jaccard_index_enriched_{element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"enriched_{element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "\n",
    " \n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            run_dict[f\"f1_enriched_{top_n_element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            run_dict[f\"precision_enriched_{top_n_element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            run_dict[f\"recall_enriched_{top_n_element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"enriched_{top_n_element_type}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "\n",
    "    \n",
    "    ####################\n",
    "    ### Get GP stats ###\n",
    "    ####################\n",
    "    # Get all genes, prior gps, and prior and de novo gp source and target genes\n",
    "    all_genes = set(model.adata.var_names)\n",
    "    \n",
    "    n_prior_gps = model.adata.varm[gp_sources_mask_key].shape[1]\n",
    "    prior_gps = set(model.adata.uns[gp_names_key][:n_prior_gps])\n",
    "    \n",
    "    if use_prior_gp_mask:\n",
    "        prior_gp_source_gene_indices = np.where(np.any(model.adata.varm[gp_sources_mask_key] == 1, axis=1))[0]\n",
    "        prior_gp_source_genes = set(model.adata.var_names[prior_gp_source_gene_indices])\n",
    "\n",
    "        prior_gp_target_gene_indices = np.where(np.any(model.adata.varm[gp_targets_mask_key] == 1, axis=1))[0]\n",
    "        prior_gp_target_genes = set(model.adata.var_names[prior_gp_target_gene_indices])\n",
    "        \n",
    "    else:\n",
    "        adata_temp = load_data(combined_gp_dict,\n",
    "                               run_dict,\n",
    "                               use_prior_gp_mask=True)\n",
    "        \n",
    "        prior_gp_source_gene_indices = np.where(np.any(adata_temp.varm[gp_sources_mask_key] == 1, axis=1))[0]\n",
    "        prior_gp_source_genes = set(adata_temp.var_names[prior_gp_source_gene_indices])\n",
    "\n",
    "        prior_gp_target_gene_indices = np.where(np.any(adata_temp.varm[gp_targets_mask_key] == 1, axis=1))[0]\n",
    "        prior_gp_target_genes = set(adata_temp.var_names[prior_gp_target_gene_indices])\n",
    "        \n",
    "        del(adata_temp)\n",
    "\n",
    "    denovo_gp_source_genes = all_genes - prior_gp_source_genes\n",
    "    denovo_gp_target_genes = all_genes - prior_gp_target_genes\n",
    "    \n",
    "    \n",
    "    ##################################################\n",
    "    ### Compute metrics for random GPs as baseline ###\n",
    "    ##################################################\n",
    "    print(\"Compute GP recovery metrics for random GPs...\")\n",
    "    random_gp_dict = {}\n",
    "    \n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        random_gp_dict[f\"f1_random_{element_type}\"] = []\n",
    "        random_gp_dict[f\"precision_random_{element_type}\"] = []\n",
    "        random_gp_dict[f\"recall_random_{element_type}\"] = []\n",
    "        random_gp_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            random_gp_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "            random_gp_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "            random_gp_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "            random_gp_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "    \n",
    "    for seed in range(n_seeds):\n",
    "        random.seed(seed)\n",
    "        # Sample random gp from prior knowledge gps so that number matches enriched gps\n",
    "        gp_eval_df[f\"random_prior_gps_seed{seed}\"] = gp_eval_df[\"enriched_prior_gps\"].apply(\n",
    "            lambda x: set(random.sample([gp for gp in prior_gps], len(x))))\n",
    "        \n",
    "        # Sample random genes from prior knowledge gps so that number matches enriched genes from prior gps\n",
    "        for n in [3, 5, 10]:\n",
    "            gp_eval_df[f\"random_prior_gp_target_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_prior_gp_target_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in prior_gp_target_genes], min(len(prior_gp_target_genes), len(x)))))\n",
    "            gp_eval_df[f\"random_prior_gp_source_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_prior_gp_source_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in prior_gp_source_genes], min(len(prior_gp_source_genes), len(x)))))\n",
    "        \n",
    "            # Sample random genes from denovo gps so that number matches enriched genes from denovo gps\n",
    "            gp_eval_df[f\"random_denovo_gp_target_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_denovo_gp_target_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in denovo_gp_target_genes], min(len(denovo_gp_target_genes), len(x)))))\n",
    "            gp_eval_df[f\"random_denovo_gp_source_genes_top{n}_seed{seed}\"]  = gp_eval_df[f\"enriched_denovo_gp_source_genes_top{n}\"].apply(\n",
    "                lambda x: set(random.sample([gene for gene in denovo_gp_source_genes], min(len(denovo_gp_source_genes), len(x)))))\n",
    "\n",
    "        # Aggregate metrics over niches\n",
    "        for element_type in [\"prior_gps\"]:\n",
    "            random_gp_dict[f\"f1_random_{element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            random_gp_dict[f\"precision_random_{element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            random_gp_dict[f\"recall_random_{element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "            random_gp_dict[f\"jaccard_index_random_{element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"random_{element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "\n",
    "        for element_type in [\"prior_gp_source_genes\",\n",
    "                             \"prior_gp_target_genes\",\n",
    "                             \"denovo_gp_source_genes\",\n",
    "                             \"denovo_gp_target_genes\"]:\n",
    "            for n in [3, 5, 10]:\n",
    "                top_n_element_type = element_type + f\"_top{n}\"\n",
    "                random_gp_dict[f\"f1_random_{top_n_element_type}\"].append(np.mean(f1_score_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "                random_gp_dict[f\"precision_random_{top_n_element_type}\"].append(np.mean(precision_score_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "                random_gp_dict[f\"recall_random_{top_n_element_type}\"].append(np.mean(recall_score_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "                random_gp_dict[f\"jaccard_index_random_{top_n_element_type}\"].append(np.mean(jaccard_index_elementwise(gp_eval_df[f\"random_{top_n_element_type}_seed{seed}\"].values, gp_eval_df[f\"incremented_{element_type}\"].values)))\n",
    "    \n",
    "    # Aggregate metrics over random seeds\n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        run_dict[f\"f1_random_{element_type}\"].append(np.mean(random_gp_dict[f\"f1_random_{element_type}\"]))\n",
    "        run_dict[f\"precision_random_{element_type}\"].append(np.mean(random_gp_dict[f\"precision_random_{element_type}\"]))\n",
    "        run_dict[f\"recall_random_{element_type}\"].append(np.mean(random_gp_dict[f\"recall_random_{element_type}\"]))\n",
    "        run_dict[f\"jaccard_index_random_{element_type}\"].append(np.mean(random_gp_dict[f\"jaccard_index_random_{element_type}\"]))\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            run_dict[f\"f1_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"f1_random_{top_n_element_type}\"]))\n",
    "            run_dict[f\"precision_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"precision_random_{top_n_element_type}\"]))\n",
    "            run_dict[f\"recall_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"recall_random_{top_n_element_type}\"]))\n",
    "            run_dict[f\"jaccard_index_random_{top_n_element_type}\"].append(np.mean(random_gp_dict[f\"jaccard_index_random_{top_n_element_type}\"]))\n",
    "\n",
    "\n",
    "    ####################################\n",
    "    ### Compute GP recovery z-scores ###\n",
    "    ####################################\n",
    "    \"\"\"\n",
    "    print(\"Compute GP recovery z-scores...\")\n",
    "    # Compute overlaps between enriched elements and incremented elements\n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        for increment_param in increment_params:\n",
    "            gp_eval_df[f\"enriched_{element_type}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                row, col1=f\"enriched_{element_type}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            for increment_param in increment_params:\n",
    "                gp_eval_df[f\"enriched_{top_n_element_type}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                    row, col1=f\"enriched_{top_n_element_type}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "    # Compute overlaps between random elements and incremented elements\n",
    "    for seed in range(n_seeds):\n",
    "        for element_type in [\"prior_gps\"]:\n",
    "            for increment_param in increment_params:\n",
    "                gp_eval_df[f\"random_{element_type}_seed{seed}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                    row, col1=f\"random_{element_type}_seed{seed}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "        for element_type in [\"prior_gp_source_genes\",\n",
    "                             \"prior_gp_target_genes\",\n",
    "                             \"denovo_gp_source_genes\",\n",
    "                             \"denovo_gp_target_genes\"]:\n",
    "            for n in [3, 5, 10]:\n",
    "                top_n_element_type = element_type + f\"_top{n}\"\n",
    "                for increment_param in increment_params:\n",
    "                    gp_eval_df[f\"random_{top_n_element_type}_seed{seed}_increment_param_{increment_param}_overlap\"] = gp_eval_df.apply(lambda row: calculate_intersection(\n",
    "                        row, col1=f\"random_{top_n_element_type}_seed{seed}\", col2=f\"increment_param_{increment_param}_{element_type}\"), axis=1)\n",
    "\n",
    "    # Compute z-scores\n",
    "    for element_type in [\"prior_gps\"]:\n",
    "        for increment_param in increment_params:\n",
    "            agg_cols = [col for col in gp_eval_df.columns if f\"random_{element_type}_seed\" in col and f\"increment_param_{increment_param}_overlap\" in col]\n",
    "\n",
    "            # Compute means and stds over random seeds\n",
    "            means = gp_eval_df[agg_cols].mean(axis=1)\n",
    "            stds = gp_eval_df[agg_cols].std(axis=1)\n",
    "\n",
    "            # Compute means over regions\n",
    "            run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"].append(\n",
    "                np.mean((gp_eval_df[f\"enriched_{element_type}_increment_param_{increment_param}_overlap\"] - means)/stds))\n",
    "\n",
    "    for element_type in [\"prior_gp_source_genes\",\n",
    "                         \"prior_gp_target_genes\",\n",
    "                         \"denovo_gp_source_genes\",\n",
    "                         \"denovo_gp_target_genes\"]:\n",
    "        for n in [3, 5, 10]:\n",
    "            top_n_element_type = element_type + f\"_top{n}\"\n",
    "            for increment_param in increment_params:\n",
    "                agg_cols = [col for col in gp_eval_df.columns if f\"random_{top_n_element_type}_seed\" in col and f\"increment_param_{increment_param}_overlap\" in col]\n",
    "    \n",
    "                # Compute means and stds over random seeds\n",
    "                means = gp_eval_df[agg_cols].mean(axis=1)\n",
    "                stds = gp_eval_df[agg_cols].std(axis=1)\n",
    "    \n",
    "                # Compute means over regions\n",
    "                run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"].append(\n",
    "                    np.mean((gp_eval_df[f\"enriched_{top_n_element_type}_increment_param_{increment_param}_overlap\"] - means)/stds))\n",
    "    \"\"\"\n",
    "\n",
    "    \n",
    "    ################################################\n",
    "    ### Compute metrics for niche identification ###\n",
    "    ################################################\n",
    "    print(\"Compute niche identification metrics...\")\n",
    "    run_dict[\"ari\"].append(sdmbench.compute_ARI(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"nmi\"].append(sdmbench.compute_NMI(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"chaos\"].append(sdmbench.compute_CHAOS(\n",
    "        model.adata,\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"pas\"].append(sdmbench.compute_PAS(\n",
    "        model.adata,\n",
    "        \"pred_niche_types\",\n",
    "        spatial_key=\"spatial\"))\n",
    "    run_dict[\"asw\"].append(sdmbench.compute_ASW(\n",
    "        model.adata,\n",
    "        \"pred_niche_types\",\n",
    "        spatial_key=\"spatial\"))\n",
    "    run_dict[\"hom\"].append(sdmbench.compute_HOM(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    run_dict[\"com\"].append(sdmbench.compute_COM(\n",
    "        model.adata,\n",
    "        \"niche_types\",\n",
    "        \"pred_niche_types\"))\n",
    "    #moranI, gearyC = sdmbench.marker_score(\n",
    "    #    model.adata,\n",
    "    #    \"niche_types\")\n",
    "    #run_dict[\"moranI\"].append(moranI)\n",
    "    #run_dict[\"gearyC\"].append(gearyC)\n",
    "\n",
    "    if debug:\n",
    "        return gp_eval_df, run_dict\n",
    "    else:\n",
    "        return run_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da5f300c-8bf2-4448-9de2-60f17f703033",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gp_genes(df,\n",
    "                 gene_type=\"source\",\n",
    "                 use_filter=True,\n",
    "                 weight_thresh=0.):\n",
    "    genes = []\n",
    "    \n",
    "    if use_filter:\n",
    "        df[f\"gp_{gene_type}_genes_incl\"] = df[f\"gp_{gene_type}_genes_weights\"].apply(\n",
    "            lambda x: [np.abs(weight) > weight_thresh for weight in x])\n",
    "    \n",
    "        for index, row in df.iterrows():\n",
    "            incl_list = row[f\"gp_{gene_type}_genes_incl\"]\n",
    "            genes_list = row[f\"gp_{gene_type}_genes\"]\n",
    "            for incl, gene in zip(incl_list, genes_list):\n",
    "                if incl:\n",
    "                    genes.append(gene)\n",
    "    else:\n",
    "        for index, row in df.iterrows():\n",
    "            genes_list = row[f\"gp_{gene_type}_genes\"]\n",
    "            for gene in genes_list:\n",
    "                genes.append(gene)        \n",
    "    return genes\n",
    "    \n",
    "\n",
    "def retrieve_gp_gene_summary(model,\n",
    "                             run_dict,\n",
    "                             weight_thresh=0.):\n",
    "    gp_summary_df = model.get_gp_summary()\n",
    "    gp_summary_df_active_gps = gp_summary_df[gp_summary_df[\"gp_active\"]]\n",
    "    \n",
    "    run_dict[\"n_total_gps\"].append(len(model.adata.uns[gp_names_key]))\n",
    "    run_dict[\"n_active_gps\"].append(len(model.adata.uns[active_gp_names_key]))\n",
    "\n",
    "    run_dict[\"n_total_source_genes\"].append(len(get_gp_genes(\n",
    "        gp_summary_df,\n",
    "        gene_type=\"source\",\n",
    "        use_filter=False)))\n",
    "    run_dict[\"n_total_target_genes\"].append(len(get_gp_genes(\n",
    "        gp_summary_df,\n",
    "        gene_type=\"target\",\n",
    "        use_filter=False)))\n",
    "    run_dict[\"n_active_source_genes\"].append(len(get_gp_genes(\n",
    "        gp_summary_df_active_gps,\n",
    "        gene_type=\"source\",\n",
    "        use_filter=True,\n",
    "        weight_thresh=weight_thresh)))\n",
    "    run_dict[\"n_active_target_genes\"].append(len(get_gp_genes(\n",
    "        gp_summary_df_active_gps,\n",
    "        gene_type=\"target\",\n",
    "        use_filter=True,\n",
    "        weight_thresh=weight_thresh)))\n",
    "    \n",
    "    return run_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab9e4f51-1512-4fcd-8206-33d866433968",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_simulation(combined_gp_dict,\n",
    "                   run_dict,\n",
    "                   mlflow_experiment_id,\n",
    "                   use_prior_gp_mask=True,\n",
    "                   model_load_timestamp=None,\n",
    "                   figure_folder_path=figure_folder_path,\n",
    "                   dataset_name=dataset_name,\n",
    "                   latent_leiden_resolution=0.3,\n",
    "                   leiden_resolution_increments=0.1,\n",
    "                   valid_niche_thresh=100,\n",
    "                   mode=\"metrics\" # \"gp_gene_analysis\"\n",
    "                  ):\n",
    "    if not model_load_timestamp:\n",
    "        # Load data\n",
    "        adata = load_data(combined_gp_dict,\n",
    "                          run_dict,\n",
    "                          use_prior_gp_mask,\n",
    "                          dataset_name)\n",
    "\n",
    "        # Train NicheCompass\n",
    "        model = train_model(adata,\n",
    "                            run_dict,\n",
    "                            mlflow_experiment_id)\n",
    "    else:\n",
    "        model_folder_path = f\"{artifacts_folder_path}/ablation/{model_load_timestamp}/model\"\n",
    "        figure_folder_path = f\"{artifacts_folder_path}/ablation/{model_load_timestamp}/figures\"\n",
    "        model = NicheCompass.load(dir_path=model_folder_path + f\"/run{run_dict['run_number'][-1]}_seed{run_dict['model_seed'][-1]}\",\n",
    "                                  adata=None,\n",
    "                                  adata_file_name=\"adata.h5ad\",\n",
    "                                  gp_names_key=gp_names_key)\n",
    "    \n",
    "    if mode == \"metrics\":\n",
    "        # Identify niches\n",
    "        model = identify_niches(model,\n",
    "                                run_dict,\n",
    "                                figure_folder_path,\n",
    "                                latent_leiden_resolution,\n",
    "                                leiden_resolution_increments,\n",
    "                                valid_niche_thresh)\n",
    "\n",
    "        # Get enriched prior and de-novo GPs\n",
    "        enriched_prior_gps_per_niche, enriched_denovo_gps_per_niche = get_enriched_gps(model,\n",
    "                                                                                       use_prior_gp_mask)\n",
    "        \n",
    "        # Compute metrics\n",
    "        run_dict = compute_metrics(model,\n",
    "                                   enriched_prior_gps_per_niche,\n",
    "                                   enriched_denovo_gps_per_niche,\n",
    "                                   run_dict,\n",
    "                                   use_prior_gp_mask)\n",
    "        \n",
    "    elif mode == \"gp_gene_analysis\":\n",
    "        run_dict = retrieve_gp_gene_summary(model,\n",
    "                                            run_dict)\n",
    "\n",
    "    return run_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "649c17ae-48c6-443c-b018-c222ce4d813c",
   "metadata": {},
   "source": [
    "## 2. Test Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "962f4778-3091-4297-8fd5-4a538d5f395f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if n_genes == 1105:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "elif n_genes == 12086:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_stats_{increment_mode}increments.csv\"\n",
    "\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "\n",
    "simulation_gps = set()\n",
    "for gp in sim_gps_stats[\"prior_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "for gp in sim_gps_stats[\"denovo_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "\n",
    "print(f\"Number of simulation GPs: {len(simulation_gps)}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92de7919-6829-4298-b69f-4d74243da1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_gp_dict = prepare_gps(species=species,\n",
    "                               omnipath_lr_network_file_path=omnipath_lr_network_file_path,\n",
    "                               nichenet_lr_network_file_path=nichenet_lr_network_file_path,\n",
    "                               nichenet_ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "                               gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c65c9bc7-b4eb-4912-9b6a-62b42aa1d012",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"n_neighbors\"] = [6]\n",
    "run_dict[\"n_latent\"] = [20]\n",
    "use_prior_gp_mask = True # True, False\n",
    "dataset_name = \"sim1_1105genes_10000locs_strongincrements\" # \"sim1_1105genes_10000locs_strongincrements\", \"sim1_1105genes_10000locs_strongincrements_binned\"\n",
    "\n",
    "adata = load_data(combined_gp_dict,\n",
    "                  run_dict,\n",
    "                  use_prior_gp_mask,\n",
    "                  dataset_name=dataset_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f07dd97f-1a32-462b-97dc-698f58fc0409",
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "torch.cuda.empty_cache()\n",
    "import gc\n",
    "gc.collect()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "023f6e5c-71d0-4a74-a3ab-b7f4c80bee1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"n_epochs\"] = [1]\n",
    "run_dict[\"model_seed\"] = [0]\n",
    "run_dict[\"run_number\"] = [42]\n",
    "run_dict[\"n_neighbors\"] = [6]\n",
    "run_dict[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])]\n",
    "run_dict[\"conv_layer_encoder\"] = [\"gatv2conv\"]\n",
    "run_dict[\"active_gp_thresh_ratio\"] = [0.01]\n",
    "run_dict[\"lambda_edge_recon\"] = [500000.]\n",
    "run_dict[\"lambda_gene_expr_recon\"] = [300.]\n",
    "run_dict[\"lambda_l1_masked\"] = [0.]\n",
    "run_dict[\"lambda_l1_addon\"] = [0.]\n",
    "run_dict[\"l1_targets_categories\"] = [\"target_gene\"]\n",
    "run_dict[\"l1_sources_categories\"] = [None]\n",
    "\n",
    "mlflow_experiment_id = None\n",
    "\n",
    "model = train_model(adata,\n",
    "                    run_dict,\n",
    "                    mlflow_experiment_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beb6625d-ba68-4c5d-801e-b4979cd76bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"model_seed\"] = [0]\n",
    "run_dict[\"run_number\"] = [42]\n",
    "run_dict[\"n_total_gps\"] = []\n",
    "run_dict[\"n_active_gps\"] = []\n",
    "run_dict[\"n_total_source_genes\"] = []\n",
    "run_dict[\"n_total_target_genes\"] = []\n",
    "run_dict[\"n_active_source_genes\"] = []\n",
    "run_dict[\"n_active_target_genes\"] = []\n",
    "\n",
    "run_dict = retrieve_gp_gene_summary(\n",
    "    model,\n",
    "    run_dict,\n",
    "    weight_thresh=0.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c392d21f-a33b-4a79-a9be-d1241bc36a43",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"model_seed\"] = [0]\n",
    "run_dict[\"run_number\"] = [42]\n",
    "\n",
    "ablation = \"test\"\n",
    "\n",
    "model = identify_niches(model,\n",
    "                        run_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a856643-040f-4512-9434-bc9dc7244769",
   "metadata": {},
   "outputs": [],
   "source": [
    "log_bayes_factor_thresh = 0. # 2.3\n",
    "use_prior_gp_mask = True # True, False\n",
    "\n",
    "enriched_prior_gps_per_niche, enriched_denovo_gps_per_niche = get_enriched_gps(model,\n",
    "                                                                               use_prior_gp_mask)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb51bee-2e66-4188-b604-95561cec58e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "enriched_denovo_gps_per_niche[\"gene_program\"].apply(lambda x: len(x)).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f334e20-bf5b-4db7-bbf0-c62321f41e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "if n_genes == 1105:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "elif n_genes == 12086:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_stats_{increment_mode}increments.csv\"\n",
    "\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "\n",
    "sim_dict = {}\n",
    "for element in [\"prior_gps\",\n",
    "                \"denovo_gps\",\n",
    "                \"prior_gp_sources\",\n",
    "                \"prior_gp_targets\",\n",
    "                \"denovo_gp_sources\",\n",
    "                \"denovo_gp_targets\"]:\n",
    "    sim_dict[element] = set()\n",
    "\n",
    "for i, row in sim_gps_stats.iterrows():\n",
    "    for element in [\"prior_gps\",\n",
    "                    \"denovo_gps\",\n",
    "                    \"prior_gp_sources\",\n",
    "                    \"prior_gp_targets\",\n",
    "                    \"denovo_gp_sources\",\n",
    "                    \"denovo_gp_targets\"]:\n",
    "        if len(row[element]) > 0:\n",
    "            sim_dict[element].update(row[element].split(\",\"))\n",
    "\n",
    "simulation_gps = sim_dict['prior_gps'] | sim_dict['denovo_gps']\n",
    "simulation_denovo_genes = sim_dict['denovo_gp_sources'] | sim_dict['denovo_gp_targets']\n",
    "simulation_prior_genes = sim_dict['prior_gp_sources'] | sim_dict['prior_gp_targets']\n",
    "\n",
    "\n",
    "print(f\"\"\"Number of GPs : {(\n",
    "    len(simulation_gps))}\"\"\")\n",
    "\n",
    "print(f\"\"\"Number of genes : {(\n",
    "    len(simulation_denovo_genes) +\n",
    "    len(simulation_prior_genes))}\"\"\")\n",
    "\n",
    "print(f\"\"\"Number of denovo genes : {len(simulation_denovo_genes)}\"\"\")\n",
    "\n",
    "print(f\"\"\"Number of prior genes : {len(simulation_prior_genes)}\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb03684a-6e84-4904-96c4-6d5c09027912",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_dict = {}\n",
    "run_dict[\"n_neighbors\"] = [6]\n",
    "\n",
    "increment_params = [3, 5, 8]\n",
    "\n",
    "for element_type in [\n",
    "    \"prior_gps\",\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "top_n_genes = 3\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 10\n",
    "use_prior_gp_mask = False # True, False\n",
    "\n",
    "eval_df, run_dict = compute_metrics(model,\n",
    "                                    enriched_prior_gps_per_niche,\n",
    "                                    enriched_denovo_gps_per_niche,\n",
    "                                    run_dict,\n",
    "                                    use_prior_gp_mask,\n",
    "                                    debug=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5c3b47-3c77-41e3-8727-54b1d0283bf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "eval_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2da62c5-5e5c-4a40-9e9b-e488ad6332dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "enriched_prior_gps_per_niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdb381ae-a558-4ca5-8bee-fd6a49b01640",
   "metadata": {},
   "outputs": [],
   "source": [
    "enriched_denovo_gps_per_niche"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eec12d64-af9c-4f49-bb12-8f12b13db4d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_summary_df = model.get_gp_summary()\n",
    "model.add_active_gp_scores_to_obs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c25e222-67e1-44d4-9dc9-905a2d647e63",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp = \"Add-on_59_GP\"\n",
    "gp_summary_df[gp_summary_df[\"gp_name\"] == gp]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe31d635-457d-4aee-bc8d-8e49937a30ce",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "idx = 0\n",
    "gene = gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].tolist()[0][idx]\n",
    "\n",
    "sc.pl.umap(adata, color=[gp], layer=\"counts\")\n",
    "\n",
    "gp_summary_df[gp_summary_df[\"gp_name\"] == gp][\"gp_target_genes\"].values.tolist()[0].index(gene)\n",
    "sc.pl.umap(adata, color=[gene], layer=\"counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d3cdd67-f375-4190-85e4-717b0fb3ef90",
   "metadata": {},
   "outputs": [],
   "source": [
    "niche_type = \"Niche 6\"\n",
    "increment_param = 8\n",
    "col = \"denovo_gp_targets\"\n",
    "genes = sim_gps_stats[(sim_gps_stats[\"regionA\"] == niche_type) & (sim_gps_stats[\"increment_param\"] == increment_param)][col].tolist()[0].split(\",\")\n",
    "\n",
    "gene in genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31c32978-764a-4fca-a3af-6eeec94914d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def f1_score(set1, set2):\n",
    "    tp = len(set1 & set2)  # true positives\n",
    "    fp = len(set1 - set2)  # false positives\n",
    "    fn = len(set2 - set1)  # false negatives\n",
    "\n",
    "    print(tp)\n",
    "    print(fp)\n",
    "    print(fn)\n",
    "    \n",
    "    if tp == 0:\n",
    "        return 0.0\n",
    "    \n",
    "    precision = tp / (tp + fp) if (tp + fp) > 0 else 0\n",
    "    recall = tp / (tp + fn) if (tp + fn) > 0 else 0\n",
    "    \n",
    "    if precision + recall == 0:\n",
    "        return 0.0\n",
    "    \n",
    "    return 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "niche = \"Niche 7\" # \"Niche 6\"\n",
    "set1_col = \"random_denovo_gp_target_genes_top3_seed0\" # enriched_denovo_gp_target_genes_top3, random_denovo_gp_target_genes_top3_seed0\n",
    "set2_col = \"incremented_denovo_gp_target_genes\"\n",
    "\n",
    "print(eval_df.loc[niche][set1_col])\n",
    "print(eval_df.loc[niche][set2_col])\n",
    "\n",
    "print(f1_score(eval_df.loc[niche][set1_col], eval_df.loc[niche][set2_col]))\n",
    "\n",
    "f1_scores = []\n",
    "for niche in eval_df.index:\n",
    "    f1_scores.append(f1_score(eval_df.loc[niche][set1_col], eval_df.loc[niche][set2_col]))\n",
    "\n",
    "print(np.mean(f1_scores))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0edeae-4452-4730-b49d-d16c498c069a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_df = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_strongincrements.csv\")\n",
    "gp_df[(gp_df[\"regionA\"] == \"Region8\") & (gp_df[\"increment_param\"] == 8)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b875e4-a6dc-4879-855a-4fd868ac9818",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "gene = \"Slc37a2\"\n",
    "\n",
    "sc.pl.umap(adata, color=[gene], layer=\"counts\")\n",
    "sq.pl.spatial_scatter(adata, color=gene, shape=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f53d723-4fd7-49dc-844b-58cf8b9585a2",
   "metadata": {},
   "source": [
    "## 3. Perform Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66340427-aed0-47d4-877c-058494f5b004",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "###############\n",
    "### Get GPs ###\n",
    "###############\n",
    "combined_gp_dict = prepare_gps(species=species,\n",
    "                               omnipath_lr_network_file_path=omnipath_lr_network_file_path,\n",
    "                               nichenet_lr_network_file_path=nichenet_lr_network_file_path,\n",
    "                               nichenet_ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "                               gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path)\n",
    "\n",
    "####################\n",
    "### Get GP stats ###\n",
    "####################\n",
    "if n_genes == 1105:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "elif n_genes == 12086:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_stats_{increment_mode}increments.csv\"\n",
    "\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "\n",
    "simulation_gps = set()\n",
    "for gp in sim_gps_stats[\"prior_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "for gp in sim_gps_stats[\"denovo_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "\n",
    "print(f\"Number of simulation GPs: {len(simulation_gps)}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d621701-56aa-4b22-90c6-20e5d1166774",
   "metadata": {},
   "source": [
    "#### 3.1.1 Loss Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10f79e70-1e93-4a73-ab34-868537ec8928",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"loss_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000., 0., 500000., 50000., 500000.]\n",
    "config[\"lambda_gene_expr_recon\"] = [300., 300., 0., 300., 30]\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lambda_l1_addon\"] = [0.] * 5\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 5\n",
    "config[\"l1_sources_categories\"] = [None] * 5\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb03f219-aef0-4bfd-9b8b-c69e00eadd92",
   "metadata": {},
   "source": [
    "#### 3.1.2 Loss Ablation Extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c429f22c-195c-406a-968e-e753043a2123",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"loss_ablation_extended\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 7\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 7\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 7\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 7\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 7\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 7\n",
    "config[\"lambda_l1_masked\"] = [0., 0., 0., 0., 3., 30., 300.]\n",
    "config[\"lambda_l1_addon\"] = [0., 3., 30., 300., 3., 30., 300.]\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 7\n",
    "config[\"l1_sources_categories\"] = [None] * 7\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44b2fd48-14f1-4f21-86b1-6e240d18ff51",
   "metadata": {},
   "source": [
    "#### 3.1.3 Encoder Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa707ec5-06a3-47c4-b821-bc6dd00e1bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"encoder_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 2\n",
    "config[\"conv_layer_encoder\"] = [\"gcnconv\", \"gatv2conv\"]\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 2\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 2\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 2\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 2\n",
    "config[\"lambda_l1_masked\"] = [0.] * 2\n",
    "config[\"lambda_l1_addon\"] = [0.] * 2\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 2\n",
    "config[\"l1_sources_categories\"] = [None] * 2\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d594652-f345-41c3-84a4-7c7384e020db",
   "metadata": {},
   "source": [
    "#### 3.1.4 Neighbor Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca1a6f5a-cb97-4efd-819b-c92a71a421a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"neighbor_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [4, 8, 12, 16, 20]\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lambda_l1_addon\"] = [0.] * 5\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 5\n",
    "config[\"l1_sources_categories\"] = [None] * 5\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9484d9e-3248-42a8-8ce3-54922895e274",
   "metadata": {},
   "source": [
    "#### 3.1.5 De-novo GP Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a75b6815-80e7-45fc-96aa-4966a3eac11e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"denovogp_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [0, 10, 30, 100, len([gp for gp in simulation_gps if \"Add-On\" in gp])]\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lambda_l1_addon\"] = [0.] * 5\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 5\n",
    "config[\"l1_sources_categories\"] = [None] * 5\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b647ac7-8abd-4fbe-8fe7-7ce7352b14b4",
   "metadata": {},
   "source": [
    "#### 3.1.6 GP Selection Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20528238-4872-41b7-82c7-a2cc973071d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"gpselection_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0., 0.01, 0.03, 0.1, 0.3]\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lambda_l1_addon\"] = [0.] * 5\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 5\n",
    "config[\"l1_sources_categories\"] = [None] * 5\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba510be8-7367-4abc-b19b-d8917e8ce0e9",
   "metadata": {},
   "source": [
    "#### 3.1.7 No Prior GP Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ab6918d-3aea-4028-8b0f-3fb91168c8b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"nopriorgp_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [0] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lambda_l1_addon\"] = [0.] * 5\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 5\n",
    "config[\"l1_sources_categories\"] = [None] * 5\n",
    "config[\"n_latent\"] = [20, 50, 100, 500, 1582] # last one is number of prior GPs\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id,\n",
    "                                      use_prior_gp_mask=False,\n",
    "                                      #model_load_timestamp=\"29072024_102431\"\n",
    "                                     )\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "311bf3cc-12b0-4609-b49b-6dba951a7828",
   "metadata": {},
   "source": [
    "#### 3.1.8 Binned Counts Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b03dd307-5a1c-4871-86ec-54a00368f457",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"binned_counts_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [4, 6, 8] * 2\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\", \"gatv2conv\", \"gatv2conv\", \"gcnconv\", \"gcnconv\", \"gcnconv\"]\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 6\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 6\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 6\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 6\n",
    "config[\"lambda_l1_masked\"] = [0.] * 6\n",
    "config[\"lambda_l1_addon\"] = [0.] * 6\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 6\n",
    "config[\"l1_sources_categories\"] = [None] * 6\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id,\n",
    "                                      dataset_name=\"sim1_1105genes_10000locs_strongincrements_binned\",\n",
    "                                      valid_niche_thresh=10)\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "834c90fc-d24f-4cb9-8a26-9db5ed3cec7f",
   "metadata": {},
   "source": [
    "#### 3.1.9 Encoder Ablation Tissue 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8c81f14-0b28-4ace-940a-017794a00884",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"encoder_tissue2_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 2\n",
    "config[\"conv_layer_encoder\"] = [\"gcnconv\", \"gatv2conv\"]\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 2\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 2\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 2\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 2\n",
    "config[\"lambda_l1_masked\"] = [0.] * 2\n",
    "config[\"lambda_l1_addon\"] = [0.] * 2\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 2\n",
    "config[\"l1_sources_categories\"] = [None] * 2\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id,\n",
    "                                      use_prior_gp_mask=True,\n",
    "                                      dataset_name=\"sim1_12086genes_10000locs_strongincrements\",\n",
    "                                     )\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2660c3f0-6dce-48bf-91b9-9b97a139d93d",
   "metadata": {},
   "source": [
    "#### 3.1.10 No Prior GP Ablation Tissue 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f438d46-34f0-43f3-a4f0-8d625302fc03",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"nopriorgp_tissue2_ablation\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = 400\n",
    "config[\"n_neighbors\"] = [6] * 5\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 5\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01] * 5\n",
    "config[\"n_addon_gp\"] = [0] * 5\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 5\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 5\n",
    "config[\"lambda_l1_masked\"] = [0.] * 5\n",
    "config[\"lambda_l1_addon\"] = [0.] * 5\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 5\n",
    "config[\"l1_sources_categories\"] = [None] * 5\n",
    "config[\"n_latent\"] = [20, 50, 100, 500, 1582] # last one is number of prior GPs\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 8\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "for element_type in [\n",
    "    \"prior_gps\"\n",
    "]:\n",
    "    run_dict[f\"f1_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"precision_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"recall_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_enriched_{element_type}\"] = []\n",
    "    run_dict[f\"f1_random_{element_type}\"] = []\n",
    "    run_dict[f\"precision_random_{element_type}\"] = []\n",
    "    run_dict[f\"recall_random_{element_type}\"] = []\n",
    "    run_dict[f\"jaccard_index_random_{element_type}\"] = []\n",
    "    for increment_param in increment_params:\n",
    "        run_dict[f\"zscore_{element_type}_increment_param_{increment_param}\"] = []\n",
    "for element_type in [\n",
    "    \"prior_gp_source_genes\",\n",
    "    \"prior_gp_target_genes\",\n",
    "    \"denovo_gp_source_genes\",\n",
    "    \"denovo_gp_target_genes\"\n",
    "]:\n",
    "    for n in [3, 5, 10]:\n",
    "        top_n_element_type = element_type + f\"_top{n}\"\n",
    "        run_dict[f\"f1_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_enriched_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"f1_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"precision_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"recall_random_{top_n_element_type}\"] = []\n",
    "        run_dict[f\"jaccard_index_random_{top_n_element_type}\"] = []\n",
    "        for increment_param in increment_params:\n",
    "            run_dict[f\"zscore_{top_n_element_type}_increment_param_{increment_param}\"] = []\n",
    "run_dict[\"ari\"] = []\n",
    "run_dict[\"nmi\"] = []\n",
    "run_dict[\"chaos\"] = []\n",
    "run_dict[\"pas\"] = []\n",
    "run_dict[\"asw\"] = []\n",
    "run_dict[\"hom\"] = []\n",
    "run_dict[\"com\"] = []\n",
    "run_dict[\"moranI\"] = []\n",
    "run_dict[\"gearyC\"] = []\n",
    "\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id,\n",
    "                                      use_prior_gp_mask=False,\n",
    "                                      dataset_name=\"sim1_12086genes_10000locs_strongincrements\",\n",
    "                                     )\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaf904cf-3d60-4150-93cb-d24d28b88080",
   "metadata": {},
   "source": [
    "## 4. Visualize Results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9de86c4-d578-4437-9d0a-6bf0e3aeff1f",
   "metadata": {},
   "source": [
    "### 4.1 Loss Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15aad70d-9112-477f-9826-939f9c89e87b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"loss_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b2cbb8f-c756-4754-84a1-3f25ec1ea8d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"Loss\"\n",
    "title = \"Edge & Gene Expression Reconstruction Ablation\"\n",
    "\n",
    "mapping_dict = {\"500000.0_300.0\": \"Balanced Edge & Gene Expr Recon\",\n",
    "                \"0.0_300.0\": \"Only Gene Expr Recon\",\n",
    "                \"500000.0_0.0\": \"Only Edge Recon\",\n",
    "                \"50000.0_300.0\": \"Weak Edge Recon\",\n",
    "                \"500000.0_30.0\": \"Weak Gene Expr Recon\"\n",
    "               }\n",
    "\n",
    "col1 = \"lambda_edge_recon\"\n",
    "col2 = \"lambda_gene_expr_recon\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get((str(row[col1]) + \"_\" + str(row[col2])), \"NA\")\n",
    "\n",
    "run_df[\"Loss\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5db79495-097e-4c75-b11e-7a70fc3b60af",
   "metadata": {},
   "source": [
    "### 4.2 Loss Ablation Extended"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "423208de-ff47-461e-9860-492bfd54f936",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"loss_ablation_extended\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48216c46-bb0c-45bd-a14a-a1a083b4e417",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"Loss\"\n",
    "title = \"Gene Expr Regularization Ablation\"\n",
    "\n",
    "mapping_dict = {\"0.0_0.0\": \"No Reg\",\n",
    "                \"0.0_3.0\": \"Only Weak De-novo Reg\",\n",
    "                \"0.0_30.0\": \"Only Medium De-novo Reg\",\n",
    "                \"0.0_300.0\": \"Only Strong De-novo Reg\",\n",
    "                \"3.0_3.0\": \"Weak Prior & De-novo Reg\",\n",
    "                \"30.0_30.0\": \"Medium Prior & De-novo Reg\",\n",
    "                \"300.0_300.0\": \"Strong Prior & De-novo Reg\",\n",
    "               }\n",
    "\n",
    "col1 = \"lambda_l1_masked\"\n",
    "col2 = \"lambda_l1_addon\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get((str(row[col1]) + \"_\" + str(row[col2])), \"NA\")\n",
    "\n",
    "run_df[\"Loss\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28f1c9eb-d16a-42b1-af83-f0991f8907c5",
   "metadata": {},
   "source": [
    "### 4.3 Encoder Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6feba5a6-2bd3-4319-91ec-adc9e51d59fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"encoder_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a1ba922-d4d5-4e00-acd9-81e0e5eb67cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"Encoder\"\n",
    "title = \"Encoder Ablation\"\n",
    "\n",
    "mapping_dict = {\"gcnconv\": \"GCNConv (NicheCompass Light)\",\n",
    "                \"gatv2conv\": \"GATv2Conv (NicheCompass)\",\n",
    "               }\n",
    "\n",
    "col1 = \"conv_layer_encoder\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get(str(row[col1]), \"NA\")\n",
    "\n",
    "run_df[\"Encoder\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a325b18-58a3-47f8-bb1b-76b7346645a4",
   "metadata": {},
   "source": [
    "### 4.4 Neighbor Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cb01da5-e9c5-46da-b982-dd9a882addae",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"neighbor_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "859b302f-8bf2-46a6-a3f9-f09487da862a",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"n_neighbors\"\n",
    "cat_label = \"KNN\"\n",
    "title = \"Neighborhood Ablation\"\n",
    "\n",
    "run_df[\"n_neighbors\"] = run_df[\"n_neighbors\"].apply(lambda x: str(x))\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "127d62c7-bb8a-4a6e-91cc-58ac2320683f",
   "metadata": {},
   "source": [
    "### 4.5 De-novo GP Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc9863c7-8872-4940-b58c-55f6a67e5b3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"denovogp_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841d83d6-a3e2-49a0-a5be-280bc25de370",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"n_addon_gp\"\n",
    "cat_label = \"De-novo GPs\"\n",
    "title = \"De-novo GP Ablation\"\n",
    "\n",
    "run_df[\"n_addon_gp\"] = run_df[\"n_addon_gp\"].apply(lambda x: str(x))\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f70b7c7-ea42-4afc-b5d8-8c84986e86fb",
   "metadata": {},
   "source": [
    "### 4.6 GP Selection Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2aea044-d79a-4ee9-acdc-8c14ba51d732",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"gpselection_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9196d651-d2da-410d-b725-7d99fabf4e80",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"active_gp_thresh_ratio\"\n",
    "cat_label = \"Active GP Threshold\"\n",
    "title = \"GP Selection Ablation\"\n",
    "\n",
    "run_df[\"active_gp_thresh_ratio\"] = run_df[\"active_gp_thresh_ratio\"].apply(lambda x: str(x))\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e31e4af5-a13e-491a-957c-7fbfbd7eebc5",
   "metadata": {},
   "source": [
    "### 4.7 No Prior GP Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30d9583a-4166-43bb-9028-6cd920e4fe08",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"nopriorgp_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c502807f-4221-49a6-baab-2ab1cf8fda53",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"n_latent\"\n",
    "cat_label = \"Embedding Size\"\n",
    "title = \"No Prior GP Ablation\"\n",
    "\n",
    "run_df[\"n_latent\"] = run_df[\"n_latent\"].apply(lambda x: str(x))\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "554522ae-db97-4ae4-ad4c-9534bd7d95cf",
   "metadata": {},
   "source": [
    "### 4.8 Binned Counts Ablation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a4e31d6-3aa6-4726-84ea-b86c5af72bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"binned_counts_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e30238c-6b84-493d-b0a5-0725eae568bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"encoder_knn\"\n",
    "cat_label = \"Encoder + KNN\"\n",
    "title = \"Binned Counts Ablation\"\n",
    "\n",
    "mapping_dict = {\"gcnconv_4\": \"GCNConv (NicheCompass Light) + 4\",\n",
    "                \"gcnconv_6\": \"GCNConv (NicheCompass Light) + 6\",\n",
    "                \"gcnconv_8\": \"GCNConv (NicheCompass Light) + 8\",\n",
    "                \"gatv2conv_4\": \"GATv2Conv (NicheCompass) + 4\",\n",
    "                \"gatv2conv_6\": \"GATv2Conv (NicheCompass) + 6\",\n",
    "                \"gatv2conv_8\": \"GATv2Conv (NicheCompass) + 8\",\n",
    "               }\n",
    "\n",
    "col1 = \"conv_layer_encoder\"\n",
    "col2 = \"n_neighbors\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get((str(row[col1]) + \"_\" + str(row[col2])), \"NA\")\n",
    "\n",
    "run_df[\"encoder_knn\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e78b3d77-8a4d-4a83-a353-2a667126706b",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_number = 2\n",
    "seed_number = 0\n",
    "leiden_resolutions = [0.034]\n",
    "\n",
    "adata = sc.read_h5ad(f\"{artifacts_folder_path}/ablation/04082024_095432/model/run{run_number}_seed{seed_number}/adata.h5ad\")\n",
    "adata.obs[\"sample\"] = \"sample1\"\n",
    "\n",
    "print(\"\\nComputing neighbor graph...\")\n",
    "# Use latent representation for UMAP generation\n",
    "sc.pp.neighbors(adata,\n",
    "                use_rep=\"nichecompass_latent\",\n",
    "                key_added=\"nichecompass_latent\")\n",
    "\n",
    "print(\"\\nComputing UMAP embedding...\")\n",
    "sc.tl.umap(adata,\n",
    "           neighbors_key=\"nichecompass_latent\")\n",
    "\n",
    "print(\"\\nComputing Leiden clustering...\")\n",
    "sc.tl.leiden(adata=adata,\n",
    "             resolution=leiden_resolutions[0],\n",
    "             key_added=f\"leiden_{leiden_resolutions[0]}\",\n",
    "             neighbors_key=\"nichecompass_latent\")\n",
    "\n",
    "niche_annotations = {\n",
    "    \"0\": \"Niche 5\",\n",
    "    \"1\": \"Niche 3\",\n",
    "    \"2\": \"Niche 6\",\n",
    "    \"3\": \"Niche 4\",        \n",
    "    \"4\": \"Niche 2\", \n",
    "    \"5\": \"Niche 1\", \n",
    "    \"6\": \"Niche 7\", \n",
    "    \"7\": \"Niche 8\",\n",
    "    \"8\": \"Niche 8 (Periphery)\"}\n",
    "\n",
    "adata.obs[\"niche\"] = adata.obs[f\"leiden_{leiden_resolutions[0]}\"].map(niche_annotations)\n",
    "\n",
    "latent_cluster_colors = create_new_color_dict(\n",
    "    adata=adata,\n",
    "    cat_key=f\"leiden_{leiden_resolutions[0]}\")\n",
    "\n",
    "niche_colors = {niche: latent_cluster_colors[cluster] for cluster, niche in niche_annotations.items()}\n",
    "\n",
    "plot_category_in_latent_and_physical_space(\n",
    "        adata=adata,\n",
    "        #figsize=(10, 20), for latent UMAP\n",
    "        plot_label=\"Latent Clusters\",\n",
    "        model_label={model},\n",
    "        cat_key=f\"niche\",\n",
    "        groups=None,\n",
    "        sample_key=\"sample\",\n",
    "        samples=[\"sample1\"], # =None for latent UMAP\n",
    "        cat_colors=niche_colors,\n",
    "        size=(720000 / len(adata)),\n",
    "        spot_size=30,\n",
    "        save_fig=True,\n",
    "        file_path=f\"{artifacts_folder_path}/ablation/latent_clusters_{ablation}.svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2f57837-5de6-423e-9fc6-d2491708d6ad",
   "metadata": {},
   "source": [
    "### 4.9 Encoder Ablation Tissue 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4a337ff-ec62-4781-8b8a-f1c8ddd313f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"encoder_tissue2_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "826f11e7-480a-450a-adf9-bf4cef7e0469",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"Encoder\"\n",
    "title = \"Encoder Ablation Tissue 2\"\n",
    "\n",
    "mapping_dict = {\"gcnconv\": \"GCNConv (NicheCompass Light)\",\n",
    "                \"gatv2conv\": \"GATv2Conv (NicheCompass)\",\n",
    "               }\n",
    "\n",
    "col1 = \"conv_layer_encoder\"\n",
    "\n",
    "def map_values(row):\n",
    "    return mapping_dict.get(str(row[col1]), \"NA\")\n",
    "\n",
    "run_df[\"Encoder\"] = run_df.apply(lambda row: map_values(row), axis=1)\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top10\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat.capitalize())\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5222f73d-67c6-498c-8246-44331ac65714",
   "metadata": {},
   "source": [
    "### 4.10 No Prior GP Ablation Tissue 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3cad4d-3701-4bab-a6c2-b57f6fbb6d48",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"nopriorgp_tissue2_ablation\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cb5aa76-3924-417a-88b3-656a81eaf663",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat = \"n_latent\"\n",
    "cat_label = \"Embedding Size\"\n",
    "title = \"No Prior GP Ablation\"\n",
    "\n",
    "run_df[\"n_latent\"] = run_df[\"n_latent\"].apply(lambda x: str(x))\n",
    "\n",
    "metric = \"nmi\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"ari\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"hom\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"com\"\n",
    "\n",
    "plt.figure(figsize=(3, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric.upper())\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gps\"\n",
    "metric_label = \"F1 Prior GPs\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_source_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Source Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_prior_gp_target_genes_top3\"\n",
    "metric_label = \"F1 Prior GP Target Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_target_genes_top3\"\n",
    "metric_label = \"F1 De-Novo GP\\nTarget Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()\n",
    "\n",
    "metric = \"f1_enriched_denovo_gp_source_genes_top3\"\n",
    "metric_label = \"F1 De-Novo GP\\nSource Genes\"\n",
    "\n",
    "plt.figure(figsize=(2.25, 1.5))\n",
    "sns.boxplot(x=metric, y=cat, data=run_df)\n",
    "plt.axvline(x=np.mean(run_df.groupby(\"run_number\")[metric.replace(\"enriched\", \"random\")].mean()), color='red', linestyle='--')\n",
    "plt.suptitle(title, x=0.09, ha=\"center\", va=\"top\", y=1.1)\n",
    "plt.xlabel(metric_label)\n",
    "plt.ylabel(cat_label)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/{ablation}_{metric}.svg\", bbox_inches=\"tight\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "144c8617-6001-44e6-8023-5b8d8ee7faf4",
   "metadata": {},
   "source": [
    "## 5. Additional Analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0c448c0-5e62-4f6c-8e1c-f1633b155853",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############\n",
    "### Get GPs ###\n",
    "###############\n",
    "combined_gp_dict = prepare_gps(species=species,\n",
    "                               omnipath_lr_network_file_path=omnipath_lr_network_file_path,\n",
    "                               nichenet_lr_network_file_path=nichenet_lr_network_file_path,\n",
    "                               nichenet_ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "                               gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path)\n",
    "\n",
    "####################\n",
    "### Get GP stats ###\n",
    "####################\n",
    "if n_genes == 1105:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_filtered_stats_{increment_mode}increments.csv\"\n",
    "elif n_genes == 12086:\n",
    "    sim_gps_stats_file_path = f\"sim_gps_stats_{increment_mode}increments.csv\"\n",
    "\n",
    "sim_gps_stats = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/{sim_gps_stats_file_path}\")\n",
    "sim_gps_stats = sim_gps_stats.applymap(lambda x: \"\" if pd.isna(x) else x)\n",
    "\n",
    "simulation_gps = set()\n",
    "for gp in sim_gps_stats[\"prior_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "for gp in sim_gps_stats[\"denovo_gps\"]:\n",
    "    if len(gp) > 0:\n",
    "        simulation_gps.update(gp.split(\",\"))\n",
    "\n",
    "print(f\"Number of simulation GPs: {len(simulation_gps)}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c0e8400-5c77-4f7f-ba50-f76c926ff11b",
   "metadata": {},
   "source": [
    "### 5.1 Analysis of GP Selection & Gene Regularization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee4bf7ac-b5b0-4381-a807-71080061a908",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################\n",
    "### Define model configuration ###\n",
    "##################################\n",
    "ablation = \"gp_gene_analysis\"\n",
    "mlflow.set_experiment(ablation)\n",
    "\n",
    "config = {}\n",
    "config[\"n_epochs\"] = [25, 25, 25, 25, 25, 25] * 2 + [400, 400, 400, 400, 400, 400] * 2\n",
    "config[\"n_neighbors\"] = [6] * 24\n",
    "config[\"conv_layer_encoder\"] = [\"gatv2conv\"] * 24\n",
    "config[\"active_gp_thresh_ratio\"] = [0.01, 0.01, 0.01, 0.01, 0.01, 0.01, 0.1, 0.1, 0.1, 0.1, 0.1, 0.1] * 2\n",
    "config[\"n_addon_gp\"] = [len([gp for gp in simulation_gps if \"Add-On\" in gp])] * 24\n",
    "config[\"lambda_edge_recon\"] = [500000.] * 24\n",
    "config[\"lambda_gene_expr_recon\"] = [300.] * 24\n",
    "config[\"lambda_l1_masked\"] = [0., 0., 0., 3., 30., 300.] * 4\n",
    "config[\"lambda_l1_addon\"] = [3., 30., 300., 3., 30., 300.] * 4\n",
    "config[\"l1_targets_categories\"] = [\"target_gene\"] * 24\n",
    "config[\"l1_sources_categories\"] = [None] * 24\n",
    "n_runs = max((len(lst) for lst in config.values() if isinstance(lst, list)), default=0)\n",
    "\n",
    "\n",
    "#################################\n",
    "### Define eval configuration ###\n",
    "#################################\n",
    "log_bayes_factor_thresh = 2.3 # (see https://en.wikipedia.org/wiki/Bayes_factor)\n",
    "increment_params = [1, 3, 5, 8]\n",
    "n_seeds = 1000\n",
    "n_model_seeds = 1\n",
    "\n",
    "\n",
    "#################################\n",
    "### Run simulation ###\n",
    "#################################\n",
    "# Initialize run dict\n",
    "run_dict = {}\n",
    "run_dict[\"run_number\"] = []\n",
    "run_dict[\"model_seed\"] = []\n",
    "for k in config.keys():\n",
    "    run_dict[k] = []\n",
    "run_dict[\"n_total_gps\"] = []\n",
    "run_dict[\"n_active_gps\"] = []\n",
    "run_dict[\"n_total_source_genes\"] = []\n",
    "run_dict[\"n_total_target_genes\"] = []\n",
    "run_dict[\"n_active_source_genes\"] = []\n",
    "run_dict[\"n_active_target_genes\"] = []\n",
    "\n",
    "# Loop through hyperparams (different model runs)\n",
    "for run_number in range(1, n_runs+1):\n",
    "\n",
    "    # Loop through random seeds\n",
    "    for model_seed in range(n_model_seeds):\n",
    "        with mlflow.start_run() as run:\n",
    "\n",
    "            run_dict[\"run_number\"].append(int(run_number))\n",
    "            run_dict[\"model_seed\"].append(int(model_seed))\n",
    "            mlflow_experiment_id = f\"{ablation}_{dataset_name}_{run_number}_{model_seed}\"\n",
    "            \n",
    "            # Add hyperparams from config\n",
    "            for k, v in config.items():\n",
    "                if isinstance(config[k], list):\n",
    "                    run_dict[k].append(config[k][run_number-1])\n",
    "                else:\n",
    "                    run_dict[k].append(v)\n",
    "                \n",
    "            run_dict = run_simulation(combined_gp_dict,\n",
    "                                      run_dict,\n",
    "                                      mlflow_experiment_id,\n",
    "                                      mode=\"gp_gene_analysis\")\n",
    "\n",
    "            # Store current results\n",
    "            run_df = pd.DataFrame(run_dict.values(), index=run_dict.keys()).T\n",
    "            run_df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5624cffd-67a2-4806-b646-e0b6da3ae50d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"gp_gene_analysis\"\n",
    "run_df = pd.read_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed4061f5-7225-4e6f-baa2-9d330f529ce8",
   "metadata": {},
   "outputs": [],
   "source": [
    "after_warmup_df = run_df[run_df[\"n_epochs\"] == 25]\n",
    "after_training_df = run_df[run_df[\"n_epochs\"] == 400]\n",
    "run_df = after_warmup_df\n",
    "\n",
    "run_df.rename(columns={\n",
    "    'n_active_gps': 'n_active_gps_warmup',\n",
    "    'n_active_source_genes': 'n_active_source_genes_warmup',\n",
    "    'n_active_target_genes': 'n_active_target_genes_warmup',\n",
    "}, inplace=True)\n",
    "\n",
    "run_df[\"n_active_gps_final\"] = after_training_df[\"n_active_gps\"].values\n",
    "run_df[\"n_active_source_genes_final\"] = after_training_df[\"n_active_source_genes\"].values\n",
    "run_df[\"n_active_target_genes_final\"] = after_training_df[\"n_active_target_genes\"].values\n",
    "\n",
    "run_df = run_df[[\n",
    "    \"active_gp_thresh_ratio\",\n",
    "    \"lambda_l1_masked\",\n",
    "    \"lambda_l1_addon\",\n",
    "    \"n_total_gps\",\n",
    "    \"n_active_gps_warmup\",\n",
    "    \"n_active_gps_final\",\n",
    "    \"n_total_source_genes\",\n",
    "    \"n_active_source_genes_warmup\",\n",
    "    \"n_active_source_genes_final\",\n",
    "    \"n_total_target_genes\",\n",
    "    \"n_active_target_genes_warmup\",\n",
    "    \"n_active_target_genes_final\"]]\n",
    "\n",
    "run_df.columns = [\"Active GP Thresh\",\n",
    "                  \"Prior GP Reg\",\n",
    "                  \"De novo GP Reg\",\n",
    "                  \"Total GPs\",\n",
    "                  \"Active GPs (W)\",\n",
    "                  \"Active GPs (F)\",\n",
    "                  \"Total Sources\",\n",
    "                  \"Active Sources (W)\",\n",
    "                  \"Active Sources (F)\",\n",
    "                  \"Total Targets\",\n",
    "                  \"Active Targets (W)\",\n",
    "                  \"Active Targets (F)\"]\n",
    "\n",
    "for col in [\"Total GPs\",\n",
    "            \"Active GPs (W)\",\n",
    "            \"Active GPs (F)\",\n",
    "            \"Total Sources\",\n",
    "            \"Active Sources (W)\",\n",
    "            \"Active Sources (F)\",\n",
    "            \"Total Targets\",\n",
    "            \"Active Targets (W)\",\n",
    "            \"Active Targets (F)\"]:\n",
    "    run_df[col] = run_df[col].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05acf55c-67aa-47ab-83fe-20722cb1cdab",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10, 2))\n",
    "ax.axis('off')\n",
    "ax.axis('tight')\n",
    "table = ax.table(\n",
    "    cellText=run_df.values,\n",
    "    colLabels=run_df.columns,\n",
    "    #rowLabels=run_df.index,\n",
    "    cellLoc='center',\n",
    "    loc='center',\n",
    "    colColours=[\"lightblue\"] * (3) + [\"lightgrey\", \"darkgrey\", \"darkgrey\"] * 3\n",
    ")\n",
    "table.auto_set_font_size(False)\n",
    "table.set_fontsize(14)\n",
    "table.scale(3., 1.7)\n",
    "plt.savefig(f\"{artifacts_folder_path}/ablation/gp_gene_analysis.svg\", bbox_inches=\"tight\", dpi=300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee9ad5e4-b900-465e-958a-70bc2b9f5a84",
   "metadata": {},
   "source": [
    "## 6. Miscellaneous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a01117bf-2c29-421a-b974-63d4aab14603",
   "metadata": {},
   "outputs": [],
   "source": [
    "ablation = \"loss_ablation\"\n",
    "\n",
    "df_list = []\n",
    "for file_path in [\n",
    "    f\"{artifacts_folder_path}/ablation/{ablation}_results_save.csv\",\n",
    "    f\"{artifacts_folder_path}/ablation/{ablation}_results_save2.csv\",\n",
    "    f\"{artifacts_folder_path}/ablation/{ablation}_results_save3.csv\", \n",
    "    f\"{artifacts_folder_path}/ablation/{ablation}_results_save4.csv\",\n",
    "]:\n",
    "    df_list.append(pd.read_csv(file_path))\n",
    "\n",
    "df = pd.concat(df_list, ignore_index=True)\n",
    "df = df.drop_duplicates()\n",
    "df.reset_index(drop=True, inplace=True)\n",
    "df.to_csv(f\"{artifacts_folder_path}/ablation/{ablation}_results.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94b580e2-6985-475c-a6a4-211404672902",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
