{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e5070953-d3a2-4b55-a413-6667b866ab11",
   "metadata": {},
   "source": [
    "# Data Simulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89a92b1d-780f-43c1-a458-3f22f12d5d9d",
   "metadata": {},
   "source": [
    "- **Creator**: Sebastian Birk (<sebastian.birk@helmholtz-munich.de>)\n",
    "- **Date of Creation:** 24.05.2024\n",
    "- **Date of Last Modification:** 09.01.2025 (Sebastian Birk; <sebastian.birk@helmholtz-munich.de>)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46492b38-b25a-4be3-9001-329f951a0622",
   "metadata": {},
   "source": [
    "## 1. Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b1fab77-a6b9-45dc-be3d-f065300c9ba1",
   "metadata": {},
   "source": [
    "- Run this notebook in the nichecompass-reproducibility environment, installable from ```('../../envs/environment.yaml')```."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe5bc485-4124-4ae2-b5d8-2ccf4d075f4e",
   "metadata": {},
   "source": [
    "### 1.1 Import Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9194a15-676d-47ed-b6c6-7a5f37f9f371",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5592c919-5b6a-4619-978e-0f6a479b579c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import os\n",
    "import pickle\n",
    "import random\n",
    "import warnings\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyensembl\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import scipy.sparse as sp\n",
    "import squidpy as sq\n",
    "\n",
    "from nichecompass.utils import (create_new_color_dict,\n",
    "                                extract_gp_dict_from_mebocost_ms_interactions,\n",
    "                                extract_gp_dict_from_nichenet_lrt_interactions,\n",
    "                                extract_gp_dict_from_omnipath_lr_interactions,\n",
    "                                filter_and_combine_gp_dict_gps_v2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0986273e-94b4-4c2f-b852-571c3f77b01f",
   "metadata": {},
   "source": [
    "### 1.2 Define Parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "933f221c-e7c5-4743-9e9b-9fcfe28effb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"mouse\" # for simulation gene programs\n",
    "tissue_type = 1 # 1, 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89cbfa37-1b4f-498d-b550-86fd19c44159",
   "metadata": {},
   "source": [
    "### 1.3 Run Notebook Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eabcd6b1-4681-465c-937f-87a398e5d2c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=(6, 6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef8e6b34-1c7d-41ea-8108-d8a8ff681f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "warnings.filterwarnings(\"ignore\")\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8be70312-e790-4744-97f6-9bdc995fb635",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "plt.rcParams['font.size'] = 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a8f2964-3139-4059-bede-e3775ca122d5",
   "metadata": {},
   "source": [
    "### 1.4 Configure Paths and Create Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57dead28-9c33-4e13-9a75-9f14e7c27858",
   "metadata": {},
   "outputs": [],
   "source": [
    "ga_data_folder_path = \"../../datasets/ga_data\"\n",
    "gp_data_folder_path = \"../../datasets/gp_data\"\n",
    "st_data_folder_path = \"../../datasets/st_data\"\n",
    "st_data_bronze_folder_path = st_data_folder_path + \"/bronze\"\n",
    "st_data_silver_folder_path = st_data_folder_path + \"/silver\"\n",
    "st_data_gold_folder_path = st_data_folder_path + \"/gold\"\n",
    "omnipath_lr_network_file_path = f\"{gp_data_folder_path}/omnipath_lr_network.csv\"\n",
    "nichenet_lr_network_file_path = f\"{gp_data_folder_path}/nichenet_lr_network_v2_{species}.csv\"\n",
    "nichenet_ligand_target_matrix_file_path = f\"{gp_data_folder_path}/nichenet_ligand_target_matrix_v2_{species}.csv\"\n",
    "mebocost_enzyme_sensor_interactions_folder_path = f\"{gp_data_folder_path}/metabolite_enzyme_sensor_gps\"\n",
    "gene_orthologs_mapping_file_path = f\"{ga_data_folder_path}/human_mouse_gene_orthologs.csv\"\n",
    "artifacts_folder_path = f\"../../artifacts\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f35d8500-cd0e-4931-8ca2-52260d2a448b",
   "metadata": {},
   "source": [
    "## 2. Simulate Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e8683c-fe1e-4caa-a76f-357055f78c0b",
   "metadata": {},
   "source": [
    "### 2.1 Simulate Cell-Cell Interactions (CCIs) Based on Prior Knowledge and De Novo Gene Programs (GPs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26fec8a8-b002-411f-ba35-fe07601b7e94",
   "metadata": {},
   "source": [
    "- Here we retrieve GPs based on three databases of prior knowledge of inter- and intracellular interaction pathways, also used as prior GPs supplied to NicheCompass:\n",
    "    - OmniPath (Ligand-Receptor GPs)\n",
    "    - NicheNet (Combined Interaction GPs)\n",
    "    - MEBOCOST (Metabolite-Sensor GPs)\n",
    "- We will artificially upregulate some GPs in the simulation data based on a SRTsim simulation backbone (through distributional gene expression increments; see './data_simulation.R' script)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca77a2d1-633a-4fe4-b24d-c7a6cc590b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve OmniPath GPs (source: ligand genes; target: receptor genes)\n",
    "omnipath_gp_dict = extract_gp_dict_from_omnipath_lr_interactions(\n",
    "    species=species,\n",
    "    load_from_disk=True,\n",
    "    save_to_disk=False,\n",
    "    lr_network_file_path=omnipath_lr_network_file_path,\n",
    "    gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "    plot_gp_gene_count_distributions=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f843a28-27d1-4916-911e-fc409586164e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve NicheNet GPs (source: ligand genes; target: receptor genes, target genes)\n",
    "nichenet_gp_dict = extract_gp_dict_from_nichenet_lrt_interactions(\n",
    "    species=species,\n",
    "    version=\"v2\",\n",
    "    load_from_disk=True,\n",
    "    save_to_disk=False,\n",
    "    lr_network_file_path=nichenet_lr_network_file_path,\n",
    "    ligand_target_matrix_file_path=nichenet_ligand_target_matrix_file_path,\n",
    "    gene_orthologs_mapping_file_path=gene_orthologs_mapping_file_path,\n",
    "    plot_gp_gene_count_distributions=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "191beb76-bf9c-41fc-96f9-d7f7fa7b06f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Retrieve MEBOCOST GPs (source: enzyme genes; target: sensor genes)\n",
    "mebocost_gp_dict = extract_gp_dict_from_mebocost_ms_interactions(\n",
    "    dir_path=mebocost_enzyme_sensor_interactions_folder_path,\n",
    "    species=species,\n",
    "    plot_gp_gene_count_distributions=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72f56e26-71e0-451e-a452-cbdea6bbbe68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter and combine GPs\n",
    "gp_dicts = [omnipath_gp_dict, nichenet_gp_dict, mebocost_gp_dict]\n",
    "combined_gp_dict = filter_and_combine_gp_dict_gps_v2(\n",
    "    gp_dicts,\n",
    "    verbose=True)\n",
    "\n",
    "print(f\"Number of gene programs after filtering and combining: \"\n",
    "      f\"{len(combined_gp_dict)}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5da730a9-c1b5-49f5-b575-8f305424e810",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create region cell type distribution for simulation (has to be the same as in './data_simulation.R' script)\n",
    "if tissue_type == 1:\n",
    "    region_celltype_arr = np.full((8, 8), 0.1)\n",
    "    np.fill_diagonal(region_celltype_arr, 0.7)\n",
    "    region_celltype_arr[0, 4:8] = 0.\n",
    "    region_celltype_arr[1, 4:8] = 0.\n",
    "    region_celltype_arr[2, 2] = 0.5\n",
    "    region_celltype_arr[2, 6:8] = 0.\n",
    "    region_celltype_arr[3, 3] = 0.5\n",
    "    region_celltype_arr[3, 6:8] = 0.\n",
    "    region_celltype_arr[4, 0:4] = 0.\n",
    "    region_celltype_arr[4, 4] = 1.\n",
    "    region_celltype_arr[4, 5] = 0.\n",
    "    region_celltype_arr[4, 6] = 0.\n",
    "    region_celltype_arr[4, 7] = 0.\n",
    "    region_celltype_arr[5, 0:5] = 0.\n",
    "    region_celltype_arr[5, 5] = 1.\n",
    "    region_celltype_arr[5, 6] = 0.\n",
    "    region_celltype_arr[5, 7] = 0.\n",
    "    region_celltype_arr[6, :] = 0.125\n",
    "    region_celltype_arr[7, 0:4] = 0.\n",
    "    region_celltype_arr[7, 4:8] = 0.25\n",
    "    print(region_celltype_arr)\n",
    "elif tissue_type == 2:\n",
    "    region_celltype_arr = np.full((8, 8), 0.)\n",
    "    region_celltype_arr[0, 0] = 0.4\n",
    "    region_celltype_arr[0, 1] = 0.3\n",
    "    region_celltype_arr[0, 2] = 0.2\n",
    "    region_celltype_arr[0, 3] = 0.1\n",
    "    region_celltype_arr[1, 0] = 0.1\n",
    "    region_celltype_arr[1, 1] = 0.2\n",
    "    region_celltype_arr[1, 2] = 0.3\n",
    "    region_celltype_arr[1, 3] = 0.4\n",
    "    region_celltype_arr[2, 3] = 0.5\n",
    "    region_celltype_arr[2, 4] = 0.25\n",
    "    region_celltype_arr[2, 5] = 0.25\n",
    "    region_celltype_arr[3, 3] = 0.25\n",
    "    region_celltype_arr[3, 4] = 0.5\n",
    "    region_celltype_arr[3, 5] = 0.25\n",
    "    region_celltype_arr[4, 3] = 0.25\n",
    "    region_celltype_arr[4, 4] = 0.25\n",
    "    region_celltype_arr[4, 5] = 0.5\n",
    "    region_celltype_arr[5, 0] = 0.4\n",
    "    region_celltype_arr[5, 4] = 0.3\n",
    "    region_celltype_arr[5, 5] = 0.2\n",
    "    region_celltype_arr[5, 6] = 0.1\n",
    "    region_celltype_arr[6, 0] = 0.4\n",
    "    region_celltype_arr[6, 5] = 0.3\n",
    "    region_celltype_arr[6, 6] = 0.2\n",
    "    region_celltype_arr[6, 7] = 0.1\n",
    "    region_celltype_arr[7, 0] = 0.4\n",
    "    region_celltype_arr[7, 1:7] = 0.1\n",
    "    print(region_celltype_arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44a908e6-752d-41a7-97db-702f5d8479d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Simulate prior knowledge CCIs ###\n",
    "np.random.seed(42)\n",
    "\n",
    "combined_gp_df = pd.DataFrame(combined_gp_dict.values(), index=combined_gp_dict.keys())\n",
    "print(len(combined_gp_df)) # 1582\n",
    "\n",
    "combined_gp_df.reset_index(inplace=True)\n",
    "combined_gp_df[\"gp\"] = combined_gp_df[\"index\"]\n",
    "combined_gp_df.drop(\"index\", axis=1, inplace=True)\n",
    "\n",
    "# Get unique number of genes in combined gps\n",
    "combined_gp_source_genes = combined_gp_df[\"sources\"]\n",
    "combined_gp_target_genes = combined_gp_df[\"targets\"]\n",
    "unique_target_genes = set([gene for genes in combined_gp_target_genes for gene in genes])\n",
    "print(len(unique_target_genes)) # 12756\n",
    "unique_source_genes = set([gene for genes in combined_gp_source_genes for gene in genes])\n",
    "print(len(unique_source_genes)) # 2369\n",
    "unique_all_genes = unique_target_genes | unique_source_genes\n",
    "print(len(unique_all_genes)) # 13209\n",
    "\n",
    "# Sample region\n",
    "sampled_idx = np.random.choice([n for n in range(1,9)], size=len(combined_gp_df))\n",
    "\n",
    "# Sample first cell type according to region cell type distributions\n",
    "sampled_idx_2 = []\n",
    "for idx in sampled_idx:\n",
    "    idx_2 = np.random.choice([n for n in range(1,9)],\n",
    "                             p=region_celltype_arr[idx-1])\n",
    "    sampled_idx_2.append(idx_2)\n",
    "sampled_idx_2 = np.array(sampled_idx_2)\n",
    "\n",
    "# Sample second cell type according to region cell type distributions\n",
    "sampled_idx_3 = []\n",
    "for idx in sampled_idx:\n",
    "    idx_3 = np.random.choice([n for n in range(1,9)],\n",
    "                             p=region_celltype_arr[idx-1])\n",
    "    sampled_idx_3.append(idx_3)\n",
    "sampled_idx_3 = np.array(sampled_idx_3)\n",
    "\n",
    "# Sample max 10 source and target genes\n",
    "np.random.seed(41)\n",
    "combined_gp_df[\"regionA\"] = [\"Region\" + str(idx) for idx in sampled_idx]\n",
    "combined_gp_df[\"celltypeA\"] = [\"Celltype\" + str(idx) for idx in sampled_idx_2]\n",
    "combined_gp_df[\"celltypeB\"] = [\"Celltype\" + str(idx) for idx in sampled_idx_3]\n",
    "combined_gp_df[\"sources\"] = combined_gp_df[\"sources\"].apply(lambda x: np.random.choice(sorted(x), size=min(10, len(x)), replace=False)) # only keep max 10 source genes\n",
    "combined_gp_df[\"targets\"] = combined_gp_df[\"targets\"].apply(lambda x: np.random.choice(sorted(x), size=min(10, len(x)), replace=False)) # only keep max 10 target genes\n",
    "combined_gp_df[\"sources\"] = combined_gp_df[\"sources\"].apply(lambda x: \",\".join(map(str, x)))\n",
    "combined_gp_df[\"targets\"] = combined_gp_df[\"targets\"].apply(lambda x: \",\".join(map(str, x)))\n",
    "combined_gp_df.drop([\"sources_categories\", \"targets_categories\"], axis=1, inplace=True)\n",
    "\n",
    "### Simulate de novo CCIs such that total number of GPs is matched between prior and de novo GPs ###\n",
    "np.random.seed(40)\n",
    "\n",
    "# Get all protein coding genes from Ensembl\n",
    "ensembl = pyensembl.EnsemblRelease(111, species=species)\n",
    "ensembl.download()\n",
    "ensembl.index()\n",
    "all_genes = ensembl.genes()\n",
    "gene_names = [gene.gene_name for gene in all_genes if gene.gene_name]\n",
    "protein_coding_genes = [gene for gene in all_genes if gene.biotype == 'protein_coding']\n",
    "protein_coding_genes = set([gene.gene_name for gene in protein_coding_genes if gene.gene_name])\n",
    "\n",
    "# Get de novo genes (protein coding genes not in prior knowledge)\n",
    "denovo_genes = protein_coding_genes - unique_all_genes\n",
    "\n",
    "n_denovo_gps = len(combined_gp_df)\n",
    "de_novo_gp_df = pd.DataFrame({\n",
    "    \"sources\": [np.random.choice(sorted(denovo_genes), 1) for i in range(n_denovo_gps)],\n",
    "    \"targets\": [np.random.choice(sorted(denovo_genes), 10, replace=False) for i in range(n_denovo_gps)]})\n",
    "de_novo_gp_df[\"sources\"] = de_novo_gp_df[\"sources\"].apply(lambda x: \",\".join(map(str, x)))\n",
    "de_novo_gp_df[\"targets\"] = de_novo_gp_df[\"targets\"].apply(lambda x: \",\".join(map(str, x)))\n",
    "de_novo_gp_df[\"gp\"] = de_novo_gp_df[\"sources\"].apply(lambda x: f\"Add-On_{x}_GP\")\n",
    "\n",
    "# Sample region\n",
    "sampled_idx_4 = np.random.choice([n for n in range(1,9)], size=len(combined_gp_df))\n",
    "\n",
    "# Sample first cell type according to region cell type distributions\n",
    "sampled_idx_5 = []\n",
    "for idx in sampled_idx_4:\n",
    "    idx_5 = np.random.choice([n for n in range(1,9)],\n",
    "                             p=region_celltype_arr[idx-1])\n",
    "    sampled_idx_5.append(idx_5)\n",
    "sampled_idx_5 = np.array(sampled_idx_5)\n",
    "\n",
    "# Sample second cell type according to region cell type distributions\n",
    "sampled_idx_6 = []\n",
    "for idx in sampled_idx_4:\n",
    "    idx_6 = np.random.choice([n for n in range(1,9)],\n",
    "                             p=region_celltype_arr[idx-1])\n",
    "    sampled_idx_6.append(idx_6)\n",
    "sampled_idx_6 = np.array(sampled_idx_6)\n",
    "\n",
    "de_novo_gp_df[\"regionA\"] = [\"Region\" + str(idx) for idx in sampled_idx_4]\n",
    "de_novo_gp_df[\"celltypeA\"] = [\"Celltype\" + str(idx) for idx in sampled_idx_5]\n",
    "de_novo_gp_df[\"celltypeB\"] = [\"Celltype\" + str(idx) for idx in sampled_idx_6]\n",
    "\n",
    "# Combine prior knowledge and de novo CCIs\n",
    "all_gp_df = pd.concat([combined_gp_df, de_novo_gp_df]).reset_index(drop=True)\n",
    "print(len(all_gp_df)) # 3164\n",
    "\n",
    "# Get unique genes for each category for full gene panel simulation\n",
    "all_gp_source_genes = all_gp_df[\"sources\"].apply(lambda x: x.split(\",\")).values\n",
    "all_gp_target_genes = all_gp_df[\"targets\"].apply(lambda x: x.split(\",\")).values\n",
    "unique_all_gp_source_genes = set([gene for genes in all_gp_source_genes for gene in genes])\n",
    "unique_all_gp_target_genes = set([gene for genes in all_gp_target_genes for gene in genes])\n",
    "unique_all_gp_all_genes = unique_all_gp_source_genes | unique_all_gp_target_genes\n",
    "print(len(unique_all_gp_all_genes)) # 12086\n",
    "prior_source_genes = all_gp_df[~all_gp_df[\"gp\"].str.contains(\"Add-On\")][\"sources\"].apply(lambda x: x.split(\",\")).values\n",
    "prior_target_genes = all_gp_df[~all_gp_df[\"gp\"].str.contains(\"Add-On\")][\"targets\"].apply(lambda x: x.split(\",\")).values\n",
    "denovo_source_genes = all_gp_df[all_gp_df[\"gp\"].str.contains(\"Add-On\")][\"sources\"].apply(lambda x: x.split(\",\")).values\n",
    "denovo_target_genes = all_gp_df[all_gp_df[\"gp\"].str.contains(\"Add-On\")][\"targets\"].apply(lambda x: x.split(\",\")).values\n",
    "unique_prior_source_genes = set([gene for genes in prior_source_genes for gene in genes])\n",
    "unique_prior_target_genes = set([gene for genes in prior_target_genes for gene in genes])\n",
    "unique_denovo_source_genes = set([gene for genes in denovo_source_genes for gene in genes])\n",
    "unique_denovo_target_genes = set([gene for genes in denovo_target_genes for gene in genes])\n",
    "\n",
    "np.random.seed(42)\n",
    "\n",
    "# Filter to keep only top 3 genes per GP and 10% of all GPs for limited gene panel simulation\n",
    "all_gp_df_filtered = all_gp_df.copy()\n",
    "all_gp_df_filtered[\"sources\"] = all_gp_df_filtered[\"sources\"].apply(\n",
    "    lambda x: x.split(\",\")[:3]).apply(lambda x: \",\".join(map(str, x)))\n",
    "all_gp_df_filtered[\"targets\"] = all_gp_df_filtered[\"targets\"].apply(\n",
    "    lambda x: x.split(\",\")[:3]).apply(lambda x: \",\".join(map(str, x)))\n",
    "all_gp_df_filtered = all_gp_df_filtered.iloc[np.random.choice(all_gp_df_filtered.index, size=316, replace=False)] # 10% of all GPs\n",
    "\n",
    "# Get unique genes for each category for limited gene panel simulation\n",
    "all_gp_filtered_source_genes = all_gp_df_filtered[\"sources\"].apply(lambda x: x.split(\",\")).values\n",
    "all_gp_filtered_target_genes = all_gp_df_filtered[\"targets\"].apply(lambda x: x.split(\",\")).values\n",
    "unique_all_gp_filtered_target_genes = set([gene for genes in all_gp_filtered_target_genes for gene in genes])\n",
    "unique_all_gp_filtered_source_genes = set([gene for genes in all_gp_filtered_source_genes for gene in genes])\n",
    "unique_all_gp_filtered_all_genes = unique_all_gp_filtered_source_genes | unique_all_gp_filtered_target_genes\n",
    "print(len(unique_all_gp_filtered_all_genes)) # 1105\n",
    "filtered_prior_source_genes = all_gp_df_filtered[~all_gp_df_filtered[\"gp\"].str.contains(\"Add-On\")][\"sources\"].apply(lambda x: x.split(\",\")).values\n",
    "filtered_prior_target_genes = all_gp_df_filtered[~all_gp_df_filtered[\"gp\"].str.contains(\"Add-On\")][\"targets\"].apply(lambda x: x.split(\",\")).values\n",
    "filtered_denovo_source_genes = all_gp_df_filtered[all_gp_df_filtered[\"gp\"].str.contains(\"Add-On\")][\"sources\"].apply(lambda x: x.split(\",\")).values\n",
    "filtered_denovo_target_genes = all_gp_df_filtered[all_gp_df_filtered[\"gp\"].str.contains(\"Add-On\")][\"targets\"].apply(lambda x: x.split(\",\")).values\n",
    "unique_filtered_prior_source_genes = set([gene for genes in filtered_prior_source_genes for gene in genes])\n",
    "unique_filtered_prior_target_genes = set([gene for genes in filtered_prior_target_genes for gene in genes])\n",
    "unique_filtered_denovo_source_genes = set([gene for genes in filtered_denovo_source_genes for gene in genes])\n",
    "unique_filtered_denovo_target_genes = set([gene for genes in filtered_denovo_target_genes for gene in genes])\n",
    "\n",
    "# Create different CCI increment parameters for each CCI\n",
    "np.random.seed(40)\n",
    "\n",
    "for increment_mode, increment_params, increment_probs in zip(\n",
    "    [\"strong\",\n",
    "     #\"weak\"\n",
    "    ], # two modes\n",
    "    [[1, 3, 5, 8],\n",
    "     [1, 2, 3, 4]\n",
    "    ],\n",
    "    [[0.5, 1/6, 1/6, 1/6], # 50% of GPs (~1582 with full gene panel and ~158 with limited gene panel) with increment\n",
    "     [0.7, 1/10, 1/10, 1/10]\n",
    "    ]):\n",
    "    all_gp_df[\"regionA\"] = all_gp_df[\"regionA\"].apply(lambda x: x.replace(\"Niche \", \"Region\"))\n",
    "    all_gp_df[\"celltypeA\"] = all_gp_df[\"celltypeA\"].apply(lambda x: x.replace(\"Cell Type \", \"Celltype\"))\n",
    "    all_gp_df[\"celltypeB\"] = all_gp_df[\"celltypeB\"].apply(lambda x: x.replace(\"Cell Type \", \"Celltype\"))\n",
    "    all_gp_df[\"increment_param\"] = np.random.choice(increment_params, size=len(all_gp_df), p=increment_probs)\n",
    "    all_gp_df.to_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_{increment_mode}increments{'_tissue2' if tissue_type == 2 else ''}.csv\", index=False)\n",
    "\n",
    "    # Save gp stats\n",
    "    all_gp_df[\"regionA\"] = all_gp_df[\"regionA\"].apply(lambda x: x.replace(\"Region\", \"Niche \"))\n",
    "    all_gp_df[\"celltypeA\"] = all_gp_df[\"celltypeA\"].apply(lambda x: x.replace(\"Celltype\", \"Cell Type \"))\n",
    "    all_gp_df[\"celltypeB\"] = all_gp_df[\"celltypeB\"].apply(lambda x: x.replace(\"Celltype\", \"Cell Type \"))\n",
    "    combined_gp_df = all_gp_df[~all_gp_df[\"gp\"].str.contains(\"Add-On\")]\n",
    "    de_novo_gp_df = all_gp_df[all_gp_df[\"gp\"].str.contains(\"Add-On\")]\n",
    "    combined_gp_agg_df = combined_gp_df.groupby([\"regionA\", \"increment_param\"]).agg(set)\n",
    "    de_novo_gp_agg_df = de_novo_gp_df.groupby([\"regionA\", \"increment_param\"]).agg(set)\n",
    "    combined_gp_agg_df.columns = [\"prior_gp_sources\", \"prior_gp_targets\", \"prior_gps\", \"prior_gp_celltypeA\", \"prior_gp_celltypeB\"]\n",
    "    de_novo_gp_agg_df.columns = [\"denovo_gp_sources\", \"denovo_gp_targets\", \"denovo_gps\", \"denovo_gp_celltypeA\", \"denovo_gp_celltypeB\"]\n",
    "    all_gp_agg_df = pd.concat([combined_gp_agg_df, de_novo_gp_agg_df], axis=1)\n",
    "    all_gp_agg_df = all_gp_agg_df.applymap(lambda x: set() if pd.isna(x) else x)\n",
    "    all_gp_agg_df = all_gp_agg_df.applymap(lambda x: ','.join(map(str, x)) if isinstance(x, set) else x)\n",
    "\n",
    "    # Update rows with increment 1 to include all valid rest elements\n",
    "    for element_type, valid_genes in zip([\"prior_gp_sources\",\n",
    "                                          \"prior_gp_targets\",\n",
    "                                          \"denovo_gp_sources\",\n",
    "                                          \"denovo_gp_targets\"],\n",
    "                                         [unique_prior_source_genes,\n",
    "                                          unique_prior_target_genes,\n",
    "                                          unique_denovo_source_genes,\n",
    "                                          unique_denovo_target_genes]):\n",
    "        all_incremented_elements = set()\n",
    "        for niche in all_gp_df[\"regionA\"].unique().tolist():\n",
    "            for increment_param in increment_params[1:]:\n",
    "                all_incremented_elements.update(set(all_gp_agg_df.loc[(niche, increment_param), element_type].split(\",\")))\n",
    "        \n",
    "            all_gp_agg_df.loc[(niche, 1), element_type] = ','.join(map(str, valid_genes - all_incremented_elements))\n",
    "    \n",
    "    all_gp_agg_df.to_csv(\n",
    "        f\"{gp_data_folder_path}/data_simulation/sim_gps_stats_{increment_mode}increments{'_tissue2' if tissue_type == 2 else ''}.csv\")\n",
    "\n",
    "    all_gp_df_filtered[\"regionA\"] = all_gp_df_filtered[\"regionA\"].apply(lambda x: x.replace(\"Niche \", \"Region\"))\n",
    "    all_gp_df_filtered[\"celltypeA\"] = all_gp_df_filtered[\"celltypeA\"].apply(lambda x: x.replace(\"Cell Type \", \"Celltype\"))\n",
    "    all_gp_df_filtered[\"celltypeB\"] = all_gp_df_filtered[\"celltypeB\"].apply(lambda x: x.replace(\"Cell Type \", \"Celltype\"))\n",
    "    all_gp_df_filtered[\"increment_param\"] = np.random.choice(increment_params, size=len(all_gp_df_filtered), p=increment_probs)\n",
    "    all_gp_df_filtered.to_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_{increment_mode}increments{'_tissue2' if tissue_type == 2 else ''}.csv\", index=False)\n",
    "\n",
    "    # Save gp stats\n",
    "    all_gp_df_filtered[\"regionA\"] = all_gp_df_filtered[\"regionA\"].apply(lambda x: x.replace(\"Region\", \"Niche \"))\n",
    "    all_gp_df_filtered[\"celltypeA\"] = all_gp_df_filtered[\"celltypeA\"].apply(lambda x: x.replace(\"Celltype\", \"Cell Type \"))\n",
    "    all_gp_df_filtered[\"celltypeB\"] = all_gp_df_filtered[\"celltypeB\"].apply(lambda x: x.replace(\"Celltype\", \"Cell Type \"))\n",
    "    combined_gp_df_filtered = all_gp_df_filtered[~all_gp_df_filtered[\"gp\"].str.contains(\"Add-On\")]\n",
    "    de_novo_gp_df_filtered = all_gp_df_filtered[all_gp_df_filtered[\"gp\"].str.contains(\"Add-On\")]\n",
    "    combined_gp_agg_df_filtered = combined_gp_df_filtered.groupby([\"regionA\", \"increment_param\"]).agg(set)\n",
    "    de_novo_gp_agg_df_filtered = de_novo_gp_df_filtered.groupby([\"regionA\", \"increment_param\"]).agg(set)\n",
    "    combined_gp_agg_df_filtered.columns = [\"prior_gp_sources\", \"prior_gp_targets\", \"prior_gps\", \"prior_gp_celltypeA\", \"prior_gp_celltypeB\"]\n",
    "    de_novo_gp_agg_df_filtered.columns = [\"denovo_gp_sources\", \"denovo_gp_targets\", \"denovo_gps\", \"denovo_gp_celltypeA\", \"denovo_gp_celltypeB\"]\n",
    "    all_gp_agg_df_filtered = pd.concat([combined_gp_agg_df_filtered, de_novo_gp_agg_df_filtered], axis=1)\n",
    "    all_gp_agg_df_filtered = all_gp_agg_df_filtered.applymap(lambda x: set() if pd.isna(x) else x)\n",
    "    all_gp_agg_df_filtered = all_gp_agg_df_filtered.applymap(lambda x: ','.join(map(str, x)) if isinstance(x, set) else x)\n",
    "\n",
    "    # Update rows with increment 1 to include all valid rest elements\n",
    "    for element_type, valid_genes in zip([\"prior_gp_sources\",\n",
    "                                          \"prior_gp_targets\",\n",
    "                                          \"denovo_gp_sources\",\n",
    "                                          \"denovo_gp_targets\"],\n",
    "                                         [unique_filtered_prior_source_genes,\n",
    "                                          unique_filtered_prior_target_genes,\n",
    "                                          unique_filtered_denovo_source_genes,\n",
    "                                          unique_filtered_denovo_target_genes]):\n",
    "        all_incremented_elements = set()\n",
    "        for niche in all_gp_df_filtered[\"regionA\"].unique().tolist():\n",
    "            for increment_param in increment_params[1:]:\n",
    "                try:\n",
    "                    all_incremented_elements.update(set(all_gp_agg_df_filtered.loc[(niche, increment_param), element_type].split(\",\")))\n",
    "                except:\n",
    "                    print(f\"Skipped increment_param '{increment_param}' for niche '{niche}'.\")\n",
    "        \n",
    "            all_gp_agg_df_filtered.loc[(niche, 1), element_type] = ','.join(map(str, valid_genes - all_incremented_elements))\n",
    "\n",
    "    all_gp_agg_df_filtered.to_csv(\n",
    "        f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_stats_{increment_mode}increments{'_tissue2' if tissue_type == 2 else ''}.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9ae2e4a-f74f-4e96-be05-a9bbc0e2f83a",
   "metadata": {},
   "source": [
    "### 2.2 Simulate Gene Expression Data Based on Simulated CCIs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0919f2a9-2217-47f1-85fa-41c0f0ddef00",
   "metadata": {},
   "source": [
    "Run ```./data_simulation.R``` script with simulated GP-based CCIs as input."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7226435-f1f9-4799-a05f-601822067889",
   "metadata": {},
   "outputs": [],
   "source": [
    "sim_names = [\"sim1_1105genes_10000locs_strongincrements\",\n",
    "             #\"sim2_1105genes_10000locs_strongincrements\",\n",
    "             #\"sim1_12086genes_10000locs_strongincrements\",\n",
    "             #\"sim2_12086genes_10000locs_strongincrements\",\n",
    "            ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04daa90c-fc83-4c9e-a3ec-7b556c5bf79f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Post-process data\n",
    "for sim_name in sim_names:\n",
    "    adata = sc.read_h5ad(f\"{st_data_silver_folder_path}/{sim_name}/{sim_name}.h5ad\")\n",
    "    \n",
    "    adata.obs[\"cell_types\"] = adata.obs[\"cell_type\"].replace({\"Celltype1\": \"Cell Type 1\",\n",
    "                                                              \"Celltype2\": \"Cell Type 2\",\n",
    "                                                              \"Celltype3\": \"Cell Type 3\",\n",
    "                                                              \"Celltype4\": \"Cell Type 4\",\n",
    "                                                              \"Celltype5\": \"Cell Type 5\",\n",
    "                                                              \"Celltype6\": \"Cell Type 6\",\n",
    "                                                              \"Celltype7\": \"Cell Type 7\",\n",
    "                                                              \"Celltype8\": \"Cell Type 8\"})\n",
    "    del(adata.obs[\"cell_type\"])\n",
    "    adata.obs[\"niche_types\"] = adata.obs[\"region_label\"].replace({\"Region1\": \"Niche 1\",\n",
    "                                                                  \"Region2\": \"Niche 2\",\n",
    "                                                                  \"Region3\": \"Niche 3\",\n",
    "                                                                  \"Region4\": \"Niche 4\",\n",
    "                                                                  \"Region5\": \"Niche 5\",\n",
    "                                                                  \"Region6\": \"Niche 6\",\n",
    "                                                                  \"Region7\": \"Niche 7\",\n",
    "                                                                  \"Region8\": \"Niche 8\"})\n",
    "    del(adata.obs[\"region_label\"])\n",
    "    adata.obsm[\"spatial\"] = np.array(adata.obs[[\"x\", \"y\"]] * 0.35) # this is set to achieve an average number of neighbors of around 6 (similar to Nanostring CosMx Human Brain data)\n",
    "    del(adata.obs[\"x\"])\n",
    "    del(adata.obs[\"y\"])\n",
    "    adata.layers[\"counts\"] = adata.X.astype(np.float32)\n",
    "    adata.obs[\"batch\"] = \"batch1\"\n",
    "    \n",
    "    adata.write(f\"{st_data_gold_folder_path}/{sim_name}.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c05ef135-634b-428b-9875-8dc68f4f0879",
   "metadata": {},
   "source": [
    "### 2.3 Explore Simulated Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81dd1c16-fdec-4db8-b654-64aa5b611fc6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "n_sim = \"1\" # 1, 2\n",
    "increment_mode = \"strong\" # \"strong\"\n",
    "n_genes = 1105 # 1105, 12086\n",
    "n_locs = 10000\n",
    "sim_name = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a578008f-a4bb-48f4-8102-fb63ac73c7e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/{sim_name}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65d270bc-963e-4c34-8443-c1b9673d642b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check counts per cell\n",
    "adata.layers[\"counts\"].toarray().sum(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "611a7487-977f-4548-a476-56451aff2d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute PCA and UMAP\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)\n",
    "sc.tl.pca(adata)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c225ed0-f3c6-41dc-9278-4c8783e7c46e",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extended Data Fig. 3a ###\n",
    "# Visualize niche types in tissue\n",
    "sc.pl.spatial(adata=adata,\n",
    "              color=\"niche_types\",                \n",
    "              palette=niche_type_colors,\n",
    "              spot_size=30,\n",
    "              show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_tissue_niche_types.svg\", format=\"svg\")\n",
    "\n",
    "# Visualize niche types in UMAP PCA-reduced gene expression space\n",
    "niche_type_colors = create_new_color_dict(\n",
    "        adata=adata,\n",
    "        color_palette=\"default\",\n",
    "        cat_key=\"niche_types\")\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"niche_types\"],\n",
    "           palette=niche_type_colors,\n",
    "           show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_umap_niche_types.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba30c112-6847-4209-85c5-28fcd30caf48",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extended Data Fig. 3b ###\n",
    "# Visualize cell types in tissue\n",
    "sc.pl.spatial(adata=adata,\n",
    "              color=\"cell_types\",                \n",
    "              palette=cell_type_colors,\n",
    "              spot_size=30,\n",
    "              show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_tissue_cell_types.svg\", format=\"svg\")\n",
    "\n",
    "# Visualize cell types in UMAP PCA-reduced gene expression space\n",
    "cell_type_colors = create_new_color_dict(\n",
    "        adata=adata,\n",
    "        color_palette=\"cell_type_10\",\n",
    "        cat_key=\"cell_types\")\n",
    "\n",
    "sc.pl.umap(adata,\n",
    "           color=[\"cell_types\"],\n",
    "           palette=cell_type_colors,\n",
    "           show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_umap_cell_types.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f077b998-bce9-48ba-8427-30018d9c9f00",
   "metadata": {},
   "outputs": [],
   "source": [
    "gp_df = pd.read_csv(f\"{gp_data_folder_path}/data_simulation/sim_gps_filtered_strongincrements.csv\")\n",
    "gp_df[(gp_df[\"regionA\"] == \"Region2\") & (gp_df[\"increment_param\"] == 8)]\n",
    "gp_df[\"celltypeA\"] = gp_df[\"celltypeA\"].replace({\"Celltype1\": \"Cell Type 1\",\n",
    "                                                 \"Celltype2\": \"Cell Type 2\",\n",
    "                                                 \"Celltype3\": \"Cell Type 3\",\n",
    "                                                 \"Celltype4\": \"Cell Type 4\",\n",
    "                                                 \"Celltype5\": \"Cell Type 5\",\n",
    "                                                 \"Celltype6\": \"Cell Type 6\",\n",
    "                                                 \"Celltype7\": \"Cell Type 7\",\n",
    "                                                 \"Celltype8\": \"Cell Type 8\"})\n",
    "gp_df[\"celltypeB\"] = gp_df[\"celltypeB\"].replace({\"Celltype1\": \"Cell Type 1\",\n",
    "                                                 \"Celltype2\": \"Cell Type 2\",\n",
    "                                                 \"Celltype3\": \"Cell Type 3\",\n",
    "                                                 \"Celltype4\": \"Cell Type 4\",\n",
    "                                                 \"Celltype5\": \"Cell Type 5\",\n",
    "                                                 \"Celltype6\": \"Cell Type 6\",\n",
    "                                                 \"Celltype7\": \"Cell Type 7\",\n",
    "                                                 \"Celltype8\": \"Cell Type 8\"})\n",
    "gp_df[\"regionA\"] = gp_df[\"regionA\"].replace({\"Region1\": \"Niche 1\",\n",
    "                                             \"Region2\": \"Niche 2\",\n",
    "                                             \"Region3\": \"Niche 3\",\n",
    "                                             \"Region4\": \"Niche 4\",\n",
    "                                             \"Region5\": \"Niche 5\",\n",
    "                                             \"Region6\": \"Niche 6\",\n",
    "                                             \"Region7\": \"Niche 7\",\n",
    "                                             \"Region8\": \"Niche 8\"})\n",
    "new_column_names = {\n",
    "    \"regionA\": \"Niche\",\n",
    "    \"celltypeA\": \"Target Cell Type\",\n",
    "    \"celltypeB\": \"Source Cell Type\",\n",
    "    \"sources\": \"Source Genes\",\n",
    "    \"targets\": \"Target Genes\",\n",
    "    \"gp\": \"GP\",\n",
    "    \"increment_param\": \"Increment Factor\",\n",
    "}\n",
    "gp_df.rename(columns=new_column_names, inplace=True)\n",
    "gp_df = gp_df[[\"GP\", \"Niche\", \"Source Genes\", \"Target Genes\", \"Source Cell Type\", \"Target Cell Type\", \"Increment Factor\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5fce17-8276-4d51-a1fd-e207af26dc71",
   "metadata": {},
   "outputs": [],
   "source": [
    "example_gp_df = gp_df[gp_df[\"GP\"] == \"Prl2c2_ligand_receptor_target_gene_GP\"]\n",
    "example_gp_df[\"GP\"] = example_gp_df[\"GP\"].apply(lambda x: x.replace(\"_ligand_receptor_target_gene\", \"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ee63efb-0001-479f-af6a-9ef6c28d7538",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Extended Data Fig. 3c ###\n",
    "transposed_df = example_gp_df.T\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 2))\n",
    "ax.axis('off')\n",
    "ax.axis('tight')\n",
    "table = ax.table(\n",
    "    cellText=transposed_df.values,\n",
    "    #colLabels=transposed_df.columns,\n",
    "    rowLabels=transposed_df.index,\n",
    "    cellLoc='center',\n",
    "    loc='center',\n",
    "    rowColours=[\"darkgrey\"]*len(transposed_df.index)\n",
    ")\n",
    "table.auto_set_font_size(False)\n",
    "table.set_fontsize(14)\n",
    "table.scale(0.25, 1.7)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/example_GP_table.svg\", bbox_inches=\"tight\", dpi=300)\n",
    "plt.show()\n",
    "\n",
    "# Visualize incremented genes in UMAP PCA-reduced gene expression space\n",
    "sc.pl.umap(adata, color=[\"Prl2c2\"], layer=\"counts\", show=False) # niche 6 cell type 6 prior source gene\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_niche2_celltype2_upregulated_prior_source_gene_umap.svg\", format=\"svg\")\n",
    "\n",
    "# Visualize incremented genes in UMAP PCA-reduced gene expression space\n",
    "sc.pl.umap(adata, color=[\"Bcl6\"], layer=\"counts\", show=False) # niche 6 cell type 6 prior target gene\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_niche2_celltype2_upregulated_prior_target_gene_umap.svg\", format=\"svg\")\n",
    "\n",
    "# Visualize niche types in UMAP PCA-reduced gene expression space\n",
    "sc.pl.umap(adata, color=[\"niche_types\"], show=False, groups=\"Niche 2\")\n",
    "\n",
    "### Extended Data Fig. 3c ### \n",
    "sc.pl.spatial(\n",
    "    adata=adata,\n",
    "    layer=\"counts\",\n",
    "    color=\"Prl2c2\",\n",
    "    color_map=\"RdPu\",\n",
    "    spot_size=50,\n",
    "    title=f\"Prl2c2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_niche2_celltype2_upregulated_prior_source_gene_tissue.svg\", format=\"svg\")\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata=adata,\n",
    "    layer=\"counts\",\n",
    "    color=\"Bcl6\",\n",
    "    color_map=\"RdPu\",\n",
    "    spot_size=50,\n",
    "    title=f\"Bcl6\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_niche2_celltype2_upregulated_prior_target_gene_tissue.svg\", format=\"svg\")\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata=adata,\n",
    "    layer=\"counts\",\n",
    "    color=\"Mmp1b\",\n",
    "    color_map=\"RdPu\",\n",
    "    spot_size=50,\n",
    "    title=f\"Mmp1b\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_niche2_celltype2_upregulated_prior_target_gene_2_tissue.svg\", format=\"svg\")\n",
    "\n",
    "sc.pl.spatial(\n",
    "    adata=adata,\n",
    "    layer=\"counts\",\n",
    "    color=\"Ccn2\",\n",
    "    color_map=\"RdPu\",\n",
    "    spot_size=50,\n",
    "    title=f\"Ccn2\",\n",
    "    legend_loc=None,\n",
    "    colorbar_loc=\"bottom\",\n",
    "    show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_niche2_celltype2_upregulated_prior_target_gene_3_tissue.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82cdcadf-d159-4833-9455-47392372b781",
   "metadata": {},
   "source": [
    "### 2.4 Create Spatially Binned Version of Simulated Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7222ed17-6713-4254-b69b-1dfd3156ca05",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "n_sim = \"1\" # 1, 2\n",
    "increment_mode = \"strong\" # \"strong\"\n",
    "n_genes = 1105 # 1105, 12086\n",
    "n_locs = 10000\n",
    "sim_name = f\"sim{n_sim}_{n_genes}genes_{n_locs}locs_{increment_mode}increments\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26d0b830-b100-4c97-9635-3f8e3725df89",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/{sim_name}.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "790dd14d-0730-46d4-a4d4-cc7d5ead94a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check counts per cell\n",
    "adata.layers[\"counts\"].toarray().sum(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88aa9494-5f5e-4677-be02-6055058aa640",
   "metadata": {},
   "outputs": [],
   "source": [
    "diameter = 55\n",
    "\n",
    "# Define grid of circle centers\n",
    "x_coords = adata.obsm[\"spatial\"][:,0] - min(adata.obsm[\"spatial\"][:,0])\n",
    "y_coords = adata.obsm[\"spatial\"][:,1] - min(adata.obsm[\"spatial\"][:,1])\n",
    "x_centers = np.arange(0 + math.ceil(max(x_coords) % 55 / 2), max(x_coords), diameter)\n",
    "y_centers = np.arange(0 + math.ceil(max(y_coords) % 55 / 2), max(y_coords), diameter)\n",
    "x_centers, y_centers = np.meshgrid(x_centers, y_centers)\n",
    "x_centers = x_centers.flatten()\n",
    "y_centers = y_centers.flatten()\n",
    "\n",
    "# Function to find the nearest center for each point\n",
    "def find_nearest_center(x, y, centers_x, centers_y):\n",
    "    distances = np.sqrt((centers_x - x)**2 + (centers_y - y)**2)\n",
    "    return np.argmin(distances)\n",
    "\n",
    "# Assign points to the nearest center/bin\n",
    "bins = []\n",
    "bin_coords = []\n",
    "for x, y in zip(x_coords, y_coords):\n",
    "    nearest_center_idx = find_nearest_center(x, y, x_centers, y_centers)\n",
    "    bins.append(nearest_center_idx)\n",
    "    bin_coords.append((x_centers[nearest_center_idx], y_centers[nearest_center_idx]))\n",
    "    \n",
    "# Create binned adata\n",
    "adata.obs[\"bin\"] = bins\n",
    "adata.obsm[\"spatial\"] = np.array(bin_coords)\n",
    "\n",
    "sq.gr.spatial_neighbors(adata,\n",
    "                        coord_type=\"generic\",\n",
    "                        spatial_key=\"spatial\",\n",
    "                        radius=0,\n",
    "                        set_diag=True)\n",
    "\n",
    "adata.layers[\"counts\"] = adata.obsp[\"spatial_connectivities\"].T @ adata.X\n",
    "\n",
    "first_occurrences = adata.obs.drop_duplicates(subset=\"bin\", keep=\"first\")\n",
    "filtered_indices = first_occurrences.index\n",
    "binned_adata = adata[filtered_indices, :]\n",
    "binned_adata.obs.index = binned_adata.obs.index.str.replace(\"Cell\", \"Spot\")\n",
    "binned_adata.obs = binned_adata.obs[[\"niche_types\", \"batch\", \"bin\"]]\n",
    "binned_adata.layers[\"counts\"] = sp.csr_matrix(binned_adata.layers[\"counts\"].astype(np.float32))\n",
    "binned_adata.X = binned_adata.layers[\"counts\"]\n",
    "\n",
    "del binned_adata.obsp[\"spatial_connectivities\"]\n",
    "del binned_adata.obsp[\"spatial_distances\"]\n",
    "del binned_adata.uns[\"spatial_neighbors\"]\n",
    "\n",
    "binned_adata.write(f\"{st_data_gold_folder_path}/{sim_name}_binned.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b76538ce-f7d2-460b-ab6d-f2abd453a93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "binned_adata = sc.read_h5ad(f\"{st_data_gold_folder_path}/{sim_name}_binned.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9bb48ea-1efe-4382-bfb7-8b7713a5b4e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 14a ###\n",
    "niche_type_colors = create_new_color_dict(\n",
    "        adata=binned_adata,\n",
    "        color_palette=\"default\",\n",
    "        cat_key=\"niche_types\")\n",
    "sc.pl.spatial(adata=binned_adata,\n",
    "              color=\"niche_types\",                \n",
    "              palette=niche_type_colors,\n",
    "              spot_size=50,\n",
    "              show=False)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_binned_tissue_niche_types.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a13f91c2-3b5d-414b-8899-110c1e927741",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Supplementary Fig. 14b ###\n",
    "print(np.mean(adata.obs[\"bin\"].value_counts()))\n",
    "sns.histplot(adata.obs[\"bin\"].value_counts(), bins=20, kde=True)\n",
    "plt.xlabel(\"Cells per Spot\", fontsize=14)\n",
    "plt.ylabel(\"Number of Spots\", fontsize=14)\n",
    "plt.title(\"Spot Cell Count Distribution\", fontsize=20)\n",
    "plt.tick_params(axis='both', which='major', labelsize=14)\n",
    "plt.savefig(f\"{artifacts_folder_path}/data_simulation/{sim_name}_binned_spot_cell_count_hist.svg\", format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1144e2cf-06bc-4af0-8af9-fcb17377e86c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
